{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "satisfied-traffic",
   "metadata": {},
   "source": [
    "# Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "nearby-variation",
   "metadata": {},
   "source": [
    "UCI has a complex google doc they've been using to track their samples so far.\n",
    "\n",
    "https://docs.google.com/spreadsheets/d/13M6-Ry6oXgkx94BHZOGioYPI6F_hWDqjGgcaNu2JNYs/edit#gid=1838362486\n",
    "\n",
    "Also there's a metadata spreadsheet for IGVF at\n",
    "https://docs.google.com/spreadsheets/d/1BLMledzmqOqXnJHzpijgw91IOs-9tSlVeZDG_MtXddk/edit#gid=1284120531"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "increased-bottle",
   "metadata": {},
   "outputs": [],
   "source": [
    "import bz2\n",
    "from collections import Counter, namedtuple\n",
    "import datetime\n",
    "import numpy\n",
    "import os\n",
    "import pandas\n",
    "from pathlib import Path\n",
    "import re\n",
    "from subprocess import run, PIPE\n",
    "import sys\n",
    "import zoneinfo\n",
    "from tqdm import tqdm\n",
    "\n",
    "os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'mousedemo.settings')\n",
    "os.environ[\"DJANGO_ALLOW_ASYNC_UNSAFE\"] = \"true\"\n",
    "\n",
    "import django\n",
    "from django.contrib.auth import get_user_model\n",
    "from django.db import DEFAULT_DB_ALIAS\n",
    "\n",
    "MOUSEDEMO = str(Path(\"mousedemo\").absolute())\n",
    "if MOUSEDEMO not in sys.path:\n",
    "    sys.path.append(MOUSEDEMO)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "divided-checkout",
   "metadata": {},
   "source": [
    "# Pre-initialization setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "orange-optics",
   "metadata": {},
   "source": [
    "Backup the old ones and make a new database file before we initialize the database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "effective-palace",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a new database\n",
    "result = run([\"python3\", \"manage.py\", \"check\"], capture_output=True)\n",
    "\n",
    "if len(result.stderr) > 0:\n",
    "    print(result.stderr.decode(\"utf-8\"))\n",
    "    \n",
    "result.check_returncode()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "documented-cross",
   "metadata": {},
   "outputs": [],
   "source": [
    "initial = Path(\"igvf_mice/migrations/0001_initial.py\")\n",
    "if initial.exists():\n",
    "    initial.unlink()\n",
    "    \n",
    "result = run([\"python3\", \"manage.py\", \"makemigrations\", \"igvf_mice\"], capture_output=True)\n",
    "\n",
    "if len(result.stderr) > 0:\n",
    "    print(result.stderr.decode(\"utf-8\"))\n",
    "    \n",
    "result.check_returncode()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "magnetic-soccer",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "renaming db3.sqlite3 to db4.sqlite3\n",
      "renaming db2.sqlite3 to db3.sqlite3\n",
      "renaming db1.sqlite3 to db2.sqlite3\n",
      "renaming db.sqlite3 to db1.sqlite3\n"
     ]
    }
   ],
   "source": [
    "# Backup several old sqlite database files\n",
    "\n",
    "db_name = Path(\"db.sqlite3\")\n",
    "\n",
    "backup_names = []\n",
    "for i in range(0, 4):\n",
    "    backup_names.append(Path(\"db{}.sqlite3\".format(i+1)))\n",
    "\n",
    "if backup_names[-1].exists():\n",
    "    backup_names[-1].unlink()\n",
    "\n",
    "if db_name.exists():    \n",
    "    for i in reversed(range(0, 3)):\n",
    "        if backup_names[i].exists():\n",
    "            print(\"renaming {} to {}\".format(backup_names[i], backup_names[i+1]))\n",
    "            backup_names[i].rename(backup_names[i+1])\n",
    "\n",
    "    db_name.rename(backup_names[0])\n",
    "    print(\"renaming {} to {}\".format(db_name, backup_names[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "organizational-resort",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a new database\n",
    "result = run([\"python3\", \"manage.py\", \"migrate\"], capture_output=True)\n",
    "\n",
    "if len(result.stderr) > 0:\n",
    "    print(result.stderr.decode(\"utf-8\"))\n",
    "    \n",
    "result.check_returncode()\n",
    "    \n",
    "assert Path(\"db.sqlite3\").exists()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "blessed-shareware",
   "metadata": {},
   "source": [
    "# Setup users"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "failing-advantage",
   "metadata": {},
   "source": [
    "Now that we have a fresh clean database, lets create user accounts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "aboriginal-reform",
   "metadata": {},
   "outputs": [],
   "source": [
    "django.setup()\n",
    "\n",
    "from mousedemo import settings\n",
    "from igvf_mice import models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "arranged-restaurant",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create accounts.\n",
    "with open(\"pw\", \"rt\") as instream:\n",
    "    pw = instream.read().strip()\n",
    "\n",
    "user_model = get_user_model()\n",
    "user_model._default_manager.db_manager(DEFAULT_DB_ALIAS).create_superuser(\n",
    "    username=\"diane\",\n",
    "    password=pw,   \n",
    ")\n",
    "pw = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "realistic-ocean",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "conservative-recycling",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "canadian-slide",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Planning\n",
      "Counting\n",
      "mice - 8 founders\n",
      "Line information\n",
      "schedule - all\n",
      "Sheet40\n",
      "Nuclei isolation Schedule with \n",
      "Samples - 8 founders\n",
      "tissue weights\n",
      "gonads plots\n",
      "Bridge Samples\n",
      "Samples into experiment\n",
      "Gastroc_numbers\n",
      "Gastroc_testing\n",
      "Gastroc_all_reps\n",
      "Bridge samples into experiment\n",
      "Plate setups\n",
      "Failed_samples\n",
      "Experiment\n",
      "Recount for gene capture\n",
      "DCC\n",
      "PBMCs\n",
      "Recount_TA\n",
      "practice mice\n",
      "IGVF002 Pilot Samples\n",
      "IGVF002 Pulverized Samples\n",
      "IGVF002 Plate Setup\n",
      "IGVF002 Metadata\n",
      "Cerebellum testing\n"
     ]
    }
   ],
   "source": [
    "#spreadsheet_name = \"IGVF_Split-seq_20230118.xlsx\"\n",
    "lizs_sheet_name = \"https://woldlab.caltech.edu/nextcloud/index.php/s/eEtjBfDqQFnLpSS/download\"\n",
    "\n",
    "book = pandas.ExcelFile(lizs_sheet_name)\n",
    "for name in book.sheet_names:\n",
    "    print(name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "golden-crack",
   "metadata": {},
   "outputs": [],
   "source": [
    "submitted_book_names = {\n",
    "    \"IGVF_003\": \"https://woldlab.caltech.edu/nextcloud/index.php/s/nBnE6j2RBMkpM9i/download\",\n",
    "    \"IGVF_b01\": \"https://woldlab.caltech.edu/nextcloud/index.php/s/5cJteSWgitN5BDM/download\",\n",
    "}\n",
    "submitted_sheets = {}\n",
    "for book_name in submitted_book_names:\n",
    "    book = pandas.ExcelFile(submitted_book_names[book_name])\n",
    "    \n",
    "    for sheet_name in book.sheet_names:\n",
    "        if sheet_name not in submitted_sheets:\n",
    "            submitted_sheets[sheet_name] = book.parse(sheet_name)\n",
    "        else:\n",
    "            submitted_sheets[sheet_name] = pandas.concat([submitted_sheets[sheet_name], book.parse(sheet_name)])\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "split-payday",
   "metadata": {},
   "source": [
    "Converters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "occupational-dealing",
   "metadata": {},
   "outputs": [],
   "source": [
    "def truncate(model):\n",
    "    table_name = model._meta.db_table\n",
    "    assert \"\\\\\" not in table_name\n",
    "    with django.db.connection.cursor() as cursor:\n",
    "        # I don't know why the sql params didn't work. this is a sql vulnerability waiting to happen\n",
    "        cursor.execute(\"delete from \\\"{}\\\"\".format(table_name))\n",
    "        cursor.execute(\"DELETE FROM SQLITE_SEQUENCE WHERE name=\\\"{}\\\"\".format(table_name))\n",
    "        cursor.fetchone()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "divine-incident",
   "metadata": {},
   "outputs": [],
   "source": [
    "def int_or_none(x):\n",
    "    if x in (\"N/A\", '#DIV/0!', '-'):\n",
    "        return None\n",
    "    elif pandas.isnull(x):\n",
    "        return None\n",
    "    else:\n",
    "        return int(x)\n",
    "    \n",
    "def int_or_0(x):\n",
    "    if pandas.isnull(x):\n",
    "        return 0\n",
    "    else:\n",
    "        return int(x)\n",
    "\n",
    "def float_or_none(x):\n",
    "    if pandas.isnull(x):\n",
    "        return None\n",
    "    else:\n",
    "        return float(x)\n",
    "    \n",
    "def str_or_empty(x):\n",
    "    if pandas.isnull(x):\n",
    "        return \"\"\n",
    "    else:\n",
    "        return x\n",
    "\n",
    "def str_or_none(x):\n",
    "    if pandas.isnull(x):\n",
    "        return None\n",
    "    else:\n",
    "        return x\n",
    "    \n",
    "def date_or_none(x):\n",
    "    if pandas.isnull(x):\n",
    "        return None\n",
    "    elif isinstance(x, datetime.datetime):\n",
    "        return x.date()\n",
    "    elif x in (\"-\"):\n",
    "        return None\n",
    "    else:\n",
    "        return x\n",
    "\n",
    "def datetime_or_none(x):\n",
    "    if pandas.isnull(x):\n",
    "        return None\n",
    "    else:\n",
    "        return x\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "noble-personal",
   "metadata": {},
   "source": [
    "# initialize general use tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "linear-civilian",
   "metadata": {},
   "outputs": [],
   "source": [
    "accession_namespaces = [\n",
    "    {\"name\": \"ENCODE\", \"homepage\": \"https://www.encodeproject.org\", \"accession_prefix\": \"encode\"},\n",
    "    {\"name\": \"ENCODE test\", \"homepage\": \"https://test.encodedcc.org\", \"accession_prefix\": \"encodetst\"},\n",
    "    {\"name\": \"IGVF\", \"homepage\": \"https://data.igvf.org\", \"accession_prefix\": \"igvf\"},\n",
    "    {\"name\": \"IGVF test\", \"homepage\": \"https://sandbox.igvf.org\", \"accession_prefix\": \"igvftst\"},\n",
    "]\n",
    "\n",
    "if models.AccessionNamespace.objects.count() > 0:\n",
    "    truncate(models.AccessionNamespace)\n",
    "\n",
    "for row in accession_namespaces:\n",
    "    record = models.AccessionNamespace(\n",
    "        name=row[\"name\"],\n",
    "        homepage=row[\"homepage\"],\n",
    "        accession_prefix=row[\"accession_prefix\"],\n",
    "    )\n",
    "    record.save()\n",
    "\n",
    "igvf_namespace = models.AccessionNamespace.objects.get(name=\"IGVF\")\n",
    "igvf_test_namespace = models.AccessionNamespace.objects.get(name=\"IGVF test\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "psychological-march",
   "metadata": {},
   "outputs": [],
   "source": [
    "source = [\n",
    "    {\n",
    "        \"name\": \"The Jackson Laboratory\", \n",
    "        \"homepage\": \"http://www.jax.org/index.html\", \n",
    "        \"igvf_id\": \"/sources/jackson-labs/\",\n",
    "    },\n",
    "    {\n",
    "        \"name\": \"Parse Bioscience\", \n",
    "        \"homepage\": \"https://www.parsebiosciences.com/\",\n",
    "        \"igvf_id\": \"/sources/parse-biosciences/\",\n",
    "    },\n",
    "    {\n",
    "        \"name\": \"Illumina\", \n",
    "        \"homepage\": \"https://www.illumina.com\", \n",
    "        \"igvf_id\": \"/sources/illumina/\",\n",
    "    },\n",
    "]\n",
    "\n",
    "if models.Source.objects.count() > 0:\n",
    "    truncate(models.Source)\n",
    "\n",
    "for row in source:\n",
    "    record = models.Source(\n",
    "        name=row[\"name\"],\n",
    "        homepage=row[\"homepage\"],\n",
    "        igvf_id=row[\"igvf_id\"])\n",
    "    record.save()\n",
    "\n",
    "jax_source = models.Source.objects.get(name=\"The Jackson Laboratory\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "metallic-spider",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Should this include machine model?\n",
    "illumina_source = models.Source.objects.get(name=\"Illumina\")\n",
    "platforms = [\n",
    "    {\"display_name\": \"Nextseq 2000\", \"name\": \"nextseq2000\", \"family\": \"illumina\", \"igvf_id\": \"/platform-terms/EFO_0010963/\", \"source\": illumina_source},\n",
    "    {\"display_name\": \"Novaseq 6000\", \"name\": \"novaseq6000\", \"family\": \"illumina\", \"igvf_id\": \"/platform-terms/EFO_0008637/\", \"source\": illumina_source},\n",
    "    {\"display_name\": \"Oxford Nanopore\", \"name\": \"nanopore\", \"family\": \"nanopore\",},\n",
    "    {\"display_name\": \"Pac Bio\", \"name\": \"pacbio\", \"family\": \"pacbio\",},\n",
    "]\n",
    "\n",
    "for row in platforms:\n",
    "    record = models.Platform(\n",
    "        name=row[\"name\"],\n",
    "        display_name=row[\"display_name\"],\n",
    "        family=row[\"family\"],\n",
    "        igvf_id=row.get(\"igvf_id\"),\n",
    "        source=row.get(\"source\"),\n",
    "    )\n",
    "    record.save()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "manufactured-account",
   "metadata": {},
   "source": [
    "# Library kit type & version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "innocent-copying",
   "metadata": {},
   "outputs": [],
   "source": [
    "library_construction_kit = [\n",
    "    {\"name\": \"WT Mega\", \"version\": \"v2\", \"source\": \"Parse Bioscience\"},\n",
    "]\n",
    "\n",
    "if models.LibraryConstructionKit.objects.count() > 0:\n",
    "    truncate(models.LibraryConstructionKit)\n",
    "\n",
    "for row in library_construction_kit:\n",
    "    try:\n",
    "        source = models.Source.objects.get(name=row[\"source\"])\n",
    "        \n",
    "        record = models.LibraryConstructionKit(\n",
    "            name=row[\"name\"],\n",
    "            version=row[\"version\"],\n",
    "            source=source,\n",
    "        )\n",
    "        record.save()        \n",
    "    except models.Source.DoesNotExist:\n",
    "        print(f\"Couldn't find {row['source']}\")\n",
    "        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "pressed-payday",
   "metadata": {},
   "source": [
    "# Load LibraryBarcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "convenient-syndication",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<LibraryConstructionKit: WT Mega v2>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "models.LibraryConstructionKit.objects.get(name=\"WT Mega\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "approved-hours",
   "metadata": {},
   "outputs": [],
   "source": [
    "if models.LibraryBarcode.objects.count() > 0:\n",
    "    truncate(models.LibraryBarcode)\n",
    "\n",
    "\n",
    "# Load WT Mega v2 first barcode\n",
    "bc1 = pandas.read_csv(\"bc_data_n192_v4.csv\")\n",
    "\n",
    "kit = models.LibraryConstructionKit.objects.get(name=\"WT Mega\", version=\"v2\")\n",
    "\n",
    "for i, row in bc1.iterrows():\n",
    "    record = models.LibraryBarcode(\n",
    "        kit=kit,\n",
    "        name=row[\"uid\"],\n",
    "        code=row[\"well\"],\n",
    "        sequence=row[\"sequence\"],\n",
    "        barcode_type=row[\"type\"], # TODO what do the codes mean again T,R?\n",
    "    )\n",
    "    record.save()\n",
    "\n",
    "subpool = [\n",
    "    (\"1\", \"CAGATC\"),\n",
    "    (\"2\", \"ACTTGA\"),\n",
    "    (\"3\", \"GATCAG\"),\n",
    "    (\"4\", \"TAGCTT\"),\n",
    "    (\"5\", \"ATGTCA\"),\n",
    "    (\"6\", \"CTTGTA\"),\n",
    "    (\"7\", \"AGTCAA\"),\n",
    "    (\"8\", \"AGTTCC\"),\n",
    "    (\"9\", \"GAGTGG\"),\n",
    "    (\"10\", \"CCGTCC\"),\n",
    "    (\"11\", \"GTAGAG\"),\n",
    "    (\"12\", \"GTCCGC\"),\n",
    "    (\"13\", \"GTGAAA\"),\n",
    "    (\"14\", \"GTGGCC\"),\n",
    "    (\"15\", \"GTTTCG\"),\n",
    "    (\"16\", \"CGTACG\"),\n",
    "]\n",
    "subpool = pandas.DataFrame(subpool, columns=[\"code\",\"sequence\"])\n",
    "\n",
    "for i, row in subpool.iterrows():\n",
    "    record = models.LibraryBarcode(\n",
    "        kit=kit,\n",
    "        code=row[\"code\"],\n",
    "        sequence=row[\"sequence\"],\n",
    "    )\n",
    "    record.save()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "hungry-guidance",
   "metadata": {},
   "source": [
    "# Initialize TissueOntology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "dimensional-paintball",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Samples - 8 founders (1621, 1) 22\n",
      "Bridge Samples (144, 1) 26\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Counter({'Hypothalamus/Pituitary': 81,\n",
       "         'Cerebellum': 81,\n",
       "         'Cortex/Hippocampus left': 81,\n",
       "         'Cortex/Hippocampus right': 81,\n",
       "         'Liver': 81,\n",
       "         'Heart': 81,\n",
       "         'Lung': 81,\n",
       "         'Adrenal': 81,\n",
       "         'Kidney': 81,\n",
       "         'Gonads -1 Ovary': 41,\n",
       "         'Gonads -2 Oviduct': 41,\n",
       "         'Perigonadal fat': 81,\n",
       "         'Brown fat': 81,\n",
       "         'Soleus': 81,\n",
       "         'Plantaris': 81,\n",
       "         'Gastrocnemius': 82,\n",
       "         'TA': 81,\n",
       "         'EDL': 81,\n",
       "         'Tail': 105,\n",
       "         'PBMC - WBC': 105,\n",
       "         'Gonads -1 Testis': 40,\n",
       "         'Gonads -2 Epididymis': 40,\n",
       "         'Left cortex': 24,\n",
       "         'Right cortex': 24,\n",
       "         'Left hippocampus': 24,\n",
       "         'Right hippocampus': 24})"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seen_tissues = Counter()\n",
    "for sample_sheet_name in [\"Samples - 8 founders\", \"Bridge Samples\"]:\n",
    "    tissue_sheet = pandas.read_excel(\n",
    "        lizs_sheet_name, \n",
    "        sample_sheet_name,\n",
    "        usecols=[\"Tissue\"]).dropna(axis=0, how='all')\n",
    "\n",
    "    for i, row in tissue_sheet.iterrows():\n",
    "        if not pandas.isnull(row[\"Tissue\"]):\n",
    "            seen_tissues[row[\"Tissue\"]] += 1\n",
    "    print(sample_sheet_name, tissue_sheet.shape, len(seen_tissues))\n",
    "\n",
    "seen_tissues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "responsible-aberdeen",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "set()\n"
     ]
    }
   ],
   "source": [
    "tissue_dissection_to_ontology_map = {\n",
    "    'Hypothalamus/Pituitary': [(\"UBERON:0001898\",\"hypothalamus\"), (\"UBERON:0000007\",\"pituitary gland\")],\n",
    "    'Cerebellum': [(\"UBERON:0002037\",\"cerebellum\")],\n",
    "    # Hippocampus might be:\n",
    "    #   Hippocampal formation UBERON:0002421 https://www.ebi.ac.uk/ols/ontologies/uberon/terms?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FUBERON_0002421\n",
    "    #   Layer of hippocampus UBERON:0002305 https://www.ebi.ac.uk/ols/ontologies/uberon/terms?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FUBERON_0002305\n",
    "    'Cortex/Hippocampus left': [(\"NTR:0000646\",\"left cerebral cortex\"), (\"NTR:0000750\", \"Hippocampal formation left\")],\n",
    "    'Cortex/Hippocampus right': [(\"NTR:0000647\",\"right cerebral cortex\"), (\"NTR:0000751\", \"Hippocampal formation right\")],\n",
    "    'Liver': [(\"UBERON:0002107\", \"liver\")],\n",
    "    'Heart': [(\"UBERON:0000948\",\"heart\")],\n",
    "    'Lung': [(\"UBERON:0002048\",\"lung\")],\n",
    "    'Adrenal': [(\"UBERON:0002369\",\"adrenal gland\")],\n",
    "    'Kidney': [(\"UBERON:0002113\",\"kidney\")],\n",
    "    'Gonads -1 Ovary': [(\"UBERON:0000992\",\"ovary\")],\n",
    "    'Gonads -2 Oviduct': [(\"UBERON:0000993\",\"oviduct\")],\n",
    "    'Perigonadal fat': [(\"UBERON:0003428\", \"gonadal fat pad\")], #\"is closer\" to dissection\n",
    "    'Brown fat': [(\"UBERON:0001348\",\"brown adipose tissue\")], \n",
    "    'Soleus': [(\"UBERON:0001389\",\"soleus muscle\")],\n",
    "    'Plantaris': [(\"UBERON:0011905\",\"plantaris\")],\n",
    "    'Gastrocnemius': [(\"UBERON:0001388\",\"gastrocnemius\")],\n",
    "    'TA': [(\"UBERON:0001385\",\"tibialis anterior\")],\n",
    "    'EDL': [(\"UBERON:0001386\",\"extensor digitorum longus\")],\n",
    "    'Tail': [(\"UBERON:0002415\",\"tail\")],\n",
    "    'PBMC - WBC': [(\"CL:2000001\", \"peripheral blood mononuclear cell\")],\n",
    "    'Gonads -1 Testis': [(\"UBERON:0000473\",\"testis\")],\n",
    "    'Gonads -2 Epididymis': [(\"UBERON:0001301\",\"epididymis\")],\n",
    "    \"Left cortex\": [(\"NTR:0000646\",\"left cerebral cortex\"),],\n",
    "    \"Right cortex\": [(\"NTR:0000647\",\"right cerebral cortex\"),],\n",
    "    \"Left hippocampus\": [(\"NTR:0000750\", \"Hippocampal formation left\")],\n",
    "    \"Right hippocampus\": [(\"NTR:0000751\", \"Hippocampal formation right\")],\n",
    "}\n",
    "\n",
    "print(set(seen_tissues).difference(tissue_dissection_to_ontology_map.keys()))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "green-orange",
   "metadata": {},
   "outputs": [],
   "source": [
    "term_details = pandas.read_csv(\"obo.tsv.bz2\", compression=\"bz2\", sep=\"\\t\", index_col=\"term_id\")\n",
    "\n",
    "if models.OntologyTerm.objects.count() > 0:\n",
    "    truncate(models.OntologyTerm)\n",
    "    \n",
    "for key in tissue_dissection_to_ontology_map:\n",
    "    for term_curie, term_name in tissue_dissection_to_ontology_map[key]:\n",
    "        if term_curie.startswith(\"NTR:\"):\n",
    "            description=None\n",
    "        else:\n",
    "            details = term_details.loc[term_curie]\n",
    "            description = details.description\n",
    "\n",
    "        record = models.OntologyTerm(\n",
    "            curie=term_curie,\n",
    "            name=term_name,\n",
    "            description=description\n",
    "        )\n",
    "        record.save()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "otherwise-runner",
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_urls = {\n",
    "    \"AJ\": \"https://www.jax.org/strain/000646\",\n",
    "    \"B6J\": \"https://www.jax.org/strain/000664\",\n",
    "    \"129S1J\": \"https://www.jax.org/strain/002448\",\n",
    "    \"NODJ\": \"https://www.jax.org/strain/001976\",\n",
    "    \"NZOJ\": \"https://www.jax.org/strain/002105\",\n",
    "    \"CASTJ\": \"https://www.jax.org/strain/000928\",\n",
    "    \"PWKJ\": \"https://www.jax.org/strain/003715\",\n",
    "    \"WSBJ\": \"https://www.jax.org/strain/001145\"\n",
    "}\n",
    "\n",
    "strain_name = {\n",
    "    \"AJ\": \"A/J (AJ)\",\n",
    "    \"B6J\": \"C57BL/6J (B6)\",\n",
    "    \"129S1J\": \"129S1/SvImJ (129)\",\n",
    "    \"NODJ\": \"NOD/ShiLtJ (NOD)\",\n",
    "    \"NZOJ\": \"NZO/HlLtJ (NZO)\",\n",
    "    \"CASTJ\": \"CAST/EiJ (CAST)\",\n",
    "    \"PWKJ\": \"PWK/PhJ (PWK)\",\n",
    "    \"WSBJ\": \"WSB/EiJ (WSB)\",\n",
    "    #\"CAST (M. m. castaneus)\",\n",
    "    #\"WSB (M. m. domesticus)\",\n",
    "    #\"PWK (M. m. musculus)\",\n",
    "}\n",
    "strain_background = {k: strain_name[k].split(\" \")[0] for k in strain_name}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "committed-dollar",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Designation</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Note</th>\n",
       "      <th>Jax Catalog No</th>\n",
       "      <th>Sample CODE</th>\n",
       "      <th>Strain notes</th>\n",
       "      <th>url</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>A/J</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>000646</td>\n",
       "      <td>AJ</td>\n",
       "      <td>Yellower adrenal gland.</td>\n",
       "      <td>https://www.jax.org/strain/000646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>B</td>\n",
       "      <td>C57BL/6J</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>000664</td>\n",
       "      <td>B6J</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/000664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>129S1/SvImJ</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>002448</td>\n",
       "      <td>129S1J</td>\n",
       "      <td>Male skin is tougher than female skin. Gallbla...</td>\n",
       "      <td>https://www.jax.org/strain/002448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D</td>\n",
       "      <td>NOD/ShiLtJ</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>001976</td>\n",
       "      <td>NODJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/001976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>NZO/HlLtJ</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>002105</td>\n",
       "      <td>NZOJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/002105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>F</td>\n",
       "      <td>CAST/EiJ</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>000928</td>\n",
       "      <td>CASTJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/000928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>G</td>\n",
       "      <td>PWK/PhJ</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>003715</td>\n",
       "      <td>PWKJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/003715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>H</td>\n",
       "      <td>WSB/EiJ</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>001145</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/001145</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Designation       Strain        Note Jax Catalog No Sample CODE  \\\n",
       "0           A          A/J  CC founder         000646          AJ   \n",
       "1           B     C57BL/6J  CC founder         000664         B6J   \n",
       "2           C  129S1/SvImJ  CC founder         002448      129S1J   \n",
       "3           D   NOD/ShiLtJ  CC founder         001976        NODJ   \n",
       "4           E    NZO/HlLtJ  CC founder         002105        NZOJ   \n",
       "5           F     CAST/EiJ  CC founder         000928       CASTJ   \n",
       "6           G      PWK/PhJ  CC founder         003715        PWKJ   \n",
       "7           H      WSB/EiJ  CC founder         001145        WSBJ   \n",
       "\n",
       "                                        Strain notes  \\\n",
       "0                            Yellower adrenal gland.   \n",
       "1                                                NaN   \n",
       "2  Male skin is tougher than female skin. Gallbla...   \n",
       "3                                                NaN   \n",
       "4                                                NaN   \n",
       "5                                                NaN   \n",
       "6                                                NaN   \n",
       "7                                                NaN   \n",
       "\n",
       "                                 url  \n",
       "0  https://www.jax.org/strain/000646  \n",
       "1  https://www.jax.org/strain/000664  \n",
       "2  https://www.jax.org/strain/002448  \n",
       "3  https://www.jax.org/strain/001976  \n",
       "4  https://www.jax.org/strain/002105  \n",
       "5  https://www.jax.org/strain/000928  \n",
       "6  https://www.jax.org/strain/003715  \n",
       "7  https://www.jax.org/strain/001145  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def int_csv_to_hex(x):\n",
    "    r, g, b = x.split(\",\")\n",
    "    \n",
    "    return \"#{:02x}{:02x}{:02x}\".format(int(r), int(g), int(b))\n",
    "\n",
    "def remove_citation(x):\n",
    "    return x.split(\" \")[0]\n",
    "\n",
    "strains = pandas.read_excel(\n",
    "    lizs_sheet_name, \n",
    "    sheet_name=\"Line information\", \n",
    "    usecols=[\n",
    "        \"Designation\",\n",
    "        \"Strain\",\n",
    "        \"Note\",\n",
    "        \"Jax Catalog No\",\n",
    "        \"Sample CODE\",\n",
    "        \"Strain notes\",\n",
    "    ],\n",
    "    converters={\n",
    "        \"Strain\": remove_citation,\n",
    "        \"Jax Catalog No\": str,\n",
    "        \"Strain notes\": str_or_empty,\n",
    "    }\n",
    ")\n",
    "\n",
    "# Remove the extra lines with the bridge samples for now. I don't want to track thta level of detail\n",
    "bridge = strains[strains[\"Designation\"] == \"Bridge sample\"].first_valid_index() - 1\n",
    "strains = strains.iloc[0:bridge]\n",
    "\n",
    "# Add in Strain URL\n",
    "strains[\"url\"] = strains[\"Sample CODE\"].apply(lambda x: strain_urls[x])\n",
    "strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "future-affiliation",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AJ': 'A/J',\n",
       " 'B6J': 'C57BL/6J',\n",
       " '129S1J': '129S1/SvImJ',\n",
       " 'NODJ': 'NOD/ShiLtJ',\n",
       " 'NZOJ': 'NZO/HlLtJ',\n",
       " 'CASTJ': 'CAST/EiJ',\n",
       " 'PWKJ': 'PWK/PhJ',\n",
       " 'WSBJ': 'WSB/EiJ'}"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "strain_background"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "indian-venture",
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_type_lookup = {\n",
    "    \"CC founder\": models.StrainType.CC_FOUNDER,\n",
    "    \"CC Cross\": models.StrainType.CC_CROSS,\n",
    "}\n",
    "\n",
    "if models.MouseStrain.objects.count() > 0:\n",
    "    truncate(models.MouseStrain)\n",
    "\n",
    "for i, row in strains.iterrows():\n",
    "    record = models.MouseStrain(\n",
    "        name=strain_name[row[\"Sample CODE\"]],\n",
    "        code=strain_background[row[\"Sample CODE\"]],\n",
    "        strain_type=strain_type_lookup[row[\"Note\"]],\n",
    "        jax_catalog_number=row[\"Jax Catalog No\"],\n",
    "        notes=str_or_empty(row[\"Strain notes\"]),\n",
    "        url=row[\"url\"],\n",
    "        source=jax_source,\n",
    "    )\n",
    "    record.save()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "private-century",
   "metadata": {},
   "source": [
    "This is closest to the Mouse table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "seven-technician",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "72\n"
     ]
    }
   ],
   "source": [
    "submitted_mice = {\n",
    "}\n",
    "\n",
    "for i, row in submitted_sheets[\"rodent_donor\"].iterrows():\n",
    "    if not pandas.isnull(row[\"accession\"]):\n",
    "        rodent_identifier = row[\"rodent_identifier\"]\n",
    "        submitted_mice[rodent_identifier] = {\n",
    "            \"namespace\": igvf_test_namespace,\n",
    "            \"name\": row[\"accession\"],\n",
    "            \"uuid\": row[\"uuid\"],\n",
    "            \"url\": \"https://api.sandbox.igvf.org/rodent-donors/{}/\".format(row[\"accession\"]),\n",
    "        }\n",
    "        \n",
    "print(len(submitted_mice))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "deluxe-spencer",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "distinct-failure",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def estrus_cycle(x):\n",
    "    if pandas.isnull(x):\n",
    "        return \"NA\"\n",
    "    else:\n",
    "        return x\n",
    "\n",
    "sex_lookup = {\n",
    "    numpy.nan: models.SexEnum.UNKNOWN,\n",
    "    \"Female\": models.SexEnum.FEMALE,\n",
    "    \"Male\": models.SexEnum.MALE,\n",
    "}\n",
    "\n",
    "estrus_stage = {\n",
    "    \"Unknown\": models.EstrusCycle.UNKNOWN,\n",
    "    \"Anestrus\": models.EstrusCycle.ANESTRUS,\n",
    "    \"Anestrus>Proestrus\": models.EstrusCycle.ANESTRUS_PROESTRUS,\n",
    "    \"Proestrus\": models.EstrusCycle.PROESTRUS,\n",
    "    \"Proestrus>Estrus\": models.EstrusCycle.PROESTRUS_ESTRUS,\n",
    "    \"Estrus\": models.EstrusCycle.ESTRUS,\n",
    "    \"Estrus>Metestrus\": models.EstrusCycle.ESTRUS_METESTRUS,\n",
    "    \"Metestrus\": models.EstrusCycle.METESTRUS,\n",
    "    \"Metestrus>Diestrus\": models.EstrusCycle.METESTRUS_DIESTRUS,\n",
    "    \"Diestrus\": models.EstrusCycle.DIESTRUS,\n",
    "    \"Diestrus>Proestrus\": models.EstrusCycle.DIESTRUS_PROESTRUS,\n",
    "}\n",
    "\n",
    "mice = pandas.read_excel(\n",
    "    lizs_sheet_name, \n",
    "    sheet_name=\"mice - 8 founders\", \n",
    "    usecols=range(0,16),\n",
    "    converters={\n",
    "        \"Estres cycle stage\": estrus_cycle,\n",
    "        \"Dissection ID\": int,\n",
    "        \"RMS number\": int_or_none,\n",
    "        \"Housing number\": int_or_none,\n",
    "    }\n",
    ")\n",
    "\n",
    "if models.Mouse.objects.count() > 0:\n",
    "    truncate(models.Mouse)\n",
    "\n",
    "los_angeles_tz = zoneinfo.ZoneInfo(\"America/Los_Angeles\")\n",
    "\n",
    "\n",
    "\n",
    "failed = False\n",
    "for i, row in mice.iterrows():\n",
    "    if not pandas.isnull(row[\"Mouse Name\"]):\n",
    "        name = row[\"Mouse Name\"]\n",
    "        strain_code = row[\"Strain\"]\n",
    "        \n",
    "        if name == \"056_WSBJ_10F->PWKJ_9F\":\n",
    "            name = \"056_PWKJ_9F\"\n",
    "\n",
    "        try:\n",
    "            strain = models.MouseStrain.objects.get(code=strain_code)\n",
    "        except models.MouseStrain.DoesNotExist as e:\n",
    "            print(f\"Unable to find {row['Strain']} for row {i+2}\")\n",
    "            failed = True\n",
    "            continue\n",
    "\n",
    "        if not name[-1] in (\"M\", \"F\"):\n",
    "            raise ValueError(f\"Unrecognized sex field {mouse_name}\")\n",
    "            \n",
    "        name_fields = name.split(\"_\")\n",
    "        \n",
    "        if len(name_fields) != 3:\n",
    "            raise ValueError(f\"Wrong number of attributes in {name} line {i+1}\")\n",
    "            \n",
    "        if not name_fields[1] in strain_background:\n",
    "            raise ValueError(f\"strain background in {name} not recognized line {i+1}\")\n",
    "        \n",
    "        sample_box = \"\" if pandas.isnull(row[\"Sample box\"]) else row[\"Sample box\"]\n",
    "        \n",
    "        record = models.Mouse(\n",
    "            # should i use liz's disection id?\n",
    "            name=name,\n",
    "            strain=strain,\n",
    "            sex=sex_lookup[row[\"Sex\"]],\n",
    "            weight_g=row[\"Weight (g)\"],\n",
    "            date_of_birth=row[\"DOB\"].date(),\n",
    "            harvest_date=row[\"Harvest Date\"].date(),\n",
    "            operator=row[\"Operator\"],\n",
    "            notes=str_or_empty(row[\"Comments\"]),\n",
    "            sample_box=sample_box,\n",
    "        )\n",
    "        \n",
    "        if not pandas.isnull(row[\"Estres cycle stage\"]):\n",
    "            # Ignore those transitions.\n",
    "            stage = row[\"Estres cycle stage\"]\n",
    "            if stage.endswith(\"?\"):\n",
    "                stage = stage[:-1]\n",
    "            record.estrus_cycle=estrus_stage[stage]\n",
    "       \n",
    "        record.save()\n",
    "        \n",
    "        if name in submitted_mice:\n",
    "            accession = models.Accession(\n",
    "                namespace=submitted_mice[name][\"namespace\"],\n",
    "                name=submitted_mice[name][\"name\"],\n",
    "                uuid=submitted_mice[name][\"uuid\"],\n",
    "                url=submitted_mice[name][\"url\"],\n",
    "            )\n",
    "            accession.save()\n",
    "            record.accession.set([accession])\n",
    "            record.save()\n",
    "        \n",
    "        \n",
    "        \n",
    "assert not failed, \"Check warning messages\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "interracial-batch",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'016_B6J_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO32293389',\n",
       "  'uuid': '5adf9704-60ab-41da-bdb6-ac8a3b156fb3',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO32293389/'},\n",
       " '017_B6J_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO90675659',\n",
       "  'uuid': '171702c9-bf6f-4759-839b-4a7946146837',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO90675659/'},\n",
       " '018_B6J_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO01760234',\n",
       "  'uuid': 'f60da031-f24c-4baf-b001-898724466c91',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO01760234/'},\n",
       " '019_B6J_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO83719175',\n",
       "  'uuid': 'd17f4e1b-ac5b-434a-be0b-fead09279d88',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO83719175/'},\n",
       " '020_B6J_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO52284503',\n",
       "  'uuid': 'b006f822-9a39-424a-a2a7-8e7d2e2f1936',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO52284503/'},\n",
       " '021_B6J_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO46924710',\n",
       "  'uuid': 'cdfbbed6-f466-4bdb-b2bd-531c5e94b13f',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO46924710/'},\n",
       " '024_B6J_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO89093339',\n",
       "  'uuid': '35b4899d-69bc-451e-90b8-df9a78d5969d',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO89093339/'},\n",
       " '025_B6J_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO29771657',\n",
       "  'uuid': '95e1cf00-0d36-4073-824b-995b8e323a01',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO29771657/'},\n",
       " '026_AJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO94208104',\n",
       "  'uuid': 'effe348f-ea31-4c75-a987-6f0279d978a7',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO94208104/'},\n",
       " '028_AJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO91714400',\n",
       "  'uuid': 'ff073641-5d3f-48b9-bf16-cd7d909f0a15',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO91714400/'},\n",
       " '029_AJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO41491740',\n",
       "  'uuid': '194391ed-d287-4c8a-8a86-e61ef864cb76',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO41491740/'},\n",
       " '030_AJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO86946008',\n",
       "  'uuid': '4feda215-69fe-4012-ab58-b708c554cfa3',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO86946008/'},\n",
       " '031_AJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO43152139',\n",
       "  'uuid': '2951ca50-5074-49ab-8126-89f4d3463bae',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO43152139/'},\n",
       " '032_AJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO00761787',\n",
       "  'uuid': '0bdaa947-9039-4c10-8ca9-f6de74c1aa21',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO00761787/'},\n",
       " '033_AJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO80530242',\n",
       "  'uuid': '3df174fc-3b8e-4d45-8baa-678f62780340',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO80530242/'},\n",
       " '035_AJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO38679396',\n",
       "  'uuid': 'f01857e6-c666-40a6-8166-c01b1498689c',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO38679396/'},\n",
       " '036_129S1J_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO24931131',\n",
       "  'uuid': '91a043af-624a-4f97-8247-9d20a8420fdb',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO24931131/'},\n",
       " '037_129S1J_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO12941763',\n",
       "  'uuid': '371c9315-de9a-4c9d-b4ff-5f7f6de8466e',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO12941763/'},\n",
       " '038_129S1J_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO36528785',\n",
       "  'uuid': '49c3b628-4076-499f-9a81-57c497ddd443',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO36528785/'},\n",
       " '040_129S1J_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO69479022',\n",
       "  'uuid': 'fda3b2db-d7ba-40d4-a75c-9ea4a8d08fcb',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO69479022/'},\n",
       " '041_129S1J_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO85973629',\n",
       "  'uuid': '497e3dfe-6c56-408d-99a8-2832a8b4d896',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO85973629/'},\n",
       " '043_129S1J_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO98710265',\n",
       "  'uuid': '58cf6e82-ebc2-4def-9a36-d57d0e7f60ab',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO98710265/'},\n",
       " '044_129S1J_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO50102840',\n",
       "  'uuid': 'c62b902e-2d3b-40e9-8322-ad7499186e32',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO50102840/'},\n",
       " '045_129S1J_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO92303956',\n",
       "  'uuid': '357c4a94-a447-4463-bf32-3801357a809f',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO92303956/'},\n",
       " '046_NZOJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO38202788',\n",
       "  'uuid': '58d2c52b-839c-4afb-97c3-520b839d5869',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO38202788/'},\n",
       " '047_NZOJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO64676241',\n",
       "  'uuid': '2c5e1cb0-3e59-410e-9b73-62107e1942bb',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO64676241/'},\n",
       " '048_NZOJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO45656391',\n",
       "  'uuid': 'f13e9959-7cb1-4360-a32f-29d0627020ff',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO45656391/'},\n",
       " '049_NZOJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO68840786',\n",
       "  'uuid': '9b0cea63-65ec-467d-a627-7c1f6f2c8857',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO68840786/'},\n",
       " '050_NZOJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO62926149',\n",
       "  'uuid': '70337d0a-3c05-4855-9d58-e52b4c7dec1c',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO62926149/'},\n",
       " '051_NZOJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO63535889',\n",
       "  'uuid': '7f3890bb-e4c3-4db8-bcbf-e7041c10f660',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO63535889/'},\n",
       " '052_NZOJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO99451541',\n",
       "  'uuid': 'dec2121d-1c68-4d1a-aaf2-242a3a1a01f8',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO99451541/'},\n",
       " '053_NZOJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO26972281',\n",
       "  'uuid': '668f320d-a2fa-4b2d-9a6f-87683e072299',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO26972281/'},\n",
       " '057_WSBJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO03113260',\n",
       "  'uuid': 'b13a3c91-47e2-49f5-92fe-793c0e0fa66f',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO03113260/'},\n",
       " '058_WSBJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO50709299',\n",
       "  'uuid': '49a338a5-d387-49f3-9958-e2826e34826d',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO50709299/'},\n",
       " '059_WSBJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO33072093',\n",
       "  'uuid': '64c16d83-0103-4a69-bc93-00b5b3457a73',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO33072093/'},\n",
       " '060_WSBJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO94485067',\n",
       "  'uuid': '1f7f110e-19a9-4493-b1f2-eef4228cde7d',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO94485067/'},\n",
       " '061_WSBJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO73673021',\n",
       "  'uuid': '54b4d27b-b60f-4f83-a0b9-bbb2c8420b75',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO73673021/'},\n",
       " '062_WSBJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO50952270',\n",
       "  'uuid': '87b47bfd-73e8-4c0d-97c2-eda6f9f28ee9',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO50952270/'},\n",
       " '063_WSBJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO26939083',\n",
       "  'uuid': 'b23e5d2a-8055-4fa9-abfa-acd8ec7af714',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO26939083/'},\n",
       " '066_NODJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO32028783',\n",
       "  'uuid': '23c6a218-7635-4e42-bf2b-d89c3484e70a',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO32028783/'},\n",
       " '067_NODJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO58645928',\n",
       "  'uuid': '1b030e75-7903-4f0d-8769-4d21aed44666',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO58645928/'},\n",
       " '068_NODJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO45253174',\n",
       "  'uuid': '57aab386-8712-45d3-8900-94adac9db0f0',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO45253174/'},\n",
       " '069_NODJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO01181719',\n",
       "  'uuid': 'c73e6c01-1592-42ea-bca9-28012fec1b9c',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO01181719/'},\n",
       " '070_NODJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO47750034',\n",
       "  'uuid': '07d56886-f206-485e-9e23-f01fac1d673d',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO47750034/'},\n",
       " '071_NODJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO21887615',\n",
       "  'uuid': '84be0923-1042-47f6-a1b9-fca71fad45bb',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO21887615/'},\n",
       " '072_NODJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO77277137',\n",
       "  'uuid': '45169eee-6121-4631-80a1-b2894f8692c6',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO77277137/'},\n",
       " '073_NODJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO07183952',\n",
       "  'uuid': '75f2a287-4562-45c0-912e-7b6e5c250a29',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO07183952/'},\n",
       " '076_PWKJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO07955113',\n",
       "  'uuid': '332cbbd4-0547-4ec7-990e-f87f1e8fe3fa',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO07955113/'},\n",
       " '077_PWKJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO33656155',\n",
       "  'uuid': 'd402772d-610d-42f8-8561-12e4d1a10a3c',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO33656155/'},\n",
       " '078_PWKJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO14034975',\n",
       "  'uuid': '3b900c90-c54f-4020-9037-aabbe560cb72',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO14034975/'},\n",
       " '079_PWKJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO22244365',\n",
       "  'uuid': 'cd1fe40e-848d-4469-a21a-1fc49a6832d9',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO22244365/'},\n",
       " '080_PWKJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO62994979',\n",
       "  'uuid': 'ca4b14cb-d663-4e07-af5c-178cd0785dd5',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO62994979/'},\n",
       " '081_PWKJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO29678506',\n",
       "  'uuid': 'd6127558-a238-4f37-b2e0-3d484c4b4d32',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO29678506/'},\n",
       " '084_PWKJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO51598158',\n",
       "  'uuid': 'fa10521b-df3a-44b3-bc8d-403f244599ee',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO51598158/'},\n",
       " '085_PWKJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO11462810',\n",
       "  'uuid': '91a2183f-6c0f-4b63-8976-9170ba8944ed',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO11462810/'},\n",
       " '086_CASTJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO12540543',\n",
       "  'uuid': '2b45f5f6-c236-414d-b236-bb88cb9943f4',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO12540543/'},\n",
       " '087_CASTJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO32517230',\n",
       "  'uuid': '076f4e93-0639-405d-b325-6dd05e12bc2a',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO32517230/'},\n",
       " '088_CASTJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO23401556',\n",
       "  'uuid': 'd46b8cba-84b7-4b9c-8da2-3bda165d4329',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO23401556/'},\n",
       " '089_CASTJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO89564798',\n",
       "  'uuid': '4ed52a3f-0231-4b4f-be23-7f19857afb50',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO89564798/'},\n",
       " '090_CASTJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO70628089',\n",
       "  'uuid': 'c612b3e4-8b25-44a0-92b4-545a1cc835fb',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO70628089/'},\n",
       " '091_CASTJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO41429670',\n",
       "  'uuid': 'e9e761f8-2b58-41e5-b087-38eb4d129dd0',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO41429670/'},\n",
       " '093_CASTJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO84986960',\n",
       "  'uuid': '9285186e-8ba4-4c03-8eb8-45844157f6c2',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO84986960/'},\n",
       " '094_CASTJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO59020001',\n",
       "  'uuid': '880aa0ba-8af9-4804-a9ec-e848d9cf2042',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO59020001/'},\n",
       " '096_WSBJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO49696857',\n",
       "  'uuid': '2859f493-3683-481e-bbb7-f60f0a26279e',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO49696857/'},\n",
       " '105_CASTJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO49460733',\n",
       "  'uuid': '1d2a58e0-9f0b-45c2-a9cb-a2a66193d616',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO49460733/'},\n",
       " '106_CASTJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO78829713',\n",
       "  'uuid': '5a0fde8b-9d8a-4be5-97c4-507f3b85babb',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO78829713/'},\n",
       " '107_CASTJ_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO92049597',\n",
       "  'uuid': 'd9ae241a-cd2f-4c35-b515-8273c9891ab3',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO92049597/'},\n",
       " '108_CASTJ_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO31719222',\n",
       "  'uuid': 'baebb5c6-bff2-4a63-9bbb-118b26c80390',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO31719222/'},\n",
       " '117_B6J_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO02830801',\n",
       "  'uuid': '0885025d-bc54-4151-aacb-2e3866e9e18d',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO02830801/'},\n",
       " '118_B6J_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO75684586',\n",
       "  'uuid': 'b41f3c2c-f15b-4973-9271-d6309ea2d325',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO75684586/'},\n",
       " '119_B6J_10M': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO85621848',\n",
       "  'uuid': 'b9a48ed0-468c-4660-94df-b331dcad8dc4',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO85621848/'},\n",
       " '120_B6J_10F': {'namespace': <AccessionNamespace: IGVF test>,\n",
       "  'name': 'TSTDO94777016',\n",
       "  'uuid': '3e862973-5993-431f-a2c6-a8656cb9a2aa',\n",
       "  'url': 'https://api.sandbox.igvf.org/rodent-donors/TSTDO94777016/'}}"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "submitted_mice"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "hybrid-destruction",
   "metadata": {},
   "source": [
    "# Tissues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "electoral-ability",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>IGVF Sample BOX</th>\n",
       "      <th>Mouse_Tissue ID</th>\n",
       "      <th>Unnamed: 2</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Tube label</th>\n",
       "      <th>tube weight (g)</th>\n",
       "      <th>tube+tissue wight (g)</th>\n",
       "      <th>tissue weight (mg) or blood volume (ul)</th>\n",
       "      <th>cryopreserved blood location</th>\n",
       "      <th>Dissection date</th>\n",
       "      <th>Approx. sac time</th>\n",
       "      <th>Body weight (g)</th>\n",
       "      <th>Dissector</th>\n",
       "      <th>Comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IGVF_Sample_5&amp;6</td>\n",
       "      <td>016_B6J_10F_01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hypothalamus/Pituitary</td>\n",
       "      <td>F</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>B6J</td>\n",
       "      <td>016-01</td>\n",
       "      <td>1.046</td>\n",
       "      <td>1.118</td>\n",
       "      <td>72.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-10-27</td>\n",
       "      <td>09:01:00</td>\n",
       "      <td>21.1</td>\n",
       "      <td>SK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IGVF_Sample_5&amp;6</td>\n",
       "      <td>016_B6J_10F_02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cerebellum</td>\n",
       "      <td>F</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>B6J</td>\n",
       "      <td>016-02</td>\n",
       "      <td>1.046</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1046.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-10-27</td>\n",
       "      <td>09:01:00</td>\n",
       "      <td>21.1</td>\n",
       "      <td>SK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IGVF_Sample_5&amp;6</td>\n",
       "      <td>016_B6J_10F_03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cortex/Hippocampus left</td>\n",
       "      <td>F</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>B6J</td>\n",
       "      <td>016-03</td>\n",
       "      <td>1.040</td>\n",
       "      <td>1.181</td>\n",
       "      <td>141.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-10-27</td>\n",
       "      <td>09:01:00</td>\n",
       "      <td>21.1</td>\n",
       "      <td>SK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IGVF_Sample_5&amp;6</td>\n",
       "      <td>016_B6J_10F_04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cortex/Hippocampus right</td>\n",
       "      <td>F</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>B6J</td>\n",
       "      <td>016-04</td>\n",
       "      <td>1.048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1048.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-10-27</td>\n",
       "      <td>09:01:00</td>\n",
       "      <td>21.1</td>\n",
       "      <td>SK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IGVF_Sample_5&amp;6</td>\n",
       "      <td>016_B6J_10F_05</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Liver</td>\n",
       "      <td>F</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>B6J</td>\n",
       "      <td>016-05</td>\n",
       "      <td>1.133</td>\n",
       "      <td>2.099</td>\n",
       "      <td>966.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-10-27</td>\n",
       "      <td>09:01:00</td>\n",
       "      <td>21.1</td>\n",
       "      <td>GM</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1616</th>\n",
       "      <td>IGVF_sample_17&amp;18</td>\n",
       "      <td>096_WSBJ_10F_17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TA</td>\n",
       "      <td>F</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>096-17</td>\n",
       "      <td>1.030</td>\n",
       "      <td>1.089</td>\n",
       "      <td>59.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-10-31</td>\n",
       "      <td>10:35:00</td>\n",
       "      <td>12.2</td>\n",
       "      <td>SK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1617</th>\n",
       "      <td>IGVF_sample_17&amp;18</td>\n",
       "      <td>096_WSBJ_10F_18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EDL</td>\n",
       "      <td>F</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>096-18</td>\n",
       "      <td>1.030</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1030.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-10-31</td>\n",
       "      <td>10:35:00</td>\n",
       "      <td>12.2</td>\n",
       "      <td>SK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1618</th>\n",
       "      <td>IGVF_sample_17&amp;18</td>\n",
       "      <td>096_WSBJ_10F_19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Tail</td>\n",
       "      <td>F</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>096-19</td>\n",
       "      <td>1.047</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1047.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-10-31</td>\n",
       "      <td>10:35:00</td>\n",
       "      <td>12.2</td>\n",
       "      <td>GM</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1619</th>\n",
       "      <td>IGVF_sample_17&amp;18</td>\n",
       "      <td>096_WSBJ_10F_20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PBMC - WBC</td>\n",
       "      <td>F</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>096-20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-10-31</td>\n",
       "      <td>10:35:00</td>\n",
       "      <td>12.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1621</th>\n",
       "      <td>IGVF_Sample_5&amp;6</td>\n",
       "      <td>122_B6J_10F_16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gastrocnemius</td>\n",
       "      <td>F</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>B6J</td>\n",
       "      <td>022-16</td>\n",
       "      <td>1.035</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1035.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-05-31</td>\n",
       "      <td>10:43:00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1621 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        IGVF Sample BOX  Mouse_Tissue ID Unnamed: 2                    Tissue  \\\n",
       "0       IGVF_Sample_5&6   016_B6J_10F_01        NaN    Hypothalamus/Pituitary   \n",
       "1       IGVF_Sample_5&6   016_B6J_10F_02        NaN                Cerebellum   \n",
       "2       IGVF_Sample_5&6   016_B6J_10F_03        NaN   Cortex/Hippocampus left   \n",
       "3       IGVF_Sample_5&6   016_B6J_10F_04        NaN  Cortex/Hippocampus right   \n",
       "4       IGVF_Sample_5&6   016_B6J_10F_05        NaN                     Liver   \n",
       "...                 ...              ...        ...                       ...   \n",
       "1616  IGVF_sample_17&18  096_WSBJ_10F_17        NaN                        TA   \n",
       "1617  IGVF_sample_17&18  096_WSBJ_10F_18        NaN                       EDL   \n",
       "1618  IGVF_sample_17&18  096_WSBJ_10F_19        NaN                      Tail   \n",
       "1619  IGVF_sample_17&18  096_WSBJ_10F_20        NaN                PBMC - WBC   \n",
       "1621    IGVF_Sample_5&6   122_B6J_10F_16        NaN             Gastrocnemius   \n",
       "\n",
       "     Sex Timepoint Genotype Tube label  tube weight (g)  \\\n",
       "0      F  10 weeks      B6J     016-01            1.046   \n",
       "1      F  10 weeks      B6J     016-02            1.046   \n",
       "2      F  10 weeks      B6J     016-03            1.040   \n",
       "3      F  10 weeks      B6J     016-04            1.048   \n",
       "4      F  10 weeks      B6J     016-05            1.133   \n",
       "...   ..       ...      ...        ...              ...   \n",
       "1616   F  10 weeks     WSBJ     096-17            1.030   \n",
       "1617   F  10 weeks     WSBJ     096-18            1.030   \n",
       "1618   F  10 weeks     WSBJ     096-19            1.047   \n",
       "1619   F  10 weeks     WSBJ     096-20              NaN   \n",
       "1621   F  10 weeks      B6J     022-16            1.035   \n",
       "\n",
       "      tube+tissue wight (g)  tissue weight (mg) or blood volume (ul)  \\\n",
       "0                     1.118                                     72.0   \n",
       "1                       NaN                                  -1046.0   \n",
       "2                     1.181                                    141.0   \n",
       "3                       NaN                                  -1048.0   \n",
       "4                     2.099                                    966.0   \n",
       "...                     ...                                      ...   \n",
       "1616                  1.089                                     59.0   \n",
       "1617                    NaN                                  -1030.0   \n",
       "1618                    NaN                                  -1047.0   \n",
       "1619                    NaN                                      NaN   \n",
       "1621                    NaN                                  -1035.0   \n",
       "\n",
       "     cryopreserved blood location Dissection date Approx. sac time  \\\n",
       "0                             NaN      2022-10-27         09:01:00   \n",
       "1                             NaN      2022-10-27         09:01:00   \n",
       "2                             NaN      2022-10-27         09:01:00   \n",
       "3                             NaN      2022-10-27         09:01:00   \n",
       "4                             NaN      2022-10-27         09:01:00   \n",
       "...                           ...             ...              ...   \n",
       "1616                          NaN      2022-10-31         10:35:00   \n",
       "1617                          NaN      2022-10-31         10:35:00   \n",
       "1618                          NaN      2022-10-31         10:35:00   \n",
       "1619                          NaN      2022-10-31         10:35:00   \n",
       "1621                          NaN      2023-05-31         10:43:00   \n",
       "\n",
       "      Body weight (g) Dissector Comment  \n",
       "0                21.1        SK     NaN  \n",
       "1                21.1        SK     NaN  \n",
       "2                21.1        SK     NaN  \n",
       "3                21.1        SK     NaN  \n",
       "4                21.1        GM     NaN  \n",
       "...               ...       ...     ...  \n",
       "1616             12.2        SK     NaN  \n",
       "1617             12.2        SK     NaN  \n",
       "1618             12.2        GM     NaN  \n",
       "1619             12.2       NaN     NaN  \n",
       "1621              NaN        SK     NaN  \n",
       "\n",
       "[1621 rows x 17 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#if models.Tissue.objects.count() > 0:\n",
    "#    truncate(models.Tissue)\n",
    "\n",
    "tissues_8cube = pandas.read_excel(\n",
    "    lizs_sheet_name, \n",
    "    sheet_name=\"Samples - 8 founders\",\n",
    "    header=0,\n",
    ").dropna(axis=0, how=\"all\")\n",
    "tissues_8cube"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "final-maximum",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>IGVF Sample BOX</th>\n",
       "      <th>Mouse_Tissue ID</th>\n",
       "      <th>Unnamed: 2</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Tube label</th>\n",
       "      <th>tube weight (g)</th>\n",
       "      <th>tube+tissue wight (g)</th>\n",
       "      <th>...</th>\n",
       "      <th>PBMC count 1</th>\n",
       "      <th>PBMC count 2</th>\n",
       "      <th>PBMC RSB after RBC lysis</th>\n",
       "      <th>Total # PBMCs</th>\n",
       "      <th>cryopreserved PBMC location</th>\n",
       "      <th>Dissection date</th>\n",
       "      <th>Approx. sac time</th>\n",
       "      <th>Mouse_Tissue ID.1</th>\n",
       "      <th>Dissector</th>\n",
       "      <th>Comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IGVF_BridgeSample_1</td>\n",
       "      <td>097_CASTJ_10M_19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Tail</td>\n",
       "      <td>M</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>CASTJ</td>\n",
       "      <td>097-19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.094</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-01-19</td>\n",
       "      <td>11:38:00</td>\n",
       "      <td>097_CASTJ_10M_19</td>\n",
       "      <td>SK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IGVF_BridgeSample_1</td>\n",
       "      <td>097_CASTJ_10M_20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PBMC - WBC</td>\n",
       "      <td>M</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>CASTJ</td>\n",
       "      <td>097-20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>125.0</td>\n",
       "      <td>103.0</td>\n",
       "      <td>300.0</td>\n",
       "      <td>684000.0</td>\n",
       "      <td>Canister 5, Cane 1</td>\n",
       "      <td>2023-01-19</td>\n",
       "      <td>11:38:00</td>\n",
       "      <td>097_CASTJ_10M_20</td>\n",
       "      <td>SK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IGVF_BridgeSample_1</td>\n",
       "      <td>097_CASTJ_10M_21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Left cortex</td>\n",
       "      <td>M</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>CASTJ</td>\n",
       "      <td>097-21</td>\n",
       "      <td>2.332</td>\n",
       "      <td>2.492</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-01-19</td>\n",
       "      <td>11:38:00</td>\n",
       "      <td>097_CASTJ_10M_21</td>\n",
       "      <td>SK</td>\n",
       "      <td>Tube wasn't weighed before tissue was added, s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IGVF_BridgeSample_1</td>\n",
       "      <td>097_CASTJ_10M_22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Right cortex</td>\n",
       "      <td>M</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>CASTJ</td>\n",
       "      <td>097-22</td>\n",
       "      <td>1.043</td>\n",
       "      <td>1.156</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-01-19</td>\n",
       "      <td>11:38:00</td>\n",
       "      <td>097_CASTJ_10M_22</td>\n",
       "      <td>SK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IGVF_BridgeSample_1</td>\n",
       "      <td>097_CASTJ_10M_23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Left hippocampus</td>\n",
       "      <td>M</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>CASTJ</td>\n",
       "      <td>097-23</td>\n",
       "      <td>1.038</td>\n",
       "      <td>1.063</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-01-19</td>\n",
       "      <td>11:38:00</td>\n",
       "      <td>097_CASTJ_10M_23</td>\n",
       "      <td>SK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>IGVF_BridgeSample_2</td>\n",
       "      <td>120_B6J_10F_20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PBMC - WBC</td>\n",
       "      <td>F</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>B6J</td>\n",
       "      <td>120-20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>79.0</td>\n",
       "      <td>110.0</td>\n",
       "      <td>300.0</td>\n",
       "      <td>567000.0</td>\n",
       "      <td>Canister 5, Cane 5</td>\n",
       "      <td>2023-02-02</td>\n",
       "      <td>10:16:00</td>\n",
       "      <td>120_B6J_10F_20</td>\n",
       "      <td>SK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>IGVF_BridgeSample_2</td>\n",
       "      <td>120_B6J_10F_21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Left cortex</td>\n",
       "      <td>F</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>B6J</td>\n",
       "      <td>120-21</td>\n",
       "      <td>2.332</td>\n",
       "      <td>2.465</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-02-02</td>\n",
       "      <td>10:16:00</td>\n",
       "      <td>120_B6J_10F_21</td>\n",
       "      <td>SK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>IGVF_BridgeSample_2</td>\n",
       "      <td>120_B6J_10F_22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Right cortex</td>\n",
       "      <td>F</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>B6J</td>\n",
       "      <td>120-22</td>\n",
       "      <td>1.041</td>\n",
       "      <td>1.154</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-02-02</td>\n",
       "      <td>10:16:00</td>\n",
       "      <td>120_B6J_10F_22</td>\n",
       "      <td>SK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>IGVF_BridgeSample_2</td>\n",
       "      <td>120_B6J_10F_23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Left hippocampus</td>\n",
       "      <td>F</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>B6J</td>\n",
       "      <td>120-23</td>\n",
       "      <td>1.042</td>\n",
       "      <td>1.063</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-02-02</td>\n",
       "      <td>10:16:00</td>\n",
       "      <td>120_B6J_10F_23</td>\n",
       "      <td>SK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>IGVF_BridgeSample_2</td>\n",
       "      <td>120_B6J_10F_24</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Right hippocampus</td>\n",
       "      <td>F</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>B6J</td>\n",
       "      <td>120-24</td>\n",
       "      <td>1.040</td>\n",
       "      <td>1.069</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-02-02</td>\n",
       "      <td>10:16:00</td>\n",
       "      <td>120_B6J_10F_24</td>\n",
       "      <td>SK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>144 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         IGVF Sample BOX   Mouse_Tissue ID Unnamed: 2             Tissue Sex  \\\n",
       "0    IGVF_BridgeSample_1  097_CASTJ_10M_19        NaN               Tail   M   \n",
       "1    IGVF_BridgeSample_1  097_CASTJ_10M_20        NaN         PBMC - WBC   M   \n",
       "2    IGVF_BridgeSample_1  097_CASTJ_10M_21        NaN        Left cortex   M   \n",
       "3    IGVF_BridgeSample_1  097_CASTJ_10M_22        NaN       Right cortex   M   \n",
       "4    IGVF_BridgeSample_1  097_CASTJ_10M_23        NaN   Left hippocampus   M   \n",
       "..                   ...               ...        ...                ...  ..   \n",
       "139  IGVF_BridgeSample_2    120_B6J_10F_20        NaN         PBMC - WBC   F   \n",
       "140  IGVF_BridgeSample_2    120_B6J_10F_21        NaN        Left cortex   F   \n",
       "141  IGVF_BridgeSample_2    120_B6J_10F_22        NaN       Right cortex   F   \n",
       "142  IGVF_BridgeSample_2    120_B6J_10F_23        NaN   Left hippocampus   F   \n",
       "143  IGVF_BridgeSample_2    120_B6J_10F_24        NaN  Right hippocampus   F   \n",
       "\n",
       "    Timepoint Genotype Tube label  tube weight (g)  tube+tissue wight (g)  \\\n",
       "0    10 weeks    CASTJ     097-19              NaN                  1.094   \n",
       "1    10 weeks    CASTJ     097-20              NaN                    NaN   \n",
       "2    10 weeks    CASTJ     097-21            2.332                  2.492   \n",
       "3    10 weeks    CASTJ     097-22            1.043                  1.156   \n",
       "4    10 weeks    CASTJ     097-23            1.038                  1.063   \n",
       "..        ...      ...        ...              ...                    ...   \n",
       "139  10 weeks      B6J     120-20              NaN                    NaN   \n",
       "140  10 weeks      B6J     120-21            2.332                  2.465   \n",
       "141  10 weeks      B6J     120-22            1.041                  1.154   \n",
       "142  10 weeks      B6J     120-23            1.042                  1.063   \n",
       "143  10 weeks      B6J     120-24            1.040                  1.069   \n",
       "\n",
       "     ...  PBMC count 1  PBMC count 2  PBMC RSB after RBC lysis  Total # PBMCs  \\\n",
       "0    ...           NaN           NaN                       NaN            NaN   \n",
       "1    ...         125.0         103.0                     300.0       684000.0   \n",
       "2    ...           NaN           NaN                       NaN            NaN   \n",
       "3    ...           NaN           NaN                       NaN            NaN   \n",
       "4    ...           NaN           NaN                       NaN            NaN   \n",
       "..   ...           ...           ...                       ...            ...   \n",
       "139  ...          79.0         110.0                     300.0       567000.0   \n",
       "140  ...           NaN           NaN                       NaN            NaN   \n",
       "141  ...           NaN           NaN                       NaN            NaN   \n",
       "142  ...           NaN           NaN                       NaN            NaN   \n",
       "143  ...           NaN           NaN                       NaN            NaN   \n",
       "\n",
       "     cryopreserved PBMC location  Dissection date Approx. sac time  \\\n",
       "0                            NaN       2023-01-19         11:38:00   \n",
       "1             Canister 5, Cane 1       2023-01-19         11:38:00   \n",
       "2                            NaN       2023-01-19         11:38:00   \n",
       "3                            NaN       2023-01-19         11:38:00   \n",
       "4                            NaN       2023-01-19         11:38:00   \n",
       "..                           ...              ...              ...   \n",
       "139           Canister 5, Cane 5       2023-02-02         10:16:00   \n",
       "140                          NaN       2023-02-02         10:16:00   \n",
       "141                          NaN       2023-02-02         10:16:00   \n",
       "142                          NaN       2023-02-02         10:16:00   \n",
       "143                          NaN       2023-02-02         10:16:00   \n",
       "\n",
       "    Mouse_Tissue ID.1 Dissector  \\\n",
       "0    097_CASTJ_10M_19        SK   \n",
       "1    097_CASTJ_10M_20        SK   \n",
       "2    097_CASTJ_10M_21        SK   \n",
       "3    097_CASTJ_10M_22        SK   \n",
       "4    097_CASTJ_10M_23        SK   \n",
       "..                ...       ...   \n",
       "139    120_B6J_10F_20        SK   \n",
       "140    120_B6J_10F_21        SK   \n",
       "141    120_B6J_10F_22        SK   \n",
       "142    120_B6J_10F_23        SK   \n",
       "143    120_B6J_10F_24        SK   \n",
       "\n",
       "                                               Comment  \n",
       "0                                                  NaN  \n",
       "1                                                  NaN  \n",
       "2    Tube wasn't weighed before tissue was added, s...  \n",
       "3                                                  NaN  \n",
       "4                                                  NaN  \n",
       "..                                                 ...  \n",
       "139                                                NaN  \n",
       "140                                                NaN  \n",
       "141                                                NaN  \n",
       "142                                                NaN  \n",
       "143                                                NaN  \n",
       "\n",
       "[144 rows x 22 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#if models.Tissue.objects.count() > 0:\n",
    "#    truncate(models.Tissue)\n",
    "\n",
    "tissues_bridge = pandas.read_excel(\n",
    "    lizs_sheet_name, \n",
    "    sheet_name=\"Bridge Samples\",\n",
    "    header=0,\n",
    ").dropna(axis=0, how=\"all\")\n",
    "tissues_bridge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "phantom-somerset",
   "metadata": {},
   "outputs": [],
   "source": [
    "tissues = pandas.concat([tissues_8cube, tissues_bridge])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "threatened-procedure",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_timepoint(value):\n",
    "    value, units = value.split(\" \")\n",
    "    value = float(value)\n",
    "    \n",
    "    unit_map = {\n",
    "        \"day\": models.AgeUnitsEnum.DAY,\n",
    "        \"days\": models.AgeUnitsEnum.DAY,\n",
    "        \"week\": models.AgeUnitsEnum.WEEK,\n",
    "        \"weeks\": models.AgeUnitsEnum.WEEK,\n",
    "        \"month\": models.AgeUnitsEnum.MONTH,\n",
    "        \"months\": models.AgeUnitsEnum.MONTH,\n",
    "    }\n",
    "    return (value, unit_map[units])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "specialized-albert",
   "metadata": {},
   "outputs": [],
   "source": [
    "ontology_map = {}\n",
    "for record in models.OntologyTerm.objects.all():\n",
    "    ontology_map[record.curie] = record"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "alleged-praise",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "136\n"
     ]
    }
   ],
   "source": [
    "submitted_tissues = {}\n",
    "\n",
    "for i, row in submitted_sheets[\"tissue\"].iterrows():\n",
    "    if not pandas.isnull(row[\"accession\"]):\n",
    "        aliases = row[\"aliases:array\"].split(\",\")\n",
    "        assert len(aliases) == 1\n",
    "        alias_fields = aliases[0][len(\"ali-mortazavi:\"):].split(\"_\")\n",
    "        tissue_id = \"_\".join(alias_fields[0:4])\n",
    "        submitted_tissues.setdefault(tissue_id, []).append({\n",
    "            \"namespace\": igvf_test_namespace,\n",
    "            \"name\": row[\"accession\"],\n",
    "            \"uuid\": row[\"uuid\"],\n",
    "            \"url\": \"https://ap.sndbox.igvf.org/tissues/{}/\".format(row[\"accession\"]),\n",
    "        })\n",
    "\n",
    "print(len(submitted_tissues))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "relative-corps",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1765/1765 [01:20<00:00, 22.01it/s]\n"
     ]
    }
   ],
   "source": [
    "if models.Tissue.objects.count() > 0:\n",
    "    truncate(models.Tissue)\n",
    "\n",
    "\n",
    "failed = False\n",
    "for i, row in tqdm(tissues.iterrows(), total=tissues.shape[0]):\n",
    "    if not pandas.isnull(row[\"Tissue\"]):\n",
    "        mouse_tissue_id_label = \"Mouse_Tissue ID\"\n",
    "        mouse_tissue_name_fields = row[mouse_tissue_id_label].split(\"_\")\n",
    "        \n",
    "        # some validation on tissue name\n",
    "        if len(mouse_tissue_name_fields) != 4:\n",
    "            raise ValueError(f\"Not enough elements in mouse tissue id {row[mouse_tissue_id_label]}\")\n",
    "            \n",
    "        if not mouse_tissue_name_fields[1] in strain_background:\n",
    "            raise ValueError(f\"tissue strain field in {row[mouse_tissue_id_label]} not recognized\")\n",
    "        \n",
    "        mouse_name = \"_\".join(mouse_tissue_name_fields[0:3])\n",
    "        try:\n",
    "            mouse = models.Mouse.objects.get(name=mouse_name)\n",
    "        except models.Mouse.DoesNotExist:\n",
    "            print(\"row {}, {} was not found\".format(i+2, mouse_name))\n",
    "            failed = True\n",
    "            continue\n",
    "\n",
    "        genotype = strain_background[row[\"Genotype\"]]\n",
    "        \n",
    "        # this is the \"label swap\" on spreadsheet rows 602-605.\n",
    "        if mouse_name == \"092_CASTJ_10F\":\n",
    "            genotype = \"CAST/EiJ\"\n",
    "        # this is the other half the swap on spreadsheet rows 1522-1525\n",
    "        elif mouse_name == \"046_NZOJ_10F\":\n",
    "            genotype = 'NZO/HlLtJ'\n",
    "\n",
    "        assert mouse.strain.code == genotype, f\"{mouse.strain.code} != {genotype}\"\n",
    "        tissue_terms = []\n",
    "        for term_curie, term_name in tissue_dissection_to_ontology_map[row[\"Tissue\"]]:\n",
    "            tissue_terms.append(ontology_map[term_curie])\n",
    "        \n",
    "        #age, age_units = parse_timepoint(row[\"Timepoint\"])\n",
    "        \n",
    "        if pandas.isnull(row[\"Approx. sac time\"]):\n",
    "            sac_time = datetime.time(0,0,0)\n",
    "        else:\n",
    "            sac_time = datetime.time(\n",
    "                row[\"Approx. sac time\"].hour,\n",
    "                row[\"Approx. sac time\"].minute,\n",
    "                row[\"Approx. sac time\"].second,    \n",
    "            )\n",
    "        dissection = datetime.datetime(\n",
    "            row[\"Dissection date\"].year,\n",
    "            row[\"Dissection date\"].month,\n",
    "            row[\"Dissection date\"].day,\n",
    "            sac_time.hour,\n",
    "            sac_time.minute,\n",
    "            sac_time.second,\n",
    "            tzinfo=los_angeles_tz,\n",
    "        )\n",
    "        \n",
    "        record = models.Tissue(\n",
    "            mouse=mouse,\n",
    "            name = row[mouse_tissue_id_label],\n",
    "            description = row[\"Tissue\"],\n",
    "            dissection_time=dissection,\n",
    "            #age=age,\n",
    "            #age_units=age_units,\n",
    "            tube_label=row[\"Tube label\"],\n",
    "            timepoint_description=row[\"Timepoint\"],\n",
    "            life_stage=models.LifeStageEnum.ADULT,\n",
    "            dissector=str_or_empty(row[\"Dissector\"]),\n",
    "            dissection_notes=str_or_empty(row[\"Comment\"]),\n",
    "        )\n",
    "\n",
    "        tube_weight_label = \"tube weight (g)\"\n",
    "        if not pandas.isnull(row[tube_weight_label]):\n",
    "            record.tube_weight_g = float(row[tube_weight_label])\n",
    "\n",
    "        total_weight_label = \"tube+tissue wight (g)\"\n",
    "        if not pandas.isnull(row[total_weight_label]):\n",
    "            record.total_weight_g = float(row[total_weight_label])\n",
    "        \n",
    "        record.save()\n",
    "        record.ontology_term.set(tissue_terms)\n",
    "        \n",
    "        if row[mouse_tissue_id_label] in submitted_tissues:\n",
    "            accessions = []\n",
    "            for accession_row in submitted_tissues[row[mouse_tissue_id_label]]:\n",
    "                accession = models.Accession(\n",
    "                    namespace=accession_row[\"namespace\"],\n",
    "                    name=accession_row[\"name\"],\n",
    "                    uuid=accession_row[\"uuid\"],\n",
    "                    url=accession_row[\"url\"],\n",
    "                )\n",
    "                accession.save()\n",
    "                accessions.append(accession)\n",
    "            record.accession.set(accessions)\n",
    "        record.save()\n",
    "\n",
    "assert not failed, \"Check warning messages.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "scenic-vermont",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "historic-stand",
   "metadata": {},
   "source": [
    "# Information from Samples into experiment\n",
    "\n",
    "- Total barcoded nuclei (Samples into experiment)\n",
    "\n",
    "Is this FixedTissue, FixedSample, FixedBiosample?\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "applied-mustang",
   "metadata": {},
   "outputs": [],
   "source": [
    "if models.FixedSample.objects.count() > 0:\n",
    "    truncate(models.FixedSample)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "lonely-spelling",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_samples_into_experiment(book_name, sheet_name, header=0):\n",
    "\n",
    "    samples_into_experiment = pandas.read_excel(\n",
    "        book_name, \n",
    "        sheet_name=sheet_name,\n",
    "        header=header,\n",
    "        index_col=None,\n",
    "    )\n",
    "\n",
    "    # was trying to normalize two row hierarchical headers, \n",
    "    #level0 = []\n",
    "    #level1 = []\n",
    "    #for header in samples_into_experiment.columns:\n",
    "    #    level0.append(\"\" if header[0].startswith(\"Unnamed: \") else header[0])\n",
    "    #    level1.append(header[1])\n",
    "    #samples_into_experiment.columns = pandas.MultiIndex.from_arrays([level0, level1], names=[\"phase\", \"name\"])\n",
    "\n",
    "    one_based = 1\n",
    "    header_lines = 2\n",
    "    failed = False\n",
    "    for i, row in tqdm(samples_into_experiment.iterrows(), total=samples_into_experiment.shape[0]):\n",
    "        line_no = i + one_based + header_lines\n",
    "        box_name = row[\"IGVF Fixation BOX\"]\n",
    "        if isinstance(box_name, str):\n",
    "            box_name = box_name.strip()\n",
    "\n",
    "        if box_name in [\"IGVF_FIX_001\", \"IGVF_FIX_002\"]:\n",
    "            continue\n",
    "\n",
    "        tissue_id = row[\"Mouse_Tissue ID\"]\n",
    "        pooled_from = row.get(\"pooled_from\")\n",
    "\n",
    "        if pandas.isnull(tissue_id):\n",
    "            continue\n",
    "\n",
    "        #if (tissue_id.endswith(\"_25\") or tissue_id.endswith(\"_26\")) and pandas.isnull(pooled_from):\n",
    "        #    print(f\"How are we going to merge these samples? {tissue_id}\")\n",
    "        #    continue\n",
    "\n",
    "        if pandas.isnull(pooled_from):\n",
    "            pooled_from = [tissue_id]\n",
    "        else:\n",
    "            pooled_from = pooled_from.split(\",\")\n",
    "\n",
    "        tissues = []\n",
    "        for pooled_id in pooled_from:\n",
    "            try:\n",
    "                tissues.append(models.Tissue.objects.get(name=pooled_id))\n",
    "            except models.Tissue.DoesNotExist:\n",
    "                print(f\"Tissue {pooled_id} not found in tissue table on {line_no}\")\n",
    "                failed = True\n",
    "\n",
    "        weight = row[\"weight (mg)\"]\n",
    "        fixation_date = row[\"Fixation date\"]\n",
    "        notes = row[\"Notes\"]\n",
    "\n",
    "        if weight == \"#VALUE!\":\n",
    "            pass\n",
    "        elif weight < 0:\n",
    "            pass\n",
    "        elif pandas.isnull(weight):\n",
    "            pass\n",
    "        else:\n",
    "            total_weight = 0\n",
    "            for tissue in tissues:\n",
    "                if not pandas.isnull(tissue.weight_mg):\n",
    "                    total_weight += tissue.weight_mg\n",
    "\n",
    "            if not numpy.isclose(total_weight, weight):\n",
    "                print(f\"Sum of tissue weights from {pooled_from} {total_weight} doesn't match {weight}. {line_no}\")\n",
    "\n",
    "        sample_id = row[\"Sample ID\"]\n",
    "        if not (pandas.isnull(sample_id) or pandas.isnull(tissue_id)):\n",
    "            if tissue_id != sample_id:\n",
    "                print(f\"If defined, {tissue_id} should equal {sample_id} line {line_no}\")\n",
    "                failed = True\n",
    "\n",
    "        cap_label = row[\"Cap label\"]\n",
    "        if not (pandas.isnull(cap_label) or pandas.isnull(sample_id)):\n",
    "            sample_fields = sample_id.split(\"_\")\n",
    "            predicted_label = \"_\".join([sample_fields[0], sample_fields[-1]])\n",
    "            if predicted_label != cap_label:\n",
    "                print(f\"{cap_label} should equal {predicted_label} line {line_no}\")\n",
    "                failed = True\n",
    "\n",
    "        well_id = row[\"wells in Barcoding plate\"]\n",
    "\n",
    "        record = models.FixedSample(\n",
    "            name=tissue_id,\n",
    "            tube_label=cap_label,\n",
    "            fixation_name=box_name,\n",
    "            fixation_date=date_or_none(row[\"Fixation date\"]),\n",
    "            starting_nuclei=int_or_none(row[\"Total nuclei (x10^6)\"]),\n",
    "            nuclei_into_fixation=int_or_none(row[\"Nuclei into fixation (x10^6)\"]),\n",
    "            fixed_nuclei=int_or_none(row[\"Total fixed nuclei (x10^6)\"]),\n",
    "            aliquots_made=int_or_none(row[\"# aliquots\"]),\n",
    "            aliquot_volume_ul=float_or_none(row[\"uL per aliquot\"]),\n",
    "        )\n",
    "        record.save()\n",
    "        record.tissue.set(tissues)\n",
    "        record.save()\n",
    "\n",
    "    assert not failed, \"Check warning messages\"\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "central-oriental",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 56%|█████▌    | 553/994 [00:11<00:16, 27.00it/s] "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sum of tissue weights from ['046_NZOJ_10F_17'] 116.0 doesn't match 190.0. 556\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 75%|███████▍  | 744/994 [00:16<00:07, 31.83it/s] "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sum of tissue weights from ['095_CASTJ_10M_08'] 11.0 doesn't match 15.0000000000001. 747\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 994/994 [00:22<00:00, 44.60it/s] \n"
     ]
    }
   ],
   "source": [
    "load_samples_into_experiment(lizs_sheet_name, \"Samples into experiment\", header=1)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "automated-security",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 8/8 [00:00<00:00, 25.14it/s]\n"
     ]
    }
   ],
   "source": [
    "load_samples_into_experiment(lizs_sheet_name, \"Bridge samples into experiment\", header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "durable-potter",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "specific-candle",
   "metadata": {},
   "source": [
    "# Plate layout\n",
    "\n",
    "(the hard thing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "spectacular-mirror",
   "metadata": {},
   "outputs": [],
   "source": [
    "WellContent = namedtuple(\"well_content\", [\"genotype\", \"tissue_id\"])\n",
    "\n",
    "def validate_tissue_ids(contents, expected_genotypes):\n",
    "    for expected, (row_index, row) in zip(expected_genotypes, contents.iterrows()):\n",
    "        for cell in row:\n",
    "            fields = cell.split(\"_\")\n",
    "            assert len(fields) == 4, f\"Field label validation fail {cell}\"\n",
    "            # there's a few sample swaps, don't validate them.\n",
    "            if fields[1] not in (\"092_CASTJ_10F_03\"):\n",
    "                assert fields[1] == expected, f\"{fields[1]} failed to match {expected}\"\n",
    "\n",
    "def is_plate_name(name):\n",
    "    return not pandas.isnull(name) and name.startswith(\"IGVF_\")\n",
    "\n",
    "def find_plate_start(sheet, offsets):\n",
    "    for plate_id_row in sheet[sheet[offsets[\"plate_label\"]].apply(is_plate_name)].index:\n",
    "        plate_name = sheet.loc[plate_id_row, offsets[\"plate_label\"]]\n",
    "        for i in range(plate_id_row, plate_id_row + 4):\n",
    "            cell = sheet.loc[i, offsets[\"well_start\"]]\n",
    "            if isinstance(cell, str) and (cell.startswith(\"Tissue\") or cell in (\"M\", \"F\")):\n",
    "                yield (plate_name, i)\n",
    "                \n",
    "def get_plate_genotype(sheet, start, offsets):\n",
    "    plate_tissues = []\n",
    "    for cell in sheet.loc[start+2:start+9, offsets[\"tissue_label\"]].tolist():\n",
    "        if not pandas.isnull(cell):\n",
    "            match = re.match(\"(B6J|NODJ|AJ|PWKJ|129S1J|CASTJ|WSBJ|NZOJ)\", cell)\n",
    "            if match is not None:\n",
    "                plate_tissues.append(match.group(0))\n",
    "            else:\n",
    "                print(\"Diffculty matching {} in {} {}\".format(cell, start+2, start+9))\n",
    "    \n",
    "    return plate_tissues\n",
    "\n",
    "def parse_single_well_block(sheet, plate_start, offsets):\n",
    "    # column ids\n",
    "    simple_plate_column_ids = sheet.loc[plate_start+1, offsets[\"well_range\"]].tolist()\n",
    "\n",
    "    plate_row_labels = sheet.loc[plate_start+2:plate_start+9, offsets[\"well_row_label\"]].tolist()\n",
    "\n",
    "    contents = sheet.loc[plate_start+2:plate_start+9, offsets[\"well_range\"]].copy()\n",
    "    contents.index = plate_row_labels\n",
    "    contents.columns = simple_plate_column_ids\n",
    "\n",
    "    genotypes = get_plate_genotype(sheet, plate_start, offsets)\n",
    "    validate_tissue_ids(contents, genotypes)\n",
    "\n",
    "    well_contents = {}\n",
    "    for genotype, (well_row, row) in zip(genotypes, contents.iterrows()):\n",
    "        for well_column, cell in zip(simple_plate_column_ids, row):\n",
    "            well_contents[(well_row, str(well_column))] = [WellContent(genotype, cell)]\n",
    "    return well_contents\n",
    "\n",
    "def parse_multiplexed_well_block(sheet, start, expected_genotypes, offsets):\n",
    "    well_re = re.compile(\"^[A-H]1?[\\d]$\")\n",
    "    wells = sheet.loc[start+3, offsets[\"well_range\"]].tolist()\n",
    "    for cell in wells:\n",
    "        if pandas.isnull(cell) or well_re.match(cell) is None:\n",
    "            raise ValueError(f\"well value {cell} in row {start+3} does not look correct\")\n",
    "\n",
    "    contents = sheet.loc[start+0:start+1, offsets[\"well_range\"]]\n",
    "    contents.columns = [(x[0], x[1:]) for x in wells]\n",
    "    validate_tissue_ids(contents, expected_genotypes)\n",
    "    contents.index = expected_genotypes\n",
    "\n",
    "    well_contents = {}\n",
    "    for genotype, row in contents.iterrows():\n",
    "        for well, tissue_id in row.iteritems():\n",
    "            well_contents.setdefault(well, []).append(WellContent(genotype, tissue_id))\n",
    "\n",
    "    return well_contents\n",
    "\n",
    "def parse_plate(sheet):\n",
    "    max_rows_from_plate_name_to_well_names = 4\n",
    "\n",
    "    offsets = {\n",
    "        \"plate_label\": 1,\n",
    "        \"tissue_label\": 1,\n",
    "        \"well_row_label\": 2,\n",
    "        \"well_start\": 3,\n",
    "        \"well_range\": slice(3, 10),\n",
    "    }\n",
    "    offsets_simple = offsets.copy()\n",
    "    offsets_simple[\"well_range\"] = slice(3, 14)\n",
    "\n",
    "    for plate_name, plate_start in find_plate_start(sheet, offsets):\n",
    "        well_contents = {}\n",
    "        genotypes = get_plate_genotype(sheet, plate_start, offsets)\n",
    "        \n",
    "        \n",
    "        if plate_name == \"IGVF_B01\":\n",
    "            well_contents.update(parse_single_well_block(sheet, plate_start, offsets_simple))\n",
    "            complex_blocks = []        \n",
    "        else:\n",
    "            well_contents.update(parse_single_well_block(sheet, plate_start, offsets))\n",
    "\n",
    "            complex_blocks = [\n",
    "                (plate_start + 12, slice(0, 2)),\n",
    "                (plate_start + 12 + 5, slice(2, 4)),\n",
    "                (plate_start + 12 + 10, slice(4, 6)),\n",
    "                (plate_start + 12 + 15, slice(6, 8)),\n",
    "            ]\n",
    "\n",
    "            for complex_start, genotype_slice in complex_blocks:\n",
    "                expected_genotypes = genotypes[genotype_slice]\n",
    "                well_contents.update(parse_multiplexed_well_block(sheet, complex_start, expected_genotypes, offsets))\n",
    "        \n",
    "        yield (plate_name, well_contents)\n",
    "        \n",
    "plate_layout = pandas.read_excel(\n",
    "    lizs_sheet_name, \n",
    "    sheet_name=\"Plate setups\",\n",
    "    header=None,\n",
    ")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "stupid-hindu",
   "metadata": {},
   "outputs": [],
   "source": [
    "#for plate_name, plate_contents in parse_plate(plate_layout):\n",
    "#    for well_id in plate_contents:\n",
    "#        barcodes = models.LibraryBarcode.objects.filter(\n",
    "#            kit=wt_mega_2_kit,\n",
    "#            code=\"{}{}\".format(well_id[0], well_id[1]),\n",
    "#        )        \n",
    "#        \n",
    "#        print(plate_name, well_id, barcodes.count())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "subject-accused",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Validate tissue references:\n",
    "failed = False\n",
    "for plate_name, plate_contents in parse_plate(plate_layout):\n",
    "    for well_id in plate_contents:\n",
    "        well_contents = plate_contents[well_id]\n",
    "        \n",
    "        for well_fraction in well_contents:\n",
    "            try:\n",
    "                biosample = models.FixedSample.objects.get(name=well_fraction.tissue_id)\n",
    "            except models.FixedSample.DoesNotExist:\n",
    "                print(f\"Unable to find {well_fraction.tissue_id} on plate {plate_name}\")\n",
    "                failed = True\n",
    "                \n",
    "#assert not failed, \"Resolve tissues\"\n",
    "\n",
    "# populate database\n",
    "if models.SplitSeqWell.biosample.through.objects.count() > 0:\n",
    "    truncate(models.SplitSeqWell.biosample.through)\n",
    "\n",
    "if models.SplitSeqWell.objects.count() > 0:\n",
    "    truncate(models.SplitSeqWell)\n",
    "    \n",
    "if models.SplitSeqPlate.objects.count() > 0:\n",
    "    truncate(models.SplitSeqPlate)\n",
    "\n",
    "wt_mega_2_kit = models.LibraryConstructionKit.objects.get(name=\"WT Mega\", version=\"v2\")\n",
    "\n",
    "errors = 0\n",
    "for plate_name, plate_contents in parse_plate(plate_layout):\n",
    "    plate_record = models.SplitSeqPlate(\n",
    "        name=plate_name,\n",
    "        size=models.PlateSizeEnum.size_96,\n",
    "        pool_location=None,\n",
    "        date_performed=None,\n",
    "    )\n",
    "    plate_record.save()\n",
    "\n",
    "    for well_id in plate_contents:\n",
    "        well_contents = plate_contents[well_id]\n",
    "        \n",
    "        biosamples = []\n",
    "        for well_content in well_contents:\n",
    "            try:\n",
    "                biosamples.append(models.FixedSample.objects.get(name=well_content.tissue_id))\n",
    "            except models.FixedSample.DoesNotExist:\n",
    "                print(f\"unable to find tissue {well_content.tissue_id} for {plate_name} {well_id}\")\n",
    "                errors += 1\n",
    "        \n",
    "        barcodes = models.LibraryBarcode.objects.filter(\n",
    "            kit=wt_mega_2_kit,\n",
    "            code=\"{}{}\".format(well_id[0], well_id[1]),\n",
    "        )\n",
    "        \n",
    "        well_record = models.SplitSeqWell(\n",
    "            plate=plate_record,\n",
    "            row=well_id[0],\n",
    "            column=well_id[1],\n",
    "        )\n",
    "        well_record.save()\n",
    "        well_record.biosample.set(biosamples)\n",
    "        well_record.barcode.set(barcodes)\n",
    "        well_record.save()\n",
    "        \n",
    "assert not errors, \"Check error messages\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "discrete-pricing",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "headed-scope",
   "metadata": {},
   "source": [
    "# Submitted measurement sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "arctic-burke",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25\n"
     ]
    }
   ],
   "source": [
    "measurement_sets = {}\n",
    "\n",
    "for i, row in submitted_sheets[\"measurement_set\"].iterrows():\n",
    "    if not pandas.isnull(row[\"accession\"]):\n",
    "        aliases = row[\"aliases:array\"].split(',')\n",
    "        name = aliases[0].replace(\"ali-mortazavi:\", \"\")\n",
    "        \n",
    "        try:\n",
    "            record = models.MeasurementSet.objects.get(name=name)\n",
    "        except models.MeasurementSet.DoesNotExist:\n",
    "            record = models.MeasurementSet(\n",
    "                name=name,\n",
    "            )\n",
    "            record.save()\n",
    "\n",
    "            accession = models.Accession(\n",
    "                namespace=igvf_test_namespace,\n",
    "                name=row[\"accession\"],\n",
    "                uuid=row[\"uuid\"],\n",
    "                url=\"https://api.sandbox.igvf.org/measurement-sets/{}/\".format(row[\"accession\"]),\n",
    "            )\n",
    "            accession.save()\n",
    "            record.accession.set([accession])\n",
    "            record.save()\n",
    "        measurement_sets[name] = record\n",
    "print(len(measurement_sets))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "endless-compromise",
   "metadata": {},
   "source": [
    "# Subpool\n",
    "\n",
    "## Extract meaning from the Experiments tab"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "permanent-italic",
   "metadata": {},
   "source": [
    "- number of nuclei per subpool (Experiment) \n",
    "- cDNA amp # PCR cycles (Experiment)\n",
    "- cDNA ng/ul in 25ul (experiment)\n",
    "- total cDNA ng (experiment)\n",
    "- Bioanalyzer cDNA ave bp length (expierment)\n",
    "- Sub library Index PCR # (experiment)\n",
    "- Sublibrary Index (experiment)\n",
    "- Sequence for SampleSheet (experiment)\n",
    "- library ng/ul (experiment)\n",
    "- Bioanalyzer library ave bp length (experiment)\n",
    "- Nextseq run number (experiment)\n",
    "\n",
    "- number of 67k aliquots  (experiment)\n",
    "- number of 8k aliquots (experiment)\n",
    "\n",
    "- QC # raw reads (1 mismatch) (experiment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "civil-musician",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Experiment</th>\n",
       "      <th>mice</th>\n",
       "      <th>Sample Box</th>\n",
       "      <th>Fixation Box</th>\n",
       "      <th>Split-seq prep start date</th>\n",
       "      <th>barcoded [nuclei/ul] cell counter</th>\n",
       "      <th>Volume of barcoded nuclei</th>\n",
       "      <th>Total barcoded nuclei (1st 3 rounds)</th>\n",
       "      <th># of backup aliquots</th>\n",
       "      <th>size of backup aliquot</th>\n",
       "      <th>...</th>\n",
       "      <th>Unnamed: 44</th>\n",
       "      <th>Unnamed: 45</th>\n",
       "      <th>Unnamed: 46</th>\n",
       "      <th>Unnamed: 47</th>\n",
       "      <th>Unnamed: 48</th>\n",
       "      <th>Unnamed: 49</th>\n",
       "      <th>Unnamed: 50</th>\n",
       "      <th>Unnamed: 51</th>\n",
       "      <th>Unnamed: 52</th>\n",
       "      <th>Unnamed: 53</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IGVF_Splitseq_002</td>\n",
       "      <td>B6 (2F+2M)</td>\n",
       "      <td>IGVF_Sample_3</td>\n",
       "      <td>IGVF_FIX_001</td>\n",
       "      <td>2022-08-24</td>\n",
       "      <td>4540.0</td>\n",
       "      <td>240.0</td>\n",
       "      <td>1089600.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>67k</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CAST (2F+2M)</td>\n",
       "      <td>IGVF_Sample_4</td>\n",
       "      <td>IGVF_FIX_002</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Fixation V1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1M V1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGVF_FIX_007</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>13k</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>214 rows × 54 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Experiment          mice     Sample Box  Fixation Box  \\\n",
       "0    IGVF_Splitseq_002    B6 (2F+2M)  IGVF_Sample_3  IGVF_FIX_001   \n",
       "1                  NaN  CAST (2F+2M)  IGVF_Sample_4  IGVF_FIX_002   \n",
       "2          Fixation V1           NaN            NaN           NaN   \n",
       "3                1M V1           NaN            NaN           NaN   \n",
       "4                  NaN           NaN            NaN           NaN   \n",
       "..                 ...           ...            ...           ...   \n",
       "209                NaN           NaN            NaN  IGVF_FIX_007   \n",
       "210                NaN           NaN            NaN           NaN   \n",
       "211                NaN           NaN            NaN           NaN   \n",
       "212                NaN           NaN            NaN           NaN   \n",
       "213                NaN           NaN            NaN           NaN   \n",
       "\n",
       "    Split-seq prep start date  barcoded [nuclei/ul] cell counter  \\\n",
       "0                  2022-08-24                             4540.0   \n",
       "1                         NaT                                NaN   \n",
       "2                         NaT                                NaN   \n",
       "3                         NaT                                NaN   \n",
       "4                         NaT                                NaN   \n",
       "..                        ...                                ...   \n",
       "209                       NaT                                NaN   \n",
       "210                       NaT                                NaN   \n",
       "211                       NaT                                NaN   \n",
       "212                       NaT                                NaN   \n",
       "213                       NaT                                NaN   \n",
       "\n",
       "     Volume of barcoded nuclei  Total barcoded nuclei (1st 3 rounds)  \\\n",
       "0                        240.0                             1089600.0   \n",
       "1                          NaN                                   NaN   \n",
       "2                          NaN                                   NaN   \n",
       "3                          NaN                                   NaN   \n",
       "4                          NaN                                   NaN   \n",
       "..                         ...                                   ...   \n",
       "209                        NaN                                   NaN   \n",
       "210                        NaN                                   NaN   \n",
       "211                        NaN                                   NaN   \n",
       "212                        NaN                                   NaN   \n",
       "213                        NaN                                   NaN   \n",
       "\n",
       "     # of backup aliquots size of backup aliquot  ...  Unnamed: 44  \\\n",
       "0                     3.0                    67k  ...          NaN   \n",
       "1                     NaN                    NaN  ...          NaN   \n",
       "2                     NaN                    NaN  ...          NaN   \n",
       "3                     NaN                    NaN  ...          NaN   \n",
       "4                     NaN                    NaN  ...          NaN   \n",
       "..                    ...                    ...  ...          ...   \n",
       "209                   2.0                    13k  ...          NaN   \n",
       "210                   NaN                    NaN  ...          NaN   \n",
       "211                   NaN                    NaN  ...          NaN   \n",
       "212                   NaN                    NaN  ...          NaN   \n",
       "213                   NaN                    NaN  ...          NaN   \n",
       "\n",
       "     Unnamed: 45  Unnamed: 46 Unnamed: 47 Unnamed: 48 Unnamed: 49  \\\n",
       "0            NaN          NaN         NaN         NaN         NaN   \n",
       "1            NaN          NaN         NaN         NaN         NaN   \n",
       "2            NaN          NaN         NaN         NaN         NaN   \n",
       "3            NaN          NaN         NaN         NaN         NaN   \n",
       "4            NaN          NaN         NaN         NaN         NaN   \n",
       "..           ...          ...         ...         ...         ...   \n",
       "209          NaN          NaN         NaN         NaN         NaN   \n",
       "210          NaN          NaN         NaN         NaN         NaN   \n",
       "211          NaN          NaN         NaN         NaN         NaN   \n",
       "212          NaN          NaN         NaN         NaN         NaN   \n",
       "213          NaN          NaN         NaN         NaN         NaN   \n",
       "\n",
       "     Unnamed: 50  Unnamed: 51 Unnamed: 52  Unnamed: 53  \n",
       "0            NaN          NaN         NaN          NaN  \n",
       "1            NaN          NaN         NaN          NaN  \n",
       "2            NaN          NaN         NaN          NaN  \n",
       "3            NaN          NaN         NaN          NaN  \n",
       "4            NaN          NaN         NaN          NaN  \n",
       "..           ...          ...         ...          ...  \n",
       "209          NaN          NaN         NaN          NaN  \n",
       "210          NaN          NaN         NaN          NaN  \n",
       "211          NaN          NaN         NaN          NaN  \n",
       "212          NaN          NaN         NaN          NaN  \n",
       "213          NaN          NaN         NaN          NaN  \n",
       "\n",
       "[214 rows x 54 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "experiment = pandas.read_excel(\n",
    "    lizs_sheet_name, \n",
    "    sheet_name=\"Experiment\",\n",
    "    header=0,\n",
    "    index_col=None,\n",
    "    #usecols=range(0, 49)\n",
    ")\n",
    "experiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "coordinated-destruction",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Experiment</th>\n",
       "      <th>mice</th>\n",
       "      <th>Sample Box</th>\n",
       "      <th>Fixation Box</th>\n",
       "      <th>Split-seq prep start date</th>\n",
       "      <th>barcoded [nuclei/ul] cell counter</th>\n",
       "      <th>Volume of barcoded nuclei</th>\n",
       "      <th>Total barcoded nuclei (1st 3 rounds)</th>\n",
       "      <th># of backup aliquots</th>\n",
       "      <th>size of backup aliquot</th>\n",
       "      <th>...</th>\n",
       "      <th>Novaseq1 L001</th>\n",
       "      <th>Novaseq1 L002</th>\n",
       "      <th>Novaseq1 L003</th>\n",
       "      <th>Novaseq1 L004</th>\n",
       "      <th>Novaseq2 L001</th>\n",
       "      <th>Novaseq2 L002</th>\n",
       "      <th>Novaseq2 L003</th>\n",
       "      <th>Novaseq2 L004</th>\n",
       "      <th>Unnamed: 52</th>\n",
       "      <th>Unnamed: 53</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IGVF_Splitseq_002</td>\n",
       "      <td>B6 (2F+2M)</td>\n",
       "      <td>IGVF_Sample_3</td>\n",
       "      <td>IGVF_FIX_001</td>\n",
       "      <td>2022-08-24</td>\n",
       "      <td>4540.0</td>\n",
       "      <td>240.0</td>\n",
       "      <td>1089600.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>67k</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CAST (2F+2M)</td>\n",
       "      <td>IGVF_Sample_4</td>\n",
       "      <td>IGVF_FIX_002</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Fixation V1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1M V1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGVF_FIX_007</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>13k</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>214 rows × 54 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Experiment          mice     Sample Box  Fixation Box  \\\n",
       "0    IGVF_Splitseq_002    B6 (2F+2M)  IGVF_Sample_3  IGVF_FIX_001   \n",
       "1                  NaN  CAST (2F+2M)  IGVF_Sample_4  IGVF_FIX_002   \n",
       "2          Fixation V1           NaN            NaN           NaN   \n",
       "3                1M V1           NaN            NaN           NaN   \n",
       "4                  NaN           NaN            NaN           NaN   \n",
       "..                 ...           ...            ...           ...   \n",
       "209                NaN           NaN            NaN  IGVF_FIX_007   \n",
       "210                NaN           NaN            NaN           NaN   \n",
       "211                NaN           NaN            NaN           NaN   \n",
       "212                NaN           NaN            NaN           NaN   \n",
       "213                NaN           NaN            NaN           NaN   \n",
       "\n",
       "    Split-seq prep start date  barcoded [nuclei/ul] cell counter  \\\n",
       "0                  2022-08-24                             4540.0   \n",
       "1                         NaT                                NaN   \n",
       "2                         NaT                                NaN   \n",
       "3                         NaT                                NaN   \n",
       "4                         NaT                                NaN   \n",
       "..                        ...                                ...   \n",
       "209                       NaT                                NaN   \n",
       "210                       NaT                                NaN   \n",
       "211                       NaT                                NaN   \n",
       "212                       NaT                                NaN   \n",
       "213                       NaT                                NaN   \n",
       "\n",
       "     Volume of barcoded nuclei  Total barcoded nuclei (1st 3 rounds)  \\\n",
       "0                        240.0                             1089600.0   \n",
       "1                          NaN                                   NaN   \n",
       "2                          NaN                                   NaN   \n",
       "3                          NaN                                   NaN   \n",
       "4                          NaN                                   NaN   \n",
       "..                         ...                                   ...   \n",
       "209                        NaN                                   NaN   \n",
       "210                        NaN                                   NaN   \n",
       "211                        NaN                                   NaN   \n",
       "212                        NaN                                   NaN   \n",
       "213                        NaN                                   NaN   \n",
       "\n",
       "     # of backup aliquots size of backup aliquot  ...  Novaseq1 L001  \\\n",
       "0                     3.0                    67k  ...            NaN   \n",
       "1                     NaN                    NaN  ...            NaN   \n",
       "2                     NaN                    NaN  ...            NaN   \n",
       "3                     NaN                    NaN  ...            NaN   \n",
       "4                     NaN                    NaN  ...            NaN   \n",
       "..                    ...                    ...  ...            ...   \n",
       "209                   2.0                    13k  ...            NaN   \n",
       "210                   NaN                    NaN  ...            NaN   \n",
       "211                   NaN                    NaN  ...            NaN   \n",
       "212                   NaN                    NaN  ...            NaN   \n",
       "213                   NaN                    NaN  ...            NaN   \n",
       "\n",
       "     Novaseq1 L002  Novaseq1 L003 Novaseq1 L004 Novaseq2 L001 Novaseq2 L002  \\\n",
       "0              NaN            NaN           NaN           NaN           NaN   \n",
       "1              NaN            NaN           NaN           NaN           NaN   \n",
       "2              NaN            NaN           NaN           NaN           NaN   \n",
       "3              NaN            NaN           NaN           NaN           NaN   \n",
       "4              NaN            NaN           NaN           NaN           NaN   \n",
       "..             ...            ...           ...           ...           ...   \n",
       "209            NaN            NaN           NaN           NaN           NaN   \n",
       "210            NaN            NaN           NaN           NaN           NaN   \n",
       "211            NaN            NaN           NaN           NaN           NaN   \n",
       "212            NaN            NaN           NaN           NaN           NaN   \n",
       "213            NaN            NaN           NaN           NaN           NaN   \n",
       "\n",
       "     Novaseq2 L003  Novaseq2 L004 Unnamed: 52  Unnamed: 53  \n",
       "0              NaN            NaN         NaN          NaN  \n",
       "1              NaN            NaN         NaN          NaN  \n",
       "2              NaN            NaN         NaN          NaN  \n",
       "3              NaN            NaN         NaN          NaN  \n",
       "4              NaN            NaN         NaN          NaN  \n",
       "..             ...            ...         ...          ...  \n",
       "209            NaN            NaN         NaN          NaN  \n",
       "210            NaN            NaN         NaN          NaN  \n",
       "211            NaN            NaN         NaN          NaN  \n",
       "212            NaN            NaN         NaN          NaN  \n",
       "213            NaN            NaN         NaN          NaN  \n",
       "\n",
       "[214 rows x 54 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "experiment = pandas.read_excel(\n",
    "    lizs_sheet_name, \n",
    "    sheet_name=\"Experiment\",\n",
    "    header=0,\n",
    "    index_col=None,\n",
    "    #usecols=range(0, 49)\n",
    ").rename(columns={\n",
    "    # Unnamed: 42 seems like an index or reference to the sample sheet?\n",
    "    \"Unnamed: 43\": \"Novaseq raw reads\",\n",
    "    \"Unnamed: 44\": \"Novaseq1 L001\",\n",
    "    \"Unnamed: 45\": \"Novaseq1 L002\",\n",
    "    \"Unnamed: 46\": \"Novaseq1 L003\",\n",
    "    \"Unnamed: 47\": \"Novaseq1 L004\",\n",
    "    \"Unnamed: 48\": \"Novaseq2 L001\",\n",
    "    \"Unnamed: 49\": \"Novaseq2 L002\",\n",
    "    \"Unnamed: 50\": \"Novaseq2 L003\",\n",
    "    \"Unnamed: 51\": \"Novaseq2 L004\",\n",
    "})\n",
    "\n",
    "experiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "gothic-robert",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IGVF_Splitseq_002 0 15\n",
      "IGVF_Splitseq_003 17 33\n",
      "IGVF_Splitseq_004 36 52\n",
      "IGVF_Splitseq_005 55 71\n",
      "IGVF_Splitseq_006 74 91\n",
      "IGVF_Splitseq_007 93 109\n",
      "IGVF_Splitseq_008 112 121\n",
      "IGVF_Splitseq_B01 124 131\n",
      "IGVF_Splitseq_008B 133 143\n",
      "IGVF_Splitseq_009 146 161\n",
      "IGVF_Splitseq_010 163 178\n",
      "IGVF_Splitseq_011 180 195\n",
      "IGVF_Splitseq_3,4,5,7 206 214\n"
     ]
    }
   ],
   "source": [
    "def find_fixation_start(sheet):\n",
    "    def is_experiment_name(name):\n",
    "        return not pandas.isnull(name) and name.startswith(\"IGVF_Splitseq\")\n",
    "\n",
    "    for fixation_id_row in sheet[sheet[\"Experiment\"].apply(is_experiment_name)].index:\n",
    "        name = sheet.loc[fixation_id_row, \"Experiment\"]\n",
    "        start = fixation_id_row\n",
    "\n",
    "        stop = sheet.shape[0]\n",
    "        for i, value in sheet.iloc[start:][\"sublibrary\"].iteritems():\n",
    "            if pandas.isnull(value):\n",
    "                stop = i - 1\n",
    "                break\n",
    "        \n",
    "        \n",
    "        yield (name, start, stop)\n",
    "\n",
    "for name, start, stop in find_fixation_start(experiment):\n",
    "    print(name, start, stop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "naval-muslim",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NO B01_13A\n",
      "NO B01_13B\n",
      "NO B01_13C\n",
      "NO B01_13D\n",
      "NO B01_13E\n",
      "NO B01_13F\n",
      "EX B01_13G\n",
      "NO B01_13H\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'experiment_name': 'IGVF_Splitseq_B01',\n",
       " 'plate_name': 'IGVF_B01',\n",
       " 'prep_date': Timestamp('2023-04-03 00:00:00'),\n",
       " 'unused_aliquots': {13000: 10},\n",
       " 'subpool': [{'name': 'B01_13A',\n",
       "   'subpool_subname': '13A',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionType.no_selection: 'NO'>,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul_in_25ul': 40,\n",
       "   'bioanalyzer_date': datetime.datetime(2023, 4, 5, 0, 0),\n",
       "   'cdna_average_bp_length': 1413,\n",
       "   'index_pcr_no': 11,\n",
       "   'index': 1,\n",
       "   'barcode': 'CAGATC',\n",
       "   'library_ng_per_ul': 25.2,\n",
       "   'library_average_bp_length': 437.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None},\n",
       "  {'name': 'B01_13B',\n",
       "   'subpool_subname': '13B',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionType.no_selection: 'NO'>,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul_in_25ul': 34.9,\n",
       "   'bioanalyzer_date': datetime.datetime(2023, 4, 5, 0, 0),\n",
       "   'cdna_average_bp_length': 1469,\n",
       "   'index_pcr_no': 11,\n",
       "   'index': 2,\n",
       "   'barcode': 'ACTTGA',\n",
       "   'library_ng_per_ul': 24.4,\n",
       "   'library_average_bp_length': 447.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None},\n",
       "  {'name': 'B01_13C',\n",
       "   'subpool_subname': '13C',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionType.no_selection: 'NO'>,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul_in_25ul': 41.4,\n",
       "   'bioanalyzer_date': datetime.datetime(2023, 4, 5, 0, 0),\n",
       "   'cdna_average_bp_length': 1424,\n",
       "   'index_pcr_no': 11,\n",
       "   'index': 3,\n",
       "   'barcode': 'GATCAG',\n",
       "   'library_ng_per_ul': 24.2,\n",
       "   'library_average_bp_length': 451.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None},\n",
       "  {'name': 'B01_13D',\n",
       "   'subpool_subname': '13D',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionType.no_selection: 'NO'>,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul_in_25ul': 43.8,\n",
       "   'bioanalyzer_date': datetime.datetime(2023, 4, 5, 0, 0),\n",
       "   'cdna_average_bp_length': 1352,\n",
       "   'index_pcr_no': 11,\n",
       "   'index': 4,\n",
       "   'barcode': 'TAGCTT',\n",
       "   'library_ng_per_ul': 27.2,\n",
       "   'library_average_bp_length': 449.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None},\n",
       "  {'name': 'B01_13E',\n",
       "   'subpool_subname': '13E',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionType.no_selection: 'NO'>,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul_in_25ul': 36,\n",
       "   'bioanalyzer_date': datetime.datetime(2023, 4, 5, 0, 0),\n",
       "   'cdna_average_bp_length': 1347,\n",
       "   'index_pcr_no': 11,\n",
       "   'index': 5,\n",
       "   'barcode': 'ATGTCA',\n",
       "   'library_ng_per_ul': 26.4,\n",
       "   'library_average_bp_length': 433.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None},\n",
       "  {'name': 'B01_13F',\n",
       "   'subpool_subname': '13F',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionType.no_selection: 'NO'>,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul_in_25ul': 44.5,\n",
       "   'bioanalyzer_date': datetime.datetime(2023, 4, 5, 0, 0),\n",
       "   'cdna_average_bp_length': 1388,\n",
       "   'index_pcr_no': 11,\n",
       "   'index': 6,\n",
       "   'barcode': 'CTTGTA',\n",
       "   'library_ng_per_ul': 28.2,\n",
       "   'library_average_bp_length': 457.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None},\n",
       "  {'name': 'B01_13G',\n",
       "   'subpool_subname': '13G',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionType.exome_capture: 'EX'>,\n",
       "   'cdna_pcr_rounds': '5 + 8',\n",
       "   'cdna_ng_per_ul_in_25ul': 48,\n",
       "   'bioanalyzer_date': datetime.datetime(2023, 4, 5, 0, 0),\n",
       "   'cdna_average_bp_length': 1473,\n",
       "   'index_pcr_no': 11,\n",
       "   'index': 7,\n",
       "   'barcode': 'AGTCAA',\n",
       "   'library_ng_per_ul': 31.2,\n",
       "   'library_average_bp_length': 444.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None},\n",
       "  {'name': 'B01_13H',\n",
       "   'subpool_subname': '13H',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionType.no_selection: 'NO'>,\n",
       "   'cdna_pcr_rounds': '5 + 8',\n",
       "   'cdna_ng_per_ul_in_25ul': 41.3,\n",
       "   'bioanalyzer_date': datetime.datetime(2023, 4, 5, 0, 0),\n",
       "   'cdna_average_bp_length': 1447,\n",
       "   'index_pcr_no': 11,\n",
       "   'index': 8,\n",
       "   'barcode': 'AGTTCC',\n",
       "   'library_ng_per_ul': 22.6,\n",
       "   'library_average_bp_length': 534.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None}]}"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def parse_selection_type(value):\n",
    "    if value is None:\n",
    "        return None\n",
    "    elif value.lower() in (\"normal\", \"no\"):\n",
    "        return models.SublibrarySelectionType.no_selection\n",
    "    elif value in (\"capture\", \"EX\"):\n",
    "        return models.SublibrarySelectionType.exome_capture\n",
    "    else:\n",
    "        raise ValueError(f\"Unrecognized capture type {value}\")\n",
    "\n",
    "\n",
    "def parse_fixation(sheet, experiment_name, start, stop):\n",
    "    block = sheet.loc[start:stop]\n",
    "\n",
    "    experiment_name_fields = experiment_name.split(\"_\")\n",
    "    plate_name = f\"{experiment_name_fields[0]}_{experiment_name_fields[-1]}\"\n",
    "    \n",
    "    aliquots = {}\n",
    "    for i, row in block.iterrows():\n",
    "        count = row[\"# of backup aliquots\"]\n",
    "        size = row[\"size of backup aliquot\"]\n",
    "        if not (pandas.isnull(count) or pandas.isnull(size)):\n",
    "            if isinstance(size, str) and size.endswith(\"k\"):\n",
    "                size = size[:-1]\n",
    "            size = int(size) * 1000\n",
    "            aliquots[size] = int(count)\n",
    "    \n",
    "    experiment = {\n",
    "        \"experiment_name\": experiment_name,\n",
    "        \"plate_name\": plate_name,\n",
    "        \"prep_date\": block.loc[start, \"Split-seq prep start date\"],\n",
    "        \"unused_aliquots\": aliquots,\n",
    "        \"subpool\": [],\n",
    "    }\n",
    "\n",
    "    for i, row in block.iterrows():\n",
    "        if not pandas.isnull(row[\"sublibrary\"]):\n",
    "            if pandas.isnull(row[\"NovaSeq Sequencing date\"]):\n",
    "                novaseq_run_id = None\n",
    "            else:\n",
    "                novaseq_run_id = row[\"NovaSeq Sequencing date\"].isoformat().split(\"T\")[0]\n",
    "\n",
    "            if pandas.isnull(row.get(\"Novaseq raw reads\")):\n",
    "                novaseq_raw_reads = None\n",
    "                novaseq1_raw_reads = None\n",
    "                novaseq2_raw_reads = None\n",
    "            elif row[\"Novaseq raw reads\"] == \"Total\":\n",
    "                novaseq_raw_reads = None\n",
    "                novaseq1_raw_reads = None\n",
    "                novaseq2_raw_reads = None\n",
    "            else:\n",
    "                novaseq_raw_reads = int_or_none(row.get(\"Novaseq raw reads\"))\n",
    "                novaseq1_raw_reads = sum([int_or_none(row.get(x)) for x in[\"Novaseq1 L001\", \"Novaseq1 L002\", \"Novaseq1 L003\", \"Novaseq1 L004\"]])\n",
    "                novaseq2_raw_reads = sum([int_or_none(row.get(x)) for x in[\"Novaseq2 L001\", \"Novaseq2 L002\", \"Novaseq2 L003\", \"Novaseq2 L004\"]])\n",
    "\n",
    "            print(row[\"sublibrary_selection_type\"], row[\"sublibrary\"])\n",
    "            subpool = {\n",
    "                \"name\": row[\"sublibrary\"],\n",
    "                \"subpool_subname\": row[\"sublibrary\"].split(\"_\")[1],\n",
    "                \"nuclei\": int_or_none(row[\"# of nuclei per sublibrary\"]),\n",
    "                \"selection_type\": parse_selection_type(row[\"sublibrary_selection_type\"]),\n",
    "                \"cdna_pcr_rounds\": row[\"cDNA amp # PCR cycles\"],\n",
    "                \"cdna_ng_per_ul_in_25ul\": row[\"[cDNA] ng/ul     in 25ul\"],\n",
    "                #total_cdna = cdna_ng_per_ul_in_25ul * 25\n",
    "                \"bioanalyzer_date\": row[\"BA date\"],\n",
    "                \"cdna_average_bp_length\": int_or_none(row[\"Bioanalyzer cDNA ave bp length\"]),\n",
    "                \"index_pcr_no\": int_or_none(row[\"Sub library Index PCR #\"]),\n",
    "                \"index\": int_or_none(row[\"Sublibrary Index\"]),\n",
    "                \"barcode\": row[\"Sequence for SampleSheet\"],\n",
    "                \"library_ng_per_ul\": float_or_none(row[\"[library] ng/ul\"]),\n",
    "                \"library_average_bp_length\": float_or_none(row[\"Bioanalyzer library ave bp length\"]),\n",
    "                \"nextseq_run_date\": date_or_none(row[\"QC Sequencing date\"]),\n",
    "                \"nextseq_run_id\": str_or_none(int_or_none(row[\"Nextseq run #\"])),\n",
    "                \"nextseq_run_raw_reads\": int_or_none(row[\"QC # raw reads  (1 mismatch)\"]),\n",
    "                \"novaseq_run_date\": date_or_none(row[\"NovaSeq Sequencing date\"]),\n",
    "                \"novaseq_run_id\": novaseq_run_id,\n",
    "                \"novaseq_run_raw_reads\": novaseq_raw_reads,\n",
    "                \"novaseq1_raw_reads\": novaseq1_raw_reads,\n",
    "                \"novaseq2_raw_reads\": novaseq2_raw_reads,\n",
    "            }\n",
    "            experiment[\"subpool\"].append(subpool)\n",
    "            #if not pandas.isnull(subpool[\"nuclei\"]):\n",
    "            #    nuclei = subpool[\"nuclei\"]\n",
    "            #    aliquots[nuclei] = aliquots.get(nuclei) + 1\n",
    "\n",
    "    return experiment\n",
    "\n",
    "#parse_fixation(experiment, \"IGVF_Splitseq_003\", 17, 32)\n",
    "#parse_fixation(experiment, \"IGVF_Splitseq_008\", 108, 117)\n",
    "parse_fixation(experiment, \"IGVF_Splitseq_B01\", 124, 131)\n",
    "#parse_fixation(experiment, \"IGVF_Splitseq_008B\", 133, 143)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dress-serbia",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "centered-float",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'nextseq2000': <Platform: Nextseq 2000>,\n",
       " 'novaseq6000': <Platform: Novaseq 6000>,\n",
       " 'nanopore': <Platform: Oxford Nanopore>,\n",
       " 'pacbio': <Platform: Pac Bio>}"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "platforms = {x.name: x for x in models.Platform.objects.all()}\n",
    "platforms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "express-zambia",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8ACZ\n",
      "decoded [8, 0, 2, 25]\n",
      "encoded 8ACZ\n",
      "8ADA\n",
      "8A 8B True False False\n",
      "8AA 8Z False False True\n",
      "8B 8E True False False\n",
      "8B 8A False False True\n",
      "8Z 8AA True False False\n",
      "67A 67A False True False\n",
      "8AA 8AB True False False\n",
      "8B\n",
      "8B\n",
      "8A\n"
     ]
    }
   ],
   "source": [
    "class SubpoolName:\n",
    "    def __init__(self, name=None):\n",
    "        if isinstance(name, str):\n",
    "            self._normalized = SubpoolName.subpool_name_to_numbers(name)\n",
    "        elif isinstance(name, list):\n",
    "            self._normalized = name\n",
    "        elif name is None:\n",
    "            self._normalized = None\n",
    "        else:\n",
    "            raise ValueError(\"Unrecognized type {} for {}\".format(type(name), name))\n",
    "\n",
    "    def __str__(self):\n",
    "        if self._normalized is None:\n",
    "            return \"None\"\n",
    "        else:\n",
    "            return SubpoolName.numbers_to_subpool_name(self._normalized)\n",
    "        \n",
    "    def __repr__(self):\n",
    "        if self._normalized is None:\n",
    "            return str(\"{}()\".format(self.__class__.__name__))\n",
    "        else:\n",
    "            return str(\"{}('{}')\".format(self.__class__.__name__, str(self)))\n",
    "        \n",
    "    def __eq__(self, cmp):\n",
    "        if self._normalized is None and cmp is None:\n",
    "            return True\n",
    "        elif self._normalized is None and cmp._normalized is None:\n",
    "            return True\n",
    "        elif self._normalized is None or cmp._normalized is None:\n",
    "            return False\n",
    "        elif len(self._normalized) != len(cmp._normalized):\n",
    "            return False\n",
    "        else:\n",
    "            for left, right in zip(self._normalized, cmp._normalized):\n",
    "                if left != right:\n",
    "                    return False\n",
    "                \n",
    "        return True\n",
    "    \n",
    "    def __lt__(self, cmp):\n",
    "        if self._normalized is None and cmp is None:\n",
    "            return False\n",
    "        elif self._normalized is None and cmp._normalized is None:\n",
    "            return False\n",
    "        elif self._normalized is None:\n",
    "            return True\n",
    "        elif cmp is None or cmp._normalized is None:\n",
    "            return False\n",
    "        elif self._normalized[0] < cmp._normalized[0]:\n",
    "            return True\n",
    "        elif len(self._normalized) < len(cmp._normalized):\n",
    "            return True\n",
    "        elif len(self._normalized) > len(cmp._normalized):\n",
    "            return False\n",
    "        else:\n",
    "            for left, right in zip(self._normalized, cmp._normalized):\n",
    "                if left == right:\n",
    "                    continue\n",
    "                elif left < right:\n",
    "                    return True\n",
    "                else:\n",
    "                    return False\n",
    "                \n",
    "        return False\n",
    "                \n",
    "    def __gt__(self, cmp):\n",
    "        if self._normalized is None and cmp is None:\n",
    "            return False\n",
    "        elif self._normalized is None and cmp._normalized is None:\n",
    "            return False\n",
    "        elif cmp is None or cmp._normalized is None:\n",
    "            return True\n",
    "        elif self._normalized[0] > cmp._normalized[0]:\n",
    "            return True\n",
    "        elif len(self._normalized) < len(cmp._normalized):\n",
    "            return False\n",
    "        elif len(self._normalized) > len(cmp._normalized):\n",
    "            return True\n",
    "        else:\n",
    "            for left, right in zip(self._normalized, cmp._normalized):\n",
    "                if left == right:\n",
    "                    continue\n",
    "                elif left < right:\n",
    "                    return False\n",
    "                else:\n",
    "                    return True\n",
    "            \n",
    "        return False\n",
    "        \n",
    "    @classmethod\n",
    "    def subpool_name_to_numbers(cls, name):\n",
    "        STATE_DIGITS = 1\n",
    "        STATE_LETTERS = 2\n",
    "        STATE_ERROR = -1\n",
    "\n",
    "        digits = []\n",
    "        letters = []\n",
    "        state = STATE_DIGITS\n",
    "        for char in name:\n",
    "            if state == STATE_DIGITS:\n",
    "                if char.isdigit():\n",
    "                    digits.append(char)\n",
    "                elif char.isalpha():\n",
    "                    state = STATE_LETTERS\n",
    "                    letters.append(char)\n",
    "                else:\n",
    "                    raise ValueError(\"Unrecognized symbol {} in {}\".format(char, name))\n",
    "            elif state == STATE_LETTERS:\n",
    "                if char.isdigit():\n",
    "                    raise ValueError(\"Unexpected digit {} in {}\".format(char, name))\n",
    "                elif char.isalpha():\n",
    "                    letters.append(char)\n",
    "                else:\n",
    "                    raise ValueError(\"Unrecognized symbol {} in {}\".format(char, name))\n",
    "\n",
    "        result = [int(\"\".join(digits))]\n",
    "        for number in letters:\n",
    "            if number < \"A\" or number > \"Z\":\n",
    "                raise ValueError(\"out of bounds letter code in {}\".format(name))\n",
    "            else:\n",
    "                result.append(ord(number) - ord(\"A\"))\n",
    "\n",
    "        return result\n",
    "\n",
    "    @classmethod\n",
    "    def numbers_to_subpool_name(cls, subpool_list):\n",
    "\n",
    "        result = [str(subpool_list[0])]\n",
    "        for char in subpool_list[1:]:\n",
    "            result.append(chr(char + ord(\"A\")))\n",
    "\n",
    "        return \"\".join(result)\n",
    "\n",
    "    def next_code(self):\n",
    "        if self._normalized is None:\n",
    "            return None\n",
    "\n",
    "        carry = 0\n",
    "        advancable = []\n",
    "        for i, place in enumerate(reversed(self._normalized[1:])):\n",
    "            if i == 0:\n",
    "                place += 1\n",
    "            else:\n",
    "                place += carry\n",
    "                carry = 0\n",
    "\n",
    "            if place > 25:\n",
    "                carry = 1\n",
    "                place = 0\n",
    "\n",
    "            advancable.append(place)\n",
    "\n",
    "        if carry > 0:\n",
    "            advancable.append(0)\n",
    "\n",
    "        result = [self._normalized[0]]\n",
    "        result.extend(reversed(advancable))\n",
    "        \n",
    "        return SubpoolName(result)\n",
    "\n",
    "s = SubpoolName(\"8ACZ\")\n",
    "print(s)\n",
    "\n",
    "decoded = SubpoolName.subpool_name_to_numbers(\"8ACZ\") \n",
    "print(\"decoded\", decoded)\n",
    "encoded = SubpoolName.numbers_to_subpool_name(decoded)\n",
    "print(\"encoded\", encoded)\n",
    "\n",
    "print(SubpoolName(\"8ACZ\").next_code())\n",
    "\n",
    "\n",
    "\n",
    "for pairs in [(\"8A\", \"8B\"), (\"8AA\", \"8Z\"), (\"8B\", \"8E\"), (\"8B\", \"8A\"), (\"8Z\", \"8AA\"), (\"67A\", \"67A\"), (\"8AA\", \"8AB\")]:\n",
    "    left = SubpoolName(pairs[0])\n",
    "    right = SubpoolName(pairs[1])\n",
    "    print(left, right, left < right, left == right, left > right)\n",
    "    \n",
    "print(max(SubpoolName(\"8A\"), SubpoolName(\"8B\")))\n",
    "print(max(None, SubpoolName(\"8B\")))\n",
    "print(max(SubpoolName(\"8A\"), None))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "parliamentary-pendant",
   "metadata": {},
   "source": [
    "# load subpools, sequencing runs and subpools in sequencing run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "hydraulic-nurse",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8\n"
     ]
    }
   ],
   "source": [
    "multiplexed_samples = {}\n",
    "\n",
    "for i, row in submitted_sheets[\"multiplexed_sample\"].iterrows():\n",
    "    if not pandas.isnull(row[\"accession\"]):\n",
    "        aliases = row[\"aliases:array\"].split(',')\n",
    "        name = aliases[0].replace(\"ali-mortazavi:\", \"\")\n",
    "\n",
    "        multiplexed_samples[name] = {\n",
    "            \"namespace\": igvf_test_namespace,\n",
    "            \"name\": row[\"accession\"],\n",
    "            \"uuid\": row[\"uuid\"],\n",
    "            \"url\": \"https://api.sandbox.igvf.org/multiplexed-samples/{}/\".format(row[\"accession\"]),\n",
    "        }\n",
    "print(len(multiplexed_samples))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "victorian-basics",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NO 002_4A\n",
      "NO 002_4B\n",
      "NO 002_64A\n",
      "NO 002_64B\n",
      "NO 002_64C\n",
      "NO 002_64D\n",
      "NO 002_64E\n",
      "NO 002_64F\n",
      "NO 002_64G\n",
      "NO 002_64H\n",
      "NO 002_64I\n",
      "NO 002_64J\n",
      "NO 002_64K\n",
      "NO 002_64L\n",
      "NO 002_64M\n",
      "NO 002_64N\n",
      "Unable to find plate IGVF_002\n",
      "NO 003_8A\n",
      "NO 003_13A\n",
      "NO 003_67A\n",
      "NO 003_67B\n",
      "NO 003_67C\n",
      "NO 003_67D\n",
      "NO 003_67E\n",
      "NO 003_67F\n",
      "NO 003_67G\n",
      "NO 003_67H\n",
      "NO 003_67I\n",
      "NO 003_67J\n",
      "NO 003_67K\n",
      "NO 003_67L\n",
      "NO 003_67M\n",
      "NO 003_67N\n",
      "NO 003_67O\n",
      "003_67P\n",
      "003_67Q\n",
      "003_67R\n",
      "003_67S\n",
      "003_67T\n",
      "003_67U\n",
      "003_67V\n",
      "003_13B\n",
      "003_8B\n",
      "003_8C\n",
      "NO 004_8A\n",
      "NO 004_13A\n",
      "NO 004_67A\n",
      "NO 004_67B\n",
      "NO 004_67C\n",
      "NO 004_67D\n",
      "NO 004_67E\n",
      "NO 004_67F\n",
      "NO 004_67G\n",
      "NO 004_67H\n",
      "NO 004_67I\n",
      "NO 004_67J\n",
      "NO 004_67K\n",
      "NO 004_67L\n",
      "NO 004_67M\n",
      "NO 004_67N\n",
      "NO 004_67O\n",
      "004_67P\n",
      "004_67Q\n",
      "004_67R\n",
      "004_67S\n",
      "004_67T\n",
      "004_13B\n",
      "004_13C\n",
      "004_8B\n",
      "004_8C\n",
      "NO 005_8A\n",
      "NO 005_13A\n",
      "NO 005_67A\n",
      "NO 005_67B\n",
      "NO 005_67C\n",
      "NO 005_67D\n",
      "NO 005_67E\n",
      "NO 005_67F\n",
      "NO 005_67G\n",
      "NO 005_67H\n",
      "NO 005_67I\n",
      "NO 005_67J\n",
      "NO 005_67K\n",
      "NO 005_67L\n",
      "NO 005_67M\n",
      "NO 005_67N\n",
      "NO 005_67O\n",
      "005_67P\n",
      "005_67Q\n",
      "005_67R\n",
      "005_67S\n",
      "005_67T\n",
      "005_67U\n",
      "005_13B\n",
      "005_13C\n",
      "005_8B\n",
      "005_8C\n",
      "005_8D\n",
      "005_8E\n",
      "005_8F\n",
      "005_8G\n",
      "005_8H\n",
      "NO 006_8A\n",
      "NO 006_67A\n",
      "NO 006_67B\n",
      "NO 006_67C\n",
      "NO 006_67D\n",
      "NO 006_67E\n",
      "NO 006_67F\n",
      "NO 006_67G\n",
      "NO 006_67H\n",
      "NO 006_67I\n",
      "NO 006_67J\n",
      "NO 006_67K\n",
      "NO 006_67L\n",
      "NO 006_67M\n",
      "NO 006_67N\n",
      "NO 006_67O\n",
      "NO 006_67P\n",
      "NO 006_67Q\n",
      "Skipping failed plate IGVF_006\n",
      "NO 007_8A\n",
      "NO 007_13A\n",
      "NO 007_67A\n",
      "NO 007_67B\n",
      "NO 007_67C\n",
      "NO 007_67D\n",
      "NO 007_67E\n",
      "NO 007_67F\n",
      "NO 007_67G\n",
      "NO 007_67H\n",
      "NO 007_67I\n",
      "NO 007_67J\n",
      "NO 007_67K\n",
      "NO 007_67L\n",
      "NO 007_67M\n",
      "NO 007_67N\n",
      "NO 007_67O\n",
      "007_67P\n",
      "007_67Q\n",
      "007_67R\n",
      "007_67S\n",
      "007_67T\n",
      "007_67U\n",
      "007_67V\n",
      "007_67W\n",
      "007_13B\n",
      "007_13C\n",
      "007_8B\n",
      "007_8C\n",
      "007_8D\n",
      "007_8E\n",
      "007_8F\n",
      "007_8G\n",
      "007_8H\n",
      "007_8I\n",
      "NO 008_8A\n",
      "NO 008_67A\n",
      "NO 008_67B\n",
      "NO 008_67C\n",
      "NO 008_67D\n",
      "NO 008_67E\n",
      "NO 008_67F\n",
      "NO 008_67G\n",
      "NO 008_67H\n",
      "NO 008_67I\n",
      "NO B01_13A\n",
      "NO B01_13B\n",
      "NO B01_13C\n",
      "NO B01_13D\n",
      "NO B01_13E\n",
      "NO B01_13F\n",
      "EX B01_13G\n",
      "NO B01_13H\n",
      "subpool_b01_13a B01_13A {'namespace': <AccessionNamespace: IGVF test>, 'name': 'TSTSM35366348', 'uuid': '77a93b75-f44a-4aa5-937b-a6788051bc18', 'url': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM35366348/'}\n",
      "subpool_b01_13b B01_13B {'namespace': <AccessionNamespace: IGVF test>, 'name': 'TSTSM99018100', 'uuid': '6cac6a8b-7750-4df6-92cf-fae41082813d', 'url': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM99018100/'}\n",
      "subpool_b01_13c B01_13C {'namespace': <AccessionNamespace: IGVF test>, 'name': 'TSTSM36900951', 'uuid': '01858fa2-3c3d-476a-8d81-e82a0612c0e8', 'url': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM36900951/'}\n",
      "subpool_b01_13d B01_13D {'namespace': <AccessionNamespace: IGVF test>, 'name': 'TSTSM27245068', 'uuid': 'b69b021b-aa90-41cf-8d1f-cfc59c726b20', 'url': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM27245068/'}\n",
      "subpool_b01_13e B01_13E {'namespace': <AccessionNamespace: IGVF test>, 'name': 'TSTSM90710268', 'uuid': '8d44590d-bcb5-46f9-8c87-ea375db12ea9', 'url': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM90710268/'}\n",
      "subpool_b01_13f B01_13F {'namespace': <AccessionNamespace: IGVF test>, 'name': 'TSTSM57985101', 'uuid': '2f34272a-6603-41e0-9820-cf2dd432244b', 'url': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM57985101/'}\n",
      "subpool_b01_13g B01_13G {'namespace': <AccessionNamespace: IGVF test>, 'name': 'TSTSM10953505', 'uuid': 'b0e63479-2caf-4aac-9c98-2ff823f92c74', 'url': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM10953505/'}\n",
      "subpool_b01_13h B01_13H {'namespace': <AccessionNamespace: IGVF test>, 'name': 'TSTSM89995437', 'uuid': '5484bec6-4b0e-4488-b3b1-fd4def0022df', 'url': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM89995437/'}\n",
      "B01_13I\n",
      "B01_13J\n",
      "B01_13K\n",
      "B01_13L\n",
      "B01_13M\n",
      "B01_13N\n",
      "B01_13O\n",
      "B01_13P\n",
      "B01_13Q\n",
      "B01_13R\n",
      "EX 008B_13A\n",
      "NO 008B_67A\n",
      "NO 008B_67B\n",
      "NO 008B_67C\n",
      "NO 008B_67D\n",
      "NO 008B_67E\n",
      "NO 008B_67F\n",
      "NO 008B_67G\n",
      "NO 008B_67H\n",
      "NO 008B_67I\n",
      "NO 008B_67J\n",
      "Unable to find plate IGVF_008B\n",
      "EX 009_13A\n",
      "NO 009_67A\n",
      "NO 009_67B\n",
      "NO 009_67C\n",
      "NO 009_67D\n",
      "NO 009_67E\n",
      "NO 009_67F\n",
      "NO 009_67G\n",
      "NO 009_67H\n",
      "NO 009_67I\n",
      "NO 009_67J\n",
      "NO 009_67K\n",
      "NO 009_67L\n",
      "NO 009_67M\n",
      "NO 009_67N\n",
      "NO 009_67O\n",
      "EX 010_13A\n",
      "NO 010_67A\n",
      "NO 010_67B\n",
      "NO 010_67C\n",
      "NO 010_67D\n",
      "NO 010_67E\n",
      "NO 010_67F\n",
      "NO 010_67G\n",
      "NO 010_67H\n",
      "NO 010_67I\n",
      "NO 010_67J\n",
      "NO 010_67K\n",
      "NO 010_67L\n",
      "NO 010_67M\n",
      "NO 010_67N\n",
      "NO 010_67O\n",
      "010_67P\n",
      "010_67Q\n",
      "010_67R\n",
      "010_13B\n",
      "010_13C\n",
      "010_13D\n",
      "010_13E\n",
      "EX 011_13A\n",
      "NO 011_67A\n",
      "NO 011_67B\n",
      "NO 011_67C\n",
      "NO 011_67D\n",
      "NO 011_67E\n",
      "NO 011_67F\n",
      "NO 011_67G\n",
      "NO 011_67H\n",
      "NO 011_67I\n",
      "NO 011_67J\n",
      "NO 011_67K\n",
      "NO 011_67L\n",
      "NO 011_67M\n",
      "NO 011_67N\n",
      "NO 011_67O\n",
      "011_67P\n",
      "011_67Q\n",
      "011_67R\n",
      "011_67S\n",
      "011_67T\n",
      "011_67U\n",
      "011_67V\n",
      "011_13B\n",
      "011_13C\n",
      "011_13D\n",
      "011_13E\n",
      "011_13F\n",
      "011_13G\n",
      "011_13H\n",
      "EX 003_13A\n",
      "EX 004_13A\n",
      "EX 005_13A\n",
      "EX 007_13A\n",
      "EX 008B_13A\n",
      "EX 009_13A\n",
      "EX 010_13A\n",
      "EX 011_13A\n",
      "Unable to find plate IGVF_3,4,5,7\n"
     ]
    }
   ],
   "source": [
    "compression=r\"(?<compression>gz|bz2|xz|zstd)\"\n",
    "nanopore_library_name_re_template = r\"igvf(?P<run>[\\d]+)_(?P<library_id>{library_name})_.*_(?P<fragment>[\\d]+)\\.fastq\\.{compression}\"\n",
    "nextseq_library_name_re_template = r\"(?P<run>[\\d]+)_(?P<library_id>{library_name})_(?P<read_id>[RI][\\d])\\.fastq\\.{compression}\"\n",
    "nextseq_library_index_re_template = r\"Sublibrary_(?P<library_id>{library_index})_S(?P<sample_id>[\\d]+)_(?P<lane_id>L[\\d]+)_(?P<read_id>[RI][\\d])_(?P<fragment>[\\d]+)\\.fastq\\.{compression}\"\n",
    "novaseq_library_index_re_template = r\"Sublibrary_(?P<library_id>{library_index})_S(?P<sample_id>[\\d]+)_(?P<lane_id>L[\\d]+)_(?P<read_id>[RI][\\d])_(?P<fragment>[\\d]+)\\.fastq\\.{compression}\"\n",
    "\n",
    "expected_names = {\n",
    "    \"IGVF_Splitseq_003\": {\n",
    "        \"nanopore\": nanopore_library_name_re_template,\n",
    "        \"nextseq\": nextseq_library_name_re_template,\n",
    "        \"novaseq\": novaseq_library_index_re_template,\n",
    "    },\n",
    "    \"IGVF_Splitseq_004\": {\n",
    "        \"nanopore\": nanopore_library_name_re_template,\n",
    "        \"nextseq\": nextseq_library_name_re_template,\n",
    "        \"novaseq\": novaseq_library_index_re_template,\n",
    "    },\n",
    "    \"IGVF_Splitseq_005\": {\n",
    "        \"nanopore\": nanopore_library_name_re_template,\n",
    "        \"nextseq\": nextseq_library_index_re_template,\n",
    "        \"novaseq\": novaseq_library_index_re_template,\n",
    "    },\n",
    "    \"IGVF_Splitseq_006\": {\n",
    "        \"nanopore\": nanopore_library_name_re_template,\n",
    "        \"nextseq\": nextseq_library_index_re_template,\n",
    "        \"novaseq\": novaseq_library_index_re_template,\n",
    "    },\n",
    "    \n",
    "}\n",
    "\n",
    "nanopore_runs = {\n",
    "    \"IGVF_Splitseq_003\": {\n",
    "        \"name\": \"2023-01-03-8A\",\n",
    "        \"run_date\": datetime.datetime(2023,1,3),\n",
    "        \"raw_reads\": 87784162,\n",
    "    }\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "failed_plates = {\"IGVF_006\",}\n",
    "failed = False\n",
    "for experiment_name, experiment_start, experiment_stop in find_fixation_start(experiment):\n",
    "    fixation = parse_fixation(experiment, experiment_name, experiment_start, experiment_stop)\n",
    "    \n",
    "    try:\n",
    "        plate = models.SplitSeqPlate.objects.get(name=fixation[\"plate_name\"])\n",
    "    except models.SplitSeqPlate.DoesNotExist:\n",
    "        print(f\"Unable to find plate {fixation['plate_name']}\")\n",
    "        continue\n",
    "        \n",
    "    if fixation[\"plate_name\"] in failed_plates:\n",
    "        print(f\"Skipping failed plate {fixation['plate_name']}\")\n",
    "        continue\n",
    "        \n",
    "    last_subpools = {}\n",
    "    \n",
    "    wt_mega_2_kit = models.LibraryConstructionKit.objects.get(name=\"WT Mega\", version=\"v2\")\n",
    "    for line_offset, subpool in enumerate(fixation[\"subpool\"]):\n",
    "        try:\n",
    "            barcodes = models.LibraryBarcode.objects.filter(\n",
    "                kit=wt_mega_2_kit, \n",
    "                barcode_type=None, \n",
    "                code=subpool[\"index\"])\n",
    "        except models.LibraryBarcode.DoesNotExist:\n",
    "            print(f\"Unable to find barcode {subpool['index']}\")\n",
    "\n",
    "        # validate barcodes\n",
    "        library_barcodes_sequence = {b.sequence for b in barcodes}\n",
    "        if not pandas.isnull(subpool[\"barcode\"]):\n",
    "            expected_sequence = set(subpool[\"barcode\"].split(\",\"))\n",
    "        else:\n",
    "            expected_sequence = set()\n",
    "        if library_barcodes_sequence != expected_sequence:\n",
    "            print(f\"{subpool['name']} Database lookup of barcodes {library_barcodes_sequence} doesn't match human entry {expected_sequence}\")\n",
    "            failed = True\n",
    "            continue\n",
    "        \n",
    "        subpool_prefix = subpool[\"name\"].split(\"_\")[0]\n",
    "        last_subpools[subpool[\"nuclei\"]] = max(\n",
    "            last_subpools.get(subpool[\"nuclei\"]), \n",
    "            SubpoolName(subpool[\"subpool_subname\"]))\n",
    "\n",
    "        record = models.Subpool(\n",
    "            name=subpool[\"name\"],\n",
    "            plate=plate,\n",
    "            nuclei=subpool[\"nuclei\"],\n",
    "            selection_type=subpool[\"selection_type\"],\n",
    "            cdna_pcr_rounds=subpool[\"cdna_pcr_rounds\"],\n",
    "            cdna_ng_per_ul_in_25ul=float_or_none(subpool[\"cdna_ng_per_ul_in_25ul\"]),\n",
    "            bioanalyzer_date=date_or_none(subpool[\"bioanalyzer_date\"]),\n",
    "            cdna_average_bp_length=int_or_none(subpool[\"cdna_average_bp_length\"]),\n",
    "            index_pcr_number=int_or_none(subpool[\"index_pcr_no\"]),\n",
    "            index=int_or_none(subpool[\"index\"]),\n",
    "            library_ng_per_ul=float_or_none(subpool[\"library_ng_per_ul\"]),\n",
    "            library_average_bp_length=int_or_none(subpool[\"library_average_bp_length\"]),\n",
    "        )\n",
    "        record.save()\n",
    "        record.barcode.set(barcodes)\n",
    "        record.save()\n",
    "        \n",
    "        name = \"subpool_{}\".format(subpool[\"name\"].lower())\n",
    "        if name in multiplexed_samples:\n",
    "            print(name, record.name, multiplexed_samples[name])\n",
    "            accession = models.Accession(**multiplexed_samples[name])\n",
    "            accession.save()\n",
    "            record.accession.set([accession])\n",
    "            record.save()            \n",
    "        \n",
    "    for nuclei in fixation[\"unused_aliquots\"]:\n",
    "        subpool_name = last_subpools[nuclei]\n",
    "        for i in range(fixation[\"unused_aliquots\"][nuclei]):\n",
    "            subpool_name = subpool_name.next_code()\n",
    "\n",
    "            record = models.Subpool(\n",
    "                name=\"{}_{}\".format(subpool_prefix, subpool_name),\n",
    "                plate=plate,\n",
    "                nuclei=nuclei,\n",
    "            )\n",
    "            print(record.name)\n",
    "            record.save()\n",
    "\n",
    "assert not failed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "sudden-illustration",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Old creating runs from the sample sheet\n",
    "\n",
    "#        if experiment_name in nanopore_runs and subpool[\"nuclei\"] == 8000:\n",
    "#            measurement = get_or_create_sequencing_run(\n",
    "#                # nanopore_runs[experiment_name][\"name\"]\n",
    "#                name=\"{}/{}\".format(plate.name.lower(), \"nanopore\"),\n",
    "#                run_date=nanopore_runs[experiment_name][\"run_date\"],\n",
    "#                platform=platforms[\"nanopore\"],\n",
    "#                plate=plate,\n",
    "#            )\n",
    "#            \n",
    "#            pattern = expected_names[experiment_name][\"nanopore\"].format(\n",
    "#                library_name=subpool[\"subpool_subname\"],\n",
    "#                library_index=line_offset+1,\n",
    "#                compression=compression,\n",
    "#            )\n",
    "#            \n",
    "#            run = models.SubpoolInRun(\n",
    "#                subpool=record,\n",
    "#                sequencing_run=measurement,\n",
    "#                raw_reads=nanopore_runs[experiment_name][\"raw_reads\"],\n",
    "#                status=models.RunStatus.PASS,\n",
    "#            )\n",
    "#            run.save()            \n",
    "#\n",
    "#        if subpool[\"nextseq_run_date\"] is not None:\n",
    "#            measurement = get_or_create_sequencing_run(\n",
    "#                #sublibrary[\"nextseq_run_id\"]\n",
    "#                name=\"{}/{}\".format(plate.name.lower(), \"nextseq\"),\n",
    "#                run_date=subpool[\"nextseq_run_date\"],\n",
    "#                platform=platforms[\"nextseq\"],\n",
    "#                plate=plate,\n",
    "#            )\n",
    "#\n",
    "#            pattern = expected_names[experiment_name][\"nextseq\"].format(\n",
    "#                library_name=subpool[\"subpool_subname\"],\n",
    "#                library_index=line_offset+1,\n",
    "#                compression=compression,\n",
    "#            )            \n",
    "#            \n",
    "#            run = models.SubpoolInRun(\n",
    "#                subpool=record,\n",
    "#                sequencing_run=measurement,\n",
    "#                raw_reads=subpool[\"nextseq_run_raw_reads\"],\n",
    "#                status=models.RunStatus.PASS,\n",
    "#            )\n",
    "#            run.save()\n",
    "#            \n",
    "#        if subpool[\"novaseq_run_date\"] is not None:\n",
    "#            for i, nova_raw_reads in enumerate([subpool[\"novaseq1_raw_reads\"], subpool[\"novaseq2_raw_reads\"]]):\n",
    "#                #nova_run_name = \"{}-{}\".format(sublibrary[\"novaseq_run_id\"], i+1)\n",
    "#                nova_run_name = \"{}/{}\".format(plate.name.lower(), \"nova{}\".format(i+1))\n",
    "#                measurement = get_or_create_sequencing_run(\n",
    "#                    name=nova_run_name,\n",
    "#                    run_date=subpool[\"novaseq_run_date\"],\n",
    "#                    platform=platforms[\"novaseq\"],\n",
    "#                    plate=plate,\n",
    "#                )\n",
    "#\n",
    "#                pattern = expected_names[experiment_name][\"novaseq\"].format(\n",
    "#                    library_name=subpool[\"subpool_subname\"],\n",
    "#                    library_index=line_offset+1,\n",
    "#                    compression=compression,\n",
    "#                )\n",
    "#\n",
    "#                run = models.SubpoolInRun(\n",
    "#                    subpool=record,\n",
    "#                    sequencing_run=measurement,\n",
    "#                    raw_reads=nova_raw_reads,\n",
    "#                    status=models.RunStatus.PASS,                    \n",
    "#                )\n",
    "#                run.save()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "choice-error",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>experiment</th>\n",
       "      <th>filename</th>\n",
       "      <th>md5sum</th>\n",
       "      <th>size</th>\n",
       "      <th>ctime</th>\n",
       "      <th>library_id</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>lane_id</th>\n",
       "      <th>read_id</th>\n",
       "      <th>fragment</th>\n",
       "      <th>compression</th>\n",
       "      <th>sequencer</th>\n",
       "      <th>sequencer_run</th>\n",
       "      <th>flowcell_id</th>\n",
       "      <th>lane</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>igvf_003</td>\n",
       "      <td>igvf_003/nextseq/003_8A_R1.fastq.gz</td>\n",
       "      <td>bb696e63b6a8710ec8435b60ef6c81f8</td>\n",
       "      <td>8886302014</td>\n",
       "      <td>2023-02-07</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gz</td>\n",
       "      <td>VH00582</td>\n",
       "      <td>1</td>\n",
       "      <td>AAATMGFHV</td>\n",
       "      <td>1</td>\n",
       "      <td>CAGATC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>igvf_003</td>\n",
       "      <td>igvf_003/nextseq/003_8A_R2.fastq.gz</td>\n",
       "      <td>21967ba1a5a3218508961fac482a2952</td>\n",
       "      <td>5866415324</td>\n",
       "      <td>2023-02-07</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gz</td>\n",
       "      <td>VH00582</td>\n",
       "      <td>1</td>\n",
       "      <td>AAATMGFHV</td>\n",
       "      <td>1</td>\n",
       "      <td>CAGATC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>igvf_003</td>\n",
       "      <td>igvf_003/nanopore/igvf003_8A_lig-ss_1.fastq.gz</td>\n",
       "      <td>3cc7eaba69775ece39914a7b4e58de8d</td>\n",
       "      <td>5479439080</td>\n",
       "      <td>2022-12-06</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>gz</td>\n",
       "      <td>nanopore</td>\n",
       "      <td>ffc0d811f8ac1aafcd29643e65cf87acf848891f</td>\n",
       "      <td>FAU11962</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>igvf_003</td>\n",
       "      <td>igvf_003/nanopore/igvf003_8A_lig-ss_10.fastq.gz</td>\n",
       "      <td>f637c61c5c3df12dc946d2d32c0bdace</td>\n",
       "      <td>7997759697</td>\n",
       "      <td>2022-12-13</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>gz</td>\n",
       "      <td>nanopore</td>\n",
       "      <td>7d6a038266e5875db51e462422179436e6c44b3a</td>\n",
       "      <td>FAV21462</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>igvf_003</td>\n",
       "      <td>igvf_003/nanopore/igvf003_8A_lig-ss_11.fastq.gz</td>\n",
       "      <td>494bd18e8eb9bcef5b79f18f8f21d569</td>\n",
       "      <td>7241241549</td>\n",
       "      <td>2022-12-13</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "      <td>gz</td>\n",
       "      <td>nanopore</td>\n",
       "      <td>e0fe4672609d60d67277a22973f5c8534df5fc4c</td>\n",
       "      <td>FAU06484</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   experiment                                         filename  \\\n",
       "15   igvf_003              igvf_003/nextseq/003_8A_R1.fastq.gz   \n",
       "16   igvf_003              igvf_003/nextseq/003_8A_R2.fastq.gz   \n",
       "0    igvf_003   igvf_003/nanopore/igvf003_8A_lig-ss_1.fastq.gz   \n",
       "1    igvf_003  igvf_003/nanopore/igvf003_8A_lig-ss_10.fastq.gz   \n",
       "2    igvf_003  igvf_003/nanopore/igvf003_8A_lig-ss_11.fastq.gz   \n",
       "\n",
       "                              md5sum        size       ctime  library_id  \\\n",
       "15  bb696e63b6a8710ec8435b60ef6c81f8  8886302014  2023-02-07           1   \n",
       "16  21967ba1a5a3218508961fac482a2952  5866415324  2023-02-07           1   \n",
       "0   3cc7eaba69775ece39914a7b4e58de8d  5479439080  2022-12-06           1   \n",
       "1   f637c61c5c3df12dc946d2d32c0bdace  7997759697  2022-12-13           1   \n",
       "2   494bd18e8eb9bcef5b79f18f8f21d569  7241241549  2022-12-13           1   \n",
       "\n",
       "   sample_id lane_id read_id fragment compression sequencer  \\\n",
       "15       NaN     NaN      R1      NaN          gz   VH00582   \n",
       "16       NaN     NaN      R2      NaN          gz   VH00582   \n",
       "0        NaN     NaN     NaN        1          gz  nanopore   \n",
       "1        NaN     NaN     NaN       10          gz  nanopore   \n",
       "2        NaN     NaN     NaN       11          gz  nanopore   \n",
       "\n",
       "                               sequencer_run flowcell_id lane barcode  \n",
       "15                                         1   AAATMGFHV    1  CAGATC  \n",
       "16                                         1   AAATMGFHV    1  CAGATC  \n",
       "0   ffc0d811f8ac1aafcd29643e65cf87acf848891f    FAU11962  NaN     NaN  \n",
       "1   7d6a038266e5875db51e462422179436e6c44b3a    FAV21462  NaN     NaN  \n",
       "2   e0fe4672609d60d67277a22973f5c8534df5fc4c    FAU06484  NaN     NaN  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fastqs = pandas.read_csv(\n",
    "    \"fastq_metadata.tsv\", sep=\"\\t\",\n",
    "    dtype={\n",
    "        \"sample_id\": str,\n",
    "        \"read_id\": str,\n",
    "        \"fragment\": str,\n",
    "        \"sequencer_run\": str,\n",
    "        \"lane\": str,\n",
    "    }).sort_values([\"experiment\", \"library_id\", \"lane_id\", \"read_id\"])\n",
    "fastqs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "parental-violin",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<SplitSeqPlate: IGVF_003>"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "models.SplitSeqPlate.objects.get(name=\"IGVF_003\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "motivated-sister",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CCGTCC'"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "models.Subpool.objects.get(plate__name=\"IGVF_003\", index=10).barcode.get().sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "national-princeton",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<QuerySet [<Subpool: B01_13A>, <Subpool: B01_13B>, <Subpool: B01_13C>, <Subpool: B01_13D>, <Subpool: B01_13E>, <Subpool: B01_13F>, <Subpool: B01_13G>, <Subpool: B01_13H>, <Subpool: B01_13I>, <Subpool: B01_13J>, <Subpool: B01_13K>, <Subpool: B01_13L>, <Subpool: B01_13M>, <Subpool: B01_13N>, <Subpool: B01_13O>, <Subpool: B01_13P>, <Subpool: B01_13Q>, <Subpool: B01_13R>]>"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "models.Subpool.objects.filter(plate__name=\"IGVF_B01\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "republican-strategy",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Barcode mismatch IGVF_003 1 CAGATC nan igvf_003/nanopore/igvf003_8A_lig-ss_1.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC nan igvf_003/nanopore/igvf003_8A_lig-ss_10.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC nan igvf_003/nanopore/igvf003_8A_lig-ss_11.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC nan igvf_003/nanopore/igvf003_8A_lig-ss_2.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC nan igvf_003/nanopore/igvf003_8A_lig-ss_3.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC nan igvf_003/nanopore/igvf003_8A_lig-ss_4.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC nan igvf_003/nanopore/igvf003_8A_lig-ss_5.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC nan igvf_003/nanopore/igvf003_8A_lig-ss_6.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC nan igvf_003/nanopore/igvf003_8A_lig-ss_7.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC nan igvf_003/nanopore/igvf003_8A_lig-ss_8.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC nan igvf_003/nanopore/igvf003_8A_lig-ss_9.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "for i, row in fastqs.iterrows():\n",
    "    plate_name = row.experiment.upper()\n",
    "    library_id = int(row.library_id)\n",
    "\n",
    "    subpool = models.Subpool.objects.get(plate__name=plate_name, index=library_id)\n",
    "    assert subpool.barcode.count() == 1, \"{} {}\".format(subpool, library_id)\n",
    "\n",
    "    barcode = subpool.barcode.get()\n",
    "\n",
    "    if barcode.sequence != row.barcode:\n",
    "        print(f\"Barcode mismatch {plate_name} {library_id} {barcode.sequence} {row.barcode} {row.filename}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "gross-cooler",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "adjustable-fifty",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('IGVF_B01', 8)"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plate_name, library_id"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "distant-atmosphere",
   "metadata": {},
   "source": [
    "Load information about previously submitted records"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "imported-placement",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loaded 278 submitted sequence_file records\n"
     ]
    }
   ],
   "source": [
    "submitted_sequence_file = {}\n",
    "\n",
    "for i, row in submitted_sheets[\"sequence_file\"].iterrows():\n",
    "    if not pandas.isnull(row[\"accession\"]):\n",
    "        submitted_sequence_file[row[\"submitted_file_name\"]] = {\n",
    "            \"namespace\": igvf_test_namespace,\n",
    "            \"name\": row[\"accession\"],\n",
    "            \"uuid\": row[\"uuid\"],\n",
    "            \"url\": \"https://api.sandbox.igvf.org/rodent-donors/{}/\".format(row[\"accession\"]),\n",
    "        }\n",
    "        \n",
    "print(\"loaded {} submitted sequence_file records\".format(len(submitted_sequence_file)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "numerous-marketplace",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['rodent_donor', 'tissue', 'measurement_set', 'sequence_file', 'multiplexed_sample'])"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "submitted_sheets.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "harmful-builder",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'a': 1}\n",
      "{'a': 1, 'b': 3}\n",
      "{'a': 1, 'b': 3, 'c': 4}\n",
      "{'a': 1, 'b': 3, 'c': 4}\n"
     ]
    }
   ],
   "source": [
    "def built_in_cache(a, b, cache={}):\n",
    "    cache.setdefault(a, b)\n",
    "    print(cache)\n",
    "    \n",
    "built_in_cache('a', 1)\n",
    "built_in_cache('b', 3)\n",
    "built_in_cache('c', 4)\n",
    "built_in_cache('a', 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "second-hopkins",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "supreme-projector",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequencing_experiments = {}\n",
    "sequencing_subpools = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "armed-education",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "updating 003_8A igvf_003/nextseq measurement set to 003_8A_illumina\n",
      "Unable to find 3cc7eaba69775ece39914a7b4e58de8d\n",
      "Unable to find f637c61c5c3df12dc946d2d32c0bdace\n",
      "Unable to find 494bd18e8eb9bcef5b79f18f8f21d569\n",
      "Unable to find 52ddf98b481bb1f5fa7e0f705a09625a\n",
      "Unable to find 83fe8790d4a03f082b2b042719302fcb\n",
      "Unable to find af524063452a305fa81585f760ad0cd4\n",
      "Unable to find 82e30f39ad4bf46b3dc4e341c891e08f\n",
      "Unable to find 747ed12f09ae3b9cf85c4ea32807c972\n",
      "Unable to find 0dbdc8a7fe445f8a9e6bbbd7a7d42cfd\n",
      "Unable to find a263e940bd4e732193b7615fc142b054\n",
      "Unable to find 968a3e1b887b164eef30f7a443d984be\n",
      "updating 003_67A igvf_003/nova1 measurement set to 003_67A_illumina\n",
      "updating 003_67A igvf_003/nova2 measurement set to 003_67A_illumina\n",
      "updating 003_67A igvf_003/nextseq measurement set to 003_67A_illumina\n",
      "updating 003_67B igvf_003/nova1 measurement set to 003_67B_illumina\n",
      "updating 003_67B igvf_003/nova2 measurement set to 003_67B_illumina\n",
      "updating 003_67B igvf_003/nextseq measurement set to 003_67B_illumina\n",
      "updating 003_67C igvf_003/nova1 measurement set to 003_67C_illumina\n",
      "updating 003_67C igvf_003/nova2 measurement set to 003_67C_illumina\n",
      "updating 003_67D igvf_003/nova1 measurement set to 003_67D_illumina\n",
      "updating 003_67D igvf_003/nova2 measurement set to 003_67D_illumina\n",
      "updating 003_67E igvf_003/nova1 measurement set to 003_67E_illumina\n",
      "updating 003_67E igvf_003/nova2 measurement set to 003_67E_illumina\n",
      "updating 003_67F igvf_003/nova1 measurement set to 003_67F_illumina\n",
      "updating 003_67F igvf_003/nova2 measurement set to 003_67F_illumina\n",
      "updating 003_67G igvf_003/nova1 measurement set to 003_67G_illumina\n",
      "updating 003_67G igvf_003/nova2 measurement set to 003_67G_illumina\n",
      "updating 003_67H igvf_003/nova1 measurement set to 003_67H_illumina\n",
      "updating 003_67H igvf_003/nova2 measurement set to 003_67H_illumina\n",
      "updating 003_67I igvf_003/nova1 measurement set to 003_67I_illumina\n",
      "updating 003_67I igvf_003/nova2 measurement set to 003_67I_illumina\n",
      "updating 003_67J igvf_003/nova1 measurement set to 003_67J_illumina\n",
      "updating 003_67J igvf_003/nova2 measurement set to 003_67J_illumina\n",
      "updating 003_67K igvf_003/nova1 measurement set to 003_67K_illumina\n",
      "updating 003_67K igvf_003/nova2 measurement set to 003_67K_illumina\n",
      "updating 003_67L igvf_003/nova1 measurement set to 003_67L_illumina\n",
      "updating 003_67L igvf_003/nova2 measurement set to 003_67L_illumina\n",
      "updating 003_67M igvf_003/nova1 measurement set to 003_67M_illumina\n",
      "updating 003_67M igvf_003/nova2 measurement set to 003_67M_illumina\n",
      "updating 003_67N igvf_003/nova1 measurement set to 003_67N_illumina\n",
      "updating 003_67N igvf_003/nova2 measurement set to 003_67N_illumina\n",
      "updating 003_67O igvf_003/nova1 measurement set to 003_67O_illumina\n",
      "updating 003_67O igvf_003/nova2 measurement set to 003_67O_illumina\n",
      "Unable to find 7c2e696b3ef32c93df6b60904e944206\n",
      "Unable to find 35a5f174f5092c77a9b49abb6e760162\n",
      "Unable to find 091ce70e2a81a67cab829a8ad950a05c\n",
      "Unable to find 8bd57a0ee351c5b6c4b75241e27841fb\n",
      "Unable to find 9fc966f9a65017ed8d08253358f9838a\n",
      "Unable to find 9f619b9573bceee16a101e896603a04b\n",
      "Unable to find a71346fbebadc9cd919371c635fbc4a7\n",
      "Unable to find be7f202f30967334d5a0c9efb1beb586\n",
      "Unable to find b7d9ee19e2e3d326d806185d309bd751\n",
      "Unable to find 70086765aaf2678119bb04b488aa9416\n",
      "Unable to find 42fc16a794c9f570eca9eca1393672a2\n",
      "Unable to find 429e388d61addcd0b0c8f97458b9cdfc\n",
      "Unable to find a898c9dbb5173d72304dcdda3e693d71\n",
      "Unable to find 9fef40606119b48ae90ecad0976c7363\n",
      "Unable to find 32113bed6290500ee58281936ebadc0a\n",
      "Unable to find 4b64f562917fd3ee2dcf40a18f525dda\n",
      "Unable to find de9bc2b72383f85a2779c6f90fc1ff96\n",
      "Unable to find 310343ca4eeaee652e4b23b1a4d117cf\n",
      "Unable to find b37eb5352cdc612eaa64099ff61fb255\n",
      "Unable to find 06eb1b5adbec7243160c976f0c27f06c\n",
      "Unable to find be0419eb87c252b1b35f7f62a23d0476\n",
      "Unable to find 595fa38ab2b447550c6e526e042b41af\n",
      "Unable to find 2219b962c5c81feddc0b63bb38605728\n",
      "Unable to find 56b640e9df1c43334458708fb9fea31d\n",
      "Unable to find 096237a6e997f226244978eb1d908a93\n",
      "Unable to find 981506bcdeb1d8821042ea38a69e6d1a\n",
      "Unable to find 9d0708779041d8315ac46b8534f2b15c\n",
      "Unable to find 8d669de24ff2ca0d34544f4153db0c4a\n",
      "Unable to find 63d8f3ec621aecf7569cf8eeb2b94ab9\n",
      "Unable to find e987b7a5dfe3cfc1c0ff7ca74ba41a44\n",
      "Unable to find 399c2a49c7ff22d3e494924371d6b4a7\n",
      "Unable to find 369f4852a1465b4f4462d118e15ee417\n",
      "Unable to find b159a646077c4ebe60baba76e03e9a30\n",
      "Unable to find efd713aeee526624b5b728e0be860729\n",
      "Unable to find 831b4c80b3855f671072e2fa9a4aa401\n",
      "Unable to find 011f1d710e047550a85b8318e9bf8c80\n",
      "Unable to find 43ca31d3604111e80572ebce19bf89e7\n",
      "Unable to find b35de1a33e6653fb87227fb3ac1af723\n",
      "Unable to find 1ae43c9897223dbb15ff3e2b0fa53008\n",
      "Unable to find 82d02585700f5ed73a176e5191148cd0\n",
      "Unable to find f01181fa31093a024ad42e6237babe8d\n",
      "Unable to find 8dc6232b46d8fbe744956bd83c1c631c\n",
      "Unable to find e4ba135dba9d7e7cd5f8f73bd0bbfd25\n",
      "Unable to find 99a51eabb1235e0f9aee6c600e71f56a\n",
      "Unable to find 67a28c6b7eee5dd5355317995eff23ca\n",
      "Unable to find 0a2b70b7943256441a6a5be216a601e3\n",
      "Unable to find ba8f751700841d0e5e35e7192aeef8b4\n",
      "Unable to find c3b2ea8f0975233864a5a88932c767b2\n",
      "Unable to find fae45bef661ebdaffc8bd988ad973bb7\n",
      "Unable to find 1379b8e9ff539751c253a374debcc7a9\n",
      "Unable to find 58b5ad20d04cca18a4f9b90e1445c1b4\n",
      "Unable to find 4cb815d97063042b17db20485580409c\n",
      "Unable to find 793ace3aab7fde1ead9d2cd803d960ce\n",
      "Unable to find 04d1c0ae847002c8242887a41549a435\n",
      "Unable to find dbde1caebe27de7d9fc029f787078559\n",
      "Unable to find 889da42024507f9928225f7251208f89\n",
      "Unable to find 11665291c0200154ce4970a564b0eebd\n",
      "Unable to find 032cf45ef0c66d69af3ec76bcb85d14a\n",
      "Unable to find c3ead83fa41c61c7543f0edad22e1795\n",
      "Unable to find 96cb83b941da4100a97684f596d743fa\n",
      "Unable to find f996476af14ef27280793059f19ce129\n",
      "Unable to find e71d9b35a78605cf9db9e2e3ee29dda8\n",
      "Unable to find a52ffd9922caa7a148f56e0ae66039da\n",
      "Unable to find 697ffaa53bb44eaec0aea4e80b6a0d90\n",
      "Unable to find 94f59a944c03d87e8d58f49032dc2ea2\n",
      "Unable to find ec86deab0b5d2a08f112a27a83eae6fe\n",
      "Unable to find 48879d0646b604ce71263578dbe97a13\n",
      "Unable to find 375d363560ff354bfdd526c223375fd9\n",
      "Unable to find 44f25854a197af581e1cb292e3480ebe\n",
      "Unable to find 07ade726d3652dfa30a9c5f4906ce76a\n",
      "Unable to find 35804dc97353ae3809b40c8ed5e3bebf\n",
      "Unable to find c1908ed5443510b0a4338d4acf8e6624\n",
      "Unable to find b49aa80b860b63c28641cd50565d131d\n",
      "Unable to find b43aaede8a66467fe1f5b521b9c131e3\n",
      "Unable to find a4d7df1b3ead6a237bbbcfc759fb9823\n",
      "Unable to find a5946ccd3e3cf64ec8b775bb80de5e90\n",
      "Unable to find bee8dbb991a30c2c578bd9c4f4f7d6bc\n",
      "Unable to find 1fb7aa2e302287a27b9f89d91818a050\n",
      "Unable to find 0dc9e3530c41962f2a8cf1ff0a2427d4\n",
      "Unable to find 56f6fb0f91ab0d3a0a3724fba9764758\n",
      "Unable to find 12f48ccc04ba566fe38d55a4721e8ad1\n",
      "Unable to find 747cff8bab00af77df2f66c488d3f30c\n",
      "Unable to find 2fe55b07c2adc4a95e34ea816d01ab46\n",
      "Unable to find f379f9a4cf81e7f545ac7786ab21fac6\n",
      "Unable to find 3c5ec1d2ff110fea05908f32c2b8ba1e\n",
      "Unable to find 608113c0adc39d432fbf35df11b0ddf2\n",
      "Unable to find c065a72bafd89e2845f8fe9a6a67cdc4\n",
      "Unable to find 1064be35e0fb959b2f2d36fe1132fcb8\n",
      "Unable to find 011eabe8b6f6643aec26ac85b8daece3\n",
      "Unable to find 6b6301c1553156da8fafaffb4ab4c48e\n",
      "Unable to find 6c830cd71124a46945223fb20546d3be\n",
      "Unable to find 8fa2c945d92794b5f9a5530039b734d1\n",
      "Unable to find 9e51fabea038a41d3515129c1f82988e\n",
      "Unable to find faa6da753981777453cf49bcf300f4af\n",
      "Unable to find 20e14ba5f06f8ad2598017c5845c9ef6\n",
      "Unable to find f287f5ef2229d0bedf071d507cf876b2\n",
      "Unable to find 8f0c9a83ada11bc987c24fbb2816b0bf\n",
      "Unable to find bea731854b76d462c18e1724c9e6832a\n",
      "Unable to find 88073e9c877089b1023835f0fb026859\n",
      "Unable to find 454ba90d4def82c9244c46b9c2243778\n",
      "Unable to find 438a06e46611f08e587cc426765528eb\n",
      "Unable to find 4a1071c08f9da3d30cea4e480ba6569e\n",
      "Unable to find b2396764f4b88372ab8ad6e05f0c534e\n",
      "Unable to find 96564eba8baf86caeaac67f20f4ee7b2\n",
      "Unable to find b3373ad49dd8ebf5819eb313a1dee17f\n",
      "Unable to find 1c5e4b02cf682ae861b7de37bcb2cbc6\n",
      "Unable to find 58305d984ccc92badc0442cfbabe085e\n",
      "Unable to find cde566cb42ac11ec97df64fa3d7c9783\n",
      "Unable to find a7a57cbc1e68ad2433f262708aae9b75\n",
      "Unable to find 0a906b1cf529ffefd696956c8dd58ea8\n",
      "Unable to find 41f73c808ff036aa602aa5074581778f\n",
      "Unable to find 533d7b784e0099e2b69f88ed0f637676\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unable to find 90073f04a446e525832b34ce26a53469\n",
      "Unable to find 0f531209b365677881ddebb0e9ba9908\n",
      "Unable to find 1d24b88b0894bb7c05caa0a330402953\n",
      "Unable to find 21e061d7d62b4a8e8257f552ab84d6f5\n",
      "Unable to find 4b2312b7762997fe621b0b708c791759\n",
      "Unable to find 2b5557cd84259124cadadd5676940e28\n",
      "Unable to find 945986935d0c8fd5886fe08086f66fd9\n",
      "Unable to find 706031683f8dc23754437d07cc454367\n",
      "Unable to find 0cf96234ee7e754e23b6cbb06db7aaf3\n",
      "Unable to find f841723261808e4d27f9325bf8593d4a\n",
      "Unable to find 0ae7e5cca34c0c15298693680a7a407a\n",
      "Unable to find 5c368497efadcfa4d9f3cf628121d199\n",
      "Unable to find f7029d4e44bda89abb221985448a2a32\n",
      "Unable to find 6db8062c9b25a7a3fd8d352347c18eed\n",
      "Unable to find ae9fd97e7fec715d455c3e190921ddaf\n",
      "Unable to find 42a13a6b9a1931d50c7507e482183957\n",
      "Unable to find 2a4d540927cf8742694b19d551907ff6\n",
      "Unable to find 2b184742a7396f596f2509daa3bcdd9c\n",
      "Unable to find 4f6afc7542acb3c613ecaf718f8b11d9\n",
      "Unable to find 20934a1f874f029d9caa3ab629fec9f3\n",
      "Unable to find 7a422ab13a08e253928acf01e119109a\n",
      "Unable to find b6abb56b5961f16d84080ce1cf974f44\n",
      "Unable to find 0833907c6c7e9fa67f31e923f8cb4dae\n",
      "Unable to find 575e0e545e413362c272546a70238bb0\n",
      "Unable to find 3301aa432c98b3a648f17ff607eca27c\n",
      "Unable to find 30f978330bcc5f8b2621ab2c7076fcbf\n",
      "Unable to find b96504a45d92491129b45b9f4e1a4ffd\n",
      "Unable to find 841f1b74af136d8c082dd9835e695096\n",
      "Unable to find 60357b65c5897d02e2539f6ace783f13\n",
      "Unable to find 6b8105a1aac2754828a77006eb4e7600\n",
      "Unable to find 1acb26ac9f3b4ef1abbcfcb5cda062e3\n",
      "Unable to find 44bd94287405d48620e81bea80077584\n",
      "Unable to find e9b292eb885ca4d43245463588db1f88\n",
      "Unable to find 500d6463a8aba1ea8d26ede99a09f9c1\n",
      "Unable to find 071ef7e8d1cd7dec921f333f72cd7b30\n",
      "Unable to find cbba592ce23b31c19baa203b3fd7246a\n",
      "Unable to find 5d18cc29ea5853f2d2e0c3d6d8756fb3\n",
      "Unable to find a2e9743d2dcd9c5b95958ab380c3b04e\n",
      "Unable to find 82bdc4a9183a88c5eb566d7cccb7e417\n",
      "Unable to find e2ad221c32ecb538062d52a3fb4daac2\n",
      "Unable to find b56f58a3ffc6d5917faacd8058a41253\n",
      "Unable to find fb8c31249fd951f8bf853f77bfd5a4d1\n",
      "Unable to find 58b7bbda0521471889206ca4cb844c41\n",
      "Unable to find 74ac26ad186c0d9ce5c697f4c1a986db\n",
      "Unable to find c8a2bdb7f8f9b225ce8f297991f7a7ce\n",
      "Unable to find a4d70856bf6c82cc2ea1d3edbb709937\n",
      "Unable to find f83fe37e564ccbfb23a24b38f4d3359f\n",
      "Unable to find c1ff99fe573397dfc8e30120b45cd6d8\n",
      "Unable to find 25941a5bb973f9648c6603941296d6d5\n",
      "Unable to find dae43810f5f12b8ea2f2ad4be28eaa5c\n",
      "Unable to find 5bd4c964df0b1edb59a02592e80c2298\n",
      "Unable to find 3199abb6ab7a74021e68ff8d82c55339\n",
      "Unable to find 3ee71f65efd4ca7a55d45608ca37d9cc\n",
      "Unable to find ca95ea918554339ee2639933db68489a\n",
      "Unable to find a355e2eda77543b1446c55569e84f082\n",
      "Unable to find b0ca0f5f3d43fd21fba8ff01405b1174\n",
      "Unable to find dc80f2b4b2a4eaaeec93528edd97e902\n",
      "Unable to find 3463674f95d7ff25cb0b90d0cb2d0ebb\n",
      "Unable to find 1bdd813ee656c008a1e67e460ca3bf48\n",
      "Unable to find 5af31b96f642918c20ae7bae95141868\n",
      "Unable to find 833f7751cccce848423a966040c962c7\n",
      "Unable to find 3c0c50d0112dfd67cc5516baf0d01e6b\n",
      "Unable to find 71f0b220f5923f3e22a3fd21e2e2be70\n",
      "Unable to find b950cc47e7b2d4767efb6dbb09f1c0a4\n",
      "Unable to find d4fa690460de2d6da24d4049dddecb7f\n",
      "Unable to find f2c6d5f0bc43a9b3c08b8dd4c39fb9b2\n",
      "Unable to find 7c4eb82b8520fdb1ccc87408d4f819f6\n",
      "Unable to find d337de8cf9dc9191277270e1dd0f63b1\n",
      "Unable to find d61f88556bce54e4eef42864f745b3fa\n",
      "Unable to find 639215b918d6dbfd8d04f0659b45832a\n",
      "Unable to find c458912637161c38ded725eb6a55ffc1\n",
      "Unable to find 98b0e1297ea681dc213b383444429545\n",
      "Unable to find 95f6616d7e397af8bdfd5705756f3a52\n",
      "Unable to find de45fa735389a21539e77bc661dfb560\n",
      "Unable to find f37a20e2ac3e8cfae798ae096c0f32f9\n",
      "Unable to find a60bb302f2ba56592f6126d4fcfdc610\n",
      "Unable to find 097e0d31ca2b61a1ccacbb0791496fb4\n",
      "Unable to find bed02a9d001781013b380bab3ff49e37\n",
      "Unable to find fa9d19728c2ffbd0bf2469147dc37923\n",
      "Unable to find 6e5db08e7c3dc2c055d3dcc8ee462ab3\n",
      "Unable to find c7684d783a2e6961f1571dea736b05c9\n",
      "Unable to find 975b136e76ff587cce6cf4d90ac42575\n",
      "Unable to find 13dbd506207489e1bfb7fb90e649a584\n",
      "Unable to find b0e62637889d5ffbc69c4c9af31ad0c3\n",
      "Unable to find 26bf10852d89435075c836c5e023deb0\n",
      "Unable to find b086714d3949ac0c869204b53cb59c7c\n",
      "Unable to find f20d4009fcad973ce697a0442eac9bab\n",
      "Unable to find 693c3223fefb1e9357a7904e9013c9e6\n",
      "Unable to find 1ab10a8ada340e5fc6fa3ab0f1872fd7\n",
      "Unable to find e95d8664a056058e24691be75cec698c\n",
      "Unable to find 94ffc3f5237a5f1a6b7def58ef902ed3\n",
      "Unable to find 8e174005925ea8007d9749b4f34daaea\n",
      "Unable to find 7ec50b47ae513298a78d69d9eb1d5327\n",
      "Unable to find abb0052faf9233ee809f42fed0227092\n",
      "Unable to find 35e2b5686a2949193a0ed1cdb7c04a76\n",
      "Unable to find d978bfe7b2bcd912b19db5a8b90ed94a\n",
      "Unable to find a6e803195c78882074158d68ae5576f3\n",
      "Unable to find cfa626ff7870403a35669d68db02b540\n",
      "Unable to find 90340f0f7f754d7340b47c61580c6bc7\n",
      "Unable to find ffbcfd49209211569a1b6754fa4ebde3\n",
      "Unable to find e7fd2d4ea79dc75023e98c870cc43f61\n",
      "Unable to find bbbfdd2696b1527c6fdc0ec1e35f8933\n",
      "Unable to find 6c7ac6397613436ba4195472273c49d2\n",
      "Unable to find 40c28dcc856213ba1144603e7f243caf\n",
      "Unable to find 72478d810741c82e8b00685bc0a6f770\n",
      "Unable to find 19d94cf2040d89ac14562e373ebaa226\n",
      "Unable to find 18d5c41cf55d35a46538c1db30fbff6d\n",
      "Unable to find 42de699134e0954ee2aef39f81db3a70\n",
      "Unable to find 346f1a54be26735545ff691037fc385d\n",
      "Unable to find 4b71d2439ac24617d9ef0d47bcc30ed0\n",
      "Unable to find a25eb265531fac082de597c20106b568\n",
      "Unable to find bc1726e8b61d5ab1b6ed56c99d21d87c\n",
      "Unable to find b2e9d73cb56073ef8a24b157f8aa5b80\n",
      "Unable to find c22a34eae50cd28e319d00a70435dc97\n",
      "Unable to find 7583b04d4ec7cdeec06e270b61219a8a\n",
      "Unable to find 955815f212e9389c30883d501e60d345\n",
      "Unable to find f4fd6f1dc375d214bde6361ef2d4eed4\n",
      "Unable to find 1cc03429039a277e5e92bd8ea44e9809\n",
      "Unable to find 0a47fd72289f4c7bdf55bdeac44d3c09\n",
      "Unable to find 1a127b65a4e6ca6ad7cdd2de9a344e32\n",
      "Unable to find 028c6220881164fc5e2a8a6ad8e974ed\n",
      "Unable to find 4a7363079ec871e27c9fd9d298016480\n",
      "Unable to find 093e74e1d2d09818b8b45041ac3122ff\n",
      "Unable to find 7ef4c6109f23f01ae2b6629bbfffc228\n",
      "Unable to find e87b44e2dbbf5e589afd971ba7469173\n",
      "Unable to find 652e77e5267544e129bc6f240aa00e64\n",
      "Unable to find 3936dee09b0f4809821f639f80e89b19\n",
      "Unable to find f13efbea063fd2809c3fbafee8ab21ce\n",
      "Unable to find 23a6ed489fe53accba6db235f307d79f\n",
      "Unable to find df9647df8d73cd2f0457f977ecd2959f\n",
      "Unable to find f851ea47ed44bfefec97ff9822c363fe\n",
      "Unable to find 2a90de5626a71e0b8f451eb13cc4cb99\n",
      "Unable to find 5d26d55eff99cda323f784b782481f6b\n",
      "Unable to find 40fc5d5db869c3b0daf2a462d1b195b5\n",
      "Unable to find 0c48d3c086d609816eaad49c6902ef60\n",
      "Unable to find 9c4888ea4dc074491340a569d99937df\n",
      "Unable to find 705f90c13dea6fead5fb0526d3f7aa9a\n",
      "Unable to find 61088283336d77046ac49db8599d0026\n",
      "Unable to find 286c2fe70cda13ec3138b7a6fd0da36b\n",
      "Unable to find 76831c060e1886edf90e2ee158ad79a4\n",
      "Unable to find b02ed3a97d1ef9c51194c2b7987de1d5\n",
      "Unable to find 00d05a91aae90cbcfcd27e7c5339ba61\n",
      "Unable to find e6fb6bf1823711d23be03e1449e6747d\n",
      "Unable to find 79a9ff5ada609cfdee87296b2fa8f5cc\n",
      "Unable to find e7525667052d4b8049269f00eb460952\n",
      "Unable to find 0d35e56ceb47313957d3161f115b0fc6\n",
      "Unable to find 702e24ab8f2fed9abf868e94a52f2b08\n",
      "Unable to find 0a19b2067d693d18078907dca5f7b564\n",
      "Unable to find e4df8cc0b1f6462d2ed662b8d98c145c\n",
      "Unable to find 7565549a49e57257ebc0cbcabd82c626\n",
      "Unable to find 7142e878729bdc333fd6fcd455658024\n",
      "Unable to find becaa28c7b9694ff4e0779f7c0c17447\n",
      "Unable to find ca739778141f70555b04a3efa596d7c9\n",
      "Unable to find 6f6afb9e6d4ea5b6ef9120deb14cce47\n",
      "Unable to find b22089272fcaa00468cc88e093f2ca08\n",
      "Unable to find 9c0707c63936b1826eb0caf519efcfb2\n",
      "Unable to find 0b5dea2906ae9ec30faadc759fb47b5c\n",
      "Unable to find d371d9731167550f3de38b800ceb6609\n",
      "Unable to find b0513ef377d413e61f1bcfaec72d7f93\n",
      "Unable to find f900d75b9c4f5352f2379bd023698c49\n",
      "Unable to find c0c00a40287045605243a403c7217593\n",
      "Unable to find 70b3d3233278e3b068b3c59fbec46b4d\n",
      "Unable to find c853bf6eeac42fd5f3bf7f9e996a064a\n",
      "Unable to find 7acc52845d9a27f64727828b31f384fc\n",
      "Unable to find a1c7dcf9302caa2b99eb481f4ae3b6de\n",
      "Unable to find 16cfc025ba266f3212088a7730ea54a6\n",
      "Unable to find 0e04e656db239dd609a1f370c680029b\n",
      "Unable to find fcbae83c93935997d31cee3a8c315418\n",
      "Unable to find 259be10bf9ce46d080525902185c00d9\n",
      "Unable to find 6728c2594b91a43a3ce08a20def15b0e\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unable to find b1288b020c222401c9672d8be4390165\n",
      "Unable to find 5651aef2d8cddb9539e370aba6d0b2d3\n",
      "Unable to find 82a02c994d0dcd4be4598756dd1b9200\n",
      "Unable to find 59040625cd953b0ff722ee21bfeac2ba\n",
      "Unable to find 9be86c7d9a3f51207afc8f6dacf213a7\n",
      "Unable to find 9031ecc1d6d2cdf1d3b90fce7fd99e7c\n",
      "Unable to find 9b80d956aa20a7215b60a235a143af5e\n",
      "Unable to find 3072a2b5c5272724d482d9c27837795b\n",
      "Unable to find 579e1e61623dea1636c2fb6ceb8a3562\n",
      "Unable to find af4a5ec5c75ec9b9a1f5059a9171c85a\n",
      "Unable to find 17cfd4e95d5a8dc1a1162b3f3e2b7c87\n",
      "Unable to find c2dc3603fe1696c2d50be06f349e4591\n",
      "Unable to find 48293a211a24dee992d8a9d6e5438dc4\n",
      "Unable to find 61ef5ea008e3b87fb68f7238baba050b\n",
      "Unable to find 492a2f6ab9f6c72d8ea1586f2cb8aeb3\n",
      "Unable to find 3aaa276c0aa30a7220c1fb53c7eaefa0\n",
      "Unable to find aefee63a46a79d03928fa05061b037dd\n",
      "Unable to find 21b38ef192a7131f3a38364700867dd9\n",
      "Unable to find e2470bd494bfdd3639a571aa836b99bc\n",
      "Unable to find 918cb4e67b394af3c356f16d1ed1296f\n",
      "Unable to find 87a8aad0f3504e8c2092d7cf3d5c0393\n",
      "Unable to find 42f0d54a72dc1754d7859e4e335fdda4\n",
      "Unable to find 1bd6a287176307c8c63e5c0c46a1abd0\n",
      "Unable to find 5d6683abaddd7d1896686ee7063394f8\n",
      "Unable to find 8196b64bbe621181a99331a040920af6\n",
      "Unable to find 8afb4ec8bfdcb32651dc4d8e62a8b73d\n",
      "Unable to find 99fb0cc82b25527a317b358d76c8bb75\n",
      "Unable to find 8c075873585461d69d9ee89906acf326\n",
      "Unable to find df9ed97ff3aad25f259206e87450678a\n",
      "Unable to find b3324ec1d011b2984e47ce18f42d11b5\n",
      "Unable to find 9d21ee483a0483c5523a545b308c15d3\n",
      "Unable to find 2cf0b4e244e8a4aa06145e0a197ee2a6\n",
      "Unable to find f07a8e61c1b2495e6dcb31bd0e1462eb\n",
      "Unable to find e10a4883d09b86890789bce0d1cb59fc\n",
      "Unable to find 411b7523cfc016d7a4fa44f178037538\n",
      "Unable to find 0eb8304cbf8ea1f51e8612c2dbd6417c\n",
      "Unable to find 01a94637eb0e0cf47e4b0da5ea5244e7\n",
      "Unable to find 190314c4f57b3309d2b3b06f01cd2b22\n",
      "Unable to find 7da447b043f8b5bf09a057a528d61e5a\n",
      "Unable to find c85a037ba327626ed656b33d8d73f39a\n",
      "Unable to find 9634ab8c8d3da3fec5941d513623a938\n",
      "Unable to find 833dfe286ca153e989cbcd547a736868\n",
      "Unable to find 5fb544386a0bab6ae7f630db313043ac\n",
      "Unable to find 3d319340255a02781a026e329ea344e5\n",
      "Unable to find 8f947dcc0ed0bafda54343ba697bae78\n",
      "Unable to find 29581adc5275742fefbaf8b81107a6b0\n",
      "Unable to find fa5cc54d55647c4ecb10526f43339f96\n",
      "Unable to find eff8415c2cd80793d251a184921b7071\n",
      "Unable to find 2141ff31d3124d55f43361e0bf541430\n",
      "Unable to find 216c5549b8e69f17f7378b4d9e3c6965\n",
      "Unable to find ab6f286b88669e2209b38a1655e52d1e\n",
      "Unable to find 720a062413dc8d0e78b5765aac8f8073\n",
      "Unable to find efe3057072170569a853310b20c7ce93\n",
      "Unable to find b6123bc64638db028684f8b8d38c6b58\n",
      "Unable to find f2f56485c8867565a5cf6c4d623ede69\n",
      "Unable to find 7427059121ebb51ec7ba85a206778147\n",
      "Unable to find ac1d30fb5d3a10343859de254463373a\n",
      "Unable to find 2b6a73404c5fe64a1026b863412825da\n",
      "Unable to find 0c0e9808989b0d894597e0980b5b1f07\n",
      "Unable to find 7223b227f25d343b30f18eb223718732\n",
      "Unable to find 561346e58b8257b3e588c159909755af\n",
      "Unable to find 9e923368e0eabdfee44cf28f862c7b2a\n",
      "Unable to find 25fb1af8540109d997dda7b6bdbaca83\n",
      "Unable to find e43f978317365eea653e92ae66cf5572\n",
      "Unable to find 88e2f7b21437945a59d902b5989d520c\n",
      "Unable to find 9b4904065c8457534b3c0020cf7d5e83\n",
      "Unable to find e099451db469777cd12a9c5eaa2d393e\n",
      "Unable to find 80d0d25eb8072c59e8e74be511a8dab2\n",
      "Unable to find 174fa4104c605eca8a9205ee9619f078\n",
      "Unable to find 25cb767ddb1c956b90c0dc4e63f50a5d\n",
      "Unable to find fb97266acd289e295a7838d7e0e60693\n",
      "Unable to find 954c6e936695636161d1d9d657ae1e2e\n",
      "Unable to find 7ca6c548e5f34ca8a49a5fd309445de0\n",
      "Unable to find 0f39ed01b8ad3750c512354d2f961c60\n",
      "Unable to find 7460852cb11f1e2f890efc06a4a1d972\n",
      "Unable to find 4c1c287aea22c5413e8bc005d5fc3f28\n",
      "Unable to find ded679536c9d783080cf9ed4ca1a02c2\n",
      "Unable to find fc6fbbdcc2efef5cef6b8a679f0a3a3d\n",
      "Unable to find 78fa59a04d383414f59e9cfda98d8871\n",
      "Unable to find 461be29d93743c729ee27b50d9ff0147\n",
      "Unable to find f9e2517986a13791880e33e02e779e42\n",
      "Unable to find f5c449c0ade8ae417875d9249e298d73\n",
      "Unable to find 9de765b8ac18784b6c872bf8ec67a907\n",
      "Unable to find 51042adf5fc8930b0043e399c0c580bf\n",
      "Unable to find 1fddd12449e1b417535d785f9da605f0\n",
      "Unable to find 340a4d74b2c78593282ba340a3bc6294\n",
      "Unable to find a99fe6bf431dbd0503d1c03fb7213ac8\n",
      "Unable to find 00733641a74a1d17de4a8092ebd977a1\n",
      "Unable to find b1f955838354e26dea5215ec589d9cd4\n",
      "Unable to find 504d6ed91812b5b931c89cd0fbbd8077\n",
      "Unable to find 3274cc512b7d083423eca10c93f5a071\n",
      "Unable to find 6dc4d90b73c2cfeeb53888fc5442d211\n",
      "Unable to find 2f2be48603af4a18389d0655a6a30d96\n",
      "Unable to find 1b9332a5df8d33784786251ccbba2a92\n",
      "Unable to find 154e0ce9d800cc83816bc494c4a35cea\n",
      "Unable to find 6a6e7a0d5d6f60c8ff6582869b936c3d\n",
      "Unable to find e84bd1ed5b025ace3d8b0d25f7f0ca0d\n",
      "Unable to find 34c351656326d57970725d5eb6276819\n",
      "Unable to find f43ca2540cfa36fe6799660fd619e4ba\n",
      "Unable to find 9031130d19aef77ac51e7967d1fd226e\n",
      "Unable to find bfb3209d4dcd131c2f64408fbf6335cb\n",
      "Unable to find 7026b3367658f48469c03ab782c16305\n",
      "Unable to find d4236a9133a6efbf939aaa75c5c03fec\n",
      "Unable to find 556bb27944c65ae940fc41971e1d8c31\n",
      "Unable to find a2e77dc109231f8a6c20a29cf84c75e6\n",
      "Unable to find 01d72d748a3ebece737d7d9036688945\n",
      "Unable to find e02991f9ccdbe1d1111b093cd4da6579\n",
      "Unable to find 7b696f1b2c55ba5b0605047ca6d212df\n",
      "Unable to find 271b2b099a3e42687a92a2e6fdc973c8\n",
      "Unable to find 1a4cc8ec4147d41e5b11396a9cf44098\n",
      "Unable to find ebbac613a886bb0963e72fde6288594c\n",
      "Unable to find 9b692d76442e0935f95f2255630ca7ca\n",
      "Unable to find cf7fb39cd888617e2571003d99101e10\n",
      "Unable to find 955143a8c2327ac492c366c5eecff1ab\n",
      "Unable to find 4dde592fbe08e7243a020df359f324c1\n",
      "Unable to find 975c34734f74b8774f08564642c0b52f\n",
      "Unable to find 8034d45902ca502b587b32dc7fcc8ddf\n",
      "Unable to find 8cda10b36f05edd117afcc7a79c423ed\n",
      "Unable to find 14430ae552dc1ddf3d02d1aa82beb4ad\n",
      "Unable to find fc9e44b1e2195416e3bada4662372847\n",
      "Unable to find e0bf0e4635367eda1b8dc679b30ad4c2\n",
      "Unable to find bf57823f6b067b31e46888d18c6d6569\n",
      "Unable to find 6371c31156ebfcdf9f45b8d524b476ab\n",
      "Unable to find f8dfb37904355508398ec1f6cbeaeb73\n",
      "Unable to find c2fd180b43fce4087bdd20f335e668ee\n",
      "Unable to find 9ed7561106f42d0af5efbee69c1da8b0\n",
      "Unable to find acea3c14b50fb3e747d3ab937510222a\n",
      "Unable to find 21e611873c273154e14d4cf8c7143ca3\n",
      "Unable to find 55eef058445869564d23709a8db7ba4d\n",
      "Unable to find 5edcccf9da8978669c049576db7d4016\n",
      "Unable to find 4378dc8c9b2ddab55020af9a4adcc61e\n",
      "Unable to find 1480eb5b91fe36a80043047674a5af03\n",
      "Unable to find 9e38f17d7d198749db579baa767add4e\n",
      "Unable to find 114497c1d87ea020a3292f33e3c861dc\n",
      "Unable to find f40d82a8d49e606e523c607edcf61010\n",
      "Unable to find 7e06b2b7fea52931e21a325205e4916e\n",
      "Unable to find 7aa0af27f8b103c9ca37b78592232afc\n",
      "Unable to find d595c37f2a889bb128279cdfea67635a\n",
      "Unable to find 090f1749037a59443d39dfd514b725be\n",
      "Unable to find 7b44b8a9229d208753c36dd6463a04e5\n",
      "Unable to find dff7b5b4ccdec762f518a0f886b186c9\n",
      "Unable to find 016a94820d0a2f2e6317e6958d5ee9c6\n",
      "Unable to find aebd701aaa4a204443bda7d82d1f0540\n",
      "Unable to find 7b7d7d9c5989b395cb897a47af285546\n",
      "Unable to find 6761810d322e1edf22402b8d2b636cd1\n",
      "Unable to find 144911cb3f9e2c24b8d70538768c2502\n",
      "Unable to find bcfe1a1176952408e6b6531941de40ab\n",
      "Unable to find f7a0562085fbf37e2d3c94106ec84190\n",
      "Unable to find 0c7441bf17e0d6c446058bba5bdb8bed\n",
      "Unable to find 3bedf329ba14a748a0035bab170c7ed0\n",
      "Unable to find 0b868cbf46df17a6ab9efdc2e3783e46\n",
      "Unable to find 23f012599dc4aa33689b400fcc2a845d\n",
      "Unable to find cb94089b8924ed72050f30029c62def3\n",
      "Unable to find 11b29e96b0f228995accd89721f4e0a0\n",
      "Unable to find 6afc8c48fab3226bb018fe05866987b9\n",
      "Unable to find b63c2f0d6f7c27a1765c8c3ed193a681\n",
      "Unable to find dee91189dc4a79ecb33da87ff22f0e64\n",
      "Unable to find d940ce954a0ef71bd8d1a97aa4da8d2f\n",
      "Unable to find 4a3c2ee8227098a755e40adf9e76c6b6\n",
      "Unable to find 3d45e5aad4123bf7ab9d981fd9c65c9c\n",
      "Unable to find 9f95377855e493f3bc078512e603272e\n",
      "Unable to find 81554a419906ef6ebf4b30d60a74f8b0\n",
      "Unable to find 67163ef42eb5e4c2611a439bab97f114\n",
      "Unable to find a017ddb6668d9f36dec1078a57a5fd7a\n",
      "Unable to find e1b7d2c326d3cd08409e74a377902721\n",
      "Unable to find bfe4d1f60936b8bfda2f695d9223cd4d\n",
      "Unable to find 558863fdbefb0abbacec51f91e44ac01\n",
      "Unable to find bff9c3e33c792c5b54791610307cd887\n",
      "Unable to find 938bb4bc738450902a293e6e5e750994\n",
      "Unable to find cad8e744cd257077972fee1942e6cbdc\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unable to find 5147aac6add5cd589d71cd5c47ddc450\n",
      "Unable to find 58190e86412f359e4a19557e1bc99af9\n",
      "Unable to find feee95654d5b45ed0475f915f6d278dc\n",
      "Unable to find 6f16af1530b4d561e876384eb9ea249d\n",
      "Unable to find 49b1497b21ebea35021c16f353aa904f\n",
      "Unable to find d954313605a35e192950bc4bf8cc90f7\n",
      "Unable to find 3a1a0792a3356e45a8bf7af45d364c15\n",
      "Unable to find a581386c2e0db4322348cae3da90b079\n",
      "Unable to find 943d6f74fe60ee4392d9c27edcef5ff8\n",
      "Unable to find b8560133ed3d9aa66662a1703d353a06\n",
      "Unable to find 7c10e7b3d3db717e48b15390f1ca6613\n",
      "Unable to find 079758e2a1b69969bc0b69ef971b22b4\n",
      "Unable to find c298ae91715ad8daa42b43a33e702452\n",
      "Unable to find 788b8f969f1fc9a6262e53341e64c42d\n",
      "Unable to find ccb9e24b351a1083cff1da116adbe767\n",
      "Unable to find 7f0e3f2138dc72f0efcdf315f7b3d6de\n",
      "Unable to find 88a6714afb2b92a587886c18c00fa330\n",
      "Unable to find b88f96ffd653d73ed4fdc0d19eae4c05\n",
      "Unable to find ead5d24cb2c6fbf7f7beaea0fbc8aba5\n",
      "Unable to find 389ecf2fafcdcf8fc1a4052433b14636\n",
      "Unable to find ca8929c9c88c56939f5caa9401c6421a\n",
      "Unable to find 0d8774b2c388fb63a1d4431c51dee59f\n",
      "Unable to find 7c5c46ed2756add5839b1bd9a745b356\n",
      "Unable to find b22c5de90e4db5924bbc360906eb81ee\n",
      "Unable to find 7b0a2df1713211ddbec2a2bd5c125695\n",
      "Unable to find bec5783cda545b96a6387469895d2a5b\n",
      "Unable to find 8008301ce64b72bf60e92a87c8b8319a\n",
      "Unable to find 21e41bb405172d11148dc40079b8a7e1\n",
      "Unable to find 04d42ccfe8d7bed0303b9db293f674b5\n",
      "Unable to find 834b995756fa7132758430ce33b32ad0\n",
      "Unable to find 001b020a22ff492c2a3e072e77ac9efa\n",
      "Unable to find 5d8c319b485b1bea5f80cf5b49726409\n",
      "Unable to find ab31cab21a5a9e117e8e914ff2134bcb\n",
      "Unable to find 682a74a22574ee1b95c407d284af2fa0\n",
      "Unable to find e3c1555f25a1b95cdd6349ffddf84862\n",
      "Unable to find ee28e2c0003466391055e848ca16da54\n",
      "Unable to find 46e6cff130137971da02f3e25334388c\n",
      "Unable to find fa9766938ac11fbffbf20366e6f3e2f4\n",
      "Unable to find e341997a36b73d7f500ef48bcf048570\n",
      "Unable to find 9b31b23a0dea52d9961ba6f8a2c84c82\n",
      "Unable to find 9acc7a9f8d1c62b7b7d504080ff6e4d9\n",
      "Unable to find 0d31c6628e9c0439a4e2301203e156bd\n",
      "Unable to find 460a9e7d05c8718723a7fb16bc1ae8fd\n",
      "Unable to find 150b69d1a185bba90d6972b39e0a1a52\n",
      "Unable to find 96e22aef2bb35de59f0a1d7981112ae6\n",
      "Unable to find b18e09fd9d874563295707ea0d190e78\n",
      "Unable to find 3799795af8c4e30073a277ba16ac704d\n",
      "Unable to find 0d86fc00bda1f7dd140bb7cbc64a3db9\n",
      "Unable to find 9c7ffb753183d74241091921bf833b77\n",
      "Unable to find b28d12a867fa621f6285781a36aa699f\n",
      "Unable to find da1e3960e6595d47558ec744431b4d69\n",
      "Unable to find 0101694eba2057221337028045aa0672\n",
      "Unable to find f02b5c5517243ed981b5e82d795e4121\n",
      "Unable to find 97de238c9c86119ce023c867e5676fe6\n",
      "Unable to find 682cc0f9f49992a85cd311df37531925\n",
      "Unable to find 39a47476cc0eeb0d88584b7889206c88\n",
      "Unable to find 884697756fa470e7110a32ee9d3d92f7\n",
      "Unable to find 5f74a16a683081419a4f29596b314d49\n",
      "Unable to find 20562b1be3d2e1858e52d79d703ae41b\n",
      "Unable to find e55501e9d6d6dfae11c623d3f642a91f\n",
      "Unable to find fa07484bb1d54c840a2ceb9e1f661511\n",
      "Unable to find ad8763ebd425c0ce25fe79bf270490f1\n",
      "Unable to find 7d1fc07641d0cf6eda68437f993a857c\n",
      "Unable to find 53f21d0f1d8d6f73e7ccb7c234e758ab\n",
      "Unable to find 0e258873566d1a84d6dc72ff948bd80d\n",
      "Unable to find 6c43c7d71123b1ec6fb8194f351a8f7d\n",
      "Unable to find e8bcdbb6d6e07725a209330cadd32b9d\n",
      "Unable to find 22ebbac2ec53810fdaf3e33430838011\n",
      "Unable to find dc82343f9c9e3f94d62a1adea6850040\n",
      "Unable to find 00bf76b18f85c7c15a2219293d3a72b2\n",
      "Unable to find af66f7b80740b2d6e624f76deb7f7960\n",
      "Unable to find e8ffa2d7b60559d896066fba93b28b22\n",
      "Unable to find 67c462afc50d9fd3eee7c8cb4121d26b\n",
      "Unable to find 10cf7b00b03677d97b0bcde9aa14ab05\n",
      "Unable to find 87a3836365215932ab60b01a1b45a3d9\n",
      "Unable to find 5fc5b2596bdb92531af0bdc3375680f4\n",
      "Unable to find cea2744fdbe97af693099a9fc844c299\n",
      "Unable to find e1d879ceb8967d0ecb8eb88992935371\n",
      "Unable to find 5a0f65b93cf5302105b29ff17f02f043\n",
      "Unable to find b05f79b00af82dfffa86211985630dbe\n",
      "Unable to find 80af6a4f3e499ff0f3af3d487d7e027b\n",
      "Unable to find 890a65a38ca6cae76a764495f7995a7a\n",
      "Unable to find c098895e19d457697cfea78430983687\n",
      "Unable to find de3ff3881c0e8dd7bd34fa834eac61cc\n",
      "Unable to find 744c6b0f3f9846be495a61e085ab40af\n",
      "Unable to find 31a28b970b606c531691eaf8ae10583c\n",
      "Unable to find 1bce5f3d7624eedcdc0ff4cee2942ba6\n",
      "Unable to find df7578afb85f52ef0d4f7dd2d6772100\n",
      "Unable to find b802cc448fdb3730e1b8f1f1720f3507\n",
      "Unable to find e3df7bafd8c16cb85476e860b9f8832e\n",
      "Unable to find 8c14036e33214a1197244b745307e7d9\n",
      "Unable to find 1fc8ec3dc112e8f5ba6a7c69b0e37e03\n",
      "Unable to find 0195f9d709e07e01523e359baa71a0fd\n",
      "Unable to find 6ea20ae3c85f43596b1dcc01bfd3a8ac\n",
      "Unable to find 51193373e4ddafa399305ba48ce70929\n",
      "Unable to find 4919baeff5b1e9e26a25054d508bebb8\n",
      "Unable to find 707a9eebcd8455036793a9e7dbfc1cc9\n",
      "Unable to find 282dfa6434f87380826e6ee099d6c670\n",
      "Unable to find 32462eca85529e540b1c4471677e7a89\n",
      "Unable to find 52e7806393ef2f52c9c6ee5cecbf0081\n",
      "Unable to find 0de7973d87c40d1a36840bb0c062c930\n",
      "Unable to find fbe19808faea2082bb08bb59b912a83c\n",
      "Unable to find 9e2d1db3970e75683953367254d2d964\n",
      "Unable to find e1211ff13ee8492222ca2093cd6195d4\n",
      "Unable to find 73b9a51b39442520327786e2506bbd0d\n",
      "Unable to find acf61a2a3940c0192526e949c7f7cdc8\n",
      "Unable to find 44fd8ed818020fb60f3800543b276465\n",
      "Unable to find 2b385af7a7ce5a9ddb1a111bd189407c\n",
      "Unable to find cfd0fb7a88f7f428cb992b8204edf5e6\n",
      "Unable to find 20b30c077ed02e043a6e4336eeeb0758\n",
      "Unable to find c29751a4bcbc4f720fd9e8123e2bbef0\n",
      "Unable to find bc3ab6a9b5a2af9b6bd0053a29fcc2af\n",
      "Unable to find a42057f8deebaca7815b42c192988e10\n",
      "Unable to find aa8d8c80aef8b67c3f40ff2344624508\n",
      "Unable to find 3688823f1e3143fdbe162d413a199d6b\n",
      "Unable to find e6f4da2063686034b290c9812d446712\n",
      "Unable to find 9ea0e3cb5bb754039a230878b156d74c\n",
      "Unable to find ae8bc33f23d665e219f96c3446ba46f4\n",
      "Unable to find 0468b4fea19bc90c8689e29f3e3ca593\n",
      "Unable to find ce7b93a768c758236c7aa78255550234\n",
      "Unable to find d5d494a650ae9c4f1f61a0a3742837bd\n",
      "Unable to find 0d893beaf788d3b48b4be413fe503075\n",
      "Unable to find 52f3c97e6ea5d472ee174cd28cc350c9\n",
      "Unable to find 600363446be3a5625bf265f153a5a25b\n",
      "Unable to find f6309ab9cd1fa5c4faa36206e3d84f03\n",
      "Unable to find 31fb566283b7200fb12014cc769d5a53\n",
      "Unable to find f730d80e1c44aa1ce80009c2187bb93f\n",
      "Unable to find abd5b66dd200e4adc31056889ef776da\n",
      "Unable to find 7e0a25fafe04a42099135132c007aed0\n",
      "Unable to find a0af8f24a69add465f439dd8542276fa\n",
      "Unable to find f98564b13df2bd544a54b5995d464e10\n",
      "Unable to find 9ce4632c75629e496cb2765c9a131e9a\n",
      "Unable to find 8bb354d41575daed8fe2186a2c68cffb\n",
      "Unable to find 0ea744abb347704ced80d9eec9540e4d\n",
      "Unable to find 2ee4ecd7434786fd80f1fa3e8f7b34b5\n",
      "Unable to find a58a059ba3ea1d1ec56057a430a5b193\n",
      "Unable to find 38acff41ba6d53aa2656b18014f26e58\n",
      "Unable to find 0bf3859b42eb6a2122dff92adb8308e8\n",
      "Unable to find 1222e3e69d8cf5b5e8c63f0393e98517\n",
      "Unable to find 59b22364e9582881ad9348ef54ec6e6a\n",
      "Unable to find 27b1cf8e0819ce0f9ae1a82c719556ef\n",
      "Unable to find 8987eba3f60c4a28aa8334ce532b3f6c\n",
      "Unable to find 0950db74808d08e98c66104359d356ab\n",
      "Unable to find 6de8b5b96bccb0d67891745bd927c1a5\n",
      "Unable to find 70a2acf5766712986dfd2205f1deec03\n",
      "Unable to find bb46d568c756b482be7edf3afdaad9c6\n",
      "Unable to find a744ec4c6c44df29f20aa3adfd6bda34\n",
      "Unable to find e1ec4e9fd174f17ca09dd970aeb25ff2\n",
      "Unable to find 44b390c51d96bd87be9ce9da991d0ddc\n",
      "Unable to find a64bee8ecb928434eeb2511ee96fad2d\n",
      "Unable to find 37070c25399637d91f9ed9eaab0b59ab\n",
      "Unable to find 1e2384c03017590494af709745f8dfd4\n",
      "Unable to find 623423cee4d6429ccb4f38a1107a5857\n",
      "Unable to find f29107159ee017ee25840cd35e33844e\n",
      "Unable to find 5a69f2682944669eba13eb324cdf490f\n",
      "Unable to find 6bac41f59051d32c6730687d8c47e9d8\n",
      "Unable to find 33665f7c9346b7e884eefa212efb801e\n",
      "Unable to find 43f1cb3996ff6d4622154bd01b26ed6a\n",
      "Unable to find 498493909f62e9af57d62604a009c0b0\n",
      "Unable to find b45926bbeb77de6ddcd8265ff39a887b\n",
      "Unable to find 8da7e1922cda57984cf38dd3c4b4d2bc\n",
      "Unable to find b546f2d1240a8556962c5d2adaa37d6f\n",
      "Unable to find ee3d4293b25a0c6dfed811d772685438\n",
      "Unable to find 1f82bfa25e1f9cccf793e7562011d435\n",
      "Unable to find 0b36ade568a183f5d20d30baf297325f\n",
      "Unable to find 4b90682e8e171cc8623523048874fff5\n",
      "Unable to find 190c87d79a127a491743b8d40a82cc4d\n",
      "Unable to find 9bc50935c1ebce2f825813732a03dab6\n",
      "Unable to find fa1ae7b0b49d85be3c446b39b2a4462f\n",
      "Unable to find fc4cd440da8c12e0ced22832d1d29cc0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unable to find c7d347791e3e9e838286cee9c06b0be9\n",
      "Unable to find 10e6f4ff20856cc58db83a1e5955b18d\n",
      "Unable to find b3c615945cfed4ac23a17d475bef4b88\n",
      "Unable to find c98c6992ba44b353b7e6add49c9b7c91\n",
      "Unable to find e769a4a7d543eaf40f151fe523f2e2bf\n",
      "Unable to find 3b8ddefdeec330acef07deb33c11f16c\n",
      "Unable to find b9cd3541618e307af55eb63eb32b54a0\n",
      "Unable to find 60c07ab18f31f2b47f999414a2efb768\n",
      "Unable to find eebb1672f5fb3a8ebeb8c3efa0e9a8be\n",
      "Unable to find a59d4a101e0cfe7ff08abfd1eb452a2f\n",
      "Unable to find 95d40e46cccfe85f43519ae8e809369d\n",
      "Unable to find df16065efa9e0fb2d36471d42d256f99\n",
      "Unable to find d4556acc09303b2b4da8cc38a7c412e9\n",
      "Unable to find a2dd7d0d880f12d6fa73eb1b5e6d612a\n",
      "Unable to find 618433ae0610a73a7e305a044b41fc87\n",
      "Unable to find c73b65be8f9949c103f89c66b5f78555\n",
      "Unable to find e1dedd342815ac3900ca23e8751f10a4\n",
      "Unable to find 77411e68d437d568a3c88fca2b7afac5\n",
      "Unable to find 2a70d91a76ebd032f126a77bc3c83c6b\n",
      "Unable to find 8a5a31db2494a44d703c4fab6ad09ded\n",
      "Unable to find faae8305d2b9ea77608e6e0a5a93887c\n",
      "Unable to find 08ce43758b54b873d59904bc5f11a93e\n",
      "Unable to find 59603efe38ce533bcee9a2f6858dc76f\n",
      "Unable to find ee412dc5c05a547684f460d03eba698f\n",
      "Unable to find 996b0bdbbe8c269d44675dbafc4ea763\n",
      "Unable to find 4eadc5cd108c6690a6c4b064bf85197d\n",
      "Unable to find c2a5bb7f64343f3bc920fd3a3847fe05\n",
      "Unable to find 3221440bf0b6398c5514a2ba28650597\n",
      "Unable to find 04528b4bacb04d3a1fb5aff8d3baa32e\n",
      "Unable to find 087ac1a049a9050525e4c7179700b02c\n",
      "Unable to find 7b2dea4aab87cf0509cb5a244415c182\n",
      "Unable to find 0d56580f132cc6510f6cf4dd89ec246e\n",
      "Unable to find 9b06f1ac27b2cb7fc8a3ae423d707bf6\n",
      "Unable to find 7ce555945181e088179898334c54f175\n",
      "Unable to find c7fb68242eebb0c7414863cd9f631f08\n",
      "Unable to find bb8c670ae2a6c23d8fdcce8cf0cdc9f2\n",
      "Unable to find 9b956be86b0528252f334249ce02cd6f\n",
      "Unable to find e35c07bb6be131fd736d56a946ae96e5\n",
      "Unable to find f735902688eeb51d7776a0356b521e82\n",
      "Unable to find 15d305a9359dc6b237962e432bf8d8a5\n",
      "Unable to find 90fcbddb0b5f1e5b19b70d4f08a6de89\n",
      "Unable to find bc98d77c98007a2a0a6f1a87c6c26ec9\n",
      "Unable to find ab521a7b7284eb22a4cc7171d281ceaa\n",
      "Unable to find 0c48928f7c37dc4e8655d17571ef230f\n",
      "Unable to find f0cc9693f09b754c639f2247677439cc\n",
      "Unable to find 6fa0147489fa12206e52e39559ea4eb2\n",
      "Unable to find 4546aa80a4dfc028eba79231ae923a1a\n",
      "Unable to find 4d2dcff1f654596199b43001505eed43\n",
      "Unable to find 605ed840eff406f1144f9155a7bbb8b7\n",
      "Unable to find 2a65a02e9bdb236f726185b06a8be2c3\n",
      "Unable to find 7608372ea4319e47aca9e40cab8fb82d\n",
      "Unable to find 4ff2d64bebbca8f945d4e454c07e1459\n",
      "Unable to find 41963a3299a192dfaeb1bbd6001b8a3e\n",
      "Unable to find 2d8118c924665b37be85ff65843a63b1\n",
      "Unable to find d9a08760a1fb633c25f1e842ebc34065\n",
      "Unable to find 9fce639669cc3e0a38f827df5c5563e6\n",
      "Unable to find 4ae1ea09c1aac49974eef578b594a2d7\n",
      "Unable to find 40ddce88bb8dc3104c02f6f3d6c0e3bc\n",
      "Unable to find d1b85aa6aefe4fc1e340f3febb0b9712\n",
      "Unable to find 6524015bde1de724e975ca06e15e4559\n",
      "Unable to find 4d5009c12b569ae6f6e8318b214c3f6b\n",
      "Unable to find 9a5cfe3e651680f1010417554c352e5b\n",
      "Unable to find 9178597fc4d88fedf497accbefb4e0c2\n",
      "Unable to find 1e56b8124128a4c72887443f91017279\n",
      "Unable to find df8025e1f6f6999740442e15e16004db\n",
      "Unable to find 9bc00e52a7af5a5d36be54340a352772\n",
      "Unable to find e8c27323a057504496c9c44bfe2ae177\n",
      "Unable to find e083772373e5fc23456ba13150880b86\n",
      "Unable to find 3c4099ffef968c7e65fe0c1bd474fe79\n",
      "Unable to find 31ee6e640df306a9520ac39e04ba7394\n",
      "Unable to find 4a411d860d19e5a71320a06d188e5fae\n",
      "Unable to find d9ea1c552cbf1f73cd123644a059a020\n",
      "Unable to find d3f419e604603b344536961e44380de2\n",
      "Unable to find a6787d60098baa2f41141c0d53e01e1c\n",
      "Unable to find 27d2a35d0e500cb7bd6af04049498a01\n",
      "Unable to find 1e592a959761c66873065cbfa44c7141\n",
      "Unable to find d51eedac57c9837d388864cb35b8af55\n",
      "Unable to find 845f48c45db6f2e57bb4563b66016ad0\n",
      "Unable to find 5b947ccd5e1671507dd40abc81a3a61e\n",
      "Unable to find dc410b4341e40b341d761a57045f8935\n",
      "Unable to find beb4a87a1279c6911a0a5476f3760134\n",
      "Unable to find cee1716266454aa94e84702a6506020a\n",
      "Unable to find 95dc28d09491d2f90e5540cc8649bb20\n",
      "Unable to find 97d574d770989db27907b10a7dffd1c6\n",
      "Unable to find 2de11259d425bcf4bbd07e5dc1148140\n",
      "Unable to find 59e0f5968672644a0876ae304390801f\n",
      "Unable to find 3ca5b2fab2e2e7001a4b45b95f9d705e\n",
      "Unable to find 7562bd1e8810b346c40127edbd3e400c\n",
      "Unable to find c23ff3785e8a9b84f923d5df9e7db25d\n",
      "Unable to find 08c8307565cf34d0a15738dcc98a1aad\n",
      "Unable to find f17fa461f3cc920672b930556fbd2156\n",
      "Unable to find f8dd3d3fd2d3faf0137bdf18be753207\n",
      "Unable to find d7160b846a2f2a768df534cba6e913e4\n",
      "Unable to find a628c353db8ae530a729a814b974f90f\n",
      "Unable to find 7e8f88267286bc33d7bcbe48ca1fcb5d\n",
      "Unable to find a4d244bef56624d1f2b624585b3a2887\n",
      "Unable to find d60372da2ca617dcdadcc908bdff2bc6\n",
      "Unable to find 3cd9ff6f326c0645ab5178a6a35d1402\n",
      "Unable to find d5fbedba3da9f02d41d72d83b1cf4f14\n",
      "Unable to find 4302b3cffa905bffb54437d830e789fa\n",
      "Unable to find 185a4323152897bdbfe621f6db467ea6\n",
      "Unable to find 30b6f51be3d409e281b534bb292b3bd0\n",
      "Unable to find d4ecd203ca465e835f5e6c48238f3dc2\n",
      "Unable to find b5adef44c28d504ee0753492d086a248\n",
      "Unable to find d9f29534e8b8d8ec9fd9d594ca5facde\n",
      "Unable to find 294d46ecf3f1cd22b18be6b0290c4819\n",
      "Unable to find 2814853e0e0c7a2ba980dc69eacab182\n",
      "Unable to find 03a11532f44e9d0a1a198e37ae8dedcd\n",
      "Unable to find b122ee2fa3a43f43d02069ed4ce5df27\n",
      "Unable to find d5ef349d096b6a84e651362da570d200\n",
      "Unable to find a4ddac1ad00557a58ae42d6c5164129c\n",
      "Unable to find 5711bfde6b1e1607cf07073c8ec076e2\n",
      "Unable to find 839bae71df01a0c2dfad6335bb724591\n",
      "Unable to find 023339f689dad564ac4ebf5d0b13cbff\n",
      "Unable to find f4a4e94e70262cde4044d8d6b5fe9021\n",
      "Unable to find 56aced50c5671e3124bd7de72148b771\n",
      "Unable to find f48ab5b40cc1e9a5ca98205e846d5f9f\n",
      "Unable to find 158a1b80cf98c78da8312e73cae08047\n",
      "Unable to find c223340f2cda7060eb3add4ef1e99095\n",
      "Unable to find cc2bc28ccabbf200b374ffe51ba13f41\n",
      "Unable to find c0f66bbe93cdc0f7050a2c747c11a606\n",
      "Unable to find 613a55e69a8be2e8410135f806cbc04c\n",
      "Unable to find f26cbeac92af39ac5f54e7a4ee7ed213\n",
      "Unable to find 4158f9cd913f88e8956457d11d0ce64d\n",
      "Unable to find d1c63e73cd04db3eb0154121f1488194\n",
      "Unable to find 4cdbf6893df3ced1ca75d6613ee20106\n",
      "Unable to find f0807d691e47c8a1c453c03b6e492b4a\n",
      "Unable to find 60c8d754363278bbf9021336430c52e9\n",
      "Unable to find b108c32341d1d3402d33f40722815386\n",
      "Unable to find 4dbb080642735267343968c6ab5751f3\n",
      "Unable to find bc832096099192422d6677db25240182\n",
      "Unable to find 69fbda9adb5157062241d2285026feec\n",
      "Unable to find 9f29559dad160cd59505405d2002bcee\n",
      "Unable to find 5d6829b17f9166d4a2a74930c4cef17d\n",
      "Unable to find 0969758250d4b33df9da2a569bd6957f\n",
      "Unable to find b383d83f113ffcdc7befa23296c4a1d8\n",
      "Unable to find 97b7dda0f41968b4a187927b2f338250\n",
      "Unable to find 4f8a8629ccd1be620ce00fac1a93bf4e\n",
      "Unable to find 6e31e7b26b008fe9cba3d3dfc189043c\n",
      "Unable to find a1e4b32c8131e24cfcecb965601953a8\n",
      "Unable to find d8acda19ef8e4cd297bb68bc6113b8fe\n",
      "Unable to find e8d9b655f2191b2af13fb8e675a9a47d\n",
      "Unable to find 1d7f99d7475bb7ae637960f9dd8a5635\n",
      "Unable to find c21f1d201f41600c2f4bd168cc30fce0\n",
      "Unable to find a1210e702ec6de976051f9008986c651\n",
      "Unable to find 754b4716cdc4c1ecfda4ba51ddae4e96\n",
      "Unable to find a6ae06a9c5e5794cb48486ec573953d3\n",
      "Unable to find 2f6276aa3c10fe85bab89a35b5a56985\n",
      "Unable to find b1d9ed298383ed1f66e49b3ec5c61e20\n",
      "Unable to find a291e51de3496ad88af5ab37c37cba24\n",
      "Unable to find 12706c7487d038a3a1a0c4fdb9b99747\n",
      "Unable to find 23136bcff2e7b56d46f0800cd2f31aa0\n",
      "Unable to find 1cf530ff35d14d2ec550e9b906a86200\n",
      "Unable to find 4e046f84676b7b06f9dd1b3ca17a1b85\n",
      "Unable to find 7f1d2573e543026d1728b2cdb71e2105\n",
      "Unable to find 60a2e9e2c4d5a96e1e8ef0a08eb65567\n",
      "Unable to find 2965746fe7d94889d742943dd97236af\n",
      "Unable to find fca3ec2c04c89e6f67e70abba264812e\n",
      "Unable to find a03429c921ea7d1a64826258187538f6\n",
      "Unable to find e6f11a4d50c5014c1ed1411ebc0fae58\n",
      "Unable to find 902c4c03b497f59e5b32323fa1b4f136\n",
      "Unable to find e496b1cd0ca221857a900d92429e5da0\n",
      "Unable to find 9b4e8481f1ac750d9af8391a9e726bc7\n",
      "Unable to find 2a75ec16e83c135c835dbadef48e05fa\n",
      "Unable to find 3e4b6a6494aeb0419d598911a351d205\n",
      "Unable to find 07cef61f688bbe64b02170191275b393\n",
      "Unable to find face672009915e31bf568203ce0fe31e\n",
      "Unable to find e29be705467c52ffc5941aed6b5a406a\n",
      "Unable to find b237afa07d6d792b9af6677e57fee811\n",
      "Unable to find c4700bfd375a5b31e85f2851b1446970\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unable to find ac6e141c11eec7f4c208fb1be83fbb38\n",
      "Unable to find 91073f62281850475f0366c8fe4ca86b\n",
      "Unable to find d1f33f3acb20b61d9e945f1a5d0cd38d\n",
      "Unable to find e69e4f765b87df6a3ce43c6461b6bd9b\n",
      "Unable to find 55b9084194ac9e77c85ed20c6a641483\n",
      "Unable to find ab9eada92e8e89dabaf4c1345d4f9c93\n",
      "Unable to find 7a9f34946e3ce06bedb3757a2e4b08f2\n",
      "Unable to find 7d2494233f0d9b04d47e6a84502a5310\n",
      "Unable to find 8b6dfb3cfcc4711a41de28ff4f5fd300\n",
      "Unable to find 30650b930fb4166d245279139ac5c4af\n",
      "Unable to find 9736c9e307040a26253b798f1b7aa441\n",
      "Unable to find 29619aa46e4701e9ff303834ada4a85e\n",
      "Unable to find eb104a248f69c8a12c0526cf5c06d38d\n",
      "Unable to find 8a2d9954deab336fa560f81f3199e88e\n",
      "Unable to find afcb29186003d50c8a62a7799641dacc\n",
      "Unable to find 56309ce7da06adfc847b0e5d3f94f1e3\n",
      "Unable to find 2bfba4c8be677dd06e8b8a9a350ae7a2\n",
      "Unable to find ea45861904f761415af6ae6c2c9a9a98\n",
      "Unable to find 363ab4f319bb4a26fe5c0d067714a672\n",
      "Unable to find b6489c3bb49b107f0ec11e8b8585e05c\n",
      "Unable to find b153c436b8ff1aeefde8874f17f85e3d\n",
      "Unable to find b2a5735603c36caa8574d2d9b0d35573\n",
      "Unable to find 7efa9dcf86dc8179eea6f79f21e9099e\n",
      "Unable to find d417a13741a6f0aad2d767062734d070\n",
      "updating B01_13A igvf_b01/next1 measurement set to B01_13A_illumina\n",
      "updating B01_13A igvf_b01/next2 measurement set to B01_13A_illumina\n",
      "updating B01_13B igvf_b01/next1 measurement set to B01_13B_illumina\n",
      "updating B01_13B igvf_b01/next2 measurement set to B01_13B_illumina\n",
      "updating B01_13C igvf_b01/next1 measurement set to B01_13C_illumina\n",
      "updating B01_13C igvf_b01/next2 measurement set to B01_13C_illumina\n",
      "updating B01_13D igvf_b01/next1 measurement set to B01_13D_illumina\n",
      "updating B01_13D igvf_b01/next2 measurement set to B01_13D_illumina\n",
      "updating B01_13E igvf_b01/next1 measurement set to B01_13E_illumina\n",
      "updating B01_13E igvf_b01/next2 measurement set to B01_13E_illumina\n",
      "updating B01_13F igvf_b01/next1 measurement set to B01_13F_illumina\n",
      "updating B01_13F igvf_b01/next2 measurement set to B01_13F_illumina\n",
      "updating B01_13G igvf_b01/next1 measurement set to B01_13G_illumina\n",
      "updating B01_13G igvf_b01/next2 measurement set to B01_13G_illumina\n",
      "updating B01_13H igvf_b01/next1 measurement set to B01_13H_illumina\n",
      "updating B01_13H igvf_b01/next2 measurement set to B01_13H_illumina\n",
      "total plates 10\n",
      "total sequencing runs 16\n",
      "total subpools 217\n",
      "total subpool in run 158\n",
      "total subpool in run files 1105\n"
     ]
    }
   ],
   "source": [
    "def get_or_create_sequencing_run(name, run_date, platform, plate, accessions=None):\n",
    "    if name in sequencing_experiments:\n",
    "        return sequencing_experiments[name]\n",
    "    \n",
    "    try:\n",
    "        sequencing_run = models.SequencingRun.objects.get(name=name)\n",
    "    except models.SequencingRun.DoesNotExist:\n",
    "        sequencing_run = models.SequencingRun(\n",
    "            name=name,\n",
    "            run_date=run_date,\n",
    "            platform=platform,\n",
    "            plate=plate,\n",
    "            stranded=models.Stranded.REVERSE,\n",
    "        )\n",
    "        sequencing_run.save()\n",
    "        if accessions is not None:\n",
    "            sequencing_run.accession.set(accessions)\n",
    "            sequencing_run.save\n",
    "        \n",
    "    sequencing_experiments[name] = sequencing_run\n",
    "    return sequencing_run\n",
    "\n",
    "def get_or_create_subpool_in_run(subpool, sequencing_run, raw_reads=None, measurement_set=None):\n",
    "    try:\n",
    "        subpool_in_run = models.SubpoolInRun.objects.get(subpool=subpool, sequencing_run=sequencing_run)\n",
    "    except models.SubpoolInRun.DoesNotExist:\n",
    "        subpool_in_run = models.SubpoolInRun(\n",
    "            subpool=subpool, \n",
    "            sequencing_run=sequencing_run,\n",
    "            raw_reads=None,\n",
    "            status=models.RunStatus.PASS,\n",
    "            measurement_set=measurement_set,\n",
    "        )\n",
    "        subpool_in_run.save()\n",
    "        \n",
    "    sequencing_subpools[name] = subpool_in_run\n",
    "    return subpool_in_run\n",
    "\n",
    "\n",
    "def lane_or_none(value):\n",
    "    if pandas.isnull(value):\n",
    "        return None\n",
    "    elif isinstance(value, float):\n",
    "        return int(value)\n",
    "    elif value.startswith(\"L\"):\n",
    "        return int(value[1:])\n",
    "    else:\n",
    "        return int(value)\n",
    "\n",
    "\n",
    "fastqs = pandas.read_csv(\n",
    "    \"fastq_metadata.tsv\", sep=\"\\t\",\n",
    "    dtype={\n",
    "        \"sample_id\": str,\n",
    "        \"read_id\": str,\n",
    "        \"fragment\": str,\n",
    "        \"sequencer_run\": str,\n",
    "        \"lane\": str,\n",
    "    }).sort_values([\"experiment\", \"library_id\", \"lane_id\", \"read_id\"])\n",
    "fastqs.head()\n",
    "\n",
    "for i, row in fastqs.iterrows():\n",
    "    plate_name = row.experiment.upper()\n",
    "    library_id = int(row.library_id)\n",
    "    fastq_relative = Path(row.filename)\n",
    "    sequencing_run_name = str(Path(fastq_relative.parts[0])/fastq_relative.parts[1])\n",
    "    sequencer = row.get(\"sequencer\")\n",
    "    run_date = row[\"ctime\"]\n",
    "    md5sum = row[\"md5sum\"]\n",
    "\n",
    "    if row[\"read_id\"] in ('I1',):\n",
    "        continue\n",
    "\n",
    "    subpool = models.Subpool.objects.get(plate__name=plate_name, index=library_id)\n",
    "    plate = subpool.plate\n",
    "\n",
    "    platform = {\n",
    "        \"A00850\": models.Platform.objects.get(display_name=\"Novaseq 6000\"),\n",
    "        \"VH00582\": models.Platform.objects.get(display_name=\"Nextseq 2000\"),\n",
    "        \"nanopore\": models.Platform.objects.get(display_name=\"Oxford Nanopore\"),  # TODO: figure out a machine id\n",
    "    }[sequencer]\n",
    "\n",
    "    sequencing_run = get_or_create_sequencing_run(sequencing_run_name, run_date, platform, plate, accessions=None)\n",
    "    subpool_run = get_or_create_subpool_in_run(subpool=subpool, sequencing_run=sequencing_run)\n",
    "    \n",
    "    file_set_name = None\n",
    "    try:\n",
    "        file_set_alias = submitted_sheets[\"sequence_file\"].set_index(\"md5sum\").loc[row.md5sum, \"file_set\"]\n",
    "        file_set_name = file_set_alias.split(\":\")[1]\n",
    "    except KeyError:\n",
    "        \n",
    "        print(f\"Unable to find {row.md5sum}\")\n",
    "\n",
    "    if file_set_name is None:\n",
    "        pass\n",
    "    elif subpool_run.measurement_set is None:\n",
    "        subpool_run.measurement_set = measurement_sets[file_set_name]\n",
    "        subpool_run.save()\n",
    "        print(f\"updating {subpool_run} measurement set to {measurement_sets[file_set_name]}\")\n",
    "    else:\n",
    "        assert subpool_run.measurement_set.name == file_set_name, f\"SubpoolInRun measurement_set mismatch {subpool_run.measurement_set.name} {file_set_name}\"\n",
    "        \n",
    "\n",
    "    accession_info = submitted_sequence_file.get(str(fastq_relative))\n",
    "    subpool_file = models.SubpoolInRunFile(\n",
    "        subpool_run=subpool_run,\n",
    "        md5sum=md5sum,\n",
    "        filename=str(fastq_relative),\n",
    "        flowcell_id=row[\"flowcell_id\"],\n",
    "        lane=lane_or_none(row[\"lane\"]),\n",
    "        fragment=int_or_none(row[\"fragment\"]),\n",
    "        read=row[\"read_id\"],\n",
    "    )\n",
    "    subpool_file.save()\n",
    "\n",
    "    if accession_info is not None:\n",
    "        accession = models.Accession(\n",
    "            namespace=accession_info[\"namespace\"],\n",
    "            name=accession_info[\"name\"],\n",
    "            uuid=accession_info[\"uuid\"],\n",
    "            url=accession_info[\"url\"],\n",
    "        )\n",
    "        accession.save()                \n",
    "        subpool_file.accession.set([accession])\n",
    "        subpool_file.save()\n",
    "\n",
    "\n",
    "        \n",
    "print(\"total plates\", models.SplitSeqPlate.objects.count())\n",
    "print(\"total sequencing runs\", models.SequencingRun.objects.count())\n",
    "print(\"total subpools\", models.Subpool.objects.count())\n",
    "print(\"total subpool in run\", models.SubpoolInRun.objects.count())\n",
    "print(\"total subpool in run files\", models.SubpoolInRunFile.objects.count())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "revolutionary-hearts",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "preliminary-washington",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
