{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "satisfied-traffic",
   "metadata": {},
   "source": [
    "# Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "nearby-variation",
   "metadata": {},
   "source": [
    "UCI has a complex google doc they've been using to track their samples so far.\n",
    "\n",
    "https://docs.google.com/spreadsheets/d/13M6-Ry6oXgkx94BHZOGioYPI6F_hWDqjGgcaNu2JNYs/edit#gid=1838362486\n",
    "\n",
    "Also there's a metadata spreadsheet for IGVF at\n",
    "https://docs.google.com/spreadsheets/d/1BLMledzmqOqXnJHzpijgw91IOs-9tSlVeZDG_MtXddk/edit#gid=1284120531"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "increased-bottle",
   "metadata": {},
   "outputs": [],
   "source": [
    "import bz2\n",
    "from collections import Counter, namedtuple\n",
    "import datetime\n",
    "from io import StringIO\n",
    "import numpy\n",
    "import os\n",
    "import pandas\n",
    "from pathlib import Path\n",
    "import re\n",
    "from subprocess import run, PIPE\n",
    "import sys\n",
    "import zoneinfo\n",
    "\n",
    "os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'mousedemo.settings')\n",
    "os.environ[\"DJANGO_ALLOW_ASYNC_UNSAFE\"] = \"true\"\n",
    "\n",
    "import django\n",
    "from django.contrib.auth import get_user_model\n",
    "from django.db import DEFAULT_DB_ALIAS\n",
    "\n",
    "MOUSEDEMO = str(Path(\"mousedemo\").absolute())\n",
    "if MOUSEDEMO not in sys.path:\n",
    "    sys.path.append(MOUSEDEMO)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "divided-checkout",
   "metadata": {},
   "source": [
    "# Pre-initialization setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "orange-optics",
   "metadata": {},
   "source": [
    "Backup the old ones and make a new database file before we initialize the database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "effective-palace",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a new database\n",
    "result = run([\"python3\", \"manage.py\", \"check\"], capture_output=True)\n",
    "\n",
    "if len(result.stderr) > 0:\n",
    "    print(result.stderr.decode(\"utf-8\"))\n",
    "    \n",
    "result.check_returncode()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "documented-cross",
   "metadata": {},
   "outputs": [],
   "source": [
    "initial = Path(\"igvf_mice/migrations/0001_initial.py\")\n",
    "if initial.exists():\n",
    "    initial.unlink()\n",
    "    \n",
    "result = run([\"python3\", \"manage.py\", \"makemigrations\", \"igvf_mice\"], capture_output=True)\n",
    "\n",
    "if len(result.stderr) > 0:\n",
    "    print(result.stderr.decode(\"utf-8\"))\n",
    "    \n",
    "result.check_returncode()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "magnetic-soccer",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "renaming db3.sqlite3 to db4.sqlite3\n",
      "renaming db2.sqlite3 to db3.sqlite3\n",
      "renaming db1.sqlite3 to db2.sqlite3\n",
      "renaming db.sqlite3 to db1.sqlite3\n"
     ]
    }
   ],
   "source": [
    "# Backup several old sqlite database files\n",
    "\n",
    "db_name = Path(\"db.sqlite3\")\n",
    "\n",
    "backup_names = []\n",
    "for i in range(0, 4):\n",
    "    backup_names.append(Path(\"db{}.sqlite3\".format(i+1)))\n",
    "\n",
    "if backup_names[-1].exists():\n",
    "    backup_names[-1].unlink()\n",
    "\n",
    "if db_name.exists():    \n",
    "    for i in reversed(range(0, 3)):\n",
    "        if backup_names[i].exists():\n",
    "            print(\"renaming {} to {}\".format(backup_names[i], backup_names[i+1]))\n",
    "            backup_names[i].rename(backup_names[i+1])\n",
    "\n",
    "    db_name.rename(backup_names[0])\n",
    "    print(\"renaming {} to {}\".format(db_name, backup_names[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "organizational-resort",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a new database\n",
    "result = run([\"python3\", \"manage.py\", \"migrate\"], capture_output=True)\n",
    "\n",
    "if len(result.stderr) > 0:\n",
    "    print(result.stderr.decode(\"utf-8\"))\n",
    "    \n",
    "result.check_returncode()\n",
    "    \n",
    "assert Path(\"db.sqlite3\").exists()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "blessed-shareware",
   "metadata": {},
   "source": [
    "# Setup users"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "failing-advantage",
   "metadata": {},
   "source": [
    "Now that we have a fresh clean database, lets create user accounts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "aboriginal-reform",
   "metadata": {},
   "outputs": [],
   "source": [
    "django.setup()\n",
    "\n",
    "from django.contrib.auth.models import ContentType\n",
    "from mousedemo import settings\n",
    "from igvf_mice import models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "arranged-restaurant",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "guest\n"
     ]
    }
   ],
   "source": [
    "# Create accounts.\n",
    "with open(\"pw\", \"rt\") as instream:\n",
    "    pw = instream.read().strip()\n",
    "\n",
    "user_model = get_user_model()\n",
    "user_model._default_manager.db_manager(DEFAULT_DB_ALIAS).create_superuser(\n",
    "    username=\"diane\",\n",
    "    password=pw,   \n",
    ")\n",
    "\n",
    "# load igvf_mice permissions    \n",
    "permissions = set()\n",
    "for ct in ContentType.objects.filter(app_label=\"igvf_mice\"):\n",
    "    permissions.update(ct.permission_set.all())\n",
    "\n",
    "with open(\"guest.pw\", \"rt\") as instream:\n",
    "    pw = instream.read().strip()\n",
    "\n",
    "user_model = get_user_model()\n",
    "guest = user_model._default_manager.db_manager(DEFAULT_DB_ALIAS).create_user(\n",
    "    username=\"guest\",\n",
    "    password=pw,\n",
    "    is_staff=True,\n",
    ")\n",
    "guest.user_permissions.set(permissions)\n",
    "print(guest)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "innovative-interim",
   "metadata": {},
   "outputs": [],
   "source": [
    "guest = user_model._default_manager.db_manager(DEFAULT_DB_ALIAS).get(username=\"guest\")\n",
    "diane = guest = user_model._default_manager.db_manager(DEFAULT_DB_ALIAS).get(username=\"diane\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "grateful-black",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert guest.is_staff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "covered-argument",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<QuerySet []>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "guest.user_permissions.all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "realistic-ocean",
   "metadata": {},
   "outputs": [],
   "source": [
    "def validate_alias(alias):\n",
    "    alias_pattern = re.compile(\"^(?:j-michael-cherry|ali-mortazavi|barbara-wold|lior-pachter|grant-macgregor|kim-green|mark-craven|qiongshi-lu|audrey-gasch|robert-steiner|jesse-engreitz|thomas-quertermous|anshul-kundaje|michael-bassik|will-greenleaf|marlene-rabinovitch|lars-steinmetz|jay-shendure|nadav-ahituv|martin-kircher|danwei-huangfu|michael-beer|anna-katerina-hadjantonakis|christina-leslie|alexander-rudensky|laura-donlin|hannah-carter|bing-ren|kyle-gaulton|maike-sander|charles-gersbach|gregory-crawford|tim-reddy|ansuman-satpathy|andrew-allen|gary-hon|nikhil-munshi|w-lee-kraus|lea-starita|doug-fowler|luca-pinello|guillaume-lettre|benhur-lee|daniel-bauer|richard-sherwood|benjamin-kleinstiver|marc-vidal|david-hill|frederick-roth|mikko-taipale|anne-carpenter|hyejung-won|karen-mohlke|michael-love|jason-buenrostro|bradley-bernstein|hilary-finucane|chongyuan-luo|noah-zaitlen|kathrin-plath|roy-wollman|jason-ernst|zhiping-weng|manuel-garber|xihong-lin|alan-boyle|ryan-mills|jie-liu|maureen-sartor|joshua-welch|stephen-montgomery|alexis-battle|livnat-jerby|jonathan-pritchard|predrag-radivojac|sean-mooney|harinder-singh|nidhi-sahni|jishnu-das|hao-wu|sreeram-kannan|hongjun-song|alkes-price|soumya-raychaudhuri|shamil-sunyaev|len-pennacchio|axel-visel|jill-moore|ting-wang|feng-yue|igvf|igvf-dacc):[a-zA-Z\\\\d_$.+!*,()'-]+(?:\\\\s[a-zA-Z\\\\d_$.+!*,()'-]+)*$\")\n",
    "    \n",
    "    if alias_pattern.match(alias) is None:\n",
    "        raise ValueError(\"Invalid alias\")\n",
    "    else:\n",
    "        return True\n",
    "\n",
    "\n",
    "try:\n",
    "    validate_alias(\"ali-mortazavi:194_B6CASTF1/J_10F_20\")\n",
    "    assert False, \"This should fail\"\n",
    "except ValueError:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "conservative-recycling",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "canadian-slide",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "schedule - all\n",
      "Line information\n",
      "mice\n",
      "Planning\n",
      "DissectionSorting\n",
      "Counting\n",
      "Sample Tissue IDs\n",
      "Samples - 8 founders\n",
      "Samples - F1s\n",
      "Samples - F1s_6mo\n",
      "Samples - CClines\n",
      "Samples - Bridge\n",
      "Founder Samples into experiment\n",
      "F1 Samples into experiment\n",
      "CC Samples into experiment\n",
      "Bridge samples into experiment\n",
      "PBMCs\n",
      "Founder Nuclei isolation Sample\n",
      "F1 Nuclei isolation SampleID\n",
      "Plate setups\n",
      "Experiment\n",
      "nreads_per_sublibrary_novaseq\n",
      "nreads_per_sublibrary_nextseq\n",
      "ONT Sequencing\n",
      "DCC\n",
      "Tissue_body_weight_metadata\n",
      "Failed_samples\n",
      "Sheet40\n",
      "tissue weights\n",
      "practice mice\n",
      "CellDrop_Demo\n",
      "Gastroc_numbers\n",
      "Gastroc_all_reps\n",
      "Recount for gene capture\n",
      "Gastroc_testing\n",
      "Cerebellum testing\n",
      "IGVF002 Pilot Samples\n",
      "IGVF002 Pulverized Samples\n",
      "IGVF002 Plate Setup\n",
      "IGVF002 Metadata\n"
     ]
    }
   ],
   "source": [
    "#spreadsheet_name = \"IGVF_Split-seq_20230118.xlsx\"\n",
    "lizs_sheet_name = \"https://woldlab.caltech.edu/nextcloud/index.php/s/eEtjBfDqQFnLpSS/download\"\n",
    "\n",
    "book = pandas.ExcelFile(lizs_sheet_name)\n",
    "for name in book.sheet_names:\n",
    "    print(name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "golden-crack",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading igvftst IGVF_003 https://woldlab.caltech.edu/nextcloud/index.php/s/nBnE6j2RBMkpM9i/download\n",
      "igvftst rodent_donor (64, 25)\n",
      "igvftst tissue (192, 9)\n",
      "igvftst measurement_set (17, 12)\n",
      "igvftst sequence_file (246, 14)\n",
      "reading igvftst IGVF_b01 https://woldlab.caltech.edu/nextcloud/index.php/s/5cJteSWgitN5BDM/download\n",
      "next_sheet rodent_donor (8, 25)\n",
      "next_sheet after duplicate filter (8, 25)\n",
      "igvftst rodent_donor (72, 25)\n",
      "next_sheet tissue (8, 9)\n",
      "next_sheet after duplicate filter (8, 9)\n",
      "igvftst tissue (200, 9)\n",
      "igvftst multiplexed_sample (8, 8)\n",
      "next_sheet measurement_set (8, 13)\n",
      "next_sheet after duplicate filter (8, 13)\n",
      "igvftst measurement_set (25, 13)\n",
      "next_sheet sequence_file (32, 14)\n",
      "next_sheet after duplicate filter (32, 14)\n",
      "igvftst sequence_file (278, 14)\n",
      "igvftst configuration_file (16, 8)\n",
      "reading igvf IGVF_b01 https://woldlab.caltech.edu/nextcloud/index.php/s/HTbfN6btm3kqJXb/download\n",
      "igvf rodent_donor (8, 25)\n",
      "igvf tissue (8, 9)\n",
      "igvf multiplexed_sample (8, 7)\n",
      "igvf measurement_set (8, 12)\n",
      "igvf sequence_file (32, 20)\n",
      "reading igvf IGVF_003 https://woldlab.caltech.edu/nextcloud/index.php/s/wRFxdcknisFfM2t/download\n",
      "next_sheet rodent_donor (65, 25)\n",
      "next_sheet after duplicate filter (65, 25)\n",
      "igvf rodent_donor (73, 25)\n",
      "next_sheet tissue (192, 9)\n",
      "next_sheet after duplicate filter (192, 9)\n",
      "igvf tissue (200, 9)\n",
      "next_sheet multiplexed_sample (16, 7)\n",
      "next_sheet after duplicate filter (16, 7)\n",
      "igvf multiplexed_sample (24, 7)\n",
      "next_sheet measurement_set (16, 13)\n",
      "next_sheet after duplicate filter (16, 13)\n",
      "igvf measurement_set (24, 13)\n",
      "next_sheet sequence_file (246, 20)\n",
      "next_sheet after duplicate filter (246, 20)\n",
      "igvf sequence_file (278, 20)\n",
      "reading igvf IGVF_004 https://woldlab.caltech.edu/nextcloud/index.php/s/8ygLzeyjAQ9CxXn/download\n",
      "next_sheet rodent_donor (64, 25)\n",
      "next_sheet after duplicate filter (0, 25)\n",
      "igvf rodent_donor (73, 25)\n",
      "next_sheet tissue (128, 9)\n",
      "next_sheet after duplicate filter (64, 9)\n",
      "igvf tissue (264, 9)\n",
      "next_sheet multiplexed_sample (16, 7)\n",
      "next_sheet after duplicate filter (16, 7)\n",
      "igvf multiplexed_sample (40, 7)\n",
      "next_sheet measurement_set (16, 13)\n",
      "next_sheet after duplicate filter (16, 13)\n",
      "igvf measurement_set (40, 13)\n",
      "next_sheet sequence_file (246, 20)\n",
      "next_sheet after duplicate filter (246, 20)\n",
      "igvf sequence_file (524, 20)\n",
      "reading igvf IGVF_005 https://woldlab.caltech.edu/nextcloud/index.php/s/2Pba74eLFwmFfgS/download\n",
      "next_sheet rodent_donor (65, 25)\n",
      "next_sheet after duplicate filter (0, 25)\n",
      "igvf rodent_donor (73, 25)\n",
      "next_sheet tissue (192, 9)\n",
      "next_sheet after duplicate filter (128, 9)\n",
      "igvf tissue (392, 9)\n",
      "next_sheet multiplexed_sample (16, 7)\n",
      "next_sheet after duplicate filter (16, 7)\n",
      "igvf multiplexed_sample (56, 7)\n",
      "next_sheet measurement_set (16, 13)\n",
      "next_sheet after duplicate filter (16, 13)\n",
      "igvf measurement_set (56, 13)\n",
      "next_sheet sequence_file (242, 20)\n",
      "next_sheet after duplicate filter (242, 20)\n",
      "igvf sequence_file (766, 20)\n",
      "reading igvf IGVF_007 https://woldlab.caltech.edu/nextcloud/index.php/s/7rQKGSTS7937kTW/download\n",
      "next_sheet rodent_donor (65, 25)\n",
      "next_sheet after duplicate filter (0, 25)\n",
      "igvf rodent_donor (73, 25)\n",
      "next_sheet tissue (256, 9)\n",
      "next_sheet after duplicate filter (128, 9)\n",
      "igvf tissue (520, 9)\n",
      "next_sheet multiplexed_sample (16, 7)\n",
      "next_sheet after duplicate filter (16, 7)\n",
      "igvf multiplexed_sample (72, 7)\n",
      "next_sheet measurement_set (16, 13)\n",
      "next_sheet after duplicate filter (16, 13)\n",
      "igvf measurement_set (72, 13)\n",
      "next_sheet sequence_file (242, 20)\n",
      "next_sheet after duplicate filter (242, 20)\n",
      "igvf sequence_file (1008, 20)\n",
      "reading igvf IGVF_008 https://woldlab.caltech.edu/nextcloud/index.php/s/iECMTJF2iPMy3kT/download\n",
      "next_sheet rodent_donor (65, 25)\n",
      "next_sheet after duplicate filter (1, 25)\n",
      "igvf rodent_donor (74, 25)\n",
      "next_sheet tissue (192, 9)\n",
      "next_sheet after duplicate filter (64, 9)\n",
      "igvf tissue (584, 9)\n",
      "next_sheet multiplexed_sample (9, 7)\n",
      "next_sheet after duplicate filter (9, 7)\n",
      "igvf multiplexed_sample (81, 7)\n",
      "next_sheet measurement_set (9, 13)\n",
      "next_sheet after duplicate filter (9, 13)\n",
      "igvf measurement_set (81, 13)\n",
      "next_sheet sequence_file (76, 20)\n",
      "next_sheet after duplicate filter (76, 20)\n",
      "igvf sequence_file (1084, 20)\n",
      "reading igvf IGVF_008B https://woldlab.caltech.edu/nextcloud/index.php/s/YF46eNqSASJC8fr/download\n",
      "next_sheet rodent_donor (65, 25)\n",
      "next_sheet after duplicate filter (0, 25)\n",
      "igvf rodent_donor (74, 25)\n",
      "next_sheet tissue (192, 9)\n",
      "next_sheet after duplicate filter (0, 9)\n",
      "igvf tissue (584, 9)\n",
      "next_sheet multiplexed_sample (11, 7)\n",
      "next_sheet after duplicate filter (11, 7)\n",
      "igvf multiplexed_sample (92, 7)\n",
      "next_sheet measurement_set (11, 13)\n",
      "next_sheet after duplicate filter (11, 13)\n",
      "igvf measurement_set (92, 13)\n",
      "next_sheet sequence_file (86, 20)\n",
      "next_sheet after duplicate filter (86, 20)\n",
      "igvf sequence_file (1170, 20)\n",
      "reading igvf IGVF_009 https://woldlab.caltech.edu/nextcloud/index.php/s/PtNn96zR9XbybFZ/download\n",
      "next_sheet rodent_donor (65, 25)\n",
      "next_sheet after duplicate filter (0, 25)\n",
      "igvf rodent_donor (74, 25)\n",
      "next_sheet tissue (128, 9)\n",
      "next_sheet after duplicate filter (64, 9)\n",
      "igvf tissue (648, 9)\n",
      "next_sheet multiplexed_sample (16, 7)\n",
      "next_sheet after duplicate filter (16, 7)\n",
      "igvf multiplexed_sample (108, 7)\n",
      "next_sheet measurement_set (16, 13)\n",
      "next_sheet after duplicate filter (16, 13)\n",
      "igvf measurement_set (108, 13)\n",
      "next_sheet sequence_file (246, 20)\n",
      "next_sheet after duplicate filter (246, 20)\n",
      "igvf sequence_file (1416, 20)\n",
      "reading igvf IGVF_010 https://woldlab.caltech.edu/nextcloud/index.php/s/CocTP9pePJobyPs/download\n",
      "next_sheet rodent_donor (70, 25)\n",
      "next_sheet after duplicate filter (5, 25)\n",
      "igvf rodent_donor (79, 25)\n",
      "next_sheet tissue (128, 9)\n",
      "next_sheet after duplicate filter (64, 9)\n",
      "igvf tissue (712, 9)\n",
      "next_sheet multiplexed_sample (16, 7)\n",
      "next_sheet after duplicate filter (16, 7)\n",
      "igvf multiplexed_sample (124, 7)\n",
      "next_sheet measurement_set (16, 13)\n",
      "next_sheet after duplicate filter (16, 13)\n",
      "igvf measurement_set (124, 13)\n",
      "next_sheet sequence_file (246, 20)\n",
      "next_sheet after duplicate filter (246, 20)\n",
      "igvf sequence_file (1662, 20)\n",
      "reading igvf IGVF_011 https://woldlab.caltech.edu/nextcloud/index.php/s/Sm72GtmdN4QsXJr/download\n",
      "next_sheet rodent_donor (70, 25)\n",
      "next_sheet after duplicate filter (0, 25)\n",
      "igvf rodent_donor (79, 25)\n",
      "next_sheet tissue (192, 9)\n",
      "next_sheet after duplicate filter (0, 9)\n",
      "igvf tissue (712, 9)\n",
      "next_sheet multiplexed_sample (16, 7)\n",
      "next_sheet after duplicate filter (16, 7)\n",
      "igvf multiplexed_sample (140, 7)\n",
      "next_sheet measurement_set (16, 13)\n",
      "next_sheet after duplicate filter (16, 13)\n",
      "igvf measurement_set (140, 13)\n",
      "next_sheet sequence_file (246, 20)\n",
      "next_sheet after duplicate filter (246, 20)\n",
      "igvf sequence_file (1908, 20)\n",
      "reading igvf IGVF_012 https://woldlab.caltech.edu/nextcloud/index.php/s/MmatjmzQgofeM2Y/download\n",
      "next_sheet rodent_donor (11, 25)\n",
      "next_sheet after duplicate filter (2, 25)\n",
      "igvf rodent_donor (81, 25)\n",
      "next_sheet tissue (14, 9)\n",
      "next_sheet after duplicate filter (6, 9)\n",
      "igvf tissue (718, 9)\n",
      "next_sheet multiplexed_sample (8, 7)\n",
      "next_sheet after duplicate filter (8, 7)\n",
      "igvf multiplexed_sample (148, 7)\n",
      "next_sheet measurement_set (8, 14)\n",
      "next_sheet after duplicate filter (8, 14)\n",
      "igvf measurement_set (148, 14)\n",
      "next_sheet sequence_file (32, 20)\n",
      "next_sheet after duplicate filter (32, 20)\n",
      "igvf sequence_file (1940, 20)\n",
      "reading igvf IGVF_013 https://woldlab.caltech.edu/nextcloud/index.php/s/EHret5a6Sjiro4q/download\n",
      "next_sheet rodent_donor (96, 25)\n",
      "next_sheet after duplicate filter (32, 25)\n",
      "igvf rodent_donor (113, 25)\n",
      "next_sheet tissue (96, 9)\n",
      "next_sheet after duplicate filter (96, 9)\n",
      "igvf tissue (814, 9)\n",
      "next_sheet multiplexed_sample (15, 7)\n",
      "next_sheet after duplicate filter (15, 7)\n",
      "igvf multiplexed_sample (163, 7)\n",
      "next_sheet measurement_set (15, 14)\n",
      "next_sheet after duplicate filter (15, 14)\n",
      "igvf measurement_set (163, 15)\n",
      "next_sheet sequence_file (174, 20)\n",
      "next_sheet after duplicate filter (174, 20)\n",
      "igvf sequence_file (2114, 20)\n",
      "reading igvf IGVF_014 https://woldlab.caltech.edu/nextcloud/index.php/s/GRQCaSZ3WgeonzK/download\n",
      "next_sheet rodent_donor (8, 25)\n",
      "next_sheet after duplicate filter (8, 25)\n",
      "igvf rodent_donor (121, 25)\n",
      "next_sheet tissue (8, 9)\n",
      "next_sheet after duplicate filter (8, 9)\n",
      "igvf tissue (822, 9)\n",
      "next_sheet multiplexed_sample (8, 7)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "next_sheet after duplicate filter (8, 7)\n",
      "igvf multiplexed_sample (171, 7)\n",
      "next_sheet measurement_set (8, 14)\n",
      "next_sheet after duplicate filter (8, 14)\n",
      "igvf measurement_set (171, 15)\n",
      "next_sheet sequence_file (16, 20)\n",
      "next_sheet after duplicate filter (16, 20)\n",
      "igvf sequence_file (2130, 20)\n"
     ]
    }
   ],
   "source": [
    "def load_submitted_sheets():\n",
    "    submitted_book_names = {\n",
    "        # Sheets on sandbox\n",
    "        \"igvftst\": {\n",
    "            \"IGVF_003\": \"https://woldlab.caltech.edu/nextcloud/index.php/s/nBnE6j2RBMkpM9i/download\",\n",
    "            \"IGVF_b01\": \"https://woldlab.caltech.edu/nextcloud/index.php/s/5cJteSWgitN5BDM/download\",\n",
    "        },\n",
    "        # Sheets on data. (production)\n",
    "        \"igvf\": {\n",
    "            \"IGVF_b01\": \"https://woldlab.caltech.edu/nextcloud/index.php/s/HTbfN6btm3kqJXb/download\",\n",
    "            \"IGVF_003\": \"https://woldlab.caltech.edu/nextcloud/index.php/s/wRFxdcknisFfM2t/download\",\n",
    "            \"IGVF_004\": \"https://woldlab.caltech.edu/nextcloud/index.php/s/8ygLzeyjAQ9CxXn/download\",\n",
    "            \"IGVF_005\": \"https://woldlab.caltech.edu/nextcloud/index.php/s/2Pba74eLFwmFfgS/download\",\n",
    "            # IGVF_006 failed.\n",
    "            \"IGVF_007\": \"https://woldlab.caltech.edu/nextcloud/index.php/s/7rQKGSTS7937kTW/download\",\n",
    "            \"IGVF_008\": \"https://woldlab.caltech.edu/nextcloud/index.php/s/iECMTJF2iPMy3kT/download\",\n",
    "            \"IGVF_008B\": \"https://woldlab.caltech.edu/nextcloud/index.php/s/YF46eNqSASJC8fr/download\",\n",
    "            \"IGVF_009\": \"https://woldlab.caltech.edu/nextcloud/index.php/s/PtNn96zR9XbybFZ/download\",\n",
    "            \"IGVF_010\": \"https://woldlab.caltech.edu/nextcloud/index.php/s/CocTP9pePJobyPs/download\",\n",
    "            \"IGVF_011\": \"https://woldlab.caltech.edu/nextcloud/index.php/s/Sm72GtmdN4QsXJr/download\",\n",
    "            \"IGVF_012\": \"https://woldlab.caltech.edu/nextcloud/index.php/s/MmatjmzQgofeM2Y/download\",\n",
    "            \"IGVF_013\": \"https://woldlab.caltech.edu/nextcloud/index.php/s/EHret5a6Sjiro4q/download\",\n",
    "            \"IGVF_014\": \"https://woldlab.caltech.edu/nextcloud/index.php/s/GRQCaSZ3WgeonzK/download\",\n",
    "        }\n",
    "    }\n",
    "    submitted_sheets = {}\n",
    "    for accession_prefix in submitted_book_names:\n",
    "        for book_name in submitted_book_names[accession_prefix]:\n",
    "            print(\"reading {} {} {}\".format(accession_prefix, book_name, submitted_book_names[accession_prefix][book_name]))\n",
    "            book = pandas.ExcelFile(submitted_book_names[accession_prefix][book_name])\n",
    "\n",
    "            for sheet_name in book.sheet_names:\n",
    "                if sheet_name not in submitted_sheets.get(accession_prefix, {}):\n",
    "                    submitted_sheets.setdefault(accession_prefix, {})[sheet_name] = book.parse(sheet_name)\n",
    "                else:\n",
    "                    previous_sheet = submitted_sheets[accession_prefix][sheet_name]\n",
    "                    next_sheet = book.parse(sheet_name)\n",
    "                    print(\"next_sheet {} {}\".format(sheet_name, next_sheet.shape))\n",
    "                    \n",
    "                    already_seen = set(previous_sheet[\"accession\"])\n",
    "                    need_to_add = set(next_sheet[\"accession\"]).difference(already_seen)\n",
    "                    next_sheet = next_sheet[next_sheet[\"accession\"].isin(need_to_add)]\n",
    "                    print(\"next_sheet after duplicate filter {}\".format(next_sheet.shape))\n",
    "                    \n",
    "                    submitted_sheets.setdefault(accession_prefix, {})[sheet_name] = pandas.concat([previous_sheet, next_sheet])\n",
    "                print(accession_prefix, sheet_name, submitted_sheets[accession_prefix][sheet_name].shape)\n",
    "\n",
    "    return submitted_sheets\n",
    "\n",
    "submitted_sheets = load_submitted_sheets()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "split-payday",
   "metadata": {},
   "source": [
    "Converters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "occupational-dealing",
   "metadata": {},
   "outputs": [],
   "source": [
    "def truncate(model):\n",
    "    table_name = model._meta.db_table\n",
    "    assert \"\\\\\" not in table_name\n",
    "    with django.db.connection.cursor() as cursor:\n",
    "        # I don't know why the sql params didn't work. this is a sql vulnerability waiting to happen\n",
    "        cursor.execute(\"delete from \\\"{}\\\"\".format(table_name))\n",
    "        cursor.execute(\"DELETE FROM SQLITE_SEQUENCE WHERE name=\\\"{}\\\"\".format(table_name))\n",
    "        cursor.fetchone()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "divine-incident",
   "metadata": {},
   "outputs": [],
   "source": [
    "def int_or_none(x):\n",
    "    if x in (\"N/A\", '#DIV/0!', '-'):\n",
    "        return None\n",
    "    elif pandas.isnull(x):\n",
    "        return None\n",
    "    else:\n",
    "        return int(x)\n",
    "    \n",
    "def int_or_0(x):\n",
    "    if pandas.isnull(x):\n",
    "        return 0\n",
    "    else:\n",
    "        return int(x)\n",
    "\n",
    "def float_or_none(x):\n",
    "    if pandas.isnull(x):\n",
    "        return None\n",
    "    else:\n",
    "        return float(x)\n",
    "    \n",
    "def str_or_empty(x):\n",
    "    if pandas.isnull(x):\n",
    "        return \"\"\n",
    "    else:\n",
    "        return x\n",
    "\n",
    "def str_or_none(x):\n",
    "    if pandas.isnull(x):\n",
    "        return None\n",
    "    else:\n",
    "        return x\n",
    "    \n",
    "def date_or_none(x):\n",
    "    if pandas.isnull(x):\n",
    "        return None\n",
    "    elif isinstance(x, datetime.datetime):\n",
    "        return x.date()\n",
    "    elif x in (\"-\"):\n",
    "        return None\n",
    "    else:\n",
    "        return x\n",
    "\n",
    "def datetime_or_none(x):\n",
    "    if pandas.isnull(x):\n",
    "        return None\n",
    "    else:\n",
    "        return x\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "parental-trigger",
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_strain(strain):\n",
    "    strain = strain.replace(\"F1/J\", \"F1J\")\n",
    "    return strain\n",
    "    \n",
    "assert normalize_strain(\"B6WSBF1/J\") == \"B6WSBF1J\"\n",
    "assert normalize_strain(\"195_B6CASTF1/J_10M_20\") == \"195_B6CASTF1J_10M_20\"\n",
    "\n",
    "def validate_mouse_age_sex(value):\n",
    "    sex = value[-1]\n",
    "    age = value[0:-1]\n",
    "    \n",
    "    assert sex in (\"M\", \"F\")\n",
    "    assert isinstance(int(age), int)\n",
    "    \n",
    "validate_mouse_age_sex(\"10F\")\n",
    "\n",
    "mouse_name_tuple = namedtuple(\"mouse_name_tuple\", [\"mouse_id\", \"mouse_strain\", \"mouse_age_sex\"])\n",
    "def parse_mouse_name(mouse_name):    \n",
    "    mouse_id_end = mouse_name.find(\"_\")\n",
    "    mouse_age_sex_start = mouse_name.rfind(\"_\") + 1\n",
    "    \n",
    "    mouse_id = mouse_name[0:mouse_id_end]\n",
    "    mouse_strain = normalize_strain(mouse_name[mouse_id_end+1:mouse_age_sex_start-1])\n",
    "    mouse_age_sex = mouse_name[mouse_age_sex_start:]\n",
    "    validate_mouse_age_sex(mouse_age_sex)\n",
    "    return mouse_name_tuple(mouse_id, mouse_strain, mouse_age_sex)\n",
    "\n",
    "assert parse_mouse_name(\"477_CC030_10M\") == (\"477\", \"CC030\", \"10M\")\n",
    "assert parse_mouse_name(\"239_TREM2R47HNSS_HO_10M\") == (\"239\", \"TREM2R47HNSS_HO\", \"10M\")\n",
    "assert parse_mouse_name(\"238_B6WSBF1/J_10F\") == (\"238\", \"B6WSBF1J\", \"10F\")\n",
    "\n",
    "mouse_tissue_tuple = namedtuple(\"mouse_tissue_tuple\", [\"mouse_id\", \"mouse_strain\", \"mouse_age_sex\", \"tissue_id\"])\n",
    "def parse_mouse_tissue(mouse_tissue):\n",
    "    fields = mouse_tissue.split(\"_\")\n",
    "    \n",
    "    mouse_id = fields.pop(0)\n",
    "    mouse_strain = fields.pop(0)\n",
    "    if mouse_strain in (\"TREM2R47HNSS\",):\n",
    "        mouse_strain = \"_\".join([mouse_strain, fields.pop(0)])\n",
    "    mouse_strain = normalize_strain(mouse_strain)\n",
    "    mouse_age_sex = fields.pop(0)\n",
    "    validate_mouse_age_sex(mouse_age_sex)\n",
    "    tissue_id = fields.pop(0)\n",
    "    assert len(fields) == 0\n",
    "    \n",
    "    return mouse_tissue_tuple(mouse_id, mouse_strain, mouse_age_sex, tissue_id)\n",
    "    \n",
    "assert parse_mouse_tissue(\"096_WSBJ_10F_15\") == (\"096\", \"WSBJ\", \"10F\", \"15\")\n",
    "assert parse_mouse_tissue(\"239_TREM2R47HNSS_HO_10M_01\") == (\"239\", \"TREM2R47HNSS_HO\", \"10M\", \"01\")\n",
    "\n",
    "def normalize_plate_name(name):\n",
    "    name = name.upper()\n",
    "\n",
    "    return name\n",
    "\n",
    "assert normalize_plate_name(\"igvf_003\") == \"IGVF_003\"\n",
    "assert normalize_plate_name(\"IGVF_008B\") == \"IGVF_008B\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "injured-smell",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'B6WSBF1J'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "normalize_strain(\"B6WSBF1/J\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "noble-personal",
   "metadata": {},
   "source": [
    "# initialize general use tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "linear-civilian",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Delete later: switch to just using an enum for the accessions.\n",
    "#\n",
    "#accession_namespaces = [\n",
    "#    {\"name\": \"ENCODE\", \"homepage\": \"https://www.encodeproject.org\", \"accession_prefix\": \"encode\"},\n",
    "#    {\"name\": \"ENCODE test\", \"homepage\": \"https://test.encodedcc.org\", \"accession_prefix\": \"encodetst\"},\n",
    "#    {\"name\": \"IGVF\", \"homepage\": \"https://data.igvf.org\", \"accession_prefix\": \"igvf\"},\n",
    "#    {\"name\": \"IGVF test\", \"homepage\": \"https://sandbox.igvf.org\", \"accession_prefix\": \"igvftst\"},\n",
    "#]\n",
    "#\n",
    "#if models.AccessionNamespace.objects.count() > 0:\n",
    "#    truncate(models.AccessionNamespace)\n",
    "#\n",
    "#for row in accession_namespaces:\n",
    "#    record = models.AccessionNamespace(\n",
    "#        name=row[\"name\"],\n",
    "#        homepage=row[\"homepage\"],\n",
    "#        accession_prefix=row[\"accession_prefix\"],\n",
    "#    )\n",
    "#    record.save()\n",
    "#\n",
    "#igvf_namespace = models.AccessionNamespace.objects.get(name=\"IGVF\")\n",
    "#igvf_test_namespace = models.AccessionNamespace.objects.get(name=\"IGVF test\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "psychological-march",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_sources():\n",
    "    source = [\n",
    "        {\n",
    "            \"name\": \"jackson-labs\",\n",
    "            \"display_name\": \"The Jackson Laboratory\", \n",
    "            \"homepage\": \"http://www.jax.org/index.html\", \n",
    "            \"igvf_id\": \"/sources/jackson-labs/\",\n",
    "        },\n",
    "        {\n",
    "            \"name\": \"parse-biosciences\",\n",
    "            \"display_name\": \"Parse Bioscience\", \n",
    "            \"homepage\": \"https://www.parsebiosciences.com/\",\n",
    "            \"igvf_id\": \"/sources/parse-biosciences/\",\n",
    "        },\n",
    "        {\n",
    "            \"name\": \"illumina\",\n",
    "            \"display_name\": \"Illumina\", \n",
    "            \"homepage\": \"https://www.illumina.com\", \n",
    "            \"igvf_id\": \"/sources/illumina/\",\n",
    "        },\n",
    "        {\n",
    "            \"name\": \"unc-csbio\",\n",
    "            \"display_name\": \"UNC Systems Genetics\",\n",
    "            \"homepage\": \"https://csbio.unc.edu\",\n",
    "            \"igvf_id\": \"/sources/unc-csbio/\",\n",
    "        }\n",
    "    ]\n",
    "\n",
    "    if models.Source.objects.count() > 0:\n",
    "        truncate(models.Source)\n",
    "\n",
    "    for row in source:\n",
    "        record = models.Source(\n",
    "            name=row[\"name\"],\n",
    "            homepage=row[\"homepage\"],\n",
    "            igvf_id=row[\"igvf_id\"])\n",
    "        record.save()\n",
    "\n",
    "load_sources()\n",
    "jax_source = models.Source.objects.get(name=\"jackson-labs\")\n",
    "unc_csbio_source = models.Source.objects.get(name=\"unc-csbio\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "metallic-spider",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Should this include machine model?\n",
    "\n",
    "def load_platforms():\n",
    "    platforms = [\n",
    "        {\"display_name\": \"Nextseq 2000\", \"name\": \"nextseq2000\", \"family\": \"illumina\", \"igvf_id\": \"/platform-terms/EFO_0010963/\"},\n",
    "        {\"display_name\": \"Novaseq 6000\", \"name\": \"novaseq6000\", \"family\": \"illumina\", \"igvf_id\": \"/platform-terms/EFO_0008637/\"},\n",
    "        {\"display_name\": \"Oxford Nanopore MinION\", \"name\": \"minion\", \"family\": \"nanopore\", \"igvf_id\": \"/platform-terms/EFO_0008632/\"},\n",
    "        {\"display_name\": \"Oxford Nanopore PromethION\", \"name\": \"promethion\", \"family\": \"nanopore\", \"igvf_id\": \"/platform-terms/EFO_0008634/\"},\n",
    "        {\"display_name\": \"Pac Bio\", \"name\": \"pacbio\", \"family\": \"pacbio\",},\n",
    "    ]\n",
    "\n",
    "    for row in platforms:\n",
    "        record = models.Platform(\n",
    "            name=row[\"name\"],\n",
    "            display_name=row[\"display_name\"],\n",
    "            family=row[\"family\"],\n",
    "            igvf_id=row.get(\"igvf_id\"),\n",
    "            #source=row.get(\"source\"),\n",
    "        )\n",
    "        record.save()\n",
    "        \n",
    "load_platforms()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "manufactured-account",
   "metadata": {},
   "source": [
    "# Library kit type & version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "innocent-copying",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_library_reagents():\n",
    "    library_construction_reagent = [\n",
    "        {\"name\": \"wt-mega\", \"display_name\": \"Parse WT Mega\", \"version\": \"v2\", \"source\": \"parse-biosciences\"},\n",
    "        {\"name\": \"wt\", \"display_name\": \"Parse WT\", \"version\": \"v2\", \"source\": \"parse-biosciences\"},\n",
    "    ]\n",
    "\n",
    "    if models.LibraryConstructionReagent.objects.count() > 0:\n",
    "        truncate(models.LibraryConstructionReagent)\n",
    "\n",
    "    for row in library_construction_reagent:\n",
    "        try:\n",
    "            source = models.Source.objects.get(name=row[\"source\"])\n",
    "\n",
    "            record = models.LibraryConstructionReagent(\n",
    "                name=row[\"name\"],\n",
    "                display_name=row[\"display_name\"],\n",
    "                version=row[\"version\"],\n",
    "                source=source,\n",
    "            )\n",
    "            record.save()        \n",
    "        except models.Source.DoesNotExist:\n",
    "            print(f\"Couldn't find {row['source']}\")\n",
    "\n",
    "load_library_reagents()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "pressed-payday",
   "metadata": {},
   "source": [
    "# Load LibraryBarcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "convenient-syndication",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<LibraryConstructionReagent: Parse WT Mega v2>"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "models.LibraryConstructionReagent.objects.get(name=\"wt-mega\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "faced-christianity",
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_udi_to_code(value):\n",
    "    prefix = \"UDI_Plate_WT_\"\n",
    "    assert value.startswith(prefix)\n",
    "    barcode_number = int(value[len(prefix):])\n",
    "    return \"UDI{:02}\".format(barcode_number)\n",
    "\n",
    "assert normalize_udi_to_code(\"UDI_Plate_WT_1\") == \"UDI01\"\n",
    "assert normalize_udi_to_code(\"UDI_Plate_WT_45\") == \"UDI45\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "approved-hours",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_barcodes():\n",
    "    if models.LibraryBarcode.objects.count() > 0:\n",
    "        truncate(models.LibraryBarcode)\n",
    "\n",
    "    # Load WT Mega v2 first barcode\n",
    "    bc1 = pandas.read_csv(\"bc_data_n192_v4.csv\")\n",
    "\n",
    "    mega_reagent = models.LibraryConstructionReagent.objects.get(name=\"wt-mega\", version=\"v2\")\n",
    "\n",
    "    for i, row in bc1.iterrows():\n",
    "        record = models.LibraryBarcode(\n",
    "            reagent=mega_reagent,\n",
    "            name=row[\"uid\"],\n",
    "            code=row[\"well\"],\n",
    "            i7_sequence=row[\"sequence\"],\n",
    "            barcode_type=row[\"type\"], # TODO what do the codes mean again T,R?\n",
    "            well_position=row[\"well\"],\n",
    "        )\n",
    "        record.save()\n",
    "\n",
    "    mega_subpool = [\n",
    "        (\"1\", \"CAGATC\"),\n",
    "        (\"2\", \"ACTTGA\"),\n",
    "        (\"3\", \"GATCAG\"),\n",
    "        (\"4\", \"TAGCTT\"),\n",
    "        (\"5\", \"ATGTCA\"),\n",
    "        (\"6\", \"CTTGTA\"),\n",
    "        (\"7\", \"AGTCAA\"),\n",
    "        (\"8\", \"AGTTCC\"),\n",
    "        (\"9\", \"GAGTGG\"),\n",
    "        (\"10\", \"CCGTCC\"),\n",
    "        (\"11\", \"GTAGAG\"),\n",
    "        (\"12\", \"GTCCGC\"),\n",
    "        (\"13\", \"GTGAAA\"),\n",
    "        (\"14\", \"GTGGCC\"),\n",
    "        (\"15\", \"GTTTCG\"),\n",
    "        (\"16\", \"CGTACG\"),\n",
    "    ]\n",
    "    subpool = pandas.DataFrame(mega_subpool, columns=[\"code\", \"i7_sequence\"])\n",
    "\n",
    "    for i, row in subpool.iterrows():\n",
    "        record = models.LibraryBarcode(\n",
    "            reagent=mega_reagent,\n",
    "            name=row[\"code\"],\n",
    "            code=row[\"code\"],\n",
    "            i7_sequence=row[\"i7_sequence\"],\n",
    "        )\n",
    "        record.save()\n",
    "\n",
    "    regular_reagent = models.LibraryConstructionReagent.objects.get(name=\"wt\", version=\"v2\")    \n",
    "    regular_subpool = [\n",
    "        (\"1\", \"CAGATC\"),\n",
    "        (\"2\", \"ACTTGA\"),\n",
    "        (\"3\", \"GATCAG\"),\n",
    "        (\"4\", \"TAGCTT\"),\n",
    "        (\"5\", \"ATGTCA\"),\n",
    "        (\"6\", \"CTTGTA\"),\n",
    "        (\"7\", \"AGTCAA\"),\n",
    "        (\"8\", \"AGTTCC\"),\n",
    "    ]\n",
    "    subpool = pandas.DataFrame(regular_subpool, columns=[\"code\",\"i7_sequence\"])\n",
    "    for i, row in subpool.iterrows():\n",
    "        record = models.LibraryBarcode(\n",
    "            reagent=regular_reagent,\n",
    "            name=row[\"code\"],\n",
    "            code=row[\"code\"],\n",
    "            i7_sequence=row[\"i7_sequence\"],\n",
    "        )\n",
    "        record.save()\n",
    "\n",
    "\n",
    "    udi_wt_reagent = models.LibraryConstructionReagent.objects.get(name=\"wt-mega\", version=\"v2\")\n",
    "    udi_wt_barcodes = pandas.read_excel(\"https://support.parsebiosciences.com/hc/en-us/article_attachments/12988277964436\")\n",
    "    for i, row in udi_wt_barcodes.iterrows():\n",
    "        record = models.LibraryBarcode(\n",
    "            reagent=udi_wt_reagent,\n",
    "            name=row[\"Index name\"],\n",
    "            code=normalize_udi_to_code(row[\"Index name\"]),\n",
    "            i7_sequence=row[\"i7_index\"],\n",
    "            i5_sequence=row[\"i5_index\"],\n",
    "            well_position=row[\"well_position\"],\n",
    "        )\n",
    "        record.save()\n",
    "        \n",
    "load_barcodes()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "hungry-guidance",
   "metadata": {},
   "source": [
    "# Initialize TissueOntology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "dimensional-paintball",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Samples - 8 founders (1621, 1) 22\n",
      "Samples - Bridge (416, 1) 44\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Counter({'Hypothalamus/Pituitary': 97,\n",
       "         'Cerebellum': 81,\n",
       "         'Cortex/Hippocampus left': 85,\n",
       "         'Cortex/Hippocampus right': 85,\n",
       "         'Liver': 97,\n",
       "         'Heart': 97,\n",
       "         'Lung': 81,\n",
       "         'Adrenal': 97,\n",
       "         'Kidney': 81,\n",
       "         'Gonads -1 Ovary': 41,\n",
       "         'Gonads -2 Oviduct': 41,\n",
       "         'Perigonadal fat': 81,\n",
       "         'Brown fat': 81,\n",
       "         'Soleus': 81,\n",
       "         'Plantaris': 81,\n",
       "         'Gastrocnemius': 82,\n",
       "         'TA': 81,\n",
       "         'EDL': 81,\n",
       "         'Tail': 121,\n",
       "         'PBMC - WBC': 121,\n",
       "         'Gonads -1 Testis': 40,\n",
       "         'Gonads -2 Epididymis': 40,\n",
       "         'Left cortex': 24,\n",
       "         'Right cortex': 24,\n",
       "         'Left hippocampus': 24,\n",
       "         'Right hippocampus': 24,\n",
       "         'Testis left': 8,\n",
       "         'Testis right': 8,\n",
       "         'Epididymis left': 8,\n",
       "         'Epididymis right': 8,\n",
       "         'Kidney left': 16,\n",
       "         'Kidney right': 16,\n",
       "         'Gastrocnemius left': 16,\n",
       "         'Gastrocnemius right': 16,\n",
       "         'TA left': 4,\n",
       "         'TA right': 4,\n",
       "         'Ovary': 1,\n",
       "         'Oviduct': 1,\n",
       "         'Gonads -1 (Ovary)': 7,\n",
       "         'Gonads - 2 (Oviduct)': 7,\n",
       "         'Cortex left': 12,\n",
       "         'Cortex right': 12,\n",
       "         'Hippocampus left': 12,\n",
       "         'Hippocampus right': 12})"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def count_seen_tissues():\n",
    "    seen_tissues = Counter()\n",
    "    for sample_sheet_name in [\"Samples - 8 founders\", \"Samples - Bridge\"]:\n",
    "        tissue_sheet = pandas.read_excel(\n",
    "            lizs_sheet_name, \n",
    "            sample_sheet_name,\n",
    "            usecols=[\"Tissue\"]).dropna(axis=0, how='all')\n",
    "\n",
    "        for i, row in tissue_sheet.iterrows():\n",
    "            if not pandas.isnull(row[\"Tissue\"]):\n",
    "                seen_tissues[row[\"Tissue\"]] += 1\n",
    "        print(sample_sheet_name, tissue_sheet.shape, len(seen_tissues))\n",
    "        \n",
    "    return seen_tissues\n",
    "\n",
    "seen_tissues = count_seen_tissues()\n",
    "seen_tissues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "responsible-aberdeen",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "set()\n"
     ]
    }
   ],
   "source": [
    "tissue_dissection_to_ontology_map = {\n",
    "    'Hypothalamus/Pituitary': [(\"UBERON:0001898\",\"hypothalamus\"), (\"UBERON:0000007\",\"pituitary gland\")],\n",
    "    'Cerebellum': [(\"UBERON:0002037\",\"cerebellum\")],\n",
    "    # Hippocampus might be:\n",
    "    #   Hippocampal formation UBERON:0002421 https://www.ebi.ac.uk/ols/ontologies/uberon/terms?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FUBERON_0002421\n",
    "    #   Layer of hippocampus UBERON:0002305 https://www.ebi.ac.uk/ols/ontologies/uberon/terms?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FUBERON_0002305\n",
    "    'Cortex/Hippocampus left': [(\"NTR:0000646\",\"left cerebral cortex\"), (\"NTR:0000750\", \"Hippocampal formation left\")],\n",
    "    'Cortex/Hippocampus right': [(\"NTR:0000647\",\"right cerebral cortex\"), (\"NTR:0000751\", \"Hippocampal formation right\")],\n",
    "    'Liver': [(\"UBERON:0002107\", \"liver\")],\n",
    "    'Heart': [(\"UBERON:0000948\",\"heart\")],\n",
    "    'Lung': [(\"UBERON:0002048\",\"lung\")],\n",
    "    'Adrenal': [(\"UBERON:0002369\",\"adrenal gland\")],\n",
    "    'Kidney': [(\"UBERON:0002113\",\"kidney\")],\n",
    "    'Kidney left': [(\"UBERON:0002113\",\"kidney\")],\n",
    "    'Kidney right': [(\"UBERON:0002113\",\"kidney\")],\n",
    "    \"Ovary\": [(\"UBERON:0000992\", \"ovary\")],    \n",
    "    'Gonads -1 Ovary': [(\"UBERON:0000992\", \"ovary\")],\n",
    "    'Gonads -1 (Ovary)': [(\"UBERON:0000992\", \"ovary\")],\n",
    "    'Oviduct': [(\"UBERON:0000993\",\"oviduct\")],\n",
    "    'Gonads -2 Oviduct': [(\"UBERON:0000993\",\"oviduct\")],\n",
    "    'Gonads -2 (Oviduct)': [(\"UBERON:0000993\",\"oviduct\")],\n",
    "    'Gonads - 2 (Oviduct)': [(\"UBERON:0000993\",\"oviduct\")],\n",
    "    'Perigonadal fat': [(\"UBERON:0003428\", \"gonadal fat pad\")], #\"is closer\" to dissection\n",
    "    'Brown fat': [(\"UBERON:0001348\",\"brown adipose tissue\")], \n",
    "    'Soleus': [(\"UBERON:0001389\",\"soleus muscle\")],\n",
    "    'Plantaris': [(\"UBERON:0011905\",\"plantaris\")],\n",
    "    'Gastrocnemius': [(\"UBERON:0001388\",\"gastrocnemius\")],\n",
    "    'Gastrocnemius left': [(\"UBERON:0001388\",\"gastrocnemius\")],\n",
    "    'Gastrocnemius right': [(\"UBERON:0001388\",\"gastrocnemius\")],\n",
    "    'TA': [(\"UBERON:0001385\",\"tibialis anterior\")],\n",
    "    'TA left': [(\"UBERON:0001385\",\"tibialis anterior\")],\n",
    "    'TA right': [(\"UBERON:0001385\",\"tibialis anterior\")],\n",
    "    'EDL': [(\"UBERON:0001386\",\"extensor digitorum longus\")],\n",
    "    'Tail': [(\"UBERON:0002415\",\"tail\")],\n",
    "    'PBMC - WBC': [(\"CL:2000001\", \"peripheral blood mononuclear cell\")],\n",
    "    'Gonads -1 Testis': [(\"UBERON:0000473\",\"testis\")],\n",
    "    'Gonads -2 Epididymis': [(\"UBERON:0001301\",\"epididymis\")],\n",
    "    \"Left cortex\": [(\"NTR:0000646\",\"left cerebral cortex\"),],\n",
    "    \"Cortex left\": [(\"NTR:0000646\",\"left cerebral cortex\"),],\n",
    "    \"Right cortex\": [(\"NTR:0000647\",\"right cerebral cortex\"),],\n",
    "    \"Cortex right\": [(\"NTR:0000647\",\"right cerebral cortex\"),],\n",
    "    \"Left hippocampus\": [(\"NTR:0000750\", \"Hippocampal formation left\")],\n",
    "    \"Hippocampus left\": [(\"NTR:0000750\", \"Hippocampal formation left\")],\n",
    "    \"Right hippocampus\": [(\"NTR:0000751\", \"Hippocampal formation right\")],\n",
    "    \"Hippocampus right\": [(\"NTR:0000751\", \"Hippocampal formation right\")],\n",
    "    # does left & right matter?\n",
    "    \"Testis left\": [(\"UBERON:0000473\", \"Testis\")],\n",
    "    \"Testis right\": [(\"UBERON:0000473\", \"Testis\")],\n",
    "    \"Epididymis left\": [(\"UBERON:0001301\", \"Epididymis\")],\n",
    "    \"Epididymis right\": [(\"UBERON:0001301\", \"Epididymis\")],\n",
    "    \"Pancreas\": [(\"UBERON:0001264\", \"Pancreas\")],\n",
    "    \"Lungs\": [(\"UBERON:0002048\", \"Lung\")],\n",
    "}\n",
    "\n",
    "print(set(seen_tissues).difference(tissue_dissection_to_ontology_map.keys()))\n",
    "\n",
    "assert len(set(seen_tissues).difference(tissue_dissection_to_ontology_map.keys())) == 0, \"Add in more names to term ids\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "green-orange",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_ontology_terms():\n",
    "    term_details = pandas.read_csv(\"obo.tsv.bz2\", compression=\"bz2\", sep=\"\\t\", index_col=\"term_id\")\n",
    "\n",
    "    if models.OntologyTerm.objects.count() > 0:\n",
    "        truncate(models.OntologyTerm)\n",
    "\n",
    "    for key in tissue_dissection_to_ontology_map:\n",
    "        for term_curie, term_name in tissue_dissection_to_ontology_map[key]:\n",
    "            if term_curie.startswith(\"NTR:\"):\n",
    "                description=None\n",
    "            else:\n",
    "                details = term_details.loc[term_curie]\n",
    "                description = details.description\n",
    "\n",
    "            record = models.OntologyTerm(\n",
    "                curie=term_curie,\n",
    "                name=term_name,\n",
    "                description=description\n",
    "            )\n",
    "            record.save()\n",
    "\n",
    "load_ontology_terms()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "otherwise-runner",
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_urls = {\n",
    "    \"AJ\": \"https://www.jax.org/strain/000646\",\n",
    "    \"B6J\": \"https://www.jax.org/strain/000664\",\n",
    "    \"129S1J\": \"https://www.jax.org/strain/002448\",\n",
    "    \"NODJ\": \"https://www.jax.org/strain/001976\",\n",
    "    \"NZOJ\": \"https://www.jax.org/strain/002105\",\n",
    "    \"CASTJ\": \"https://www.jax.org/strain/000928\",\n",
    "    \"PWKJ\": \"https://www.jax.org/strain/003715\",\n",
    "    \"WSBJ\": \"https://www.jax.org/strain/001145\",\n",
    "    'B6129SF1J': \"https://www.jax.org/strain/101043\",\n",
    "    'B6AF1J': \"https://www.jax.org/strain/100002\",\n",
    "    'B6CASTF1J': None,\n",
    "    'B6NODF1J': None,\n",
    "    'B6NZOF1J': None,\n",
    "    'B6PWKF1J': None,\n",
    "    'B6WSBF1J': \"https://www.jax.org/strain/019019\",\n",
    "    'TREM2R47HNSS_HO': 'https://www.jax.org/strain/034036',\n",
    "    'CC001': \"https://www.jax.org/strain/021238\",\n",
    "    'CC002': \"https://www.jax.org/strain/021236\",\n",
    "    'CC003': \"https://www.jax.org/strain/021237\",\n",
    "    'CC004': \"https://www.jax.org/strain/020944\",\n",
    "    'CC005': \"https://www.jax.org/strain/020945\",\n",
    "    'CC006': \"https://www.jax.org/strain/022869\",\n",
    "    'CC007': \"https://www.jax.org/strain/029625\",\n",
    "    'CC008': \"https://www.jax.org/strain/026971\",\n",
    "    'CC009': \"https://www.jax.org/strain/018856\",\n",
    "    'CC010': \"https://www.jax.org/strain/021889\",\n",
    "    'CC011': \"https://www.jax.org/strain/018854\",\n",
    "    'CC012': \"https://www.jax.org/strain/028409\",\n",
    "    'CC013': \"https://www.jax.org/strain/021892\",\n",
    "    'CC015': \"https://www.jax.org/strain/018859\",\n",
    "    'CC017': \"https://www.jax.org/strain/022870\",\n",
    "    'CC018': \"https://www.jax.org/strain/021890\",\n",
    "    'CC024': \"https://www.jax.org/strain/021891\",\n",
    "    'CC025': \"https://www.jax.org/strain/018857\",\n",
    "    'CC028': \"https://www.jax.org/strain/025126\",\n",
    "    'CC029': \"https://www.jax.org/strain/026972\",\n",
    "    'CC030': \"https://www.jax.org/strain/025426\",\n",
    "    'CC032': \"https://www.jax.org/strain/020946\",\n",
    "    'CC036': \"https://www.jax.org/strain/025127\",\n",
    "    'CC037': \"https://www.jax.org/strain/025423\",\n",
    "    'CC038': None,\n",
    "    'CC041': \"https://www.jax.org/strain/021893\",\n",
    "    'CC055': None,\n",
    "    'CC057': \"https://www.jax.org/strain/024683\",\n",
    "    'CC060': \"https://www.jax.org/strain/026427\",\n",
    "    'CC062': None,\n",
    "    'CC065': None,\n",
    "    'CC071': None,\n",
    "    'CC074': \"https://www.jax.org/strain/018855\",\n",
    "}\n",
    "\n",
    "# used as name. Should this be \n",
    "strain_igvf_id = {\n",
    "    \"AJ\": \"A/J (AJ)\",\n",
    "    \"B6J\": \"C57BL/6J (B6)\",\n",
    "    \"129S1J\": \"129S1/SvImJ (129)\",\n",
    "    \"NODJ\": \"NOD/ShiLtJ (NOD)\",\n",
    "    \"NZOJ\": \"NZO/HlLtJ (NZO)\",\n",
    "    \"CASTJ\": \"CAST/EiJ (CAST)\",\n",
    "    \"PWKJ\": \"PWK/PhJ (PWK)\",\n",
    "    \"WSBJ\": \"WSB/EiJ (WSB)\",\n",
    "    'B6129SF1J': None, #\"B6129SF1/J (B6 female x 129S male F1)\"\n",
    "    'B6AF1J':    None, #\"B6AF1/J (B6 female x AJ male F1)\"\n",
    "    'B6CASTF1J': None, #\"B6CASTF1/J (B6 female x CAST male F1)\"\n",
    "    'B6NODF1J':  None, #\"B6NODF1/J (B6 female x NOD male F1)\"\n",
    "    'B6NZOF1J':  None, #\"B6NZOF1/J (B6 female x NZO male F1)\"\n",
    "    'B6PWKF1J':  None, #\"B6PWKF1/J (B6 female x PWK male F1)\"\n",
    "    'B6WSBF1J':  None, #\"B6WSBF1/J (B6 female x WSB male F1)\"\n",
    "    \"TREM2R47HNSS_HO\": \"B6(SJL)-Trem2<em1Aduci>/J\", #\"B6(SJL)-Tre2ᵉᵐ¹ᴬᵈᵘᶜⁱ/J\",\n",
    "    # commented out CC lines are ones we're not currently using.\n",
    "    'CC001': 'CC001/Unc',\n",
    "    'CC002': 'CC002/Unc',\n",
    "    'CC003': 'CC003/Unc',\n",
    "    'CC004': 'CC004/TauUnc',\n",
    "    'CC005': 'CC005/TauUnc',\n",
    "    'CC006': 'CC006/TauUnc',\n",
    "    'CC007': 'CC007/Unc',\n",
    "    'CC008': 'CC008/GeniUnc',\n",
    "    'CC009': 'CC009/Unc',\n",
    "    'CC010': 'CC010/GeniUnc',\n",
    "    'CC011': 'CC011/Unc',\n",
    "    'CC012': 'CC012/GeniUnc',\n",
    "    'CC013': 'CC013/GeniUnc',\n",
    "    'CC015': 'CC015/Unc',\n",
    "    #'CC016': 'CC016/GeniUnc',\n",
    "    'CC017': 'CC017/Unc',\n",
    "    'CC018': 'CC018/Unc',\n",
    "    #'CC019': 'CC019/TauUnc',\n",
    "    #'CC020': 'CC020/GeniUnc',\n",
    "    #'CC021': 'CC021/Unc',\n",
    "    #'CC022': 'CC022/GeniUnc',\n",
    "    #'CC023': 'CC023/GeniUnc',\n",
    "    'CC024': 'CC024/GeniUnc',\n",
    "    'CC025': 'CC025/GeniUnc',\n",
    "    #'CC026': 'CC026/GeniUnc',\n",
    "    #'CC027': 'CC027/GeniUnc',\n",
    "    'CC028': 'CC028/GeniUnc',\n",
    "    'CC029': 'CC029/Unc',\n",
    "    'CC030': 'CC030/GeniUnc',\n",
    "    #'CC031': 'CC031/GeniUnc',\n",
    "    'CC032': 'CC032/GeniUnc',\n",
    "    #'CC033': 'CC033/GeniUnc',\n",
    "    #'CC034': 'CC034/Unc',\n",
    "    #'CC035': 'CC035/Unc',\n",
    "    'CC036': 'CC036/Unc',\n",
    "    'CC037': 'CC037/TauUnc',\n",
    "    'CC038': 'CC038/GeniUnc',\n",
    "    #'CC039': 'CC039/Unc',\n",
    "    #'CC040': 'CC040/TauUnc',\n",
    "    'CC041': 'CC041/TauUnc',\n",
    "    #'CC042': 'CC042/GeniUnc',\n",
    "    'CC043': 'CC043/GeniUnc',\n",
    "    #'CC044': 'CC044/Unc',\n",
    "    #'CC045': 'CC045/GeniUnc',\n",
    "    #'CC046': 'CC046/Unc',\n",
    "    #'CC047': 'CC047/Unc',\n",
    "    #'CC048': 'CC048/Unc',\n",
    "    #'CC049': 'CC049/TauUnc',\n",
    "    #'CC050': 'CC050/Unc',\n",
    "    #'CC051': 'CC051/TauUnc',\n",
    "    #'CC052': 'CC052/GeniUnc',\n",
    "    #'CC053': 'CC053/Unc',\n",
    "    #'CC054': 'CC054/GeniUnc',\n",
    "    'CC055': 'CC055/TauUnc',\n",
    "    #'CC056': 'CC056/GeniUnc',\n",
    "    'CC057': 'CC057/Unc',\n",
    "    #'CC058': 'CC058/Unc',\n",
    "    #'CC059': 'CC059/TauUnc',\n",
    "    'CC060': 'CC060/Unc',\n",
    "    #'CC061': 'CC061/GeniUnc',\n",
    "    'CC062': 'CC062/Unc',\n",
    "    #'CC063': 'CC063/Unc',\n",
    "    'CC065': 'CC065/Unc',\n",
    "    #'CC068': 'CC068/TauUnc',\n",
    "    #'CC070': 'CC070/TauUnc',\n",
    "    'CC071': 'CC071/TauUnc',\n",
    "    #'CC072': 'CC072/TauUnc',\n",
    "    #'CC073': 'CC073/Unc',\n",
    "    'CC074': 'CC074/Unc',\n",
    "    #'CC075': 'CC075/Unc',\n",
    "    #'CC076': 'CC076/Unc',\n",
    "    #'CC078': 'CC078/TauUnc',\n",
    "    #'CC079': 'CC079/TauUnc',\n",
    "    #'CC080': 'CC080/TauUnc',\n",
    "    #'CC081': 'CC081/Unc',\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "committed-dollar",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Designation</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Note</th>\n",
       "      <th>Jax Catalog No</th>\n",
       "      <th>Sample CODE</th>\n",
       "      <th>Strain notes</th>\n",
       "      <th>see_also</th>\n",
       "      <th>Source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>A/J</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>000646</td>\n",
       "      <td>AJ</td>\n",
       "      <td>Yellower adrenal gland.</td>\n",
       "      <td>https://www.jax.org/strain/000646</td>\n",
       "      <td>jackson-labs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>B</td>\n",
       "      <td>C57BL/6J</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>000664</td>\n",
       "      <td>B6J</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/000664</td>\n",
       "      <td>jackson-labs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>129S1/SvImJ</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>002448</td>\n",
       "      <td>129S1J</td>\n",
       "      <td>Male skin is tougher than female skin. Gallbla...</td>\n",
       "      <td>https://www.jax.org/strain/002448</td>\n",
       "      <td>jackson-labs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D</td>\n",
       "      <td>NOD/ShiLtJ</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>001976</td>\n",
       "      <td>NODJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/001976</td>\n",
       "      <td>jackson-labs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>NZO/HlLtJ</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>002105</td>\n",
       "      <td>NZOJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/002105</td>\n",
       "      <td>jackson-labs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>F</td>\n",
       "      <td>CAST/EiJ</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>000928</td>\n",
       "      <td>CASTJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/000928</td>\n",
       "      <td>jackson-labs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>G</td>\n",
       "      <td>PWK/PhJ</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>003715</td>\n",
       "      <td>PWKJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/003715</td>\n",
       "      <td>jackson-labs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>H</td>\n",
       "      <td>WSB/EiJ</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>001145</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/001145</td>\n",
       "      <td>jackson-labs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NaN</td>\n",
       "      <td>B6129SF1/J</td>\n",
       "      <td>CC F1</td>\n",
       "      <td>101043</td>\n",
       "      <td>B6129SF1J</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/101043</td>\n",
       "      <td>jackson-labs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NaN</td>\n",
       "      <td>B6AF1/J</td>\n",
       "      <td>CC F1</td>\n",
       "      <td>101043</td>\n",
       "      <td>B6AF1J</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/100002</td>\n",
       "      <td>jackson-labs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NaN</td>\n",
       "      <td>B6CASTF1/J</td>\n",
       "      <td>CC F1</td>\n",
       "      <td>Custom C56BL/6J crossed CAST/EiJ</td>\n",
       "      <td>B6CASTF1J</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>jackson-labs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NaN</td>\n",
       "      <td>B6NODF1/J</td>\n",
       "      <td>CC F1</td>\n",
       "      <td>Custom C56BL/6J crossed NOD/ShiLtJ</td>\n",
       "      <td>B6NODF1J</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>jackson-labs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NaN</td>\n",
       "      <td>B6NZOF1/J</td>\n",
       "      <td>CC F1</td>\n",
       "      <td>Custom C56BL/6J crossed NZO/HlLtJ</td>\n",
       "      <td>B6NZOF1J</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>jackson-labs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NaN</td>\n",
       "      <td>B6PWKF1/J</td>\n",
       "      <td>CC F1</td>\n",
       "      <td>Custom C56BL/6J crossed PWK/PhJ</td>\n",
       "      <td>B6PWKF1J</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>jackson-labs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NaN</td>\n",
       "      <td>B6WSBF1/J</td>\n",
       "      <td>CC F1</td>\n",
       "      <td>019019</td>\n",
       "      <td>B6WSBF1J</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/019019</td>\n",
       "      <td>jackson-labs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TREM2R47HNSS_HO</td>\n",
       "      <td>CC Mutant</td>\n",
       "      <td>033781</td>\n",
       "      <td>TREM2R47HNSS_HO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/033781</td>\n",
       "      <td>jackson-labs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC001</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>021238</td>\n",
       "      <td>CC001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/021238</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC002</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>021236</td>\n",
       "      <td>CC002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/021236</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC003</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>021237</td>\n",
       "      <td>CC003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/021237</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC004</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>020944</td>\n",
       "      <td>CC004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/020944</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC005</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>020945</td>\n",
       "      <td>CC005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/020945</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC006</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>022869</td>\n",
       "      <td>CC006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/022869</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC007</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>029625</td>\n",
       "      <td>CC007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/029625</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC008</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>026971</td>\n",
       "      <td>CC008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/026971</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC009</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>018856</td>\n",
       "      <td>CC009</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/018856</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC010</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>021889</td>\n",
       "      <td>CC010</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/021889</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC011</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>018854</td>\n",
       "      <td>CC011</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/018854</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC012</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>028409</td>\n",
       "      <td>CC012</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/028409</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC013</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>021892</td>\n",
       "      <td>CC013</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/021892</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC015</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>018859</td>\n",
       "      <td>CC015</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/018859</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC017</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>022870</td>\n",
       "      <td>CC017</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/022870</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC018</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>021890</td>\n",
       "      <td>CC018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/021890</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC024</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>021891</td>\n",
       "      <td>CC024</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/021891</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC025</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>018857</td>\n",
       "      <td>CC025</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/018857</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC028</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>025126</td>\n",
       "      <td>CC028</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/025126</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC029</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>026972</td>\n",
       "      <td>CC029</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/026972</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC030</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>025426</td>\n",
       "      <td>CC030</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/025426</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC032</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>020946</td>\n",
       "      <td>CC032</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/020946</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC036</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>025127</td>\n",
       "      <td>CC036</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/025127</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC037</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>025423</td>\n",
       "      <td>CC037</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/025423</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC038</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td></td>\n",
       "      <td>CC038</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC041</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>021893</td>\n",
       "      <td>CC041</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/021893</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC043</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>023828</td>\n",
       "      <td>CC043</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/023828</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC055</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td></td>\n",
       "      <td>CC055</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC057</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>024683</td>\n",
       "      <td>CC057</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/024683</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC060</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>026427</td>\n",
       "      <td>CC060</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/026427</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC062</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td></td>\n",
       "      <td>CC062</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC065</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td></td>\n",
       "      <td>CC065</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC071</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td></td>\n",
       "      <td>CC071</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC074</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>018855</td>\n",
       "      <td>CC074</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/018855</td>\n",
       "      <td>unc-csbio</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Designation           Strain        Note  \\\n",
       "0            A              A/J  CC founder   \n",
       "1            B         C57BL/6J  CC founder   \n",
       "2            C      129S1/SvImJ  CC founder   \n",
       "3            D       NOD/ShiLtJ  CC founder   \n",
       "4            E        NZO/HlLtJ  CC founder   \n",
       "5            F         CAST/EiJ  CC founder   \n",
       "6            G          PWK/PhJ  CC founder   \n",
       "7            H          WSB/EiJ  CC founder   \n",
       "8          NaN       B6129SF1/J       CC F1   \n",
       "9          NaN          B6AF1/J       CC F1   \n",
       "10         NaN       B6CASTF1/J       CC F1   \n",
       "11         NaN        B6NODF1/J       CC F1   \n",
       "12         NaN        B6NZOF1/J       CC F1   \n",
       "13         NaN        B6PWKF1/J       CC F1   \n",
       "14         NaN        B6WSBF1/J       CC F1   \n",
       "15         NaN  TREM2R47HNSS_HO   CC Mutant   \n",
       "16         NaN            CC001    CC Cross   \n",
       "17         NaN            CC002    CC Cross   \n",
       "18         NaN            CC003    CC Cross   \n",
       "19         NaN            CC004    CC Cross   \n",
       "20         NaN            CC005    CC Cross   \n",
       "21         NaN            CC006    CC Cross   \n",
       "22         NaN            CC007    CC Cross   \n",
       "23         NaN            CC008    CC Cross   \n",
       "24         NaN            CC009    CC Cross   \n",
       "25         NaN            CC010    CC Cross   \n",
       "26         NaN            CC011    CC Cross   \n",
       "27         NaN            CC012    CC Cross   \n",
       "28         NaN            CC013    CC Cross   \n",
       "29         NaN            CC015    CC Cross   \n",
       "30         NaN            CC017    CC Cross   \n",
       "31         NaN            CC018    CC Cross   \n",
       "32         NaN            CC024    CC Cross   \n",
       "33         NaN            CC025    CC Cross   \n",
       "34         NaN            CC028    CC Cross   \n",
       "35         NaN            CC029    CC Cross   \n",
       "36         NaN            CC030    CC Cross   \n",
       "37         NaN            CC032    CC Cross   \n",
       "38         NaN            CC036    CC Cross   \n",
       "39         NaN            CC037    CC Cross   \n",
       "40         NaN            CC038    CC Cross   \n",
       "41         NaN            CC041    CC Cross   \n",
       "42         NaN            CC043    CC Cross   \n",
       "43         NaN            CC055    CC Cross   \n",
       "44         NaN            CC057    CC Cross   \n",
       "45         NaN            CC060    CC Cross   \n",
       "46         NaN            CC062    CC Cross   \n",
       "47         NaN            CC065    CC Cross   \n",
       "48         NaN            CC071    CC Cross   \n",
       "49         NaN            CC074    CC Cross   \n",
       "\n",
       "                        Jax Catalog No      Sample CODE  \\\n",
       "0                               000646               AJ   \n",
       "1                               000664              B6J   \n",
       "2                               002448           129S1J   \n",
       "3                               001976             NODJ   \n",
       "4                               002105             NZOJ   \n",
       "5                               000928            CASTJ   \n",
       "6                               003715             PWKJ   \n",
       "7                               001145             WSBJ   \n",
       "8                               101043        B6129SF1J   \n",
       "9                               101043           B6AF1J   \n",
       "10    Custom C56BL/6J crossed CAST/EiJ        B6CASTF1J   \n",
       "11  Custom C56BL/6J crossed NOD/ShiLtJ         B6NODF1J   \n",
       "12   Custom C56BL/6J crossed NZO/HlLtJ         B6NZOF1J   \n",
       "13     Custom C56BL/6J crossed PWK/PhJ         B6PWKF1J   \n",
       "14                              019019         B6WSBF1J   \n",
       "15                              033781  TREM2R47HNSS_HO   \n",
       "16                              021238            CC001   \n",
       "17                              021236            CC002   \n",
       "18                              021237            CC003   \n",
       "19                              020944            CC004   \n",
       "20                              020945            CC005   \n",
       "21                              022869            CC006   \n",
       "22                              029625            CC007   \n",
       "23                              026971            CC008   \n",
       "24                              018856            CC009   \n",
       "25                              021889            CC010   \n",
       "26                              018854            CC011   \n",
       "27                              028409            CC012   \n",
       "28                              021892            CC013   \n",
       "29                              018859            CC015   \n",
       "30                              022870            CC017   \n",
       "31                              021890            CC018   \n",
       "32                              021891            CC024   \n",
       "33                              018857            CC025   \n",
       "34                              025126            CC028   \n",
       "35                              026972            CC029   \n",
       "36                              025426            CC030   \n",
       "37                              020946            CC032   \n",
       "38                              025127            CC036   \n",
       "39                              025423            CC037   \n",
       "40                                                CC038   \n",
       "41                              021893            CC041   \n",
       "42                              023828            CC043   \n",
       "43                                                CC055   \n",
       "44                              024683            CC057   \n",
       "45                              026427            CC060   \n",
       "46                                                CC062   \n",
       "47                                                CC065   \n",
       "48                                                CC071   \n",
       "49                              018855            CC074   \n",
       "\n",
       "                                         Strain notes  \\\n",
       "0                             Yellower adrenal gland.   \n",
       "1                                                 NaN   \n",
       "2   Male skin is tougher than female skin. Gallbla...   \n",
       "3                                                 NaN   \n",
       "4                                                 NaN   \n",
       "5                                                 NaN   \n",
       "6                                                 NaN   \n",
       "7                                                 NaN   \n",
       "8                                                 NaN   \n",
       "9                                                 NaN   \n",
       "10                                                NaN   \n",
       "11                                                NaN   \n",
       "12                                                NaN   \n",
       "13                                                NaN   \n",
       "14                                                NaN   \n",
       "15                                                NaN   \n",
       "16                                                NaN   \n",
       "17                                                NaN   \n",
       "18                                                NaN   \n",
       "19                                                NaN   \n",
       "20                                                NaN   \n",
       "21                                                NaN   \n",
       "22                                                NaN   \n",
       "23                                                NaN   \n",
       "24                                                NaN   \n",
       "25                                                NaN   \n",
       "26                                                NaN   \n",
       "27                                                NaN   \n",
       "28                                                NaN   \n",
       "29                                                NaN   \n",
       "30                                                NaN   \n",
       "31                                                NaN   \n",
       "32                                                NaN   \n",
       "33                                                NaN   \n",
       "34                                                NaN   \n",
       "35                                                NaN   \n",
       "36                                                NaN   \n",
       "37                                                NaN   \n",
       "38                                                NaN   \n",
       "39                                                NaN   \n",
       "40                                                NaN   \n",
       "41                                                NaN   \n",
       "42                                                NaN   \n",
       "43                                                NaN   \n",
       "44                                                NaN   \n",
       "45                                                NaN   \n",
       "46                                                NaN   \n",
       "47                                                NaN   \n",
       "48                                                NaN   \n",
       "49                                                NaN   \n",
       "\n",
       "                             see_also        Source  \n",
       "0   https://www.jax.org/strain/000646  jackson-labs  \n",
       "1   https://www.jax.org/strain/000664  jackson-labs  \n",
       "2   https://www.jax.org/strain/002448  jackson-labs  \n",
       "3   https://www.jax.org/strain/001976  jackson-labs  \n",
       "4   https://www.jax.org/strain/002105  jackson-labs  \n",
       "5   https://www.jax.org/strain/000928  jackson-labs  \n",
       "6   https://www.jax.org/strain/003715  jackson-labs  \n",
       "7   https://www.jax.org/strain/001145  jackson-labs  \n",
       "8   https://www.jax.org/strain/101043  jackson-labs  \n",
       "9   https://www.jax.org/strain/100002  jackson-labs  \n",
       "10                                NaN  jackson-labs  \n",
       "11                                NaN  jackson-labs  \n",
       "12                                NaN  jackson-labs  \n",
       "13                                NaN  jackson-labs  \n",
       "14  https://www.jax.org/strain/019019  jackson-labs  \n",
       "15  https://www.jax.org/strain/033781  jackson-labs  \n",
       "16  https://www.jax.org/strain/021238     unc-csbio  \n",
       "17  https://www.jax.org/strain/021236     unc-csbio  \n",
       "18  https://www.jax.org/strain/021237     unc-csbio  \n",
       "19  https://www.jax.org/strain/020944     unc-csbio  \n",
       "20  https://www.jax.org/strain/020945     unc-csbio  \n",
       "21  https://www.jax.org/strain/022869     unc-csbio  \n",
       "22  https://www.jax.org/strain/029625     unc-csbio  \n",
       "23  https://www.jax.org/strain/026971     unc-csbio  \n",
       "24  https://www.jax.org/strain/018856     unc-csbio  \n",
       "25  https://www.jax.org/strain/021889     unc-csbio  \n",
       "26  https://www.jax.org/strain/018854     unc-csbio  \n",
       "27  https://www.jax.org/strain/028409     unc-csbio  \n",
       "28  https://www.jax.org/strain/021892     unc-csbio  \n",
       "29  https://www.jax.org/strain/018859     unc-csbio  \n",
       "30  https://www.jax.org/strain/022870     unc-csbio  \n",
       "31  https://www.jax.org/strain/021890     unc-csbio  \n",
       "32  https://www.jax.org/strain/021891     unc-csbio  \n",
       "33  https://www.jax.org/strain/018857     unc-csbio  \n",
       "34  https://www.jax.org/strain/025126     unc-csbio  \n",
       "35  https://www.jax.org/strain/026972     unc-csbio  \n",
       "36  https://www.jax.org/strain/025426     unc-csbio  \n",
       "37  https://www.jax.org/strain/020946     unc-csbio  \n",
       "38  https://www.jax.org/strain/025127     unc-csbio  \n",
       "39  https://www.jax.org/strain/025423     unc-csbio  \n",
       "40                                NaN     unc-csbio  \n",
       "41  https://www.jax.org/strain/021893     unc-csbio  \n",
       "42  https://www.jax.org/strain/023828     unc-csbio  \n",
       "43                                NaN     unc-csbio  \n",
       "44  https://www.jax.org/strain/024683     unc-csbio  \n",
       "45  https://www.jax.org/strain/026427     unc-csbio  \n",
       "46                                NaN     unc-csbio  \n",
       "47                                NaN     unc-csbio  \n",
       "48                                NaN     unc-csbio  \n",
       "49  https://www.jax.org/strain/018855     unc-csbio  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def int_csv_to_hex(x):\n",
    "    r, g, b = x.split(\",\")\n",
    "    \n",
    "    return \"#{:02x}{:02x}{:02x}\".format(int(r), int(g), int(b))\n",
    "\n",
    "def remove_citation(x):\n",
    "    return x.split(\" \")[0]\n",
    "\n",
    "#strains = pandas.read_excel(\n",
    "#    lizs_sheet_name, \n",
    "#    sheet_name=\"Line information\", \n",
    "#    usecols=[\n",
    "#        \"Designation\",\n",
    "#        \"Strain\",\n",
    "#        \"Note\",\n",
    "#        \"Jax Catalog No\",\n",
    "#        \"Sample CODE\",\n",
    "#        \"Strain notes\",\n",
    "#    ],\n",
    "#    converters={\n",
    "#        \"Strain\": remove_citation,\n",
    "#        \"Jax Catalog No\": str,\n",
    "#        \"Strain notes\": str_or_empty,\n",
    "#    }\n",
    "#).dropna(how=\"all\")\n",
    "# Remove the extra lines with the bridge samples for now. I don't want to track thta level of detail\n",
    "#bridge = strains[strains[\"Designation\"] == \"Bridge sample\"].first_valid_index() - 1\n",
    "#strains = strains.iloc[0:bridge]\n",
    "\n",
    "# Add in Strain URL\n",
    "#strains[\"see_also\"] = strains[\"Sample CODE\"].apply(lambda x: strain_urls.get(x, x))\n",
    "#strains\n",
    "\n",
    "#print(strains.to_csv(index=False))\n",
    "\n",
    "strains_csv = StringIO(\"\"\"Designation,Strain,Note,Jax Catalog No,Sample CODE,Strain notes,see_also,Source\n",
    "A,A/J,CC founder,000646,AJ,Yellower adrenal gland.,https://www.jax.org/strain/000646,jackson-labs\n",
    "B,C57BL/6J,CC founder,000664,B6J,,https://www.jax.org/strain/000664,jackson-labs\n",
    "C,129S1/SvImJ,CC founder,002448,129S1J,Male skin is tougher than female skin. Gallbladder is more filled in this strain. Adrenals are pale and smaller than A6J.,https://www.jax.org/strain/002448,jackson-labs\n",
    "D,NOD/ShiLtJ,CC founder,001976,NODJ,,https://www.jax.org/strain/001976,jackson-labs\n",
    "E,NZO/HlLtJ,CC founder,002105,NZOJ,,https://www.jax.org/strain/002105,jackson-labs\n",
    "F,CAST/EiJ,CC founder,000928,CASTJ,,https://www.jax.org/strain/000928,jackson-labs\n",
    "G,PWK/PhJ,CC founder,003715,PWKJ,,https://www.jax.org/strain/003715,jackson-labs\n",
    "H,WSB/EiJ,CC founder,001145,WSBJ,,https://www.jax.org/strain/001145,jackson-labs\n",
    ",B6129SF1/J,CC F1,101043,B6129SF1J,,https://www.jax.org/strain/101043,jackson-labs\n",
    ",B6AF1/J,CC F1,101043,B6AF1J,,https://www.jax.org/strain/100002,jackson-labs\n",
    ",B6CASTF1/J,CC F1,Custom C56BL/6J crossed CAST/EiJ,B6CASTF1J,,,jackson-labs\n",
    ",B6NODF1/J,CC F1,Custom C56BL/6J crossed NOD/ShiLtJ,B6NODF1J,,,jackson-labs\n",
    ",B6NZOF1/J,CC F1,Custom C56BL/6J crossed NZO/HlLtJ,B6NZOF1J,,,jackson-labs\n",
    ",B6PWKF1/J,CC F1,Custom C56BL/6J crossed PWK/PhJ,B6PWKF1J,,,jackson-labs\n",
    ",B6WSBF1/J,CC F1,019019,B6WSBF1J,,https://www.jax.org/strain/019019,jackson-labs\n",
    ",TREM2R47HNSS_HO,CC Mutant,033781,TREM2R47HNSS_HO,,https://www.jax.org/strain/033781,jackson-labs\n",
    ",CC001,CC Cross,021238,CC001,,https://www.jax.org/strain/021238,unc-csbio\n",
    ",CC002,CC Cross,021236,CC002,,https://www.jax.org/strain/021236,unc-csbio\n",
    ",CC003,CC Cross,021237,CC003,,https://www.jax.org/strain/021237,unc-csbio\n",
    ",CC004,CC Cross,020944,CC004,,https://www.jax.org/strain/020944,unc-csbio\n",
    ",CC005,CC Cross,020945,CC005,,https://www.jax.org/strain/020945,unc-csbio\n",
    ",CC006,CC Cross,022869,CC006,,https://www.jax.org/strain/022869,unc-csbio\n",
    ",CC007,CC Cross,029625,CC007,,https://www.jax.org/strain/029625,unc-csbio\n",
    ",CC008,CC Cross,026971,CC008,,https://www.jax.org/strain/026971,unc-csbio\n",
    ",CC009,CC Cross,018856,CC009,,https://www.jax.org/strain/018856,unc-csbio\n",
    ",CC010,CC Cross,021889,CC010,,https://www.jax.org/strain/021889,unc-csbio\n",
    ",CC011,CC Cross,018854,CC011,,https://www.jax.org/strain/018854,unc-csbio\n",
    ",CC012,CC Cross,028409,CC012,,https://www.jax.org/strain/028409,unc-csbio\n",
    ",CC013,CC Cross,021892,CC013,,https://www.jax.org/strain/021892,unc-csbio\n",
    ",CC015,CC Cross,018859,CC015,,https://www.jax.org/strain/018859,unc-csbio\n",
    ",CC017,CC Cross,022870,CC017,,https://www.jax.org/strain/022870,unc-csbio\n",
    ",CC018,CC Cross,021890,CC018,,https://www.jax.org/strain/021890,unc-csbio\n",
    ",CC024,CC Cross,021891,CC024,,https://www.jax.org/strain/021891,unc-csbio\n",
    ",CC025,CC Cross,018857,CC025,,https://www.jax.org/strain/018857,unc-csbio\n",
    ",CC028,CC Cross,025126,CC028,,https://www.jax.org/strain/025126,unc-csbio\n",
    ",CC029,CC Cross,026972,CC029,,https://www.jax.org/strain/026972,unc-csbio\n",
    ",CC030,CC Cross,025426,CC030,,https://www.jax.org/strain/025426,unc-csbio\n",
    ",CC032,CC Cross,020946,CC032,,https://www.jax.org/strain/020946,unc-csbio\n",
    ",CC036,CC Cross,025127,CC036,,https://www.jax.org/strain/025127,unc-csbio\n",
    ",CC037,CC Cross,025423,CC037,,https://www.jax.org/strain/025423,unc-csbio\n",
    ",CC038,CC Cross,,CC038,,,unc-csbio\n",
    ",CC041,CC Cross,021893,CC041,,https://www.jax.org/strain/021893,unc-csbio\n",
    ",CC043,CC Cross,023828,CC043,,https://www.jax.org/strain/023828,unc-csbio\n",
    ",CC055,CC Cross,,CC055,,,unc-csbio\n",
    ",CC057,CC Cross,024683,CC057,,https://www.jax.org/strain/024683,unc-csbio\n",
    ",CC060,CC Cross,026427,CC060,,https://www.jax.org/strain/026427,unc-csbio\n",
    ",CC062,CC Cross,,CC062,,,unc-csbio\n",
    ",CC065,CC Cross,,CC065,,,unc-csbio\n",
    ",CC071,CC Cross,,CC071,,,unc-csbio\n",
    ",CC074,CC Cross,018855,CC074,,https://www.jax.org/strain/018855,unc-csbio\"\"\")\n",
    "\n",
    "strains = pandas.read_csv(\n",
    "    strains_csv,\n",
    "    converters={\"Jax Catalog No\": str_or_none}\n",
    ")\n",
    "\n",
    "for i, row in strains.iterrows():\n",
    "    assert pandas.notnull(row[\"Strain\"]), f\"{row['Strain']} was null {i} {row}\"\n",
    "    assert pandas.notnull(row[\"Sample CODE\"]), f\"{row['Sample CODE']} was null {i} {row}\"\n",
    "    assert pandas.notnull(row[\"Source\"]), f\"{row['Source']} was null {i} {row}\"\n",
    "\n",
    "strains\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "purple-adult",
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_code_to_name = strains[[\"Strain\", \"Sample CODE\"]].set_index(\"Sample CODE\")[\"Strain\"].to_dict()\n",
    "strain_name_to_code = strains[[\"Strain\", \"Sample CODE\"]].set_index(\"Strain\")[\"Sample CODE\"].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "digital-specific",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AJ': 'A/J',\n",
       " 'B6J': 'C57BL/6J',\n",
       " '129S1J': '129S1/SvImJ',\n",
       " 'NODJ': 'NOD/ShiLtJ',\n",
       " 'NZOJ': 'NZO/HlLtJ',\n",
       " 'CASTJ': 'CAST/EiJ',\n",
       " 'PWKJ': 'PWK/PhJ',\n",
       " 'WSBJ': 'WSB/EiJ',\n",
       " 'B6129SF1J': 'B6129SF1/J',\n",
       " 'B6AF1J': 'B6AF1/J',\n",
       " 'B6CASTF1J': 'B6CASTF1/J',\n",
       " 'B6NODF1J': 'B6NODF1/J',\n",
       " 'B6NZOF1J': 'B6NZOF1/J',\n",
       " 'B6PWKF1J': 'B6PWKF1/J',\n",
       " 'B6WSBF1J': 'B6WSBF1/J',\n",
       " 'TREM2R47HNSS_HO': 'TREM2R47HNSS_HO',\n",
       " 'CC001': 'CC001',\n",
       " 'CC002': 'CC002',\n",
       " 'CC003': 'CC003',\n",
       " 'CC004': 'CC004',\n",
       " 'CC005': 'CC005',\n",
       " 'CC006': 'CC006',\n",
       " 'CC007': 'CC007',\n",
       " 'CC008': 'CC008',\n",
       " 'CC009': 'CC009',\n",
       " 'CC010': 'CC010',\n",
       " 'CC011': 'CC011',\n",
       " 'CC012': 'CC012',\n",
       " 'CC013': 'CC013',\n",
       " 'CC015': 'CC015',\n",
       " 'CC017': 'CC017',\n",
       " 'CC018': 'CC018',\n",
       " 'CC024': 'CC024',\n",
       " 'CC025': 'CC025',\n",
       " 'CC028': 'CC028',\n",
       " 'CC029': 'CC029',\n",
       " 'CC030': 'CC030',\n",
       " 'CC032': 'CC032',\n",
       " 'CC036': 'CC036',\n",
       " 'CC037': 'CC037',\n",
       " 'CC038': 'CC038',\n",
       " 'CC041': 'CC041',\n",
       " 'CC043': 'CC043',\n",
       " 'CC055': 'CC055',\n",
       " 'CC057': 'CC057',\n",
       " 'CC060': 'CC060',\n",
       " 'CC062': 'CC062',\n",
       " 'CC065': 'CC065',\n",
       " 'CC071': 'CC071',\n",
       " 'CC074': 'CC074'}"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#strain_name_to_code\n",
    "strain_code_to_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "fiscal-income",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Designation,Strain,Note,Jax Catalog No,Sample CODE,Strain notes,see_also,Source\n",
      "A,A/J,CC founder,000646,AJ,Yellower adrenal gland.,https://www.jax.org/strain/000646,jackson-labs\n",
      "B,C57BL/6J,CC founder,000664,B6J,,https://www.jax.org/strain/000664,jackson-labs\n",
      "C,129S1/SvImJ,CC founder,002448,129S1J,Male skin is tougher than female skin. Gallbladder is more filled in this strain. Adrenals are pale and smaller than A6J.,https://www.jax.org/strain/002448,jackson-labs\n",
      "D,NOD/ShiLtJ,CC founder,001976,NODJ,,https://www.jax.org/strain/001976,jackson-labs\n",
      "E,NZO/HlLtJ,CC founder,002105,NZOJ,,https://www.jax.org/strain/002105,jackson-labs\n",
      "F,CAST/EiJ,CC founder,000928,CASTJ,,https://www.jax.org/strain/000928,jackson-labs\n",
      "G,PWK/PhJ,CC founder,003715,PWKJ,,https://www.jax.org/strain/003715,jackson-labs\n",
      "H,WSB/EiJ,CC founder,001145,WSBJ,,https://www.jax.org/strain/001145,jackson-labs\n",
      ",B6129SF1/J,CC F1,101043,B6129SF1J,,https://www.jax.org/strain/101043,jackson-labs\n",
      ",B6AF1/J,CC F1,101043,B6AF1J,,https://www.jax.org/strain/100002,jackson-labs\n",
      ",B6CASTF1/J,CC F1,Custom C56BL/6J crossed CAST/EiJ,B6CASTF1J,,,jackson-labs\n",
      ",B6NODF1/J,CC F1,Custom C56BL/6J crossed NOD/ShiLtJ,B6NODF1J,,,jackson-labs\n",
      ",B6NZOF1/J,CC F1,Custom C56BL/6J crossed NZO/HlLtJ,B6NZOF1J,,,jackson-labs\n",
      ",B6PWKF1/J,CC F1,Custom C56BL/6J crossed PWK/PhJ,B6PWKF1J,,,jackson-labs\n",
      ",B6WSBF1/J,CC F1,019019,B6WSBF1J,,https://www.jax.org/strain/019019,jackson-labs\n",
      ",TREM2R47HNSS_HO,CC Mutant,033781,TREM2R47HNSS_HO,,https://www.jax.org/strain/033781,jackson-labs\n",
      ",CC001,CC Cross,021238,CC001,,https://www.jax.org/strain/021238,unc-csbio\n",
      ",CC002,CC Cross,021236,CC002,,https://www.jax.org/strain/021236,unc-csbio\n",
      ",CC003,CC Cross,021237,CC003,,https://www.jax.org/strain/021237,unc-csbio\n",
      ",CC004,CC Cross,020944,CC004,,https://www.jax.org/strain/020944,unc-csbio\n",
      ",CC005,CC Cross,020945,CC005,,https://www.jax.org/strain/020945,unc-csbio\n",
      ",CC006,CC Cross,022869,CC006,,https://www.jax.org/strain/022869,unc-csbio\n",
      ",CC007,CC Cross,029625,CC007,,https://www.jax.org/strain/029625,unc-csbio\n",
      ",CC008,CC Cross,026971,CC008,,https://www.jax.org/strain/026971,unc-csbio\n",
      ",CC009,CC Cross,018856,CC009,,https://www.jax.org/strain/018856,unc-csbio\n",
      ",CC010,CC Cross,021889,CC010,,https://www.jax.org/strain/021889,unc-csbio\n",
      ",CC011,CC Cross,018854,CC011,,https://www.jax.org/strain/018854,unc-csbio\n",
      ",CC012,CC Cross,028409,CC012,,https://www.jax.org/strain/028409,unc-csbio\n",
      ",CC013,CC Cross,021892,CC013,,https://www.jax.org/strain/021892,unc-csbio\n",
      ",CC015,CC Cross,018859,CC015,,https://www.jax.org/strain/018859,unc-csbio\n",
      ",CC017,CC Cross,022870,CC017,,https://www.jax.org/strain/022870,unc-csbio\n",
      ",CC018,CC Cross,021890,CC018,,https://www.jax.org/strain/021890,unc-csbio\n",
      ",CC024,CC Cross,021891,CC024,,https://www.jax.org/strain/021891,unc-csbio\n",
      ",CC025,CC Cross,018857,CC025,,https://www.jax.org/strain/018857,unc-csbio\n",
      ",CC028,CC Cross,025126,CC028,,https://www.jax.org/strain/025126,unc-csbio\n",
      ",CC029,CC Cross,026972,CC029,,https://www.jax.org/strain/026972,unc-csbio\n",
      ",CC030,CC Cross,025426,CC030,,https://www.jax.org/strain/025426,unc-csbio\n",
      ",CC032,CC Cross,020946,CC032,,https://www.jax.org/strain/020946,unc-csbio\n",
      ",CC036,CC Cross,025127,CC036,,https://www.jax.org/strain/025127,unc-csbio\n",
      ",CC037,CC Cross,025423,CC037,,https://www.jax.org/strain/025423,unc-csbio\n",
      ",CC038,CC Cross,,CC038,,,unc-csbio\n",
      ",CC041,CC Cross,021893,CC041,,https://www.jax.org/strain/021893,unc-csbio\n",
      ",CC043,CC Cross,023828,CC043,,https://www.jax.org/strain/023828,unc-csbio\n",
      ",CC055,CC Cross,,CC055,,,unc-csbio\n",
      ",CC057,CC Cross,024683,CC057,,https://www.jax.org/strain/024683,unc-csbio\n",
      ",CC060,CC Cross,026427,CC060,,https://www.jax.org/strain/026427,unc-csbio\n",
      ",CC062,CC Cross,,CC062,,,unc-csbio\n",
      ",CC065,CC Cross,,CC065,,,unc-csbio\n",
      ",CC071,CC Cross,,CC071,,,unc-csbio\n",
      ",CC074,CC Cross,018855,CC074,,https://www.jax.org/strain/018855,unc-csbio\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(strains.to_csv(index=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "indian-venture",
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_type_lookup = {\n",
    "    \"CC founder\": models.StrainType.FOUNDER,\n",
    "    \"CC F1\": models.StrainType.F1,\n",
    "    \"CC Cross\": models.StrainType.CROSS,\n",
    "    \"CC Mutant\": models.StrainType.MUTANT,\n",
    "}\n",
    "\n",
    "if models.MouseStrain.objects.count() > 0:\n",
    "    truncate(models.MouseStrain)\n",
    "\n",
    "# This is the version where I was pulling from Liz's spreadsheet\n",
    "#for i, row in strains.iterrows():\n",
    "#    record = models.MouseStrain(\n",
    "#        name=row[\"Sample CODE\"],\n",
    "#        display_name=strain_name[row[\"Sample CODE\"]],\n",
    "#        igvf_id=strain_igvf_id[row[\"Sample CODE\"]],\n",
    "#        strain_type=strain_type_lookup[row[\"Note\"]],\n",
    "#        jax_catalog_number=row[\"Jax Catalog No\"],\n",
    "#        notes=str_or_empty(row[\"Strain notes\"]),\n",
    "#        see_also=row[\"see_also\"],\n",
    "#        source=jax_source,\n",
    "#    )\n",
    "#    record.save()\n",
    "\n",
    "sources = {\n",
    "    \"jackson-labs\": jax_source,\n",
    "    \"unc-csbio\": unc_csbio_source,\n",
    "}\n",
    "\n",
    "# this version is from my embedded csv file, so less adjusting needed\n",
    "for i, row in strains.iterrows():\n",
    "    record = models.MouseStrain(\n",
    "        name=row[\"Sample CODE\"],\n",
    "        display_name=row[\"Strain\"],\n",
    "        igvf_id=strain_igvf_id[row[\"Sample CODE\"]] if pandas.notnull(row[\"Sample CODE\"]) else None,\n",
    "        strain_type=strain_type_lookup[row[\"Note\"]],\n",
    "        jax_catalog_number=str_or_none(row[\"Jax Catalog No\"]),\n",
    "        notes=str_or_empty(row[\"Strain notes\"]),\n",
    "        see_also=str_or_none(row[\"see_also\"]),\n",
    "        source=sources[row[\"Source\"]],\n",
    "    )\n",
    "    record.save()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "fleet-company",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Designation                                     NaN\n",
       "Strain                                        CC074\n",
       "Note                                       CC Cross\n",
       "Jax Catalog No                               018855\n",
       "Sample CODE                                   CC074\n",
       "Strain notes                                    NaN\n",
       "see_also          https://www.jax.org/strain/018855\n",
       "Source                                    unc-csbio\n",
       "Name: 49, dtype: object"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "private-century",
   "metadata": {},
   "source": [
    "This is closest to the Mouse table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "seven-technician",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_submitted_mice():\n",
    "    submitted_mice = {\n",
    "    }\n",
    "\n",
    "    for accession_prefix in submitted_sheets:\n",
    "        for i, row in submitted_sheets[accession_prefix][\"rodent_donor\"].iterrows():\n",
    "            if not pandas.isnull(row[\"accession\"]):\n",
    "                rodent_identifier = row[\"rodent_identifier\"]\n",
    "                if accession_prefix == \"igvftst\":\n",
    "                    see_also_template = \"https://api.sandbox.igvf.org/rodent-donors/{}/\"\n",
    "                elif accession_prefix == \"igvf\":\n",
    "                    see_also_template = \"https://api.data.igvf.org/rodent-donors/{}/\"\n",
    "                if pandas.notnull(row[\"accession\"]):\n",
    "                    submitted_mice.setdefault(rodent_identifier, []).append({\n",
    "                        \"accession_prefix\": accession_prefix,\n",
    "                        \"name\": row[\"accession\"],\n",
    "                        \"uuid\": row[\"uuid\"],\n",
    "                        \"see_also\": see_also_template.format(row[\"accession\"]),\n",
    "                    })\n",
    "\n",
    "    print(len(submitted_mice))\n",
    "    return submitted_mice\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "distinct-failure",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "121\n"
     ]
    }
   ],
   "source": [
    "def estrus_cycle(x):\n",
    "    if pandas.isnull(x):\n",
    "        return \"NA\"\n",
    "    else:\n",
    "        return x\n",
    "\n",
    "    \n",
    "def load_mouse_tables():\n",
    "    submitted_mice = load_submitted_mice()\n",
    "    \n",
    "    sex_lookup = {\n",
    "        numpy.nan: models.SexEnum.UNKNOWN,\n",
    "        \"Female\": models.SexEnum.FEMALE,\n",
    "        \"Male\": models.SexEnum.MALE,\n",
    "    }\n",
    "\n",
    "    estrus_stage = {\n",
    "        \"Unknown\": models.EstrusCycle.UNKNOWN,\n",
    "        \"Anestrus\": models.EstrusCycle.ANESTRUS,\n",
    "        \"Anestrus>Proestrus\": models.EstrusCycle.ANESTRUS_PROESTRUS,\n",
    "        \"Proestrus\": models.EstrusCycle.PROESTRUS,\n",
    "        \"Proestrus>Estrus\": models.EstrusCycle.PROESTRUS_ESTRUS,\n",
    "        \"Estrus\": models.EstrusCycle.ESTRUS,\n",
    "        \"Estrus>Metestrus\": models.EstrusCycle.ESTRUS_METESTRUS,\n",
    "        \"Metestrus\": models.EstrusCycle.METESTRUS,\n",
    "        \"Metestrus>Diestrus\": models.EstrusCycle.METESTRUS_DIESTRUS,\n",
    "        \"Diestrus\": models.EstrusCycle.DIESTRUS,\n",
    "        \"Diestrus>Proestrus\": models.EstrusCycle.DIESTRUS_PROESTRUS,\n",
    "    }\n",
    "\n",
    "    mice = pandas.read_excel(\n",
    "        lizs_sheet_name, \n",
    "        sheet_name=\"mice\", \n",
    "        usecols=range(0,16),\n",
    "        converters={\n",
    "            \"Estres cycle stage\": estrus_cycle,\n",
    "            \"Dissection ID\": int,\n",
    "            \"RMS number\": int_or_none,\n",
    "            \"Housing number\": int_or_none,\n",
    "        }\n",
    "    )\n",
    "\n",
    "    if models.Mouse.objects.count() > 0:\n",
    "        truncate(models.Mouse)\n",
    "\n",
    "    los_angeles_tz = zoneinfo.ZoneInfo(\"America/Los_Angeles\")\n",
    "\n",
    "\n",
    "    failed = False\n",
    "    for i, row in mice.iterrows():\n",
    "        if not (pandas.isnull(row[\"Mouse Name\"]) or pandas.isnull(row[\"Dissection ID\"])):\n",
    "            name = normalize_strain(row[\"Mouse Name\"])\n",
    "            strain_code = strain_name_to_code[row[\"Strain\"]]\n",
    "\n",
    "            if name == \"056_WSBJ_10F->PWKJ_9F\":\n",
    "                name = \"056_PWKJ_9F\"\n",
    "\n",
    "            try:\n",
    "                strain = models.MouseStrain.objects.get(name=strain_code)\n",
    "            except models.MouseStrain.DoesNotExist as e:\n",
    "                print(f\"Unable to find {row['Strain']} for row {i+2} with code {strain_code}\")\n",
    "                failed = True\n",
    "                continue\n",
    "\n",
    "            if not name[-1] in (\"M\", \"F\"):\n",
    "                raise ValueError(f\"Unrecognized sex field {mouse_name}\")\n",
    "\n",
    "            name_fields = name.split(\"_\")\n",
    "\n",
    "            # deal with TERM2R47HNSS_HO\n",
    "            if len(name_fields) > 3:\n",
    "                name_fields = [name_fields[0], \"_\".join(name_fields[1:-1]), name_fields[-1]]\n",
    "            #if len(name_fields) != 3:\n",
    "            #    raise ValueError(f\"Wrong number of attributes in {name} line {i+1}\")\n",
    "\n",
    "            if not name_fields[1] in strain_igvf_id:\n",
    "                raise ValueError(f\"strain background in {name} not recognized line {i+1}\")\n",
    "\n",
    "            housing_number = None if pandas.isnull(row[\"Housing number\"]) else row[\"Housing number\"]\n",
    "\n",
    "            record = models.Mouse(\n",
    "                # should i use liz's disection id?\n",
    "                name=name,\n",
    "                strain=strain,\n",
    "                sex=sex_lookup[row[\"Sex\"]],\n",
    "                weight_g=row[\"Weight (g)\"],\n",
    "                date_of_birth=row[\"DOB\"].date() if not pandas.isnull(row[\"DOB\"]) else None,\n",
    "                harvest_date=row[\"Harvest Date\"].date() if not pandas.isnull(row[\"Harvest Date\"]) else None,\n",
    "                timepoint_description=row[\"Timepoint\"],\n",
    "                life_stage=models.LifeStageEnum.ADULT,                \n",
    "                operator=str_or_empty(row[\"Operator\"]),\n",
    "                notes=str_or_empty(row[\"Comments\"]),\n",
    "                housing_number=housing_number,\n",
    "            )\n",
    "\n",
    "            if not pandas.isnull(row[\"Estres cycle stage\"]):\n",
    "                # Ignore those transitions.\n",
    "                stage = row[\"Estres cycle stage\"]\n",
    "                if stage.endswith(\"?\"):\n",
    "                    stage = stage[:-1]\n",
    "                record.estrus_cycle=estrus_stage[stage]\n",
    "\n",
    "            record.save()\n",
    "\n",
    "            if name in submitted_mice:\n",
    "                accessions = []\n",
    "                for accession_record in submitted_mice[name]:\n",
    "                    accession = models.Accession(\n",
    "                        accession_prefix=accession_record[\"accession_prefix\"],\n",
    "                        name=accession_record[\"name\"],\n",
    "                        uuid=accession_record[\"uuid\"],\n",
    "                        see_also=accession_record[\"see_also\"],\n",
    "                    )\n",
    "                    accession.save()\n",
    "                    accessions.append(accession)\n",
    "                record.accession.set(accessions)\n",
    "                record.save()\n",
    "\n",
    "    assert not failed, \"Check warning messages\"\n",
    "    \n",
    "load_mouse_tables()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "perceived-jerusalem",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:No traceback has been produced, nothing to debug.\n"
     ]
    }
   ],
   "source": [
    "%debug"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "hybrid-destruction",
   "metadata": {},
   "source": [
    "# Tissues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "electoral-ability",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Samples - 8 founders (1621, 16)\n",
      "Samples - Bridge (416, 23)\n",
      "Samples - F1s (1920, 18)\n",
      "Samples - CClines (5720, 18)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def load_tissue_sheets():\n",
    "    tissue_sheets = [\n",
    "        \"Samples - 8 founders\", \n",
    "        \"Samples - Bridge\", \n",
    "        \"Samples - F1s\",\n",
    "        \"Samples - CClines\",\n",
    "    ]\n",
    "    tissues = []\n",
    "    for tissue_sheet in tissue_sheets:\n",
    "        tissue = pandas.read_excel(\n",
    "            lizs_sheet_name, \n",
    "            sheet_name=tissue_sheet,\n",
    "            header=0,\n",
    "        ).dropna(axis=0, how=\"all\")\n",
    "        if \"tube+tissue wight (g)\" in tissue.columns:\n",
    "            print(f\"{tissue_sheet} has typoed wight column, renaming to weight\")\n",
    "            tissue = tissue.rename(columns={\n",
    "                \"tube+tissue wight (g)\": \"tube+tissue weight (g)\"\n",
    "            })\n",
    "        print(tissue_sheet, tissue.shape)\n",
    "        tissues.append(tissue)\n",
    "\n",
    "    combined = pandas.concat(tissues)\n",
    "    assert \"tube+tissue wight (g)\" not in combined.columns\n",
    "    return combined\n",
    "\n",
    "_ = load_tissue_sheets()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "threatened-procedure",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_timepoint(value):\n",
    "    value, units = value.split(\" \")\n",
    "    value = float(value)\n",
    "    \n",
    "    unit_map = {\n",
    "        \"day\": models.AgeUnitsEnum.DAY,\n",
    "        \"days\": models.AgeUnitsEnum.DAY,\n",
    "        \"week\": models.AgeUnitsEnum.WEEK,\n",
    "        \"weeks\": models.AgeUnitsEnum.WEEK,\n",
    "        \"month\": models.AgeUnitsEnum.MONTH,\n",
    "        \"months\": models.AgeUnitsEnum.MONTH,\n",
    "    }\n",
    "    return (value, unit_map[units])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "specialized-albert",
   "metadata": {},
   "outputs": [],
   "source": [
    "ontology_map = {}\n",
    "for record in models.OntologyTerm.objects.all():\n",
    "    ontology_map[record.curie] = record"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "alleged-praise",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "692\n"
     ]
    }
   ],
   "source": [
    "def load_submitted_tissues():\n",
    "    submitted_tissues = {}\n",
    "\n",
    "    for accession_prefix in submitted_sheets:\n",
    "        for i, row in submitted_sheets[accession_prefix][\"tissue\"].iterrows():\n",
    "            if not pandas.isnull(row[\"accession\"]):\n",
    "                if accession_prefix == \"igvftst\":\n",
    "                    see_also_template = \"https://api.sandbox.igvf.org/tissues/{}/\"\n",
    "                elif accession_prefix == \"igvf\":\n",
    "                    see_also_template = \"https://api.data.igvf.org/tissues/{}/\"\n",
    "\n",
    "                aliases = row[\"aliases:array\"].split(\",\")\n",
    "                assert len(aliases) == 1\n",
    "                alias_fields = aliases[0][len(\"ali-mortazavi:\"):].split(\"_\")\n",
    "                tissue_id = \"_\".join(alias_fields[0:4])\n",
    "                if pandas.notnull(row[\"accession\"]):\n",
    "                    submitted_tissues.setdefault(tissue_id, []).append({\n",
    "                        \"accession_prefix\": accession_prefix,\n",
    "                        \"name\": row[\"accession\"],\n",
    "                        \"uuid\": row[\"uuid\"],\n",
    "                        \"see_also\": see_also_template.format(row[\"accession\"]),\n",
    "                    })\n",
    "\n",
    "    print(len(submitted_tissues))\n",
    "    return submitted_tissues\n",
    "\n",
    "submitted_tissues = load_submitted_tissues()\n",
    "#len(submitted_tissues)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "sweet-imaging",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "016_B6J_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM21751020', 'uuid': '5093a872-bf44-42c5-abfd-a99dd3d6df2c', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM21751020/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM20935867', 'uuid': '8ac461bf-8d25-4f0e-bd17-f5043cb96ce5', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM20935867/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2747ZOJJ', 'uuid': '12089b43-dc3c-40cd-b7b7-de5934949a94', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2747ZOJJ/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9909OKAR', 'uuid': '3ecee650-301b-481f-bc5e-1a46dafd4542', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9909OKAR/'}]\n",
      "017_B6J_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM58133825', 'uuid': '56f28252-6e1a-4367-bd0d-3bdd75abb6f2', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM58133825/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM51811428', 'uuid': 'ce1bfdf4-cacb-4859-a79f-5da312178a27', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM51811428/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM5751LMFR', 'uuid': '489b79b4-c432-4f92-b1ae-9e3b765b8dbf', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5751LMFR/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9616JAGA', 'uuid': '74a4f3c4-5173-4a0f-b684-1cf988a49c7a', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9616JAGA/'}]\n",
      "018_B6J_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM10325421', 'uuid': 'db754d0d-e36b-49b6-b0fa-53eaf25fd744', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM10325421/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM33316466', 'uuid': 'e9114d98-ffe6-447a-9f6f-9a6c4bb8ddd6', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM33316466/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM4584AIMA', 'uuid': '13650fd4-0573-4e27-92d2-c42c7b5eaaf6', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4584AIMA/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6999WOUL', 'uuid': 'a71cb08f-2788-461d-9c5c-079ed8b1e4a7', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6999WOUL/'}]\n",
      "019_B6J_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM26473010', 'uuid': '8863dd4f-198b-4d15-9067-d0561c39734b', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM26473010/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM18283790', 'uuid': '83cee65e-441a-4703-8a05-68a89b7499ee', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM18283790/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM3502KEWR', 'uuid': 'e48b8327-9d8d-41d9-8735-b924df3bfaf1', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3502KEWR/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9565QVTN', 'uuid': '552b6e0a-454f-4900-87b0-e7405e6a494b', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9565QVTN/'}]\n",
      "020_B6J_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM01160367', 'uuid': '3e69fe23-5e69-4b91-973e-2728f79d718e', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM01160367/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM33524527', 'uuid': '99b62f36-444b-4649-9e4e-26a663a7c252', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM33524527/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6578SAUC', 'uuid': 'c170a792-994e-4c37-bc3b-ef49897c5de4', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6578SAUC/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9764WOWB', 'uuid': 'd861a51f-57ff-4222-820a-b46a4f9907d4', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9764WOWB/'}]\n",
      "021_B6J_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM42088479', 'uuid': '9ab88858-088f-4095-8651-5cc42419a0dd', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM42088479/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM99278471', 'uuid': '924e79ca-d960-42ab-acde-1a9d85fdbb78', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM99278471/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2795QIAK', 'uuid': '164c4760-ba0e-4424-9ab6-0f37bda42607', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2795QIAK/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM3360ORUF', 'uuid': 'cf407c40-57a8-4ee4-b7a3-15be85574c60', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3360ORUF/'}]\n",
      "024_B6J_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM41442798', 'uuid': 'c06d0f9d-6fee-4b01-8401-14c26f84c1af', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM41442798/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM92339352', 'uuid': '71a9c6e3-4013-456f-b1cd-f547ef8caf95', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM92339352/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM7596ZIPA', 'uuid': '8654187a-0b2a-4326-b464-09fb102bddd2', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7596ZIPA/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0587AFGX', 'uuid': 'c4e66921-3298-4ffc-9a43-f6ce8698ae7e', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0587AFGX/'}]\n",
      "025_B6J_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM53700181', 'uuid': 'b779a774-03fd-4495-9e11-43d1b886816e', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM53700181/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM24966581', 'uuid': '4191c958-5276-480b-8242-b11faae829d6', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM24966581/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM4108RLII', 'uuid': 'f814d7f1-3bfc-448a-853a-b4b0e9ecd877', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4108RLII/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1457MNSR', 'uuid': 'f1f113d7-662a-43d9-aef2-e9b508aaeef0', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1457MNSR/'}]\n",
      "066_NODJ_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM00342720', 'uuid': '5a77d96b-0dbf-4f69-882c-405bc8c5fa46', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM00342720/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM62163894', 'uuid': '2fb0690b-46dd-4370-bc3c-0f0591bc3316', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM62163894/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1810KYOB', 'uuid': 'c221bf2e-917b-46b2-88e1-a2cb67ba636d', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1810KYOB/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM5439QLWY', 'uuid': '14ed6824-2dd2-42b5-9ea5-68eed279b527', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5439QLWY/'}]\n",
      "067_NODJ_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM05994267', 'uuid': 'dac6d232-6f33-48d0-81d7-dea1ee5b6204', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM05994267/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM96811554', 'uuid': '1629913b-4e9c-4032-a115-98757739d445', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM96811554/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6514BRQT', 'uuid': '2821b73b-5d02-4ef6-ac0d-badd69b82778', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6514BRQT/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM3477RPDP', 'uuid': 'aa241577-33cb-48c6-bac9-221cda21cbde', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3477RPDP/'}]\n",
      "068_NODJ_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM72460407', 'uuid': '6012f47d-d05c-41f1-b154-3130845e9979', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM72460407/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM52294448', 'uuid': '940cd826-45e8-495a-a1a2-1fec4f617810', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM52294448/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0483CXTW', 'uuid': 'ef6a8399-83b9-4160-9fbb-47e1a18b0a43', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0483CXTW/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2315DNPR', 'uuid': '6cd5eed6-35dc-46e7-9dba-5f566f014066', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2315DNPR/'}]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "069_NODJ_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM70776988', 'uuid': '64c37487-bc8c-4fbf-a70d-4815014ee4e3', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM70776988/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM04240735', 'uuid': 'b54afd50-51d6-4b6f-b4ae-41e5fb04a26c', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM04240735/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2210CUOC', 'uuid': '3237d42d-9f30-412c-be67-97f43cd605bb', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2210CUOC/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM3655BTER', 'uuid': '2b03b94c-c611-44c1-98c6-7aa18f1641c3', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3655BTER/'}]\n",
      "070_NODJ_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM16736591', 'uuid': '7f713377-43fd-4ea7-901c-35c3e50e8a77', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM16736591/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM83682213', 'uuid': '7f390718-5684-4145-b3b5-85284e482727', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM83682213/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM4132TDOU', 'uuid': '8293abba-add2-4a82-affb-d0af66f38a2e', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4132TDOU/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1916FXAR', 'uuid': '269c39b2-e726-4fe5-851b-5b175979b936', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1916FXAR/'}]\n",
      "071_NODJ_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM76160667', 'uuid': '8736e722-d0be-49cc-acb9-db2e38629d4c', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM76160667/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM28533810', 'uuid': '6e39b3b2-a5ea-47b2-81af-4dc0fcec779a', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM28533810/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9423CHGE', 'uuid': '2a2997b4-56ee-4217-ae0d-a8268af47355', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9423CHGE/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2219SJUB', 'uuid': '6eaec560-1fcf-4769-a9e9-9c060f2af1a6', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2219SJUB/'}]\n",
      "072_NODJ_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM79168608', 'uuid': 'd4df962c-c9b9-47a7-adf5-6f622e84f7bb', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM79168608/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM85453964', 'uuid': '0c60d4a3-dbbb-4750-853e-fabe199470b8', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM85453964/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0749UUSU', 'uuid': 'ad59bbd5-96bf-4aa0-873a-543c3ce74588', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0749UUSU/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1724CQMW', 'uuid': '23e32920-8fa8-4b7d-af9b-6b5676edc4b8', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1724CQMW/'}]\n",
      "073_NODJ_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM45296598', 'uuid': '35b4d0f5-213c-4c5d-8c9c-40d66423b038', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM45296598/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM90705852', 'uuid': '00f594bc-3ed3-4072-a22c-2ed304759f85', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM90705852/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9890CMQT', 'uuid': '11041153-5c6f-48c0-976a-f4926a829f3b', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9890CMQT/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM5506IVPG', 'uuid': 'b285ef2b-32dd-4c44-864f-8b9537d0c231', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5506IVPG/'}]\n",
      "026_AJ_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM12615846', 'uuid': '7e5de2b8-9185-4700-a665-548914c62015', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM12615846/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM92907388', 'uuid': 'a58fe6cd-8956-4170-8e2d-350485ac1e7b', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM92907388/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM4370URZQ', 'uuid': '7e2f83bb-686d-4cdf-ad60-06a415149844', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4370URZQ/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2423BNZZ', 'uuid': '3a8b2024-b8da-46db-954d-4551c348f9c3', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2423BNZZ/'}]\n",
      "029_AJ_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM38994111', 'uuid': '3b680c8b-146b-4310-8f08-94bd2dee5119', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM38994111/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM47454871', 'uuid': 'f857066f-cf89-45a9-a653-7795e030373c', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM47454871/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM7221OIET', 'uuid': 'a56ecec3-101a-42cc-8007-4ba96b3b2e6d', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7221OIET/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2957XPNR', 'uuid': 'fb7f4566-4a49-4a1e-9ff7-eac9d16e5f97', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2957XPNR/'}]\n",
      "028_AJ_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM30161161', 'uuid': 'a8e03d55-deaf-413f-b933-c93e793d0732', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM30161161/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM51463318', 'uuid': '76ebeea1-04a3-40f0-a7a6-9c821f2d8959', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM51463318/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM7981HLTV', 'uuid': 'caff5d29-ca29-4fe8-a5db-35bff6384de2', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7981HLTV/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8529NIEC', 'uuid': '74efe2ba-adde-4ceb-ab49-0d88526b8648', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8529NIEC/'}]\n",
      "031_AJ_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM02881437', 'uuid': 'ffea0493-9cbc-4ac0-a6c4-3a2fe4a486b4', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM02881437/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM84736502', 'uuid': '3f29043c-477a-4e5e-a777-7cea6f1ea34a', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM84736502/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6144DRYM', 'uuid': 'c75fc1e6-69fd-4b8f-9081-ca906ede0630', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6144DRYM/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8120ZQGC', 'uuid': '15ff0666-156d-47a7-8b25-9e64067a5d2e', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8120ZQGC/'}]\n",
      "030_AJ_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM83815196', 'uuid': '9bcf7b19-cc01-4089-b0f1-e5bd7927f864', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM83815196/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM03250341', 'uuid': '0ede3985-50d8-4b16-89e8-ccf024fc756d', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM03250341/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9442VMKA', 'uuid': '355d36a6-82db-4e97-96b2-64f56363f4c9', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9442VMKA/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0425PEQP', 'uuid': 'ac533e1f-824f-4852-81a3-38c47b9b9cfc', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0425PEQP/'}]\n",
      "033_AJ_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM66295363', 'uuid': '9074c218-f526-44f4-86e7-e2a5cf2ee7eb', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM66295363/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM89879775', 'uuid': '91defe4a-7bb6-49af-8dc8-32043ab5ce1c', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM89879775/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9867TADD', 'uuid': '29ae3b1b-7632-4146-a2af-5dfdf9eeca62', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9867TADD/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0406DUWH', 'uuid': 'ff71c9ee-8f76-42e6-9768-9bfa1f33ff04', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0406DUWH/'}]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "032_AJ_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM80577942', 'uuid': '6d8e091d-a63e-4464-b944-2785039b98a6', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM80577942/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM57113277', 'uuid': '26bdccbd-1bba-4691-aab9-8e6228442673', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM57113277/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM3195CJCF', 'uuid': 'cf59909d-c385-4e5c-a1c7-10dcdc2661f2', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3195CJCF/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6726YGIG', 'uuid': '18411a3d-e5dc-40bd-afe5-fcc9d0009ea4', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6726YGIG/'}]\n",
      "035_AJ_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM59466810', 'uuid': 'de3f2622-b21c-4521-9496-04c96805f19f', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM59466810/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM93284518', 'uuid': '2eda64f7-e9e3-42fe-a6a7-c68fa659e718', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM93284518/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM4131CTSJ', 'uuid': '9ba11cdd-82ff-4407-aab5-f168da714c97', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4131CTSJ/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM4651QBMQ', 'uuid': 'fe1f3ec4-8eab-442c-816d-0efd1897ecce', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4651QBMQ/'}]\n",
      "076_PWKJ_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM52954034', 'uuid': '88e8d180-0338-4f12-b49a-6a7ad3e0e549', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM52954034/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM15705729', 'uuid': 'e7356ef8-de67-4d6e-b725-58038bd875b4', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM15705729/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0918MYKM', 'uuid': 'd325cdfe-00a6-4f23-aa04-774de52ada7d', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0918MYKM/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM7550NDQP', 'uuid': '480e10d4-1d83-4bd6-ae69-5ec584ea6ded', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7550NDQP/'}]\n",
      "077_PWKJ_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM83815676', 'uuid': 'aac6da89-ae7f-4eb5-9f65-0da6d1b2bc66', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM83815676/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM04954230', 'uuid': '0320b949-703d-40bf-a9f2-bf15e61ca548', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM04954230/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6013VRIK', 'uuid': '270b496d-e5d7-4b81-893e-bdd40fa7971a', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6013VRIK/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0563DEOO', 'uuid': '7fc28775-ddf6-4c14-9793-312b4d48932f', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0563DEOO/'}]\n",
      "078_PWKJ_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM03938139', 'uuid': '007e2129-8560-49e8-b925-07f2697c961a', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM03938139/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM20539280', 'uuid': '7c485858-6b2a-4a8b-86de-da2885b89f33', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM20539280/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6356IYPX', 'uuid': '3c9c357e-4e8d-4e28-b027-9b31a21fde39', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6356IYPX/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0623LMKH', 'uuid': '6333e18a-0755-4f18-a887-070ae5f5270e', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0623LMKH/'}]\n",
      "079_PWKJ_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM22223422', 'uuid': '8d7f1305-4634-4489-aef7-4f79cea13d1a', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM22223422/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM12427616', 'uuid': '08e893c3-e408-4f54-a1a7-9c76bafa848f', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM12427616/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM4224SVKL', 'uuid': '5e631ee5-4547-4d35-b02f-985c79180b90', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4224SVKL/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2048SMTE', 'uuid': '9d37ed77-929e-4b8d-9341-ddd293311ca0', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2048SMTE/'}]\n",
      "080_PWKJ_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM65126292', 'uuid': '136be24f-8e79-4884-9edc-18f66ef76790', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM65126292/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM82793177', 'uuid': '8e27cc65-2496-4be7-9b0e-d47f51139dee', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM82793177/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1463UXNU', 'uuid': '134ff3a6-3b32-4313-9874-82dd71a65433', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1463UXNU/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM3905UDST', 'uuid': '99197a01-b977-4482-8338-303fcd21b12d', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3905UDST/'}]\n",
      "081_PWKJ_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM69336152', 'uuid': '46b85ce0-63ae-4c76-92e5-0e3aced3ac31', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM69336152/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM16556661', 'uuid': '3e1a251c-10a1-47bd-b9dd-21ccb493ac9c', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM16556661/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1764RZHQ', 'uuid': '4f1ac155-9874-402b-b63b-d14fd8d775ee', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1764RZHQ/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0454IWHA', 'uuid': '4f486bac-7118-422f-bdd7-026191f62168', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0454IWHA/'}]\n",
      "084_PWKJ_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM60112939', 'uuid': 'b031b861-4912-4013-ab3f-6f2ae8b5893e', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM60112939/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM20313266', 'uuid': 'aa5409f3-a67d-4784-8388-6943a519043d', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM20313266/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2111IGAF', 'uuid': '21318f88-293b-47b4-b042-d1518651b155', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2111IGAF/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM7111UROK', 'uuid': 'd36d756f-c0a5-40aa-8927-c68ade6feb0c', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7111UROK/'}]\n",
      "085_PWKJ_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM07209305', 'uuid': '1b973b12-10e5-4834-aaf0-f68c55199e4e', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM07209305/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM51740762', 'uuid': '8f4529af-8e03-4309-92f5-c073a259883a', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM51740762/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM5036ECAW', 'uuid': '61d1e625-09fe-427e-aa53-9036eb57f8af', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5036ECAW/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1227KXDL', 'uuid': 'bef60b67-6f91-42a4-a157-2f12e8cfa467', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1227KXDL/'}]\n",
      "036_129S1J_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM53940869', 'uuid': '475429e8-7e03-4377-973d-e89e2e22b86b', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM53940869/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM96114334', 'uuid': '994c3ec7-bd62-47ac-b44b-c9ec2a48e492', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM96114334/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9227EIZD', 'uuid': '5b35e6a7-6648-4c4c-b78f-3125c84efe97', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9227EIZD/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8124ZHLJ', 'uuid': 'd1d75f8f-3d4a-43d0-8654-add43f26d44f', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8124ZHLJ/'}]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "037_129S1J_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM79359987', 'uuid': '24278b79-3600-4fd8-8ef3-55f78bbe77be', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM79359987/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM14025651', 'uuid': '2657fe8f-9409-4cdd-b8d3-522bd1135eac', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM14025651/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0860OBUU', 'uuid': '9ce6f4ec-0e70-4cce-ab32-b2fe0bc9e97b', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0860OBUU/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2419NKNV', 'uuid': '44fe44ad-3ce9-466c-97c5-8b8936d14106', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2419NKNV/'}]\n",
      "038_129S1J_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM46396178', 'uuid': '8ef01892-7f1b-4ea2-9b50-5d827e1f1d72', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM46396178/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM52735813', 'uuid': '21859d64-6f62-4cfc-b51e-fb4280809d4a', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM52735813/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0225XPQG', 'uuid': 'b7867990-c177-4b2c-8a9f-3a405f1f442f', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0225XPQG/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1783ISPH', 'uuid': 'ee5197d5-82f3-4875-ad1a-d73a53ee3e21', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1783ISPH/'}]\n",
      "041_129S1J_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM92914084', 'uuid': 'ede30dbb-c084-44ab-9d54-3ad6e25349c5', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM92914084/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM90714157', 'uuid': '90f34008-76b4-4f73-bb48-391c510e2f59', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM90714157/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9603NVUP', 'uuid': 'e64b3d46-b5a9-44ab-9697-b1fc4556f9a2', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9603NVUP/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM4535GZXJ', 'uuid': 'c7faf615-7151-4f54-b063-8bdf71adb031', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4535GZXJ/'}]\n",
      "040_129S1J_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM77387655', 'uuid': '349daa31-7d4d-4a56-89e0-86085d48998e', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM77387655/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM74867542', 'uuid': '888fdbfc-bd1c-4439-b43b-f4d714acb690', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM74867542/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1865RBAP', 'uuid': '26661547-9c86-4f4e-8054-4ea59d555095', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1865RBAP/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM3592ZCYS', 'uuid': 'e54e61d7-f4ea-48ac-b5a2-cd0074470279', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3592ZCYS/'}]\n",
      "043_129S1J_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM16786304', 'uuid': 'f42e340d-b2a6-4083-9c7c-f36f460e71c6', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM16786304/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM54542240', 'uuid': '827a41c6-c206-4325-b751-1f78666c2355', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM54542240/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2592YCXF', 'uuid': '84d33c54-3f48-4f22-9f11-574d30d2a987', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2592YCXF/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM4046THJJ', 'uuid': '162cb92b-9e88-4b16-98c5-af314807c787', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4046THJJ/'}]\n",
      "044_129S1J_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM33999140', 'uuid': 'a53d9f0d-ad97-464e-a137-1b8364eddaf3', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM33999140/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM35170934', 'uuid': 'd6c65361-4d12-40e0-b535-4ebcc66000e8', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM35170934/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8263PWJC', 'uuid': '6ab6613c-7057-4ddc-9d18-99940dbabe51', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8263PWJC/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0923PDGB', 'uuid': '63a0fa92-389a-4b5d-b4c5-cc22fbf02c84', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0923PDGB/'}]\n",
      "045_129S1J_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM44620762', 'uuid': '7f863cef-6c3d-4285-9de1-d35f4ba1bf04', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM44620762/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM93636128', 'uuid': '17226952-dfa8-400b-a4a2-b3e1daa18930', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM93636128/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9934DIHO', 'uuid': 'f1b33610-3df0-49f8-8990-3d90c1382fb1', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9934DIHO/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0221ZKMO', 'uuid': '2ba4661d-8e4d-451b-b69b-7bfced3c9e4e', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0221ZKMO/'}]\n",
      "086_CASTJ_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM23724063', 'uuid': '463fd79b-5e4f-4012-bac2-b7ba1a3394f9', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM23724063/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM36392432', 'uuid': 'cfc0cb17-14f6-42e7-b98d-8b3dc24906bb', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM36392432/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM7195PKWY', 'uuid': 'a9637f98-9c95-4a35-9b95-c62680707a51', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7195PKWY/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM7666QOZY', 'uuid': '1e6db460-5b80-4fc5-9052-29e86eb9685c', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7666QOZY/'}]\n",
      "087_CASTJ_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM32674535', 'uuid': '2e6aeee2-cb21-4df9-9bba-6afceea05c14', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM32674535/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM01448874', 'uuid': 'a3c0df52-96f5-409d-8faf-f6eda174909d', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM01448874/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM7317YYMK', 'uuid': 'a8ad5cd0-22e9-4bfb-b6d2-cee8fce9c2f5', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7317YYMK/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8253LIZU', 'uuid': '35c7329b-3ca6-4a7f-a9a2-befb336a3905', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8253LIZU/'}]\n",
      "088_CASTJ_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM79946851', 'uuid': '3febf347-cd39-4718-8c5d-bae336c84849', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM79946851/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM57968715', 'uuid': '45d6a976-c79d-4462-812e-573781dda9fa', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM57968715/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8386KIXU', 'uuid': '09d4d693-82d7-4cec-8a5e-55469af01157', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8386KIXU/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM3708BQMH', 'uuid': '437f21ae-ba2a-44fb-82ef-9693b4a7058c', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3708BQMH/'}]\n",
      "089_CASTJ_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM49388204', 'uuid': 'fe4a6e87-bb8c-4611-b089-23af834ff814', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM49388204/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM57395435', 'uuid': 'd9f5df98-f0bd-4c06-b486-c4d3ea8ab57b', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM57395435/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6275QJJU', 'uuid': 'd9e9bee7-7412-4a35-bf17-00054975f090', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6275QJJU/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM7470VZNE', 'uuid': '8a298c03-1cdb-4e38-aa3d-00afb7dcc00d', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7470VZNE/'}]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "090_CASTJ_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM38489848', 'uuid': '5ec2ee20-ad68-4a12-ada9-ba63bd84b113', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM38489848/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM57565332', 'uuid': '03d61e1c-9c02-4669-9129-2c0c9ad0ad21', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM57565332/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9194WULA', 'uuid': 'e97872f8-da47-4980-8a75-0426a69ea189', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9194WULA/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM3391TANO', 'uuid': '4a773267-7169-43a1-9a01-67fdd84f0351', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3391TANO/'}]\n",
      "091_CASTJ_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM62101456', 'uuid': 'd7a9a951-4843-4ff0-b909-26d3098c87f7', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM62101456/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM66782725', 'uuid': 'a9aa2edc-a47e-4dce-9b4e-daa393396a2d', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM66782725/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM3523HTHO', 'uuid': '47365040-d1c5-4892-8306-66647229345c', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3523HTHO/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM4618XBBU', 'uuid': 'c195863e-d49c-4347-8e62-91e9640ba62b', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4618XBBU/'}]\n",
      "094_CASTJ_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM90443496', 'uuid': '6532ab58-9141-4dd0-a1a1-8c4aff5a036e', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM90443496/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM57771690', 'uuid': '764c1f5a-1e05-433d-9a66-2650db1344b5', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM57771690/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM7747QNMD', 'uuid': '277353ca-1540-4289-9140-695f62a03bf4', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7747QNMD/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8043HCTC', 'uuid': '87ac318f-a8e4-43ea-8a47-bf0c88efbf86', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8043HCTC/'}]\n",
      "093_CASTJ_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM79094150', 'uuid': '6c4d166d-8657-45fa-90ab-fc11165968dd', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM79094150/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM84121976', 'uuid': '6c941741-096c-4dee-8e98-b8f10a9495e0', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM84121976/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8601VZGE', 'uuid': '5bf3801d-a019-4678-a9f6-f62356fb7054', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8601VZGE/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0611PVMT', 'uuid': '5cee9cef-0e3c-482d-a066-2d0e09d71972', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0611PVMT/'}]\n",
      "058_WSBJ_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM32200435', 'uuid': '5f9db1c1-0496-43c4-a7c7-e8bbd688b306', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM32200435/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM15885234', 'uuid': '891e5899-4b30-4e71-97b4-78adb49f216c', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM15885234/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2105LIWI', 'uuid': '9862f105-760c-47ea-8ccd-8f6a39861e6c', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2105LIWI/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1395QLCT', 'uuid': '96ef3f40-91b9-4148-a466-20c5e952f03b', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1395QLCT/'}]\n",
      "057_WSBJ_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM43608351', 'uuid': 'b8b00f4b-d2e7-45e9-b468-0356c2407109', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM43608351/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM11842313', 'uuid': '84725cb2-a0e8-4c20-8170-a04749b56c9a', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM11842313/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6575IFHC', 'uuid': '2eb3f073-7c09-4f5f-b23c-93e17861d5f5', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6575IFHC/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1203GHYB', 'uuid': 'd9277809-9203-4102-aea1-503ea29d675a', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1203GHYB/'}]\n",
      "060_WSBJ_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM63825870', 'uuid': 'a3680081-8f54-46ef-93f9-0ecf53648ede', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM63825870/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM41460346', 'uuid': 'c06ebf0e-5ed2-4702-ae7c-990101c9cd9e', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM41460346/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1996USDH', 'uuid': '1d4ce3c7-6c99-4af8-949a-d6e64c99a2ab', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1996USDH/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9916UWTV', 'uuid': '20342be2-d3a6-4b39-9cd2-cf24dd47ed52', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9916UWTV/'}]\n",
      "059_WSBJ_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM19613428', 'uuid': '769d284b-fda3-487b-a42e-bdcbdf5dafda', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM19613428/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM97826128', 'uuid': 'fb0f8bf0-fa87-4841-a460-2d098f594018', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM97826128/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1116FVJP', 'uuid': '3e5ad4a3-73a4-459b-9812-0a46b1dd4e56', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1116FVJP/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1295CUJX', 'uuid': '9389066e-f233-4503-8970-c483ab60f03d', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1295CUJX/'}]\n",
      "062_WSBJ_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM60971839', 'uuid': '596e9af4-9cd9-400d-ac18-a24a3495481b', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM60971839/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM19608989', 'uuid': 'b8f87ea3-7fb2-46d1-9797-ad31db6a6a7c', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM19608989/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2903TFAJ', 'uuid': 'c640cd48-1289-4197-aa47-9e029bd9bec3', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2903TFAJ/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM5785PIAM', 'uuid': 'e1122092-0ca4-4b95-ae27-ba161a5bbb3d', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5785PIAM/'}]\n",
      "061_WSBJ_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM45945257', 'uuid': '56bf203f-9664-4f91-a207-5fb79c05fc06', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM45945257/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM00084391', 'uuid': 'de93915a-493e-4bb1-ab5f-c9253ad89da7', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM00084391/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0898KHQX', 'uuid': 'f2daad20-7004-4546-9a46-c7bfe9fa9259', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0898KHQX/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0431DSEH', 'uuid': '17a3349e-8c25-4030-ae76-3944ae5fe95c', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0431DSEH/'}]\n",
      "096_WSBJ_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM60767863', 'uuid': '2cdfca7a-8472-4245-ba95-408585983f6a', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM60767863/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM64908116', 'uuid': '54b91c6d-2fb3-4a08-a3b4-e8813d254eec', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM64908116/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM7994PCAR', 'uuid': 'aa248864-199b-4f5b-b651-b6888fc80652', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7994PCAR/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1901NWCD', 'uuid': 'a9047a77-3864-4ff7-9432-45479de846b3', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1901NWCD/'}]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "063_WSBJ_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM55897972', 'uuid': '1e20fa9c-ea22-4766-a1df-d8b53af753b6', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM55897972/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM15834985', 'uuid': '07bb5012-908c-416a-95ba-64010ff3cf20', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM15834985/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8727ZEXW', 'uuid': '857c0d4c-64a5-42bb-8d6f-189ce3460106', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8727ZEXW/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1039KCGC', 'uuid': 'bbc4bbb1-c7ef-4c6b-9526-ed8daae6c024', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1039KCGC/'}]\n",
      "046_NZOJ_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM24255023', 'uuid': 'dfff0205-ca9b-47a5-8e77-41fc29ffecf1', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM24255023/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM72017255', 'uuid': '3adad841-b470-457f-986e-c4b150107e70', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM72017255/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1839OFVA', 'uuid': 'c429089a-98e9-4a8c-bdf6-1106fa9ea3df', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1839OFVA/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8201XPIS', 'uuid': '204cd5a3-ccfd-4103-8f93-b2da7d330414', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8201XPIS/'}]\n",
      "047_NZOJ_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM33306992', 'uuid': 'e0b436be-bc8a-4a6a-bd93-8824288cf1e8', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM33306992/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM30685000', 'uuid': '93291861-0f79-4db9-a95a-dd740c8ac3cf', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM30685000/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9072COND', 'uuid': 'e07c3c4b-afb7-488c-af4a-8608fa7775ba', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9072COND/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM7371QPPK', 'uuid': '6346bcc3-f4fd-4c35-b799-40366f19d461', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7371QPPK/'}]\n",
      "048_NZOJ_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM61430175', 'uuid': 'd6185f78-2e47-486e-8f94-bfd321670386', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM61430175/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM44199147', 'uuid': '03729597-907a-4e2f-87cd-767063d30158', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM44199147/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9852CDAM', 'uuid': '9472075d-4236-4d24-ad21-157d223e2924', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9852CDAM/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8490EBUM', 'uuid': '6ea0d036-6a67-4665-8bea-41e55ba98cb2', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8490EBUM/'}]\n",
      "049_NZOJ_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM42091421', 'uuid': 'b86c8f2f-dc4f-47e2-8352-2b7de835a7f1', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM42091421/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM87330877', 'uuid': '422462b9-34b2-4e5f-a269-3895cbafcb1d', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM87330877/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM4528ONTR', 'uuid': '4affe146-58d4-4da6-9e4e-e17dbd25e404', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4528ONTR/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0101ITFE', 'uuid': '39b241d7-c097-44d3-9878-fa241c4caeec', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0101ITFE/'}]\n",
      "050_NZOJ_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM85978816', 'uuid': 'd6101be5-d671-4afe-b71a-2f4f9f5e3d9a', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM85978816/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM80562094', 'uuid': 'a8f27223-3eef-442b-9af3-82db0350ab2e', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM80562094/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM5306QAYC', 'uuid': 'b885aaf8-f080-487c-8c13-7891bdc200ee', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5306QAYC/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0561GQRI', 'uuid': 'df11d12b-7b4c-4733-ab6c-0920a7afb777', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0561GQRI/'}]\n",
      "051_NZOJ_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM86161501', 'uuid': 'caa2889a-9757-4b18-bdbb-a57a9f02b18e', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM86161501/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM11875348', 'uuid': '1d24833e-56eb-40a0-a3e7-93ca8d398a9d', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM11875348/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1083MRIY', 'uuid': 'c1197245-2e20-4320-9328-06d734c8bd2b', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1083MRIY/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM5958RVMJ', 'uuid': 'fe1488d9-96a6-44a3-9013-84d0c07813f9', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5958RVMJ/'}]\n",
      "052_NZOJ_10F_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM13439390', 'uuid': '4e281e53-7fbd-4b87-aaeb-a6d108e43c83', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM13439390/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM67022673', 'uuid': '301fe56b-2bee-4940-89eb-18f578e9c717', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM67022673/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6096DCNO', 'uuid': '25e08fc5-bee5-45c7-b518-9676a9876cff', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6096DCNO/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8959FRNP', 'uuid': 'a7b92ad5-8354-4c6a-8ac7-6d72aa4ca991', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8959FRNP/'}]\n",
      "053_NZOJ_10M_03 [{'accession_prefix': 'igvftst', 'name': 'TSTSM78719590', 'uuid': '0c77a8fa-65cf-47de-94b5-4c993e37bda8', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM78719590/'}, {'accession_prefix': 'igvftst', 'name': 'TSTSM86500952', 'uuid': 'e009dbb6-d7a5-4df1-bad3-9cb56e2119a3', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM86500952/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9234QGBN', 'uuid': '033dea5f-6701-46d3-b469-6d21f6671e61', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9234QGBN/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8856ZHGA', 'uuid': '1b0f45b7-d996-470e-940e-eb7c05b554de', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8856ZHGA/'}]\n",
      "016_B6J_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM95666276', 'uuid': '480ad476-6353-41a2-b178-0540ec4f29a8', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM95666276/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2397OXJE', 'uuid': '24774070-a134-42fa-8ac9-9f4bb6865876', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2397OXJE/'}]\n",
      "066_NODJ_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM73842863', 'uuid': '34657832-2051-4ac4-b86f-037718503bfb', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM73842863/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0400KSXT', 'uuid': '64e5242d-ceee-48dd-9bbb-d921618ea6b2', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0400KSXT/'}]\n",
      "017_B6J_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM15336485', 'uuid': 'c85ee703-cd8c-4595-9999-7916e378be5d', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM15336485/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6821QIOI', 'uuid': 'dc9711ba-0e84-43f1-b835-f1056fe91bec', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6821QIOI/'}]\n",
      "067_NODJ_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM13603592', 'uuid': 'c8b3907d-2c7e-499c-8246-bd37cccbe629', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM13603592/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8486IRSE', 'uuid': 'be507859-5265-4f59-a3f7-8b134dcecda1', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8486IRSE/'}]\n",
      "018_B6J_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM53191815', 'uuid': '8ba491dd-cd43-4768-9ed9-ff40b57eeab7', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM53191815/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9283STYP', 'uuid': '314ead3d-f3b8-4ab6-947f-e5e86a6da896', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9283STYP/'}]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "068_NODJ_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM71072025', 'uuid': '21c29c17-6f6e-47f2-a23e-430fd41d9e30', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM71072025/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6061AIFT', 'uuid': 'c8dedd9c-8424-4846-a347-fbfe73c2531a', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6061AIFT/'}]\n",
      "019_B6J_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM46884051', 'uuid': '4f8c3b79-f670-4d89-adbe-95eb3131854d', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM46884051/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM3819GVRA', 'uuid': '17a90e67-0181-48a4-ad10-b4eacd8b632d', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3819GVRA/'}]\n",
      "069_NODJ_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM04142593', 'uuid': '26ec404b-af28-46af-b84c-0300057c8a06', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM04142593/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9555SHCP', 'uuid': '2cd63182-4ad1-4d86-8217-4e497e3e5d0a', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9555SHCP/'}]\n",
      "020_B6J_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM65099751', 'uuid': '07b47c5e-ded9-4cca-b291-766e2dc1e038', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM65099751/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9188TGLO', 'uuid': 'bd84bcd9-a766-4025-bfe5-3a471cb41046', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9188TGLO/'}]\n",
      "070_NODJ_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM09266471', 'uuid': '35f8a786-3cba-4327-a717-16c0dc64968e', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM09266471/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2548HETT', 'uuid': '4ac3d10e-fcf2-46fb-9403-aae342a65a8d', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2548HETT/'}]\n",
      "021_B6J_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM19996820', 'uuid': '427c9ec9-97af-4b1e-a5bf-b82dccd2754a', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM19996820/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM4152MQAV', 'uuid': 'cc4ae4eb-c2d8-45a3-8b09-0b45e33ce54b', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4152MQAV/'}]\n",
      "071_NODJ_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM25609101', 'uuid': '9afaeb66-fd25-491c-ab1c-71d00e2f188d', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM25609101/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6976MFYC', 'uuid': '0aed675d-d46c-4403-bba4-8600b978caa2', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6976MFYC/'}]\n",
      "024_B6J_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM29667933', 'uuid': 'fcf2d581-281c-4876-a058-83cdea8bed02', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM29667933/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1715AZGK', 'uuid': 'd858a27a-c195-4857-ad73-410237b64fa3', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1715AZGK/'}]\n",
      "072_NODJ_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM63950592', 'uuid': '86e17b30-c60d-4ce9-a6ca-f653597e405d', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM63950592/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1795GGVQ', 'uuid': '905e7b42-4332-4767-ad18-d1c591f20030', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1795GGVQ/'}]\n",
      "025_B6J_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM68343290', 'uuid': 'c8db47b9-3d10-4e79-adf8-7f724a459801', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM68343290/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0659JNLJ', 'uuid': '60d3dab0-39a0-4854-9e9e-8b69af8fe024', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0659JNLJ/'}]\n",
      "073_NODJ_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM90886123', 'uuid': 'a7520b24-3765-4d7a-9862-2ab5eb569cbb', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM90886123/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9138HBOW', 'uuid': '799953a5-80b9-4b06-be04-0c18db5877d2', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9138HBOW/'}]\n",
      "026_AJ_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM95867687', 'uuid': '6d656b9d-2f31-4866-9635-26108815c44f', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM95867687/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM7523CIFZ', 'uuid': 'f382de20-701d-49de-84f9-adb20aef9739', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7523CIFZ/'}]\n",
      "076_PWKJ_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM94943437', 'uuid': '9802d57b-a598-4566-bf54-27f48b7b0bf9', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM94943437/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2489HRKN', 'uuid': 'bc051825-57b4-45f1-9984-70c854b193a1', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2489HRKN/'}]\n",
      "029_AJ_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM01591084', 'uuid': 'd190eee8-fe5c-4141-a3f4-39c8a95fd434', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM01591084/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM5294ZSIL', 'uuid': 'b8a82c4a-e8ca-41ca-8dae-9fe6daeb167d', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5294ZSIL/'}]\n",
      "077_PWKJ_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM08966472', 'uuid': 'd5b53443-c893-4b53-885a-dab5be2215ac', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM08966472/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9967KFUW', 'uuid': '952e454e-d3e8-462f-80cf-5dd427ae4972', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9967KFUW/'}]\n",
      "028_AJ_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM31122757', 'uuid': '820244a7-ec88-496c-ba00-eda81f46a10d', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM31122757/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1802ZPXC', 'uuid': '0eb73cca-b1fe-4633-b90a-7850ff2c5157', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1802ZPXC/'}]\n",
      "078_PWKJ_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM26212740', 'uuid': '08e8b673-c7fe-4537-8faa-ccb919403b43', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM26212740/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9748KTTY', 'uuid': 'daead910-2d55-4d20-a4ff-bcbbe1b7370f', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9748KTTY/'}]\n",
      "031_AJ_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM93419295', 'uuid': '4386c5b7-f72e-4ace-ae7a-9418d16c0b61', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM93419295/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM4608HMDV', 'uuid': 'b67c46ec-d705-43c8-9535-6b1aad588fa5', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4608HMDV/'}]\n",
      "079_PWKJ_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM33203788', 'uuid': '323a9fb2-a007-438b-bc78-0990d636eaad', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM33203788/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6943ARXK', 'uuid': '67c529cd-1460-4fb8-9be7-73a8c044eb92', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6943ARXK/'}]\n",
      "030_AJ_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM24985438', 'uuid': 'bfdcd53b-9b04-479e-a4ee-1cae448d9475', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM24985438/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM5549SWHL', 'uuid': 'b9144ccb-6175-4b03-8ab1-289bf594a1fe', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5549SWHL/'}]\n",
      "080_PWKJ_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM62982620', 'uuid': 'f56a146c-9c9d-4794-9436-81730a941933', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM62982620/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8513LDPM', 'uuid': '0a3f1d15-a7f8-4c09-a87e-23966a8feec6', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8513LDPM/'}]\n",
      "033_AJ_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM66449571', 'uuid': '418a43e3-f725-4c54-a659-99c25bf1a6f8', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM66449571/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM7128WVEJ', 'uuid': '0738e6de-dcab-49ae-86c9-a6ed1d69eccb', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7128WVEJ/'}]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "081_PWKJ_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM92869047', 'uuid': '118f6311-3fc4-4006-8245-e3f16c44616a', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM92869047/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM3449RSOM', 'uuid': 'aba5e223-75b7-4180-8d99-dc96c28c05cd', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3449RSOM/'}]\n",
      "032_AJ_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM79829816', 'uuid': '2072f0ab-f6b8-4937-8bc4-d0d441894a51', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM79829816/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2048BQHM', 'uuid': 'bc124c8c-160f-4e8b-8cd0-b2d91cc7e03a', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2048BQHM/'}]\n",
      "084_PWKJ_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM52490317', 'uuid': '3b050af2-f7dd-4ee0-bb77-2480a4b4c565', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM52490317/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM5687CRRX', 'uuid': '8cc97f35-a489-446c-a1b9-20d29144df7a', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5687CRRX/'}]\n",
      "035_AJ_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM39935197', 'uuid': '32b3ecf4-d62f-4ae8-b1c5-12d7dae9b3a5', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM39935197/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2812EMHZ', 'uuid': '3eb7e846-2b60-4d3f-bf7f-5f276c2caa97', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2812EMHZ/'}]\n",
      "085_PWKJ_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM36796607', 'uuid': 'cf3a6deb-227d-46cf-99fc-091dd7b5fe57', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM36796607/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2657UOQA', 'uuid': '594d8a7e-eefd-49eb-b4a3-8ef201a55899', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2657UOQA/'}]\n",
      "036_129S1J_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM40397039', 'uuid': '0734aa5d-4700-4d8e-93f1-2fdd547e364b', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM40397039/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9118PKWD', 'uuid': '72904729-798f-46f2-83b3-6d9bb9be239a', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9118PKWD/'}]\n",
      "086_CASTJ_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM14666728', 'uuid': 'bb8e38e0-fbbf-4d0b-b25d-e2aa3d30bdf7', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM14666728/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM5819SOLY', 'uuid': '2bdf312f-3696-4437-a50b-5de51f96d293', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5819SOLY/'}]\n",
      "037_129S1J_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM46868917', 'uuid': '51ea91a6-fc1c-41f1-87e4-eeb4c068e258', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM46868917/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8184LOKP', 'uuid': '31239451-0a63-4ad0-a1cd-682fd6dd7ef9', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8184LOKP/'}]\n",
      "087_CASTJ_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM44794195', 'uuid': '4f4b73e6-c90e-4f08-b430-88092f68306d', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM44794195/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0487BHBU', 'uuid': 'e483631b-70a4-491c-9636-3454dec4f92f', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0487BHBU/'}]\n",
      "038_129S1J_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM91678148', 'uuid': 'f0250054-82d6-47b5-9eb1-4ab4cccda44b', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM91678148/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM3629MXTV', 'uuid': '01421cce-634f-441e-b3fc-3456fc41505f', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3629MXTV/'}]\n",
      "088_CASTJ_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM06875364', 'uuid': '6b2e42ac-8252-4684-8668-d2474222cebd', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM06875364/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM5847JXKA', 'uuid': '66de4d37-023a-432e-8be8-56d2a16a3576', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5847JXKA/'}]\n",
      "041_129S1J_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM23818121', 'uuid': 'e9ba223a-25ab-4a2d-a67a-99fb7a159c70', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM23818121/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9345WJHN', 'uuid': '1a5679fe-419d-4ca7-bb10-8213c3548d87', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9345WJHN/'}]\n",
      "089_CASTJ_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM54707771', 'uuid': '1abd5819-2a46-461c-b726-2de2ca6de0e1', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM54707771/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM4021BOPQ', 'uuid': 'efd0e2ef-0e7c-4338-8ce9-977e23afb522', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4021BOPQ/'}]\n",
      "040_129S1J_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM15974006', 'uuid': '08cbac9b-2910-4850-a92f-7de4b3c7ea09', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM15974006/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM5009WOZO', 'uuid': '4964639a-6440-461d-9c7f-61e8dc85e9eb', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5009WOZO/'}]\n",
      "090_CASTJ_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM73074702', 'uuid': 'd710a38b-498f-4780-8636-e593cc546c64', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM73074702/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM4983NNIU', 'uuid': '6a0b4d8e-0f7a-4d9a-84a2-5a2b9993deb2', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4983NNIU/'}]\n",
      "043_129S1J_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM15577726', 'uuid': '2da81692-5891-4672-aaf2-401f63ce292d', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM15577726/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM5004FHRH', 'uuid': '1be23e33-fe8f-4aca-bc19-c12344be2dbc', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5004FHRH/'}]\n",
      "091_CASTJ_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM18004864', 'uuid': '100e69a4-43b3-4d94-9455-32ac224f8e09', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM18004864/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2511KNVM', 'uuid': '09888e86-39f2-4173-89c9-ecaaca0c2bda', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2511KNVM/'}]\n",
      "044_129S1J_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM05869884', 'uuid': 'eba9851b-062f-4c3a-90c7-a9eeb2b65b9c', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM05869884/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0279NVKF', 'uuid': '8850f6b8-70c2-488a-a3e4-a0073e6be535', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0279NVKF/'}]\n",
      "094_CASTJ_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM35713742', 'uuid': 'bb8c02e5-ec0c-4782-840c-10d513d70611', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM35713742/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8621PXJN', 'uuid': 'bf102af3-c393-430f-8e7b-75541d7169be', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8621PXJN/'}]\n",
      "045_129S1J_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM76586550', 'uuid': 'c7b1d7c6-2ea9-4856-b94a-2f6cee7191a9', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM76586550/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM5577TDWZ', 'uuid': 'e15dad49-60fa-43a7-8de3-74bdbe9ede4a', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5577TDWZ/'}]\n",
      "093_CASTJ_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM92959368', 'uuid': 'dc2e6674-848d-4048-a845-be6a9a973194', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM92959368/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0998PUXL', 'uuid': '6a09b51e-7ea8-4e2e-915d-e12e68d597b9', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0998PUXL/'}]\n",
      "058_WSBJ_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM97534373', 'uuid': '62610a5d-8a2f-4f23-aad7-8856c86bfaa1', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM97534373/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM3432ZCOX', 'uuid': '41ea5d49-b645-4e6c-88f1-8e5d96b213ab', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3432ZCOX/'}]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "046_NZOJ_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM56267464', 'uuid': '7cbf667d-8e6b-4136-a86d-39736d913051', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM56267464/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0510YNAJ', 'uuid': '935b1909-c93c-452f-8511-001e4d66958f', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0510YNAJ/'}]\n",
      "057_WSBJ_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM25214382', 'uuid': 'fe31cebf-74f0-452c-a375-8c39654fc11f', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM25214382/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1456DXFQ', 'uuid': 'cd8afd32-4cdd-496a-9147-1c7f9f5f7c49', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1456DXFQ/'}]\n",
      "047_NZOJ_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM83837273', 'uuid': 'c5def42a-2e25-493d-b51b-fc1b48e0331f', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM83837273/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM7577ZSZL', 'uuid': '1991df86-8766-43b4-819b-a0229f771252', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7577ZSZL/'}]\n",
      "060_WSBJ_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM47500306', 'uuid': '6f942d7c-7e24-45eb-aff3-282ac15dfc66', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM47500306/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM7766OOSU', 'uuid': '8172c6fb-790f-46d5-a542-a89c7cf099fa', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7766OOSU/'}]\n",
      "048_NZOJ_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM08931386', 'uuid': 'a429ca65-df0b-41c7-be04-fc953ec09cf4', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM08931386/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM4204PXKJ', 'uuid': '94d46c13-198e-4269-838a-644f91cf1215', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4204PXKJ/'}]\n",
      "059_WSBJ_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM46730920', 'uuid': '3d0430f9-a23b-40d4-884c-8961dded82a9', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM46730920/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6894ALPA', 'uuid': '42f7e1da-7d83-47a0-847b-6fa97350ce5d', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6894ALPA/'}]\n",
      "049_NZOJ_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM83313462', 'uuid': '86b5c320-2320-4ef6-afd8-c75aa2f1e1c7', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM83313462/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8056EVEE', 'uuid': '8ce544ea-16d8-4790-8335-0b2cecf8c3d4', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8056EVEE/'}]\n",
      "062_WSBJ_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM39521310', 'uuid': 'b27595d4-c5fa-450e-9254-66d69134c0dc', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM39521310/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM3113AHHA', 'uuid': 'fbdc87c8-e144-4acb-b62e-99e4eec5bf52', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3113AHHA/'}]\n",
      "050_NZOJ_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM60684929', 'uuid': '46a58b33-c0fb-4637-9d89-a931bf3cbb6b', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM60684929/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8878JBBO', 'uuid': 'fc878f40-fe2a-49b8-9359-2b83f6a88ec4', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8878JBBO/'}]\n",
      "061_WSBJ_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM91863238', 'uuid': '536f0fbc-1e95-4070-aee2-1be519a4f6e8', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM91863238/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8737KUQW', 'uuid': '7b21b0fd-747e-46aa-aa29-e61d533adadf', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8737KUQW/'}]\n",
      "051_NZOJ_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM65100276', 'uuid': 'ec917d24-1857-4710-8a03-0e44ae56208b', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM65100276/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2214BDBR', 'uuid': 'd6de7007-8add-4482-aa9e-3d4a857d451e', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2214BDBR/'}]\n",
      "096_WSBJ_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM31948477', 'uuid': '25e334fc-7fef-440a-87e7-42fb2b4407bd', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM31948477/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6802LFPK', 'uuid': '50c99171-17dc-4a28-ab0d-37c7e744559c', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6802LFPK/'}]\n",
      "052_NZOJ_10F_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM35827492', 'uuid': '89ad5f34-2764-4e32-9923-28a51efc3096', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM35827492/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8146OBNU', 'uuid': 'd6776701-67b0-4502-8e93-e6e5db2595eb', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8146OBNU/'}]\n",
      "063_WSBJ_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM43274889', 'uuid': '98b82cbe-a826-412f-9f84-c3979cd45f10', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM43274889/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6965RRET', 'uuid': 'f670bdaa-d614-42c6-a192-c9a943fa8c43', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6965RRET/'}]\n",
      "053_NZOJ_10M_06 [{'accession_prefix': 'igvftst', 'name': 'TSTSM51587050', 'uuid': '14c9cba3-d3fe-4709-8f08-6a78830b2c3f', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM51587050/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM4952NHSA', 'uuid': 'f7a17647-e601-438e-9934-db753663e417', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4952NHSA/'}]\n",
      "118_B6J_10F_21 [{'accession_prefix': 'igvftst', 'name': 'TSTSM75253545', 'uuid': 'db1101a0-69ff-4bab-906c-7a3ee2011713', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM75253545/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM7772DGMA', 'uuid': 'bcde37eb-fbd6-45cc-83e1-d9aba738e997', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7772DGMA/'}]\n",
      "117_B6J_10M_21 [{'accession_prefix': 'igvftst', 'name': 'TSTSM62420264', 'uuid': '63c625ae-1dfa-431f-95d9-b5d19fe1fd9b', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM62420264/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2279TQMZ', 'uuid': 'c9502e1f-2b03-4e21-a073-0be53e87c45f', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2279TQMZ/'}]\n",
      "106_CASTJ_10F_21 [{'accession_prefix': 'igvftst', 'name': 'TSTSM30808904', 'uuid': 'bc94df23-8c9a-4cf8-a0da-414cc717bced', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM30808904/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM3421UUOE', 'uuid': '08119648-380d-4d0e-b1cb-6dfc6af375eb', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3421UUOE/'}]\n",
      "105_CASTJ_10M_21 [{'accession_prefix': 'igvftst', 'name': 'TSTSM06196990', 'uuid': 'f08d15dc-e3e1-4135-b572-612a87156ace', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM06196990/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6969SYOD', 'uuid': '21fbdda9-f0a4-4f4c-b303-46b6d6766b4c', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6969SYOD/'}]\n",
      "120_B6J_10F_21 [{'accession_prefix': 'igvftst', 'name': 'TSTSM33801505', 'uuid': 'dfc12fcf-cad3-4fba-af48-38436eb07943', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM33801505/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM7094ZNGP', 'uuid': '062d0409-f18b-4e32-b550-f2b7e6febcb7', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7094ZNGP/'}]\n",
      "119_B6J_10M_21 [{'accession_prefix': 'igvftst', 'name': 'TSTSM44207215', 'uuid': 'c08dea93-759b-49d2-b192-6796fd940ae0', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM44207215/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM5782OIYL', 'uuid': '2dfe1823-10a6-42db-8113-7fa478df3d31', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5782OIYL/'}]\n",
      "108_CASTJ_10F_21 [{'accession_prefix': 'igvftst', 'name': 'TSTSM54562708', 'uuid': '8c0e5a9e-71d8-4d23-8fe7-fdb9c5ef3ca0', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM54562708/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0447EABS', 'uuid': 'b61a1809-38bf-4b13-8442-c71e9038d276', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0447EABS/'}]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "107_CASTJ_10M_21 [{'accession_prefix': 'igvftst', 'name': 'TSTSM56825988', 'uuid': 'cfd1ca9e-6ed3-4eb1-bb1d-7b954b5d58e0', 'see_also': 'https://api.sandbox.igvf.org/tissues/TSTSM56825988/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM5157GFSA', 'uuid': 'edd70761-6e0b-46d7-bda1-e34c5a3b2266', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5157GFSA/'}]\n",
      "092_CASTJ_10F_03 [{'accession_prefix': 'igvf', 'name': 'IGVFSM4189AKUL', 'uuid': '8cb2f317-879e-4ab1-82fd-0b95c3e660e0', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4189AKUL/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0322EOHO', 'uuid': '862fc166-261c-4e78-b219-611fe387a33d', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0322EOHO/'}]\n",
      "016_B6J_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM4465YOJN', 'uuid': 'd296f0d1-265c-426b-a808-95c00b827243', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4465YOJN/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM5984CJUU', 'uuid': 'b28f2e52-2d05-4743-865b-1cbfef1844b7', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5984CJUU/'}]\n",
      "066_NODJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM6418OJGM', 'uuid': '440f3c2f-f9e8-431d-8857-98ad971969e5', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6418OJGM/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM3183PSGJ', 'uuid': '46633bce-f56c-499e-b78a-c553a1ebe129', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3183PSGJ/'}]\n",
      "017_B6J_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM5230YSVJ', 'uuid': '46a10a17-94d9-4918-b264-f9c2d37b4af5', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5230YSVJ/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6071JHCX', 'uuid': 'e2e57b68-169f-4d42-b023-d7a9d93b95bd', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6071JHCX/'}]\n",
      "067_NODJ_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM1910QUCB', 'uuid': '995ed9a4-fa5e-4b67-88f0-1dd04b9fd652', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1910QUCB/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6194RATY', 'uuid': '81cd53bd-4f6d-4ed5-a190-59abd89f5730', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6194RATY/'}]\n",
      "018_B6J_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM5188HZIG', 'uuid': '72c9d73f-ba1b-4bfd-a479-070c16edb35b', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5188HZIG/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9214NKTQ', 'uuid': '587fc623-bc5e-4c8d-992e-8bc0061db504', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9214NKTQ/'}]\n",
      "068_NODJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM4182XRQE', 'uuid': '0df4e41b-7154-42e1-b2cf-b67e3c577301', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4182XRQE/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8455TNVO', 'uuid': 'e37fc772-2f5f-471e-bc59-7ac5e14d7f15', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8455TNVO/'}]\n",
      "019_B6J_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM6699LJHC', 'uuid': '052dfa8a-01ac-4dfb-869d-d4e9ee289a9d', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6699LJHC/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0166IBCH', 'uuid': 'b28ebfd4-496c-455c-85e4-9a93543400ec', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0166IBCH/'}]\n",
      "069_NODJ_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM6842OMYX', 'uuid': '074e7cd8-952c-4775-b90b-ea4c12703ad8', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6842OMYX/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8873UKSN', 'uuid': '0528ceaf-a31d-475d-9d3f-5ec93a8e6034', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8873UKSN/'}]\n",
      "020_B6J_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM8528YUHU', 'uuid': 'adf0896f-8640-4f7f-9e97-02df92ce8a42', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8528YUHU/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM3470VQGS', 'uuid': '81233ee3-0e9f-40fd-b633-57175ab76b6d', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3470VQGS/'}]\n",
      "070_NODJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM0017ZTCP', 'uuid': '72b5c8a5-0da1-4d74-92c9-d02c06f565c0', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0017ZTCP/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1602MMIE', 'uuid': 'ac2f20e4-6f48-4560-955d-54674635261e', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1602MMIE/'}]\n",
      "021_B6J_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM8952OBMF', 'uuid': '2c9a0d8c-6a94-42b7-8868-8ae980e92955', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8952OBMF/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM7323JYDH', 'uuid': '4befccea-a196-4637-b5ad-8dac3c87396e', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7323JYDH/'}]\n",
      "071_NODJ_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM8886HJDO', 'uuid': '0ad97244-1760-4aec-a5e6-b493bbfd37a3', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8886HJDO/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6636PLRM', 'uuid': '2e7f40b4-ceff-4929-8243-fb20525d6f9d', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6636PLRM/'}]\n",
      "024_B6J_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM1381VQIA', 'uuid': 'f5a4a2eb-a566-4c7e-afc3-2e1bad239247', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1381VQIA/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8656OHKM', 'uuid': 'c3606b99-5215-4de4-ab2e-f72605517f96', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8656OHKM/'}]\n",
      "072_NODJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM0442MPEV', 'uuid': 'c92f503a-163e-46fb-af62-1ccf44e42505', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0442MPEV/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9173JXFM', 'uuid': '67a6dcd9-7c85-47f9-9ab6-c5eb73b4e5d7', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9173JXFM/'}]\n",
      "025_B6J_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM4934KHPJ', 'uuid': 'ce71f2c6-e245-4cb8-8b76-1c7ec4766e8f', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4934KHPJ/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1156KMPC', 'uuid': 'd8ed5567-a834-47d8-88d4-f38a15971c2b', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1156KMPC/'}]\n",
      "073_NODJ_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM0489RZXG', 'uuid': 'fd63a7d2-7e5a-4838-867a-dfbec8a5604f', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0489RZXG/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8174YDBO', 'uuid': 'd67fb1b8-02aa-4495-b1e5-650f87e5b955', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8174YDBO/'}]\n",
      "026_AJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM8677EQEG', 'uuid': 'c133084a-a1ac-4a11-9c67-4117a111990a', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8677EQEG/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2617YKKQ', 'uuid': 'bd5dafed-5fa0-44c6-b0d3-6ca7a6a397d7', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2617YKKQ/'}]\n",
      "076_PWKJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM5417LNGC', 'uuid': '4501f5ab-feb3-4ec6-8bd6-89577f444ac6', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5417LNGC/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6924RRKD', 'uuid': '457e115a-26a7-4b69-9b8a-3a9dc59d40dd', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6924RRKD/'}]\n",
      "029_AJ_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM1458SAIZ', 'uuid': '7490152e-3f0d-4e01-b0d3-c09003b0cae8', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1458SAIZ/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8548SQFG', 'uuid': 'e9e8411d-7338-42c5-83bd-7150c52024bb', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8548SQFG/'}]\n",
      "077_PWKJ_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM4103LTZB', 'uuid': '8000bba1-9994-43ad-a127-660789bc6cf9', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4103LTZB/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM5500OUCV', 'uuid': 'ae3d00e6-1f7f-442e-8c1e-980356f9b4c9', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5500OUCV/'}]\n",
      "028_AJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM4796DKJO', 'uuid': 'bd5c5c6d-83ae-4993-80d9-ac43288ad9d9', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4796DKJO/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM4881USYE', 'uuid': '5f8caf9f-e6b6-4607-b59c-ee695b60e3e3', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4881USYE/'}]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "078_PWKJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM9111MIVH', 'uuid': '09bdfe37-27e8-4e14-a8b7-d2ddd7bbc7ac', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9111MIVH/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2146SQAW', 'uuid': '2524a727-bbc3-48b8-a40f-fadf802233ad', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2146SQAW/'}]\n",
      "031_AJ_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM6893IKIH', 'uuid': '61a2b143-b341-437e-8117-ea685322676d', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6893IKIH/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM3351WODP', 'uuid': '7e190df1-ae82-4925-bc22-1ec75270b252', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3351WODP/'}]\n",
      "079_PWKJ_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM3536WDNR', 'uuid': 'd18320cc-ed7a-4d10-8a8f-d918b492ab8d', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3536WDNR/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM3744QPBB', 'uuid': 'eaab5bfc-ca1e-415b-b41d-c652530c842b', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3744QPBB/'}]\n",
      "030_AJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM4955LORB', 'uuid': 'c0dcbc3b-a748-4e07-99c2-cda6cedb63a4', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4955LORB/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9545EZNW', 'uuid': 'a7a8e27d-4f85-4335-afc0-bf861208d9ba', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9545EZNW/'}]\n",
      "080_PWKJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM0567BKYP', 'uuid': 'ec84894a-3d99-4c0c-94b8-2822cc643237', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0567BKYP/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1552JBSI', 'uuid': 'dde86bc8-bba1-41d2-97d3-970bb8128491', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1552JBSI/'}]\n",
      "033_AJ_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM2119ANJP', 'uuid': '11b78286-7eae-49b9-a664-6f5d3c18e0c0', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2119ANJP/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1755SWJI', 'uuid': '692ee9c5-aad7-4df2-a223-ab3a4c910ac5', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1755SWJI/'}]\n",
      "081_PWKJ_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM8490BJOU', 'uuid': '068d55a1-c4d3-4626-a515-f341b62a3feb', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8490BJOU/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6086KVUS', 'uuid': '137daebd-0599-427a-a61e-c2b0ce4451a4', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6086KVUS/'}]\n",
      "032_AJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM3957TLDF', 'uuid': 'a4b79c62-05d8-483d-91e2-19c4fb3b8922', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3957TLDF/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6828FNBA', 'uuid': 'df65d29a-952c-49f9-8a26-9041a4b77c13', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6828FNBA/'}]\n",
      "084_PWKJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM4249JBEE', 'uuid': '1ba4b0aa-986d-48f4-a657-0cb7b9683ac3', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4249JBEE/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM7955ABKB', 'uuid': '31586d6d-13e8-46bc-babf-5353f58cb4a1', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7955ABKB/'}]\n",
      "035_AJ_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM3012KPRJ', 'uuid': 'c3c03275-a389-40be-82f5-591d3c902716', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3012KPRJ/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1951AMWE', 'uuid': '966a66ee-8fb7-43e1-a45b-87dce5031e97', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1951AMWE/'}]\n",
      "085_PWKJ_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM3050OPDO', 'uuid': '96e18fd2-04a5-4e56-802f-99cf2c959b8b', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3050OPDO/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM4510XDYL', 'uuid': 'c4301ca1-fa47-4472-8f2b-28c0220995fd', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4510XDYL/'}]\n",
      "036_129S1J_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM9398YHXY', 'uuid': '2796ef8e-c018-46e3-868a-bcf5a5e69947', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9398YHXY/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM4167SEXH', 'uuid': '47c3f193-a365-4c91-9499-514f17567b8c', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4167SEXH/'}]\n",
      "086_CASTJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM9286SBAR', 'uuid': 'be30909f-0d2d-47b2-b3c5-058531f7b39e', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9286SBAR/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM7092FDTP', 'uuid': '46a08032-bca7-4e1f-bc6f-85a9e53da1f7', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7092FDTP/'}]\n",
      "037_129S1J_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM3753OVHM', 'uuid': '29892f46-8636-4e08-b35e-b99160f47926', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3753OVHM/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9806TYXF', 'uuid': 'fa3e0716-0e6f-4bf2-ad13-94d94aa94018', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9806TYXF/'}]\n",
      "087_CASTJ_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM5457PTKY', 'uuid': '778c18be-698e-4ff9-b328-3f4dadf18381', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5457PTKY/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM4956HIAH', 'uuid': '5d1e9d06-169a-4804-b750-751490291cd2', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4956HIAH/'}]\n",
      "038_129S1J_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM9530NJWB', 'uuid': 'ea7743e9-9f48-494f-94d8-2266750dec1d', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9530NJWB/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM5458IIQG', 'uuid': 'a3154a90-e6c9-4527-bda5-1e5572b001b1', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5458IIQG/'}]\n",
      "088_CASTJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM0690VJRZ', 'uuid': '7296f8ef-b4d3-4e7a-adbe-3302f64d8ea9', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0690VJRZ/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9111THIZ', 'uuid': 'ae9db008-41bb-47a0-bc66-a926ae61c174', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9111THIZ/'}]\n",
      "041_129S1J_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM4269MUAL', 'uuid': '7806d098-aa8d-414f-9527-eb359ccb6128', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4269MUAL/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0340DRJV', 'uuid': 'f9a3a29d-08c5-40f3-97e0-86f7f78b20d8', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0340DRJV/'}]\n",
      "089_CASTJ_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM8458TCXX', 'uuid': '8be9fbcb-ad91-48b1-b347-0a343f440317', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8458TCXX/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6740OVAR', 'uuid': '04c1fc90-89a5-4d8c-921b-026667ef8a3f', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6740OVAR/'}]\n",
      "040_129S1J_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM7479OZRI', 'uuid': 'cd77ff2b-d024-43aa-8895-7c5ceb8bcc66', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7479OZRI/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6249RJQV', 'uuid': '69619cc3-2e7c-4c54-a5a6-77b816868c12', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6249RJQV/'}]\n",
      "090_CASTJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM1264GFRJ', 'uuid': '8c6a0077-6ddd-4f58-b9da-f8322e48ed3e', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1264GFRJ/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM5214BQRA', 'uuid': 'b3fbbfa8-a5da-4702-a00a-05abc74a9156', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5214BQRA/'}]\n",
      "043_129S1J_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM6782GARZ', 'uuid': '463b6afc-3248-4592-bdfe-03a7692ed950', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6782GARZ/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8930SAOK', 'uuid': '6300e920-318f-4cbf-a6f4-6843e98fce0d', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8930SAOK/'}]\n",
      "091_CASTJ_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM4321DFFH', 'uuid': '1e9ffafe-f59c-40ae-8eca-97a9b5ef57ad', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4321DFFH/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8357WYSN', 'uuid': 'c85274aa-790b-413d-b3b2-7a79329c0d32', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8357WYSN/'}]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "044_129S1J_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM6738OHZO', 'uuid': '83e218de-7b42-426b-81be-65127043e3e0', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6738OHZO/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0539MTAN', 'uuid': 'b49b3933-3e97-4507-9a8a-c122e6a8bc06', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0539MTAN/'}]\n",
      "094_CASTJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM0021WQYA', 'uuid': 'da950e21-1ab3-4705-9511-5dfe7dc4176f', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0021WQYA/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1257AOMZ', 'uuid': 'd959cea7-a785-401c-aa5e-89c07cdb5de7', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1257AOMZ/'}]\n",
      "045_129S1J_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM6742HLKN', 'uuid': '649ec13f-97a1-4398-9148-f381fddd6df0', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6742HLKN/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1996CIIR', 'uuid': '9fce9d03-d51b-49cf-b8e1-da310b5fe339', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1996CIIR/'}]\n",
      "093_CASTJ_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM5017BJEE', 'uuid': '42a10680-c23d-4055-a9fe-24f2b800960a', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5017BJEE/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1436ECIH', 'uuid': 'f3bf5922-4e53-4db9-a888-1f7aecb85752', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1436ECIH/'}]\n",
      "058_WSBJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM4130KIEV', 'uuid': '548222ac-d86c-4d6e-95ba-7b7530ab3d0c', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4130KIEV/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0966WWIZ', 'uuid': '1103358a-4d00-4b62-b064-7afa51d094a1', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0966WWIZ/'}]\n",
      "092_CASTJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM8134HPUH', 'uuid': '4bce5bed-6719-4220-9d3f-81bfed153cf9', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8134HPUH/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1078KTYU', 'uuid': '1639516f-a979-4240-a8eb-3454a0957d3f', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1078KTYU/'}]\n",
      "047_NZOJ_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM4107ZNSE', 'uuid': '986a5e15-5553-48f5-a898-643632262c44', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4107ZNSE/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6703AANL', 'uuid': '0f90a9cb-a38b-4fef-945b-ae3297b89f13', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6703AANL/'}]\n",
      "057_WSBJ_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM8498UCVA', 'uuid': 'f3fefb5c-a44f-4436-acf4-b3effb637d1b', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8498UCVA/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM7450WJIV', 'uuid': 'b40b8e46-0b01-42e8-8a5e-d98a9b37bb41', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7450WJIV/'}]\n",
      "048_NZOJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM0298LFOQ', 'uuid': '6a108204-df5b-48f3-ad48-4082ae333c0b', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0298LFOQ/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8754ELWG', 'uuid': 'c78670f9-3e00-4d19-8031-d9520b2c468f', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8754ELWG/'}]\n",
      "060_WSBJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM3371TFMC', 'uuid': 'f97bd66c-bae5-44f4-9535-d12e4e144f62', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3371TFMC/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM0752ZRMG', 'uuid': '6856aaab-7fec-4b0d-b0db-479e1d8c20a6', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM0752ZRMG/'}]\n",
      "049_NZOJ_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM9197PRMM', 'uuid': '38ea1daf-8a5a-48ea-8693-d50a47cbb0f8', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9197PRMM/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2572HXXE', 'uuid': '7ac92b44-01cb-46ac-a5ef-c0acb4ec2489', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2572HXXE/'}]\n",
      "059_WSBJ_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM9356QKBM', 'uuid': '96cacd9f-2134-4ad6-8bfd-1fc3db857b77', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9356QKBM/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2287EHGA', 'uuid': 'e0d1f6d6-2e05-4b9f-880b-b9d8f55f32ee', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2287EHGA/'}]\n",
      "050_NZOJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM2120LYVH', 'uuid': '7e5f6196-dd64-4376-947d-0709f4f5e126', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2120LYVH/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8205AVFV', 'uuid': 'e65ecb63-e05b-4de4-9a9c-dd93f621c017', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8205AVFV/'}]\n",
      "062_WSBJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM7560RVGB', 'uuid': '75b3af62-8e22-4fef-8944-460d7fefc0d4', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7560RVGB/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2047VYYJ', 'uuid': '59cda413-8413-49d9-81f2-d23cfdf01085', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2047VYYJ/'}]\n",
      "051_NZOJ_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM3556PAEN', 'uuid': 'b13f1408-ec31-4c0e-9cd7-2bc032536fc3', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3556PAEN/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9711JMMI', 'uuid': '7afe7c7a-4d03-42c3-9ef4-14246f0930df', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9711JMMI/'}]\n",
      "061_WSBJ_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM3364FXFB', 'uuid': '37270f8d-8baa-4190-83d7-b5d48b971ef4', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3364FXFB/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM4632PCNC', 'uuid': 'e04ac839-27cc-45e7-8180-c01365935d41', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM4632PCNC/'}]\n",
      "052_NZOJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM5374RXFR', 'uuid': '0fb589eb-73cf-4b62-8ef1-4e6a59a8a315', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM5374RXFR/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9813BNZU', 'uuid': '52d7ba9c-24a0-4ae1-860f-9b19c6fd3a53', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9813BNZU/'}]\n",
      "096_WSBJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM6620ZEBK', 'uuid': '5278d9e7-f87b-4403-aacd-30133b4068d8', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM6620ZEBK/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1348QFIK', 'uuid': '30245671-8960-4ce9-b7c6-28a71f921adf', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM1348QFIK/'}]\n",
      "053_NZOJ_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM9817PYLC', 'uuid': 'd9e6d3fc-d959-49aa-9e5a-b1cbbd3db4a8', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9817PYLC/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM3167TBKE', 'uuid': 'f2a79240-36bb-4126-9440-3f63fdebdacd', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM3167TBKE/'}]\n",
      "063_WSBJ_10M_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM7877XJMN', 'uuid': 'd9becf1f-3dcb-434f-ae18-78195c9f1eba', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM7877XJMN/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9074ZTED', 'uuid': '32feebe5-5fd9-437e-8ab3-dc7e74b46720', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM9074ZTED/'}]\n",
      "046_NZOJ_10F_01 [{'accession_prefix': 'igvf', 'name': 'IGVFSM8078MQQC', 'uuid': 'aaaa720e-a504-426b-b79d-fa8190d4523f', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM8078MQQC/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2791YRHP', 'uuid': 'aace359e-5ae0-404c-930d-1490c1196452', 'see_also': 'https://api.data.igvf.org/tissues/IGVFSM2791YRHP/'}]\n"
     ]
    }
   ],
   "source": [
    "for key in submitted_tissues:\n",
    "    value = submitted_tissues[key]\n",
    "    if len(value) != 1:\n",
    "        print(key, value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "legislative-shape",
   "metadata": {},
   "outputs": [],
   "source": [
    "def validate_tissue_genotype(tissue, genotype):\n",
    "    normalized_genotype = normalize_strain(genotype)\n",
    "    assert tissue == normalized_genotype, f\"{tissue} != {normalized_genotype} ({genotype})\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "relative-corps",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Samples - 8 founders (1621, 16)\n",
      "Samples - Bridge (416, 23)\n",
      "Samples - F1s (1920, 18)\n",
      "Samples - CClines (5720, 18)\n",
      "692\n"
     ]
    }
   ],
   "source": [
    "def load_tissues(strain_code_to_name):\n",
    "    if models.Tissue.objects.count() > 0:\n",
    "        truncate(models.Tissue)\n",
    "\n",
    "    tissues = load_tissue_sheets()\n",
    "    submitted_tissues = load_submitted_tissues()\n",
    "    los_angeles_tz = zoneinfo.ZoneInfo(\"America/Los_Angeles\")\n",
    "    \n",
    "    failed = False\n",
    "    for i, row in tissues.iterrows():\n",
    "        if not pandas.isnull(row[\"Tissue\"]) and row[\"IGVF Sample BOX\"] != \"DNA Extraction Optimization\":\n",
    "            mouse_tissue_id_label = \"Mouse_Tissue ID\"\n",
    "\n",
    "            mouse_tissue_name = normalize_strain(row[mouse_tissue_id_label])\n",
    "            mouse_tissue_name_fields = parse_mouse_tissue(mouse_tissue_name)\n",
    "\n",
    "            if not mouse_tissue_name_fields.mouse_strain in strain_code_to_name:\n",
    "                raise ValueError(f\"tissue strain field in {mouse_tissue_name} not recognized\")\n",
    "\n",
    "            mouse_name = \"_\".join(mouse_tissue_name_fields[0:3])\n",
    "            try:\n",
    "                mouse = models.Mouse.objects.get(name=mouse_name)\n",
    "            except models.Mouse.DoesNotExist:\n",
    "                print(\"row {}, {} was not found\".format(i+2, mouse_name))\n",
    "                failed = True\n",
    "                continue\n",
    "\n",
    "            mouse_tissue = \"_\".join(mouse_tissue_name_fields)\n",
    "\n",
    "            genotype = row[\"Genotype\"]\n",
    "\n",
    "            # this is the \"label swap\" on spreadsheet rows 602-605.\n",
    "            if mouse_name == \"092_CASTJ_10F\":\n",
    "                genotype = \"CASTJ\"\n",
    "            # this is the other half the swap on spreadsheet rows 1522-1525\n",
    "            elif mouse_name == \"046_NZOJ_10F\":\n",
    "                genotype = 'NZOJ'\n",
    "\n",
    "            validate_tissue_genotype(mouse.strain.name, genotype)\n",
    "\n",
    "            tissue_terms = []\n",
    "            for term_curie, term_name in tissue_dissection_to_ontology_map[row[\"Tissue\"]]:\n",
    "                tissue_terms.append(ontology_map[term_curie])\n",
    "\n",
    "            #age, age_units = parse_timepoint(row[\"Timepoint\"])\n",
    "\n",
    "            if pandas.isnull(row[\"Approx. sac time\"]):\n",
    "                sac_time = datetime.time(0,0,0)\n",
    "            else:\n",
    "                sac_time = datetime.time(\n",
    "                    row[\"Approx. sac time\"].hour,\n",
    "                    row[\"Approx. sac time\"].minute,\n",
    "                    row[\"Approx. sac time\"].second,    \n",
    "                )\n",
    "\n",
    "\n",
    "            record = models.Tissue(\n",
    "                mouse=mouse,\n",
    "                name = mouse_tissue,\n",
    "                description = row[\"Tissue\"],\n",
    "                #dissection_time=dissection,\n",
    "                #age=age,\n",
    "                #age_units=age_units,\n",
    "                tube_label=row[\"Tube label\"],\n",
    "                dissector=str_or_empty(row[\"Dissector\"]),\n",
    "                dissection_notes=str_or_empty(row[\"Comment\"]),\n",
    "            )\n",
    "\n",
    "            if not pandas.isnull(row[\"Dissection date\"]):\n",
    "                record.dissection = datetime.datetime(\n",
    "                    row[\"Dissection date\"].year,\n",
    "                    row[\"Dissection date\"].month,\n",
    "                    row[\"Dissection date\"].day,\n",
    "                    sac_time.hour,\n",
    "                    sac_time.minute,\n",
    "                    sac_time.second,\n",
    "                    tzinfo=los_angeles_tz,\n",
    "                )\n",
    "\n",
    "            tube_weight_label = \"tube weight (g)\"\n",
    "            if not pandas.isnull(row[tube_weight_label]):\n",
    "                record.tube_weight_g = float(row[tube_weight_label])\n",
    "\n",
    "            total_weight_label = \"tube+tissue weight (g)\"\n",
    "            if not pandas.isnull(row[total_weight_label]):\n",
    "                record.total_weight_g = float(row[total_weight_label])\n",
    "\n",
    "            record.save()\n",
    "            record.ontology_term.set(tissue_terms)\n",
    "\n",
    "            if row[mouse_tissue_id_label] in submitted_tissues:\n",
    "                accessions = []\n",
    "                for accession_row in submitted_tissues[row[mouse_tissue_id_label]]:\n",
    "                    if pandas.notnull(accession_row[\"name\"]):\n",
    "                        accession = models.Accession(\n",
    "                            accession_prefix=accession_row[\"accession_prefix\"],\n",
    "                            name=accession_row[\"name\"],\n",
    "                            uuid=accession_row[\"uuid\"],\n",
    "                            see_also=accession_row[\"see_also\"],\n",
    "                        )\n",
    "                        accession.save()\n",
    "                        accessions.append(accession)\n",
    "                record.accession.set(accessions)\n",
    "            record.save()\n",
    "\n",
    "    assert not failed, \"Check warning messages.\"\n",
    "    \n",
    "load_tissues(strain_code_to_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "developing-contents",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "comprehensive-worker",
   "metadata": {},
   "outputs": [],
   "source": [
    "#submitted_tissues[row[mouse_tissue_id_label]], type(submitted_tissues[row[mouse_tissue_id_label]])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "historic-stand",
   "metadata": {},
   "source": [
    "# Information from Samples into experiment\n",
    "\n",
    "- Total barcoded nuclei (Samples into experiment)\n",
    "\n",
    "Is this FixedTissue, FixedSample, FixedBiosample?\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "applied-mustang",
   "metadata": {},
   "outputs": [],
   "source": [
    "if models.FixedSample.objects.count() > 0:\n",
    "    truncate(models.FixedSample)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "lonely-spelling",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_samples_into_experiment(book_name, sheet_name, header=0):\n",
    "    samples_into_experiment = pandas.read_excel(\n",
    "        book_name, \n",
    "        sheet_name=sheet_name,\n",
    "        header=header,\n",
    "        index_col=None,\n",
    "    ).dropna(how=\"all\")\n",
    "\n",
    "    # was trying to normalize two row hierarchical headers, \n",
    "    #level0 = []\n",
    "    #level1 = []\n",
    "    #for header in samples_into_experiment.columns:\n",
    "    #    level0.append(\"\" if header[0].startswith(\"Unnamed: \") else header[0])\n",
    "    #    level1.append(header[1])\n",
    "    #samples_into_experiment.columns = pandas.MultiIndex.from_arrays([level0, level1], names=[\"phase\", \"name\"])\n",
    "\n",
    "    one_based = 1\n",
    "    header_lines = 2\n",
    "    failed = False\n",
    "    for i, row in samples_into_experiment.iterrows():\n",
    "        line_no = i + one_based + header_lines\n",
    "        box_name = row[\"IGVF Fixation BOX\"]\n",
    "        if isinstance(box_name, str):\n",
    "            box_name = box_name.strip()\n",
    "\n",
    "        if box_name in [\"IGVF_FIX_001\", \"IGVF_FIX_002\"]:\n",
    "            continue\n",
    "            \n",
    "        if pandas.isnull(row[\"Mouse_Tissue ID\"]):\n",
    "            continue\n",
    "\n",
    "        tissue_id = normalize_strain(row[\"Mouse_Tissue ID\"])\n",
    "        pooled_from = row.get(\"pooled_from\")\n",
    "\n",
    "        if pandas.isnull(tissue_id):\n",
    "            continue\n",
    "\n",
    "        #if (tissue_id.endswith(\"_25\") or tissue_id.endswith(\"_26\")) and pandas.isnull(pooled_from):\n",
    "        #    print(f\"How are we going to merge these samples? {tissue_id}\")\n",
    "        #    continue\n",
    "\n",
    "        if pandas.isnull(pooled_from):\n",
    "            pooled_from = [tissue_id]\n",
    "        else:\n",
    "            pooled_from = [normalize_strain(pooled_id.strip()) for pooled_id in pooled_from.split(\",\")]\n",
    "\n",
    "        tissues = []\n",
    "        for pooled_id in pooled_from:\n",
    "            try:\n",
    "                tissues.append(models.Tissue.objects.get(name=pooled_id))\n",
    "            except models.Tissue.DoesNotExist:\n",
    "                print(f\"Tissue {pooled_id} not found in tissue table on {line_no}\")\n",
    "                failed = True\n",
    "\n",
    "        weight = row[\"weight (mg)\"]\n",
    "        fixation_date = row[\"Fixation date\"]\n",
    "        notes = row[\"Notes\"]\n",
    "\n",
    "        if weight == \"#VALUE!\":\n",
    "            pass\n",
    "        elif weight < 0:\n",
    "            pass\n",
    "        elif pandas.isnull(weight):\n",
    "            pass\n",
    "        else:\n",
    "            total_weight = 0\n",
    "            for tissue in tissues:\n",
    "                if not pandas.isnull(tissue.weight_mg):\n",
    "                    total_weight += tissue.weight_mg\n",
    "\n",
    "            if not numpy.isclose(total_weight, weight):\n",
    "                print(f\"Sum of tissue weights from {pooled_from} {total_weight} doesn't match {weight}. {line_no}\")\n",
    "\n",
    "        sample_id = normalize_strain(row[\"Sample ID\"]) if not pandas.isnull(row[\"Sample ID\"]) else None\n",
    "        if not (pandas.isnull(sample_id) or pandas.isnull(tissue_id)):\n",
    "            if tissue_id != sample_id:\n",
    "                print(f\"If defined, {tissue_id} should equal {sample_id} line {line_no}\")\n",
    "                failed = True\n",
    "\n",
    "        cap_label = row[\"Cap label\"]\n",
    "        if not (pandas.isnull(cap_label) or pandas.isnull(sample_id)):\n",
    "            sample_fields = sample_id.split(\"_\")\n",
    "            predicted_label = \"_\".join([sample_fields[0], sample_fields[-1]])\n",
    "            if predicted_label != cap_label:\n",
    "                print(f\"{cap_label} should equal {predicted_label} line {line_no}\")\n",
    "                failed = True\n",
    "\n",
    "        well_id = row[\"wells in Barcoding plate\"]\n",
    "\n",
    "        record = models.FixedSample(\n",
    "            name=tissue_id,\n",
    "            tube_label=cap_label,\n",
    "            fixation_name=box_name,\n",
    "            fixation_date=date_or_none(row[\"Fixation date\"]),\n",
    "            starting_nuclei=int_or_none(row[\"Total nuclei (x10^6)\"]),\n",
    "            nuclei_into_fixation=int_or_none(row[\"Nuclei into fixation (x10^6)\"]),\n",
    "            fixed_nuclei=int_or_none(row[\"Total fixed nuclei (x10^6)\"]),\n",
    "            aliquots_made=int_or_none(row[\"# aliquots\"]),\n",
    "            aliquot_volume_ul=float_or_none(row[\"uL per aliquot\"]),\n",
    "        )\n",
    "        record.save()\n",
    "        record.tissue.set(tissues)\n",
    "        record.save()\n",
    "\n",
    "    assert not failed, \"Check warning messages\"\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "central-oriental",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sum of tissue weights from ['046_NZOJ_10F_17'] 116.0 doesn't match 190.0. 558\n",
      "Sum of tissue weights from ['095_CASTJ_10M_08'] 11.0 doesn't match 15.0. 749\n"
     ]
    }
   ],
   "source": [
    "load_samples_into_experiment(lizs_sheet_name, \"Founder Samples into experiment\", header=1)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "automated-security",
   "metadata": {},
   "outputs": [],
   "source": [
    "load_samples_into_experiment(lizs_sheet_name, \"Bridge samples into experiment\", header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "durable-potter",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sum of tissue weights from ['127_B6AF1J_10M_05'] 1075.0 doesn't match 966.0. 205\n",
      "Sum of tissue weights from ['128_B6AF1J_10F_05'] 873.0 doesn't match 1078.0. 206\n",
      "Sum of tissue weights from ['129_B6AF1J_10M_05'] 1139.0 doesn't match 946.0. 207\n",
      "Sum of tissue weights from ['130_B6AF1J_10F_05'] 66.1 doesn't match 1577.0. 208\n",
      "Sum of tissue weights from ['131_B6AF1J_10M_05'] 1525.0 doesn't match 748.0. 209\n",
      "Sum of tissue weights from ['132_B6AF1J_10F_05'] 888.0 doesn't match 1418.0. 210\n",
      "Sum of tissue weights from ['133_B6AF1J_10M_05'] 0 doesn't match 857.0. 211\n",
      "Sum of tissue weights from ['134_B6AF1J_10F_05'] 0 doesn't match 1351.0. 212\n",
      "Sum of tissue weights from ['143_B6129SF1J_10M_05'] 1288.0 doesn't match 603.0. 213\n",
      "Sum of tissue weights from ['144_B6129SF1J_10F_05'] 995.0 doesn't match 1248.0. 214\n",
      "Sum of tissue weights from ['145_B6129SF1J_10M_05'] 1259.0 doesn't match 776.0. 215\n",
      "Sum of tissue weights from ['146_B6129SF1J_10F_05'] 974.0 doesn't match 1216.0. 216\n",
      "Sum of tissue weights from ['147_B6129SF1J_10M_05'] 1368.0 doesn't match 798.0. 217\n",
      "Sum of tissue weights from ['148_B6129SF1J_10F_05'] 977.0 doesn't match 1243.0. 218\n",
      "Sum of tissue weights from ['149_B6129SF1J_10M_05'] 1385.0 doesn't match 800.0. 219\n",
      "Sum of tissue weights from ['150_B6129SF1J_10F_05'] 870.0 doesn't match 1343.0. 220\n",
      "Sum of tissue weights from ['159_B6NODF1J_10M_05'] 1343.0 doesn't match 1189.0. 221\n",
      "Sum of tissue weights from ['160_B6NODF1J_10F_05'] 1247.0 doesn't match 828.0. 222\n",
      "Sum of tissue weights from ['161_B6NODF1J_10M_05'] 1238.0 doesn't match 1071.0. 223\n",
      "Sum of tissue weights from ['164_B6NODF1J_10F_05'] 861.0 doesn't match 626.0. 224\n",
      "Sum of tissue weights from ['165_B6NODF1J_10M_05'] 1237.0 doesn't match 1039.0. 225\n",
      "Sum of tissue weights from ['166_B6NODF1J_10F_05'] 1053.0 doesn't match 1065.0. 226\n",
      "Sum of tissue weights from ['167_B6NODF1J_10M_05'] 1343.0 doesn't match 1189.0. 227\n",
      "Sum of tissue weights from ['168_B6NODF1J_10F_05'] 1033.0 doesn't match 1157.0. 228\n",
      "Sum of tissue weights from ['175_B6NZOF1J_10M_05'] 1941.0 doesn't match 721.0. 229\n",
      "Sum of tissue weights from ['176_B6NZOF1J_10F_05'] 1142.0 doesn't match 984.0. 230\n",
      "Sum of tissue weights from ['177_B6NZOF1J_10M_05'] 1861.0 doesn't match 846.0. 231\n",
      "Sum of tissue weights from ['178_B6NZOF1J_10F_05'] 1467.0 doesn't match 707.0. 232\n",
      "Sum of tissue weights from ['181_B6NZOF1J_10M_05'] 1774.0 doesn't match 702.0. 233\n",
      "Sum of tissue weights from ['180_B6NZOF1J_10F_05'] 1528.0 doesn't match 841.0. 234\n",
      "Sum of tissue weights from ['183_B6NZOF1J_10M_05'] 1545.0 doesn't match 782.0. 235\n",
      "Sum of tissue weights from ['182_B6NZOF1J_10F_05'] 1600.0 doesn't match 927.0. 236\n",
      "Sum of tissue weights from ['191_B6CASTF1J_10M_05'] 1765.0 doesn't match 610.0. 237\n",
      "Sum of tissue weights from ['192_B6CASTF1J_10F_05'] 882.0 doesn't match 1005.0. 238\n",
      "Sum of tissue weights from ['193_B6CASTF1J_10M_05'] 1214.0 doesn't match 786.0. 239\n",
      "Sum of tissue weights from ['194_B6CASTF1J_10F_05'] 816.0 doesn't match 1050.0. 240\n",
      "Sum of tissue weights from ['195_B6CASTF1J_10M_05'] 1306.0 doesn't match 836.0. 241\n",
      "Sum of tissue weights from ['196_B6CASTF1J_10F_05'] 896.0 doesn't match 1123.0. 242\n",
      "Sum of tissue weights from ['197_B6CASTF1J_10M_05'] 1615.0 doesn't match 863.0. 243\n",
      "Sum of tissue weights from ['198_B6CASTF1J_10F_05'] 887.0 doesn't match 934.0. 244\n",
      "Sum of tissue weights from ['207_B6PWKF1J_10M_05'] 0 doesn't match 439.0. 245\n",
      "Sum of tissue weights from ['208_B6PWKF1J_10F_05'] 0 doesn't match 635.0. 246\n",
      "Sum of tissue weights from ['213_B6PWKF1J_10M_05'] 1235.0 doesn't match 411.0. 247\n",
      "Sum of tissue weights from ['210_B6PWKF1J_10F_05'] 0 doesn't match 638.0. 248\n",
      "Sum of tissue weights from ['215_B6PWKF1J_10M_05'] 1094.0 doesn't match 474.0. 249\n",
      "Sum of tissue weights from ['212_B6PWKF1J_10F_05'] 0 doesn't match 691.0. 250\n",
      "Sum of tissue weights from ['217_B6PWKF1J_10M_05'] 1236.0 doesn't match 520.0. 251\n",
      "Sum of tissue weights from ['214_B6PWKF1J_10F_05'] 984.0 doesn't match 658.0. 252\n",
      "Sum of tissue weights from ['223_B6WSBF1J_10M_05'] 1141.0 doesn't match 549.0. 253\n",
      "Sum of tissue weights from ['228_B6WSBF1J_10F_05'] 1743.0 doesn't match 651.0. 254\n",
      "Sum of tissue weights from ['225_B6WSBF1J_10M_05'] 991.0 doesn't match 522.0. 255\n",
      "Sum of tissue weights from ['232_B6WSBF1J_10F_05'] 977.0 doesn't match 691.0. 256\n",
      "Sum of tissue weights from ['227_B6WSBF1J_10M_05'] 950.0 doesn't match 656.0. 257\n",
      "Sum of tissue weights from ['234_B6WSBF1J_10F_05'] 0 doesn't match 682.0. 258\n",
      "Sum of tissue weights from ['229_B6WSBF1J_10M_05'] 962.0 doesn't match 482.0. 259\n",
      "Sum of tissue weights from ['236_B6WSBF1J_10F_05'] 0 doesn't match 607.0. 260\n",
      "Sum of tissue weights from ['239_TREM2R47HNSS_HO_10M_05'] 1262.0 doesn't match 2017.0. 261\n",
      "Sum of tissue weights from ['240_TREM2R47HNSS_HO_10F_05'] 1039.0 doesn't match 2064.0. 262\n",
      "Sum of tissue weights from ['241_TREM2R47HNSS_HO_10M_05'] 1176.0 doesn't match 2032.0. 263\n",
      "Sum of tissue weights from ['242_TREM2R47HNSS_HO_10F_05'] 1134.0 doesn't match 2001.0. 264\n",
      "Sum of tissue weights from ['243_TREM2R47HNSS_HO_10M_05'] 1128.0 doesn't match 1968.0. 265\n",
      "Sum of tissue weights from ['244_TREM2R47HNSS_HO_10F_05'] 1138.0 doesn't match 2157.0. 266\n",
      "Sum of tissue weights from ['245_TREM2R47HNSS_HO_10M_05'] 1311.0 doesn't match 1810.0. 267\n",
      "Sum of tissue weights from ['248_TREM2R47HNSS_HO_10F_05'] 981.0 doesn't match 1963.0. 268\n",
      "Sum of tissue weights from ['127_B6AF1J_10M_01'] 82.0 doesn't match 72.0. 305\n",
      "Sum of tissue weights from ['128_B6AF1J_10F_01'] 81.0 doesn't match 85.0. 306\n",
      "Sum of tissue weights from ['129_B6AF1J_10M_01'] 77.0 doesn't match 63.0. 307\n",
      "Sum of tissue weights from ['130_B6AF1J_10F_01'] 79.0 doesn't match 63.0. 308\n",
      "Sum of tissue weights from ['131_B6AF1J_10M_01'] 78.0 doesn't match 71.0. 309\n",
      "Sum of tissue weights from ['132_B6AF1J_10F_01'] 76.0 doesn't match 78.0. 310\n",
      "Sum of tissue weights from ['133_B6AF1J_10M_01'] 0 doesn't match 62.0. 311\n",
      "Sum of tissue weights from ['134_B6AF1J_10F_01'] 0 doesn't match 61.0. 312\n",
      "Sum of tissue weights from ['143_B6129SF1J_10M_01'] 99.0 doesn't match 68.0. 313\n",
      "Sum of tissue weights from ['144_B6129SF1J_10F_01'] 89.0 doesn't match 57.0. 314\n",
      "Sum of tissue weights from ['145_B6129SF1J_10M_01'] 88.0 doesn't match 69.0. 315\n",
      "Sum of tissue weights from ['146_B6129SF1J_10F_01'] 89.0 doesn't match 56.0. 316\n",
      "Sum of tissue weights from ['147_B6129SF1J_10M_01'] 92.0 doesn't match 51.0. 317\n",
      "Sum of tissue weights from ['148_B6129SF1J_10F_01'] 83.0 doesn't match 72.0. 318\n",
      "Sum of tissue weights from ['149_B6129SF1J_10M_01'] 83.0 doesn't match 52.0. 319\n",
      "Sum of tissue weights from ['150_B6129SF1J_10F_01'] 84.0 doesn't match 82.0. 320\n",
      "Sum of tissue weights from ['159_B6NODF1J_10M_01'] 96.0 doesn't match 42.0. 321\n",
      "Sum of tissue weights from ['160_B6NODF1J_10F_01'] 103.0 doesn't match 73.0. 322\n",
      "Sum of tissue weights from ['161_B6NODF1J_10M_01'] 107.0 doesn't match 51.0. 323\n",
      "Sum of tissue weights from ['164_B6NODF1J_10F_01'] 93.0 doesn't match 59.0. 324\n",
      "Sum of tissue weights from ['165_B6NODF1J_10M_01'] 79.0 doesn't match 55.0. 325\n",
      "Sum of tissue weights from ['166_B6NODF1J_10F_01'] 99.0 doesn't match 57.0. 326\n",
      "Sum of tissue weights from ['167_B6NODF1J_10M_01'] 105.0 doesn't match 65.0. 327\n",
      "Sum of tissue weights from ['168_B6NODF1J_10F_01'] 93.0 doesn't match 80.0. 328\n",
      "Sum of tissue weights from ['175_B6NZOF1J_10M_01'] 108.0 doesn't match 55.0. 329\n",
      "Sum of tissue weights from ['176_B6NZOF1J_10F_01'] 101.0 doesn't match 54.0. 330\n",
      "Sum of tissue weights from ['177_B6NZOF1J_10M_01'] 101.0 doesn't match 58.0. 331\n",
      "Sum of tissue weights from ['178_B6NZOF1J_10F_01'] 99.0 doesn't match 50.0. 332\n",
      "Sum of tissue weights from ['181_B6NZOF1J_10M_01'] 92.0 doesn't match 56.0. 333\n",
      "Sum of tissue weights from ['180_B6NZOF1J_10F_01'] 99.0 doesn't match 50.0. 334\n",
      "Sum of tissue weights from ['183_B6NZOF1J_10M_01'] 103.0 doesn't match 56.0. 335\n",
      "Sum of tissue weights from ['182_B6NZOF1J_10F_01'] 103.0 doesn't match 50.0. 336\n",
      "Sum of tissue weights from ['191_B6CASTF1J_10M_01'] 130.0 doesn't match 66.0. 337\n",
      "Sum of tissue weights from ['192_B6CASTF1J_10F_01'] 72.0 doesn't match 57.0. 338\n",
      "Sum of tissue weights from ['193_B6CASTF1J_10M_01'] 116.0 doesn't match 75.0. 339\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sum of tissue weights from ['194_B6CASTF1J_10F_01'] 90.0 doesn't match 63.0. 340\n",
      "Sum of tissue weights from ['195_B6CASTF1J_10M_01'] 87.0 doesn't match 66.0. 341\n",
      "Sum of tissue weights from ['196_B6CASTF1J_10F_01'] 86.0 doesn't match 82.0. 342\n",
      "Sum of tissue weights from ['197_B6CASTF1J_10M_01'] 98.0 doesn't match 54.0. 343\n",
      "Sum of tissue weights from ['198_B6CASTF1J_10F_01'] 84.0 doesn't match 60.0. 344\n",
      "Sum of tissue weights from ['207_B6PWKF1J_10M_01'] 0 doesn't match 47.0. 345\n",
      "Sum of tissue weights from ['208_B6PWKF1J_10F_01'] 0 doesn't match 52.0. 346\n",
      "Sum of tissue weights from ['213_B6PWKF1J_10M_01'] 93.0 doesn't match 66.0. 347\n",
      "Sum of tissue weights from ['210_B6PWKF1J_10F_01'] 0 doesn't match 57.0. 348\n",
      "Sum of tissue weights from ['215_B6PWKF1J_10M_01'] 96.0 doesn't match 53.0. 349\n",
      "Sum of tissue weights from ['212_B6PWKF1J_10F_01'] 0 doesn't match 86.0. 350\n",
      "Sum of tissue weights from ['217_B6PWKF1J_10M_01'] 107.0 doesn't match 53.0. 351\n",
      "Sum of tissue weights from ['214_B6PWKF1J_10F_01'] 109.0 doesn't match 63.0. 352\n",
      "Sum of tissue weights from ['223_B6WSBF1J_10M_01'] 119.0 doesn't match 52.0. 353\n",
      "Sum of tissue weights from ['228_B6WSBF1J_10F_01'] 122.0 doesn't match 61.0. 354\n",
      "Sum of tissue weights from ['225_B6WSBF1J_10M_01'] 118.0 doesn't match 52.0. 355\n",
      "Sum of tissue weights from ['232_B6WSBF1J_10F_01'] 118.0 doesn't match 61.0. 356\n",
      "Sum of tissue weights from ['227_B6WSBF1J_10M_01'] 115.0 doesn't match 60.0. 357\n",
      "Sum of tissue weights from ['234_B6WSBF1J_10F_01'] 0 doesn't match 54.0. 358\n",
      "Sum of tissue weights from ['229_B6WSBF1J_10M_01'] 101.0 doesn't match 49.0. 359\n",
      "Sum of tissue weights from ['236_B6WSBF1J_10F_01'] 0 doesn't match 68.0. 360\n",
      "Sum of tissue weights from ['239_TREM2R47HNSS_HO_10M_01'] 121.0 doesn't match 42.0. 361\n",
      "Sum of tissue weights from ['240_TREM2R47HNSS_HO_10F_01'] 124.0 doesn't match 58.0. 362\n",
      "Sum of tissue weights from ['241_TREM2R47HNSS_HO_10M_01'] 112.0 doesn't match 54.0. 363\n",
      "Sum of tissue weights from ['242_TREM2R47HNSS_HO_10F_01'] 107.0 doesn't match 49.0. 364\n",
      "Sum of tissue weights from ['243_TREM2R47HNSS_HO_10M_01'] 117.0 doesn't match 64.0. 365\n",
      "Sum of tissue weights from ['244_TREM2R47HNSS_HO_10F_01'] 122.0 doesn't match 48.0. 366\n",
      "Sum of tissue weights from ['245_TREM2R47HNSS_HO_10M_01'] 100.0 doesn't match 56.0. 367\n",
      "Sum of tissue weights from ['248_TREM2R47HNSS_HO_10F_01'] 111.0 doesn't match 59.0. 368\n",
      "Sum of tissue weights from ['127_B6AF1J_10M_27', '127_B6AF1J_10M_29'] 0 doesn't match 33.0. 405\n",
      "Sum of tissue weights from ['128_B6AF1J_10F_10', '128_B6AF1J_10F_11'] 0 doesn't match 263.0. 406\n",
      "Sum of tissue weights from ['129_B6AF1J_10M_27', '129_B6AF1J_10M_29'] 0 doesn't match 42.0. 407\n",
      "Sum of tissue weights from ['130_B6AF1J_10F_10', '130_B6AF1J_10F_11'] 0 doesn't match 280.0. 408\n",
      "Sum of tissue weights from ['131_B6AF1J_10M_27', '131_B6AF1J_10M_29'] 0 doesn't match 40.0. 409\n",
      "Sum of tissue weights from ['132_B6AF1J_10F_10', '132_B6AF1J_10F_11'] 0 doesn't match 334.0. 410\n",
      "Sum of tissue weights from ['133_B6AF1J_10M_27', '133_B6AF1J_10M_29'] 0 doesn't match 30.0. 411\n",
      "Sum of tissue weights from ['134_B6AF1J_10F_10', '134_B6AF1J_10F_11'] 0 doesn't match 320.0. 412\n",
      "Sum of tissue weights from ['143_B6129SF1J_10M_27', '143_B6129SF1J_10M_29'] 0 doesn't match 30.0. 413\n",
      "Sum of tissue weights from ['144_B6129SF1J_10F_10', '144_B6129SF1J_10F_11'] 0 doesn't match 336.0. 414\n",
      "Sum of tissue weights from ['145_B6129SF1J_10M_27', '145_B6129SF1J_10M_29'] 0 doesn't match 28.0. 415\n",
      "Sum of tissue weights from ['146_B6129SF1J_10F_10', '146_B6129SF1J_10F_11'] 0 doesn't match 324.0. 416\n",
      "Sum of tissue weights from ['147_B6129SF1J_10M_27', '147_B6129SF1J_10M_29'] 0 doesn't match 26.0. 417\n",
      "Sum of tissue weights from ['148_B6129SF1J_10F_10', '148_B6129SF1J_10F_11'] 0 doesn't match 342.0. 418\n",
      "Sum of tissue weights from ['149_B6129SF1J_10M_27', '149_B6129SF1J_10M_29'] 0 doesn't match 31.0. 419\n",
      "Sum of tissue weights from ['150_B6129SF1J_10F_10', '150_B6129SF1J_10F_11'] 0 doesn't match 332.0. 420\n",
      "Sum of tissue weights from ['159_B6NODF1J_10M_27', '159_B6NODF1J_10M_29'] 0 doesn't match 41.0. 421\n",
      "Sum of tissue weights from ['160_B6NODF1J_10F_10', '160_B6NODF1J_10F_11'] 0 doesn't match 231.0. 422\n",
      "Sum of tissue weights from ['161_B6NODF1J_10M_27', '161_B6NODF1J_10M_29'] 0 doesn't match 39.0. 423\n",
      "Sum of tissue weights from ['164_B6NODF1J_10F_10', '164_B6NODF1J_10F_11'] 0 doesn't match 216.0. 424\n",
      "Sum of tissue weights from ['165_B6NODF1J_10M_27', '165_B6NODF1J_10M_29'] 0 doesn't match 35.0. 425\n",
      "Sum of tissue weights from ['166_B6NODF1J_10F_10', '166_B6NODF1J_10F_11'] 0 doesn't match 218.0. 426\n",
      "Sum of tissue weights from ['167_B6NODF1J_10M_27', '167_B6NODF1J_10M_29'] 0 doesn't match 31.0. 427\n",
      "Sum of tissue weights from ['168_B6NODF1J_10F_10', '168_B6NODF1J_10F_11'] 0 doesn't match 264.0. 428\n",
      "Sum of tissue weights from ['175_B6NZOF1J_10M_27', '175_B6NZOF1J_10M_29'] 0 doesn't match 24.0. 429\n",
      "Sum of tissue weights from ['176_B6NZOF1J_10F_10', '176_B6NZOF1J_10F_11'] 0 doesn't match 249.0. 430\n",
      "Sum of tissue weights from ['177_B6NZOF1J_10M_27', '177_B6NZOF1J_10M_29'] 0 doesn't match 28.0. 431\n",
      "Sum of tissue weights from ['178_B6NZOF1J_10F_10', '178_B6NZOF1J_10F_11'] 0 doesn't match 208.0. 432\n",
      "Sum of tissue weights from ['181_B6NZOF1J_10M_27', '181_B6NZOF1J_10M_29'] 0 doesn't match 33.0. 433\n",
      "Sum of tissue weights from ['180_B6NZOF1J_10F_10', '180_B6NZOF1J_10F_11'] 0 doesn't match 242.0. 434\n",
      "Sum of tissue weights from ['183_B6NZOF1J_10M_27', '183_B6NZOF1J_10M_29'] 0 doesn't match 19.0. 435\n",
      "Sum of tissue weights from ['182_B6NZOF1J_10F_10', '182_B6NZOF1J_10F_11'] 0 doesn't match 266.0. 436\n",
      "Sum of tissue weights from ['191_B6CASTF1J_10M_27', '191_B6CASTF1J_10M_29'] 0 doesn't match 27.0. 437\n",
      "Sum of tissue weights from ['192_B6CASTF1J_10F_10', '192_B6CASTF1J_10F_11'] 0 doesn't match 282.0. 438\n",
      "Sum of tissue weights from ['193_B6CASTF1J_10M_27', '193_B6CASTF1J_10M_29'] 0 doesn't match 36.0. 439\n",
      "Sum of tissue weights from ['194_B6CASTF1J_10F_10', '194_B6CASTF1J_10F_11'] 0 doesn't match 316.0. 440\n",
      "Sum of tissue weights from ['195_B6CASTF1J_10M_27', '195_B6CASTF1J_10M_29'] 0 doesn't match 34.0. 441\n",
      "Sum of tissue weights from ['196_B6CASTF1J_10F_10', '196_B6CASTF1J_10F_11'] 0 doesn't match 285.0. 442\n",
      "Sum of tissue weights from ['197_B6CASTF1J_10M_27', '197_B6CASTF1J_10M_29'] 0 doesn't match 35.0. 443\n",
      "Sum of tissue weights from ['198_B6CASTF1J_10F_10', '198_B6CASTF1J_10F_11'] 0 doesn't match 315.0. 444\n",
      "Sum of tissue weights from ['207_B6PWKF1J_10M_27', '207_B6PWKF1J_10M_29'] 0 doesn't match 28.0. 445\n",
      "Sum of tissue weights from ['208_B6PWKF1J_10F_10', '208_B6PWKF1J_10F_11'] 0 doesn't match 171.0. 446\n",
      "Sum of tissue weights from ['213_B6PWKF1J_10M_27', '213_B6PWKF1J_10M_29'] 0 doesn't match 28.0. 447\n",
      "Sum of tissue weights from ['210_B6PWKF1J_10F_10', '210_B6PWKF1J_10F_11'] 0 doesn't match 166.0. 448\n",
      "Sum of tissue weights from ['215_B6PWKF1J_10M_27', '215_B6PWKF1J_10M_29'] 0 doesn't match 24.0. 449\n",
      "Sum of tissue weights from ['212_B6PWKF1J_10F_10', '212_B6PWKF1J_10F_11'] 0 doesn't match 245.0. 450\n",
      "Sum of tissue weights from ['217_B6PWKF1J_10M_27', '217_B6PWKF1J_10M_29'] 0 doesn't match 23.0. 451\n",
      "Sum of tissue weights from ['214_B6PWKF1J_10F_10', '214_B6PWKF1J_10F_11'] 0 doesn't match 178.0. 452\n",
      "Sum of tissue weights from ['223_B6WSBF1J_10M_27', '223_B6WSBF1J_10M_29'] 0 doesn't match 11.0. 453\n",
      "Sum of tissue weights from ['228_B6WSBF1J_10F_10', '228_B6WSBF1J_10F_11'] 0 doesn't match 168.0. 454\n",
      "Sum of tissue weights from ['225_B6WSBF1J_10M_27', '225_B6WSBF1J_10M_29'] 0 doesn't match 22.0. 455\n",
      "Sum of tissue weights from ['232_B6WSBF1J_10F_10', '232_B6WSBF1J_10F_11'] 0 doesn't match 246.0. 456\n",
      "Sum of tissue weights from ['227_B6WSBF1J_10M_27', '227_B6WSBF1J_10M_29'] 0 doesn't match 19.0. 457\n",
      "Sum of tissue weights from ['234_B6WSBF1J_10F_10', '234_B6WSBF1J_10F_11'] 0 doesn't match 213.0. 458\n",
      "Sum of tissue weights from ['229_B6WSBF1J_10M_27', '229_B6WSBF1J_10M_29'] 0 doesn't match 20.0. 459\n",
      "Sum of tissue weights from ['236_B6WSBF1J_10F_10', '236_B6WSBF1J_10F_11'] 0 doesn't match 170.0. 460\n",
      "Sum of tissue weights from ['239_TREM2R47HNSS_HO_10M_27', '239_TREM2R47HNSS_HO_10M_29'] 0 doesn't match 44.0. 461\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sum of tissue weights from ['240_TREM2R47HNSS_HO_10F_10', '240_TREM2R47HNSS_HO_10F_11'] 0 doesn't match 334.0. 462\n",
      "Sum of tissue weights from ['241_TREM2R47HNSS_HO_10M_27', '241_TREM2R47HNSS_HO_10M_29'] 0 doesn't match 43.0. 463\n",
      "Sum of tissue weights from ['242_TREM2R47HNSS_HO_10F_10', '242_TREM2R47HNSS_HO_10F_11'] 0 doesn't match 365.0. 464\n",
      "Sum of tissue weights from ['243_TREM2R47HNSS_HO_10M_27', '243_TREM2R47HNSS_HO_10M_29'] 0 doesn't match 30.0. 465\n",
      "Sum of tissue weights from ['244_TREM2R47HNSS_HO_10F_10', '244_TREM2R47HNSS_HO_10F_11'] 0 doesn't match 341.0. 466\n",
      "Sum of tissue weights from ['245_TREM2R47HNSS_HO_10M_27', '245_TREM2R47HNSS_HO_10M_29'] 0 doesn't match 40.0. 467\n",
      "Sum of tissue weights from ['248_TREM2R47HNSS_HO_10F_10', '248_TREM2R47HNSS_HO_10F_11'] 0 doesn't match 370.0. 468\n",
      "Sum of tissue weights from ['127_B6AF1J_10M_08'] 0 doesn't match 15.0. 505\n",
      "Sum of tissue weights from ['128_B6AF1J_10F_08'] 0 doesn't match 13.0. 506\n",
      "Sum of tissue weights from ['129_B6AF1J_10M_08'] 0 doesn't match 17.0. 507\n",
      "Sum of tissue weights from ['130_B6AF1J_10F_08'] 0 doesn't match 14.0. 508\n",
      "Sum of tissue weights from ['131_B6AF1J_10M_08'] 0 doesn't match 19.0. 509\n",
      "Sum of tissue weights from ['132_B6AF1J_10F_08'] 0 doesn't match 13.0. 510\n",
      "Sum of tissue weights from ['133_B6AF1J_10M_08'] 0 doesn't match 12.0. 511\n",
      "Sum of tissue weights from ['134_B6AF1J_10F_08'] 0 doesn't match 23.0. 512\n",
      "Sum of tissue weights from ['143_B6129SF1J_10M_08'] 0 doesn't match 19.0. 513\n",
      "Sum of tissue weights from ['144_B6129SF1J_10F_08'] 0 doesn't match 14.0. 514\n",
      "Sum of tissue weights from ['145_B6129SF1J_10M_08'] 0 doesn't match 20.0. 515\n",
      "Sum of tissue weights from ['146_B6129SF1J_10F_08'] 0 doesn't match 18.0. 516\n",
      "Sum of tissue weights from ['147_B6129SF1J_10M_08'] 0 doesn't match 18.0. 517\n",
      "Sum of tissue weights from ['148_B6129SF1J_10F_08'] 0 doesn't match 19.0. 518\n",
      "Sum of tissue weights from ['149_B6129SF1J_10M_08'] 0 doesn't match 17.0. 519\n",
      "Sum of tissue weights from ['150_B6129SF1J_10F_08'] 0 doesn't match 16.0. 520\n",
      "Sum of tissue weights from ['159_B6NODF1J_10M_08'] 0 doesn't match 18.0. 521\n",
      "Sum of tissue weights from ['160_B6NODF1J_10F_08'] 0 doesn't match 21.0. 522\n",
      "Sum of tissue weights from ['161_B6NODF1J_10M_08'] 0 doesn't match 27.0. 523\n",
      "Sum of tissue weights from ['164_B6NODF1J_10F_08'] 0 doesn't match 16.0. 524\n",
      "Sum of tissue weights from ['165_B6NODF1J_10M_08'] 0 doesn't match 23.0. 525\n",
      "Sum of tissue weights from ['166_B6NODF1J_10F_08'] 0 doesn't match 14.0. 526\n",
      "Sum of tissue weights from ['167_B6NODF1J_10M_08'] 0 doesn't match 16.0. 527\n",
      "Sum of tissue weights from ['168_B6NODF1J_10F_08'] 0 doesn't match 18.0. 528\n",
      "Sum of tissue weights from ['175_B6NZOF1J_10M_08'] 0 doesn't match 15.0. 529\n",
      "Sum of tissue weights from ['176_B6NZOF1J_10F_08'] 0 doesn't match 13.0. 530\n",
      "Sum of tissue weights from ['177_B6NZOF1J_10M_08'] 0 doesn't match 14.0. 531\n",
      "Sum of tissue weights from ['178_B6NZOF1J_10F_08'] 0 doesn't match 16.0. 532\n",
      "Sum of tissue weights from ['181_B6NZOF1J_10M_08'] 0 doesn't match 18.0. 533\n",
      "Sum of tissue weights from ['180_B6NZOF1J_10F_08'] 0 doesn't match 17.0. 534\n",
      "Sum of tissue weights from ['183_B6NZOF1J_10M_08'] 0 doesn't match 17.0. 535\n",
      "Sum of tissue weights from ['182_B6NZOF1J_10F_08'] 0 doesn't match 13.0. 536\n",
      "Sum of tissue weights from ['191_B6CASTF1J_10M_08'] 0 doesn't match 19.0. 537\n",
      "Sum of tissue weights from ['192_B6CASTF1J_10F_08'] 0 doesn't match 19.0. 538\n",
      "Sum of tissue weights from ['193_B6CASTF1J_10M_08'] 0 doesn't match 14.0. 539\n",
      "Sum of tissue weights from ['194_B6CASTF1J_10F_08'] 0 doesn't match 16.0. 540\n",
      "Sum of tissue weights from ['195_B6CASTF1J_10M_08'] 0 doesn't match 20.0. 541\n",
      "Sum of tissue weights from ['196_B6CASTF1J_10F_08'] 0 doesn't match 16.0. 542\n",
      "Sum of tissue weights from ['197_B6CASTF1J_10M_08'] 0 doesn't match 21.0. 543\n",
      "Sum of tissue weights from ['198_B6CASTF1J_10F_08'] 0 doesn't match 19.0. 544\n",
      "Sum of tissue weights from ['207_B6PWKF1J_10M_08'] 0 doesn't match 24.0. 545\n",
      "Sum of tissue weights from ['208_B6PWKF1J_10F_08'] 0 doesn't match 13.0. 546\n",
      "Sum of tissue weights from ['213_B6PWKF1J_10M_08'] 0 doesn't match 19.0. 547\n",
      "Sum of tissue weights from ['210_B6PWKF1J_10F_08'] 0 doesn't match 20.0. 548\n",
      "Sum of tissue weights from ['215_B6PWKF1J_10M_08'] 0 doesn't match 17.0. 549\n",
      "Sum of tissue weights from ['212_B6PWKF1J_10F_08'] 0 doesn't match 43.0. 550\n",
      "Sum of tissue weights from ['217_B6PWKF1J_10M_08'] 0 doesn't match 17.0. 551\n",
      "Sum of tissue weights from ['214_B6PWKF1J_10F_08'] 0 doesn't match 15.0. 552\n",
      "Sum of tissue weights from ['223_B6WSBF1J_10M_08'] 0 doesn't match 13.0. 553\n",
      "Sum of tissue weights from ['228_B6WSBF1J_10F_08'] 0 doesn't match 17.0. 554\n",
      "Sum of tissue weights from ['225_B6WSBF1J_10M_08'] 0 doesn't match 20.0. 555\n",
      "Sum of tissue weights from ['232_B6WSBF1J_10F_08'] 0 doesn't match 23.0. 556\n",
      "Sum of tissue weights from ['227_B6WSBF1J_10M_08'] 0 doesn't match 24.0. 557\n",
      "Sum of tissue weights from ['234_B6WSBF1J_10F_08'] 0 doesn't match 21.0. 558\n",
      "Sum of tissue weights from ['229_B6WSBF1J_10M_08'] 0 doesn't match 12.0. 559\n",
      "Sum of tissue weights from ['236_B6WSBF1J_10F_08'] 0 doesn't match 23.0. 560\n",
      "Sum of tissue weights from ['239_TREM2R47HNSS_HO_10M_08'] 0 doesn't match 24.0. 561\n",
      "Sum of tissue weights from ['240_TREM2R47HNSS_HO_10F_08'] 0 doesn't match 15.0. 562\n",
      "Sum of tissue weights from ['241_TREM2R47HNSS_HO_10M_08'] 0 doesn't match 33.0. 563\n",
      "Sum of tissue weights from ['242_TREM2R47HNSS_HO_10F_08'] 0 doesn't match 20.0. 564\n",
      "Sum of tissue weights from ['243_TREM2R47HNSS_HO_10M_08'] 0 doesn't match 35.0. 565\n",
      "Sum of tissue weights from ['244_TREM2R47HNSS_HO_10F_08'] 0 doesn't match 20.0. 566\n",
      "Sum of tissue weights from ['245_TREM2R47HNSS_HO_10M_08'] 0 doesn't match 27.0. 567\n",
      "Sum of tissue weights from ['248_TREM2R47HNSS_HO_10F_08'] 0 doesn't match 21.0. 568\n",
      "Sum of tissue weights from ['127_B6AF1J_10M_31'] 0 doesn't match 301.0. 605\n",
      "Sum of tissue weights from ['128_B6AF1J_10F_31'] 0 doesn't match 431.0. 606\n",
      "Sum of tissue weights from ['129_B6AF1J_10M_31'] 0 doesn't match 287.0. 607\n",
      "Sum of tissue weights from ['130_B6AF1J_10F_31'] 0 doesn't match 390.0. 608\n",
      "Sum of tissue weights from ['131_B6AF1J_10M_31'] 0 doesn't match 266.0. 609\n",
      "Sum of tissue weights from ['132_B6AF1J_10F_31'] 0 doesn't match 381.0. 610\n",
      "Sum of tissue weights from ['133_B6AF1J_10M_31'] 0 doesn't match 277.0. 611\n",
      "Sum of tissue weights from ['134_B6AF1J_10F_31'] 0 doesn't match 401.0. 612\n",
      "Sum of tissue weights from ['143_B6129SF1J_10M_31'] 0 doesn't match 234.0. 613\n",
      "Sum of tissue weights from ['144_B6129SF1J_10F_31'] 0 doesn't match 465.0. 614\n",
      "Sum of tissue weights from ['145_B6129SF1J_10M_31'] 0 doesn't match 269.0. 615\n",
      "Sum of tissue weights from ['146_B6129SF1J_10F_31'] 0 doesn't match 435.0. 616\n",
      "Sum of tissue weights from ['147_B6129SF1J_10M_31'] 0 doesn't match 264.0. 617\n",
      "Sum of tissue weights from ['148_B6129SF1J_10F_31'] 0 doesn't match 464.0. 618\n",
      "Sum of tissue weights from ['149_B6129SF1J_10M_31'] 0 doesn't match 267.0. 619\n",
      "Sum of tissue weights from ['150_B6129SF1J_10F_31'] 0 doesn't match 439.0. 620\n",
      "Sum of tissue weights from ['159_B6NODF1J_10M_31'] 0 doesn't match 320.0. 621\n",
      "Sum of tissue weights from ['160_B6NODF1J_10F_31'] 0 doesn't match 323.0. 622\n",
      "Sum of tissue weights from ['161_B6NODF1J_10M_31'] 0 doesn't match 299.0. 623\n",
      "Sum of tissue weights from ['164_B6NODF1J_10F_31'] 0 doesn't match 240.0. 624\n",
      "Sum of tissue weights from ['165_B6NODF1J_10M_31'] 0 doesn't match 314.0. 625\n",
      "Sum of tissue weights from ['166_B6NODF1J_10F_31'] 0 doesn't match 382.0. 626\n",
      "Sum of tissue weights from ['167_B6NODF1J_10M_31'] 0 doesn't match 320.0. 627\n",
      "Sum of tissue weights from ['168_B6NODF1J_10F_31'] 0 doesn't match 396.0. 628\n",
      "Sum of tissue weights from ['175_B6NZOF1J_10M_31'] 0 doesn't match 231.0. 629\n",
      "Sum of tissue weights from ['176_B6NZOF1J_10F_31'] 0 doesn't match 291.0. 630\n",
      "Sum of tissue weights from ['177_B6NZOF1J_10M_31'] 0 doesn't match 234.0. 631\n",
      "Sum of tissue weights from ['178_B6NZOF1J_10F_31'] 0 doesn't match 246.0. 632\n",
      "Sum of tissue weights from ['181_B6NZOF1J_10M_31'] 0 doesn't match 225.0. 633\n",
      "Sum of tissue weights from ['180_B6NZOF1J_10F_31'] 0 doesn't match 271.0. 634\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sum of tissue weights from ['183_B6NZOF1J_10M_31'] 0 doesn't match 240.0. 635\n",
      "Sum of tissue weights from ['182_B6NZOF1J_10F_31'] 0 doesn't match 326.0. 636\n",
      "Sum of tissue weights from ['191_B6CASTF1J_10M_31'] 0 doesn't match 267.0. 637\n",
      "Sum of tissue weights from ['192_B6CASTF1J_10F_31'] 0 doesn't match 400.0. 638\n",
      "Sum of tissue weights from ['193_B6CASTF1J_10M_31'] 0 doesn't match 316.0. 639\n",
      "Sum of tissue weights from ['194_B6CASTF1J_10F_31'] 0 doesn't match 415.0. 640\n",
      "Sum of tissue weights from ['195_B6CASTF1J_10M_31'] 0 doesn't match 276.0. 641\n",
      "Sum of tissue weights from ['196_B6CASTF1J_10F_31'] 0 doesn't match 409.0. 642\n",
      "Sum of tissue weights from ['197_B6CASTF1J_10M_31'] 0 doesn't match 310.0. 643\n",
      "Sum of tissue weights from ['198_B6CASTF1J_10F_31'] 0 doesn't match 410.0. 644\n",
      "Sum of tissue weights from ['207_B6PWKF1J_10M_31'] 0 doesn't match 178.0. 645\n",
      "Sum of tissue weights from ['208_B6PWKF1J_10F_31'] 0 doesn't match 230.0. 646\n",
      "Sum of tissue weights from ['213_B6PWKF1J_10M_31'] 0 doesn't match 172.0. 647\n",
      "Sum of tissue weights from ['210_B6PWKF1J_10F_31'] 0 doesn't match 239.0. 648\n",
      "Sum of tissue weights from ['215_B6PWKF1J_10M_31'] 0 doesn't match 190.0. 649\n",
      "Sum of tissue weights from ['212_B6PWKF1J_10F_31'] 0 doesn't match 286.0. 650\n",
      "Sum of tissue weights from ['217_B6PWKF1J_10M_31'] 0 doesn't match 189.0. 651\n",
      "Sum of tissue weights from ['214_B6PWKF1J_10F_31'] 0 doesn't match 242.0. 652\n",
      "Sum of tissue weights from ['223_B6WSBF1J_10M_31'] 0 doesn't match 226.0. 653\n",
      "Sum of tissue weights from ['228_B6WSBF1J_10F_31'] 0 doesn't match 331.0. 654\n",
      "Sum of tissue weights from ['225_B6WSBF1J_10M_31'] 0 doesn't match 258.0. 655\n",
      "Sum of tissue weights from ['232_B6WSBF1J_10F_31'] 0 doesn't match 357.0. 656\n",
      "Sum of tissue weights from ['227_B6WSBF1J_10M_31'] 0 doesn't match 233.0. 657\n",
      "Sum of tissue weights from ['234_B6WSBF1J_10F_31'] 0 doesn't match 325.0. 658\n",
      "Sum of tissue weights from ['229_B6WSBF1J_10M_31'] 0 doesn't match 218.0. 659\n",
      "Sum of tissue weights from ['236_B6WSBF1J_10F_31'] 0 doesn't match 302.0. 660\n",
      "Sum of tissue weights from ['239_TREM2R47HNSS_HO_10M_31'] 0 doesn't match 443.0. 661\n",
      "Sum of tissue weights from ['240_TREM2R47HNSS_HO_10F_31'] 0 doesn't match 612.0. 662\n",
      "Sum of tissue weights from ['241_TREM2R47HNSS_HO_10M_31'] 0 doesn't match 479.0. 663\n",
      "Sum of tissue weights from ['242_TREM2R47HNSS_HO_10F_31'] 0 doesn't match 622.0. 664\n",
      "Sum of tissue weights from ['243_TREM2R47HNSS_HO_10M_31'] 0 doesn't match 440.0. 665\n",
      "Sum of tissue weights from ['244_TREM2R47HNSS_HO_10F_31'] 0 doesn't match 616.0. 666\n",
      "Sum of tissue weights from ['245_TREM2R47HNSS_HO_10M_31'] 0 doesn't match 420.0. 667\n",
      "Sum of tissue weights from ['248_TREM2R47HNSS_HO_10F_31'] 0 doesn't match 602.0. 668\n",
      "Sum of tissue weights from ['127_B6AF1J_10M_33'] 0 doesn't match 226.0. 705\n",
      "Sum of tissue weights from ['128_B6AF1J_10F_33'] 0 doesn't match 336.0. 706\n",
      "Sum of tissue weights from ['129_B6AF1J_10M_33'] 0 doesn't match 277.0. 707\n",
      "Sum of tissue weights from ['130_B6AF1J_10F_33'] 0 doesn't match 353.0. 708\n",
      "Sum of tissue weights from ['131_B6AF1J_10M_33'] 0 doesn't match 248.0. 709\n",
      "Sum of tissue weights from ['132_B6AF1J_10F_33'] 0 doesn't match 330.0. 710\n",
      "Sum of tissue weights from ['133_B6AF1J_10M_33'] 0 doesn't match 253.0. 711\n",
      "Sum of tissue weights from ['134_B6AF1J_10F_33'] 0 doesn't match 357.0. 712\n",
      "Sum of tissue weights from ['143_B6129SF1J_10M_33'] 0 doesn't match 188.0. 713\n",
      "Sum of tissue weights from ['144_B6129SF1J_10F_33'] 0 doesn't match 289.0. 714\n",
      "Sum of tissue weights from ['146_B6129SF1J_10F_33'] 0 doesn't match 276.0. 716\n",
      "Sum of tissue weights from ['147_B6129SF1J_10M_33'] 0 doesn't match 228.0. 717\n",
      "Sum of tissue weights from ['148_B6129SF1J_10F_33'] 0 doesn't match 316.0. 718\n",
      "Sum of tissue weights from ['149_B6129SF1J_10M_33'] 0 doesn't match 192.0. 719\n",
      "Sum of tissue weights from ['150_B6129SF1J_10F_33'] 0 doesn't match 283.0. 720\n",
      "Sum of tissue weights from ['159_B6NODF1J_10M_33'] 0 doesn't match 252.0. 721\n",
      "Sum of tissue weights from ['160_B6NODF1J_10F_33'] 0 doesn't match 280.0. 722\n",
      "Sum of tissue weights from ['161_B6NODF1J_10M_33'] 0 doesn't match 222.0. 723\n",
      "Sum of tissue weights from ['164_B6NODF1J_10F_33'] 0 doesn't match 220.0. 724\n",
      "Sum of tissue weights from ['165_B6NODF1J_10M_33'] 0 doesn't match 235.0. 725\n",
      "Sum of tissue weights from ['166_B6NODF1J_10F_33'] 0 doesn't match 258.0. 726\n",
      "Sum of tissue weights from ['175_B6NZOF1J_10M_33'] 0 doesn't match 146.0. 729\n",
      "Sum of tissue weights from ['176_B6NZOF1J_10F_33'] 0 doesn't match 176.0. 730\n",
      "Sum of tissue weights from ['177_B6NZOF1J_10M_33'] 0 doesn't match 167.0. 731\n",
      "Sum of tissue weights from ['178_B6NZOF1J_10F_33'] 0 doesn't match 154.0. 732\n",
      "Sum of tissue weights from ['181_B6NZOF1J_10M_33'] 0 doesn't match 154.0. 733\n",
      "Sum of tissue weights from ['180_B6NZOF1J_10F_33'] 0 doesn't match 182.0. 734\n",
      "Sum of tissue weights from ['183_B6NZOF1J_10M_33'] 0 doesn't match 150.0. 735\n",
      "Sum of tissue weights from ['182_B6NZOF1J_10F_33'] 0 doesn't match 211.0. 736\n",
      "Sum of tissue weights from ['192_B6CASTF1J_10F_33'] 0 doesn't match 262.0. 738\n",
      "Sum of tissue weights from ['193_B6CASTF1J_10M_33'] 0 doesn't match 238.0. 739\n",
      "Sum of tissue weights from ['194_B6CASTF1J_10F_33'] 0 doesn't match 258.0. 740\n",
      "Sum of tissue weights from ['195_B6CASTF1J_10M_33'] 0 doesn't match 218.0. 741\n",
      "Sum of tissue weights from ['196_B6CASTF1J_10F_33'] 0 doesn't match 257.0. 742\n",
      "Sum of tissue weights from ['197_B6CASTF1J_10M_33'] 0 doesn't match 214.0. 743\n",
      "Sum of tissue weights from ['198_B6CASTF1J_10F_33'] 0 doesn't match 239.0. 744\n",
      "Sum of tissue weights from ['207_B6PWKF1J_10M_33'] 0 doesn't match 193.0. 745\n",
      "Sum of tissue weights from ['208_B6PWKF1J_10F_33'] 0 doesn't match 207.0. 746\n",
      "Sum of tissue weights from ['210_B6PWKF1J_10F_33'] 0 doesn't match 225.0. 748\n",
      "Sum of tissue weights from ['215_B6PWKF1J_10M_33'] 0 doesn't match 203.0. 749\n",
      "Sum of tissue weights from ['212_B6PWKF1J_10F_33'] 0 doesn't match 242.0. 750\n",
      "Sum of tissue weights from ['217_B6PWKF1J_10M_33'] 0 doesn't match 195.0. 751\n",
      "Sum of tissue weights from ['214_B6PWKF1J_10F_33'] 0 doesn't match 224.0. 752\n",
      "Sum of tissue weights from ['223_B6WSBF1J_10M_33'] 0 doesn't match 162.0. 753\n",
      "Sum of tissue weights from ['228_B6WSBF1J_10F_33'] 0 doesn't match 216.0. 754\n",
      "Sum of tissue weights from ['232_B6WSBF1J_10F_33'] 0 doesn't match 237.0. 756\n",
      "Sum of tissue weights from ['227_B6WSBF1J_10M_33'] 0 doesn't match 192.0. 757\n",
      "Sum of tissue weights from ['234_B6WSBF1J_10F_33'] 0 doesn't match 241.0. 758\n",
      "Sum of tissue weights from ['229_B6WSBF1J_10M_33'] 0 doesn't match 183.0. 759\n",
      "Sum of tissue weights from ['236_B6WSBF1J_10F_33'] 0 doesn't match 236.0. 760\n",
      "Sum of tissue weights from ['239_TREM2R47HNSS_HO_10M_33'] 0 doesn't match 270.0. 761\n",
      "Sum of tissue weights from ['240_TREM2R47HNSS_HO_10F_33'] 0 doesn't match 359.0. 762\n",
      "Sum of tissue weights from ['241_TREM2R47HNSS_HO_10M_33'] 0 doesn't match 298.0. 763\n",
      "Sum of tissue weights from ['242_TREM2R47HNSS_HO_10F_33'] 0 doesn't match 340.0. 764\n",
      "Sum of tissue weights from ['243_TREM2R47HNSS_HO_10M_33'] 0 doesn't match 268.0. 765\n",
      "Sum of tissue weights from ['244_TREM2R47HNSS_HO_10F_33'] 0 doesn't match 308.0. 766\n",
      "Sum of tissue weights from ['245_TREM2R47HNSS_HO_10M_33'] 0 doesn't match 286.0. 767\n",
      "Sum of tissue weights from ['248_TREM2R47HNSS_HO_10F_33'] 0 doesn't match 330.0. 768\n"
     ]
    }
   ],
   "source": [
    "load_samples_into_experiment(lizs_sheet_name, \"F1 Samples into experiment\", header=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "entertaining-fantasy",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# According to Liz this sheet isn't ready yet.\n",
    "# load_samples_into_experiment(lizs_sheet_name, \"CC Samples into experiment\", header=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "specific-candle",
   "metadata": {},
   "source": [
    "# Plate layout\n",
    "\n",
    "(the hard thing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "celtic-mystery",
   "metadata": {},
   "outputs": [],
   "source": [
    "WellContent = namedtuple(\"well_content\", [\"genotype\", \"tissue_id\"])\n",
    "\n",
    "\n",
    "def get_genotype_from_tissue(tissue_id):\n",
    "    assert pandas.notnull(tissue_id)\n",
    "\n",
    "    fields = tissue_id.split(\"_\")\n",
    "    assert len(fields) == 4, f\"Field label validation fail {tissue_id}, not enough fields {len(fields)}\"\n",
    "\n",
    "    return normalize_strain(fields[1])\n",
    "\n",
    "assert get_genotype_from_tissue(\"016_B6J_10F_20\") == \"B6J\"\n",
    "assert get_genotype_from_tissue(\"046_NZOJ_10F_03\") == \"NZOJ\"\n",
    "assert get_genotype_from_tissue(\"198_B6CASTF1/J_10F_20\") == \"B6CASTF1J\"\n",
    "\n",
    "\n",
    "\n",
    "def validate_tissue_ids(contents, expected_genotypes):\n",
    "    for expected, (row_index, row) in zip(expected_genotypes, contents.iterrows()):\n",
    "        for cell in row:\n",
    "            fields = cell.split(\"_\")\n",
    "            assert len(fields) == 4, f\"Field label validation fail {cell}\"\n",
    "            # there's a few sample swaps, don't validate them.\n",
    "            if expected == \"Mixed\":\n",
    "                continue\n",
    "            elif fields[1] not in (\"092_CASTJ_10F_03\"):\n",
    "                assert fields[1] == expected, f\"{fields[1]} failed to match {expected}\"\n",
    "\n",
    "\n",
    "def is_plate_name(name):\n",
    "    return not pandas.isnull(name) and name.startswith(\"IGVF_\")\n",
    "\n",
    "def find_plate_start(sheet, offsets):\n",
    "    block_headers = [\n",
    "        \"M\",\"F\",\"Mixed\",\"B6J\",\"NODJ\",\"AJ\",\"PWKJ\",\"129S1J\",\"CASTJ\",\"WSBJ\",\"NZOJ\",\"B6CASTF1/J\",\"B6PWKF1/J\",\"B6WSBF1/J\",\"B6NODF1/J\"\n",
    "    ]\n",
    "    for plate_id_row in sheet[sheet[offsets[\"plate_label\"]].apply(is_plate_name)].index:\n",
    "        plate_name = normalize_plate_name(sheet.loc[plate_id_row, offsets[\"plate_label\"]])\n",
    "        for i in range(plate_id_row, plate_id_row + 4):\n",
    "            cell = sheet.loc[i, offsets[\"well_start\"]]\n",
    "            if isinstance(cell, str) and (cell.startswith(\"Tissue\") or cell in block_headers):\n",
    "                yield (plate_name, i)\n",
    "                \n",
    "def get_plate_genotype(sheet, start, offsets):\n",
    "    plate_tissues = []\n",
    "    for cell in sheet.loc[start+2:start+9, offsets[\"tissue_label\"]].tolist():\n",
    "        if not pandas.isnull(cell):\n",
    "            match = re.match(\"(B6J|NODJ|AJ|PWKJ|129S1J|CASTJ|WSBJ|NZOJ|Mixed)\", cell, re.IGNORECASE)\n",
    "            if match is not None:\n",
    "                plate_tissues.append(match.group(0))\n",
    "            else:\n",
    "                print(\"Diffculty matching {} in {} {}\".format(cell, start+2, start+9))\n",
    "    \n",
    "    return plate_tissues\n",
    "\n",
    "\n",
    "def validate_plate_column_ids(plate_column_ids):\n",
    "    expected_column_ids = set(range(1,13))\n",
    "    for column_id in plate_column_ids:\n",
    "        assert column_id in expected_column_ids, f\"{column_id} {type(column_id)} not in expected_column_ids\"\n",
    "    \n",
    "def validate_plate_row_ids(plate_row_labels):\n",
    "    expected_row_labels = set((\"A\", \"B\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\"))\n",
    "    for row_label in plate_row_labels:\n",
    "        assert row_label in expected_row_labels, f\"{row_label} not in expected labels\"\n",
    "    \n",
    "def parse_single_well_block(sheet, plate_start, offsets):\n",
    "    # column ids\n",
    "    simple_plate_column_ids = sheet.loc[plate_start+1, offsets[\"well_range\"]].tolist()\n",
    "    validate_plate_column_ids(simple_plate_column_ids)\n",
    "\n",
    "    plate_row_labels = sheet.loc[plate_start+2:plate_start+9, offsets[\"well_row_label\"]].dropna().tolist()\n",
    "    validate_plate_row_ids(plate_row_labels)\n",
    "\n",
    "    contents = sheet.loc[plate_start+2:plate_start+len(plate_row_labels)+1, offsets[\"well_range\"]].copy()\n",
    "    contents.index = plate_row_labels\n",
    "    contents.columns = simple_plate_column_ids\n",
    "    contents = contents.dropna()\n",
    "\n",
    "    genotypes = get_plate_genotype(sheet, plate_start, offsets)\n",
    "    validate_tissue_ids(contents, genotypes)\n",
    "\n",
    "    well_contents = {}\n",
    "    for well_row, row in contents.iterrows():\n",
    "        for well_column, tissue_cell in zip(simple_plate_column_ids, row):\n",
    "            genotype = get_genotype_from_tissue(tissue_cell)\n",
    "            well_contents[(well_row, str(well_column))] = [WellContent(genotype, tissue_cell)]\n",
    "\n",
    "    return well_contents\n",
    "\n",
    "def parse_multiplexed_well_block(sheet, start, expected_genotypes, offsets):\n",
    "    well_re = re.compile(\"^[A-H]1?[\\d]$\")\n",
    "    wells = sheet.loc[start+3, offsets[\"well_range\"]].tolist()\n",
    "    for cell in wells:\n",
    "        if pandas.isnull(cell) or well_re.match(cell) is None:\n",
    "            raise ValueError(f\"well value {cell} in row {start+3} does not look correct\")\n",
    "\n",
    "    contents = sheet.loc[start+0:start+1, offsets[\"well_range\"]]\n",
    "    contents.columns = [(x[0], x[1:]) for x in wells]\n",
    "    validate_tissue_ids(contents, expected_genotypes)\n",
    "    contents.index = expected_genotypes\n",
    "\n",
    "    well_contents = {}\n",
    "    for expected_genotype, row in contents.dropna().iterrows():\n",
    "        for well, tissue_id in row.iteritems():\n",
    "            genotype = get_genotype_from_tissue(tissue_id)\n",
    "            well_contents.setdefault(well, []).append(WellContent(genotype, tissue_id))\n",
    "\n",
    "    return well_contents\n",
    "\n",
    "def parse_plate(sheet):\n",
    "    max_rows_from_plate_name_to_well_names = 4\n",
    "\n",
    "    offsets = {\n",
    "        \"plate_label\": 1,\n",
    "        \"tissue_label\": 1,\n",
    "        \"well_row_label\": 2,\n",
    "        \"well_start\": 3,\n",
    "        \"well_range\": slice(3, 10),\n",
    "    }\n",
    "    offsets_simple = offsets.copy()\n",
    "    offsets_simple[\"well_range\"] = slice(3, 14)\n",
    "\n",
    "    for plate_name, plate_start in find_plate_start(sheet, offsets):\n",
    "        if plate_name.endswith(\"XX\"):\n",
    "            return\n",
    "\n",
    "        well_contents = {}\n",
    "        genotypes = get_plate_genotype(sheet, plate_start, offsets)\n",
    "        \n",
    "        if plate_name in (\"IGVF_B01\", \"IGVF_012\", \"IGVF_013\", \"IGVF_014\"):\n",
    "            well_contents.update(parse_single_well_block(sheet, plate_start, offsets_simple))\n",
    "            complex_blocks = []        \n",
    "        else:\n",
    "            well_contents.update(parse_single_well_block(sheet, plate_start, offsets))\n",
    "\n",
    "            complex_blocks = [\n",
    "                (plate_start + 12, slice(0, 2)),\n",
    "                (plate_start + 12 + 5, slice(2, 4)),\n",
    "                (plate_start + 12 + 10, slice(4, 6)),\n",
    "                (plate_start + 12 + 15, slice(6, 8)),\n",
    "            ]\n",
    "\n",
    "            for complex_start, genotype_slice in complex_blocks:\n",
    "                expected_genotypes = genotypes[genotype_slice]\n",
    "                well_contents.update(parse_multiplexed_well_block(sheet, complex_start, expected_genotypes, offsets))\n",
    "        \n",
    "        yield (plate_name, well_contents)\n",
    "        \n",
    "offsets = {\n",
    "    \"plate_label\": 1,\n",
    "    \"tissue_label\": 1,\n",
    "    \"well_row_label\": 2,\n",
    "    \"well_start\": 3,\n",
    "    \"well_range\": slice(3, 14),\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "accessory-venue",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IGVF_003 96\n",
      "IGVF_004 96\n",
      "IGVF_005 96\n",
      "IGVF_006 96\n",
      "IGVF_007 96\n",
      "IGVF_008 96\n",
      "IGVF_008B 96\n",
      "IGVF_B01 48\n",
      "IGVF_009 96\n",
      "IGVF_010 96\n",
      "IGVF_011 96\n",
      "IGVF_012 48\n",
      "IGVF_013 96\n",
      "IGVF_014 48\n"
     ]
    }
   ],
   "source": [
    "        \n",
    "plate_layout = pandas.read_excel(\n",
    "    lizs_sheet_name, \n",
    "    sheet_name=\"Plate setups\",\n",
    "    header=None,\n",
    ")\n",
    "\n",
    "\n",
    "for plate_name, plate_contents in parse_plate(plate_layout):\n",
    "    print(plate_name, len(plate_contents))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "progressive-purchase",
   "metadata": {},
   "outputs": [],
   "source": [
    "# half height 4 row plates\n",
    "well_contents = parse_single_well_block(plate_layout, 257, offsets)\n",
    "assert len(well_contents) == 48, f\"{len(well_contents)}\"\n",
    "\n",
    "well_contents = parse_single_well_block(plate_layout, 395, offsets)\n",
    "assert len(well_contents) == 48, f\"{len(well_contents)}\"\n",
    "\n",
    "# Full length 8 row plate\n",
    "well_contents = parse_single_well_block(plate_layout, 406, offsets)\n",
    "assert len(well_contents) == 96, f\"{len(well_contents)}\"\n",
    "#print(well_contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "streaming-savings",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<QuerySet [<LibraryConstructionReagent: Parse WT Mega v2>, <LibraryConstructionReagent: Parse WT v2>]>"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "models.LibraryConstructionReagent.objects.all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "subject-accused",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def guess_barcode_reagent_from_plate(plate_name):\n",
    "    # Depend on a global to avoid hitting the database a bunch\n",
    "    wt_mega_2_reagent = models.LibraryConstructionReagent.objects.get(name=\"wt-mega\", version=\"v2\")\n",
    "    wt_regular_2_reagent = models.LibraryConstructionReagent.objects.get(name=\"wt\", version=\"v2\")    \n",
    "    \n",
    "    if plate_name in (\"IGVF_B01\", \"IGVF_012\"):\n",
    "        reagent = wt_regular_2_reagent\n",
    "    else:\n",
    "        reagent = wt_mega_2_reagent\n",
    "\n",
    "    return reagent\n",
    "\n",
    "assert str(guess_barcode_reagent_from_plate(\"IGVF_003\")) == \"Parse WT Mega v2\"\n",
    "assert str(guess_barcode_reagent_from_plate(\"IGVF_012\")) == \"Parse WT v2\"\n",
    "\n",
    "def load_plates(plate_layout):\n",
    "    # Validate tissue references:\n",
    "    failed = False\n",
    "    for plate_name, plate_contents in parse_plate(plate_layout):\n",
    "        for well_id in plate_contents:\n",
    "            well_contents = plate_contents[well_id]\n",
    "\n",
    "            for well_fraction in well_contents:\n",
    "                tissue_id = normalize_strain(well_fraction.tissue_id)\n",
    "                try:\n",
    "                    biosample = models.FixedSample.objects.get(name=tissue_id)\n",
    "                except models.FixedSample.DoesNotExist:\n",
    "                    print(f\"Unable to find {tissue_id} on plate {plate_name}\")\n",
    "                    failed = True\n",
    "\n",
    "    #assert not failed, \"Resolve tissues\"\n",
    "\n",
    "    # populate database\n",
    "    if models.SplitSeqWell.biosample.through.objects.count() > 0:\n",
    "        truncate(models.SplitSeqWell.biosample.through)\n",
    "\n",
    "    if models.SplitSeqWell.objects.count() > 0:\n",
    "        truncate(models.SplitSeqWell)\n",
    "\n",
    "    if models.SplitSeqPlate.objects.count() > 0:\n",
    "        truncate(models.SplitSeqPlate)\n",
    "\n",
    "    errors = 0\n",
    "    plates_loaded = set()\n",
    "\n",
    "    for plate_name, plate_contents in parse_plate(plate_layout):\n",
    "        plates_loaded.add(plate_name)\n",
    "        plate_record = models.SplitSeqPlate(\n",
    "            name=plate_name,\n",
    "            size=models.PlateSizeEnum.size_96,\n",
    "            pool_location=None,\n",
    "            date_performed=None,\n",
    "        )\n",
    "        plate_record.save()\n",
    "\n",
    "        for well_id in plate_contents:\n",
    "            well_contents = plate_contents[well_id]\n",
    "\n",
    "            biosamples = []\n",
    "            for well_content in well_contents:\n",
    "                tissue_id = normalize_strain(well_content.tissue_id)\n",
    "                try:\n",
    "                    biosamples.append(models.FixedSample.objects.get(name=tissue_id))\n",
    "                except models.FixedSample.DoesNotExist:\n",
    "                    print(f\"unable to find tissue {tissue_id} for {plate_name} {well_id}\")\n",
    "                    errors += 1\n",
    "\n",
    "\n",
    "            reagent = guess_barcode_reagent_from_plate(plate_name)\n",
    "\n",
    "            barcodes = models.LibraryBarcode.objects.filter(\n",
    "                reagent=reagent,\n",
    "                code=\"{}{}\".format(well_id[0], well_id[1]),\n",
    "            )\n",
    "\n",
    "            well_record = models.SplitSeqWell(\n",
    "                plate=plate_record,\n",
    "                row=well_id[0],\n",
    "                column=well_id[1],\n",
    "            )\n",
    "            well_record.save()\n",
    "            well_record.biosample.set(biosamples)\n",
    "            well_record.barcode.set(barcodes)\n",
    "            well_record.save()\n",
    "\n",
    "    assert not errors, \"Check error messages\"\n",
    "    assert \"IGVF_008B\" in plates_loaded, \"Remember to make copied IGVF_008B plate\"\n",
    "    assert \"IGVF_012\" in plates_loaded, \"Still having trouble with IGVF_012\"\n",
    "    \n",
    "load_plates(plate_layout)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "premium-wiring",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<Tissue: 191_B6CASTF1J_10M_01 Hypothalamus/Pituitary>,\n",
       " <Tissue: 191_B6CASTF1J_10M_03 Cortex/Hippocampus left>,\n",
       " <Tissue: 191_B6CASTF1J_10M_04 Cortex/Hippocampus right>,\n",
       " <Tissue: 191_B6CASTF1J_10M_05 Liver>,\n",
       " <Tissue: 191_B6CASTF1J_10M_06 Heart>,\n",
       " <Tissue: 191_B6CASTF1J_10M_08 Adrenal>,\n",
       " <Tissue: 191_B6CASTF1J_10M_19 Tail>,\n",
       " <Tissue: 191_B6CASTF1J_10M_20 PBMC - WBC>,\n",
       " <Tissue: 191_B6CASTF1J_10M_27 Testis left>,\n",
       " <Tissue: 191_B6CASTF1J_10M_28 Testis right>,\n",
       " <Tissue: 191_B6CASTF1J_10M_29 Epididymis left>,\n",
       " <Tissue: 191_B6CASTF1J_10M_30 Epididymis right>,\n",
       " <Tissue: 191_B6CASTF1J_10M_31 Kidney left>,\n",
       " <Tissue: 191_B6CASTF1J_10M_32 Kidney right>,\n",
       " <Tissue: 191_B6CASTF1J_10M_33 Gastrocnemius left>,\n",
       " <Tissue: 191_B6CASTF1J_10M_34 Gastrocnemius right>]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[x for x in models.Tissue.objects.all() if x.name.startswith(\"191_B6CAST\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "perceived-albany",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "headed-scope",
   "metadata": {},
   "source": [
    "# Submitted measurement sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "arctic-burke",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_submitted_measurement_sets():\n",
    "    measurement_sets = {}\n",
    "\n",
    "    for accession_prefix in submitted_sheets:\n",
    "        for i, row in submitted_sheets[accession_prefix][\"measurement_set\"].iterrows():\n",
    "            if not pandas.isnull(row[\"accession\"]):\n",
    "                aliases = row[\"aliases:array\"].split(',')\n",
    "                name = aliases[0].replace(\"ali-mortazavi:\", \"\")\n",
    "\n",
    "                try:\n",
    "                    record = models.MeasurementSet.objects.get(name=name)\n",
    "                except models.MeasurementSet.DoesNotExist:\n",
    "                    record = models.MeasurementSet(\n",
    "                        name=name,\n",
    "                    )\n",
    "                    record.save()\n",
    "\n",
    "                    if accession_prefix == \"igvftst\":\n",
    "                        see_also_template = \"https://api.sandbox.igvf.org/measurment-sets/{}/\"\n",
    "                    elif accession_prefix == \"igvf\":\n",
    "                        see_also_template = \"https://api.data.igvf.org/measurement-sets/{}/\"\n",
    "\n",
    "                    accession = models.Accession(\n",
    "                        accession_prefix=accession_prefix,\n",
    "                        name=row[\"accession\"],\n",
    "                        uuid=row[\"uuid\"],\n",
    "                        see_also=see_also_template.format(row[\"accession\"]),\n",
    "                    )\n",
    "                    accession.save()\n",
    "                    record.accession.add(accession)\n",
    "                    record.save()\n",
    "                measurement_sets[name] = record\n",
    "    print(\"loaded\", len(measurement_sets), \"measurement_sets\")\n",
    "    return measurement_sets\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "endless-compromise",
   "metadata": {},
   "source": [
    "# Subpool\n",
    "\n",
    "## Extract meaning from the Experiments tab"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "permanent-italic",
   "metadata": {},
   "source": [
    "- number of nuclei per subpool (Experiment) \n",
    "- cDNA amp # PCR cycles (Experiment)\n",
    "- cDNA ng/ul in 25ul (experiment)\n",
    "- total cDNA ng (experiment)\n",
    "- Bioanalyzer cDNA ave bp length (expierment)\n",
    "- Sub library Index PCR # (experiment)\n",
    "- Sublibrary Index (experiment)\n",
    "- Sequence for SampleSheet (experiment)\n",
    "- library ng/ul (experiment)\n",
    "- Bioanalyzer library ave bp length (experiment)\n",
    "- Nextseq run number (experiment)\n",
    "\n",
    "- number of 67k aliquots  (experiment)\n",
    "- number of 8k aliquots (experiment)\n",
    "\n",
    "- QC # raw reads (1 mismatch) (experiment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "civil-musician",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Experiment</th>\n",
       "      <th>mice</th>\n",
       "      <th>Fixation Box</th>\n",
       "      <th>Split-seq prep start date</th>\n",
       "      <th>Submit</th>\n",
       "      <th>Sublibrary</th>\n",
       "      <th>[Barcoded Nuclei] (nuclei/uL)</th>\n",
       "      <th>Barcoded Nuclei Vol. (uL)</th>\n",
       "      <th>Total barcoded nuclei (1st 3 rounds)</th>\n",
       "      <th># of backup aliquots</th>\n",
       "      <th>...</th>\n",
       "      <th>Unnamed: 54</th>\n",
       "      <th>Unnamed: 55</th>\n",
       "      <th>Unnamed: 56</th>\n",
       "      <th>Unnamed: 57</th>\n",
       "      <th>Unnamed: 58</th>\n",
       "      <th>Unnamed: 59</th>\n",
       "      <th>Unnamed: 60</th>\n",
       "      <th>Unnamed: 61</th>\n",
       "      <th>Unnamed: 62</th>\n",
       "      <th>Unnamed: 63</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IGVF_Splitseq_002</td>\n",
       "      <td>B6 (2F+2M)</td>\n",
       "      <td>IGVF_FIX_001</td>\n",
       "      <td>2022-08-24</td>\n",
       "      <td>No</td>\n",
       "      <td>002_4A</td>\n",
       "      <td>4540</td>\n",
       "      <td>240.0</td>\n",
       "      <td>1089600</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CAST (2F+2M)</td>\n",
       "      <td>IGVF_FIX_002</td>\n",
       "      <td>NaT</td>\n",
       "      <td>No</td>\n",
       "      <td>002_4B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Fixation V1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>No</td>\n",
       "      <td>002_64A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Target Nuclei</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1M V1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>No</td>\n",
       "      <td>002_64B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1018000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>No</td>\n",
       "      <td>002_64C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Excess Nuclei</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1009</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1010</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1011</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1012</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1013</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1014 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Experiment          mice  Fixation Box Split-seq prep start date  \\\n",
       "0     IGVF_Splitseq_002    B6 (2F+2M)  IGVF_FIX_001                2022-08-24   \n",
       "1                   NaN  CAST (2F+2M)  IGVF_FIX_002                       NaT   \n",
       "2           Fixation V1           NaN           NaN                       NaT   \n",
       "3                 1M V1           NaN           NaN                       NaT   \n",
       "4                   NaN           NaN           NaN                       NaT   \n",
       "...                 ...           ...           ...                       ...   \n",
       "1009                NaN           NaN           NaN                       NaT   \n",
       "1010                NaN           NaN           NaN                       NaT   \n",
       "1011                NaN           NaN           NaN                       NaT   \n",
       "1012                NaN           NaN           NaN                       NaT   \n",
       "1013                NaN           NaN           NaN                       NaT   \n",
       "\n",
       "     Submit Sublibrary [Barcoded Nuclei] (nuclei/uL)  \\\n",
       "0        No     002_4A                          4540   \n",
       "1        No     002_4B                           NaN   \n",
       "2        No    002_64A                           NaN   \n",
       "3        No    002_64B                           NaN   \n",
       "4        No    002_64C                           NaN   \n",
       "...     ...        ...                           ...   \n",
       "1009    NaN        NaN                           NaN   \n",
       "1010    NaN        NaN                           NaN   \n",
       "1011    NaN        NaN                           NaN   \n",
       "1012    NaN        NaN                           NaN   \n",
       "1013    NaN        NaN                           NaN   \n",
       "\n",
       "      Barcoded Nuclei Vol. (uL) Total barcoded nuclei (1st 3 rounds)  \\\n",
       "0                         240.0                              1089600   \n",
       "1                           NaN                                  NaN   \n",
       "2                           NaN                        Target Nuclei   \n",
       "3                           NaN                              1018000   \n",
       "4                           NaN                        Excess Nuclei   \n",
       "...                         ...                                  ...   \n",
       "1009                        NaN                                  NaN   \n",
       "1010                        NaN                                  NaN   \n",
       "1011                        NaN                                  NaN   \n",
       "1012                        NaN                                  NaN   \n",
       "1013                        NaN                                  NaN   \n",
       "\n",
       "      # of backup aliquots  ... Unnamed: 54  Unnamed: 55  Unnamed: 56  \\\n",
       "0                      3.0  ...         NaN          NaN          NaN   \n",
       "1                      NaN  ...         NaN          NaN          NaN   \n",
       "2                      NaN  ...         NaN          NaN          NaN   \n",
       "3                      NaN  ...         NaN          NaN          NaN   \n",
       "4                      NaN  ...         NaN          NaN          NaN   \n",
       "...                    ...  ...         ...          ...          ...   \n",
       "1009                   NaN  ...         NaN          NaN          NaN   \n",
       "1010                   NaN  ...         NaN          NaN          NaN   \n",
       "1011                   NaN  ...         NaN          NaN          NaN   \n",
       "1012                   NaN  ...         NaN          NaN          NaN   \n",
       "1013                   NaN  ...         NaN          NaN          NaN   \n",
       "\n",
       "      Unnamed: 57 Unnamed: 58 Unnamed: 59 Unnamed: 60 Unnamed: 61  \\\n",
       "0             NaN         NaN         NaN         NaN         NaN   \n",
       "1             NaN         NaN         NaN         NaN         NaN   \n",
       "2             NaN         NaN         NaN         NaN         NaN   \n",
       "3             NaN         NaN         NaN         NaN         NaN   \n",
       "4             NaN         NaN         NaN         NaN         NaN   \n",
       "...           ...         ...         ...         ...         ...   \n",
       "1009          NaN         NaN         NaN         NaN         NaN   \n",
       "1010          NaN         NaN         NaN         NaN         NaN   \n",
       "1011          NaN         NaN         NaN         NaN         NaN   \n",
       "1012          NaN         NaN         NaN         NaN         NaN   \n",
       "1013          NaN         NaN         NaN         NaN         NaN   \n",
       "\n",
       "      Unnamed: 62  Unnamed: 63  \n",
       "0             NaN          NaN  \n",
       "1             NaN          NaN  \n",
       "2             NaN          NaN  \n",
       "3             NaN          NaN  \n",
       "4             NaN          NaN  \n",
       "...           ...          ...  \n",
       "1009          NaN          NaN  \n",
       "1010          NaN          NaN  \n",
       "1011          NaN          NaN  \n",
       "1012          NaN          NaN  \n",
       "1013          NaN          NaN  \n",
       "\n",
       "[1014 rows x 64 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "experiment = pandas.read_excel(\n",
    "    lizs_sheet_name, \n",
    "    sheet_name=\"Experiment\",\n",
    "    header=0,\n",
    "    index_col=None,\n",
    "    #usecols=range(0, 49)\n",
    ")\n",
    "experiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "coordinated-destruction",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Experiment</th>\n",
       "      <th>mice</th>\n",
       "      <th>Fixation Box</th>\n",
       "      <th>Split-seq prep start date</th>\n",
       "      <th>Submit</th>\n",
       "      <th>Sublibrary</th>\n",
       "      <th>[Barcoded Nuclei] (nuclei/uL)</th>\n",
       "      <th>Barcoded Nuclei Vol. (uL)</th>\n",
       "      <th>Total barcoded nuclei (1st 3 rounds)</th>\n",
       "      <th># of backup aliquots</th>\n",
       "      <th>...</th>\n",
       "      <th>Unnamed: 54</th>\n",
       "      <th>Unnamed: 55</th>\n",
       "      <th>Unnamed: 56</th>\n",
       "      <th>Unnamed: 57</th>\n",
       "      <th>Unnamed: 58</th>\n",
       "      <th>Unnamed: 59</th>\n",
       "      <th>Unnamed: 60</th>\n",
       "      <th>Unnamed: 61</th>\n",
       "      <th>Unnamed: 62</th>\n",
       "      <th>Unnamed: 63</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IGVF_Splitseq_002</td>\n",
       "      <td>B6 (2F+2M)</td>\n",
       "      <td>IGVF_FIX_001</td>\n",
       "      <td>2022-08-24</td>\n",
       "      <td>No</td>\n",
       "      <td>002_4A</td>\n",
       "      <td>4540</td>\n",
       "      <td>240.0</td>\n",
       "      <td>1089600</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CAST (2F+2M)</td>\n",
       "      <td>IGVF_FIX_002</td>\n",
       "      <td>NaT</td>\n",
       "      <td>No</td>\n",
       "      <td>002_4B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Fixation V1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>No</td>\n",
       "      <td>002_64A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Target Nuclei</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1M V1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>No</td>\n",
       "      <td>002_64B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1018000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>No</td>\n",
       "      <td>002_64C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Excess Nuclei</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1009</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1010</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1011</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1012</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1013</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1014 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Experiment          mice  Fixation Box Split-seq prep start date  \\\n",
       "0     IGVF_Splitseq_002    B6 (2F+2M)  IGVF_FIX_001                2022-08-24   \n",
       "1                   NaN  CAST (2F+2M)  IGVF_FIX_002                       NaT   \n",
       "2           Fixation V1           NaN           NaN                       NaT   \n",
       "3                 1M V1           NaN           NaN                       NaT   \n",
       "4                   NaN           NaN           NaN                       NaT   \n",
       "...                 ...           ...           ...                       ...   \n",
       "1009                NaN           NaN           NaN                       NaT   \n",
       "1010                NaN           NaN           NaN                       NaT   \n",
       "1011                NaN           NaN           NaN                       NaT   \n",
       "1012                NaN           NaN           NaN                       NaT   \n",
       "1013                NaN           NaN           NaN                       NaT   \n",
       "\n",
       "     Submit Sublibrary [Barcoded Nuclei] (nuclei/uL)  \\\n",
       "0        No     002_4A                          4540   \n",
       "1        No     002_4B                           NaN   \n",
       "2        No    002_64A                           NaN   \n",
       "3        No    002_64B                           NaN   \n",
       "4        No    002_64C                           NaN   \n",
       "...     ...        ...                           ...   \n",
       "1009    NaN        NaN                           NaN   \n",
       "1010    NaN        NaN                           NaN   \n",
       "1011    NaN        NaN                           NaN   \n",
       "1012    NaN        NaN                           NaN   \n",
       "1013    NaN        NaN                           NaN   \n",
       "\n",
       "      Barcoded Nuclei Vol. (uL) Total barcoded nuclei (1st 3 rounds)  \\\n",
       "0                         240.0                              1089600   \n",
       "1                           NaN                                  NaN   \n",
       "2                           NaN                        Target Nuclei   \n",
       "3                           NaN                              1018000   \n",
       "4                           NaN                        Excess Nuclei   \n",
       "...                         ...                                  ...   \n",
       "1009                        NaN                                  NaN   \n",
       "1010                        NaN                                  NaN   \n",
       "1011                        NaN                                  NaN   \n",
       "1012                        NaN                                  NaN   \n",
       "1013                        NaN                                  NaN   \n",
       "\n",
       "      # of backup aliquots  ... Unnamed: 54  Unnamed: 55  Unnamed: 56  \\\n",
       "0                      3.0  ...         NaN          NaN          NaN   \n",
       "1                      NaN  ...         NaN          NaN          NaN   \n",
       "2                      NaN  ...         NaN          NaN          NaN   \n",
       "3                      NaN  ...         NaN          NaN          NaN   \n",
       "4                      NaN  ...         NaN          NaN          NaN   \n",
       "...                    ...  ...         ...          ...          ...   \n",
       "1009                   NaN  ...         NaN          NaN          NaN   \n",
       "1010                   NaN  ...         NaN          NaN          NaN   \n",
       "1011                   NaN  ...         NaN          NaN          NaN   \n",
       "1012                   NaN  ...         NaN          NaN          NaN   \n",
       "1013                   NaN  ...         NaN          NaN          NaN   \n",
       "\n",
       "      Unnamed: 57 Unnamed: 58 Unnamed: 59 Unnamed: 60 Unnamed: 61  \\\n",
       "0             NaN         NaN         NaN         NaN         NaN   \n",
       "1             NaN         NaN         NaN         NaN         NaN   \n",
       "2             NaN         NaN         NaN         NaN         NaN   \n",
       "3             NaN         NaN         NaN         NaN         NaN   \n",
       "4             NaN         NaN         NaN         NaN         NaN   \n",
       "...           ...         ...         ...         ...         ...   \n",
       "1009          NaN         NaN         NaN         NaN         NaN   \n",
       "1010          NaN         NaN         NaN         NaN         NaN   \n",
       "1011          NaN         NaN         NaN         NaN         NaN   \n",
       "1012          NaN         NaN         NaN         NaN         NaN   \n",
       "1013          NaN         NaN         NaN         NaN         NaN   \n",
       "\n",
       "      Unnamed: 62  Unnamed: 63  \n",
       "0             NaN          NaN  \n",
       "1             NaN          NaN  \n",
       "2             NaN          NaN  \n",
       "3             NaN          NaN  \n",
       "4             NaN          NaN  \n",
       "...           ...          ...  \n",
       "1009          NaN          NaN  \n",
       "1010          NaN          NaN  \n",
       "1011          NaN          NaN  \n",
       "1012          NaN          NaN  \n",
       "1013          NaN          NaN  \n",
       "\n",
       "[1014 rows x 64 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "experiment = pandas.read_excel(\n",
    "    lizs_sheet_name, \n",
    "    sheet_name=\"Experiment\",\n",
    "    header=0,\n",
    "    index_col=None,\n",
    "    #usecols=range(0, 49)\n",
    ").rename(columns={\n",
    "    # Unnamed: 42 seems like an index or reference to the sample sheet?\n",
    "    \"Unnamed: 43\": \"Novaseq raw reads\",\n",
    "    \"Unnamed: 44\": \"Novaseq1 L001\",\n",
    "    \"Unnamed: 45\": \"Novaseq1 L002\",\n",
    "    \"Unnamed: 46\": \"Novaseq1 L003\",\n",
    "    \"Unnamed: 47\": \"Novaseq1 L004\",\n",
    "    \"Unnamed: 48\": \"Novaseq2 L001\",\n",
    "    \"Unnamed: 49\": \"Novaseq2 L002\",\n",
    "    \"Unnamed: 50\": \"Novaseq2 L003\",\n",
    "    \"Unnamed: 51\": \"Novaseq2 L004\",\n",
    "})\n",
    "\n",
    "experiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "subjective-medicine",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Experiment</th>\n",
       "      <th>mice</th>\n",
       "      <th>Fixation Box</th>\n",
       "      <th>Split-seq prep start date</th>\n",
       "      <th>Submit</th>\n",
       "      <th>Sublibrary</th>\n",
       "      <th>[Barcoded Nuclei] (nuclei/uL)</th>\n",
       "      <th>Barcoded Nuclei Vol. (uL)</th>\n",
       "      <th>Total barcoded nuclei (1st 3 rounds)</th>\n",
       "      <th># of backup aliquots</th>\n",
       "      <th>...</th>\n",
       "      <th>Unnamed: 54</th>\n",
       "      <th>Unnamed: 55</th>\n",
       "      <th>Unnamed: 56</th>\n",
       "      <th>Unnamed: 57</th>\n",
       "      <th>Unnamed: 58</th>\n",
       "      <th>Unnamed: 59</th>\n",
       "      <th>Unnamed: 60</th>\n",
       "      <th>Unnamed: 61</th>\n",
       "      <th>Unnamed: 62</th>\n",
       "      <th>Unnamed: 63</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IGVF_Splitseq_002</td>\n",
       "      <td>B6 (2F+2M)</td>\n",
       "      <td>IGVF_FIX_001</td>\n",
       "      <td>2022-08-24</td>\n",
       "      <td>No</td>\n",
       "      <td>002_4A</td>\n",
       "      <td>4540</td>\n",
       "      <td>240.0</td>\n",
       "      <td>1089600</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>IGVF_Splitseq_003</td>\n",
       "      <td>8 founders (4M+4F)</td>\n",
       "      <td>IGVF_FIX_003</td>\n",
       "      <td>2022-12-01</td>\n",
       "      <td>No</td>\n",
       "      <td>003_8A</td>\n",
       "      <td>6860</td>\n",
       "      <td>250.0</td>\n",
       "      <td>1715000</td>\n",
       "      <td>7.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>IGVF_Splitseq_004</td>\n",
       "      <td>8 founders (4M+4F)</td>\n",
       "      <td>IGVF_FIX_005</td>\n",
       "      <td>2022-12-15</td>\n",
       "      <td>No</td>\n",
       "      <td>004_8A</td>\n",
       "      <td>2460</td>\n",
       "      <td>640.0</td>\n",
       "      <td>1574400</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>IGVF_Splitseq_005</td>\n",
       "      <td>8 founders (4M+4F)</td>\n",
       "      <td>IGVF_FIX_010</td>\n",
       "      <td>2023-01-23</td>\n",
       "      <td>No</td>\n",
       "      <td>005_8A</td>\n",
       "      <td>4675</td>\n",
       "      <td>390.0</td>\n",
       "      <td>1823250</td>\n",
       "      <td>6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>IGVF_Splitseq_006</td>\n",
       "      <td>8 founders (4M+4F)</td>\n",
       "      <td>IGVF_FIX_006</td>\n",
       "      <td>2023-02-10</td>\n",
       "      <td>No</td>\n",
       "      <td>006_8A</td>\n",
       "      <td>3160</td>\n",
       "      <td>780.0</td>\n",
       "      <td>2464800</td>\n",
       "      <td>8.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>IGVF_Splitseq_007</td>\n",
       "      <td>8 founders (4M+4F)</td>\n",
       "      <td>IGVF_FIX_006</td>\n",
       "      <td>2023-03-07</td>\n",
       "      <td>No</td>\n",
       "      <td>007_8A</td>\n",
       "      <td>2810</td>\n",
       "      <td>580.0</td>\n",
       "      <td>1629800</td>\n",
       "      <td>8.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>IGVF_Splitseq_008</td>\n",
       "      <td>8 founders (4M+4F)</td>\n",
       "      <td>IGVF_FIX_008</td>\n",
       "      <td>2023-03-24</td>\n",
       "      <td>No</td>\n",
       "      <td>008_8A</td>\n",
       "      <td>2600</td>\n",
       "      <td>250.0</td>\n",
       "      <td>650000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>IGVF_Splitseq_B01</td>\n",
       "      <td>CASTJ + B6J (2M+2F)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-04-03</td>\n",
       "      <td>Yes</td>\n",
       "      <td>B01_13A</td>\n",
       "      <td>4920</td>\n",
       "      <td>83.6</td>\n",
       "      <td>411312</td>\n",
       "      <td>10.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>IGVF_Splitseq_008B</td>\n",
       "      <td>8 founders (4M+4F)</td>\n",
       "      <td>IGVF_FIX_008</td>\n",
       "      <td>2023-04-19</td>\n",
       "      <td>Yes</td>\n",
       "      <td>008B_67A</td>\n",
       "      <td>3330</td>\n",
       "      <td>250.0</td>\n",
       "      <td>832500</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1417529714</td>\n",
       "      <td>355914199</td>\n",
       "      <td>355917666</td>\n",
       "      <td>358863075</td>\n",
       "      <td>346834774</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>IGVF_Splitseq_009</td>\n",
       "      <td>8 founders (4M+4F)</td>\n",
       "      <td>IGVF_FIX_004</td>\n",
       "      <td>2023-05-02</td>\n",
       "      <td>Yes</td>\n",
       "      <td>009_67A</td>\n",
       "      <td>3370</td>\n",
       "      <td>285.0</td>\n",
       "      <td>960450</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1474101220</td>\n",
       "      <td>177858630</td>\n",
       "      <td>182128617</td>\n",
       "      <td>178436907</td>\n",
       "      <td>177628982</td>\n",
       "      <td>189105418</td>\n",
       "      <td>193500915</td>\n",
       "      <td>191649075</td>\n",
       "      <td>183792676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>IGVF_Splitseq_010</td>\n",
       "      <td>8 founders (4M+4F)</td>\n",
       "      <td>IGVF_FIX_011</td>\n",
       "      <td>NaT</td>\n",
       "      <td>Yes</td>\n",
       "      <td>010_67A</td>\n",
       "      <td>5440</td>\n",
       "      <td>250.0</td>\n",
       "      <td>1360000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1276835447</td>\n",
       "      <td>173224540</td>\n",
       "      <td>171233145</td>\n",
       "      <td>172829595</td>\n",
       "      <td>168517796</td>\n",
       "      <td>145490904</td>\n",
       "      <td>150435757</td>\n",
       "      <td>148667669</td>\n",
       "      <td>146436041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>IGVF_Splitseq_011</td>\n",
       "      <td>8 founders (4M+4F)</td>\n",
       "      <td>IGVF_FIX_012</td>\n",
       "      <td>NaT</td>\n",
       "      <td>Yes</td>\n",
       "      <td>011_67A</td>\n",
       "      <td>6620</td>\n",
       "      <td>250.0</td>\n",
       "      <td>1655000</td>\n",
       "      <td>7.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1316269124</td>\n",
       "      <td>173623468</td>\n",
       "      <td>170663719</td>\n",
       "      <td>172876343</td>\n",
       "      <td>165573211</td>\n",
       "      <td>158569678</td>\n",
       "      <td>162211611</td>\n",
       "      <td>156551618</td>\n",
       "      <td>156199476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>IGVF_Splitseq_EX</td>\n",
       "      <td>8 founders (4M+4F)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>Yes</td>\n",
       "      <td>003_13A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>IGVF_Splitseq_012</td>\n",
       "      <td>B6J</td>\n",
       "      <td>IGVF_Repeats</td>\n",
       "      <td>NaT</td>\n",
       "      <td>Yes</td>\n",
       "      <td>012_13A</td>\n",
       "      <td>3220</td>\n",
       "      <td>81.0</td>\n",
       "      <td>260820</td>\n",
       "      <td>12.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>IGVF_Splitseq_013</td>\n",
       "      <td>8 founders + 4 F1s (4M+4F)</td>\n",
       "      <td>IGVF_FIX_021</td>\n",
       "      <td>2023-09-08</td>\n",
       "      <td>Yes</td>\n",
       "      <td>013_13A</td>\n",
       "      <td>4160</td>\n",
       "      <td>225.0</td>\n",
       "      <td>936000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>IGVF_Splitseq_014</td>\n",
       "      <td>B6J</td>\n",
       "      <td>IGVF_FIX_021</td>\n",
       "      <td>2023-09-25</td>\n",
       "      <td>Yes</td>\n",
       "      <td>014_13A</td>\n",
       "      <td>1040</td>\n",
       "      <td>97.5</td>\n",
       "      <td>101400</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>IGVF_Splitseq_015</td>\n",
       "      <td>8 F1s (4M+4F)</td>\n",
       "      <td>IGVF_FIX_013</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>015_13A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>250.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253</th>\n",
       "      <td>IGVF_Splitseq_016</td>\n",
       "      <td>8 F1s (4M+4F)</td>\n",
       "      <td>IGVF_FIX_015</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>016_13A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>250.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271</th>\n",
       "      <td>IGVF_Splitseq_017</td>\n",
       "      <td>8 F1s (4M+4F)</td>\n",
       "      <td>IGVF_FIX_014</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>017_13A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>250.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289</th>\n",
       "      <td>IGVF_Splitseq_018</td>\n",
       "      <td>8 F1s (4M+4F)</td>\n",
       "      <td>IGVF_FIX_012</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>018_13A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>250.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>307</th>\n",
       "      <td>IGVF_Splitseq_019</td>\n",
       "      <td>8 F1s (4M+4F)</td>\n",
       "      <td>IGVF_FIX_029</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>019_13A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>250.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325</th>\n",
       "      <td>IGVF_Splitseq_020</td>\n",
       "      <td>8 F1s (4M+4F)</td>\n",
       "      <td>IGVF_FIX_016</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>020_13A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>250.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>343</th>\n",
       "      <td>IGVF_Splitseq_021</td>\n",
       "      <td>8 F1s (4M+4F)</td>\n",
       "      <td>IGVF_FIX_017</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>021_13A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>250.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>IGVF_Splitseq_022</td>\n",
       "      <td>8 F1s (4M+4F)</td>\n",
       "      <td>IGVF_FIX_031</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>022_13A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>250.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Experiment                         mice  Fixation Box  \\\n",
       "0     IGVF_Splitseq_002                   B6 (2F+2M)  IGVF_FIX_001   \n",
       "17    IGVF_Splitseq_003           8 founders (4M+4F)  IGVF_FIX_003   \n",
       "34    IGVF_Splitseq_004           8 founders (4M+4F)  IGVF_FIX_005   \n",
       "51    IGVF_Splitseq_005           8 founders (4M+4F)  IGVF_FIX_010   \n",
       "68    IGVF_Splitseq_006           8 founders (4M+4F)  IGVF_FIX_006   \n",
       "87    IGVF_Splitseq_007           8 founders (4M+4F)  IGVF_FIX_006   \n",
       "105   IGVF_Splitseq_008           8 founders (4M+4F)  IGVF_FIX_008   \n",
       "117   IGVF_Splitseq_B01          CASTJ + B6J (2M+2F)           NaN   \n",
       "126  IGVF_Splitseq_008B           8 founders (4M+4F)  IGVF_FIX_008   \n",
       "137   IGVF_Splitseq_009           8 founders (4M+4F)  IGVF_FIX_004   \n",
       "153   IGVF_Splitseq_010           8 founders (4M+4F)  IGVF_FIX_011   \n",
       "169   IGVF_Splitseq_011           8 founders (4M+4F)  IGVF_FIX_012   \n",
       "185    IGVF_Splitseq_EX           8 founders (4M+4F)           NaN   \n",
       "196   IGVF_Splitseq_012                          B6J  IGVF_Repeats   \n",
       "208   IGVF_Splitseq_013  8 founders + 4 F1s (4M+4F)   IGVF_FIX_021   \n",
       "225   IGVF_Splitseq_014                          B6J  IGVF_FIX_021   \n",
       "235   IGVF_Splitseq_015                8 F1s (4M+4F)  IGVF_FIX_013   \n",
       "253   IGVF_Splitseq_016                8 F1s (4M+4F)  IGVF_FIX_015   \n",
       "271   IGVF_Splitseq_017                8 F1s (4M+4F)  IGVF_FIX_014   \n",
       "289   IGVF_Splitseq_018                8 F1s (4M+4F)  IGVF_FIX_012   \n",
       "307   IGVF_Splitseq_019                8 F1s (4M+4F)  IGVF_FIX_029   \n",
       "325   IGVF_Splitseq_020                8 F1s (4M+4F)  IGVF_FIX_016   \n",
       "343   IGVF_Splitseq_021                8 F1s (4M+4F)  IGVF_FIX_017   \n",
       "361   IGVF_Splitseq_022                8 F1s (4M+4F)  IGVF_FIX_031   \n",
       "\n",
       "    Split-seq prep start date Submit Sublibrary [Barcoded Nuclei] (nuclei/uL)  \\\n",
       "0                  2022-08-24     No     002_4A                          4540   \n",
       "17                 2022-12-01     No     003_8A                          6860   \n",
       "34                 2022-12-15     No     004_8A                          2460   \n",
       "51                 2023-01-23     No     005_8A                          4675   \n",
       "68                 2023-02-10     No     006_8A                          3160   \n",
       "87                 2023-03-07     No     007_8A                          2810   \n",
       "105                2023-03-24     No     008_8A                          2600   \n",
       "117                2023-04-03    Yes    B01_13A                          4920   \n",
       "126                2023-04-19    Yes   008B_67A                          3330   \n",
       "137                2023-05-02    Yes    009_67A                          3370   \n",
       "153                       NaT    Yes    010_67A                          5440   \n",
       "169                       NaT    Yes    011_67A                          6620   \n",
       "185                       NaT    Yes    003_13A                           NaN   \n",
       "196                       NaT    Yes    012_13A                          3220   \n",
       "208                2023-09-08    Yes    013_13A                          4160   \n",
       "225                2023-09-25    Yes    014_13A                          1040   \n",
       "235                       NaT    NaN    015_13A                           NaN   \n",
       "253                       NaT    NaN    016_13A                           NaN   \n",
       "271                       NaT    NaN    017_13A                           NaN   \n",
       "289                       NaT    NaN    018_13A                           NaN   \n",
       "307                       NaT    NaN    019_13A                           NaN   \n",
       "325                       NaT    NaN    020_13A                           NaN   \n",
       "343                       NaT    NaN    021_13A                           NaN   \n",
       "361                       NaT    NaN    022_13A                           NaN   \n",
       "\n",
       "     Barcoded Nuclei Vol. (uL) Total barcoded nuclei (1st 3 rounds)  \\\n",
       "0                        240.0                              1089600   \n",
       "17                       250.0                              1715000   \n",
       "34                       640.0                              1574400   \n",
       "51                       390.0                              1823250   \n",
       "68                       780.0                              2464800   \n",
       "87                       580.0                              1629800   \n",
       "105                      250.0                               650000   \n",
       "117                       83.6                               411312   \n",
       "126                      250.0                               832500   \n",
       "137                      285.0                               960450   \n",
       "153                      250.0                              1360000   \n",
       "169                      250.0                              1655000   \n",
       "185                        NaN                                  NaN   \n",
       "196                       81.0                               260820   \n",
       "208                      225.0                               936000   \n",
       "225                       97.5                               101400   \n",
       "235                      250.0                                  NaN   \n",
       "253                      250.0                                  NaN   \n",
       "271                      250.0                                  NaN   \n",
       "289                      250.0                                  NaN   \n",
       "307                      250.0                                  NaN   \n",
       "325                      250.0                                  NaN   \n",
       "343                      250.0                                  NaN   \n",
       "361                      250.0                                  NaN   \n",
       "\n",
       "     # of backup aliquots  ... Unnamed: 54  Unnamed: 55  Unnamed: 56  \\\n",
       "0                     3.0  ...         NaN          NaN          NaN   \n",
       "17                    7.0  ...         NaN        Total         L001   \n",
       "34                    5.0  ...         NaN        Total         L001   \n",
       "51                    6.0  ...         NaN        Total         L001   \n",
       "68                    8.0  ...         NaN        Total         L001   \n",
       "87                    8.0  ...         NaN        Total         L001   \n",
       "105                   0.0  ...         NaN        Total         L001   \n",
       "117                  10.0  ...         NaN          NaN          NaN   \n",
       "126                   0.0  ...           2   1417529714    355914199   \n",
       "137                   0.0  ...           2   1474101220    177858630   \n",
       "153                   3.0  ...           2   1276835447    173224540   \n",
       "169                   7.0  ...           2   1316269124    173623468   \n",
       "185                   1.0  ...         NaN          NaN          NaN   \n",
       "196                  12.0  ...         NaN          NaN          NaN   \n",
       "208                   0.0  ...         NaN        Total         L001   \n",
       "225                   0.0  ...         NaN          NaN          NaN   \n",
       "235                   NaN  ...         NaN        Total         L001   \n",
       "253                   NaN  ...         NaN        Total         L001   \n",
       "271                   NaN  ...         NaN        Total         L001   \n",
       "289                   NaN  ...         NaN        Total         L001   \n",
       "307                   NaN  ...         NaN        Total         L001   \n",
       "325                   NaN  ...         NaN        Total         L001   \n",
       "343                   NaN  ...         NaN        Total         L001   \n",
       "361                   NaN  ...         NaN        Total         L001   \n",
       "\n",
       "     Unnamed: 57 Unnamed: 58 Unnamed: 59 Unnamed: 60 Unnamed: 61  Unnamed: 62  \\\n",
       "0            NaN         NaN         NaN         NaN         NaN          NaN   \n",
       "17          L002        L003        L004        L001        L002         L003   \n",
       "34          L002        L003        L004        L001        L002         L003   \n",
       "51          L002        L003        L004        L001        L002         L003   \n",
       "68          L002        L003        L004        L001        L002         L003   \n",
       "87          L002        L003        L004        L001        L002         L003   \n",
       "105         L002        L003        L004         NaN         NaN          NaN   \n",
       "117          NaN         NaN         NaN         NaN         NaN          NaN   \n",
       "126    355917666   358863075   346834774         NaN         NaN          NaN   \n",
       "137    182128617   178436907   177628982   189105418   193500915    191649075   \n",
       "153    171233145   172829595   168517796   145490904   150435757    148667669   \n",
       "169    170663719   172876343   165573211   158569678   162211611    156551618   \n",
       "185          NaN         NaN         NaN         NaN         NaN          NaN   \n",
       "196          NaN         NaN         NaN         NaN         NaN          NaN   \n",
       "208         L002        L003        L004         NaN         NaN          NaN   \n",
       "225          NaN         NaN         NaN         NaN         NaN          NaN   \n",
       "235         L002        L003        L004        L001        L002         L003   \n",
       "253         L002        L003        L004        L001        L002         L003   \n",
       "271         L002        L003        L004        L001        L002         L003   \n",
       "289         L002        L003        L004        L001        L002         L003   \n",
       "307         L002        L003        L004        L001        L002         L003   \n",
       "325         L002        L003        L004        L001        L002         L003   \n",
       "343         L002        L003        L004        L001        L002         L003   \n",
       "361         L002        L003        L004        L001        L002         L003   \n",
       "\n",
       "     Unnamed: 63  \n",
       "0            NaN  \n",
       "17          L004  \n",
       "34          L004  \n",
       "51          L004  \n",
       "68          L004  \n",
       "87          L004  \n",
       "105          NaN  \n",
       "117          NaN  \n",
       "126          NaN  \n",
       "137    183792676  \n",
       "153    146436041  \n",
       "169    156199476  \n",
       "185          NaN  \n",
       "196          NaN  \n",
       "208          NaN  \n",
       "225          NaN  \n",
       "235         L004  \n",
       "253         L004  \n",
       "271         L004  \n",
       "289         L004  \n",
       "307         L004  \n",
       "325         L004  \n",
       "343         L004  \n",
       "361         L004  \n",
       "\n",
       "[24 rows x 64 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def is_experiment_name(name):\n",
    "    not_submittable = [\"IGVF_Splitseq_EX\"]\n",
    "    return not pandas.isnull(name) and name.startswith(\"IGVF_Splitseq\")\n",
    "\n",
    "experiment[experiment[\"Experiment\"].apply(is_experiment_name) & (pandas.notnull(experiment[\"mice\"]))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "gothic-robert",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IGVF_Splitseq_002 0 15\n",
      "IGVF_Splitseq_003 17 32\n",
      "IGVF_Splitseq_004 34 49\n",
      "IGVF_Splitseq_005 51 66\n",
      "IGVF_Splitseq_006 68 85\n",
      "IGVF_Splitseq_007 87 102\n",
      "IGVF_Splitseq_008 105 114\n",
      "IGVF_Splitseq_B01 117 124\n",
      "IGVF_Splitseq_008B 126 135\n",
      "IGVF_Splitseq_009 137 151\n",
      "IGVF_Splitseq_010 153 167\n",
      "IGVF_Splitseq_011 169 183\n",
      "IGVF_Splitseq_EX 185 192\n",
      "IGVF_Splitseq_012 196 203\n",
      "IGVF_Splitseq_013 208 222\n",
      "IGVF_Splitseq_014 225 232\n",
      "IGVF_Splitseq_015 235 250\n",
      "IGVF_Splitseq_016 253 268\n",
      "IGVF_Splitseq_017 271 286\n",
      "IGVF_Splitseq_018 289 304\n",
      "IGVF_Splitseq_019 307 322\n",
      "IGVF_Splitseq_020 325 340\n",
      "IGVF_Splitseq_021 343 358\n",
      "IGVF_Splitseq_022 361 376\n"
     ]
    }
   ],
   "source": [
    "def find_fixation_start(sheet):\n",
    "    def is_experiment_name(name):\n",
    "        not_submittable = [\"IGVF_Splitseq_EX\"]\n",
    "        return not pandas.isnull(name) and name.startswith(\"IGVF_Splitseq\") # and not name in not_submittable\n",
    "\n",
    "    for fixation_id_row in sheet[sheet[\"Experiment\"].apply(is_experiment_name) & (pandas.notnull(experiment[\"mice\"]))].index:\n",
    "        name = sheet.loc[fixation_id_row, \"Experiment\"]\n",
    "        start = fixation_id_row\n",
    "\n",
    "        stop = sheet.shape[0]\n",
    "        for i, value in sheet.iloc[start:][\"Sublibrary\"].iteritems():\n",
    "            if pandas.isnull(value):\n",
    "                stop = i - 1\n",
    "                break\n",
    "        \n",
    "        \n",
    "        yield (name, start, stop)\n",
    "\n",
    "for name, start, stop in find_fixation_start(experiment):\n",
    "    print(name, start, stop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "naval-muslim",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EX 004_13A\n",
      "EX 005_13A\n",
      "EX 007_13A\n",
      "EX 008B_13A\n",
      "EX 009_13A\n",
      "EX 010_13A\n",
      "EX 011_13A\n",
      "NO 012_13A\n",
      "NO 012_13B\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'experiment_name': 'IGVF_Splitseq_EX',\n",
       " 'plate_name': 'IGVF_EX',\n",
       " 'prep_date': NaT,\n",
       " 'unused_aliquots': {13000: 12},\n",
       " 'subpool': [{'name': '004_13A',\n",
       "   'plate_name': 'IGVF_004',\n",
       "   'subpool_subname': '13A',\n",
       "   'submit': 'Yes',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionTypeEnum.exome_capture: 'EX'>,\n",
       "   'subcellular_component': <SubcellularComponentEnum.nuclei: 'N'>,\n",
       "   'cdna_pcr_rounds': '5 + 8',\n",
       "   'cdna_ng_per_ul': 23.8,\n",
       "   'cdna_volume': 25.0,\n",
       "   'bioanalyzer_date': NaT,\n",
       "   'cdna_average_bp_length': None,\n",
       "   'index_pcr_no': 12,\n",
       "   'index': 2,\n",
       "   'i7_barcode': 'ACTTGA',\n",
       "   'i5_barcode': nan,\n",
       "   'library_ng_per_ul': 24.4,\n",
       "   'library_average_bp_length': None,\n",
       "   'nextseq_run_date': datetime.date(2023, 5, 30),\n",
       "   'nextseq_run_id': 1070,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None},\n",
       "  {'name': '005_13A',\n",
       "   'plate_name': 'IGVF_005',\n",
       "   'subpool_subname': '13A',\n",
       "   'submit': 'Yes',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionTypeEnum.exome_capture: 'EX'>,\n",
       "   'subcellular_component': <SubcellularComponentEnum.nuclei: 'N'>,\n",
       "   'cdna_pcr_rounds': '5 + 8',\n",
       "   'cdna_ng_per_ul': 52.0,\n",
       "   'cdna_volume': 25.0,\n",
       "   'bioanalyzer_date': NaT,\n",
       "   'cdna_average_bp_length': None,\n",
       "   'index_pcr_no': 12,\n",
       "   'index': 3,\n",
       "   'i7_barcode': 'GATCAG',\n",
       "   'i5_barcode': nan,\n",
       "   'library_ng_per_ul': 19.8,\n",
       "   'library_average_bp_length': None,\n",
       "   'nextseq_run_date': datetime.date(2023, 5, 30),\n",
       "   'nextseq_run_id': 1070,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None},\n",
       "  {'name': '007_13A',\n",
       "   'plate_name': 'IGVF_007',\n",
       "   'subpool_subname': '13A',\n",
       "   'submit': 'Yes',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionTypeEnum.exome_capture: 'EX'>,\n",
       "   'subcellular_component': <SubcellularComponentEnum.nuclei: 'N'>,\n",
       "   'cdna_pcr_rounds': '5 + 8',\n",
       "   'cdna_ng_per_ul': 62.4,\n",
       "   'cdna_volume': 25.0,\n",
       "   'bioanalyzer_date': NaT,\n",
       "   'cdna_average_bp_length': None,\n",
       "   'index_pcr_no': 12,\n",
       "   'index': 4,\n",
       "   'i7_barcode': 'TAGCTT',\n",
       "   'i5_barcode': nan,\n",
       "   'library_ng_per_ul': 17.5,\n",
       "   'library_average_bp_length': None,\n",
       "   'nextseq_run_date': datetime.date(2023, 5, 30),\n",
       "   'nextseq_run_id': 1070,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None},\n",
       "  {'name': '008B_13A',\n",
       "   'plate_name': 'IGVF_008B',\n",
       "   'subpool_subname': '13A',\n",
       "   'submit': 'Yes',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionTypeEnum.exome_capture: 'EX'>,\n",
       "   'subcellular_component': <SubcellularComponentEnum.nuclei: 'N'>,\n",
       "   'cdna_pcr_rounds': '5 + 8',\n",
       "   'cdna_ng_per_ul': 57.6,\n",
       "   'cdna_volume': 25.0,\n",
       "   'bioanalyzer_date': NaT,\n",
       "   'cdna_average_bp_length': None,\n",
       "   'index_pcr_no': 12,\n",
       "   'index': 5,\n",
       "   'i7_barcode': 'ATGTCA',\n",
       "   'i5_barcode': nan,\n",
       "   'library_ng_per_ul': 16.8,\n",
       "   'library_average_bp_length': None,\n",
       "   'nextseq_run_date': datetime.date(2023, 6, 2),\n",
       "   'nextseq_run_id': 1071,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None},\n",
       "  {'name': '009_13A',\n",
       "   'plate_name': 'IGVF_009',\n",
       "   'subpool_subname': '13A',\n",
       "   'submit': 'Yes',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionTypeEnum.exome_capture: 'EX'>,\n",
       "   'subcellular_component': <SubcellularComponentEnum.nuclei: 'N'>,\n",
       "   'cdna_pcr_rounds': '5 + 8',\n",
       "   'cdna_ng_per_ul': 84.5,\n",
       "   'cdna_volume': 25.0,\n",
       "   'bioanalyzer_date': NaT,\n",
       "   'cdna_average_bp_length': None,\n",
       "   'index_pcr_no': 12,\n",
       "   'index': 6,\n",
       "   'i7_barcode': 'CTTGTA',\n",
       "   'i5_barcode': nan,\n",
       "   'library_ng_per_ul': 20.8,\n",
       "   'library_average_bp_length': None,\n",
       "   'nextseq_run_date': datetime.date(2023, 6, 2),\n",
       "   'nextseq_run_id': 1071,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None},\n",
       "  {'name': '010_13A',\n",
       "   'plate_name': 'IGVF_010',\n",
       "   'subpool_subname': '13A',\n",
       "   'submit': 'Yes',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionTypeEnum.exome_capture: 'EX'>,\n",
       "   'subcellular_component': <SubcellularComponentEnum.nuclei: 'N'>,\n",
       "   'cdna_pcr_rounds': '5 + 8',\n",
       "   'cdna_ng_per_ul': 25.8,\n",
       "   'cdna_volume': 25.0,\n",
       "   'bioanalyzer_date': NaT,\n",
       "   'cdna_average_bp_length': None,\n",
       "   'index_pcr_no': 12,\n",
       "   'index': 7,\n",
       "   'i7_barcode': 'AGTCAA',\n",
       "   'i5_barcode': nan,\n",
       "   'library_ng_per_ul': 27.6,\n",
       "   'library_average_bp_length': None,\n",
       "   'nextseq_run_date': datetime.date(2023, 6, 2),\n",
       "   'nextseq_run_id': 1071,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None},\n",
       "  {'name': '011_13A',\n",
       "   'plate_name': 'IGVF_011',\n",
       "   'subpool_subname': '13A',\n",
       "   'submit': 'Yes',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionTypeEnum.exome_capture: 'EX'>,\n",
       "   'subcellular_component': <SubcellularComponentEnum.nuclei: 'N'>,\n",
       "   'cdna_pcr_rounds': '5 + 8',\n",
       "   'cdna_ng_per_ul': 98.0,\n",
       "   'cdna_volume': 25.0,\n",
       "   'bioanalyzer_date': NaT,\n",
       "   'cdna_average_bp_length': None,\n",
       "   'index_pcr_no': 12,\n",
       "   'index': 8,\n",
       "   'i7_barcode': 'AGTTCC',\n",
       "   'i5_barcode': nan,\n",
       "   'library_ng_per_ul': 17.9,\n",
       "   'library_average_bp_length': None,\n",
       "   'nextseq_run_date': datetime.date(2023, 6, 2),\n",
       "   'nextseq_run_id': 1071,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None},\n",
       "  {'name': '012_13A',\n",
       "   'plate_name': 'IGVF_012',\n",
       "   'subpool_subname': '13A',\n",
       "   'submit': 'Yes',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionTypeEnum.no_selection: 'NO'>,\n",
       "   'subcellular_component': <SubcellularComponentEnum.nuclei: 'N'>,\n",
       "   'cdna_pcr_rounds': '5 + 8',\n",
       "   'cdna_ng_per_ul': 72.1,\n",
       "   'cdna_volume': 20.0,\n",
       "   'bioanalyzer_date': Timestamp('2023-06-13 00:00:00'),\n",
       "   'cdna_average_bp_length': 1368,\n",
       "   'index_pcr_no': 11,\n",
       "   'index': 1,\n",
       "   'i7_barcode': 'CAGATC',\n",
       "   'i5_barcode': nan,\n",
       "   'library_ng_per_ul': 19.2,\n",
       "   'library_average_bp_length': 446.0,\n",
       "   'nextseq_run_date': datetime.date(2023, 6, 16),\n",
       "   'nextseq_run_id': 1072,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None},\n",
       "  {'name': '012_13B',\n",
       "   'plate_name': 'IGVF_012',\n",
       "   'subpool_subname': '13B',\n",
       "   'submit': 'Yes',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionTypeEnum.no_selection: 'NO'>,\n",
       "   'subcellular_component': <SubcellularComponentEnum.nuclei: 'N'>,\n",
       "   'cdna_pcr_rounds': '5 + 8',\n",
       "   'cdna_ng_per_ul': 95.2,\n",
       "   'cdna_volume': 20.0,\n",
       "   'bioanalyzer_date': Timestamp('2023-06-13 00:00:00'),\n",
       "   'cdna_average_bp_length': 1458,\n",
       "   'index_pcr_no': 11,\n",
       "   'index': 2,\n",
       "   'i7_barcode': 'ACTTGA',\n",
       "   'i5_barcode': nan,\n",
       "   'library_ng_per_ul': 14.8,\n",
       "   'library_average_bp_length': 451.0,\n",
       "   'nextseq_run_date': datetime.date(2023, 6, 16),\n",
       "   'nextseq_run_id': 1072,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None}]}"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def parse_selection_type(value):\n",
    "    if value is None:\n",
    "        return None\n",
    "    elif value.lower() in (\"normal\", \"no\"):\n",
    "        return models.SublibrarySelectionTypeEnum.no_selection\n",
    "    elif value in (\"capture\", \"EX\"):\n",
    "        return models.SublibrarySelectionTypeEnum.exome_capture\n",
    "    else:\n",
    "        raise ValueError(f\"Unrecognized capture type {value}\")\n",
    "        \n",
    "def parse_sample_type(value):\n",
    "    if value is None:\n",
    "        return None\n",
    "    elif value.lower() in (\"nuclei\",):\n",
    "        return models.SubcellularComponentEnum.nuclei\n",
    "    elif value.lower() in (\"cells\",):\n",
    "        return models.SubcellularComponentEnum.cellular\n",
    "    else:\n",
    "        raise ValueError(f\"Unrecognized sample type {value}\")\n",
    "\n",
    "\n",
    "def parse_fixation(sheet, experiment_name, start, stop):\n",
    "    block = sheet.loc[start:stop]\n",
    "\n",
    "    experiment_name_fields = experiment_name.split(\"_\")\n",
    "    plate_name = f\"{experiment_name_fields[0]}_{experiment_name_fields[-1]}\"\n",
    "    \n",
    "    aliquots = {}\n",
    "    for i, row in block.iterrows():\n",
    "        count = row[\"# of backup aliquots\"]\n",
    "        size = row[\"Backup Aliquot Size\"]\n",
    "        if not (pandas.isnull(count) or pandas.isnull(size)):\n",
    "            if isinstance(size, str) and size.endswith(\"k\"):\n",
    "                size = size[:-1]\n",
    "            size = int(size) * 1000\n",
    "            aliquots[size] = int(count)\n",
    "    \n",
    "    experiment = {\n",
    "        \"experiment_name\": experiment_name,\n",
    "        \"plate_name\": plate_name,\n",
    "        \"prep_date\": block.loc[start, \"Split-seq prep start date\"],\n",
    "        \"unused_aliquots\": aliquots,\n",
    "        \"subpool\": [],\n",
    "    }\n",
    "\n",
    "    for i, row in block.iterrows():\n",
    "        if not pandas.isnull(row[\"Sublibrary\"]):\n",
    "            if pandas.isnull(row[\"NovaSeq Sequencing date\"]):\n",
    "                novaseq_run_id = None\n",
    "            else:\n",
    "                novaseq_run_id = row[\"NovaSeq Sequencing date\"].isoformat().split(\"T\")[0]\n",
    "\n",
    "            if pandas.isnull(row.get(\"Novaseq raw reads\")):\n",
    "                novaseq_raw_reads = None\n",
    "                novaseq1_raw_reads = None\n",
    "                novaseq2_raw_reads = None\n",
    "            elif row[\"Novaseq raw reads\"] == \"Total\":\n",
    "                novaseq_raw_reads = None\n",
    "                novaseq1_raw_reads = None\n",
    "                novaseq2_raw_reads = None\n",
    "            else:\n",
    "                novaseq_raw_reads = int_or_none(row.get(\"Novaseq raw reads\"))\n",
    "                novaseq1_raw_reads = sum([int_or_none(row.get(x)) for x in[\"Novaseq1 L001\", \"Novaseq1 L002\", \"Novaseq1 L003\", \"Novaseq1 L004\"]])\n",
    "                novaseq2_raw_reads = sum([int_or_none(row.get(x)) for x in[\"Novaseq2 L001\", \"Novaseq2 L002\", \"Novaseq2 L003\", \"Novaseq2 L004\"]])\n",
    "\n",
    "            plate_name, subname = row[\"Sublibrary\"].split(\"_\")\n",
    "            print(row[\"Sublibrary Type\"], row[\"Sublibrary\"])\n",
    "            subpool = {\n",
    "                \"name\": row[\"Sublibrary\"],\n",
    "                \"plate_name\": \"IGVF_{}\".format(plate_name),\n",
    "                \"subpool_subname\": subname,\n",
    "                \"submit\": row[\"Submit\"],\n",
    "                \"nuclei\": int_or_none(row[\"Nuclei / Sublibrary\"]),\n",
    "                \"selection_type\": parse_selection_type(row[\"Sublibrary Type\"]),\n",
    "                \"subcellular_component\": parse_sample_type(row[\"Sample Type\"]),\n",
    "                \"cdna_pcr_rounds\": row[\"cDNA Amp # PCR Cycles\"],\n",
    "                \"cdna_ng_per_ul\": row[\"[cDNA] (ng/uL)\"],\n",
    "                \"cdna_volume\": row[\"cDNA Vol. (uL)\"],\n",
    "                #total_cdna = cdna_ng_per_ul_in_25ul * 25\n",
    "                \"bioanalyzer_date\": row[\"BA date\"],\n",
    "                \"cdna_average_bp_length\": int_or_none(row[\"cDNA Ave. Length (bp)\"]),\n",
    "                \"index_pcr_no\": int_or_none(row[\"Sub library Index PCR #\"]),\n",
    "                \"index\": str_or_none(row[\"Sublibrary Index\"]),\n",
    "                \"i7_barcode\": row[\"i7 Forward Sequence\"],\n",
    "                \"i5_barcode\": row[\"i5 Forward Sequence\"],\n",
    "                \"library_ng_per_ul\": float_or_none(row[\"[Library] (ng/ul)\"]),\n",
    "                \"library_average_bp_length\": float_or_none(row[\"Library Ave. Length (bp)\"]),\n",
    "                \"nextseq_run_date\": date_or_none(row[\"QC Sequencing date\"]),\n",
    "                \"nextseq_run_id\": str_or_none(int_or_none(row[\"Nextseq run #\"])),\n",
    "                \"nextseq_run_raw_reads\": int_or_none(row[\"QC # raw reads  (1 mismatch)\"]),\n",
    "                \"novaseq_run_date\": date_or_none(row[\"NovaSeq Sequencing date\"]),\n",
    "                \"novaseq_run_id\": novaseq_run_id,\n",
    "                \"novaseq_run_raw_reads\": novaseq_raw_reads,\n",
    "                \"novaseq1_raw_reads\": novaseq1_raw_reads,\n",
    "                \"novaseq2_raw_reads\": novaseq2_raw_reads,\n",
    "            }\n",
    "            experiment[\"subpool\"].append(subpool)\n",
    "            #if not pandas.isnull(subpool[\"nuclei\"]):\n",
    "            #    nuclei = subpool[\"nuclei\"]\n",
    "            #    aliquots[nuclei] = aliquots.get(nuclei) + 1\n",
    "\n",
    "    return experiment\n",
    "\n",
    "#parse_fixation(experiment, \"IGVF_Splitseq_003\", 17, 32)\n",
    "#parse_fixation(experiment, \"IGVF_Splitseq_008\", 108, 117)\n",
    "#parse_fixation(experiment, \"IGVF_Splitseq_B01\", 124, 131)\n",
    "#parse_fixation(experiment, \"IGVF_Splitseq_008B\", 133, 143)\n",
    "parse_fixation(experiment, \"IGVF_Splitseq_EX\", 186, 197)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dress-serbia",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "centered-float",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'nextseq2000': <Platform: Nextseq 2000>,\n",
       " 'novaseq6000': <Platform: Novaseq 6000>,\n",
       " 'minion': <Platform: Oxford Nanopore MinION>,\n",
       " 'promethion': <Platform: Oxford Nanopore PromethION>,\n",
       " 'pacbio': <Platform: Pac Bio>}"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "platforms = {x.name: x for x in models.Platform.objects.all()}\n",
    "platforms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "express-zambia",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8ACZ\n",
      "decoded [8, 0, 2, 25]\n",
      "encoded 8ACZ\n",
      "8ADA\n",
      "8A 8B True False False\n",
      "8AA 8Z False False True\n",
      "8B 8E True False False\n",
      "8B 8A False False True\n",
      "8Z 8AA True False False\n",
      "67A 67A False True False\n",
      "8AA 8AB True False False\n",
      "8B\n",
      "8B\n",
      "8A\n"
     ]
    }
   ],
   "source": [
    "class SubpoolName:\n",
    "    def __init__(self, name=None):\n",
    "        if isinstance(name, str):\n",
    "            self._normalized = SubpoolName.subpool_name_to_numbers(name)\n",
    "        elif isinstance(name, list):\n",
    "            self._normalized = name\n",
    "        elif name is None:\n",
    "            self._normalized = None\n",
    "        else:\n",
    "            raise ValueError(\"Unrecognized type {} for {}\".format(type(name), name))\n",
    "\n",
    "    def __str__(self):\n",
    "        if self._normalized is None:\n",
    "            return \"None\"\n",
    "        else:\n",
    "            return SubpoolName.numbers_to_subpool_name(self._normalized)\n",
    "        \n",
    "    def __repr__(self):\n",
    "        if self._normalized is None:\n",
    "            return str(\"{}()\".format(self.__class__.__name__))\n",
    "        else:\n",
    "            return str(\"{}('{}')\".format(self.__class__.__name__, str(self)))\n",
    "        \n",
    "    def __eq__(self, cmp):\n",
    "        if self._normalized is None and cmp is None:\n",
    "            return True\n",
    "        elif self._normalized is None and cmp._normalized is None:\n",
    "            return True\n",
    "        elif self._normalized is None or cmp._normalized is None:\n",
    "            return False\n",
    "        elif len(self._normalized) != len(cmp._normalized):\n",
    "            return False\n",
    "        else:\n",
    "            for left, right in zip(self._normalized, cmp._normalized):\n",
    "                if left != right:\n",
    "                    return False\n",
    "                \n",
    "        return True\n",
    "    \n",
    "    def __lt__(self, cmp):\n",
    "        if self._normalized is None and cmp is None:\n",
    "            return False\n",
    "        elif self._normalized is None and cmp._normalized is None:\n",
    "            return False\n",
    "        elif self._normalized is None:\n",
    "            return True\n",
    "        elif cmp is None or cmp._normalized is None:\n",
    "            return False\n",
    "        elif self._normalized[0] < cmp._normalized[0]:\n",
    "            return True\n",
    "        elif len(self._normalized) < len(cmp._normalized):\n",
    "            return True\n",
    "        elif len(self._normalized) > len(cmp._normalized):\n",
    "            return False\n",
    "        else:\n",
    "            for left, right in zip(self._normalized, cmp._normalized):\n",
    "                if left == right:\n",
    "                    continue\n",
    "                elif left < right:\n",
    "                    return True\n",
    "                else:\n",
    "                    return False\n",
    "                \n",
    "        return False\n",
    "                \n",
    "    def __gt__(self, cmp):\n",
    "        if self._normalized is None and cmp is None:\n",
    "            return False\n",
    "        elif self._normalized is None and cmp._normalized is None:\n",
    "            return False\n",
    "        elif cmp is None or cmp._normalized is None:\n",
    "            return True\n",
    "        elif self._normalized[0] > cmp._normalized[0]:\n",
    "            return True\n",
    "        elif len(self._normalized) < len(cmp._normalized):\n",
    "            return False\n",
    "        elif len(self._normalized) > len(cmp._normalized):\n",
    "            return True\n",
    "        else:\n",
    "            for left, right in zip(self._normalized, cmp._normalized):\n",
    "                if left == right:\n",
    "                    continue\n",
    "                elif left < right:\n",
    "                    return False\n",
    "                else:\n",
    "                    return True\n",
    "            \n",
    "        return False\n",
    "        \n",
    "    @classmethod\n",
    "    def subpool_name_to_numbers(cls, name):\n",
    "        STATE_DIGITS = 1\n",
    "        STATE_LETTERS = 2\n",
    "        STATE_ERROR = -1\n",
    "\n",
    "        digits = []\n",
    "        letters = []\n",
    "        state = STATE_DIGITS\n",
    "        for char in name:\n",
    "            if state == STATE_DIGITS:\n",
    "                if char.isdigit():\n",
    "                    digits.append(char)\n",
    "                elif char.isalpha():\n",
    "                    state = STATE_LETTERS\n",
    "                    letters.append(char)\n",
    "                else:\n",
    "                    raise ValueError(\"Unrecognized symbol {} in {}\".format(char, name))\n",
    "            elif state == STATE_LETTERS:\n",
    "                if char.isdigit():\n",
    "                    raise ValueError(\"Unexpected digit {} in {}\".format(char, name))\n",
    "                elif char.isalpha():\n",
    "                    letters.append(char)\n",
    "                else:\n",
    "                    raise ValueError(\"Unrecognized symbol {} in {}\".format(char, name))\n",
    "\n",
    "        result = [int(\"\".join(digits))]\n",
    "        for number in letters:\n",
    "            if number < \"A\" or number > \"Z\":\n",
    "                raise ValueError(\"out of bounds letter code in {}\".format(name))\n",
    "            else:\n",
    "                result.append(ord(number) - ord(\"A\"))\n",
    "\n",
    "        return result\n",
    "\n",
    "    @classmethod\n",
    "    def numbers_to_subpool_name(cls, subpool_list):\n",
    "\n",
    "        result = [str(subpool_list[0])]\n",
    "        for char in subpool_list[1:]:\n",
    "            result.append(chr(char + ord(\"A\")))\n",
    "\n",
    "        return \"\".join(result)\n",
    "\n",
    "    def next_code(self):\n",
    "        if self._normalized is None:\n",
    "            return None\n",
    "\n",
    "        carry = 0\n",
    "        advancable = []\n",
    "        for i, place in enumerate(reversed(self._normalized[1:])):\n",
    "            if i == 0:\n",
    "                place += 1\n",
    "            else:\n",
    "                place += carry\n",
    "                carry = 0\n",
    "\n",
    "            if place > 25:\n",
    "                carry = 1\n",
    "                place = 0\n",
    "\n",
    "            advancable.append(place)\n",
    "\n",
    "        if carry > 0:\n",
    "            advancable.append(0)\n",
    "\n",
    "        result = [self._normalized[0]]\n",
    "        result.extend(reversed(advancable))\n",
    "        \n",
    "        return SubpoolName(result)\n",
    "\n",
    "s = SubpoolName(\"8ACZ\")\n",
    "print(s)\n",
    "\n",
    "decoded = SubpoolName.subpool_name_to_numbers(\"8ACZ\") \n",
    "print(\"decoded\", decoded)\n",
    "encoded = SubpoolName.numbers_to_subpool_name(decoded)\n",
    "print(\"encoded\", encoded)\n",
    "\n",
    "print(SubpoolName(\"8ACZ\").next_code())\n",
    "\n",
    "\n",
    "\n",
    "for pairs in [(\"8A\", \"8B\"), (\"8AA\", \"8Z\"), (\"8B\", \"8E\"), (\"8B\", \"8A\"), (\"8Z\", \"8AA\"), (\"67A\", \"67A\"), (\"8AA\", \"8AB\")]:\n",
    "    left = SubpoolName(pairs[0])\n",
    "    right = SubpoolName(pairs[1])\n",
    "    print(left, right, left < right, left == right, left > right)\n",
    "    \n",
    "print(max(SubpoolName(\"8A\"), SubpoolName(\"8B\")))\n",
    "print(max(None, SubpoolName(\"8B\")))\n",
    "print(max(SubpoolName(\"8A\"), None))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "parliamentary-pendant",
   "metadata": {},
   "source": [
    "# load subpools, sequencing runs and subpools in sequencing run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "hydraulic-nurse",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_submitted_multiplexed_samples():\n",
    "    multiplexed_samples = {}\n",
    "\n",
    "    for accession_prefix in submitted_sheets:\n",
    "        for i, row in submitted_sheets[accession_prefix][\"multiplexed_sample\"].iterrows():\n",
    "            if not pandas.isnull(row[\"accession\"]):\n",
    "                aliases = row[\"aliases:array\"].split(',')\n",
    "                name = aliases[0].replace(\"ali-mortazavi:\", \"\")\n",
    "                if accession_prefix == \"igvftst\":\n",
    "                    see_also_template = \"https://api.sandbox.igvf.org/multiplexed-samples/{}/\"\n",
    "                elif accession_prefix == \"igvf\":\n",
    "                    see_also_template = \"https://api.data.igvf.org/multiplexed-samples/{}/\"\n",
    "\n",
    "                multiplexed_samples.setdefault(name, []).append({\n",
    "                    \"accession_prefix\": accession_prefix,\n",
    "                    \"name\": row[\"accession\"],\n",
    "                    \"uuid\": row[\"uuid\"],\n",
    "                    \"see_also\": see_also_template.format(row[\"accession\"]),\n",
    "                })\n",
    "    print(len(multiplexed_samples))\n",
    "    \n",
    "    return multiplexed_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "assured-provincial",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_split_seq_plates_as_dict():\n",
    "    plates = {}\n",
    "    for current_plate in models.SplitSeqPlate.objects.all():\n",
    "        plates[current_plate.name] = current_plate\n",
    "    return plates\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "oriented-manual",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'IGVF_003': <SplitSeqPlate: IGVF_003>,\n",
       " 'IGVF_004': <SplitSeqPlate: IGVF_004>,\n",
       " 'IGVF_005': <SplitSeqPlate: IGVF_005>,\n",
       " 'IGVF_006': <SplitSeqPlate: IGVF_006>,\n",
       " 'IGVF_007': <SplitSeqPlate: IGVF_007>,\n",
       " 'IGVF_008': <SplitSeqPlate: IGVF_008>,\n",
       " 'IGVF_008B': <SplitSeqPlate: IGVF_008B>,\n",
       " 'IGVF_B01': <SplitSeqPlate: IGVF_B01>,\n",
       " 'IGVF_009': <SplitSeqPlate: IGVF_009>,\n",
       " 'IGVF_010': <SplitSeqPlate: IGVF_010>,\n",
       " 'IGVF_011': <SplitSeqPlate: IGVF_011>,\n",
       " 'IGVF_012': <SplitSeqPlate: IGVF_012>,\n",
       " 'IGVF_013': <SplitSeqPlate: IGVF_013>,\n",
       " 'IGVF_014': <SplitSeqPlate: IGVF_014>}"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "load_split_seq_plates_as_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "complete-wellington",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'_state': <django.db.models.base.ModelState at 0x7efdcf27ea00>,\n",
       "  'id': 202,\n",
       "  'reagent_id': 'wt-mega',\n",
       "  'name': '10',\n",
       "  'code': '10',\n",
       "  'i7_sequence': 'CCGTCC',\n",
       "  'i5_sequence': None,\n",
       "  'barcode_type': None,\n",
       "  'well_position': None}]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[vars(x) for x in models.LibraryBarcode.objects.filter(\n",
    "    reagent=guess_barcode_reagent_from_plate(\"IGVF_004\"),\n",
    "    barcode_type=None,\n",
    "    code=\"10\",\n",
    ")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "victorian-basics",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "171\n",
      "NO 002_4A\n",
      "NO 002_4B\n",
      "NO 002_64A\n",
      "NO 002_64B\n",
      "NO 002_64C\n",
      "NO 002_64D\n",
      "NO 002_64E\n",
      "NO 002_64F\n",
      "NO 002_64G\n",
      "NO 002_64H\n",
      "NO 002_64I\n",
      "NO 002_64J\n",
      "NO 002_64K\n",
      "NO 002_64L\n",
      "NO 002_64M\n",
      "NO 002_64N\n",
      "Unable to find plate for IGVF_002\n",
      "Unable to find plate for IGVF_002\n",
      "Unable to find plate for IGVF_002\n",
      "Unable to find plate for IGVF_002\n",
      "Unable to find plate for IGVF_002\n",
      "Unable to find plate for IGVF_002\n",
      "Unable to find plate for IGVF_002\n",
      "Unable to find plate for IGVF_002\n",
      "Unable to find plate for IGVF_002\n",
      "Unable to find plate for IGVF_002\n",
      "Unable to find plate for IGVF_002\n",
      "Unable to find plate for IGVF_002\n",
      "Unable to find plate for IGVF_002\n",
      "Unable to find plate for IGVF_002\n",
      "Unable to find plate for IGVF_002\n",
      "Unable to find plate for IGVF_002\n",
      "No unused aliquots 67000 {}\n",
      "NO 003_8A\n",
      "NO 003_67A\n",
      "NO 003_67B\n",
      "NO 003_67C\n",
      "NO 003_67D\n",
      "NO 003_67E\n",
      "NO 003_67F\n",
      "NO 003_67G\n",
      "NO 003_67H\n",
      "NO 003_67I\n",
      "NO 003_67J\n",
      "NO 003_67K\n",
      "NO 003_67L\n",
      "NO 003_67M\n",
      "NO 003_67N\n",
      "NO 003_67O\n",
      "003_67P\n",
      "003_67Q\n",
      "003_67R\n",
      "003_67S\n",
      "003_67T\n",
      "003_67U\n",
      "003_67V\n",
      "003_8B\n",
      "003_8C\n",
      "NO 004_8A\n",
      "NO 004_67A\n",
      "NO 004_67B\n",
      "NO 004_67C\n",
      "NO 004_67D\n",
      "NO 004_67E\n",
      "NO 004_67F\n",
      "NO 004_67G\n",
      "NO 004_67H\n",
      "NO 004_67I\n",
      "NO 004_67J\n",
      "NO 004_67K\n",
      "NO 004_67L\n",
      "NO 004_67M\n",
      "NO 004_67N\n",
      "NO 004_67O\n",
      "004_67P\n",
      "004_67Q\n",
      "004_67R\n",
      "004_67S\n",
      "004_67T\n",
      "004_8B\n",
      "004_8C\n",
      "NO 005_8A\n",
      "NO 005_67A\n",
      "NO 005_67B\n",
      "NO 005_67C\n",
      "NO 005_67D\n",
      "NO 005_67E\n",
      "NO 005_67F\n",
      "NO 005_67G\n",
      "NO 005_67H\n",
      "NO 005_67I\n",
      "NO 005_67J\n",
      "NO 005_67K\n",
      "NO 005_67L\n",
      "NO 005_67M\n",
      "NO 005_67N\n",
      "NO 005_67O\n",
      "005_67P\n",
      "005_67Q\n",
      "005_67R\n",
      "005_67S\n",
      "005_67T\n",
      "005_67U\n",
      "005_8B\n",
      "005_8C\n",
      "005_8D\n",
      "005_8E\n",
      "005_8F\n",
      "005_8G\n",
      "005_8H\n",
      "NO 006_8A\n",
      "NO 006_67A\n",
      "NO 006_67B\n",
      "NO 006_67C\n",
      "NO 006_67D\n",
      "NO 006_67E\n",
      "NO 006_67F\n",
      "NO 006_67G\n",
      "NO 006_67H\n",
      "NO 006_67I\n",
      "NO 006_67J\n",
      "NO 006_67K\n",
      "NO 006_67L\n",
      "NO 006_67M\n",
      "NO 006_67N\n",
      "NO 006_67O\n",
      "NO 006_67P\n",
      "NO 006_67Q\n",
      "Skipping failed plate IGVF_014\n",
      "NO 007_8A\n",
      "NO 007_67A\n",
      "NO 007_67B\n",
      "NO 007_67C\n",
      "NO 007_67D\n",
      "NO 007_67E\n",
      "NO 007_67F\n",
      "NO 007_67G\n",
      "NO 007_67H\n",
      "NO 007_67I\n",
      "NO 007_67J\n",
      "NO 007_67K\n",
      "NO 007_67L\n",
      "NO 007_67M\n",
      "NO 007_67N\n",
      "NO 007_67O\n",
      "007_67P\n",
      "007_67Q\n",
      "007_67R\n",
      "007_67S\n",
      "007_67T\n",
      "007_67U\n",
      "007_67V\n",
      "007_67W\n",
      "007_8B\n",
      "007_8C\n",
      "007_8D\n",
      "007_8E\n",
      "007_8F\n",
      "007_8G\n",
      "007_8H\n",
      "007_8I\n",
      "NO 008_8A\n",
      "NO 008_67A\n",
      "NO 008_67B\n",
      "NO 008_67C\n",
      "NO 008_67D\n",
      "NO 008_67E\n",
      "NO 008_67F\n",
      "NO 008_67G\n",
      "NO 008_67H\n",
      "NO 008_67I\n",
      "NO B01_13A\n",
      "NO B01_13B\n",
      "NO B01_13C\n",
      "NO B01_13D\n",
      "NO B01_13E\n",
      "NO B01_13F\n",
      "EX B01_13G\n",
      "NO B01_13H\n",
      "subpool_b01_13a B01_13A [{'accession_prefix': 'igvftst', 'name': 'TSTSM35366348', 'uuid': '77a93b75-f44a-4aa5-937b-a6788051bc18', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM35366348/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1676SYRC', 'uuid': 'efad3fdd-f466-4c59-94e0-8791e2bfaa93', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM1676SYRC/'}]\n",
      "subpool_b01_13a B01_13A [{'accession_prefix': 'igvftst', 'name': 'TSTSM35366348', 'uuid': '77a93b75-f44a-4aa5-937b-a6788051bc18', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM35366348/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1676SYRC', 'uuid': 'efad3fdd-f466-4c59-94e0-8791e2bfaa93', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM1676SYRC/'}]\n",
      "subpool_b01_13b B01_13B [{'accession_prefix': 'igvftst', 'name': 'TSTSM99018100', 'uuid': '6cac6a8b-7750-4df6-92cf-fae41082813d', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM99018100/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1955RVUC', 'uuid': 'f45edb75-5e28-4578-a2d5-395546f64b2a', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM1955RVUC/'}]\n",
      "subpool_b01_13b B01_13B [{'accession_prefix': 'igvftst', 'name': 'TSTSM99018100', 'uuid': '6cac6a8b-7750-4df6-92cf-fae41082813d', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM99018100/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1955RVUC', 'uuid': 'f45edb75-5e28-4578-a2d5-395546f64b2a', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM1955RVUC/'}]\n",
      "subpool_b01_13c B01_13C [{'accession_prefix': 'igvftst', 'name': 'TSTSM36900951', 'uuid': '01858fa2-3c3d-476a-8d81-e82a0612c0e8', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM36900951/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8093RPKR', 'uuid': '33b1fa8c-53ac-4db9-81e2-df137881c434', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM8093RPKR/'}]\n",
      "subpool_b01_13c B01_13C [{'accession_prefix': 'igvftst', 'name': 'TSTSM36900951', 'uuid': '01858fa2-3c3d-476a-8d81-e82a0612c0e8', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM36900951/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8093RPKR', 'uuid': '33b1fa8c-53ac-4db9-81e2-df137881c434', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM8093RPKR/'}]\n",
      "subpool_b01_13d B01_13D [{'accession_prefix': 'igvftst', 'name': 'TSTSM27245068', 'uuid': 'b69b021b-aa90-41cf-8d1f-cfc59c726b20', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM27245068/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9603DIDI', 'uuid': 'abb7e2a6-c08b-4a15-88fe-94c5c38a3400', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM9603DIDI/'}]\n",
      "subpool_b01_13d B01_13D [{'accession_prefix': 'igvftst', 'name': 'TSTSM27245068', 'uuid': 'b69b021b-aa90-41cf-8d1f-cfc59c726b20', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM27245068/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9603DIDI', 'uuid': 'abb7e2a6-c08b-4a15-88fe-94c5c38a3400', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM9603DIDI/'}]\n",
      "subpool_b01_13e B01_13E [{'accession_prefix': 'igvftst', 'name': 'TSTSM90710268', 'uuid': '8d44590d-bcb5-46f9-8c87-ea375db12ea9', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM90710268/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6958ESMJ', 'uuid': '92fcc1a8-9170-4a7c-90f8-5ec339a987d1', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM6958ESMJ/'}]\n",
      "subpool_b01_13e B01_13E [{'accession_prefix': 'igvftst', 'name': 'TSTSM90710268', 'uuid': '8d44590d-bcb5-46f9-8c87-ea375db12ea9', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM90710268/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6958ESMJ', 'uuid': '92fcc1a8-9170-4a7c-90f8-5ec339a987d1', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM6958ESMJ/'}]\n",
      "subpool_b01_13f B01_13F [{'accession_prefix': 'igvftst', 'name': 'TSTSM57985101', 'uuid': '2f34272a-6603-41e0-9820-cf2dd432244b', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM57985101/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6546CQJB', 'uuid': 'b2e62292-5163-4a73-9861-23a535a97dac', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM6546CQJB/'}]\n",
      "subpool_b01_13f B01_13F [{'accession_prefix': 'igvftst', 'name': 'TSTSM57985101', 'uuid': '2f34272a-6603-41e0-9820-cf2dd432244b', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM57985101/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6546CQJB', 'uuid': 'b2e62292-5163-4a73-9861-23a535a97dac', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM6546CQJB/'}]\n",
      "subpool_b01_13g B01_13G [{'accession_prefix': 'igvftst', 'name': 'TSTSM10953505', 'uuid': 'b0e63479-2caf-4aac-9c98-2ff823f92c74', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM10953505/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8722YQAW', 'uuid': '00179685-55cf-4dc1-a00a-3f5fb9133288', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM8722YQAW/'}]\n",
      "subpool_b01_13g B01_13G [{'accession_prefix': 'igvftst', 'name': 'TSTSM10953505', 'uuid': 'b0e63479-2caf-4aac-9c98-2ff823f92c74', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM10953505/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8722YQAW', 'uuid': '00179685-55cf-4dc1-a00a-3f5fb9133288', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM8722YQAW/'}]\n",
      "subpool_b01_13h B01_13H [{'accession_prefix': 'igvftst', 'name': 'TSTSM89995437', 'uuid': '5484bec6-4b0e-4488-b3b1-fd4def0022df', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM89995437/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2294BOOU', 'uuid': 'f25eb07b-ed33-496f-a8d7-83814b4b642f', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM2294BOOU/'}]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "subpool_b01_13h B01_13H [{'accession_prefix': 'igvftst', 'name': 'TSTSM89995437', 'uuid': '5484bec6-4b0e-4488-b3b1-fd4def0022df', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM89995437/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2294BOOU', 'uuid': 'f25eb07b-ed33-496f-a8d7-83814b4b642f', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM2294BOOU/'}]\n",
      "B01_13I\n",
      "B01_13J\n",
      "B01_13K\n",
      "B01_13L\n",
      "B01_13M\n",
      "B01_13N\n",
      "B01_13O\n",
      "B01_13P\n",
      "B01_13Q\n",
      "B01_13R\n",
      "NO 008B_67A\n",
      "NO 008B_67B\n",
      "NO 008B_67C\n",
      "NO 008B_67D\n",
      "NO 008B_67E\n",
      "NO 008B_67F\n",
      "NO 008B_67G\n",
      "NO 008B_67H\n",
      "NO 008B_67I\n",
      "NO 008B_67J\n",
      "No unused aliquots 13000 {67000: SubpoolName('67J')}\n",
      "NO 009_67A\n",
      "NO 009_67B\n",
      "NO 009_67C\n",
      "NO 009_67D\n",
      "NO 009_67E\n",
      "NO 009_67F\n",
      "NO 009_67G\n",
      "NO 009_67H\n",
      "NO 009_67I\n",
      "NO 009_67J\n",
      "NO 009_67K\n",
      "NO 009_67L\n",
      "NO 009_67M\n",
      "NO 009_67N\n",
      "NO 009_67O\n",
      "No unused aliquots 13000 {67000: SubpoolName('67O')}\n",
      "NO 010_67A\n",
      "NO 010_67B\n",
      "NO 010_67C\n",
      "NO 010_67D\n",
      "NO 010_67E\n",
      "NO 010_67F\n",
      "NO 010_67G\n",
      "NO 010_67H\n",
      "NO 010_67I\n",
      "NO 010_67J\n",
      "NO 010_67K\n",
      "NO 010_67L\n",
      "NO 010_67M\n",
      "NO 010_67N\n",
      "NO 010_67O\n",
      "010_67P\n",
      "010_67Q\n",
      "010_67R\n",
      "No unused aliquots 13000 {67000: SubpoolName('67O')}\n",
      "NO 011_67A\n",
      "NO 011_67B\n",
      "NO 011_67C\n",
      "NO 011_67D\n",
      "NO 011_67E\n",
      "NO 011_67F\n",
      "NO 011_67G\n",
      "NO 011_67H\n",
      "NO 011_67I\n",
      "NO 011_67J\n",
      "NO 011_67K\n",
      "NO 011_67L\n",
      "NO 011_67M\n",
      "NO 011_67N\n",
      "NO 011_67O\n",
      "011_67P\n",
      "011_67Q\n",
      "011_67R\n",
      "011_67S\n",
      "011_67T\n",
      "011_67U\n",
      "011_67V\n",
      "No unused aliquots 13000 {67000: SubpoolName('67O')}\n",
      "EX 003_13A\n",
      "EX 004_13A\n",
      "EX 005_13A\n",
      "EX 007_13A\n",
      "EX 008B_13A\n",
      "EX 009_13A\n",
      "EX 010_13A\n",
      "EX 011_13A\n",
      "011_13B\n",
      "011_13C\n",
      "NO 012_13A\n",
      "NO 012_13B\n",
      "NO 012_13C\n",
      "NO 012_13D\n",
      "NO 012_13E\n",
      "NO 012_13F\n",
      "NO 012_13G\n",
      "NO 012_13H\n",
      "012_13I\n",
      "012_13J\n",
      "012_13K\n",
      "012_13L\n",
      "012_13M\n",
      "012_13N\n",
      "012_13O\n",
      "012_13P\n",
      "012_13Q\n",
      "012_13R\n",
      "012_13S\n",
      "012_13T\n",
      "EX 013_13A\n",
      "NO 013_67A\n",
      "NO 013_67B\n",
      "NO 013_67C\n",
      "NO 013_67D\n",
      "NO 013_67E\n",
      "NO 013_67F\n",
      "NO 013_67G\n",
      "NO 013_67H\n",
      "NO 013_67I\n",
      "NO 013_67J\n",
      "NO 013_67K\n",
      "NO 013_67L\n",
      "NO 013_67M\n",
      "NO 013_67N\n",
      "NO 014_13A\n",
      "NO 014_13B\n",
      "NO 014_13C\n",
      "NO 014_13D\n",
      "NO 014_13E\n",
      "NO 014_13F\n",
      "NO 014_13G\n",
      "NO 014_13H\n",
      "EX 015_13A\n",
      "NO 015_67A\n",
      "NO 015_67B\n",
      "NO 015_67C\n",
      "NO 015_67D\n",
      "NO 015_67E\n",
      "NO 015_67F\n",
      "NO 015_67G\n",
      "NO 015_67H\n",
      "NO 015_67I\n",
      "NO 015_67J\n",
      "NO 015_67K\n",
      "NO 015_67L\n",
      "NO 015_67M\n",
      "NO 015_67N\n",
      "NO 015_67O\n",
      "Unable to find plate for IGVF_015\n",
      "Unable to find plate for IGVF_015\n",
      "Unable to find plate for IGVF_015\n",
      "Unable to find plate for IGVF_015\n",
      "Unable to find plate for IGVF_015\n",
      "Unable to find plate for IGVF_015\n",
      "Unable to find plate for IGVF_015\n",
      "Unable to find plate for IGVF_015\n",
      "Unable to find plate for IGVF_015\n",
      "Unable to find plate for IGVF_015\n",
      "Unable to find plate for IGVF_015\n",
      "Unable to find plate for IGVF_015\n",
      "Unable to find plate for IGVF_015\n",
      "Unable to find plate for IGVF_015\n",
      "Unable to find plate for IGVF_015\n",
      "Unable to find plate for IGVF_015\n",
      "EX 016_13A\n",
      "NO 016_67A\n",
      "NO 016_67B\n",
      "NO 016_67C\n",
      "NO 016_67D\n",
      "NO 016_67E\n",
      "NO 016_67F\n",
      "NO 016_67G\n",
      "NO 016_67H\n",
      "NO 016_67I\n",
      "NO 016_67J\n",
      "NO 016_67K\n",
      "NO 016_67L\n",
      "NO 016_67M\n",
      "NO 016_67N\n",
      "NO 016_67O\n",
      "Unable to find plate for IGVF_016\n",
      "Unable to find plate for IGVF_016\n",
      "Unable to find plate for IGVF_016\n",
      "Unable to find plate for IGVF_016\n",
      "Unable to find plate for IGVF_016\n",
      "Unable to find plate for IGVF_016\n",
      "Unable to find plate for IGVF_016\n",
      "Unable to find plate for IGVF_016\n",
      "Unable to find plate for IGVF_016\n",
      "Unable to find plate for IGVF_016\n",
      "Unable to find plate for IGVF_016\n",
      "Unable to find plate for IGVF_016\n",
      "Unable to find plate for IGVF_016\n",
      "Unable to find plate for IGVF_016\n",
      "Unable to find plate for IGVF_016\n",
      "Unable to find plate for IGVF_016\n",
      "EX 017_13A\n",
      "NO 017_67A\n",
      "NO 017_67B\n",
      "NO 017_67C\n",
      "NO 017_67D\n",
      "NO 017_67E\n",
      "NO 017_67F\n",
      "NO 017_67G\n",
      "NO 017_67H\n",
      "NO 017_67I\n",
      "NO 017_67J\n",
      "NO 017_67K\n",
      "NO 017_67L\n",
      "NO 017_67M\n",
      "NO 017_67N\n",
      "NO 017_67O\n",
      "Unable to find plate for IGVF_017\n",
      "Unable to find plate for IGVF_017\n",
      "Unable to find plate for IGVF_017\n",
      "Unable to find plate for IGVF_017\n",
      "Unable to find plate for IGVF_017\n",
      "Unable to find plate for IGVF_017\n",
      "Unable to find plate for IGVF_017\n",
      "Unable to find plate for IGVF_017\n",
      "Unable to find plate for IGVF_017\n",
      "Unable to find plate for IGVF_017\n",
      "Unable to find plate for IGVF_017\n",
      "Unable to find plate for IGVF_017\n",
      "Unable to find plate for IGVF_017\n",
      "Unable to find plate for IGVF_017\n",
      "Unable to find plate for IGVF_017\n",
      "Unable to find plate for IGVF_017\n",
      "EX 018_13A\n",
      "NO 018_67A\n",
      "NO 018_67B\n",
      "NO 018_67C\n",
      "NO 018_67D\n",
      "NO 018_67E\n",
      "NO 018_67F\n",
      "NO 018_67G\n",
      "NO 018_67H\n",
      "NO 018_67I\n",
      "NO 018_67J\n",
      "NO 018_67K\n",
      "NO 018_67L\n",
      "NO 018_67M\n",
      "NO 018_67N\n",
      "NO 018_67O\n",
      "Unable to find plate for IGVF_018\n",
      "Unable to find plate for IGVF_018\n",
      "Unable to find plate for IGVF_018\n",
      "Unable to find plate for IGVF_018\n",
      "Unable to find plate for IGVF_018\n",
      "Unable to find plate for IGVF_018\n",
      "Unable to find plate for IGVF_018\n",
      "Unable to find plate for IGVF_018\n",
      "Unable to find plate for IGVF_018\n",
      "Unable to find plate for IGVF_018\n",
      "Unable to find plate for IGVF_018\n",
      "Unable to find plate for IGVF_018\n",
      "Unable to find plate for IGVF_018\n",
      "Unable to find plate for IGVF_018\n",
      "Unable to find plate for IGVF_018\n",
      "Unable to find plate for IGVF_018\n",
      "EX 019_13A\n",
      "NO 019_67A\n",
      "NO 019_67B\n",
      "NO 019_67C\n",
      "NO 019_67D\n",
      "NO 019_67E\n",
      "NO 019_67F\n",
      "NO 019_67G\n",
      "NO 019_67H\n",
      "NO 019_67I\n",
      "NO 019_67J\n",
      "NO 019_67K\n",
      "NO 019_67L\n",
      "NO 019_67M\n",
      "NO 019_67N\n",
      "NO 019_67O\n",
      "Unable to find plate for IGVF_019\n",
      "Unable to find plate for IGVF_019\n",
      "Unable to find plate for IGVF_019\n",
      "Unable to find plate for IGVF_019\n",
      "Unable to find plate for IGVF_019\n",
      "Unable to find plate for IGVF_019\n",
      "Unable to find plate for IGVF_019\n",
      "Unable to find plate for IGVF_019\n",
      "Unable to find plate for IGVF_019\n",
      "Unable to find plate for IGVF_019\n",
      "Unable to find plate for IGVF_019\n",
      "Unable to find plate for IGVF_019\n",
      "Unable to find plate for IGVF_019\n",
      "Unable to find plate for IGVF_019\n",
      "Unable to find plate for IGVF_019\n",
      "Unable to find plate for IGVF_019\n",
      "EX 020_13A\n",
      "NO 020_67A\n",
      "NO 020_67B\n",
      "NO 020_67C\n",
      "NO 020_67D\n",
      "NO 020_67E\n",
      "NO 020_67F\n",
      "NO 020_67G\n",
      "NO 020_67H\n",
      "NO 020_67I\n",
      "NO 020_67J\n",
      "NO 020_67K\n",
      "NO 020_67L\n",
      "NO 020_67M\n",
      "NO 020_67N\n",
      "NO 020_67O\n",
      "Unable to find plate for IGVF_020\n",
      "Unable to find plate for IGVF_020\n",
      "Unable to find plate for IGVF_020\n",
      "Unable to find plate for IGVF_020\n",
      "Unable to find plate for IGVF_020\n",
      "Unable to find plate for IGVF_020\n",
      "Unable to find plate for IGVF_020\n",
      "Unable to find plate for IGVF_020\n",
      "Unable to find plate for IGVF_020\n",
      "Unable to find plate for IGVF_020\n",
      "Unable to find plate for IGVF_020\n",
      "Unable to find plate for IGVF_020\n",
      "Unable to find plate for IGVF_020\n",
      "Unable to find plate for IGVF_020\n",
      "Unable to find plate for IGVF_020\n",
      "Unable to find plate for IGVF_020\n",
      "EX 021_13A\n",
      "NO 021_67A\n",
      "NO 021_67B\n",
      "NO 021_67C\n",
      "NO 021_67D\n",
      "NO 021_67E\n",
      "NO 021_67F\n",
      "NO 021_67G\n",
      "NO 021_67H\n",
      "NO 021_67I\n",
      "NO 021_67J\n",
      "NO 021_67K\n",
      "NO 021_67L\n",
      "NO 021_67M\n",
      "NO 021_67N\n",
      "NO 021_67O\n",
      "Unable to find plate for IGVF_021\n",
      "Unable to find plate for IGVF_021\n",
      "Unable to find plate for IGVF_021\n",
      "Unable to find plate for IGVF_021\n",
      "Unable to find plate for IGVF_021\n",
      "Unable to find plate for IGVF_021\n",
      "Unable to find plate for IGVF_021\n",
      "Unable to find plate for IGVF_021\n",
      "Unable to find plate for IGVF_021\n",
      "Unable to find plate for IGVF_021\n",
      "Unable to find plate for IGVF_021\n",
      "Unable to find plate for IGVF_021\n",
      "Unable to find plate for IGVF_021\n",
      "Unable to find plate for IGVF_021\n",
      "Unable to find plate for IGVF_021\n",
      "Unable to find plate for IGVF_021\n",
      "EX 022_13A\n",
      "NO 022_67A\n",
      "NO 022_67B\n",
      "NO 022_67C\n",
      "NO 022_67D\n",
      "NO 022_67E\n",
      "NO 022_67F\n",
      "NO 022_67G\n",
      "NO 022_67H\n",
      "NO 022_67I\n",
      "NO 022_67J\n",
      "NO 022_67K\n",
      "NO 022_67L\n",
      "NO 022_67M\n",
      "NO 022_67N\n",
      "NO 022_67O\n",
      "Unable to find plate for IGVF_022\n",
      "Unable to find plate for IGVF_022\n",
      "Unable to find plate for IGVF_022\n",
      "Unable to find plate for IGVF_022\n",
      "Unable to find plate for IGVF_022\n",
      "Unable to find plate for IGVF_022\n",
      "Unable to find plate for IGVF_022\n",
      "Unable to find plate for IGVF_022\n",
      "Unable to find plate for IGVF_022\n",
      "Unable to find plate for IGVF_022\n",
      "Unable to find plate for IGVF_022\n",
      "Unable to find plate for IGVF_022\n",
      "Unable to find plate for IGVF_022\n",
      "Unable to find plate for IGVF_022\n",
      "Unable to find plate for IGVF_022\n",
      "Unable to find plate for IGVF_022\n"
     ]
    }
   ],
   "source": [
    "def load_experiments_return_submittable_subpools():\n",
    "    compression=r\"(?<compression>gz|bz2|xz|zstd)\"\n",
    "    nanopore_library_name_re_template = r\"igvf(?P<run>[\\d]+)_(?P<library_id>{library_name})_.*_(?P<fragment>[\\d]+)\\.fastq\\.{compression}\"\n",
    "    nextseq_library_name_re_template = r\"(?P<run>[\\d]+)_(?P<library_id>{library_name})_(?P<read_id>[RI][\\d])\\.fastq\\.{compression}\"\n",
    "    nextseq_library_index_re_template = r\"Sublibrary_(?P<library_id>{library_index})_S(?P<sample_id>[\\d]+)_(?P<lane_id>L[\\d]+)_(?P<read_id>[RI][\\d])_(?P<fragment>[\\d]+)\\.fastq\\.{compression}\"\n",
    "    novaseq_library_index_re_template = r\"Sublibrary_(?P<library_id>{library_index})_S(?P<sample_id>[\\d]+)_(?P<lane_id>L[\\d]+)_(?P<read_id>[RI][\\d])_(?P<fragment>[\\d]+)\\.fastq\\.{compression}\"\n",
    "\n",
    "    multiplexed_samples = load_submitted_multiplexed_samples()\n",
    "    split_seq_plates = load_split_seq_plates_as_dict()\n",
    "\n",
    "    failed_plates = {\"IGVF_006\",}\n",
    "    failed = False\n",
    "    submittable_subpools = {}\n",
    "    for experiment_name, experiment_start, experiment_stop in find_fixation_start(experiment):\n",
    "        fixation = parse_fixation(experiment, experiment_name, experiment_start, experiment_stop)\n",
    "\n",
    "        if fixation[\"plate_name\"] in failed_plates:\n",
    "            print(f\"Skipping failed plate {plate_name}\")\n",
    "            continue\n",
    "\n",
    "        last_subpools = {}\n",
    "\n",
    "        reagent = guess_barcode_reagent_from_plate(fixation[\"plate_name\"])\n",
    "\n",
    "        for line_offset, subpool in enumerate(fixation[\"subpool\"]):\n",
    "            if subpool[\"plate_name\"] not in split_seq_plates:\n",
    "                print(\"Unable to find plate for {}\".format(subpool[\"plate_name\"]))\n",
    "                continue\n",
    "\n",
    "            try:\n",
    "                barcodes = models.LibraryBarcode.objects.filter(\n",
    "                    reagent=reagent, \n",
    "                    barcode_type=None, \n",
    "                    code=subpool[\"index\"])\n",
    "            except models.LibraryBarcode.DoesNotExist:\n",
    "                print(f\"Unable to find barcode {subpool['index']}\")\n",
    "                \n",
    "            if len(barcodes) == 0:\n",
    "                print(f\"Unable to find barcodes with {reagent} None {subpool['index']}\")\n",
    "\n",
    "            # validate i7 barcodes\n",
    "            library_i7_barcodes_sequence = {b.i7_sequence for b in barcodes if pandas.notnull(b.i7_sequence)}\n",
    "            if not pandas.isnull(subpool[\"i7_barcode\"]):\n",
    "                i7_expected_sequence = set(subpool[\"i7_barcode\"].split(\",\"))\n",
    "            else:\n",
    "                i7_expected_sequence = set()\n",
    "            if library_i7_barcodes_sequence != i7_expected_sequence:\n",
    "                print(f\"{subpool['name']} Database lookup of {subpool['index']} i7_barcodes {library_i7_barcodes_sequence} doesn't match human entry {i7_expected_sequence}\")\n",
    "                print(reagent, subpool[\"index\"], barcodes)\n",
    "                failed = True\n",
    "                continue\n",
    "\n",
    "            # validate i5 barcodes\n",
    "            library_i5_barcodes_sequence = {b.i5_sequence for b in barcodes if pandas.notnull(b.i5_sequence)}\n",
    "            if len(library_i5_barcodes_sequence) > 0:\n",
    "                if not pandas.isnull(subpool[\"i5_barcode\"]):\n",
    "                    i5_expected_sequence = set(subpool[\"i5_barcode\"].split(\",\"))\n",
    "                else:\n",
    "                    i5_expected_sequence = set()\n",
    "                if library_i5_barcodes_sequence != i5_expected_sequence:\n",
    "                    print(f\"{subpool['name']} Database lookup of i5_barcodes {library_i5_barcodes_sequence} doesn't match human entry {i5_expected_sequence}\")\n",
    "                    failed = True\n",
    "                    continue\n",
    "\n",
    "            plate = split_seq_plates[subpool[\"plate_name\"]]\n",
    "            subpool_prefix = subpool[\"name\"].split(\"_\")[0]\n",
    "            last_subpools[subpool[\"nuclei\"]] = max(\n",
    "                last_subpools.get(subpool[\"nuclei\"]), \n",
    "                SubpoolName(subpool[\"subpool_subname\"]))\n",
    "\n",
    "            record = models.Subpool(\n",
    "                name=subpool[\"name\"],\n",
    "                plate=plate,\n",
    "                nuclei=subpool[\"nuclei\"],\n",
    "                selection_type=subpool[\"selection_type\"],\n",
    "                cdna_pcr_rounds=subpool[\"cdna_pcr_rounds\"],\n",
    "                cdna_ng_per_ul=float_or_none(subpool[\"cdna_ng_per_ul\"]),\n",
    "                cdna_volume=float_or_none(subpool[\"cdna_volume\"]),\n",
    "                bioanalyzer_date=date_or_none(subpool[\"bioanalyzer_date\"]),\n",
    "                cdna_average_bp_length=int_or_none(subpool[\"cdna_average_bp_length\"]),\n",
    "                index_pcr_number=int_or_none(subpool[\"index_pcr_no\"]),\n",
    "                index=str_or_none(subpool[\"index\"]),\n",
    "                library_ng_per_ul=float_or_none(subpool[\"library_ng_per_ul\"]),\n",
    "                library_average_bp_length=int_or_none(subpool[\"library_average_bp_length\"]),\n",
    "            )\n",
    "            record.save()\n",
    "            record.barcode.set(barcodes)\n",
    "            record.save()\n",
    "\n",
    "            if pandas.notnull(subpool[\"submit\"]):\n",
    "                submittable_subpools[subpool[\"name\"]] = subpool[\"submit\"]\n",
    "\n",
    "            name = \"subpool_{}\".format(subpool[\"name\"].lower())\n",
    "            if name in multiplexed_samples:\n",
    "                for accession_row in multiplexed_samples[name]:\n",
    "                    if pandas.notnull(accession_row[\"name\"]):\n",
    "                        print(name, record.name, multiplexed_samples[name])\n",
    "                        accession = models.Accession(**accession_row)\n",
    "                        accession.save()\n",
    "                        record.accession.add(accession)\n",
    "                        record.save()            \n",
    "\n",
    "        for nuclei in fixation[\"unused_aliquots\"]:\n",
    "            if nuclei in last_subpools:\n",
    "                subpool_name = last_subpools[nuclei]\n",
    "                for i in range(fixation[\"unused_aliquots\"][nuclei]):\n",
    "                    subpool_name = subpool_name.next_code()\n",
    "\n",
    "                    record = models.Subpool(\n",
    "                        name=\"{}_{}\".format(subpool_prefix, subpool_name),\n",
    "                        plate=plate,\n",
    "                        nuclei=nuclei,\n",
    "                    )\n",
    "                    print(record.name)\n",
    "                    record.save()\n",
    "            else:\n",
    "                print(\"No unused aliquots {} {}\".format(nuclei, last_subpools))\n",
    "\n",
    "    assert not failed\n",
    "    return submittable_subpools\n",
    "    \n",
    "submittable_subpools = load_experiments_return_submittable_subpools()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "electoral-township",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<LibraryBarcode: UDI_Plate_WT_20 UDI20 TAATCGAC+TAATCGAC>"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "models.LibraryBarcode.objects.get(i7_sequence=\"TAATCGAC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "sudden-illustration",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Old creating runs from the sample sheet\n",
    "\n",
    "#    expected_names = {\n",
    "#        \"IGVF_Splitseq_003\": {\n",
    "#            \"nanopore\": nanopore_library_name_re_template,\n",
    "#            \"nextseq\": nextseq_library_name_re_template,\n",
    "#            \"novaseq\": novaseq_library_index_re_template,\n",
    "#        },\n",
    "#        \"IGVF_Splitseq_004\": {\n",
    "#            \"nanopore\": nanopore_library_name_re_template,\n",
    "#            \"nextseq\": nextseq_library_name_re_template,\n",
    "#            \"novaseq\": novaseq_library_index_re_template,\n",
    "#        },\n",
    "#        \"IGVF_Splitseq_005\": {\n",
    "#            \"nanopore\": nanopore_library_name_re_template,\n",
    "#            \"nextseq\": nextseq_library_index_re_template,\n",
    "#            \"novaseq\": novaseq_library_index_re_template,\n",
    "#        },\n",
    "#        \"IGVF_Splitseq_006\": {\n",
    "#            \"nanopore\": nanopore_library_name_re_template,\n",
    "#            \"nextseq\": nextseq_library_index_re_template,\n",
    "#            \"novaseq\": novaseq_library_index_re_template,\n",
    "#        },\n",
    "#\n",
    "#    }\n",
    "#\n",
    "#    nanopore_runs = {\n",
    "#        \"IGVF_Splitseq_003\": {\n",
    "#            \"name\": \"2023-01-03-8A\",\n",
    "#            \"run_date\": datetime.datetime(2023,1,3),\n",
    "#            \"raw_reads\": 87784162,\n",
    "#        }\n",
    "#    }\n",
    "\n",
    "#        if experiment_name in nanopore_runs and subpool[\"nuclei\"] == 8000:\n",
    "#            measurement = get_or_create_sequencing_run(\n",
    "#                # nanopore_runs[experiment_name][\"name\"]\n",
    "#                name=\"{}/{}\".format(plate.name.lower(), \"nanopore\"),\n",
    "#                run_date=nanopore_runs[experiment_name][\"run_date\"],\n",
    "#                platform=platforms[\"nanopore\"],\n",
    "#                plate=plate,\n",
    "#            )\n",
    "#            \n",
    "#            pattern = expected_names[experiment_name][\"nanopore\"].format(\n",
    "#                library_name=subpool[\"subpool_subname\"],\n",
    "#                library_index=line_offset+1,\n",
    "#                compression=compression,\n",
    "#            )\n",
    "#            \n",
    "#            run = models.SubpoolInRun(\n",
    "#                subpool=record,\n",
    "#                sequencing_run=measurement,\n",
    "#                raw_reads=nanopore_runs[experiment_name][\"raw_reads\"],\n",
    "#                status=models.RunStatus.PASS,\n",
    "#            )\n",
    "#            run.save()            \n",
    "#\n",
    "#        if subpool[\"nextseq_run_date\"] is not None:\n",
    "#            measurement = get_or_create_sequencing_run(\n",
    "#                #sublibrary[\"nextseq_run_id\"]\n",
    "#                name=\"{}/{}\".format(plate.name.lower(), \"nextseq\"),\n",
    "#                run_date=subpool[\"nextseq_run_date\"],\n",
    "#                platform=platforms[\"nextseq\"],\n",
    "#                plate=plate,\n",
    "#            )\n",
    "#\n",
    "#            pattern = expected_names[experiment_name][\"nextseq\"].format(\n",
    "#                library_name=subpool[\"subpool_subname\"],\n",
    "#                library_index=line_offset+1,\n",
    "#                compression=compression,\n",
    "#            )            \n",
    "#            \n",
    "#            run = models.SubpoolInRun(\n",
    "#                subpool=record,\n",
    "#                sequencing_run=measurement,\n",
    "#                raw_reads=subpool[\"nextseq_run_raw_reads\"],\n",
    "#                status=models.RunStatus.PASS,\n",
    "#            )\n",
    "#            run.save()\n",
    "#            \n",
    "#        if subpool[\"novaseq_run_date\"] is not None:\n",
    "#            for i, nova_raw_reads in enumerate([subpool[\"novaseq1_raw_reads\"], subpool[\"novaseq2_raw_reads\"]]):\n",
    "#                #nova_run_name = \"{}-{}\".format(sublibrary[\"novaseq_run_id\"], i+1)\n",
    "#                nova_run_name = \"{}/{}\".format(plate.name.lower(), \"nova{}\".format(i+1))\n",
    "#                measurement = get_or_create_sequencing_run(\n",
    "#                    name=nova_run_name,\n",
    "#                    run_date=subpool[\"novaseq_run_date\"],\n",
    "#                    platform=platforms[\"novaseq\"],\n",
    "#                    plate=plate,\n",
    "#                )\n",
    "#\n",
    "#                pattern = expected_names[experiment_name][\"novaseq\"].format(\n",
    "#                    library_name=subpool[\"subpool_subname\"],\n",
    "#                    library_index=line_offset+1,\n",
    "#                    compression=compression,\n",
    "#                )\n",
    "#\n",
    "#                run = models.SubpoolInRun(\n",
    "#                    subpool=record,\n",
    "#                    sequencing_run=measurement,\n",
    "#                    raw_reads=nova_raw_reads,\n",
    "#                    status=models.RunStatus.PASS,                    \n",
    "#                )\n",
    "#                run.save()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "choice-error",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>experiment</th>\n",
       "      <th>family</th>\n",
       "      <th>filename</th>\n",
       "      <th>md5sum</th>\n",
       "      <th>size</th>\n",
       "      <th>ctime</th>\n",
       "      <th>plate_id</th>\n",
       "      <th>subpool_name</th>\n",
       "      <th>fragment</th>\n",
       "      <th>compression</th>\n",
       "      <th>sequencer</th>\n",
       "      <th>sequencer_run</th>\n",
       "      <th>flowcell_id</th>\n",
       "      <th>nuclei</th>\n",
       "      <th>read</th>\n",
       "      <th>barcode</th>\n",
       "      <th>index_id</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>lane</th>\n",
       "      <th>exome_capture</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>igvf_003</td>\n",
       "      <td>illumina</td>\n",
       "      <td>igvf_003/nextseq2/003_13A_R1.fastq.gz</td>\n",
       "      <td>57e977fdfe32850db83ee5f7addb7165</td>\n",
       "      <td>17707636592</td>\n",
       "      <td>2023-10-02</td>\n",
       "      <td>003</td>\n",
       "      <td>13A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gz</td>\n",
       "      <td>VH00582</td>\n",
       "      <td>6</td>\n",
       "      <td>AAC3V7HHV</td>\n",
       "      <td>13</td>\n",
       "      <td>R1</td>\n",
       "      <td>CAGATC</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377</th>\n",
       "      <td>igvf_003</td>\n",
       "      <td>illumina</td>\n",
       "      <td>igvf_003/nextseq2/003_13A_R2.fastq.gz</td>\n",
       "      <td>148c181e7d81e4aa15611fb2086d5971</td>\n",
       "      <td>10322410034</td>\n",
       "      <td>2023-10-02</td>\n",
       "      <td>003</td>\n",
       "      <td>13A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gz</td>\n",
       "      <td>VH00582</td>\n",
       "      <td>6</td>\n",
       "      <td>AAC3V7HHV</td>\n",
       "      <td>13</td>\n",
       "      <td>R2</td>\n",
       "      <td>CAGATC</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>igvf_003</td>\n",
       "      <td>illumina</td>\n",
       "      <td>igvf_003/nextseq/003_67A_R1.fastq.gz</td>\n",
       "      <td>35350d53cf858267d294bce69e475c83</td>\n",
       "      <td>48483277280</td>\n",
       "      <td>2023-09-08</td>\n",
       "      <td>003</td>\n",
       "      <td>67A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gz</td>\n",
       "      <td>VH00582</td>\n",
       "      <td>1</td>\n",
       "      <td>AAATMGFHV</td>\n",
       "      <td>67</td>\n",
       "      <td>R1</td>\n",
       "      <td>ACTTGA</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>igvf_003</td>\n",
       "      <td>illumina</td>\n",
       "      <td>igvf_003/nextseq/003_67A_R2.fastq.gz</td>\n",
       "      <td>00c68ff0f9e0217d422c57e8948d4bb4</td>\n",
       "      <td>32327542019</td>\n",
       "      <td>2023-09-08</td>\n",
       "      <td>003</td>\n",
       "      <td>67A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gz</td>\n",
       "      <td>VH00582</td>\n",
       "      <td>1</td>\n",
       "      <td>AAATMGFHV</td>\n",
       "      <td>67</td>\n",
       "      <td>R2</td>\n",
       "      <td>ACTTGA</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>igvf_003</td>\n",
       "      <td>illumina</td>\n",
       "      <td>igvf_003/nextseq/003_67B_R1.fastq.gz</td>\n",
       "      <td>79db765143bb23e14c613c9e1339e9f7</td>\n",
       "      <td>44182320136</td>\n",
       "      <td>2023-09-08</td>\n",
       "      <td>003</td>\n",
       "      <td>67B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gz</td>\n",
       "      <td>VH00582</td>\n",
       "      <td>1</td>\n",
       "      <td>AAATMGFHV</td>\n",
       "      <td>67</td>\n",
       "      <td>R1</td>\n",
       "      <td>GATCAG</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    experiment    family                               filename  \\\n",
       "378   igvf_003  illumina  igvf_003/nextseq2/003_13A_R1.fastq.gz   \n",
       "377   igvf_003  illumina  igvf_003/nextseq2/003_13A_R2.fastq.gz   \n",
       "13    igvf_003  illumina   igvf_003/nextseq/003_67A_R1.fastq.gz   \n",
       "12    igvf_003  illumina   igvf_003/nextseq/003_67A_R2.fastq.gz   \n",
       "14    igvf_003  illumina   igvf_003/nextseq/003_67B_R1.fastq.gz   \n",
       "\n",
       "                               md5sum         size       ctime plate_id  \\\n",
       "378  57e977fdfe32850db83ee5f7addb7165  17707636592  2023-10-02      003   \n",
       "377  148c181e7d81e4aa15611fb2086d5971  10322410034  2023-10-02      003   \n",
       "13   35350d53cf858267d294bce69e475c83  48483277280  2023-09-08      003   \n",
       "12   00c68ff0f9e0217d422c57e8948d4bb4  32327542019  2023-09-08      003   \n",
       "14   79db765143bb23e14c613c9e1339e9f7  44182320136  2023-09-08      003   \n",
       "\n",
       "    subpool_name fragment compression sequencer sequencer_run flowcell_id  \\\n",
       "378          13A      NaN          gz   VH00582             6   AAC3V7HHV   \n",
       "377          13A      NaN          gz   VH00582             6   AAC3V7HHV   \n",
       "13           67A      NaN          gz   VH00582             1   AAATMGFHV   \n",
       "12           67A      NaN          gz   VH00582             1   AAATMGFHV   \n",
       "14           67B      NaN          gz   VH00582             1   AAATMGFHV   \n",
       "\n",
       "     nuclei read barcode index_id sample_id lane exome_capture  \n",
       "378      13   R1  CAGATC        1       NaN  NaN            EX  \n",
       "377      13   R2  CAGATC        1       NaN  NaN            EX  \n",
       "13       67   R1  ACTTGA        2       NaN  NaN           NaN  \n",
       "12       67   R2  ACTTGA        2       NaN  NaN           NaN  \n",
       "14       67   R1  GATCAG        3       NaN  NaN           NaN  "
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def load_fastq_csv():\n",
    "    fastqs = pandas.read_csv(\n",
    "        \"fastq_metadata.tsv\", sep=\"\\t\",\n",
    "        dtype={\n",
    "            \"sample_id\": str,\n",
    "            \"read_id\": str,\n",
    "            \"fragment\": str,\n",
    "            \"sequencer_run\": str,\n",
    "            \"lane\": str,\n",
    "        }).sort_values([\"experiment\", \"subpool_name\", \"index_id\", \"lane\", \"read\"])\n",
    "    return fastqs\n",
    "\n",
    "fastqs = load_fastq_csv()\n",
    "fastqs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "republican-strategy",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: database says dual index, file says single index: GTGAAACT AGTCTGTA CTTGTA\n",
      "Finish check_fastq_barcode_is_equal test assertions\n",
      "Barcode mismatch IGVF_003 1 CAGATC None nan igvf_003/nanopore/igvf003_8A_lig-ss_11.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC None nan igvf_003/nanopore/igvf003_8A_lig-ss_4.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC None nan igvf_003/nanopore/igvf003_8A_lig-ss_7.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC None nan igvf_003/nanopore/igvf003_8A_lig-ss_6.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC None nan igvf_003/nanopore/igvf003_8A_lig-ss_8.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC None nan igvf_003/nanopore/igvf003_8A_lig-ss_5.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC None nan igvf_003/nanopore/igvf003_8A_lig-ss_1.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC None nan igvf_003/nanopore/igvf003_8A_lig-ss_2.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC None nan igvf_003/nanopore/igvf003_8A_lig-ss_10.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC None nan igvf_003/nanopore/igvf003_8A_lig-ss_9.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC None nan igvf_003/nanopore/igvf003_8A_lig-ss_3.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "def fastq_metadata_row_to_subpool_name(row):\n",
    "    if pandas.isnull(row.subpool_name):\n",
    "        return None\n",
    "    else:\n",
    "        return \"{}_{}\".format(row.plate_id, row.subpool_name)\n",
    "\n",
    "def get_subpool_from_fastq_row(row):\n",
    "    plate_name = normalize_plate_name(row.experiment)\n",
    "    subpool_name = fastq_metadata_row_to_subpool_name(row)\n",
    "\n",
    "    query = {\n",
    "        \"plate__name\": plate_name,\n",
    "    }\n",
    "    \n",
    "    if pandas.notnull(subpool_name):\n",
    "        query[\"name\"] = subpool_name\n",
    "    if pandas.notnull(row.index_id):\n",
    "        query[\"index\"] = str_or_none(row.index_id)\n",
    "    \n",
    "    # these are labeled as having X nuclei but actually have fewer nuclei. \n",
    "    # Not that you can tell that from the filename\n",
    "    nuclei_dont_match = {\"013_67N\", \"014_13H\"}\n",
    "    if pandas.notnull(row.nuclei) and subpool_name not in nuclei_dont_match:\n",
    "        query[\"nuclei\"] = int(row.nuclei)*1000\n",
    "    if pandas.notnull(row.exome_capture):\n",
    "        query[\"selection_type\"] = row.exome_capture\n",
    "\n",
    "    subpools = models.Subpool.objects.filter(**query)\n",
    "    \n",
    "    if len(subpools) == 0:\n",
    "        print(\"Nothing found for {} {}\".format(query, row.filename))\n",
    "    elif len(subpools) > 1:\n",
    "        print(\"Multiple hits for {}: {} {}\".format(query, subpools, row.filename))\n",
    "    else:\n",
    "        return subpools[0]\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "def check_fastq_barcode_is_equal(i7_sequence, i5_sequence, barcode):\n",
    "    if pandas.isnull(i5_sequence):\n",
    "        return i7_sequence == barcode\n",
    "    \n",
    "    sequences = barcode.split(\"+\")\n",
    "    if len(sequences) != 2:\n",
    "        print(\"Warning: database says dual index, file says single index: {} {} {}\".format(i7_sequence, i5_sequence, barcode))\n",
    "        return False\n",
    "\n",
    "    i7, i5 = sequences\n",
    "    # barcodes in the fastq file are reverse complimented.\n",
    "    i5 = Seq(i5).reverse_complement()\n",
    "    return i7_sequence == i7 and i5_sequence == i5\n",
    "    \n",
    "assert check_fastq_barcode_is_equal(\"ACTTGA\", None, \"ACTTGA\")\n",
    "assert not check_fastq_barcode_is_equal(\"ACTTGA\", None, \"CTTGTA\")\n",
    "assert check_fastq_barcode_is_equal(\"GTGAAACT\", \"AGTCTGTA\", \"GTGAAACT+TACAGACT\")\n",
    "assert not check_fastq_barcode_is_equal(\"GTGAAACT\", \"AGTCTGTA\", \"ACTTGATC+TTTGGGTG\")\n",
    "assert not check_fastq_barcode_is_equal(\"GTGAAACT\", \"AGTCTGTA\", \"CTTGTA\")\n",
    "print(\"Finish check_fastq_barcode_is_equal test assertions\")\n",
    "\n",
    "def test_subpool_lookup(fastqs):\n",
    "    for i, row in fastqs.iterrows():\n",
    "        subpool = get_subpool_from_fastq_row(row)\n",
    "        if subpool is None:\n",
    "            continue\n",
    "        barcodes = subpool.barcode.filter(barcode_type=None)\n",
    "        if barcodes.count() != 1:\n",
    "            print(\"Barcode count wrong {}\".format(barcode))\n",
    "\n",
    "        barcode = barcodes[0]\n",
    "        if not check_fastq_barcode_is_equal(barcode.i7_sequence, barcode.i5_sequence, row[\"barcode\"]):\n",
    "            print(f\"Barcode mismatch {subpool.plate_name()} {subpool.index} {barcode.i7_sequence} {barcode.i5_sequence} {row.barcode} {row.filename}\")\n",
    "\n",
    "test_subpool_lookup(fastqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "excess-vanilla",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('UDI01', 13000, 'EX'),\n",
       " ('UDI02', 67000, 'NO'),\n",
       " ('UDI03', 67000, 'NO'),\n",
       " ('UDI04', 67000, 'NO'),\n",
       " ('UDI05', 67000, 'NO'),\n",
       " ('UDI06', 67000, 'NO'),\n",
       " ('UDI07', 67000, 'NO'),\n",
       " ('UDI08', 67000, 'NO'),\n",
       " ('UDI09', 67000, 'NO'),\n",
       " ('UDI10', 67000, 'NO'),\n",
       " ('UDI11', 67000, 'NO'),\n",
       " ('UDI12', 67000, 'NO'),\n",
       " ('UDI13', 67000, 'NO'),\n",
       " ('UDI14', 67000, 'NO'),\n",
       " ('UDI15', 48256, 'NO')]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[(x.index, x.nuclei, x.selection_type) for x in models.Subpool.objects.filter(plate__name=\"IGVF_013\")]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "distant-atmosphere",
   "metadata": {},
   "source": [
    "Load information about previously submitted records"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "imported-placement",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loaded 2132 submitted sequence_file records\n"
     ]
    }
   ],
   "source": [
    "submitted_sequence_file = {}\n",
    "\n",
    "for accession_prefix in submitted_sheets:\n",
    "    for i, row in submitted_sheets[accession_prefix][\"sequence_file\"].iterrows():\n",
    "        if not pandas.isnull(row[\"accession\"]):\n",
    "            if accession_prefix == \"igvftst\":\n",
    "                see_also_template = \"https://api.sandbox.igvf.org/sequence-files/{}/\"\n",
    "            elif accession_prefix == \"igvf\":\n",
    "                see_also_template = \"https://api.data.igvf.org/sequence-files/{}/\"\n",
    "            \n",
    "            if pandas.notnull(row[\"accession\"]):\n",
    "                submitted_sequence_file.setdefault(row[\"submitted_file_name\"], []).append({\n",
    "                    \"accession_prefix\": accession_prefix,\n",
    "                    \"name\": row[\"accession\"],\n",
    "                    \"uuid\": row[\"uuid\"],\n",
    "                    \"see_also\": see_also_template.format(row[\"accession\"]),\n",
    "                })\n",
    "        \n",
    "print(\"loaded {} submitted sequence_file records\".format(len(submitted_sequence_file)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "numerous-marketplace",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['igvftst', 'igvf'])"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "submitted_sheets.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "harmful-builder",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'a': 1}\n",
      "{'a': 1, 'b': 3}\n",
      "{'a': 1, 'b': 3, 'c': 4}\n",
      "{'a': 1, 'b': 3, 'c': 4}\n"
     ]
    }
   ],
   "source": [
    "def built_in_cache(a, b, cache={}):\n",
    "    cache.setdefault(a, b)\n",
    "    print(cache)\n",
    "    \n",
    "built_in_cache('a', 1)\n",
    "built_in_cache('b', 3)\n",
    "built_in_cache('c', 4)\n",
    "built_in_cache('a', 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "second-hopkins",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "supreme-projector",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequencing_experiments = {}\n",
    "sequencing_subpools = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "armed-education",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loaded 173 measurement_sets\n",
      "updating 003_13A igvf_003/nextseq2 measurement set to 003_13A_illumina\n",
      "updating 003_67A igvf_003/nextseq measurement set to 003_67A_illumina\n",
      "updating 003_67B igvf_003/nextseq measurement set to 003_67B_illumina\n",
      "updating 003_67I igvf_003/nova1 measurement set to 003_67I_illumina\n",
      "updating 003_67I igvf_003/nova2 measurement set to 003_67I_illumina\n",
      "updating 003_67J igvf_003/nova1 measurement set to 003_67J_illumina\n",
      "updating 003_67J igvf_003/nova2 measurement set to 003_67J_illumina\n",
      "updating 003_67K igvf_003/nova1 measurement set to 003_67K_illumina\n",
      "updating 003_67K igvf_003/nova2 measurement set to 003_67K_illumina\n",
      "updating 003_67L igvf_003/nova1 measurement set to 003_67L_illumina\n",
      "updating 003_67L igvf_003/nova2 measurement set to 003_67L_illumina\n",
      "updating 003_67M igvf_003/nova1 measurement set to 003_67M_illumina\n",
      "updating 003_67M igvf_003/nova2 measurement set to 003_67M_illumina\n",
      "updating 003_67N igvf_003/nova1 measurement set to 003_67N_illumina\n",
      "updating 003_67N igvf_003/nova2 measurement set to 003_67N_illumina\n",
      "updating 003_67O igvf_003/nova1 measurement set to 003_67O_illumina\n",
      "updating 003_67O igvf_003/nova2 measurement set to 003_67O_illumina\n",
      "updating 003_67A igvf_003/nova1 measurement set to 003_67A_illumina\n",
      "updating 003_67A igvf_003/nova2 measurement set to 003_67A_illumina\n",
      "updating 003_67B igvf_003/nova1 measurement set to 003_67B_illumina\n",
      "updating 003_67B igvf_003/nova2 measurement set to 003_67B_illumina\n",
      "updating 003_67C igvf_003/nova1 measurement set to 003_67C_illumina\n",
      "updating 003_67C igvf_003/nova2 measurement set to 003_67C_illumina\n",
      "updating 003_67D igvf_003/nova1 measurement set to 003_67D_illumina\n",
      "updating 003_67D igvf_003/nova2 measurement set to 003_67D_illumina\n",
      "updating 003_67E igvf_003/nova1 measurement set to 003_67E_illumina\n",
      "updating 003_67E igvf_003/nova2 measurement set to 003_67E_illumina\n",
      "updating 003_67F igvf_003/nova1 measurement set to 003_67F_illumina\n",
      "updating 003_67F igvf_003/nova2 measurement set to 003_67F_illumina\n",
      "updating 003_67G igvf_003/nova1 measurement set to 003_67G_illumina\n",
      "updating 003_67G igvf_003/nova2 measurement set to 003_67G_illumina\n",
      "updating 003_67H igvf_003/nova1 measurement set to 003_67H_illumina\n",
      "updating 003_67H igvf_003/nova2 measurement set to 003_67H_illumina\n",
      "updating 004_13A igvf_004/nextseq2 measurement set to 004_13A_illumina\n",
      "updating 004_67A igvf_004/nextseq measurement set to 004_67A_illumina\n",
      "updating 004_67B igvf_004/nextseq measurement set to 004_67B_illumina\n",
      "updating 004_67I igvf_004/nova1 measurement set to 004_67I_illumina\n",
      "updating 004_67I igvf_004/nova2 measurement set to 004_67I_illumina\n",
      "updating 004_67J igvf_004/nova1 measurement set to 004_67J_illumina\n",
      "updating 004_67J igvf_004/nova2 measurement set to 004_67J_illumina\n",
      "updating 004_67K igvf_004/nova1 measurement set to 004_67K_illumina\n",
      "updating 004_67K igvf_004/nova2 measurement set to 004_67K_illumina\n",
      "updating 004_67L igvf_004/nova1 measurement set to 004_67L_illumina\n",
      "updating 004_67L igvf_004/nova2 measurement set to 004_67L_illumina\n",
      "updating 004_67M igvf_004/nova1 measurement set to 004_67M_illumina\n",
      "updating 004_67M igvf_004/nova2 measurement set to 004_67M_illumina\n",
      "updating 004_67N igvf_004/nova1 measurement set to 004_67N_illumina\n",
      "updating 004_67N igvf_004/nova2 measurement set to 004_67N_illumina\n",
      "updating 004_67O igvf_004/nova1 measurement set to 004_67O_illumina\n",
      "updating 004_67O igvf_004/nova2 measurement set to 004_67O_illumina\n",
      "updating 004_67A igvf_004/nova1 measurement set to 004_67A_illumina\n",
      "updating 004_67A igvf_004/nova2 measurement set to 004_67A_illumina\n",
      "updating 004_67B igvf_004/nova1 measurement set to 004_67B_illumina\n",
      "updating 004_67B igvf_004/nova2 measurement set to 004_67B_illumina\n",
      "updating 004_67C igvf_004/nova1 measurement set to 004_67C_illumina\n",
      "updating 004_67C igvf_004/nova2 measurement set to 004_67C_illumina\n",
      "updating 004_67D igvf_004/nova1 measurement set to 004_67D_illumina\n",
      "updating 004_67D igvf_004/nova2 measurement set to 004_67D_illumina\n",
      "updating 004_67E igvf_004/nova1 measurement set to 004_67E_illumina\n",
      "updating 004_67E igvf_004/nova2 measurement set to 004_67E_illumina\n",
      "updating 004_67F igvf_004/nova1 measurement set to 004_67F_illumina\n",
      "updating 004_67F igvf_004/nova2 measurement set to 004_67F_illumina\n",
      "updating 004_67G igvf_004/nova1 measurement set to 004_67G_illumina\n",
      "updating 004_67G igvf_004/nova2 measurement set to 004_67G_illumina\n",
      "updating 004_67H igvf_004/nova1 measurement set to 004_67H_illumina\n",
      "updating 004_67H igvf_004/nova2 measurement set to 004_67H_illumina\n",
      "updating 005_13A igvf_005/nextseq2 measurement set to 005_13A_illumina\n",
      "updating 005_67I igvf_005/nova1 measurement set to 005_67I_illumina\n",
      "updating 005_67I igvf_005/nova2 measurement set to 005_67I_illumina\n",
      "updating 005_67J igvf_005/nova1 measurement set to 005_67J_illumina\n",
      "updating 005_67J igvf_005/nova2 measurement set to 005_67J_illumina\n",
      "updating 005_67K igvf_005/nova1 measurement set to 005_67K_illumina\n",
      "updating 005_67K igvf_005/nova2 measurement set to 005_67K_illumina\n",
      "updating 005_67L igvf_005/nova1 measurement set to 005_67L_illumina\n",
      "updating 005_67L igvf_005/nova2 measurement set to 005_67L_illumina\n",
      "updating 005_67M igvf_005/nova1 measurement set to 005_67M_illumina\n",
      "updating 005_67M igvf_005/nova2 measurement set to 005_67M_illumina\n",
      "updating 005_67N igvf_005/nova1 measurement set to 005_67N_illumina\n",
      "updating 005_67N igvf_005/nova2 measurement set to 005_67N_illumina\n",
      "updating 005_67O igvf_005/nova1 measurement set to 005_67O_illumina\n",
      "updating 005_67O igvf_005/nova2 measurement set to 005_67O_illumina\n",
      "updating 005_67A igvf_005/nova1 measurement set to 005_67A_illumina\n",
      "updating 005_67A igvf_005/nova2 measurement set to 005_67A_illumina\n",
      "updating 005_67B igvf_005/nova1 measurement set to 005_67B_illumina\n",
      "updating 005_67B igvf_005/nova2 measurement set to 005_67B_illumina\n",
      "updating 005_67C igvf_005/nova1 measurement set to 005_67C_illumina\n",
      "updating 005_67C igvf_005/nova2 measurement set to 005_67C_illumina\n",
      "updating 005_67D igvf_005/nova1 measurement set to 005_67D_illumina\n",
      "updating 005_67D igvf_005/nova2 measurement set to 005_67D_illumina\n",
      "updating 005_67E igvf_005/nova1 measurement set to 005_67E_illumina\n",
      "updating 005_67E igvf_005/nova2 measurement set to 005_67E_illumina\n",
      "updating 005_67F igvf_005/nova1 measurement set to 005_67F_illumina\n",
      "updating 005_67F igvf_005/nova2 measurement set to 005_67F_illumina\n",
      "updating 005_67G igvf_005/nova1 measurement set to 005_67G_illumina\n",
      "updating 005_67G igvf_005/nova2 measurement set to 005_67G_illumina\n",
      "updating 005_67H igvf_005/nova1 measurement set to 005_67H_illumina\n",
      "updating 005_67H igvf_005/nova2 measurement set to 005_67H_illumina\n",
      "updating 007_13A igvf_007/nextseq2 measurement set to 007_13A_illumina\n",
      "updating 007_67I igvf_007/nova1 measurement set to 007_67I_illumina\n",
      "updating 007_67I igvf_007/nova2 measurement set to 007_67I_illumina\n",
      "updating 007_67J igvf_007/nova1 measurement set to 007_67J_illumina\n",
      "updating 007_67J igvf_007/nova2 measurement set to 007_67J_illumina\n",
      "updating 007_67K igvf_007/nova1 measurement set to 007_67K_illumina\n",
      "updating 007_67K igvf_007/nova2 measurement set to 007_67K_illumina\n",
      "updating 007_67L igvf_007/nova1 measurement set to 007_67L_illumina\n",
      "updating 007_67L igvf_007/nova2 measurement set to 007_67L_illumina\n",
      "updating 007_67M igvf_007/nova1 measurement set to 007_67M_illumina\n",
      "updating 007_67M igvf_007/nova2 measurement set to 007_67M_illumina\n",
      "updating 007_67N igvf_007/nova1 measurement set to 007_67N_illumina\n",
      "updating 007_67N igvf_007/nova2 measurement set to 007_67N_illumina\n",
      "updating 007_67O igvf_007/nova1 measurement set to 007_67O_illumina\n",
      "updating 007_67O igvf_007/nova2 measurement set to 007_67O_illumina\n",
      "updating 007_67A igvf_007/nova1 measurement set to 007_67A_illumina\n",
      "updating 007_67A igvf_007/nova2 measurement set to 007_67A_illumina\n",
      "updating 007_67B igvf_007/nova1 measurement set to 007_67B_illumina\n",
      "updating 007_67B igvf_007/nova2 measurement set to 007_67B_illumina\n",
      "updating 007_67C igvf_007/nova1 measurement set to 007_67C_illumina\n",
      "updating 007_67C igvf_007/nova2 measurement set to 007_67C_illumina\n",
      "updating 007_67D igvf_007/nova1 measurement set to 007_67D_illumina\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "updating 007_67D igvf_007/nova2 measurement set to 007_67D_illumina\n",
      "updating 007_67E igvf_007/nova1 measurement set to 007_67E_illumina\n",
      "updating 007_67E igvf_007/nova2 measurement set to 007_67E_illumina\n",
      "updating 007_67F igvf_007/nova1 measurement set to 007_67F_illumina\n",
      "updating 007_67F igvf_007/nova2 measurement set to 007_67F_illumina\n",
      "updating 007_67G igvf_007/nova1 measurement set to 007_67G_illumina\n",
      "updating 007_67G igvf_007/nova2 measurement set to 007_67G_illumina\n",
      "updating 007_67H igvf_007/nova1 measurement set to 007_67H_illumina\n",
      "updating 007_67H igvf_007/nova2 measurement set to 007_67H_illumina\n",
      "updating 008_67C igvf_008/nextseq measurement set to 008_67C_illumina\n",
      "updating 008_67G igvf_008/nextseq measurement set to 008_67G_illumina\n",
      "updating 008_67I igvf_008/nova1 measurement set to 008_67I_illumina\n",
      "updating 008_67A igvf_008/nova1 measurement set to 008_67A_illumina\n",
      "updating 008_67B igvf_008/nova1 measurement set to 008_67B_illumina\n",
      "updating 008_67C igvf_008/nova1 measurement set to 008_67C_illumina\n",
      "updating 008_67D igvf_008/nova1 measurement set to 008_67D_illumina\n",
      "updating 008_67E igvf_008/nova1 measurement set to 008_67E_illumina\n",
      "updating 008_67F igvf_008/nova1 measurement set to 008_67F_illumina\n",
      "updating 008_67G igvf_008/nova1 measurement set to 008_67G_illumina\n",
      "updating 008_67H igvf_008/nova1 measurement set to 008_67H_illumina\n",
      "updating 008B_13A igvf_008b/nextseq2 measurement set to 008B_13A_illumina\n",
      "updating 008B_67G igvf_008b/nextseq measurement set to 008B_67G_illumina\n",
      "updating 008B_67I igvf_008b/nextseq measurement set to 008B_67I_illumina\n",
      "updating 008B_67I igvf_008b/nova1 measurement set to 008B_67I_illumina\n",
      "updating 008B_67J igvf_008b/nova1 measurement set to 008B_67J_illumina\n",
      "updating 008B_67A igvf_008b/nova1 measurement set to 008B_67A_illumina\n",
      "updating 008B_67B igvf_008b/nova1 measurement set to 008B_67B_illumina\n",
      "updating 008B_67C igvf_008b/nova1 measurement set to 008B_67C_illumina\n",
      "updating 008B_67D igvf_008b/nova1 measurement set to 008B_67D_illumina\n",
      "updating 008B_67E igvf_008b/nova1 measurement set to 008B_67E_illumina\n",
      "updating 008B_67F igvf_008b/nova1 measurement set to 008B_67F_illumina\n",
      "updating 008B_67G igvf_008b/nova1 measurement set to 008B_67G_illumina\n",
      "updating 008B_67H igvf_008b/nova1 measurement set to 008B_67H_illumina\n",
      "updating 009_13A igvf_009/nextseq2 measurement set to 009_13A_illumina\n",
      "updating 009_67A igvf_009/nextseq measurement set to 009_67A_illumina\n",
      "updating 009_67O igvf_009/nextseq measurement set to 009_67O_illumina\n",
      "updating 009_67I igvf_009/nova1 measurement set to 009_67I_illumina\n",
      "updating 009_67I igvf_009/nova2 measurement set to 009_67I_illumina\n",
      "updating 009_67J igvf_009/nova1 measurement set to 009_67J_illumina\n",
      "updating 009_67J igvf_009/nova2 measurement set to 009_67J_illumina\n",
      "updating 009_67K igvf_009/nova1 measurement set to 009_67K_illumina\n",
      "updating 009_67K igvf_009/nova2 measurement set to 009_67K_illumina\n",
      "updating 009_67L igvf_009/nova1 measurement set to 009_67L_illumina\n",
      "updating 009_67L igvf_009/nova2 measurement set to 009_67L_illumina\n",
      "updating 009_67M igvf_009/nova1 measurement set to 009_67M_illumina\n",
      "updating 009_67M igvf_009/nova2 measurement set to 009_67M_illumina\n",
      "updating 009_67N igvf_009/nova1 measurement set to 009_67N_illumina\n",
      "updating 009_67N igvf_009/nova2 measurement set to 009_67N_illumina\n",
      "updating 009_67O igvf_009/nova1 measurement set to 009_67O_illumina\n",
      "updating 009_67O igvf_009/nova2 measurement set to 009_67O_illumina\n",
      "updating 009_67A igvf_009/nova1 measurement set to 009_67A_illumina\n",
      "updating 009_67A igvf_009/nova2 measurement set to 009_67A_illumina\n",
      "updating 009_67B igvf_009/nova1 measurement set to 009_67B_illumina\n",
      "updating 009_67B igvf_009/nova2 measurement set to 009_67B_illumina\n",
      "updating 009_67C igvf_009/nova1 measurement set to 009_67C_illumina\n",
      "updating 009_67C igvf_009/nova2 measurement set to 009_67C_illumina\n",
      "updating 009_67D igvf_009/nova1 measurement set to 009_67D_illumina\n",
      "updating 009_67D igvf_009/nova2 measurement set to 009_67D_illumina\n",
      "updating 009_67E igvf_009/nova1 measurement set to 009_67E_illumina\n",
      "updating 009_67E igvf_009/nova2 measurement set to 009_67E_illumina\n",
      "updating 009_67F igvf_009/nova1 measurement set to 009_67F_illumina\n",
      "updating 009_67F igvf_009/nova2 measurement set to 009_67F_illumina\n",
      "updating 009_67G igvf_009/nova1 measurement set to 009_67G_illumina\n",
      "updating 009_67G igvf_009/nova2 measurement set to 009_67G_illumina\n",
      "updating 009_67H igvf_009/nova1 measurement set to 009_67H_illumina\n",
      "updating 009_67H igvf_009/nova2 measurement set to 009_67H_illumina\n",
      "updating 010_13A igvf_010/nextseq2 measurement set to 010_13A_illumina\n",
      "updating 010_67D igvf_010/nextseq measurement set to 010_67D_illumina\n",
      "updating 010_67O igvf_010/nextseq measurement set to 010_67O_illumina\n",
      "updating 010_67I igvf_010/nova1 measurement set to 010_67I_illumina\n",
      "updating 010_67I igvf_010/nova2 measurement set to 010_67I_illumina\n",
      "updating 010_67J igvf_010/nova1 measurement set to 010_67J_illumina\n",
      "updating 010_67J igvf_010/nova2 measurement set to 010_67J_illumina\n",
      "updating 010_67K igvf_010/nova1 measurement set to 010_67K_illumina\n",
      "updating 010_67K igvf_010/nova2 measurement set to 010_67K_illumina\n",
      "updating 010_67L igvf_010/nova1 measurement set to 010_67L_illumina\n",
      "updating 010_67L igvf_010/nova2 measurement set to 010_67L_illumina\n",
      "updating 010_67M igvf_010/nova1 measurement set to 010_67M_illumina\n",
      "updating 010_67M igvf_010/nova2 measurement set to 010_67M_illumina\n",
      "updating 010_67N igvf_010/nova1 measurement set to 010_67N_illumina\n",
      "updating 010_67N igvf_010/nova2 measurement set to 010_67N_illumina\n",
      "updating 010_67O igvf_010/nova1 measurement set to 010_67O_illumina\n",
      "updating 010_67O igvf_010/nova2 measurement set to 010_67O_illumina\n",
      "updating 010_67A igvf_010/nova1 measurement set to 010_67A_illumina\n",
      "updating 010_67A igvf_010/nova2 measurement set to 010_67A_illumina\n",
      "updating 010_67B igvf_010/nova1 measurement set to 010_67B_illumina\n",
      "updating 010_67B igvf_010/nova2 measurement set to 010_67B_illumina\n",
      "updating 010_67C igvf_010/nova1 measurement set to 010_67C_illumina\n",
      "updating 010_67C igvf_010/nova2 measurement set to 010_67C_illumina\n",
      "updating 010_67D igvf_010/nova1 measurement set to 010_67D_illumina\n",
      "updating 010_67D igvf_010/nova2 measurement set to 010_67D_illumina\n",
      "updating 010_67E igvf_010/nova1 measurement set to 010_67E_illumina\n",
      "updating 010_67E igvf_010/nova2 measurement set to 010_67E_illumina\n",
      "updating 010_67F igvf_010/nova1 measurement set to 010_67F_illumina\n",
      "updating 010_67F igvf_010/nova2 measurement set to 010_67F_illumina\n",
      "updating 010_67G igvf_010/nova1 measurement set to 010_67G_illumina\n",
      "updating 010_67G igvf_010/nova2 measurement set to 010_67G_illumina\n",
      "updating 010_67H igvf_010/nova1 measurement set to 010_67H_illumina\n",
      "updating 010_67H igvf_010/nova2 measurement set to 010_67H_illumina\n",
      "updating 011_13A igvf_011/nextseq2 measurement set to 011_13A_illumina\n",
      "updating 011_67M igvf_011/nextseq measurement set to 011_67M_illumina\n",
      "updating 011_67O igvf_011/nextseq measurement set to 011_67O_illumina\n",
      "updating 011_67I igvf_011/nova1 measurement set to 011_67I_illumina\n",
      "updating 011_67I igvf_011/nova2 measurement set to 011_67I_illumina\n",
      "updating 011_67J igvf_011/nova1 measurement set to 011_67J_illumina\n",
      "updating 011_67J igvf_011/nova2 measurement set to 011_67J_illumina\n",
      "updating 011_67K igvf_011/nova1 measurement set to 011_67K_illumina\n",
      "updating 011_67K igvf_011/nova2 measurement set to 011_67K_illumina\n",
      "updating 011_67L igvf_011/nova1 measurement set to 011_67L_illumina\n",
      "updating 011_67L igvf_011/nova2 measurement set to 011_67L_illumina\n",
      "updating 011_67M igvf_011/nova1 measurement set to 011_67M_illumina\n",
      "updating 011_67M igvf_011/nova2 measurement set to 011_67M_illumina\n",
      "updating 011_67N igvf_011/nova1 measurement set to 011_67N_illumina\n",
      "updating 011_67N igvf_011/nova2 measurement set to 011_67N_illumina\n",
      "updating 011_67O igvf_011/nova1 measurement set to 011_67O_illumina\n",
      "updating 011_67O igvf_011/nova2 measurement set to 011_67O_illumina\n",
      "updating 011_67A igvf_011/nova1 measurement set to 011_67A_illumina\n",
      "updating 011_67A igvf_011/nova2 measurement set to 011_67A_illumina\n",
      "updating 011_67B igvf_011/nova1 measurement set to 011_67B_illumina\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "updating 011_67B igvf_011/nova2 measurement set to 011_67B_illumina\n",
      "updating 011_67C igvf_011/nova1 measurement set to 011_67C_illumina\n",
      "updating 011_67C igvf_011/nova2 measurement set to 011_67C_illumina\n",
      "updating 011_67D igvf_011/nova1 measurement set to 011_67D_illumina\n",
      "updating 011_67D igvf_011/nova2 measurement set to 011_67D_illumina\n",
      "updating 011_67E igvf_011/nova1 measurement set to 011_67E_illumina\n",
      "updating 011_67E igvf_011/nova2 measurement set to 011_67E_illumina\n",
      "updating 011_67F igvf_011/nova1 measurement set to 011_67F_illumina\n",
      "updating 011_67F igvf_011/nova2 measurement set to 011_67F_illumina\n",
      "updating 011_67G igvf_011/nova1 measurement set to 011_67G_illumina\n",
      "updating 011_67G igvf_011/nova2 measurement set to 011_67G_illumina\n",
      "updating 011_67H igvf_011/nova1 measurement set to 011_67H_illumina\n",
      "updating 011_67H igvf_011/nova2 measurement set to 011_67H_illumina\n",
      "updating 012_13A igvf_012/nextseq measurement set to 012_13A_illumina\n",
      "updating 012_13B igvf_012/nextseq measurement set to 012_13B_illumina\n",
      "updating 012_13C igvf_012/nextseq measurement set to 012_13C_illumina\n",
      "updating 012_13D igvf_012/nextseq measurement set to 012_13D_illumina\n",
      "updating 012_13E igvf_012/nextseq measurement set to 012_13E_illumina\n",
      "updating 012_13F igvf_012/nextseq measurement set to 012_13F_illumina\n",
      "updating 012_13G igvf_012/nextseq measurement set to 012_13G_illumina\n",
      "updating 012_13H igvf_012/nextseq measurement set to 012_13H_illumina\n",
      "updating 013_13A igvf_013/nextseq measurement set to 013_13A_illumina\n",
      "updating 013_67A igvf_013/nova1 measurement set to 013_67A_illumina\n",
      "updating 013_67B igvf_013/nova1 measurement set to 013_67B_illumina\n",
      "updating 013_67C igvf_013/nova1 measurement set to 013_67C_illumina\n",
      "updating 013_67D igvf_013/nova1 measurement set to 013_67D_illumina\n",
      "updating 013_67E igvf_013/nova1 measurement set to 013_67E_illumina\n",
      "updating 013_67F igvf_013/nova1 measurement set to 013_67F_illumina\n",
      "updating 013_67G igvf_013/nova1 measurement set to 013_67G_illumina\n",
      "updating 013_67H igvf_013/nova1 measurement set to 013_67H_illumina\n",
      "updating 013_67H igvf_013/nextseq measurement set to 013_67H_illumina\n",
      "updating 013_67I igvf_013/nova1 measurement set to 013_67I_illumina\n",
      "updating 013_67J igvf_013/nova1 measurement set to 013_67J_illumina\n",
      "updating 013_67J igvf_013/nextseq measurement set to 013_67J_illumina\n",
      "updating 013_67K igvf_013/nova1 measurement set to 013_67K_illumina\n",
      "updating 013_67L igvf_013/nova1 measurement set to 013_67L_illumina\n",
      "updating 013_67M igvf_013/nova1 measurement set to 013_67M_illumina\n",
      "updating 013_67N igvf_013/nova1 measurement set to 013_67N_illumina\n",
      "updating 014_13A igvf_014/nextseq measurement set to 014_13A_illumina\n",
      "updating 014_13B igvf_014/nextseq measurement set to 014_13B_illumina\n",
      "updating 014_13C igvf_014/nextseq measurement set to 014_13C_illumina\n",
      "updating 014_13D igvf_014/nextseq measurement set to 014_13D_illumina\n",
      "updating 014_13E igvf_014/nextseq measurement set to 014_13E_illumina\n",
      "updating 014_13F igvf_014/nextseq measurement set to 014_13F_illumina\n",
      "updating 014_13G igvf_014/nextseq measurement set to 014_13G_illumina\n",
      "updating 014_13H igvf_014/nextseq measurement set to 014_13H_illumina\n",
      "updating B01_13A igvf_b01/next2 measurement set to B01_13A_illumina\n",
      "updating B01_13A igvf_b01/next1 measurement set to B01_13A_illumina\n",
      "updating B01_13B igvf_b01/next2 measurement set to B01_13B_illumina\n",
      "updating B01_13B igvf_b01/next1 measurement set to B01_13B_illumina\n",
      "updating B01_13C igvf_b01/next2 measurement set to B01_13C_illumina\n",
      "updating B01_13C igvf_b01/next1 measurement set to B01_13C_illumina\n",
      "updating B01_13D igvf_b01/next2 measurement set to B01_13D_illumina\n",
      "updating B01_13D igvf_b01/next1 measurement set to B01_13D_illumina\n",
      "updating B01_13E igvf_b01/next2 measurement set to B01_13E_illumina\n",
      "updating B01_13E igvf_b01/next1 measurement set to B01_13E_illumina\n",
      "updating B01_13F igvf_b01/next2 measurement set to B01_13F_illumina\n",
      "updating B01_13F igvf_b01/next1 measurement set to B01_13F_illumina\n",
      "updating B01_13G igvf_b01/next2 measurement set to B01_13G_illumina\n",
      "updating B01_13G igvf_b01/next1 measurement set to B01_13G_illumina\n",
      "updating B01_13H igvf_b01/next2 measurement set to B01_13H_illumina\n",
      "updating B01_13H igvf_b01/next1 measurement set to B01_13H_illumina\n",
      "total plates 14\n",
      "total sequencing runs 38\n",
      "total subpools 255\n",
      "total subpool in run 304\n",
      "total subpool in run files 3119\n"
     ]
    }
   ],
   "source": [
    "def get_or_create_sequencing_run(name, run_date, platform, plate, accessions=None):\n",
    "    if name in sequencing_experiments:\n",
    "        return sequencing_experiments[name]\n",
    "    \n",
    "    try:\n",
    "        sequencing_run = models.SequencingRun.objects.get(name=name)\n",
    "    except models.SequencingRun.DoesNotExist:\n",
    "        sequencing_run = models.SequencingRun(\n",
    "            name=name,\n",
    "            run_date=run_date,\n",
    "            platform=platform,\n",
    "            plate=plate,\n",
    "            stranded=models.StrandedEnum.REVERSE,\n",
    "        )\n",
    "        sequencing_run.save()\n",
    "        if accessions is not None:\n",
    "            sequencing_run.accession.set(accessions)\n",
    "            sequencing_run.save\n",
    "        \n",
    "    sequencing_experiments[name] = sequencing_run\n",
    "    return sequencing_run\n",
    "\n",
    "def get_or_create_subpool_in_run(subpool, sequencing_run, status, raw_reads=None, measurement_set=None):\n",
    "    try:\n",
    "        subpool_in_run = models.SubpoolInRun.objects.get(subpool=subpool, sequencing_run=sequencing_run)\n",
    "    except models.SubpoolInRun.DoesNotExist:\n",
    "        subpool_in_run = models.SubpoolInRun(\n",
    "            subpool=subpool, \n",
    "            sequencing_run=sequencing_run,\n",
    "            raw_reads=None,\n",
    "            status=status,\n",
    "            measurement_set=measurement_set,\n",
    "        )\n",
    "        subpool_in_run.save()\n",
    "        \n",
    "    sequencing_subpools[name] = subpool_in_run\n",
    "    return subpool_in_run\n",
    "\n",
    "\n",
    "def lane_or_none(value):\n",
    "    if pandas.isnull(value):\n",
    "        return None\n",
    "    elif isinstance(value, float):\n",
    "        return int(value)\n",
    "    elif value.startswith(\"L\"):\n",
    "        return int(value[1:])\n",
    "    else:\n",
    "        return int(value)\n",
    "\n",
    "def load_fastqs(submittable_subpools):\n",
    "    fastqs = load_fastq_csv()\n",
    "    measurement_sets = load_submitted_measurement_sets()\n",
    "\n",
    "    for i, row in fastqs.iterrows():\n",
    "        #plate_name = normalize_plate_name(row.experiment)\n",
    "        #subpool_name = \"{}_{}\".format(row.plate_id, row.subpool_name)\n",
    "        #index_id = int(row.index_id)\n",
    "        fastq_relative = Path(row.filename)\n",
    "        sequencing_run_name = str(Path(fastq_relative.parts[0])/fastq_relative.parts[1])\n",
    "        sequencer = row.get(\"sequencer\")\n",
    "        run_date = row[\"ctime\"]\n",
    "        md5sum = row[\"md5sum\"]\n",
    "\n",
    "        #if row[\"read_id\"] in ('I1',):\n",
    "        #    continue\n",
    "\n",
    "        subpool = get_subpool_from_fastq_row(row)\n",
    "\n",
    "        if subpool is None:\n",
    "            print(\"Unable to import {row.filename}\")\n",
    "            continue\n",
    "\n",
    "        current_plate = subpool.plate\n",
    "\n",
    "        platform = {\n",
    "            \"A00850\": models.Platform.objects.get(display_name=\"Novaseq 6000\"),\n",
    "            \"VH00582\": models.Platform.objects.get(display_name=\"Nextseq 2000\"),\n",
    "            \"minion\": models.Platform.objects.get(name=\"minion\"),\n",
    "            \"promethion\": models.Platform.objects.get(name=\"promethion\"),\n",
    "        }[sequencer]\n",
    "\n",
    "        subpool_status = {\n",
    "            \"Yes\": models.RunStatusEnum.PASS,\n",
    "            \"No\": models.RunStatusEnum.FAILED,\n",
    "            None: models.RunStatusEnum.FAILED,\n",
    "        }[submittable_subpools.get(subpool.name)]\n",
    "\n",
    "        sequencing_run = get_or_create_sequencing_run(sequencing_run_name, run_date, platform, current_plate, accessions=None)\n",
    "        subpool_run = get_or_create_subpool_in_run(subpool=subpool, sequencing_run=sequencing_run, status=subpool_status)\n",
    "\n",
    "        for accession_prefix in [\"igvftst\", \"igvf\"]:\n",
    "            file_set_name = None\n",
    "            submitted_md5s = submitted_sheets[accession_prefix][\"sequence_file\"].set_index(\"md5sum\")\n",
    "            if row.md5sum in submitted_md5s.index:\n",
    "                file_set_alias = submitted_md5s.loc[row.md5sum][\"file_set\"]\n",
    "                file_set_name = file_set_alias.split(\":\")[1]\n",
    "\n",
    "        if file_set_name is None:\n",
    "            pass\n",
    "        elif file_set_name not in measurement_sets:\n",
    "            # skip if there hasn't been a submitted measurement_set\n",
    "            pass\n",
    "        elif subpool_run.measurement_set is None:\n",
    "            subpool_run.measurement_set = measurement_sets[file_set_name]\n",
    "            subpool_run.save()\n",
    "            print(f\"updating {subpool_run} measurement set to {measurement_sets[file_set_name]}\")\n",
    "        else:\n",
    "            assert subpool_run.measurement_set.name == file_set_name, f\"SubpoolInRun measurement_set mismatch {subpool_run.measurement_set.name} {file_set_name}\"\n",
    "\n",
    "\n",
    "        subpool_file = models.SubpoolInRunFile(\n",
    "            sequencing_run=sequencing_run,\n",
    "            subpool_run=subpool_run,\n",
    "            md5sum=md5sum,\n",
    "            filename=str(fastq_relative),\n",
    "            flowcell_id=row[\"flowcell_id\"],\n",
    "        )\n",
    "        if not pandas.isnull(row[\"lane\"]):\n",
    "            subpool_file.lane = lane_or_none(row[\"lane\"])\n",
    "        if not pandas.isnull(row[\"read\"]):\n",
    "            subpool_file.read = row[\"read\"]\n",
    "        subpool_file.save()\n",
    "\n",
    "        for accession_info in submitted_sequence_file.get(str(fastq_relative), []):\n",
    "            if pandas.notnull(accession_info[\"name\"]):\n",
    "                accession = models.Accession(\n",
    "                    accession_prefix=accession_info[\"accession_prefix\"],\n",
    "                    name=accession_info[\"name\"],\n",
    "                    uuid=accession_info[\"uuid\"],\n",
    "                    see_also=accession_info[\"see_also\"],\n",
    "                )\n",
    "                accession.save()                \n",
    "                subpool_file.accession.add(accession)\n",
    "                subpool_file.save()\n",
    "\n",
    "load_fastqs(submittable_subpools)\n",
    "\n",
    "print(\"total plates\", models.SplitSeqPlate.objects.count())\n",
    "print(\"total sequencing runs\", models.SequencingRun.objects.count())\n",
    "print(\"total subpools\", models.Subpool.objects.count())\n",
    "print(\"total subpool in run\", models.SubpoolInRun.objects.count())\n",
    "print(\"total subpool in run files\", models.SubpoolInRunFile.objects.count())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "cosmetic-calculation",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "014_13H UDI24 10400 NO\n"
     ]
    }
   ],
   "source": [
    "for q in models.Subpool.objects.filter(plate__name=\"IGVF_014\", index=\"UDI24\"):\n",
    "    print(q.name, q.index, q.nuclei, q.selection_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "atomic-intake",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('013_13A', 'UDI01', 'EX'),\n",
       " ('013_67A', 'UDI02', 'NO'),\n",
       " ('013_67B', 'UDI03', 'NO'),\n",
       " ('013_67C', 'UDI04', 'NO'),\n",
       " ('013_67D', 'UDI05', 'NO'),\n",
       " ('013_67E', 'UDI06', 'NO'),\n",
       " ('013_67F', 'UDI07', 'NO'),\n",
       " ('013_67G', 'UDI08', 'NO'),\n",
       " ('013_67H', 'UDI09', 'NO'),\n",
       " ('013_67I', 'UDI10', 'NO'),\n",
       " ('013_67J', 'UDI11', 'NO'),\n",
       " ('013_67K', 'UDI12', 'NO'),\n",
       " ('013_67L', 'UDI13', 'NO'),\n",
       " ('013_67M', 'UDI14', 'NO'),\n",
       " ('013_67N', 'UDI15', 'NO')]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[(x.name, x.index, x.selection_type) for x in models.Subpool.objects.filter(plate__name=\"IGVF_013\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "retired-geography",
   "metadata": {},
   "outputs": [],
   "source": [
    "def print_subpool_submission_status(fastqs):\n",
    "    for i, row in fastqs.iterrows():\n",
    "        plate_name = normalize_plate_name(row.experiment)\n",
    "        index_id = row.index_id\n",
    "        fastq_relative = Path(row.filename)\n",
    "        sequencing_run_name = str(Path(fastq_relative.parts[0])/fastq_relative.parts[1])\n",
    "        sequencer = row.get(\"sequencer\")\n",
    "        run_date = row[\"ctime\"]\n",
    "        md5sum = row[\"md5sum\"]\n",
    "\n",
    "        #if row[\"read_id\"] in ('I1',):\n",
    "        #    continue\n",
    "\n",
    "        subpool = get_subpool_from_fastq_row(row)\n",
    "        current_plate = subpool.plate\n",
    "\n",
    "        platform = {\n",
    "            \"A00850\": models.Platform.objects.get(display_name=\"Novaseq 6000\"),\n",
    "            \"VH00582\": models.Platform.objects.get(display_name=\"Nextseq 2000\"),\n",
    "            \"minion\": models.Platform.objects.get(name=\"minion\"),\n",
    "            \"promethion\": models.Platform.objects.get(name=\"promethion\"),\n",
    "        }[sequencer]\n",
    "\n",
    "        subpool_status = {\n",
    "            \"Yes\": models.RunStatus.PASS,\n",
    "            \"No\": models.RunStatus.FAILED,\n",
    "            None: models.RunStatus.FAILED,\n",
    "        }[submittable_subpools.get(subpool.name)]\n",
    "\n",
    "        #sequencing_run = get_or_create_sequencing_run(sequencing_run_name, run_date, platform, current_plate, accessions=None)\n",
    "        #subpool_run = get_or_create_subpool_in_run(subpool=subpool, sequencing_run=sequencing_run, status=subpool_status)\n",
    "        print(row.filename, row.index_id, sequencing_run_name, run_date, subpool, subpool_status)\n",
    "\n",
    "#print_subpool_submission_status(fastqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "regular-interstate",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0, 20)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>experiment</th>\n",
       "      <th>family</th>\n",
       "      <th>filename</th>\n",
       "      <th>md5sum</th>\n",
       "      <th>size</th>\n",
       "      <th>ctime</th>\n",
       "      <th>plate_id</th>\n",
       "      <th>subpool_name</th>\n",
       "      <th>fragment</th>\n",
       "      <th>compression</th>\n",
       "      <th>sequencer</th>\n",
       "      <th>sequencer_run</th>\n",
       "      <th>flowcell_id</th>\n",
       "      <th>nuclei</th>\n",
       "      <th>read</th>\n",
       "      <th>barcode</th>\n",
       "      <th>index_id</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>lane</th>\n",
       "      <th>exome_capture</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [experiment, family, filename, md5sum, size, ctime, plate_id, subpool_name, fragment, compression, sequencer, sequencer_run, flowcell_id, nuclei, read, barcode, index_id, sample_id, lane, exome_capture]\n",
       "Index: []"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "md5counts = Counter(fastqs[\"md5sum\"])\n",
    "dupmd5s = set()\n",
    "for md5sum in md5counts:\n",
    "    if md5counts[md5sum] > 1:\n",
    "        dupmd5s.add(md5sum)\n",
    "        \n",
    "dupfiles = fastqs[fastqs[\"md5sum\"].isin(dupmd5s)].sort_values([\"experiment\", \"md5sum\", \"filename\"])\n",
    "print(dupfiles.shape)\n",
    "dupfiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "blind-package",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "experiment,family,filename,md5sum,size,ctime,plate_id,subpool_name,fragment,compression,sequencer,sequencer_run,flowcell_id,nuclei,read,barcode,index_id,sample_id,lane,exome_capture\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(dupfiles.to_csv(index=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "shaped-effect",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "sufficient-orlando",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('M', 'male'), ('F', 'female'), ('U', 'unknown'), ('O', 'other')]"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "models.SexEnum.choices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cleared-standing",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "marine-november",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "adequate-proposal",
   "metadata": {},
   "outputs": [],
   "source": [
    "# In case of interrupted upload\n",
    "\n",
    "if 0:\n",
    "    EC = str(Path(\"~/proj/encoded_client\").expanduser())\n",
    "    if EC not in sys.path:\n",
    "        sys.path.append(EC)\n",
    "\n",
    "    from encoded_client.encoded import ENCODED\n",
    "\n",
    "    server = ENCODED(\"api.data.igvf.org\")\n",
    "\n",
    "    submitted_tissues = load_submitted_tissues()\n",
    "    for tissue_id in submitted_tissues:\n",
    "        for record in submitted_tissues[tissue_id]:\n",
    "            if pandas.isnull(record[\"uuid\"] ):\n",
    "                tissue = server.get_json(record[\"see_also\"])\n",
    "                print(\"{}\\t{}\".format(record[\"name\"], tissue[\"uuid\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "circular-ranch",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "occupational-opposition",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<SubpoolInRunFile: igvf_b01/next2/B01_13H_R2.fastq.gz>"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "models.SubpoolInRunFile.objects.get(md5sum=\"2a5a07155767a2f541cfb20e7d0636e0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "wound-movie",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
