{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "satisfied-traffic",
   "metadata": {},
   "source": [
    "# Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "nearby-variation",
   "metadata": {},
   "source": [
    "UCI has a complex google doc they've been using to track their samples so far.\n",
    "\n",
    "https://docs.google.com/spreadsheets/d/13M6-Ry6oXgkx94BHZOGioYPI6F_hWDqjGgcaNu2JNYs/edit#gid=1838362486\n",
    "\n",
    "Also there's a metadata spreadsheet for IGVF at\n",
    "https://docs.google.com/spreadsheets/d/1BLMledzmqOqXnJHzpijgw91IOs-9tSlVeZDG_MtXddk/edit#gid=1284120531"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "increased-bottle",
   "metadata": {},
   "outputs": [],
   "source": [
    "import bz2\n",
    "from collections import Counter, namedtuple\n",
    "import datetime\n",
    "import numpy\n",
    "import os\n",
    "import pandas\n",
    "from pathlib import Path\n",
    "import re\n",
    "from subprocess import run, PIPE\n",
    "import sys\n",
    "import zoneinfo\n",
    "\n",
    "os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'mousedemo.settings')\n",
    "os.environ[\"DJANGO_ALLOW_ASYNC_UNSAFE\"] = \"true\"\n",
    "\n",
    "import django\n",
    "from django.contrib.auth import get_user_model\n",
    "from django.db import DEFAULT_DB_ALIAS\n",
    "\n",
    "MOUSEDEMO = str(Path(\"mousedemo\").absolute())\n",
    "if MOUSEDEMO not in sys.path:\n",
    "    sys.path.append(MOUSEDEMO)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "divided-checkout",
   "metadata": {},
   "source": [
    "# Pre-initialization setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "orange-optics",
   "metadata": {},
   "source": [
    "Backup the old ones and make a new database file before we initialize the database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "effective-palace",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a new database\n",
    "result = run([\"python3\", \"manage.py\", \"check\"], capture_output=True)\n",
    "\n",
    "if len(result.stderr) > 0:\n",
    "    print(result.stderr.decode(\"utf-8\"))\n",
    "    \n",
    "result.check_returncode()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "documented-cross",
   "metadata": {},
   "outputs": [],
   "source": [
    "initial = Path(\"igvf_mice/migrations/0001_initial.py\")\n",
    "if initial.exists():\n",
    "    initial.unlink()\n",
    "    \n",
    "result = run([\"python3\", \"manage.py\", \"makemigrations\", \"igvf_mice\"], capture_output=True)\n",
    "\n",
    "if len(result.stderr) > 0:\n",
    "    print(result.stderr.decode(\"utf-8\"))\n",
    "    \n",
    "result.check_returncode()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "magnetic-soccer",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "renaming db3.sqlite3 to db4.sqlite3\n",
      "renaming db2.sqlite3 to db3.sqlite3\n",
      "renaming db1.sqlite3 to db2.sqlite3\n",
      "renaming db.sqlite3 to db1.sqlite3\n"
     ]
    }
   ],
   "source": [
    "# Backup several old sqlite database files\n",
    "\n",
    "db_name = Path(\"db.sqlite3\")\n",
    "\n",
    "backup_names = []\n",
    "for i in range(0, 4):\n",
    "    backup_names.append(Path(\"db{}.sqlite3\".format(i+1)))\n",
    "\n",
    "if backup_names[-1].exists():\n",
    "    backup_names[-1].unlink()\n",
    "\n",
    "if db_name.exists():    \n",
    "    for i in reversed(range(0, 3)):\n",
    "        if backup_names[i].exists():\n",
    "            print(\"renaming {} to {}\".format(backup_names[i], backup_names[i+1]))\n",
    "            backup_names[i].rename(backup_names[i+1])\n",
    "\n",
    "    db_name.rename(backup_names[0])\n",
    "    print(\"renaming {} to {}\".format(db_name, backup_names[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "organizational-resort",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a new database\n",
    "result = run([\"python3\", \"manage.py\", \"migrate\"], capture_output=True)\n",
    "\n",
    "if len(result.stderr) > 0:\n",
    "    print(result.stderr.decode(\"utf-8\"))\n",
    "    \n",
    "result.check_returncode()\n",
    "    \n",
    "assert Path(\"db.sqlite3\").exists()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "blessed-shareware",
   "metadata": {},
   "source": [
    "# Setup users"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "failing-advantage",
   "metadata": {},
   "source": [
    "Now that we have a fresh clean database, lets create user accounts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "aboriginal-reform",
   "metadata": {},
   "outputs": [],
   "source": [
    "django.setup()\n",
    "\n",
    "from mousedemo import settings\n",
    "from igvf_mice import models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "arranged-restaurant",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create accounts.\n",
    "with open(\"pw\", \"rt\") as instream:\n",
    "    pw = instream.read().strip()\n",
    "\n",
    "user_model = get_user_model()\n",
    "user_model._default_manager.db_manager(DEFAULT_DB_ALIAS).create_superuser(\n",
    "    username=\"diane\",\n",
    "    password=pw,   \n",
    ")\n",
    "\n",
    "with open(\"guest.pw\", \"rt\") as instream:\n",
    "    pw = instream.read().strip()\n",
    "\n",
    "user_model = get_user_model()\n",
    "user_model._default_manager.db_manager(DEFAULT_DB_ALIAS).create_user(\n",
    "    username=\"guest\",\n",
    "    password=pw,   \n",
    ")\n",
    "\n",
    "pw = None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "realistic-ocean",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "conservative-recycling",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "canadian-slide",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample Tissue IDs\n",
      "Planning\n",
      "Counting\n",
      "DissectionSorting\n",
      "schedule - all\n",
      "Line information\n",
      "mice\n",
      "Samples - 8 founders\n",
      "Samples - F1s\n",
      "Samples - CClines\n",
      "Samples - Bridge\n",
      "Founder Samples into experiment\n",
      "F1 Samples into experiment\n",
      "Bridge samples into experiment\n",
      "PBMCs\n",
      "Founder Nuclei isolation Sample\n",
      "F1 Nuclei isolation SampleID\n",
      "Plate setups\n",
      "Experiment\n",
      "ONT Sequencing\n",
      "DCC\n",
      "Tissue_body_weight_metadata\n",
      "Failed_samples\n",
      "Sheet40\n",
      "tissue weights\n",
      "practice mice\n",
      "CellDrop_Demo\n",
      "Gastroc_numbers\n",
      "Gastroc_all_reps\n",
      "Recount for gene capture\n",
      "Gastroc_testing\n",
      "Cerebellum testing\n",
      "IGVF002 Pilot Samples\n",
      "IGVF002 Pulverized Samples\n",
      "IGVF002 Plate Setup\n",
      "IGVF002 Metadata\n"
     ]
    }
   ],
   "source": [
    "#spreadsheet_name = \"IGVF_Split-seq_20230118.xlsx\"\n",
    "lizs_sheet_name = \"https://woldlab.caltech.edu/nextcloud/index.php/s/eEtjBfDqQFnLpSS/download\"\n",
    "\n",
    "book = pandas.ExcelFile(lizs_sheet_name)\n",
    "for name in book.sheet_names:\n",
    "    print(name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "golden-crack",
   "metadata": {},
   "outputs": [],
   "source": [
    "submitted_book_names = {\n",
    "    \"igvftst\": {\n",
    "        \"IGVF_003\": \"https://woldlab.caltech.edu/nextcloud/index.php/s/nBnE6j2RBMkpM9i/download\",\n",
    "        \"IGVF_b01\": \"https://woldlab.caltech.edu/nextcloud/index.php/s/5cJteSWgitN5BDM/download\",\n",
    "    },\n",
    "    \"igvf\": {\n",
    "        \"IGVF_b01\": \"https://woldlab.caltech.edu/nextcloud/index.php/s/HTbfN6btm3kqJXb/download\",        \n",
    "    }\n",
    "}\n",
    "submitted_sheets = {}\n",
    "for accession_prefix in submitted_book_names:\n",
    "    for book_name in submitted_book_names[accession_prefix]:\n",
    "        book = pandas.ExcelFile(submitted_book_names[accession_prefix][book_name])\n",
    "\n",
    "        for sheet_name in book.sheet_names:\n",
    "            if sheet_name not in submitted_sheets:\n",
    "                submitted_sheets.setdefault(accession_prefix, {})[sheet_name] = book.parse(sheet_name)\n",
    "            else:\n",
    "                submitted_sheets.setdefault(accession_prefix, {})[sheet_name] = pandas.concat([submitted_sheets[sheet_name], book.parse(sheet_name)])\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "split-payday",
   "metadata": {},
   "source": [
    "Converters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "occupational-dealing",
   "metadata": {},
   "outputs": [],
   "source": [
    "def truncate(model):\n",
    "    table_name = model._meta.db_table\n",
    "    assert \"\\\\\" not in table_name\n",
    "    with django.db.connection.cursor() as cursor:\n",
    "        # I don't know why the sql params didn't work. this is a sql vulnerability waiting to happen\n",
    "        cursor.execute(\"delete from \\\"{}\\\"\".format(table_name))\n",
    "        cursor.execute(\"DELETE FROM SQLITE_SEQUENCE WHERE name=\\\"{}\\\"\".format(table_name))\n",
    "        cursor.fetchone()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "divine-incident",
   "metadata": {},
   "outputs": [],
   "source": [
    "def int_or_none(x):\n",
    "    if x in (\"N/A\", '#DIV/0!', '-'):\n",
    "        return None\n",
    "    elif pandas.isnull(x):\n",
    "        return None\n",
    "    else:\n",
    "        return int(x)\n",
    "    \n",
    "def int_or_0(x):\n",
    "    if pandas.isnull(x):\n",
    "        return 0\n",
    "    else:\n",
    "        return int(x)\n",
    "\n",
    "def float_or_none(x):\n",
    "    if pandas.isnull(x):\n",
    "        return None\n",
    "    else:\n",
    "        return float(x)\n",
    "    \n",
    "def str_or_empty(x):\n",
    "    if pandas.isnull(x):\n",
    "        return \"\"\n",
    "    else:\n",
    "        return x\n",
    "\n",
    "def str_or_none(x):\n",
    "    if pandas.isnull(x):\n",
    "        return None\n",
    "    else:\n",
    "        return x\n",
    "    \n",
    "def date_or_none(x):\n",
    "    if pandas.isnull(x):\n",
    "        return None\n",
    "    elif isinstance(x, datetime.datetime):\n",
    "        return x.date()\n",
    "    elif x in (\"-\"):\n",
    "        return None\n",
    "    else:\n",
    "        return x\n",
    "\n",
    "def datetime_or_none(x):\n",
    "    if pandas.isnull(x):\n",
    "        return None\n",
    "    else:\n",
    "        return x\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "noble-personal",
   "metadata": {},
   "source": [
    "# initialize general use tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "linear-civilian",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Delete later: switch to just using an enum for the accessions.\n",
    "#\n",
    "#accession_namespaces = [\n",
    "#    {\"name\": \"ENCODE\", \"homepage\": \"https://www.encodeproject.org\", \"accession_prefix\": \"encode\"},\n",
    "#    {\"name\": \"ENCODE test\", \"homepage\": \"https://test.encodedcc.org\", \"accession_prefix\": \"encodetst\"},\n",
    "#    {\"name\": \"IGVF\", \"homepage\": \"https://data.igvf.org\", \"accession_prefix\": \"igvf\"},\n",
    "#    {\"name\": \"IGVF test\", \"homepage\": \"https://sandbox.igvf.org\", \"accession_prefix\": \"igvftst\"},\n",
    "#]\n",
    "#\n",
    "#if models.AccessionNamespace.objects.count() > 0:\n",
    "#    truncate(models.AccessionNamespace)\n",
    "#\n",
    "#for row in accession_namespaces:\n",
    "#    record = models.AccessionNamespace(\n",
    "#        name=row[\"name\"],\n",
    "#        homepage=row[\"homepage\"],\n",
    "#        accession_prefix=row[\"accession_prefix\"],\n",
    "#    )\n",
    "#    record.save()\n",
    "#\n",
    "#igvf_namespace = models.AccessionNamespace.objects.get(name=\"IGVF\")\n",
    "#igvf_test_namespace = models.AccessionNamespace.objects.get(name=\"IGVF test\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "psychological-march",
   "metadata": {},
   "outputs": [],
   "source": [
    "source = [\n",
    "    {\n",
    "        \"name\": \"jackson-labs\",\n",
    "        \"display_name\": \"The Jackson Laboratory\", \n",
    "        \"homepage\": \"http://www.jax.org/index.html\", \n",
    "        \"igvf_id\": \"/sources/jackson-labs/\",\n",
    "    },\n",
    "    {\n",
    "        \"name\": \"parse-biosciences\",\n",
    "        \"display_name\": \"Parse Bioscience\", \n",
    "        \"homepage\": \"https://www.parsebiosciences.com/\",\n",
    "        \"igvf_id\": \"/sources/parse-biosciences/\",\n",
    "    },\n",
    "    {\n",
    "        \"name\": \"illumina\",\n",
    "        \"display_name\": \"Illumina\", \n",
    "        \"homepage\": \"https://www.illumina.com\", \n",
    "        \"igvf_id\": \"/sources/illumina/\",\n",
    "    },\n",
    "]\n",
    "\n",
    "if models.Source.objects.count() > 0:\n",
    "    truncate(models.Source)\n",
    "\n",
    "for row in source:\n",
    "    record = models.Source(\n",
    "        name=row[\"name\"],\n",
    "        homepage=row[\"homepage\"],\n",
    "        igvf_id=row[\"igvf_id\"])\n",
    "    record.save()\n",
    "\n",
    "jax_source = models.Source.objects.get(name=\"jackson-labs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "metallic-spider",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Should this include machine model?\n",
    "illumina_source = models.Source.objects.get(name=\"illumina\")\n",
    "platforms = [\n",
    "    {\"display_name\": \"Nextseq 2000\", \"name\": \"nextseq2000\", \"family\": \"illumina\", \"igvf_id\": \"/platform-terms/EFO_0010963/\", \"source\": illumina_source},\n",
    "    {\"display_name\": \"Novaseq 6000\", \"name\": \"novaseq6000\", \"family\": \"illumina\", \"igvf_id\": \"/platform-terms/EFO_0008637/\", \"source\": illumina_source},\n",
    "    {\"display_name\": \"Oxford Nanopore\", \"name\": \"nanopore\", \"family\": \"nanopore\",},\n",
    "    {\"display_name\": \"Pac Bio\", \"name\": \"pacbio\", \"family\": \"pacbio\",},\n",
    "]\n",
    "\n",
    "for row in platforms:\n",
    "    record = models.Platform(\n",
    "        name=row[\"name\"],\n",
    "        display_name=row[\"display_name\"],\n",
    "        family=row[\"family\"],\n",
    "        igvf_id=row.get(\"igvf_id\"),\n",
    "        source=row.get(\"source\"),\n",
    "    )\n",
    "    record.save()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "manufactured-account",
   "metadata": {},
   "source": [
    "# Library kit type & version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "innocent-copying",
   "metadata": {},
   "outputs": [],
   "source": [
    "library_construction_kit = [\n",
    "    {\"name\": \"wt-mega\", \"display_name\": \"Parse WT Mega\", \"version\": \"v2\", \"source\": \"parse-biosciences\"},\n",
    "    {\"name\": \"wt\", \"display_name\": \"Parse WT\", \"version\": \"v2\", \"source\": \"parse-biosciences\"},\n",
    "]\n",
    "\n",
    "if models.LibraryConstructionKit.objects.count() > 0:\n",
    "    truncate(models.LibraryConstructionKit)\n",
    "\n",
    "for row in library_construction_kit:\n",
    "    try:\n",
    "        source = models.Source.objects.get(name=row[\"source\"])\n",
    "        \n",
    "        record = models.LibraryConstructionKit(\n",
    "            name=row[\"name\"],\n",
    "            display_name=row[\"display_name\"],\n",
    "            version=row[\"version\"],\n",
    "            source=source,\n",
    "        )\n",
    "        record.save()        \n",
    "    except models.Source.DoesNotExist:\n",
    "        print(f\"Couldn't find {row['source']}\")\n",
    "        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "pressed-payday",
   "metadata": {},
   "source": [
    "# Load LibraryBarcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "convenient-syndication",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<LibraryConstructionKit: Parse WT Mega v2>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "models.LibraryConstructionKit.objects.get(name=\"wt-mega\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "approved-hours",
   "metadata": {},
   "outputs": [],
   "source": [
    "if models.LibraryBarcode.objects.count() > 0:\n",
    "    truncate(models.LibraryBarcode)\n",
    "\n",
    "\n",
    "# Load WT Mega v2 first barcode\n",
    "bc1 = pandas.read_csv(\"bc_data_n192_v4.csv\")\n",
    "\n",
    "mega_kit = models.LibraryConstructionKit.objects.get(name=\"wt-mega\", version=\"v2\")\n",
    "\n",
    "for i, row in bc1.iterrows():\n",
    "    record = models.LibraryBarcode(\n",
    "        kit=mega_kit,\n",
    "        name=row[\"uid\"],\n",
    "        code=row[\"well\"],\n",
    "        sequence=row[\"sequence\"],\n",
    "        barcode_type=row[\"type\"], # TODO what do the codes mean again T,R?\n",
    "    )\n",
    "    record.save()\n",
    "\n",
    "mega_subpool = [\n",
    "    (\"1\", \"CAGATC\"),\n",
    "    (\"2\", \"ACTTGA\"),\n",
    "    (\"3\", \"GATCAG\"),\n",
    "    (\"4\", \"TAGCTT\"),\n",
    "    (\"5\", \"ATGTCA\"),\n",
    "    (\"6\", \"CTTGTA\"),\n",
    "    (\"7\", \"AGTCAA\"),\n",
    "    (\"8\", \"AGTTCC\"),\n",
    "    (\"9\", \"GAGTGG\"),\n",
    "    (\"10\", \"CCGTCC\"),\n",
    "    (\"11\", \"GTAGAG\"),\n",
    "    (\"12\", \"GTCCGC\"),\n",
    "    (\"13\", \"GTGAAA\"),\n",
    "    (\"14\", \"GTGGCC\"),\n",
    "    (\"15\", \"GTTTCG\"),\n",
    "    (\"16\", \"CGTACG\"),\n",
    "]\n",
    "subpool = pandas.DataFrame(mega_subpool, columns=[\"code\",\"sequence\"])\n",
    "\n",
    "for i, row in subpool.iterrows():\n",
    "    record = models.LibraryBarcode(\n",
    "        kit=mega_kit,\n",
    "        code=row[\"code\"],\n",
    "        sequence=row[\"sequence\"],\n",
    "    )\n",
    "    record.save()\n",
    "\n",
    "regular_kit = models.LibraryConstructionKit.objects.get(name=\"wt\", version=\"v2\")    \n",
    "    \n",
    "regular_subpool = [\n",
    "    (\"1\", \"CAGATC\"),\n",
    "    (\"2\", \"ACTTGA\"),\n",
    "    (\"3\", \"GATCAG\"),\n",
    "    (\"4\", \"TAGCTT\"),\n",
    "    (\"5\", \"ATGTCA\"),\n",
    "    (\"6\", \"CTTGTA\"),\n",
    "    (\"7\", \"AGTCAA\"),\n",
    "    (\"8\", \"AGTTCC\"),\n",
    "]\n",
    "subpool = pandas.DataFrame(regular_subpool, columns=[\"code\",\"sequence\"])\n",
    "\n",
    "for i, row in subpool.iterrows():\n",
    "    record = models.LibraryBarcode(\n",
    "        kit=regular_kit,\n",
    "        code=row[\"code\"],\n",
    "        sequence=row[\"sequence\"],\n",
    "    )\n",
    "    record.save()\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "hungry-guidance",
   "metadata": {},
   "source": [
    "# Initialize TissueOntology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "dimensional-paintball",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Samples - 8 founders (1621, 1) 22\n",
      "Samples - Bridge (416, 1) 44\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Counter({'Hypothalamus/Pituitary': 97,\n",
       "         'Cerebellum': 81,\n",
       "         'Cortex/Hippocampus left': 85,\n",
       "         'Cortex/Hippocampus right': 85,\n",
       "         'Liver': 97,\n",
       "         'Heart': 97,\n",
       "         'Lung': 81,\n",
       "         'Adrenal': 97,\n",
       "         'Kidney': 81,\n",
       "         'Gonads -1 Ovary': 41,\n",
       "         'Gonads -2 Oviduct': 41,\n",
       "         'Perigonadal fat': 81,\n",
       "         'Brown fat': 81,\n",
       "         'Soleus': 81,\n",
       "         'Plantaris': 81,\n",
       "         'Gastrocnemius': 82,\n",
       "         'TA': 81,\n",
       "         'EDL': 81,\n",
       "         'Tail': 121,\n",
       "         'PBMC - WBC': 121,\n",
       "         'Gonads -1 Testis': 40,\n",
       "         'Gonads -2 Epididymis': 40,\n",
       "         'Left cortex': 24,\n",
       "         'Right cortex': 24,\n",
       "         'Left hippocampus': 24,\n",
       "         'Right hippocampus': 24,\n",
       "         'Testis left': 8,\n",
       "         'Testis right': 8,\n",
       "         'Epididymis left': 8,\n",
       "         'Epididymis right': 8,\n",
       "         'Kidney left': 16,\n",
       "         'Kidney right': 16,\n",
       "         'Gastrocnemius left': 16,\n",
       "         'Gastrocnemius right': 16,\n",
       "         'TA left': 4,\n",
       "         'TA right': 4,\n",
       "         'Ovary': 1,\n",
       "         'Oviduct': 1,\n",
       "         'Gonads -1 (Ovary)': 7,\n",
       "         'Gonads - 2 (Oviduct)': 7,\n",
       "         'Cortex left': 12,\n",
       "         'Cortex right': 12,\n",
       "         'Hippocampus left': 12,\n",
       "         'Hippocampus right': 12})"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seen_tissues = Counter()\n",
    "for sample_sheet_name in [\"Samples - 8 founders\", \"Samples - Bridge\"]:\n",
    "    tissue_sheet = pandas.read_excel(\n",
    "        lizs_sheet_name, \n",
    "        sample_sheet_name,\n",
    "        usecols=[\"Tissue\"]).dropna(axis=0, how='all')\n",
    "\n",
    "    for i, row in tissue_sheet.iterrows():\n",
    "        if not pandas.isnull(row[\"Tissue\"]):\n",
    "            seen_tissues[row[\"Tissue\"]] += 1\n",
    "    print(sample_sheet_name, tissue_sheet.shape, len(seen_tissues))\n",
    "\n",
    "seen_tissues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "responsible-aberdeen",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "set()\n"
     ]
    }
   ],
   "source": [
    "tissue_dissection_to_ontology_map = {\n",
    "    'Hypothalamus/Pituitary': [(\"UBERON:0001898\",\"hypothalamus\"), (\"UBERON:0000007\",\"pituitary gland\")],\n",
    "    'Cerebellum': [(\"UBERON:0002037\",\"cerebellum\")],\n",
    "    # Hippocampus might be:\n",
    "    #   Hippocampal formation UBERON:0002421 https://www.ebi.ac.uk/ols/ontologies/uberon/terms?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FUBERON_0002421\n",
    "    #   Layer of hippocampus UBERON:0002305 https://www.ebi.ac.uk/ols/ontologies/uberon/terms?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FUBERON_0002305\n",
    "    'Cortex/Hippocampus left': [(\"NTR:0000646\",\"left cerebral cortex\"), (\"NTR:0000750\", \"Hippocampal formation left\")],\n",
    "    'Cortex/Hippocampus right': [(\"NTR:0000647\",\"right cerebral cortex\"), (\"NTR:0000751\", \"Hippocampal formation right\")],\n",
    "    'Liver': [(\"UBERON:0002107\", \"liver\")],\n",
    "    'Heart': [(\"UBERON:0000948\",\"heart\")],\n",
    "    'Lung': [(\"UBERON:0002048\",\"lung\")],\n",
    "    'Adrenal': [(\"UBERON:0002369\",\"adrenal gland\")],\n",
    "    'Kidney': [(\"UBERON:0002113\",\"kidney\")],\n",
    "    'Kidney left': [(\"UBERON:0002113\",\"kidney\")],\n",
    "    'Kidney right': [(\"UBERON:0002113\",\"kidney\")],\n",
    "    \"Ovary\": [(\"UBERON:0000992\", \"ovary\")],    \n",
    "    'Gonads -1 Ovary': [(\"UBERON:0000992\", \"ovary\")],\n",
    "    'Gonads -1 (Ovary)': [(\"UBERON:0000992\", \"ovary\")],\n",
    "    'Oviduct': [(\"UBERON:0000993\",\"oviduct\")],\n",
    "    'Gonads -2 Oviduct': [(\"UBERON:0000993\",\"oviduct\")],\n",
    "    'Gonads -2 (Oviduct)': [(\"UBERON:0000993\",\"oviduct\")],\n",
    "    'Gonads - 2 (Oviduct)': [(\"UBERON:0000993\",\"oviduct\")],\n",
    "    'Perigonadal fat': [(\"UBERON:0003428\", \"gonadal fat pad\")], #\"is closer\" to dissection\n",
    "    'Brown fat': [(\"UBERON:0001348\",\"brown adipose tissue\")], \n",
    "    'Soleus': [(\"UBERON:0001389\",\"soleus muscle\")],\n",
    "    'Plantaris': [(\"UBERON:0011905\",\"plantaris\")],\n",
    "    'Gastrocnemius': [(\"UBERON:0001388\",\"gastrocnemius\")],\n",
    "    'Gastrocnemius left': [(\"UBERON:0001388\",\"gastrocnemius\")],\n",
    "    'Gastrocnemius right': [(\"UBERON:0001388\",\"gastrocnemius\")],\n",
    "    'TA': [(\"UBERON:0001385\",\"tibialis anterior\")],\n",
    "    'TA left': [(\"UBERON:0001385\",\"tibialis anterior\")],\n",
    "    'TA right': [(\"UBERON:0001385\",\"tibialis anterior\")],\n",
    "    'EDL': [(\"UBERON:0001386\",\"extensor digitorum longus\")],\n",
    "    'Tail': [(\"UBERON:0002415\",\"tail\")],\n",
    "    'PBMC - WBC': [(\"CL:2000001\", \"peripheral blood mononuclear cell\")],\n",
    "    'Gonads -1 Testis': [(\"UBERON:0000473\",\"testis\")],\n",
    "    'Gonads -2 Epididymis': [(\"UBERON:0001301\",\"epididymis\")],\n",
    "    \"Left cortex\": [(\"NTR:0000646\",\"left cerebral cortex\"),],\n",
    "    \"Cortex left\": [(\"NTR:0000646\",\"left cerebral cortex\"),],\n",
    "    \"Right cortex\": [(\"NTR:0000647\",\"right cerebral cortex\"),],\n",
    "    \"Cortex right\": [(\"NTR:0000647\",\"right cerebral cortex\"),],\n",
    "    \"Left hippocampus\": [(\"NTR:0000750\", \"Hippocampal formation left\")],\n",
    "    \"Hippocampus left\": [(\"NTR:0000750\", \"Hippocampal formation left\")],\n",
    "    \"Right hippocampus\": [(\"NTR:0000751\", \"Hippocampal formation right\")],\n",
    "    \"Hippocampus right\": [(\"NTR:0000751\", \"Hippocampal formation right\")],\n",
    "    # does left & right matter?\n",
    "    \"Testis left\": [(\"UBERON:0000473\", \"Testis\")],\n",
    "    \"Testis right\": [(\"UBERON:0000473\", \"Testis\")],\n",
    "    \"Epididymis left\": [(\"UBERON:0001301\", \"Epididymis\")],\n",
    "    \"Epididymis right\": [(\"UBERON:0001301\", \"Epididymis\")],\n",
    "}\n",
    "\n",
    "print(set(seen_tissues).difference(tissue_dissection_to_ontology_map.keys()))\n",
    "\n",
    "assert len(set(seen_tissues).difference(tissue_dissection_to_ontology_map.keys())) == 0, \"Add in more names to term ids\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "green-orange",
   "metadata": {},
   "outputs": [],
   "source": [
    "term_details = pandas.read_csv(\"obo.tsv.bz2\", compression=\"bz2\", sep=\"\\t\", index_col=\"term_id\")\n",
    "\n",
    "if models.OntologyTerm.objects.count() > 0:\n",
    "    truncate(models.OntologyTerm)\n",
    "    \n",
    "for key in tissue_dissection_to_ontology_map:\n",
    "    for term_curie, term_name in tissue_dissection_to_ontology_map[key]:\n",
    "        if term_curie.startswith(\"NTR:\"):\n",
    "            description=None\n",
    "        else:\n",
    "            details = term_details.loc[term_curie]\n",
    "            description = details.description\n",
    "\n",
    "        record = models.OntologyTerm(\n",
    "            curie=term_curie,\n",
    "            name=term_name,\n",
    "            description=description\n",
    "        )\n",
    "        record.save()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "otherwise-runner",
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_urls = {\n",
    "    \"AJ\": \"https://www.jax.org/strain/000646\",\n",
    "    \"B6J\": \"https://www.jax.org/strain/000664\",\n",
    "    \"129S1J\": \"https://www.jax.org/strain/002448\",\n",
    "    \"NODJ\": \"https://www.jax.org/strain/001976\",\n",
    "    \"NZOJ\": \"https://www.jax.org/strain/002105\",\n",
    "    \"CASTJ\": \"https://www.jax.org/strain/000928\",\n",
    "    \"PWKJ\": \"https://www.jax.org/strain/003715\",\n",
    "    \"WSBJ\": \"https://www.jax.org/strain/001145\",\n",
    "    'B6129SF1J': \"https://www.jax.org/strain/101043\",\n",
    "    'B6AF1J': \"https://www.jax.org/strain/100002\",\n",
    "    'B6CASTF1J': None,\n",
    "    'B6NODF1J': None,\n",
    "    'B6NZOF1J': None,\n",
    "    'B6PWKF1J': None,\n",
    "    'B6WSBF1J': \"https://www.jax.org/strain/019019\",\n",
    "    'TREM2R47HNSS_HO': 'https://www.jax.org/strain/033781',\n",
    "    'CC001': \"https://www.jax.org/strain/021238\",\n",
    "    'CC002': \"https://www.jax.org/strain/021236\",\n",
    "    'CC003': \"https://www.jax.org/strain/021237\",\n",
    "    'CC004': \"https://www.jax.org/strain/020944\",\n",
    "    'CC005': \"https://www.jax.org/strain/020945\",\n",
    "    'CC006': \"https://www.jax.org/strain/022869\",\n",
    "    'CC007': \"https://www.jax.org/strain/029625\",\n",
    "    'CC008': \"https://www.jax.org/strain/026971\",\n",
    "    'CC009': \"https://www.jax.org/strain/018856\",\n",
    "    'CC010': \"https://www.jax.org/strain/021889\",\n",
    "    'CC011': \"https://www.jax.org/strain/018854\",\n",
    "    'CC012': \"https://www.jax.org/strain/028409\",\n",
    "    'CC013': \"https://www.jax.org/strain/021892\",\n",
    "    'CC015': \"https://www.jax.org/strain/018859\",\n",
    "    'CC017': \"https://www.jax.org/strain/022870\",\n",
    "    'CC018': \"https://www.jax.org/strain/021890\",\n",
    "    'CC024': \"https://www.jax.org/strain/021891\",\n",
    "    'CC025': \"https://www.jax.org/strain/018857\",\n",
    "    'CC028': \"https://www.jax.org/strain/025126\",\n",
    "    'CC029': \"https://www.jax.org/strain/026972\",\n",
    "    'CC030': \"https://www.jax.org/strain/025426\",\n",
    "    'CC032': \"https://www.jax.org/strain/020946\",\n",
    "    'CC036': \"https://www.jax.org/strain/025127\",\n",
    "    'CC037': \"https://www.jax.org/strain/025423\",\n",
    "    'CC038': None,\n",
    "    'CC041': \"https://www.jax.org/strain/021893\",\n",
    "    'CC055': None,\n",
    "    'CC057': \"https://www.jax.org/strain/024683\",\n",
    "    'CC060': \"https://www.jax.org/strain/026427\",\n",
    "    'CC062': None,\n",
    "    'CC065': None,\n",
    "    'CC071': None,\n",
    "    'CC074': \"https://www.jax.org/strain/018855\",\n",
    "}\n",
    "\n",
    "# used as name. Should this be \n",
    "strain_igvf_id = {\n",
    "    \"AJ\": \"A/J (AJ)\",\n",
    "    \"B6J\": \"C57BL/6J (B6)\",\n",
    "    \"129S1J\": \"129S1/SvImJ (129)\",\n",
    "    \"NODJ\": \"NOD/ShiLtJ (NOD)\",\n",
    "    \"NZOJ\": \"NZO/HlLtJ (NZO)\",\n",
    "    \"CASTJ\": \"CAST/EiJ (CAST)\",\n",
    "    \"PWKJ\": \"PWK/PhJ (PWK)\",\n",
    "    \"WSBJ\": \"WSB/EiJ (WSB)\",\n",
    "    'B6129SF1J': \"B6129SF1/J\",\n",
    "    'B6AF1J': \"B6AF1/J\",\n",
    "    'B6CASTF1J': \"B6CASTF1/J\",\n",
    "    'B6NODF1J': \"B6NODF1/J\",\n",
    "    'B6NZOF1J': \"B6NZOF1/J\",\n",
    "    'B6PWKF1J': \"B6PWKF1/J\",\n",
    "    'B6WSBF1J': \"B6WSBF1/J\",\n",
    "    \"TREM2R47HNSS_HO\": \"TREM2R47HNSS_HO\",\n",
    "    'CC001': \"CC001\",\n",
    "    'CC002': \"CC002\",\n",
    "    'CC003': \"CC003\",\n",
    "    'CC004': \"CC004\",\n",
    "    'CC005': \"CC005\",\n",
    "    'CC006': \"CC006\",\n",
    "    'CC007': \"CC007\",\n",
    "    'CC008': \"CC008\",\n",
    "    'CC009': \"CC009\",\n",
    "    'CC010': \"CC010\",\n",
    "    'CC011': \"CC011\",\n",
    "    'CC012': \"CC012\",\n",
    "    'CC013': \"CC013\",\n",
    "    'CC015': \"CC015\",\n",
    "    'CC017': \"CC017\",\n",
    "    'CC018': \"CC018\",\n",
    "    'CC024': \"CC024\",\n",
    "    'CC025': \"CC025\",\n",
    "    'CC028': \"CC028\",\n",
    "    'CC029': \"CC029\",\n",
    "    'CC030': \"CC030\",\n",
    "    'CC032': \"CC032\",\n",
    "    'CC036': \"CC036\",\n",
    "    'CC037': \"CC037\",\n",
    "    'CC038': \"CC038\",\n",
    "    'CC041': \"CC041\",\n",
    "    'CC055': \"CC055\",\n",
    "    'CC057': \"CC057\",\n",
    "    'CC060': \"CC060\",\n",
    "    'CC062': \"CC062\",\n",
    "    'CC065': \"CC065\",\n",
    "    'CC071': \"CC071\",\n",
    "    'CC074': \"CC074\",\n",
    "}\n",
    "strain_name = {k: strain_igvf_id[k].split(\" \")[0] for k in strain_igvf_id}\n",
    "strain_name_to_code = {strain_name[k]: k for k in strain_name}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "committed-dollar",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Designation</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Note</th>\n",
       "      <th>Jax Catalog No</th>\n",
       "      <th>Sample CODE</th>\n",
       "      <th>Strain notes</th>\n",
       "      <th>see_also</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>A/J</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>000646</td>\n",
       "      <td>AJ</td>\n",
       "      <td>Yellower adrenal gland.</td>\n",
       "      <td>https://www.jax.org/strain/000646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>B</td>\n",
       "      <td>C57BL/6J</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>000664</td>\n",
       "      <td>B6J</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/000664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>129S1/SvImJ</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>002448</td>\n",
       "      <td>129S1J</td>\n",
       "      <td>Male skin is tougher than female skin. Gallbla...</td>\n",
       "      <td>https://www.jax.org/strain/002448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D</td>\n",
       "      <td>NOD/ShiLtJ</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>001976</td>\n",
       "      <td>NODJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/001976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>NZO/HlLtJ</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>002105</td>\n",
       "      <td>NZOJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/002105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>F</td>\n",
       "      <td>CAST/EiJ</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>000928</td>\n",
       "      <td>CASTJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/000928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>G</td>\n",
       "      <td>PWK/PhJ</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>003715</td>\n",
       "      <td>PWKJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/003715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>H</td>\n",
       "      <td>WSB/EiJ</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>001145</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/001145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NaN</td>\n",
       "      <td>B6129SF1/J</td>\n",
       "      <td>CC F1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B6129SF1J</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/101043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NaN</td>\n",
       "      <td>B6AF1/J</td>\n",
       "      <td>CC F1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B6AF1J</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/100002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NaN</td>\n",
       "      <td>B6CASTF1/J</td>\n",
       "      <td>CC F1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B6CASTF1J</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NaN</td>\n",
       "      <td>B6NODF1/J</td>\n",
       "      <td>CC F1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B6NODF1J</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NaN</td>\n",
       "      <td>B6NZOF1/J</td>\n",
       "      <td>CC F1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B6NZOF1J</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NaN</td>\n",
       "      <td>B6PWKF1/J</td>\n",
       "      <td>CC F1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B6PWKF1J</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NaN</td>\n",
       "      <td>B6WSBF1/J</td>\n",
       "      <td>CC F1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B6WSBF1J</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/019019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TREM2R47HNSS_HO</td>\n",
       "      <td>CC Mutant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TREM2R47HNSS_HO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/033781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC001</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/021238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC002</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/021236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC003</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/021237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC004</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/020944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC005</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/020945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC006</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/022869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC007</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/029625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC008</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/026971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC009</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC009</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/018856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC010</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC010</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/021889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC011</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC011</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/018854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC012</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC012</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/028409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC013</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC013</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/021892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC015</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC015</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/018859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC017</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC017</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/022870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC018</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/021890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC024</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC024</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/021891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC025</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC025</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/018857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC028</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC028</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/025126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC029</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC029</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/026972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC030</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC030</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/025426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC032</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC032</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/020946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC036</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC036</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/025127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC037</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC037</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/025423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC038</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC038</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC041</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC041</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/021893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC055</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC055</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC057</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC057</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/024683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC060</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC060</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/026427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC062</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC062</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC065</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC065</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC071</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC071</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CC074</td>\n",
       "      <td>CC Cross</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC074</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.jax.org/strain/018855</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Designation           Strain        Note Jax Catalog No      Sample CODE  \\\n",
       "0            A              A/J  CC founder         000646               AJ   \n",
       "1            B         C57BL/6J  CC founder         000664              B6J   \n",
       "2            C      129S1/SvImJ  CC founder         002448           129S1J   \n",
       "3            D       NOD/ShiLtJ  CC founder         001976             NODJ   \n",
       "4            E        NZO/HlLtJ  CC founder         002105             NZOJ   \n",
       "5            F         CAST/EiJ  CC founder         000928            CASTJ   \n",
       "6            G          PWK/PhJ  CC founder         003715             PWKJ   \n",
       "7            H          WSB/EiJ  CC founder         001145             WSBJ   \n",
       "8          NaN       B6129SF1/J       CC F1            NaN        B6129SF1J   \n",
       "9          NaN          B6AF1/J       CC F1            NaN           B6AF1J   \n",
       "10         NaN       B6CASTF1/J       CC F1            NaN        B6CASTF1J   \n",
       "11         NaN        B6NODF1/J       CC F1            NaN         B6NODF1J   \n",
       "12         NaN        B6NZOF1/J       CC F1            NaN         B6NZOF1J   \n",
       "13         NaN        B6PWKF1/J       CC F1            NaN         B6PWKF1J   \n",
       "14         NaN        B6WSBF1/J       CC F1            NaN         B6WSBF1J   \n",
       "15         NaN  TREM2R47HNSS_HO   CC Mutant            NaN  TREM2R47HNSS_HO   \n",
       "16         NaN            CC001    CC Cross            NaN            CC001   \n",
       "17         NaN            CC002    CC Cross            NaN            CC002   \n",
       "18         NaN            CC003    CC Cross            NaN            CC003   \n",
       "19         NaN            CC004    CC Cross            NaN            CC004   \n",
       "20         NaN            CC005    CC Cross            NaN            CC005   \n",
       "21         NaN            CC006    CC Cross            NaN            CC006   \n",
       "22         NaN            CC007    CC Cross            NaN            CC007   \n",
       "23         NaN            CC008    CC Cross            NaN            CC008   \n",
       "24         NaN            CC009    CC Cross            NaN            CC009   \n",
       "25         NaN            CC010    CC Cross            NaN            CC010   \n",
       "26         NaN            CC011    CC Cross            NaN            CC011   \n",
       "27         NaN            CC012    CC Cross            NaN            CC012   \n",
       "28         NaN            CC013    CC Cross            NaN            CC013   \n",
       "29         NaN            CC015    CC Cross            NaN            CC015   \n",
       "30         NaN            CC017    CC Cross            NaN            CC017   \n",
       "31         NaN            CC018    CC Cross            NaN            CC018   \n",
       "32         NaN            CC024    CC Cross            NaN            CC024   \n",
       "33         NaN            CC025    CC Cross            NaN            CC025   \n",
       "34         NaN            CC028    CC Cross            NaN            CC028   \n",
       "35         NaN            CC029    CC Cross            NaN            CC029   \n",
       "36         NaN            CC030    CC Cross            NaN            CC030   \n",
       "37         NaN            CC032    CC Cross            NaN            CC032   \n",
       "38         NaN            CC036    CC Cross            NaN            CC036   \n",
       "39         NaN            CC037    CC Cross            NaN            CC037   \n",
       "40         NaN            CC038    CC Cross            NaN            CC038   \n",
       "41         NaN            CC041    CC Cross            NaN            CC041   \n",
       "42         NaN            CC055    CC Cross            NaN            CC055   \n",
       "43         NaN            CC057    CC Cross            NaN            CC057   \n",
       "44         NaN            CC060    CC Cross            NaN            CC060   \n",
       "45         NaN            CC062    CC Cross            NaN            CC062   \n",
       "46         NaN            CC065    CC Cross            NaN            CC065   \n",
       "47         NaN            CC071    CC Cross            NaN            CC071   \n",
       "48         NaN            CC074    CC Cross            NaN            CC074   \n",
       "\n",
       "                                         Strain notes  \\\n",
       "0                             Yellower adrenal gland.   \n",
       "1                                                 NaN   \n",
       "2   Male skin is tougher than female skin. Gallbla...   \n",
       "3                                                 NaN   \n",
       "4                                                 NaN   \n",
       "5                                                 NaN   \n",
       "6                                                 NaN   \n",
       "7                                                 NaN   \n",
       "8                                                 NaN   \n",
       "9                                                 NaN   \n",
       "10                                                NaN   \n",
       "11                                                NaN   \n",
       "12                                                NaN   \n",
       "13                                                NaN   \n",
       "14                                                NaN   \n",
       "15                                                NaN   \n",
       "16                                                NaN   \n",
       "17                                                NaN   \n",
       "18                                                NaN   \n",
       "19                                                NaN   \n",
       "20                                                NaN   \n",
       "21                                                NaN   \n",
       "22                                                NaN   \n",
       "23                                                NaN   \n",
       "24                                                NaN   \n",
       "25                                                NaN   \n",
       "26                                                NaN   \n",
       "27                                                NaN   \n",
       "28                                                NaN   \n",
       "29                                                NaN   \n",
       "30                                                NaN   \n",
       "31                                                NaN   \n",
       "32                                                NaN   \n",
       "33                                                NaN   \n",
       "34                                                NaN   \n",
       "35                                                NaN   \n",
       "36                                                NaN   \n",
       "37                                                NaN   \n",
       "38                                                NaN   \n",
       "39                                                NaN   \n",
       "40                                                NaN   \n",
       "41                                                NaN   \n",
       "42                                                NaN   \n",
       "43                                                NaN   \n",
       "44                                                NaN   \n",
       "45                                                NaN   \n",
       "46                                                NaN   \n",
       "47                                                NaN   \n",
       "48                                                NaN   \n",
       "\n",
       "                             see_also  \n",
       "0   https://www.jax.org/strain/000646  \n",
       "1   https://www.jax.org/strain/000664  \n",
       "2   https://www.jax.org/strain/002448  \n",
       "3   https://www.jax.org/strain/001976  \n",
       "4   https://www.jax.org/strain/002105  \n",
       "5   https://www.jax.org/strain/000928  \n",
       "6   https://www.jax.org/strain/003715  \n",
       "7   https://www.jax.org/strain/001145  \n",
       "8   https://www.jax.org/strain/101043  \n",
       "9   https://www.jax.org/strain/100002  \n",
       "10                               None  \n",
       "11                               None  \n",
       "12                               None  \n",
       "13                               None  \n",
       "14  https://www.jax.org/strain/019019  \n",
       "15  https://www.jax.org/strain/033781  \n",
       "16  https://www.jax.org/strain/021238  \n",
       "17  https://www.jax.org/strain/021236  \n",
       "18  https://www.jax.org/strain/021237  \n",
       "19  https://www.jax.org/strain/020944  \n",
       "20  https://www.jax.org/strain/020945  \n",
       "21  https://www.jax.org/strain/022869  \n",
       "22  https://www.jax.org/strain/029625  \n",
       "23  https://www.jax.org/strain/026971  \n",
       "24  https://www.jax.org/strain/018856  \n",
       "25  https://www.jax.org/strain/021889  \n",
       "26  https://www.jax.org/strain/018854  \n",
       "27  https://www.jax.org/strain/028409  \n",
       "28  https://www.jax.org/strain/021892  \n",
       "29  https://www.jax.org/strain/018859  \n",
       "30  https://www.jax.org/strain/022870  \n",
       "31  https://www.jax.org/strain/021890  \n",
       "32  https://www.jax.org/strain/021891  \n",
       "33  https://www.jax.org/strain/018857  \n",
       "34  https://www.jax.org/strain/025126  \n",
       "35  https://www.jax.org/strain/026972  \n",
       "36  https://www.jax.org/strain/025426  \n",
       "37  https://www.jax.org/strain/020946  \n",
       "38  https://www.jax.org/strain/025127  \n",
       "39  https://www.jax.org/strain/025423  \n",
       "40                               None  \n",
       "41  https://www.jax.org/strain/021893  \n",
       "42                               None  \n",
       "43  https://www.jax.org/strain/024683  \n",
       "44  https://www.jax.org/strain/026427  \n",
       "45                               None  \n",
       "46                               None  \n",
       "47                               None  \n",
       "48  https://www.jax.org/strain/018855  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def int_csv_to_hex(x):\n",
    "    r, g, b = x.split(\",\")\n",
    "    \n",
    "    return \"#{:02x}{:02x}{:02x}\".format(int(r), int(g), int(b))\n",
    "\n",
    "def remove_citation(x):\n",
    "    return x.split(\" \")[0]\n",
    "\n",
    "strains = pandas.read_excel(\n",
    "    lizs_sheet_name, \n",
    "    sheet_name=\"Line information\", \n",
    "    usecols=[\n",
    "        \"Designation\",\n",
    "        \"Strain\",\n",
    "        \"Note\",\n",
    "        \"Jax Catalog No\",\n",
    "        \"Sample CODE\",\n",
    "        \"Strain notes\",\n",
    "    ],\n",
    "    converters={\n",
    "        \"Strain\": remove_citation,\n",
    "        \"Jax Catalog No\": str,\n",
    "        \"Strain notes\": str_or_empty,\n",
    "    }\n",
    ").dropna(how=\"all\")\n",
    "\n",
    "# Remove the extra lines with the bridge samples for now. I don't want to track thta level of detail\n",
    "#bridge = strains[strains[\"Designation\"] == \"Bridge sample\"].first_valid_index() - 1\n",
    "#strains = strains.iloc[0:bridge]\n",
    "\n",
    "# Add in Strain URL\n",
    "strains[\"see_also\"] = strains[\"Sample CODE\"].apply(lambda x: strain_urls.get(x, x))\n",
    "strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "indian-venture",
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_type_lookup = {\n",
    "    \"CC founder\": models.StrainType.FOUNDER,\n",
    "    \"CC F1\": models.StrainType.F1,\n",
    "    \"CC Cross\": models.StrainType.CROSS,\n",
    "    \"CC Mutant\": models.StrainType.MUTANT,\n",
    "}\n",
    "\n",
    "if models.MouseStrain.objects.count() > 0:\n",
    "    truncate(models.MouseStrain)\n",
    "\n",
    "for i, row in strains.iterrows():\n",
    "    record = models.MouseStrain(\n",
    "        name=row[\"Sample CODE\"],\n",
    "        display_name=strain_name[row[\"Sample CODE\"]],\n",
    "        igvf_id=strain_igvf_id[row[\"Sample CODE\"]],\n",
    "        strain_type=strain_type_lookup[row[\"Note\"]],\n",
    "        jax_catalog_number=row[\"Jax Catalog No\"],\n",
    "        notes=str_or_empty(row[\"Strain notes\"]),\n",
    "        see_also=row[\"see_also\"],\n",
    "        source=jax_source,\n",
    "    )\n",
    "    record.save()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "private-century",
   "metadata": {},
   "source": [
    "This is closest to the Mouse table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "seven-technician",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8\n"
     ]
    }
   ],
   "source": [
    "submitted_mice = {\n",
    "}\n",
    "\n",
    "for accession_prefix in submitted_sheets:\n",
    "    for i, row in submitted_sheets[accession_prefix][\"rodent_donor\"].iterrows():\n",
    "        if not pandas.isnull(row[\"accession\"]):\n",
    "            rodent_identifier = row[\"rodent_identifier\"]\n",
    "            if accession_prefix == \"igvftst\":\n",
    "                see_also_template = \"https://api.sandbox.igvf.org/rodent-donors/{}/\"\n",
    "            elif accession_prefix == \"igvf\":\n",
    "                see_also_template = \"https://api.data.igvf.org/rodent-donors/{}/\"\n",
    "            submitted_mice.setdefault(rodent_identifier, []).append({\n",
    "                \"accession_prefix\": accession_prefix,\n",
    "                \"name\": row[\"accession\"],\n",
    "                \"uuid\": row[\"uuid\"],\n",
    "                \"see_also\": see_also_template.format(row[\"accession\"]),\n",
    "            })\n",
    "        \n",
    "print(len(submitted_mice))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "deluxe-spencer",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "distinct-failure",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def estrus_cycle(x):\n",
    "    if pandas.isnull(x):\n",
    "        return \"NA\"\n",
    "    else:\n",
    "        return x\n",
    "\n",
    "sex_lookup = {\n",
    "    numpy.nan: models.SexEnum.UNKNOWN,\n",
    "    \"Female\": models.SexEnum.FEMALE,\n",
    "    \"Male\": models.SexEnum.MALE,\n",
    "}\n",
    "\n",
    "estrus_stage = {\n",
    "    \"Unknown\": models.EstrusCycle.UNKNOWN,\n",
    "    \"Anestrus\": models.EstrusCycle.ANESTRUS,\n",
    "    \"Anestrus>Proestrus\": models.EstrusCycle.ANESTRUS_PROESTRUS,\n",
    "    \"Proestrus\": models.EstrusCycle.PROESTRUS,\n",
    "    \"Proestrus>Estrus\": models.EstrusCycle.PROESTRUS_ESTRUS,\n",
    "    \"Estrus\": models.EstrusCycle.ESTRUS,\n",
    "    \"Estrus>Metestrus\": models.EstrusCycle.ESTRUS_METESTRUS,\n",
    "    \"Metestrus\": models.EstrusCycle.METESTRUS,\n",
    "    \"Metestrus>Diestrus\": models.EstrusCycle.METESTRUS_DIESTRUS,\n",
    "    \"Diestrus\": models.EstrusCycle.DIESTRUS,\n",
    "    \"Diestrus>Proestrus\": models.EstrusCycle.DIESTRUS_PROESTRUS,\n",
    "}\n",
    "\n",
    "mice = pandas.read_excel(\n",
    "    lizs_sheet_name, \n",
    "    sheet_name=\"mice\", \n",
    "    usecols=range(0,16),\n",
    "    converters={\n",
    "        \"Estres cycle stage\": estrus_cycle,\n",
    "        \"Dissection ID\": int,\n",
    "        \"RMS number\": int_or_none,\n",
    "        \"Housing number\": int_or_none,\n",
    "    }\n",
    ")\n",
    "\n",
    "if models.Mouse.objects.count() > 0:\n",
    "    truncate(models.Mouse)\n",
    "\n",
    "los_angeles_tz = zoneinfo.ZoneInfo(\"America/Los_Angeles\")\n",
    "\n",
    "\n",
    "\n",
    "failed = False\n",
    "for i, row in mice.iterrows():\n",
    "    if not (pandas.isnull(row[\"Mouse Name\"]) or pandas.isnull(row[\"Dissection ID\"])):\n",
    "        name = row[\"Mouse Name\"].replace(\"F1/J\", \"F1J\")\n",
    "        strain_code = strain_name_to_code[row[\"Strain\"]]\n",
    "        \n",
    "        if name == \"056_WSBJ_10F->PWKJ_9F\":\n",
    "            name = \"056_PWKJ_9F\"\n",
    "\n",
    "        try:\n",
    "            strain = models.MouseStrain.objects.get(name=strain_code)\n",
    "        except models.MouseStrain.DoesNotExist as e:\n",
    "            print(f\"Unable to find {row['Strain']} for row {i+2} with code {strain_code}\")\n",
    "            failed = True\n",
    "            continue\n",
    "\n",
    "        if not name[-1] in (\"M\", \"F\"):\n",
    "            raise ValueError(f\"Unrecognized sex field {mouse_name}\")\n",
    "            \n",
    "        name_fields = name.split(\"_\")\n",
    "        \n",
    "        # deal with TERM2R47HNSS_HO\n",
    "        if len(name_fields) > 3:\n",
    "            name_fields = [name_fields[0], \"_\".join(name_fields[1:-1]), name_fields[-1]]\n",
    "        #if len(name_fields) != 3:\n",
    "        #    raise ValueError(f\"Wrong number of attributes in {name} line {i+1}\")\n",
    "            \n",
    "        if not name_fields[1] in strain_igvf_id:\n",
    "            raise ValueError(f\"strain background in {name} not recognized line {i+1}\")\n",
    "        \n",
    "        sample_box = \"\" if pandas.isnull(row[\"Sample box\"]) else row[\"Sample box\"]\n",
    "        \n",
    "        record = models.Mouse(\n",
    "            # should i use liz's disection id?\n",
    "            name=name,\n",
    "            strain=strain,\n",
    "            sex=sex_lookup[row[\"Sex\"]],\n",
    "            weight_g=row[\"Weight (g)\"],\n",
    "            date_of_birth=row[\"DOB\"].date() if not pandas.isnull(row[\"DOB\"]) else None,\n",
    "            harvest_date=row[\"Harvest Date\"].date() if not pandas.isnull(row[\"Harvest Date\"]) else None,\n",
    "            operator=row[\"Operator\"],\n",
    "            notes=str_or_empty(row[\"Comments\"]),\n",
    "            sample_box=sample_box,\n",
    "        )\n",
    "        \n",
    "        if not pandas.isnull(row[\"Estres cycle stage\"]):\n",
    "            # Ignore those transitions.\n",
    "            stage = row[\"Estres cycle stage\"]\n",
    "            if stage.endswith(\"?\"):\n",
    "                stage = stage[:-1]\n",
    "            record.estrus_cycle=estrus_stage[stage]\n",
    "       \n",
    "        record.save()\n",
    "        \n",
    "        if name in submitted_mice:\n",
    "            accessions = []\n",
    "            for accession_record in submitted_mice[name]:\n",
    "                accession = models.Accession(\n",
    "                    accession_prefix=accession_record[\"accession_prefix\"],\n",
    "                    name=accession_record[\"name\"],\n",
    "                    uuid=accession_record[\"uuid\"],\n",
    "                    see_also=accession_record[\"see_also\"],\n",
    "                )\n",
    "                accession.save()\n",
    "                accessions.append(accession)\n",
    "            record.accession.set(accessions)\n",
    "            record.save()\n",
    "        \n",
    "        \n",
    "        \n",
    "assert not failed, \"Check warning messages\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "sporting-orbit",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['627', 'B6J', '10M']"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "name_fields"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "hybrid-destruction",
   "metadata": {},
   "source": [
    "# Tissues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "electoral-ability",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Samples - 8 founders (1621, 17)\n",
      "Samples - Bridge (416, 24)\n"
     ]
    }
   ],
   "source": [
    "#if models.Tissue.objects.count() > 0:\n",
    "#    truncate(models.Tissue)\n",
    "\n",
    "tissue_sheets = [\"Samples - 8 founders\", \"Samples - Bridge\"]\n",
    "tissues = []\n",
    "for tissue_sheet in tissue_sheets:\n",
    "    tissue = pandas.read_excel(\n",
    "        lizs_sheet_name, \n",
    "        sheet_name=tissue_sheet,\n",
    "        header=0,\n",
    "    ).dropna(axis=0, how=\"all\")\n",
    "    print(tissue_sheet, tissue.shape)\n",
    "    tissues.append(tissue)\n",
    "\n",
    "tissues = pandas.concat(tissues)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "threatened-procedure",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_timepoint(value):\n",
    "    value, units = value.split(\" \")\n",
    "    value = float(value)\n",
    "    \n",
    "    unit_map = {\n",
    "        \"day\": models.AgeUnitsEnum.DAY,\n",
    "        \"days\": models.AgeUnitsEnum.DAY,\n",
    "        \"week\": models.AgeUnitsEnum.WEEK,\n",
    "        \"weeks\": models.AgeUnitsEnum.WEEK,\n",
    "        \"month\": models.AgeUnitsEnum.MONTH,\n",
    "        \"months\": models.AgeUnitsEnum.MONTH,\n",
    "    }\n",
    "    return (value, unit_map[units])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "specialized-albert",
   "metadata": {},
   "outputs": [],
   "source": [
    "ontology_map = {}\n",
    "for record in models.OntologyTerm.objects.all():\n",
    "    ontology_map[record.curie] = record"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "alleged-praise",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8\n"
     ]
    }
   ],
   "source": [
    "submitted_tissues = {}\n",
    "\n",
    "for accession_prefix in submitted_sheets:\n",
    "    for i, row in submitted_sheets[accession_prefix][\"tissue\"].iterrows():\n",
    "        if not pandas.isnull(row[\"accession\"]):\n",
    "            if accession_prefix == \"igvftst\":\n",
    "                see_also_template = \"https://api.sandbox.igvf.org/tissues/{}/\"\n",
    "            elif accession_prefix == \"igvf\":\n",
    "                see_also_template = \"https://api.data.igvf.org/tissues/{}/\"\n",
    "            \n",
    "            aliases = row[\"aliases:array\"].split(\",\")\n",
    "            assert len(aliases) == 1\n",
    "            alias_fields = aliases[0][len(\"ali-mortazavi:\"):].split(\"_\")\n",
    "            tissue_id = \"_\".join(alias_fields[0:4])\n",
    "            submitted_tissues.setdefault(tissue_id, []).append({\n",
    "                \"accession_prefix\": accession_prefix,\n",
    "                \"name\": row[\"accession\"],\n",
    "                \"uuid\": row[\"uuid\"],\n",
    "                \"see_also\": see_also_template.format(row[\"accession\"]),\n",
    "            })\n",
    "\n",
    "print(len(submitted_tissues))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "relative-corps",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "if models.Tissue.objects.count() > 0:\n",
    "    truncate(models.Tissue)\n",
    "\n",
    "\n",
    "failed = False\n",
    "for i, row in tissues.iterrows():\n",
    "    if not pandas.isnull(row[\"Tissue\"]) and row[\"IGVF Sample BOX\"] != \"DNA Extraction Optimization\":\n",
    "        mouse_tissue_id_label = \"Mouse_Tissue ID\"\n",
    "        mouse_tissue_name_fields = row[mouse_tissue_id_label].split(\"_\")\n",
    "        \n",
    "        # some validation on tissue name\n",
    "        if len(mouse_tissue_name_fields) != 4:\n",
    "            raise ValueError(f\"Not enough elements in mouse tissue id {row[mouse_tissue_id_label]}\")\n",
    "            \n",
    "        if not mouse_tissue_name_fields[1] in strain_name:\n",
    "            raise ValueError(f\"tissue strain field in {row[mouse_tissue_id_label]} not recognized\")\n",
    "        \n",
    "        mouse_name = \"_\".join(mouse_tissue_name_fields[0:3])\n",
    "        try:\n",
    "            mouse = models.Mouse.objects.get(name=mouse_name)\n",
    "        except models.Mouse.DoesNotExist:\n",
    "            print(\"row {}, {} was not found\".format(i+2, mouse_name))\n",
    "            failed = True\n",
    "            continue\n",
    "\n",
    "        genotype = row[\"Genotype\"]\n",
    "        \n",
    "        # this is the \"label swap\" on spreadsheet rows 602-605.\n",
    "        if mouse_name == \"092_CASTJ_10F\":\n",
    "            genotype = \"CASTJ\"\n",
    "        # this is the other half the swap on spreadsheet rows 1522-1525\n",
    "        elif mouse_name == \"046_NZOJ_10F\":\n",
    "            genotype = 'NZOJ'\n",
    "\n",
    "        assert mouse.strain.name == genotype, f\"{mouse.strain.name} != {genotype}\"\n",
    "        tissue_terms = []\n",
    "        for term_curie, term_name in tissue_dissection_to_ontology_map[row[\"Tissue\"]]:\n",
    "            tissue_terms.append(ontology_map[term_curie])\n",
    "        \n",
    "        #age, age_units = parse_timepoint(row[\"Timepoint\"])\n",
    "        \n",
    "        if pandas.isnull(row[\"Approx. sac time\"]):\n",
    "            sac_time = datetime.time(0,0,0)\n",
    "        else:\n",
    "            sac_time = datetime.time(\n",
    "                row[\"Approx. sac time\"].hour,\n",
    "                row[\"Approx. sac time\"].minute,\n",
    "                row[\"Approx. sac time\"].second,    \n",
    "            )\n",
    "\n",
    "        \n",
    "        record = models.Tissue(\n",
    "            mouse=mouse,\n",
    "            name = row[mouse_tissue_id_label],\n",
    "            description = row[\"Tissue\"],\n",
    "            #dissection_time=dissection,\n",
    "            #age=age,\n",
    "            #age_units=age_units,\n",
    "            tube_label=row[\"Tube label\"],\n",
    "            timepoint_description=row[\"Timepoint\"],\n",
    "            life_stage=models.LifeStageEnum.ADULT,\n",
    "            dissector=str_or_empty(row[\"Dissector\"]),\n",
    "            dissection_notes=str_or_empty(row[\"Comment\"]),\n",
    "        )\n",
    "\n",
    "        if not pandas.isnull(row[\"Dissection date\"]):\n",
    "            record.dissection = datetime.datetime(\n",
    "                row[\"Dissection date\"].year,\n",
    "                row[\"Dissection date\"].month,\n",
    "                row[\"Dissection date\"].day,\n",
    "                sac_time.hour,\n",
    "                sac_time.minute,\n",
    "                sac_time.second,\n",
    "                tzinfo=los_angeles_tz,\n",
    "            )\n",
    "        \n",
    "        tube_weight_label = \"tube weight (g)\"\n",
    "        if not pandas.isnull(row[tube_weight_label]):\n",
    "            record.tube_weight_g = float(row[tube_weight_label])\n",
    "\n",
    "        total_weight_label = \"tube+tissue wight (g)\"\n",
    "        if not pandas.isnull(row[total_weight_label]):\n",
    "            record.total_weight_g = float(row[total_weight_label])\n",
    "        \n",
    "        record.save()\n",
    "        record.ontology_term.set(tissue_terms)\n",
    "        \n",
    "        if row[mouse_tissue_id_label] in submitted_tissues:\n",
    "            accessions = []\n",
    "            for accession_row in submitted_tissues[row[mouse_tissue_id_label]]:\n",
    "                accession = models.Accession(\n",
    "                    accession_prefix=accession_row[\"accession_prefix\"],\n",
    "                    name=accession_row[\"name\"],\n",
    "                    uuid=accession_row[\"uuid\"],\n",
    "                    see_also=accession_row[\"see_also\"],\n",
    "                )\n",
    "                accession.save()\n",
    "                accessions.append(accession)\n",
    "            record.accession.set(accessions)\n",
    "        record.save()\n",
    "\n",
    "assert not failed, \"Check warning messages.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "blocked-council",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IGVF Sample BOX                            IGVF_BridgeSample_5\n",
       "Mouse_Tissue ID                               266_CASTJ_10F_34\n",
       "Unnamed: 2                                                 NaN\n",
       "Tissue                                     Gastrocnemius right\n",
       "Sex                                                          F\n",
       "Timepoint                                             10 weeks\n",
       "Genotype                                                 CASTJ\n",
       "Tube label                                              266-34\n",
       "tube weight (g)                                          1.034\n",
       "tube+tissue wight (g)                                      NaN\n",
       "tissue weight (mg) or blood volume (ul)                 -1.034\n",
       "cryopreserved blood location                               NaN\n",
       "Dissection date                                            NaN\n",
       "Approx. sac time                                           NaN\n",
       "Body weight (g)                                            NaN\n",
       "Dissector                                                  NaN\n",
       "Comment                                                    NaN\n",
       "blood volume (uL)                                          NaN\n",
       "PBMC count 1                                               NaN\n",
       "PBMC count 2                                               NaN\n",
       "PBMC RSB after RBC lysis                                   NaN\n",
       "Total # PBMCs                                              NaN\n",
       "cryopreserved PBMC location                                NaN\n",
       "Dissection Start                                           NaN\n",
       "Dissection End                                             NaN\n",
       "Body Weight (g)                                            NaN\n",
       "Collector                                                  NaN\n",
       "Photographer                                               NaN\n",
       "Name: 415, dtype: object"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "comprehensive-worker",
   "metadata": {},
   "outputs": [],
   "source": [
    "#submitted_tissues[row[mouse_tissue_id_label]], type(submitted_tissues[row[mouse_tissue_id_label]])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "historic-stand",
   "metadata": {},
   "source": [
    "# Information from Samples into experiment\n",
    "\n",
    "- Total barcoded nuclei (Samples into experiment)\n",
    "\n",
    "Is this FixedTissue, FixedSample, FixedBiosample?\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "applied-mustang",
   "metadata": {},
   "outputs": [],
   "source": [
    "if models.FixedSample.objects.count() > 0:\n",
    "    truncate(models.FixedSample)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "lonely-spelling",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_samples_into_experiment(book_name, sheet_name, header=0):\n",
    "\n",
    "    samples_into_experiment = pandas.read_excel(\n",
    "        book_name, \n",
    "        sheet_name=sheet_name,\n",
    "        header=header,\n",
    "        index_col=None,\n",
    "    ).dropna(how=\"all\")\n",
    "\n",
    "    # was trying to normalize two row hierarchical headers, \n",
    "    #level0 = []\n",
    "    #level1 = []\n",
    "    #for header in samples_into_experiment.columns:\n",
    "    #    level0.append(\"\" if header[0].startswith(\"Unnamed: \") else header[0])\n",
    "    #    level1.append(header[1])\n",
    "    #samples_into_experiment.columns = pandas.MultiIndex.from_arrays([level0, level1], names=[\"phase\", \"name\"])\n",
    "\n",
    "    one_based = 1\n",
    "    header_lines = 2\n",
    "    failed = False\n",
    "    for i, row in samples_into_experiment.iterrows():\n",
    "        line_no = i + one_based + header_lines\n",
    "        box_name = row[\"IGVF Fixation BOX\"]\n",
    "        if isinstance(box_name, str):\n",
    "            box_name = box_name.strip()\n",
    "\n",
    "        if box_name in [\"IGVF_FIX_001\", \"IGVF_FIX_002\"]:\n",
    "            continue\n",
    "            \n",
    "        if pandas.isnull(row[\"Mouse_Tissue ID\"]):\n",
    "            continue\n",
    "\n",
    "        tissue_id = row[\"Mouse_Tissue ID\"].replace(\"F1/J\", \"F1J\")\n",
    "        pooled_from = row.get(\"pooled_from\")\n",
    "\n",
    "        if pandas.isnull(tissue_id):\n",
    "            continue\n",
    "\n",
    "        #if (tissue_id.endswith(\"_25\") or tissue_id.endswith(\"_26\")) and pandas.isnull(pooled_from):\n",
    "        #    print(f\"How are we going to merge these samples? {tissue_id}\")\n",
    "        #    continue\n",
    "\n",
    "        if pandas.isnull(pooled_from):\n",
    "            pooled_from = [tissue_id]\n",
    "        else:\n",
    "            pooled_from = pooled_from.split(\",\")\n",
    "\n",
    "        tissues = []\n",
    "        for pooled_id in pooled_from:\n",
    "            try:\n",
    "                tissues.append(models.Tissue.objects.get(name=pooled_id))\n",
    "            except models.Tissue.DoesNotExist:\n",
    "                print(f\"Tissue {pooled_id} not found in tissue table on {line_no}\")\n",
    "                failed = True\n",
    "\n",
    "        weight = row[\"weight (mg)\"]\n",
    "        fixation_date = row[\"Fixation date\"]\n",
    "        notes = row[\"Notes\"]\n",
    "\n",
    "        if weight == \"#VALUE!\":\n",
    "            pass\n",
    "        elif weight < 0:\n",
    "            pass\n",
    "        elif pandas.isnull(weight):\n",
    "            pass\n",
    "        else:\n",
    "            total_weight = 0\n",
    "            for tissue in tissues:\n",
    "                if not pandas.isnull(tissue.weight_mg):\n",
    "                    total_weight += tissue.weight_mg\n",
    "\n",
    "            if not numpy.isclose(total_weight, weight):\n",
    "                print(f\"Sum of tissue weights from {pooled_from} {total_weight} doesn't match {weight}. {line_no}\")\n",
    "\n",
    "        sample_id = row[\"Sample ID\"].replace(\"F1/J\", \"F1J\") if not pandas.isnull(row[\"Sample ID\"]) else None\n",
    "        if not (pandas.isnull(sample_id) or pandas.isnull(tissue_id)):\n",
    "            if tissue_id != sample_id:\n",
    "                print(f\"If defined, {tissue_id} should equal {sample_id} line {line_no}\")\n",
    "                failed = True\n",
    "\n",
    "        cap_label = row[\"Cap label\"]\n",
    "        if not (pandas.isnull(cap_label) or pandas.isnull(sample_id)):\n",
    "            sample_fields = sample_id.split(\"_\")\n",
    "            predicted_label = \"_\".join([sample_fields[0], sample_fields[-1]])\n",
    "            if predicted_label != cap_label:\n",
    "                print(f\"{cap_label} should equal {predicted_label} line {line_no}\")\n",
    "                failed = True\n",
    "\n",
    "        well_id = row[\"wells in Barcoding plate\"]\n",
    "\n",
    "        record = models.FixedSample(\n",
    "            name=tissue_id,\n",
    "            tube_label=cap_label,\n",
    "            fixation_name=box_name,\n",
    "            fixation_date=date_or_none(row[\"Fixation date\"]),\n",
    "            starting_nuclei=int_or_none(row[\"Total nuclei (x10^6)\"]),\n",
    "            nuclei_into_fixation=int_or_none(row[\"Nuclei into fixation (x10^6)\"]),\n",
    "            fixed_nuclei=int_or_none(row[\"Total fixed nuclei (x10^6)\"]),\n",
    "            aliquots_made=int_or_none(row[\"# aliquots\"]),\n",
    "            aliquot_volume_ul=float_or_none(row[\"uL per aliquot\"]),\n",
    "        )\n",
    "        record.save()\n",
    "        record.tissue.set(tissues)\n",
    "        record.save()\n",
    "\n",
    "    assert not failed, \"Check warning messages\"\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "central-oriental",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sum of tissue weights from ['046_NZOJ_10F_17'] 116.0 doesn't match 190.0. 557\n",
      "Sum of tissue weights from ['095_CASTJ_10M_08'] 11.0 doesn't match 15.0. 748\n"
     ]
    }
   ],
   "source": [
    "load_samples_into_experiment(lizs_sheet_name, \"Founder Samples into experiment\", header=1)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "automated-security",
   "metadata": {},
   "outputs": [],
   "source": [
    "load_samples_into_experiment(lizs_sheet_name, \"Bridge samples into experiment\", header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "durable-potter",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load_samples_into_experiment(lizs_sheet_name, \"F1 Samples into experiment\", header=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "specific-candle",
   "metadata": {},
   "source": [
    "# Plate layout\n",
    "\n",
    "(the hard thing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "spectacular-mirror",
   "metadata": {},
   "outputs": [],
   "source": [
    "WellContent = namedtuple(\"well_content\", [\"genotype\", \"tissue_id\"])\n",
    "\n",
    "def validate_tissue_ids(contents, expected_genotypes):\n",
    "    for expected, (row_index, row) in zip(expected_genotypes, contents.iterrows()):\n",
    "        for cell in row:\n",
    "            fields = cell.split(\"_\")\n",
    "            assert len(fields) == 4, f\"Field label validation fail {cell}\"\n",
    "            # there's a few sample swaps, don't validate them.\n",
    "            if fields[1] not in (\"092_CASTJ_10F_03\"):\n",
    "                assert fields[1] == expected, f\"{fields[1]} failed to match {expected}\"\n",
    "\n",
    "def is_plate_name(name):\n",
    "    return not pandas.isnull(name) and name.startswith(\"IGVF_\")\n",
    "\n",
    "def find_plate_start(sheet, offsets):\n",
    "    for plate_id_row in sheet[sheet[offsets[\"plate_label\"]].apply(is_plate_name)].index:\n",
    "        plate_name = sheet.loc[plate_id_row, offsets[\"plate_label\"]]\n",
    "        for i in range(plate_id_row, plate_id_row + 4):\n",
    "            cell = sheet.loc[i, offsets[\"well_start\"]]\n",
    "            if isinstance(cell, str) and (cell.startswith(\"Tissue\") or cell in (\"M\", \"F\")):\n",
    "                yield (plate_name, i)\n",
    "                \n",
    "def get_plate_genotype(sheet, start, offsets):\n",
    "    plate_tissues = []\n",
    "    for cell in sheet.loc[start+2:start+9, offsets[\"tissue_label\"]].tolist():\n",
    "        if not pandas.isnull(cell):\n",
    "            match = re.match(\"(B6J|NODJ|AJ|PWKJ|129S1J|CASTJ|WSBJ|NZOJ)\", cell)\n",
    "            if match is not None:\n",
    "                plate_tissues.append(match.group(0))\n",
    "            else:\n",
    "                print(\"Diffculty matching {} in {} {}\".format(cell, start+2, start+9))\n",
    "    \n",
    "    return plate_tissues\n",
    "\n",
    "def parse_single_well_block(sheet, plate_start, offsets):\n",
    "    # column ids\n",
    "    simple_plate_column_ids = sheet.loc[plate_start+1, offsets[\"well_range\"]].tolist()\n",
    "\n",
    "    plate_row_labels = sheet.loc[plate_start+2:plate_start+9, offsets[\"well_row_label\"]].tolist()\n",
    "\n",
    "    contents = sheet.loc[plate_start+2:plate_start+9, offsets[\"well_range\"]].copy()\n",
    "    contents.index = plate_row_labels\n",
    "    contents.columns = simple_plate_column_ids\n",
    "\n",
    "    genotypes = get_plate_genotype(sheet, plate_start, offsets)\n",
    "    validate_tissue_ids(contents, genotypes)\n",
    "\n",
    "    well_contents = {}\n",
    "    for genotype, (well_row, row) in zip(genotypes, contents.iterrows()):\n",
    "        for well_column, cell in zip(simple_plate_column_ids, row):\n",
    "            well_contents[(well_row, str(well_column))] = [WellContent(genotype, cell)]\n",
    "    return well_contents\n",
    "\n",
    "def parse_multiplexed_well_block(sheet, start, expected_genotypes, offsets):\n",
    "    well_re = re.compile(\"^[A-H]1?[\\d]$\")\n",
    "    wells = sheet.loc[start+3, offsets[\"well_range\"]].tolist()\n",
    "    for cell in wells:\n",
    "        if pandas.isnull(cell) or well_re.match(cell) is None:\n",
    "            raise ValueError(f\"well value {cell} in row {start+3} does not look correct\")\n",
    "\n",
    "    contents = sheet.loc[start+0:start+1, offsets[\"well_range\"]]\n",
    "    contents.columns = [(x[0], x[1:]) for x in wells]\n",
    "    validate_tissue_ids(contents, expected_genotypes)\n",
    "    contents.index = expected_genotypes\n",
    "\n",
    "    well_contents = {}\n",
    "    for genotype, row in contents.iterrows():\n",
    "        for well, tissue_id in row.iteritems():\n",
    "            well_contents.setdefault(well, []).append(WellContent(genotype, tissue_id))\n",
    "\n",
    "    return well_contents\n",
    "\n",
    "def parse_plate(sheet):\n",
    "    max_rows_from_plate_name_to_well_names = 4\n",
    "\n",
    "    offsets = {\n",
    "        \"plate_label\": 1,\n",
    "        \"tissue_label\": 1,\n",
    "        \"well_row_label\": 2,\n",
    "        \"well_start\": 3,\n",
    "        \"well_range\": slice(3, 10),\n",
    "    }\n",
    "    offsets_simple = offsets.copy()\n",
    "    offsets_simple[\"well_range\"] = slice(3, 14)\n",
    "\n",
    "    for plate_name, plate_start in find_plate_start(sheet, offsets):\n",
    "        if plate_name.endswith(\"XX\"):\n",
    "            return\n",
    "\n",
    "        well_contents = {}\n",
    "        genotypes = get_plate_genotype(sheet, plate_start, offsets)        \n",
    "        \n",
    "        if plate_name in (\"IGVF_B01\", \"IGVF_012\"):\n",
    "            well_contents.update(parse_single_well_block(sheet, plate_start, offsets_simple))\n",
    "            complex_blocks = []        \n",
    "        else:\n",
    "            well_contents.update(parse_single_well_block(sheet, plate_start, offsets))\n",
    "\n",
    "            complex_blocks = [\n",
    "                (plate_start + 12, slice(0, 2)),\n",
    "                (plate_start + 12 + 5, slice(2, 4)),\n",
    "                (plate_start + 12 + 10, slice(4, 6)),\n",
    "                (plate_start + 12 + 15, slice(6, 8)),\n",
    "            ]\n",
    "\n",
    "            for complex_start, genotype_slice in complex_blocks:\n",
    "                expected_genotypes = genotypes[genotype_slice]\n",
    "                well_contents.update(parse_multiplexed_well_block(sheet, complex_start, expected_genotypes, offsets))\n",
    "        \n",
    "        yield (plate_name, well_contents)\n",
    "        \n",
    "plate_layout = pandas.read_excel(\n",
    "    lizs_sheet_name, \n",
    "    sheet_name=\"Plate setups\",\n",
    "    header=None,\n",
    ")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "subject-accused",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Validate tissue references:\n",
    "failed = False\n",
    "for plate_name, plate_contents in parse_plate(plate_layout):\n",
    "    for well_id in plate_contents:\n",
    "        well_contents = plate_contents[well_id]\n",
    "        \n",
    "        for well_fraction in well_contents:\n",
    "            try:\n",
    "                biosample = models.FixedSample.objects.get(name=well_fraction.tissue_id)\n",
    "            except models.FixedSample.DoesNotExist:\n",
    "                print(f\"Unable to find {well_fraction.tissue_id} on plate {plate_name}\")\n",
    "                failed = True\n",
    "                \n",
    "#assert not failed, \"Resolve tissues\"\n",
    "\n",
    "# populate database\n",
    "if models.SplitSeqWell.biosample.through.objects.count() > 0:\n",
    "    truncate(models.SplitSeqWell.biosample.through)\n",
    "\n",
    "if models.SplitSeqWell.objects.count() > 0:\n",
    "    truncate(models.SplitSeqWell)\n",
    "    \n",
    "if models.SplitSeqPlate.objects.count() > 0:\n",
    "    truncate(models.SplitSeqPlate)\n",
    "\n",
    "wt_mega_2_kit = models.LibraryConstructionKit.objects.get(name=\"wt-mega\", version=\"v2\")\n",
    "wt_regular_2_kit = models.LibraryConstructionKit.objects.get(name=\"wt\", version=\"v2\")\n",
    "\n",
    "errors = 0\n",
    "for plate_name, plate_contents in parse_plate(plate_layout):\n",
    "    plate_record = models.SplitSeqPlate(\n",
    "        name=plate_name,\n",
    "        size=models.PlateSizeEnum.size_96,\n",
    "        pool_location=None,\n",
    "        date_performed=None,\n",
    "    )\n",
    "    plate_record.save()\n",
    "\n",
    "    for well_id in plate_contents:\n",
    "        well_contents = plate_contents[well_id]\n",
    "        \n",
    "        biosamples = []\n",
    "        for well_content in well_contents:\n",
    "            try:\n",
    "                biosamples.append(models.FixedSample.objects.get(name=well_content.tissue_id))\n",
    "            except models.FixedSample.DoesNotExist:\n",
    "                print(f\"unable to find tissue {well_content.tissue_id} for {plate_name} {well_id}\")\n",
    "                errors += 1\n",
    "        \n",
    "        if plate_name in (\"IGVF_B01\", \"IGVF_012\"):\n",
    "            kit = wt_regular_2_kit\n",
    "        else:\n",
    "            kit = wt_mega_2_kit\n",
    "        \n",
    "        barcodes = models.LibraryBarcode.objects.filter(\n",
    "            kit=kit,\n",
    "            code=\"{}{}\".format(well_id[0], well_id[1]),\n",
    "        )\n",
    "        \n",
    "        well_record = models.SplitSeqWell(\n",
    "            plate=plate_record,\n",
    "            row=well_id[0],\n",
    "            column=well_id[1],\n",
    "        )\n",
    "        well_record.save()\n",
    "        well_record.biosample.set(biosamples)\n",
    "        well_record.barcode.set(barcodes)\n",
    "        well_record.save()\n",
    "        \n",
    "assert not errors, \"Check error messages\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "discrete-pricing",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "headed-scope",
   "metadata": {},
   "source": [
    "# Submitted measurement sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "arctic-burke",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8\n"
     ]
    }
   ],
   "source": [
    "measurement_sets = {}\n",
    "\n",
    "for accession_prefix in submitted_sheets:\n",
    "    for i, row in submitted_sheets[accession_prefix][\"measurement_set\"].iterrows():\n",
    "        if not pandas.isnull(row[\"accession\"]):\n",
    "            aliases = row[\"aliases:array\"].split(',')\n",
    "            name = aliases[0].replace(\"ali-mortazavi:\", \"\")\n",
    "\n",
    "            try:\n",
    "                record = models.MeasurementSet.objects.get(name=name)\n",
    "            except models.MeasurementSet.DoesNotExist:\n",
    "                record = models.MeasurementSet(\n",
    "                    name=name,\n",
    "                )\n",
    "                record.save()\n",
    "\n",
    "                if accession_prefix == \"igvftst\":\n",
    "                    see_also_template = \"https://api.sandbox.igvf.org/measurment-sets/{}/\"\n",
    "                elif accession_prefix == \"igvf\":\n",
    "                    see_also_template = \"https://api.data.igvf.org/measurement-sets/{}/\"\n",
    "                \n",
    "                accession = models.Accession(\n",
    "                    accession_prefix=accession_prefix,\n",
    "                    name=row[\"accession\"],\n",
    "                    uuid=row[\"uuid\"],\n",
    "                    see_also=see_also_template.format(row[\"accession\"]),\n",
    "                )\n",
    "                accession.save()\n",
    "                record.accession.add(accession)\n",
    "                record.save()\n",
    "            measurement_sets[name] = record\n",
    "print(len(measurement_sets))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "endless-compromise",
   "metadata": {},
   "source": [
    "# Subpool\n",
    "\n",
    "## Extract meaning from the Experiments tab"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "permanent-italic",
   "metadata": {},
   "source": [
    "- number of nuclei per subpool (Experiment) \n",
    "- cDNA amp # PCR cycles (Experiment)\n",
    "- cDNA ng/ul in 25ul (experiment)\n",
    "- total cDNA ng (experiment)\n",
    "- Bioanalyzer cDNA ave bp length (expierment)\n",
    "- Sub library Index PCR # (experiment)\n",
    "- Sublibrary Index (experiment)\n",
    "- Sequence for SampleSheet (experiment)\n",
    "- library ng/ul (experiment)\n",
    "- Bioanalyzer library ave bp length (experiment)\n",
    "- Nextseq run number (experiment)\n",
    "\n",
    "- number of 67k aliquots  (experiment)\n",
    "- number of 8k aliquots (experiment)\n",
    "\n",
    "- QC # raw reads (1 mismatch) (experiment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "civil-musician",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Experiment</th>\n",
       "      <th>mice</th>\n",
       "      <th>Sample Box</th>\n",
       "      <th>Fixation Box</th>\n",
       "      <th>Split-seq prep start date</th>\n",
       "      <th>Sublibrary</th>\n",
       "      <th>[Barcoded Nuclei] (nuclei/uL)</th>\n",
       "      <th>Barcoded Nuclei Vol. (uL)</th>\n",
       "      <th>Total barcoded nuclei (1st 3 rounds)</th>\n",
       "      <th># of backup aliquots</th>\n",
       "      <th>...</th>\n",
       "      <th>Unnamed: 47</th>\n",
       "      <th>Unnamed: 48</th>\n",
       "      <th>Unnamed: 49</th>\n",
       "      <th>Unnamed: 50</th>\n",
       "      <th>Unnamed: 51</th>\n",
       "      <th>Unnamed: 52</th>\n",
       "      <th>Unnamed: 53</th>\n",
       "      <th>Unnamed: 54</th>\n",
       "      <th>Unnamed: 55</th>\n",
       "      <th>Unnamed: 56</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IGVF_Splitseq_002</td>\n",
       "      <td>B6 (2F+2M)</td>\n",
       "      <td>IGVF_Sample_3</td>\n",
       "      <td>IGVF_FIX_001</td>\n",
       "      <td>2022-08-24</td>\n",
       "      <td>002_4A</td>\n",
       "      <td>4540.0</td>\n",
       "      <td>240.0</td>\n",
       "      <td>1089600.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CAST (2F+2M)</td>\n",
       "      <td>IGVF_Sample_4</td>\n",
       "      <td>IGVF_FIX_002</td>\n",
       "      <td>NaT</td>\n",
       "      <td>002_4B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Fixation V1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>002_64A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1M V1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>002_64B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>002_64C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1033</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1034</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1035</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1036</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1037 rows  57 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Experiment          mice     Sample Box  Fixation Box  \\\n",
       "0     IGVF_Splitseq_002    B6 (2F+2M)  IGVF_Sample_3  IGVF_FIX_001   \n",
       "1                   NaN  CAST (2F+2M)  IGVF_Sample_4  IGVF_FIX_002   \n",
       "2           Fixation V1           NaN            NaN           NaN   \n",
       "3                 1M V1           NaN            NaN           NaN   \n",
       "4                   NaN           NaN            NaN           NaN   \n",
       "...                 ...           ...            ...           ...   \n",
       "1032                NaN           NaN            NaN           NaN   \n",
       "1033                NaN           NaN            NaN           NaN   \n",
       "1034                NaN           NaN            NaN           NaN   \n",
       "1035                NaN           NaN            NaN           NaN   \n",
       "1036                NaN           NaN            NaN           NaN   \n",
       "\n",
       "     Split-seq prep start date Sublibrary  [Barcoded Nuclei] (nuclei/uL)  \\\n",
       "0                   2022-08-24     002_4A                         4540.0   \n",
       "1                          NaT     002_4B                            NaN   \n",
       "2                          NaT    002_64A                            NaN   \n",
       "3                          NaT    002_64B                            NaN   \n",
       "4                          NaT    002_64C                            NaN   \n",
       "...                        ...        ...                            ...   \n",
       "1032                       NaT        NaN                            NaN   \n",
       "1033                       NaT        NaN                            NaN   \n",
       "1034                       NaT        NaN                            NaN   \n",
       "1035                       NaT        NaN                            NaN   \n",
       "1036                       NaT        NaN                            NaN   \n",
       "\n",
       "      Barcoded Nuclei Vol. (uL)  Total barcoded nuclei (1st 3 rounds)  \\\n",
       "0                         240.0                             1089600.0   \n",
       "1                           NaN                                   NaN   \n",
       "2                           NaN                                   NaN   \n",
       "3                           NaN                                   NaN   \n",
       "4                           NaN                                   NaN   \n",
       "...                         ...                                   ...   \n",
       "1032                        NaN                                   NaN   \n",
       "1033                        NaN                                   NaN   \n",
       "1034                        NaN                                   NaN   \n",
       "1035                        NaN                                   NaN   \n",
       "1036                        NaN                                   NaN   \n",
       "\n",
       "      # of backup aliquots  ... Unnamed: 47  Unnamed: 48  Unnamed: 49  \\\n",
       "0                      3.0  ...         NaN          NaN          NaN   \n",
       "1                      NaN  ...         NaN          NaN          NaN   \n",
       "2                      NaN  ...         NaN          NaN          NaN   \n",
       "3                      NaN  ...         NaN          NaN          NaN   \n",
       "4                      NaN  ...         NaN          NaN          NaN   \n",
       "...                    ...  ...         ...          ...          ...   \n",
       "1032                   NaN  ...         NaN          NaN          NaN   \n",
       "1033                   NaN  ...         NaN          NaN          NaN   \n",
       "1034                   NaN  ...         NaN          NaN          NaN   \n",
       "1035                   NaN  ...         NaN          NaN          NaN   \n",
       "1036                   NaN  ...         NaN          NaN          NaN   \n",
       "\n",
       "      Unnamed: 50 Unnamed: 51 Unnamed: 52 Unnamed: 53  Unnamed: 54  \\\n",
       "0             NaN         NaN         NaN         NaN          NaN   \n",
       "1             NaN         NaN         NaN         NaN          NaN   \n",
       "2             NaN         NaN         NaN         NaN          NaN   \n",
       "3             NaN         NaN         NaN         NaN          NaN   \n",
       "4             NaN         NaN         NaN         NaN          NaN   \n",
       "...           ...         ...         ...         ...          ...   \n",
       "1032          NaN         NaN         NaN         NaN          NaN   \n",
       "1033          NaN         NaN         NaN         NaN          NaN   \n",
       "1034          NaN         NaN         NaN         NaN          NaN   \n",
       "1035          NaN         NaN         NaN         NaN          NaN   \n",
       "1036          NaN         NaN         NaN         NaN          NaN   \n",
       "\n",
       "      Unnamed: 55 Unnamed: 56  \n",
       "0             NaN         NaN  \n",
       "1             NaN         NaN  \n",
       "2             NaN         NaN  \n",
       "3             NaN         NaN  \n",
       "4             NaN         NaN  \n",
       "...           ...         ...  \n",
       "1032          NaN         NaN  \n",
       "1033          NaN         NaN  \n",
       "1034          NaN         NaN  \n",
       "1035          NaN         NaN  \n",
       "1036          NaN         NaN  \n",
       "\n",
       "[1037 rows x 57 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "experiment = pandas.read_excel(\n",
    "    lizs_sheet_name, \n",
    "    sheet_name=\"Experiment\",\n",
    "    header=0,\n",
    "    index_col=None,\n",
    "    #usecols=range(0, 49)\n",
    ")\n",
    "experiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "coordinated-destruction",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Experiment</th>\n",
       "      <th>mice</th>\n",
       "      <th>Sample Box</th>\n",
       "      <th>Fixation Box</th>\n",
       "      <th>Split-seq prep start date</th>\n",
       "      <th>Sublibrary</th>\n",
       "      <th>[Barcoded Nuclei] (nuclei/uL)</th>\n",
       "      <th>Barcoded Nuclei Vol. (uL)</th>\n",
       "      <th>Total barcoded nuclei (1st 3 rounds)</th>\n",
       "      <th># of backup aliquots</th>\n",
       "      <th>...</th>\n",
       "      <th>Novaseq1 L004</th>\n",
       "      <th>Novaseq2 L001</th>\n",
       "      <th>Novaseq2 L002</th>\n",
       "      <th>Novaseq2 L003</th>\n",
       "      <th>Novaseq2 L004</th>\n",
       "      <th>Unnamed: 52</th>\n",
       "      <th>Unnamed: 53</th>\n",
       "      <th>Unnamed: 54</th>\n",
       "      <th>Unnamed: 55</th>\n",
       "      <th>Unnamed: 56</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IGVF_Splitseq_002</td>\n",
       "      <td>B6 (2F+2M)</td>\n",
       "      <td>IGVF_Sample_3</td>\n",
       "      <td>IGVF_FIX_001</td>\n",
       "      <td>2022-08-24</td>\n",
       "      <td>002_4A</td>\n",
       "      <td>4540.0</td>\n",
       "      <td>240.0</td>\n",
       "      <td>1089600.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CAST (2F+2M)</td>\n",
       "      <td>IGVF_Sample_4</td>\n",
       "      <td>IGVF_FIX_002</td>\n",
       "      <td>NaT</td>\n",
       "      <td>002_4B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Fixation V1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>002_64A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1M V1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>002_64B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>002_64C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1033</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1034</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1035</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1036</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1037 rows  57 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Experiment          mice     Sample Box  Fixation Box  \\\n",
       "0     IGVF_Splitseq_002    B6 (2F+2M)  IGVF_Sample_3  IGVF_FIX_001   \n",
       "1                   NaN  CAST (2F+2M)  IGVF_Sample_4  IGVF_FIX_002   \n",
       "2           Fixation V1           NaN            NaN           NaN   \n",
       "3                 1M V1           NaN            NaN           NaN   \n",
       "4                   NaN           NaN            NaN           NaN   \n",
       "...                 ...           ...            ...           ...   \n",
       "1032                NaN           NaN            NaN           NaN   \n",
       "1033                NaN           NaN            NaN           NaN   \n",
       "1034                NaN           NaN            NaN           NaN   \n",
       "1035                NaN           NaN            NaN           NaN   \n",
       "1036                NaN           NaN            NaN           NaN   \n",
       "\n",
       "     Split-seq prep start date Sublibrary  [Barcoded Nuclei] (nuclei/uL)  \\\n",
       "0                   2022-08-24     002_4A                         4540.0   \n",
       "1                          NaT     002_4B                            NaN   \n",
       "2                          NaT    002_64A                            NaN   \n",
       "3                          NaT    002_64B                            NaN   \n",
       "4                          NaT    002_64C                            NaN   \n",
       "...                        ...        ...                            ...   \n",
       "1032                       NaT        NaN                            NaN   \n",
       "1033                       NaT        NaN                            NaN   \n",
       "1034                       NaT        NaN                            NaN   \n",
       "1035                       NaT        NaN                            NaN   \n",
       "1036                       NaT        NaN                            NaN   \n",
       "\n",
       "      Barcoded Nuclei Vol. (uL)  Total barcoded nuclei (1st 3 rounds)  \\\n",
       "0                         240.0                             1089600.0   \n",
       "1                           NaN                                   NaN   \n",
       "2                           NaN                                   NaN   \n",
       "3                           NaN                                   NaN   \n",
       "4                           NaN                                   NaN   \n",
       "...                         ...                                   ...   \n",
       "1032                        NaN                                   NaN   \n",
       "1033                        NaN                                   NaN   \n",
       "1034                        NaN                                   NaN   \n",
       "1035                        NaN                                   NaN   \n",
       "1036                        NaN                                   NaN   \n",
       "\n",
       "      # of backup aliquots  ... Novaseq1 L004  Novaseq2 L001  Novaseq2 L002  \\\n",
       "0                      3.0  ...           NaN            NaN            NaN   \n",
       "1                      NaN  ...           NaN            NaN            NaN   \n",
       "2                      NaN  ...           NaN            NaN            NaN   \n",
       "3                      NaN  ...           NaN            NaN            NaN   \n",
       "4                      NaN  ...           NaN            NaN            NaN   \n",
       "...                    ...  ...           ...            ...            ...   \n",
       "1032                   NaN  ...           NaN            NaN            NaN   \n",
       "1033                   NaN  ...           NaN            NaN            NaN   \n",
       "1034                   NaN  ...           NaN            NaN            NaN   \n",
       "1035                   NaN  ...           NaN            NaN            NaN   \n",
       "1036                   NaN  ...           NaN            NaN            NaN   \n",
       "\n",
       "      Novaseq2 L003 Novaseq2 L004 Unnamed: 52 Unnamed: 53  Unnamed: 54  \\\n",
       "0               NaN           NaN         NaN         NaN          NaN   \n",
       "1               NaN           NaN         NaN         NaN          NaN   \n",
       "2               NaN           NaN         NaN         NaN          NaN   \n",
       "3               NaN           NaN         NaN         NaN          NaN   \n",
       "4               NaN           NaN         NaN         NaN          NaN   \n",
       "...             ...           ...         ...         ...          ...   \n",
       "1032            NaN           NaN         NaN         NaN          NaN   \n",
       "1033            NaN           NaN         NaN         NaN          NaN   \n",
       "1034            NaN           NaN         NaN         NaN          NaN   \n",
       "1035            NaN           NaN         NaN         NaN          NaN   \n",
       "1036            NaN           NaN         NaN         NaN          NaN   \n",
       "\n",
       "      Unnamed: 55 Unnamed: 56  \n",
       "0             NaN         NaN  \n",
       "1             NaN         NaN  \n",
       "2             NaN         NaN  \n",
       "3             NaN         NaN  \n",
       "4             NaN         NaN  \n",
       "...           ...         ...  \n",
       "1032          NaN         NaN  \n",
       "1033          NaN         NaN  \n",
       "1034          NaN         NaN  \n",
       "1035          NaN         NaN  \n",
       "1036          NaN         NaN  \n",
       "\n",
       "[1037 rows x 57 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "experiment = pandas.read_excel(\n",
    "    lizs_sheet_name, \n",
    "    sheet_name=\"Experiment\",\n",
    "    header=0,\n",
    "    index_col=None,\n",
    "    #usecols=range(0, 49)\n",
    ").rename(columns={\n",
    "    # Unnamed: 42 seems like an index or reference to the sample sheet?\n",
    "    \"Unnamed: 43\": \"Novaseq raw reads\",\n",
    "    \"Unnamed: 44\": \"Novaseq1 L001\",\n",
    "    \"Unnamed: 45\": \"Novaseq1 L002\",\n",
    "    \"Unnamed: 46\": \"Novaseq1 L003\",\n",
    "    \"Unnamed: 47\": \"Novaseq1 L004\",\n",
    "    \"Unnamed: 48\": \"Novaseq2 L001\",\n",
    "    \"Unnamed: 49\": \"Novaseq2 L002\",\n",
    "    \"Unnamed: 50\": \"Novaseq2 L003\",\n",
    "    \"Unnamed: 51\": \"Novaseq2 L004\",\n",
    "})\n",
    "\n",
    "experiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "subjective-medicine",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Experiment</th>\n",
       "      <th>mice</th>\n",
       "      <th>Sample Box</th>\n",
       "      <th>Fixation Box</th>\n",
       "      <th>Split-seq prep start date</th>\n",
       "      <th>Sublibrary</th>\n",
       "      <th>[Barcoded Nuclei] (nuclei/uL)</th>\n",
       "      <th>Barcoded Nuclei Vol. (uL)</th>\n",
       "      <th>Total barcoded nuclei (1st 3 rounds)</th>\n",
       "      <th># of backup aliquots</th>\n",
       "      <th>...</th>\n",
       "      <th>Novaseq1 L004</th>\n",
       "      <th>Novaseq2 L001</th>\n",
       "      <th>Novaseq2 L002</th>\n",
       "      <th>Novaseq2 L003</th>\n",
       "      <th>Novaseq2 L004</th>\n",
       "      <th>Unnamed: 52</th>\n",
       "      <th>Unnamed: 53</th>\n",
       "      <th>Unnamed: 54</th>\n",
       "      <th>Unnamed: 55</th>\n",
       "      <th>Unnamed: 56</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IGVF_Splitseq_002</td>\n",
       "      <td>B6 (2F+2M)</td>\n",
       "      <td>IGVF_Sample_3</td>\n",
       "      <td>IGVF_FIX_001</td>\n",
       "      <td>2022-08-24</td>\n",
       "      <td>002_4A</td>\n",
       "      <td>4540.0</td>\n",
       "      <td>240.0</td>\n",
       "      <td>1089600.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>IGVF_Splitseq_003</td>\n",
       "      <td>8 founders (4M+4F)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGVF_FIX_003</td>\n",
       "      <td>2022-12-01</td>\n",
       "      <td>003_8A</td>\n",
       "      <td>6860.0</td>\n",
       "      <td>250.0</td>\n",
       "      <td>1715000.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>IGVF_Splitseq_004</td>\n",
       "      <td>8 founders (4M+4F)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGVF_FIX_005</td>\n",
       "      <td>2022-12-15</td>\n",
       "      <td>004_8A</td>\n",
       "      <td>2460.0</td>\n",
       "      <td>640.0</td>\n",
       "      <td>1574400.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>IGVF_Splitseq_005</td>\n",
       "      <td>8 founders (4M+4F)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGVF_FIX_010</td>\n",
       "      <td>2023-01-23</td>\n",
       "      <td>005_8A</td>\n",
       "      <td>4675.0</td>\n",
       "      <td>390.0</td>\n",
       "      <td>1823250.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>IGVF_Splitseq_006</td>\n",
       "      <td>8 founders (4M+4F)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGVF_FIX_006</td>\n",
       "      <td>2023-02-10</td>\n",
       "      <td>006_8A</td>\n",
       "      <td>3160.0</td>\n",
       "      <td>780.0</td>\n",
       "      <td>2464800.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>IGVF_Splitseq_007</td>\n",
       "      <td>8 founders (4M+4F)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGVF_FIX_006</td>\n",
       "      <td>2023-03-07</td>\n",
       "      <td>007_8A</td>\n",
       "      <td>2810.0</td>\n",
       "      <td>580.0</td>\n",
       "      <td>1629800.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>IGVF_Splitseq_008</td>\n",
       "      <td>8 founders (4M+4F)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGVF_FIX_008</td>\n",
       "      <td>2023-03-24</td>\n",
       "      <td>008_8A</td>\n",
       "      <td>2600.0</td>\n",
       "      <td>250.0</td>\n",
       "      <td>650000.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>IGVF_Splitseq_B01</td>\n",
       "      <td>CASTJ + B6J (2M+2F)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-04-03</td>\n",
       "      <td>B01_13A</td>\n",
       "      <td>4920.0</td>\n",
       "      <td>83.6</td>\n",
       "      <td>411312.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>IGVF_Splitseq_008B</td>\n",
       "      <td>8 founders (4M+4F)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGVF_FIX_008</td>\n",
       "      <td>2023-04-19</td>\n",
       "      <td>008B_13A</td>\n",
       "      <td>3330.0</td>\n",
       "      <td>250.0</td>\n",
       "      <td>832500.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>IGVF_Splitseq_009</td>\n",
       "      <td>8 founders (4M+4F)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGVF_FIX_004</td>\n",
       "      <td>2023-05-02</td>\n",
       "      <td>009_13A</td>\n",
       "      <td>3370.0</td>\n",
       "      <td>285.0</td>\n",
       "      <td>960450.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>IGVF_Splitseq_010</td>\n",
       "      <td>8 founders (4M+4F)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGVF_FIX_011</td>\n",
       "      <td>NaT</td>\n",
       "      <td>010_13A</td>\n",
       "      <td>5440.0</td>\n",
       "      <td>250.0</td>\n",
       "      <td>1360000.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>IGVF_Splitseq_011</td>\n",
       "      <td>8 founders (4M+4F)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGVF_FIX_012</td>\n",
       "      <td>NaT</td>\n",
       "      <td>011_13A</td>\n",
       "      <td>6620.0</td>\n",
       "      <td>250.0</td>\n",
       "      <td>1655000.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "      <td>L001</td>\n",
       "      <td>L002</td>\n",
       "      <td>L003</td>\n",
       "      <td>L004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>IGVF_Splitseq_EX</td>\n",
       "      <td>8 founders (4M+4F)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>003_13A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>IGVF_Splitseq_012</td>\n",
       "      <td>B6J</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGVF_Repeats</td>\n",
       "      <td>NaT</td>\n",
       "      <td>012_13A</td>\n",
       "      <td>3220.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>260820.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14 rows  57 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Experiment                 mice     Sample Box  Fixation Box  \\\n",
       "0     IGVF_Splitseq_002           B6 (2F+2M)  IGVF_Sample_3  IGVF_FIX_001   \n",
       "17    IGVF_Splitseq_003   8 founders (4M+4F)            NaN  IGVF_FIX_003   \n",
       "36    IGVF_Splitseq_004   8 founders (4M+4F)            NaN  IGVF_FIX_005   \n",
       "55    IGVF_Splitseq_005   8 founders (4M+4F)            NaN  IGVF_FIX_010   \n",
       "74    IGVF_Splitseq_006   8 founders (4M+4F)            NaN  IGVF_FIX_006   \n",
       "93    IGVF_Splitseq_007   8 founders (4M+4F)            NaN  IGVF_FIX_006   \n",
       "112   IGVF_Splitseq_008   8 founders (4M+4F)            NaN  IGVF_FIX_008   \n",
       "124   IGVF_Splitseq_B01  CASTJ + B6J (2M+2F)            NaN           NaN   \n",
       "133  IGVF_Splitseq_008B   8 founders (4M+4F)            NaN  IGVF_FIX_008   \n",
       "146   IGVF_Splitseq_009   8 founders (4M+4F)            NaN  IGVF_FIX_004   \n",
       "163   IGVF_Splitseq_010   8 founders (4M+4F)            NaN  IGVF_FIX_011   \n",
       "180   IGVF_Splitseq_011   8 founders (4M+4F)            NaN  IGVF_FIX_012   \n",
       "207    IGVF_Splitseq_EX   8 founders (4M+4F)            NaN           NaN   \n",
       "218   IGVF_Splitseq_012                  B6J            NaN  IGVF_Repeats   \n",
       "\n",
       "    Split-seq prep start date Sublibrary  [Barcoded Nuclei] (nuclei/uL)  \\\n",
       "0                  2022-08-24     002_4A                         4540.0   \n",
       "17                 2022-12-01     003_8A                         6860.0   \n",
       "36                 2022-12-15     004_8A                         2460.0   \n",
       "55                 2023-01-23     005_8A                         4675.0   \n",
       "74                 2023-02-10     006_8A                         3160.0   \n",
       "93                 2023-03-07     007_8A                         2810.0   \n",
       "112                2023-03-24     008_8A                         2600.0   \n",
       "124                2023-04-03    B01_13A                         4920.0   \n",
       "133                2023-04-19   008B_13A                         3330.0   \n",
       "146                2023-05-02    009_13A                         3370.0   \n",
       "163                       NaT    010_13A                         5440.0   \n",
       "180                       NaT    011_13A                         6620.0   \n",
       "207                       NaT    003_13A                            NaN   \n",
       "218                       NaT    012_13A                         3220.0   \n",
       "\n",
       "     Barcoded Nuclei Vol. (uL)  Total barcoded nuclei (1st 3 rounds)  \\\n",
       "0                        240.0                             1089600.0   \n",
       "17                       250.0                             1715000.0   \n",
       "36                       640.0                             1574400.0   \n",
       "55                       390.0                             1823250.0   \n",
       "74                       780.0                             2464800.0   \n",
       "93                       580.0                             1629800.0   \n",
       "112                      250.0                              650000.0   \n",
       "124                       83.6                              411312.0   \n",
       "133                      250.0                              832500.0   \n",
       "146                      285.0                              960450.0   \n",
       "163                      250.0                             1360000.0   \n",
       "180                      250.0                             1655000.0   \n",
       "207                        NaN                                   NaN   \n",
       "218                       81.0                              260820.0   \n",
       "\n",
       "     # of backup aliquots  ... Novaseq1 L004  Novaseq2 L001  Novaseq2 L002  \\\n",
       "0                     3.0  ...           NaN            NaN            NaN   \n",
       "17                    7.0  ...           NaN          Total           L001   \n",
       "36                    5.0  ...           NaN          Total           L001   \n",
       "55                    6.0  ...           NaN          Total           L001   \n",
       "74                    8.0  ...           NaN          Total           L001   \n",
       "93                    8.0  ...           NaN          Total           L001   \n",
       "112                   0.0  ...           NaN          Total           L001   \n",
       "124                  10.0  ...           NaN            NaN            NaN   \n",
       "133                   0.0  ...           NaN          Total           L001   \n",
       "146                   0.0  ...           NaN          Total           L001   \n",
       "163                   3.0  ...           NaN          Total           L001   \n",
       "180                   7.0  ...           NaN          Total           L001   \n",
       "207                   1.0  ...           NaN            NaN            NaN   \n",
       "218                  12.0  ...           NaN            NaN            NaN   \n",
       "\n",
       "     Novaseq2 L003 Novaseq2 L004 Unnamed: 52 Unnamed: 53  Unnamed: 54  \\\n",
       "0              NaN           NaN         NaN         NaN          NaN   \n",
       "17            L002          L003        L004        L001         L002   \n",
       "36            L002          L003        L004        L001         L002   \n",
       "55            L002          L003        L004        L001         L002   \n",
       "74            L002          L003        L004        L001         L002   \n",
       "93            L002          L003        L004        L001         L002   \n",
       "112           L002          L003        L004         NaN          NaN   \n",
       "124            NaN           NaN         NaN         NaN          NaN   \n",
       "133           L002          L003        L004         NaN          NaN   \n",
       "146           L002          L003        L004        L001         L002   \n",
       "163           L002          L003        L004        L001         L002   \n",
       "180           L002          L003        L004        L001         L002   \n",
       "207            NaN           NaN         NaN         NaN          NaN   \n",
       "218            NaN           NaN         NaN         NaN          NaN   \n",
       "\n",
       "     Unnamed: 55 Unnamed: 56  \n",
       "0            NaN         NaN  \n",
       "17          L003        L004  \n",
       "36          L003        L004  \n",
       "55          L003        L004  \n",
       "74          L003        L004  \n",
       "93          L003        L004  \n",
       "112          NaN         NaN  \n",
       "124          NaN         NaN  \n",
       "133          NaN         NaN  \n",
       "146         L003        L004  \n",
       "163         L003        L004  \n",
       "180         L003        L004  \n",
       "207          NaN         NaN  \n",
       "218          NaN         NaN  \n",
       "\n",
       "[14 rows x 57 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def is_experiment_name(name):\n",
    "    not_submittable = [\"IGVF_Splitseq_EX\"]\n",
    "    return not pandas.isnull(name) and name.startswith(\"IGVF_Splitseq\")\n",
    "\n",
    "experiment[experiment[\"Experiment\"].apply(is_experiment_name) & (pandas.notnull(experiment[\"mice\"]))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "gothic-robert",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IGVF_Splitseq_002 0 15\n",
      "IGVF_Splitseq_003 17 33\n",
      "IGVF_Splitseq_004 36 52\n",
      "IGVF_Splitseq_005 55 71\n",
      "IGVF_Splitseq_006 74 91\n",
      "IGVF_Splitseq_007 93 109\n",
      "IGVF_Splitseq_008 112 121\n",
      "IGVF_Splitseq_B01 124 131\n",
      "IGVF_Splitseq_008B 133 143\n",
      "IGVF_Splitseq_009 146 161\n",
      "IGVF_Splitseq_010 163 178\n",
      "IGVF_Splitseq_011 180 195\n",
      "IGVF_Splitseq_EX 207 214\n",
      "IGVF_Splitseq_012 218 225\n"
     ]
    }
   ],
   "source": [
    "def find_fixation_start(sheet):\n",
    "    def is_experiment_name(name):\n",
    "        not_submittable = [\"IGVF_Splitseq_EX\"]\n",
    "        return not pandas.isnull(name) and name.startswith(\"IGVF_Splitseq\") # and not name in not_submittable\n",
    "\n",
    "    for fixation_id_row in sheet[sheet[\"Experiment\"].apply(is_experiment_name) & (pandas.notnull(experiment[\"mice\"]))].index:\n",
    "        name = sheet.loc[fixation_id_row, \"Experiment\"]\n",
    "        start = fixation_id_row\n",
    "\n",
    "        stop = sheet.shape[0]\n",
    "        for i, value in sheet.iloc[start:][\"Sublibrary\"].iteritems():\n",
    "            if pandas.isnull(value):\n",
    "                stop = i - 1\n",
    "                break\n",
    "        \n",
    "        \n",
    "        yield (name, start, stop)\n",
    "\n",
    "for name, start, stop in find_fixation_start(experiment):\n",
    "    print(name, start, stop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "naval-muslim",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NO B01_13A\n",
      "NO B01_13B\n",
      "NO B01_13C\n",
      "NO B01_13D\n",
      "NO B01_13E\n",
      "NO B01_13F\n",
      "EX B01_13G\n",
      "NO B01_13H\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'experiment_name': 'IGVF_Splitseq_B01',\n",
       " 'plate_name': 'IGVF_B01',\n",
       " 'prep_date': Timestamp('2023-04-03 00:00:00'),\n",
       " 'unused_aliquots': {13000: 10},\n",
       " 'subpool': [{'name': 'B01_13A',\n",
       "   'subpool_subname': '13A',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionType.no_selection: 'NO'>,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul': 40,\n",
       "   'cdna_volume': 25.0,\n",
       "   'bioanalyzer_date': datetime.datetime(2023, 4, 5, 0, 0),\n",
       "   'cdna_average_bp_length': 1413,\n",
       "   'index_pcr_no': 11,\n",
       "   'index': 1,\n",
       "   'barcode': 'CAGATC',\n",
       "   'library_ng_per_ul': 25.2,\n",
       "   'library_average_bp_length': 437.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None},\n",
       "  {'name': 'B01_13B',\n",
       "   'subpool_subname': '13B',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionType.no_selection: 'NO'>,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul': 34.9,\n",
       "   'cdna_volume': 25.0,\n",
       "   'bioanalyzer_date': datetime.datetime(2023, 4, 5, 0, 0),\n",
       "   'cdna_average_bp_length': 1469,\n",
       "   'index_pcr_no': 11,\n",
       "   'index': 2,\n",
       "   'barcode': 'ACTTGA',\n",
       "   'library_ng_per_ul': 24.4,\n",
       "   'library_average_bp_length': 447.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None},\n",
       "  {'name': 'B01_13C',\n",
       "   'subpool_subname': '13C',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionType.no_selection: 'NO'>,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul': 41.4,\n",
       "   'cdna_volume': 25.0,\n",
       "   'bioanalyzer_date': datetime.datetime(2023, 4, 5, 0, 0),\n",
       "   'cdna_average_bp_length': 1424,\n",
       "   'index_pcr_no': 11,\n",
       "   'index': 3,\n",
       "   'barcode': 'GATCAG',\n",
       "   'library_ng_per_ul': 24.2,\n",
       "   'library_average_bp_length': 451.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None},\n",
       "  {'name': 'B01_13D',\n",
       "   'subpool_subname': '13D',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionType.no_selection: 'NO'>,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul': 43.8,\n",
       "   'cdna_volume': 25.0,\n",
       "   'bioanalyzer_date': datetime.datetime(2023, 4, 5, 0, 0),\n",
       "   'cdna_average_bp_length': 1352,\n",
       "   'index_pcr_no': 11,\n",
       "   'index': 4,\n",
       "   'barcode': 'TAGCTT',\n",
       "   'library_ng_per_ul': 27.2,\n",
       "   'library_average_bp_length': 449.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None},\n",
       "  {'name': 'B01_13E',\n",
       "   'subpool_subname': '13E',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionType.no_selection: 'NO'>,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul': 36,\n",
       "   'cdna_volume': 25.0,\n",
       "   'bioanalyzer_date': datetime.datetime(2023, 4, 5, 0, 0),\n",
       "   'cdna_average_bp_length': 1347,\n",
       "   'index_pcr_no': 11,\n",
       "   'index': 5,\n",
       "   'barcode': 'ATGTCA',\n",
       "   'library_ng_per_ul': 26.4,\n",
       "   'library_average_bp_length': 433.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None},\n",
       "  {'name': 'B01_13F',\n",
       "   'subpool_subname': '13F',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionType.no_selection: 'NO'>,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul': 44.5,\n",
       "   'cdna_volume': 25.0,\n",
       "   'bioanalyzer_date': datetime.datetime(2023, 4, 5, 0, 0),\n",
       "   'cdna_average_bp_length': 1388,\n",
       "   'index_pcr_no': 11,\n",
       "   'index': 6,\n",
       "   'barcode': 'CTTGTA',\n",
       "   'library_ng_per_ul': 28.2,\n",
       "   'library_average_bp_length': 457.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None},\n",
       "  {'name': 'B01_13G',\n",
       "   'subpool_subname': '13G',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionType.exome_capture: 'EX'>,\n",
       "   'cdna_pcr_rounds': '5 + 8',\n",
       "   'cdna_ng_per_ul': 48,\n",
       "   'cdna_volume': 25.0,\n",
       "   'bioanalyzer_date': datetime.datetime(2023, 4, 5, 0, 0),\n",
       "   'cdna_average_bp_length': 1473,\n",
       "   'index_pcr_no': 11,\n",
       "   'index': 7,\n",
       "   'barcode': 'AGTCAA',\n",
       "   'library_ng_per_ul': 31.2,\n",
       "   'library_average_bp_length': 444.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None},\n",
       "  {'name': 'B01_13H',\n",
       "   'subpool_subname': '13H',\n",
       "   'nuclei': 13000,\n",
       "   'selection_type': <SublibrarySelectionType.no_selection: 'NO'>,\n",
       "   'cdna_pcr_rounds': '5 + 8',\n",
       "   'cdna_ng_per_ul': 41.3,\n",
       "   'cdna_volume': 25.0,\n",
       "   'bioanalyzer_date': datetime.datetime(2023, 4, 5, 0, 0),\n",
       "   'cdna_average_bp_length': 1447,\n",
       "   'index_pcr_no': 11,\n",
       "   'index': 8,\n",
       "   'barcode': 'AGTTCC',\n",
       "   'library_ng_per_ul': 22.6,\n",
       "   'library_average_bp_length': 534.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None,\n",
       "   'novaseq1_raw_reads': None,\n",
       "   'novaseq2_raw_reads': None}]}"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def parse_selection_type(value):\n",
    "    if value is None:\n",
    "        return None\n",
    "    elif value.lower() in (\"normal\", \"no\"):\n",
    "        return models.SublibrarySelectionType.no_selection\n",
    "    elif value in (\"capture\", \"EX\"):\n",
    "        return models.SublibrarySelectionType.exome_capture\n",
    "    else:\n",
    "        raise ValueError(f\"Unrecognized capture type {value}\")\n",
    "\n",
    "\n",
    "def parse_fixation(sheet, experiment_name, start, stop):\n",
    "    block = sheet.loc[start:stop]\n",
    "\n",
    "    experiment_name_fields = experiment_name.split(\"_\")\n",
    "    plate_name = f\"{experiment_name_fields[0]}_{experiment_name_fields[-1]}\"\n",
    "    \n",
    "    aliquots = {}\n",
    "    for i, row in block.iterrows():\n",
    "        count = row[\"# of backup aliquots\"]\n",
    "        size = row[\"Backup Aliquot Size\"]\n",
    "        if not (pandas.isnull(count) or pandas.isnull(size)):\n",
    "            if isinstance(size, str) and size.endswith(\"k\"):\n",
    "                size = size[:-1]\n",
    "            size = int(size) * 1000\n",
    "            aliquots[size] = int(count)\n",
    "    \n",
    "    experiment = {\n",
    "        \"experiment_name\": experiment_name,\n",
    "        \"plate_name\": plate_name,\n",
    "        \"prep_date\": block.loc[start, \"Split-seq prep start date\"],\n",
    "        \"unused_aliquots\": aliquots,\n",
    "        \"subpool\": [],\n",
    "    }\n",
    "\n",
    "    for i, row in block.iterrows():\n",
    "        if not pandas.isnull(row[\"Sublibrary\"]):\n",
    "            if pandas.isnull(row[\"NovaSeq Sequencing date\"]):\n",
    "                novaseq_run_id = None\n",
    "            else:\n",
    "                novaseq_run_id = row[\"NovaSeq Sequencing date\"].isoformat().split(\"T\")[0]\n",
    "\n",
    "            if pandas.isnull(row.get(\"Novaseq raw reads\")):\n",
    "                novaseq_raw_reads = None\n",
    "                novaseq1_raw_reads = None\n",
    "                novaseq2_raw_reads = None\n",
    "            elif row[\"Novaseq raw reads\"] == \"Total\":\n",
    "                novaseq_raw_reads = None\n",
    "                novaseq1_raw_reads = None\n",
    "                novaseq2_raw_reads = None\n",
    "            else:\n",
    "                novaseq_raw_reads = int_or_none(row.get(\"Novaseq raw reads\"))\n",
    "                novaseq1_raw_reads = sum([int_or_none(row.get(x)) for x in[\"Novaseq1 L001\", \"Novaseq1 L002\", \"Novaseq1 L003\", \"Novaseq1 L004\"]])\n",
    "                novaseq2_raw_reads = sum([int_or_none(row.get(x)) for x in[\"Novaseq2 L001\", \"Novaseq2 L002\", \"Novaseq2 L003\", \"Novaseq2 L004\"]])\n",
    "\n",
    "            print(row[\"Sublibrary Type\"], row[\"Sublibrary\"])\n",
    "            subpool = {\n",
    "                \"name\": row[\"Sublibrary\"],\n",
    "                \"subpool_subname\": row[\"Sublibrary\"].split(\"_\")[1],\n",
    "                \"nuclei\": int_or_none(row[\"Nuclei / Sublibrary\"]),\n",
    "                \"selection_type\": parse_selection_type(row[\"Sublibrary Type\"]),\n",
    "                \"cdna_pcr_rounds\": row[\"cDNA Amp # PCR Cycles\"],\n",
    "                \"cdna_ng_per_ul\": row[\"[cDNA] (ng/uL)\"],\n",
    "                \"cdna_volume\": row[\"cDNA Vol. (uL)\"],\n",
    "                #total_cdna = cdna_ng_per_ul_in_25ul * 25\n",
    "                \"bioanalyzer_date\": row[\"BA date\"],\n",
    "                \"cdna_average_bp_length\": int_or_none(row[\"cDNA Ave. Length (bp)\"]),\n",
    "                \"index_pcr_no\": int_or_none(row[\"Sub library Index PCR #\"]),\n",
    "                \"index\": int_or_none(row[\"Sublibrary Index\"]),\n",
    "                \"barcode\": row[\"Sequence for SampleSheet\"],\n",
    "                \"library_ng_per_ul\": float_or_none(row[\"[Library] (ng/ul)\"]),\n",
    "                \"library_average_bp_length\": float_or_none(row[\"Library Ave. Length (bp)\"]),\n",
    "                \"nextseq_run_date\": date_or_none(row[\"QC Sequencing date\"]),\n",
    "                \"nextseq_run_id\": str_or_none(int_or_none(row[\"Nextseq run #\"])),\n",
    "                \"nextseq_run_raw_reads\": int_or_none(row[\"QC # raw reads  (1 mismatch)\"]),\n",
    "                \"novaseq_run_date\": date_or_none(row[\"NovaSeq Sequencing date\"]),\n",
    "                \"novaseq_run_id\": novaseq_run_id,\n",
    "                \"novaseq_run_raw_reads\": novaseq_raw_reads,\n",
    "                \"novaseq1_raw_reads\": novaseq1_raw_reads,\n",
    "                \"novaseq2_raw_reads\": novaseq2_raw_reads,\n",
    "            }\n",
    "            experiment[\"subpool\"].append(subpool)\n",
    "            #if not pandas.isnull(subpool[\"nuclei\"]):\n",
    "            #    nuclei = subpool[\"nuclei\"]\n",
    "            #    aliquots[nuclei] = aliquots.get(nuclei) + 1\n",
    "\n",
    "    return experiment\n",
    "\n",
    "#parse_fixation(experiment, \"IGVF_Splitseq_003\", 17, 32)\n",
    "#parse_fixation(experiment, \"IGVF_Splitseq_008\", 108, 117)\n",
    "parse_fixation(experiment, \"IGVF_Splitseq_B01\", 124, 131)\n",
    "#parse_fixation(experiment, \"IGVF_Splitseq_008B\", 133, 143)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dress-serbia",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "centered-float",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'nextseq2000': <Platform: Nextseq 2000>,\n",
       " 'novaseq6000': <Platform: Novaseq 6000>,\n",
       " 'nanopore': <Platform: Oxford Nanopore>,\n",
       " 'pacbio': <Platform: Pac Bio>}"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "platforms = {x.name: x for x in models.Platform.objects.all()}\n",
    "platforms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "express-zambia",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8ACZ\n",
      "decoded [8, 0, 2, 25]\n",
      "encoded 8ACZ\n",
      "8ADA\n",
      "8A 8B True False False\n",
      "8AA 8Z False False True\n",
      "8B 8E True False False\n",
      "8B 8A False False True\n",
      "8Z 8AA True False False\n",
      "67A 67A False True False\n",
      "8AA 8AB True False False\n",
      "8B\n",
      "8B\n",
      "8A\n"
     ]
    }
   ],
   "source": [
    "class SubpoolName:\n",
    "    def __init__(self, name=None):\n",
    "        if isinstance(name, str):\n",
    "            self._normalized = SubpoolName.subpool_name_to_numbers(name)\n",
    "        elif isinstance(name, list):\n",
    "            self._normalized = name\n",
    "        elif name is None:\n",
    "            self._normalized = None\n",
    "        else:\n",
    "            raise ValueError(\"Unrecognized type {} for {}\".format(type(name), name))\n",
    "\n",
    "    def __str__(self):\n",
    "        if self._normalized is None:\n",
    "            return \"None\"\n",
    "        else:\n",
    "            return SubpoolName.numbers_to_subpool_name(self._normalized)\n",
    "        \n",
    "    def __repr__(self):\n",
    "        if self._normalized is None:\n",
    "            return str(\"{}()\".format(self.__class__.__name__))\n",
    "        else:\n",
    "            return str(\"{}('{}')\".format(self.__class__.__name__, str(self)))\n",
    "        \n",
    "    def __eq__(self, cmp):\n",
    "        if self._normalized is None and cmp is None:\n",
    "            return True\n",
    "        elif self._normalized is None and cmp._normalized is None:\n",
    "            return True\n",
    "        elif self._normalized is None or cmp._normalized is None:\n",
    "            return False\n",
    "        elif len(self._normalized) != len(cmp._normalized):\n",
    "            return False\n",
    "        else:\n",
    "            for left, right in zip(self._normalized, cmp._normalized):\n",
    "                if left != right:\n",
    "                    return False\n",
    "                \n",
    "        return True\n",
    "    \n",
    "    def __lt__(self, cmp):\n",
    "        if self._normalized is None and cmp is None:\n",
    "            return False\n",
    "        elif self._normalized is None and cmp._normalized is None:\n",
    "            return False\n",
    "        elif self._normalized is None:\n",
    "            return True\n",
    "        elif cmp is None or cmp._normalized is None:\n",
    "            return False\n",
    "        elif self._normalized[0] < cmp._normalized[0]:\n",
    "            return True\n",
    "        elif len(self._normalized) < len(cmp._normalized):\n",
    "            return True\n",
    "        elif len(self._normalized) > len(cmp._normalized):\n",
    "            return False\n",
    "        else:\n",
    "            for left, right in zip(self._normalized, cmp._normalized):\n",
    "                if left == right:\n",
    "                    continue\n",
    "                elif left < right:\n",
    "                    return True\n",
    "                else:\n",
    "                    return False\n",
    "                \n",
    "        return False\n",
    "                \n",
    "    def __gt__(self, cmp):\n",
    "        if self._normalized is None and cmp is None:\n",
    "            return False\n",
    "        elif self._normalized is None and cmp._normalized is None:\n",
    "            return False\n",
    "        elif cmp is None or cmp._normalized is None:\n",
    "            return True\n",
    "        elif self._normalized[0] > cmp._normalized[0]:\n",
    "            return True\n",
    "        elif len(self._normalized) < len(cmp._normalized):\n",
    "            return False\n",
    "        elif len(self._normalized) > len(cmp._normalized):\n",
    "            return True\n",
    "        else:\n",
    "            for left, right in zip(self._normalized, cmp._normalized):\n",
    "                if left == right:\n",
    "                    continue\n",
    "                elif left < right:\n",
    "                    return False\n",
    "                else:\n",
    "                    return True\n",
    "            \n",
    "        return False\n",
    "        \n",
    "    @classmethod\n",
    "    def subpool_name_to_numbers(cls, name):\n",
    "        STATE_DIGITS = 1\n",
    "        STATE_LETTERS = 2\n",
    "        STATE_ERROR = -1\n",
    "\n",
    "        digits = []\n",
    "        letters = []\n",
    "        state = STATE_DIGITS\n",
    "        for char in name:\n",
    "            if state == STATE_DIGITS:\n",
    "                if char.isdigit():\n",
    "                    digits.append(char)\n",
    "                elif char.isalpha():\n",
    "                    state = STATE_LETTERS\n",
    "                    letters.append(char)\n",
    "                else:\n",
    "                    raise ValueError(\"Unrecognized symbol {} in {}\".format(char, name))\n",
    "            elif state == STATE_LETTERS:\n",
    "                if char.isdigit():\n",
    "                    raise ValueError(\"Unexpected digit {} in {}\".format(char, name))\n",
    "                elif char.isalpha():\n",
    "                    letters.append(char)\n",
    "                else:\n",
    "                    raise ValueError(\"Unrecognized symbol {} in {}\".format(char, name))\n",
    "\n",
    "        result = [int(\"\".join(digits))]\n",
    "        for number in letters:\n",
    "            if number < \"A\" or number > \"Z\":\n",
    "                raise ValueError(\"out of bounds letter code in {}\".format(name))\n",
    "            else:\n",
    "                result.append(ord(number) - ord(\"A\"))\n",
    "\n",
    "        return result\n",
    "\n",
    "    @classmethod\n",
    "    def numbers_to_subpool_name(cls, subpool_list):\n",
    "\n",
    "        result = [str(subpool_list[0])]\n",
    "        for char in subpool_list[1:]:\n",
    "            result.append(chr(char + ord(\"A\")))\n",
    "\n",
    "        return \"\".join(result)\n",
    "\n",
    "    def next_code(self):\n",
    "        if self._normalized is None:\n",
    "            return None\n",
    "\n",
    "        carry = 0\n",
    "        advancable = []\n",
    "        for i, place in enumerate(reversed(self._normalized[1:])):\n",
    "            if i == 0:\n",
    "                place += 1\n",
    "            else:\n",
    "                place += carry\n",
    "                carry = 0\n",
    "\n",
    "            if place > 25:\n",
    "                carry = 1\n",
    "                place = 0\n",
    "\n",
    "            advancable.append(place)\n",
    "\n",
    "        if carry > 0:\n",
    "            advancable.append(0)\n",
    "\n",
    "        result = [self._normalized[0]]\n",
    "        result.extend(reversed(advancable))\n",
    "        \n",
    "        return SubpoolName(result)\n",
    "\n",
    "s = SubpoolName(\"8ACZ\")\n",
    "print(s)\n",
    "\n",
    "decoded = SubpoolName.subpool_name_to_numbers(\"8ACZ\") \n",
    "print(\"decoded\", decoded)\n",
    "encoded = SubpoolName.numbers_to_subpool_name(decoded)\n",
    "print(\"encoded\", encoded)\n",
    "\n",
    "print(SubpoolName(\"8ACZ\").next_code())\n",
    "\n",
    "\n",
    "\n",
    "for pairs in [(\"8A\", \"8B\"), (\"8AA\", \"8Z\"), (\"8B\", \"8E\"), (\"8B\", \"8A\"), (\"8Z\", \"8AA\"), (\"67A\", \"67A\"), (\"8AA\", \"8AB\")]:\n",
    "    left = SubpoolName(pairs[0])\n",
    "    right = SubpoolName(pairs[1])\n",
    "    print(left, right, left < right, left == right, left > right)\n",
    "    \n",
    "print(max(SubpoolName(\"8A\"), SubpoolName(\"8B\")))\n",
    "print(max(None, SubpoolName(\"8B\")))\n",
    "print(max(SubpoolName(\"8A\"), None))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "parliamentary-pendant",
   "metadata": {},
   "source": [
    "# load subpools, sequencing runs and subpools in sequencing run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "hydraulic-nurse",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8\n"
     ]
    }
   ],
   "source": [
    "multiplexed_samples = {}\n",
    "\n",
    "for accession_prefix in submitted_sheets:\n",
    "    for i, row in submitted_sheets[accession_prefix][\"multiplexed_sample\"].iterrows():\n",
    "        if not pandas.isnull(row[\"accession\"]):\n",
    "            aliases = row[\"aliases:array\"].split(',')\n",
    "            name = aliases[0].replace(\"ali-mortazavi:\", \"\")\n",
    "            if accession_prefix == \"igvftst\":\n",
    "                see_also_template = \"https://api.sandbox.igvf.org/multiplexed-samples/{}/\"\n",
    "            elif accession_prefix == \"igvf\":\n",
    "                see_also_template = \"https://api.data.igvf.org/multiplexed-samples/{}/\"\n",
    "\n",
    "            multiplexed_samples.setdefault(name, []).append({\n",
    "                \"accession_prefix\": accession_prefix,\n",
    "                \"name\": row[\"accession\"],\n",
    "                \"uuid\": row[\"uuid\"],\n",
    "                \"see_also\": see_also_template.format(row[\"accession\"]),\n",
    "            })\n",
    "print(len(multiplexed_samples))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "victorian-basics",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NO 002_4A\n",
      "NO 002_4B\n",
      "NO 002_64A\n",
      "NO 002_64B\n",
      "NO 002_64C\n",
      "NO 002_64D\n",
      "NO 002_64E\n",
      "NO 002_64F\n",
      "NO 002_64G\n",
      "NO 002_64H\n",
      "NO 002_64I\n",
      "NO 002_64J\n",
      "NO 002_64K\n",
      "NO 002_64L\n",
      "NO 002_64M\n",
      "NO 002_64N\n",
      "Unable to find plate IGVF_002\n",
      "NO 003_8A\n",
      "NO 003_13A\n",
      "NO 003_67A\n",
      "NO 003_67B\n",
      "NO 003_67C\n",
      "NO 003_67D\n",
      "NO 003_67E\n",
      "NO 003_67F\n",
      "NO 003_67G\n",
      "NO 003_67H\n",
      "NO 003_67I\n",
      "NO 003_67J\n",
      "NO 003_67K\n",
      "NO 003_67L\n",
      "NO 003_67M\n",
      "NO 003_67N\n",
      "NO 003_67O\n",
      "003_67P\n",
      "003_67Q\n",
      "003_67R\n",
      "003_67S\n",
      "003_67T\n",
      "003_67U\n",
      "003_67V\n",
      "003_13B\n",
      "003_8B\n",
      "003_8C\n",
      "NO 004_8A\n",
      "NO 004_13A\n",
      "NO 004_67A\n",
      "NO 004_67B\n",
      "NO 004_67C\n",
      "NO 004_67D\n",
      "NO 004_67E\n",
      "NO 004_67F\n",
      "NO 004_67G\n",
      "NO 004_67H\n",
      "NO 004_67I\n",
      "NO 004_67J\n",
      "NO 004_67K\n",
      "NO 004_67L\n",
      "NO 004_67M\n",
      "NO 004_67N\n",
      "NO 004_67O\n",
      "004_67P\n",
      "004_67Q\n",
      "004_67R\n",
      "004_67S\n",
      "004_67T\n",
      "004_13B\n",
      "004_13C\n",
      "004_8B\n",
      "004_8C\n",
      "NO 005_8A\n",
      "NO 005_13A\n",
      "NO 005_67A\n",
      "NO 005_67B\n",
      "NO 005_67C\n",
      "NO 005_67D\n",
      "NO 005_67E\n",
      "NO 005_67F\n",
      "NO 005_67G\n",
      "NO 005_67H\n",
      "NO 005_67I\n",
      "NO 005_67J\n",
      "NO 005_67K\n",
      "NO 005_67L\n",
      "NO 005_67M\n",
      "NO 005_67N\n",
      "NO 005_67O\n",
      "005_67P\n",
      "005_67Q\n",
      "005_67R\n",
      "005_67S\n",
      "005_67T\n",
      "005_67U\n",
      "005_13B\n",
      "005_13C\n",
      "005_8B\n",
      "005_8C\n",
      "005_8D\n",
      "005_8E\n",
      "005_8F\n",
      "005_8G\n",
      "005_8H\n",
      "NO 006_8A\n",
      "NO 006_67A\n",
      "NO 006_67B\n",
      "NO 006_67C\n",
      "NO 006_67D\n",
      "NO 006_67E\n",
      "NO 006_67F\n",
      "NO 006_67G\n",
      "NO 006_67H\n",
      "NO 006_67I\n",
      "NO 006_67J\n",
      "NO 006_67K\n",
      "NO 006_67L\n",
      "NO 006_67M\n",
      "NO 006_67N\n",
      "NO 006_67O\n",
      "NO 006_67P\n",
      "NO 006_67Q\n",
      "Skipping failed plate IGVF_006\n",
      "NO 007_8A\n",
      "NO 007_13A\n",
      "NO 007_67A\n",
      "NO 007_67B\n",
      "NO 007_67C\n",
      "NO 007_67D\n",
      "NO 007_67E\n",
      "NO 007_67F\n",
      "NO 007_67G\n",
      "NO 007_67H\n",
      "NO 007_67I\n",
      "NO 007_67J\n",
      "NO 007_67K\n",
      "NO 007_67L\n",
      "NO 007_67M\n",
      "NO 007_67N\n",
      "NO 007_67O\n",
      "007_67P\n",
      "007_67Q\n",
      "007_67R\n",
      "007_67S\n",
      "007_67T\n",
      "007_67U\n",
      "007_67V\n",
      "007_67W\n",
      "007_13B\n",
      "007_13C\n",
      "007_8B\n",
      "007_8C\n",
      "007_8D\n",
      "007_8E\n",
      "007_8F\n",
      "007_8G\n",
      "007_8H\n",
      "007_8I\n",
      "NO 008_8A\n",
      "NO 008_67A\n",
      "NO 008_67B\n",
      "NO 008_67C\n",
      "NO 008_67D\n",
      "NO 008_67E\n",
      "NO 008_67F\n",
      "NO 008_67G\n",
      "NO 008_67H\n",
      "NO 008_67I\n",
      "NO B01_13A\n",
      "NO B01_13B\n",
      "NO B01_13C\n",
      "NO B01_13D\n",
      "NO B01_13E\n",
      "NO B01_13F\n",
      "EX B01_13G\n",
      "NO B01_13H\n",
      "subpool_b01_13a B01_13A [{'accession_prefix': 'igvftst', 'name': 'TSTSM35366348', 'uuid': '77a93b75-f44a-4aa5-937b-a6788051bc18', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM35366348/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1676SYRC', 'uuid': 'efad3fdd-f466-4c59-94e0-8791e2bfaa93', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM1676SYRC/'}]\n",
      "subpool_b01_13a B01_13A [{'accession_prefix': 'igvftst', 'name': 'TSTSM35366348', 'uuid': '77a93b75-f44a-4aa5-937b-a6788051bc18', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM35366348/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1676SYRC', 'uuid': 'efad3fdd-f466-4c59-94e0-8791e2bfaa93', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM1676SYRC/'}]\n",
      "subpool_b01_13b B01_13B [{'accession_prefix': 'igvftst', 'name': 'TSTSM99018100', 'uuid': '6cac6a8b-7750-4df6-92cf-fae41082813d', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM99018100/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1955RVUC', 'uuid': 'f45edb75-5e28-4578-a2d5-395546f64b2a', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM1955RVUC/'}]\n",
      "subpool_b01_13b B01_13B [{'accession_prefix': 'igvftst', 'name': 'TSTSM99018100', 'uuid': '6cac6a8b-7750-4df6-92cf-fae41082813d', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM99018100/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM1955RVUC', 'uuid': 'f45edb75-5e28-4578-a2d5-395546f64b2a', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM1955RVUC/'}]\n",
      "subpool_b01_13c B01_13C [{'accession_prefix': 'igvftst', 'name': 'TSTSM36900951', 'uuid': '01858fa2-3c3d-476a-8d81-e82a0612c0e8', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM36900951/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8093RPKR', 'uuid': '33b1fa8c-53ac-4db9-81e2-df137881c434', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM8093RPKR/'}]\n",
      "subpool_b01_13c B01_13C [{'accession_prefix': 'igvftst', 'name': 'TSTSM36900951', 'uuid': '01858fa2-3c3d-476a-8d81-e82a0612c0e8', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM36900951/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8093RPKR', 'uuid': '33b1fa8c-53ac-4db9-81e2-df137881c434', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM8093RPKR/'}]\n",
      "subpool_b01_13d B01_13D [{'accession_prefix': 'igvftst', 'name': 'TSTSM27245068', 'uuid': 'b69b021b-aa90-41cf-8d1f-cfc59c726b20', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM27245068/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9603DIDI', 'uuid': 'abb7e2a6-c08b-4a15-88fe-94c5c38a3400', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM9603DIDI/'}]\n",
      "subpool_b01_13d B01_13D [{'accession_prefix': 'igvftst', 'name': 'TSTSM27245068', 'uuid': 'b69b021b-aa90-41cf-8d1f-cfc59c726b20', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM27245068/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM9603DIDI', 'uuid': 'abb7e2a6-c08b-4a15-88fe-94c5c38a3400', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM9603DIDI/'}]\n",
      "subpool_b01_13e B01_13E [{'accession_prefix': 'igvftst', 'name': 'TSTSM90710268', 'uuid': '8d44590d-bcb5-46f9-8c87-ea375db12ea9', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM90710268/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6958ESMJ', 'uuid': '92fcc1a8-9170-4a7c-90f8-5ec339a987d1', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM6958ESMJ/'}]\n",
      "subpool_b01_13e B01_13E [{'accession_prefix': 'igvftst', 'name': 'TSTSM90710268', 'uuid': '8d44590d-bcb5-46f9-8c87-ea375db12ea9', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM90710268/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6958ESMJ', 'uuid': '92fcc1a8-9170-4a7c-90f8-5ec339a987d1', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM6958ESMJ/'}]\n",
      "subpool_b01_13f B01_13F [{'accession_prefix': 'igvftst', 'name': 'TSTSM57985101', 'uuid': '2f34272a-6603-41e0-9820-cf2dd432244b', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM57985101/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6546CQJB', 'uuid': 'b2e62292-5163-4a73-9861-23a535a97dac', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM6546CQJB/'}]\n",
      "subpool_b01_13f B01_13F [{'accession_prefix': 'igvftst', 'name': 'TSTSM57985101', 'uuid': '2f34272a-6603-41e0-9820-cf2dd432244b', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM57985101/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM6546CQJB', 'uuid': 'b2e62292-5163-4a73-9861-23a535a97dac', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM6546CQJB/'}]\n",
      "subpool_b01_13g B01_13G [{'accession_prefix': 'igvftst', 'name': 'TSTSM10953505', 'uuid': 'b0e63479-2caf-4aac-9c98-2ff823f92c74', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM10953505/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8722YQAW', 'uuid': '00179685-55cf-4dc1-a00a-3f5fb9133288', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM8722YQAW/'}]\n",
      "subpool_b01_13g B01_13G [{'accession_prefix': 'igvftst', 'name': 'TSTSM10953505', 'uuid': 'b0e63479-2caf-4aac-9c98-2ff823f92c74', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM10953505/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM8722YQAW', 'uuid': '00179685-55cf-4dc1-a00a-3f5fb9133288', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM8722YQAW/'}]\n",
      "subpool_b01_13h B01_13H [{'accession_prefix': 'igvftst', 'name': 'TSTSM89995437', 'uuid': '5484bec6-4b0e-4488-b3b1-fd4def0022df', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM89995437/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2294BOOU', 'uuid': 'f25eb07b-ed33-496f-a8d7-83814b4b642f', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM2294BOOU/'}]\n",
      "subpool_b01_13h B01_13H [{'accession_prefix': 'igvftst', 'name': 'TSTSM89995437', 'uuid': '5484bec6-4b0e-4488-b3b1-fd4def0022df', 'see_also': 'https://api.sandbox.igvf.org/multiplexed-samples/TSTSM89995437/'}, {'accession_prefix': 'igvf', 'name': 'IGVFSM2294BOOU', 'uuid': 'f25eb07b-ed33-496f-a8d7-83814b4b642f', 'see_also': 'https://api.data.igvf.org/multiplexed-samples/IGVFSM2294BOOU/'}]\n",
      "B01_13I\n",
      "B01_13J\n",
      "B01_13K\n",
      "B01_13L\n",
      "B01_13M\n",
      "B01_13N\n",
      "B01_13O\n",
      "B01_13P\n",
      "B01_13Q\n",
      "B01_13R\n",
      "EX 008B_13A\n",
      "NO 008B_67A\n",
      "NO 008B_67B\n",
      "NO 008B_67C\n",
      "NO 008B_67D\n",
      "NO 008B_67E\n",
      "NO 008B_67F\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NO 008B_67G\n",
      "NO 008B_67H\n",
      "NO 008B_67I\n",
      "NO 008B_67J\n",
      "Unable to find plate IGVF_008B\n",
      "EX 009_13A\n",
      "NO 009_67A\n",
      "NO 009_67B\n",
      "NO 009_67C\n",
      "NO 009_67D\n",
      "NO 009_67E\n",
      "NO 009_67F\n",
      "NO 009_67G\n",
      "NO 009_67H\n",
      "NO 009_67I\n",
      "NO 009_67J\n",
      "NO 009_67K\n",
      "NO 009_67L\n",
      "NO 009_67M\n",
      "NO 009_67N\n",
      "NO 009_67O\n",
      "EX 010_13A\n",
      "NO 010_67A\n",
      "NO 010_67B\n",
      "NO 010_67C\n",
      "NO 010_67D\n",
      "NO 010_67E\n",
      "NO 010_67F\n",
      "NO 010_67G\n",
      "NO 010_67H\n",
      "NO 010_67I\n",
      "NO 010_67J\n",
      "NO 010_67K\n",
      "NO 010_67L\n",
      "NO 010_67M\n",
      "NO 010_67N\n",
      "NO 010_67O\n",
      "010_67P\n",
      "010_67Q\n",
      "010_67R\n",
      "010_13B\n",
      "010_13C\n",
      "010_13D\n",
      "010_13E\n",
      "EX 011_13A\n",
      "NO 011_67A\n",
      "NO 011_67B\n",
      "NO 011_67C\n",
      "NO 011_67D\n",
      "NO 011_67E\n",
      "NO 011_67F\n",
      "NO 011_67G\n",
      "NO 011_67H\n",
      "NO 011_67I\n",
      "NO 011_67J\n",
      "NO 011_67K\n",
      "NO 011_67L\n",
      "NO 011_67M\n",
      "NO 011_67N\n",
      "NO 011_67O\n",
      "011_67P\n",
      "011_67Q\n",
      "011_67R\n",
      "011_67S\n",
      "011_67T\n",
      "011_67U\n",
      "011_67V\n",
      "011_13B\n",
      "011_13C\n",
      "011_13D\n",
      "011_13E\n",
      "011_13F\n",
      "011_13G\n",
      "011_13H\n",
      "EX 003_13A\n",
      "EX 004_13A\n",
      "EX 005_13A\n",
      "EX 007_13A\n",
      "EX 008B_13A\n",
      "EX 009_13A\n",
      "EX 010_13A\n",
      "EX 011_13A\n",
      "Unable to find plate IGVF_EX\n",
      "NO 012_13A\n",
      "NO 012_13B\n",
      "NO 012_13C\n",
      "NO 012_13D\n",
      "NO 012_13E\n",
      "NO 012_13F\n",
      "NO 012_13G\n",
      "NO 012_13H\n",
      "Unable to find plate IGVF_012\n"
     ]
    }
   ],
   "source": [
    "compression=r\"(?<compression>gz|bz2|xz|zstd)\"\n",
    "nanopore_library_name_re_template = r\"igvf(?P<run>[\\d]+)_(?P<library_id>{library_name})_.*_(?P<fragment>[\\d]+)\\.fastq\\.{compression}\"\n",
    "nextseq_library_name_re_template = r\"(?P<run>[\\d]+)_(?P<library_id>{library_name})_(?P<read_id>[RI][\\d])\\.fastq\\.{compression}\"\n",
    "nextseq_library_index_re_template = r\"Sublibrary_(?P<library_id>{library_index})_S(?P<sample_id>[\\d]+)_(?P<lane_id>L[\\d]+)_(?P<read_id>[RI][\\d])_(?P<fragment>[\\d]+)\\.fastq\\.{compression}\"\n",
    "novaseq_library_index_re_template = r\"Sublibrary_(?P<library_id>{library_index})_S(?P<sample_id>[\\d]+)_(?P<lane_id>L[\\d]+)_(?P<read_id>[RI][\\d])_(?P<fragment>[\\d]+)\\.fastq\\.{compression}\"\n",
    "\n",
    "expected_names = {\n",
    "    \"IGVF_Splitseq_003\": {\n",
    "        \"nanopore\": nanopore_library_name_re_template,\n",
    "        \"nextseq\": nextseq_library_name_re_template,\n",
    "        \"novaseq\": novaseq_library_index_re_template,\n",
    "    },\n",
    "    \"IGVF_Splitseq_004\": {\n",
    "        \"nanopore\": nanopore_library_name_re_template,\n",
    "        \"nextseq\": nextseq_library_name_re_template,\n",
    "        \"novaseq\": novaseq_library_index_re_template,\n",
    "    },\n",
    "    \"IGVF_Splitseq_005\": {\n",
    "        \"nanopore\": nanopore_library_name_re_template,\n",
    "        \"nextseq\": nextseq_library_index_re_template,\n",
    "        \"novaseq\": novaseq_library_index_re_template,\n",
    "    },\n",
    "    \"IGVF_Splitseq_006\": {\n",
    "        \"nanopore\": nanopore_library_name_re_template,\n",
    "        \"nextseq\": nextseq_library_index_re_template,\n",
    "        \"novaseq\": novaseq_library_index_re_template,\n",
    "    },\n",
    "    \n",
    "}\n",
    "\n",
    "nanopore_runs = {\n",
    "    \"IGVF_Splitseq_003\": {\n",
    "        \"name\": \"2023-01-03-8A\",\n",
    "        \"run_date\": datetime.datetime(2023,1,3),\n",
    "        \"raw_reads\": 87784162,\n",
    "    }\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "failed_plates = {\"IGVF_006\",}\n",
    "failed = False\n",
    "for experiment_name, experiment_start, experiment_stop in find_fixation_start(experiment):\n",
    "    fixation = parse_fixation(experiment, experiment_name, experiment_start, experiment_stop)\n",
    "    \n",
    "    try:\n",
    "        plate = models.SplitSeqPlate.objects.get(name=fixation[\"plate_name\"])\n",
    "    except models.SplitSeqPlate.DoesNotExist:\n",
    "        print(f\"Unable to find plate {fixation['plate_name']}\")\n",
    "        continue\n",
    "        \n",
    "    if fixation[\"plate_name\"] in failed_plates:\n",
    "        print(f\"Skipping failed plate {fixation['plate_name']}\")\n",
    "        continue\n",
    "        \n",
    "    last_subpools = {}\n",
    "    \n",
    "    wt_mega_2_kit = models.LibraryConstructionKit.objects.get(name=\"wt-mega\", version=\"v2\")\n",
    "    for line_offset, subpool in enumerate(fixation[\"subpool\"]):\n",
    "        try:\n",
    "            barcodes = models.LibraryBarcode.objects.filter(\n",
    "                kit=wt_mega_2_kit, \n",
    "                barcode_type=None, \n",
    "                code=subpool[\"index\"])\n",
    "        except models.LibraryBarcode.DoesNotExist:\n",
    "            print(f\"Unable to find barcode {subpool['index']}\")\n",
    "\n",
    "        # validate barcodes\n",
    "        library_barcodes_sequence = {b.sequence for b in barcodes}\n",
    "        if not pandas.isnull(subpool[\"barcode\"]):\n",
    "            expected_sequence = set(subpool[\"barcode\"].split(\",\"))\n",
    "        else:\n",
    "            expected_sequence = set()\n",
    "        if library_barcodes_sequence != expected_sequence:\n",
    "            print(f\"{subpool['name']} Database lookup of barcodes {library_barcodes_sequence} doesn't match human entry {expected_sequence}\")\n",
    "            failed = True\n",
    "            continue\n",
    "        \n",
    "        subpool_prefix = subpool[\"name\"].split(\"_\")[0]\n",
    "        last_subpools[subpool[\"nuclei\"]] = max(\n",
    "            last_subpools.get(subpool[\"nuclei\"]), \n",
    "            SubpoolName(subpool[\"subpool_subname\"]))\n",
    "\n",
    "        record = models.Subpool(\n",
    "            name=subpool[\"name\"],\n",
    "            plate=plate,\n",
    "            nuclei=subpool[\"nuclei\"],\n",
    "            selection_type=subpool[\"selection_type\"],\n",
    "            cdna_pcr_rounds=subpool[\"cdna_pcr_rounds\"],\n",
    "            cdna_ng_per_ul=float_or_none(subpool[\"cdna_ng_per_ul\"]),\n",
    "            cdna_volume=float_or_none(subpool[\"cdna_volume\"]),\n",
    "            bioanalyzer_date=date_or_none(subpool[\"bioanalyzer_date\"]),\n",
    "            cdna_average_bp_length=int_or_none(subpool[\"cdna_average_bp_length\"]),\n",
    "            index_pcr_number=int_or_none(subpool[\"index_pcr_no\"]),\n",
    "            index=int_or_none(subpool[\"index\"]),\n",
    "            library_ng_per_ul=float_or_none(subpool[\"library_ng_per_ul\"]),\n",
    "            library_average_bp_length=int_or_none(subpool[\"library_average_bp_length\"]),\n",
    "        )\n",
    "        record.save()\n",
    "        record.barcode.set(barcodes)\n",
    "        record.save()\n",
    "        \n",
    "        name = \"subpool_{}\".format(subpool[\"name\"].lower())\n",
    "        if name in multiplexed_samples:\n",
    "            for accession_row in multiplexed_samples[name]:\n",
    "                print(name, record.name, multiplexed_samples[name])\n",
    "                accession = models.Accession(**accession_row)\n",
    "                accession.save()\n",
    "                record.accession.add(accession)\n",
    "                record.save()            \n",
    "        \n",
    "    for nuclei in fixation[\"unused_aliquots\"]:\n",
    "        subpool_name = last_subpools[nuclei]\n",
    "        for i in range(fixation[\"unused_aliquots\"][nuclei]):\n",
    "            subpool_name = subpool_name.next_code()\n",
    "\n",
    "            record = models.Subpool(\n",
    "                name=\"{}_{}\".format(subpool_prefix, subpool_name),\n",
    "                plate=plate,\n",
    "                nuclei=nuclei,\n",
    "            )\n",
    "            print(record.name)\n",
    "            record.save()\n",
    "\n",
    "assert not failed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "sudden-illustration",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Old creating runs from the sample sheet\n",
    "\n",
    "#        if experiment_name in nanopore_runs and subpool[\"nuclei\"] == 8000:\n",
    "#            measurement = get_or_create_sequencing_run(\n",
    "#                # nanopore_runs[experiment_name][\"name\"]\n",
    "#                name=\"{}/{}\".format(plate.name.lower(), \"nanopore\"),\n",
    "#                run_date=nanopore_runs[experiment_name][\"run_date\"],\n",
    "#                platform=platforms[\"nanopore\"],\n",
    "#                plate=plate,\n",
    "#            )\n",
    "#            \n",
    "#            pattern = expected_names[experiment_name][\"nanopore\"].format(\n",
    "#                library_name=subpool[\"subpool_subname\"],\n",
    "#                library_index=line_offset+1,\n",
    "#                compression=compression,\n",
    "#            )\n",
    "#            \n",
    "#            run = models.SubpoolInRun(\n",
    "#                subpool=record,\n",
    "#                sequencing_run=measurement,\n",
    "#                raw_reads=nanopore_runs[experiment_name][\"raw_reads\"],\n",
    "#                status=models.RunStatus.PASS,\n",
    "#            )\n",
    "#            run.save()            \n",
    "#\n",
    "#        if subpool[\"nextseq_run_date\"] is not None:\n",
    "#            measurement = get_or_create_sequencing_run(\n",
    "#                #sublibrary[\"nextseq_run_id\"]\n",
    "#                name=\"{}/{}\".format(plate.name.lower(), \"nextseq\"),\n",
    "#                run_date=subpool[\"nextseq_run_date\"],\n",
    "#                platform=platforms[\"nextseq\"],\n",
    "#                plate=plate,\n",
    "#            )\n",
    "#\n",
    "#            pattern = expected_names[experiment_name][\"nextseq\"].format(\n",
    "#                library_name=subpool[\"subpool_subname\"],\n",
    "#                library_index=line_offset+1,\n",
    "#                compression=compression,\n",
    "#            )            \n",
    "#            \n",
    "#            run = models.SubpoolInRun(\n",
    "#                subpool=record,\n",
    "#                sequencing_run=measurement,\n",
    "#                raw_reads=subpool[\"nextseq_run_raw_reads\"],\n",
    "#                status=models.RunStatus.PASS,\n",
    "#            )\n",
    "#            run.save()\n",
    "#            \n",
    "#        if subpool[\"novaseq_run_date\"] is not None:\n",
    "#            for i, nova_raw_reads in enumerate([subpool[\"novaseq1_raw_reads\"], subpool[\"novaseq2_raw_reads\"]]):\n",
    "#                #nova_run_name = \"{}-{}\".format(sublibrary[\"novaseq_run_id\"], i+1)\n",
    "#                nova_run_name = \"{}/{}\".format(plate.name.lower(), \"nova{}\".format(i+1))\n",
    "#                measurement = get_or_create_sequencing_run(\n",
    "#                    name=nova_run_name,\n",
    "#                    run_date=subpool[\"novaseq_run_date\"],\n",
    "#                    platform=platforms[\"novaseq\"],\n",
    "#                    plate=plate,\n",
    "#                )\n",
    "#\n",
    "#                pattern = expected_names[experiment_name][\"novaseq\"].format(\n",
    "#                    library_name=subpool[\"subpool_subname\"],\n",
    "#                    library_index=line_offset+1,\n",
    "#                    compression=compression,\n",
    "#                )\n",
    "#\n",
    "#                run = models.SubpoolInRun(\n",
    "#                    subpool=record,\n",
    "#                    sequencing_run=measurement,\n",
    "#                    raw_reads=nova_raw_reads,\n",
    "#                    status=models.RunStatus.PASS,                    \n",
    "#                )\n",
    "#                run.save()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "choice-error",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>experiment</th>\n",
       "      <th>filename</th>\n",
       "      <th>md5sum</th>\n",
       "      <th>size</th>\n",
       "      <th>ctime</th>\n",
       "      <th>library_id</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>lane_id</th>\n",
       "      <th>read_id</th>\n",
       "      <th>fragment</th>\n",
       "      <th>compression</th>\n",
       "      <th>sequencer</th>\n",
       "      <th>sequencer_run</th>\n",
       "      <th>flowcell_id</th>\n",
       "      <th>lane</th>\n",
       "      <th>barcode</th>\n",
       "      <th>plate_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1587</th>\n",
       "      <td>igvf_003</td>\n",
       "      <td>igvf_003/nextseq/003_8A_R1.fastq.gz</td>\n",
       "      <td>bb696e63b6a8710ec8435b60ef6c81f8</td>\n",
       "      <td>8886302014</td>\n",
       "      <td>2023-02-07</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gz</td>\n",
       "      <td>VH00582</td>\n",
       "      <td>1</td>\n",
       "      <td>AAATMGFHV</td>\n",
       "      <td>1</td>\n",
       "      <td>CAGATC</td>\n",
       "      <td>003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1592</th>\n",
       "      <td>igvf_003</td>\n",
       "      <td>igvf_003/nextseq/003_8A_R2.fastq.gz</td>\n",
       "      <td>21967ba1a5a3218508961fac482a2952</td>\n",
       "      <td>5866415324</td>\n",
       "      <td>2023-02-07</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gz</td>\n",
       "      <td>VH00582</td>\n",
       "      <td>1</td>\n",
       "      <td>AAATMGFHV</td>\n",
       "      <td>1</td>\n",
       "      <td>CAGATC</td>\n",
       "      <td>003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1576</th>\n",
       "      <td>igvf_003</td>\n",
       "      <td>igvf_003/nanopore/igvf003_8A_lig-ss_11.fastq.gz</td>\n",
       "      <td>494bd18e8eb9bcef5b79f18f8f21d569</td>\n",
       "      <td>7241241549</td>\n",
       "      <td>2022-12-13</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "      <td>gz</td>\n",
       "      <td>nanopore</td>\n",
       "      <td>e0fe4672609d60d67277a22973f5c8534df5fc4c</td>\n",
       "      <td>FAU06484</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1577</th>\n",
       "      <td>igvf_003</td>\n",
       "      <td>igvf_003/nanopore/igvf003_8A_lig-ss_4.fastq.gz</td>\n",
       "      <td>af524063452a305fa81585f760ad0cd4</td>\n",
       "      <td>1241893857</td>\n",
       "      <td>2022-12-07</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>gz</td>\n",
       "      <td>nanopore</td>\n",
       "      <td>1ab5f6b09adc4af2c1c9a1f93c7a752037dbdfce</td>\n",
       "      <td>FAU11624</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1578</th>\n",
       "      <td>igvf_003</td>\n",
       "      <td>igvf_003/nanopore/igvf003_8A_lig-ss_7.fastq.gz</td>\n",
       "      <td>0dbdc8a7fe445f8a9e6bbbd7a7d42cfd</td>\n",
       "      <td>7602537389</td>\n",
       "      <td>2022-12-08</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7</td>\n",
       "      <td>gz</td>\n",
       "      <td>nanopore</td>\n",
       "      <td>c3ebdb16d712e04e2ff2f2ea4e7b83be4de41c77</td>\n",
       "      <td>FAU06496</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>003</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     experiment                                         filename  \\\n",
       "1587   igvf_003              igvf_003/nextseq/003_8A_R1.fastq.gz   \n",
       "1592   igvf_003              igvf_003/nextseq/003_8A_R2.fastq.gz   \n",
       "1576   igvf_003  igvf_003/nanopore/igvf003_8A_lig-ss_11.fastq.gz   \n",
       "1577   igvf_003   igvf_003/nanopore/igvf003_8A_lig-ss_4.fastq.gz   \n",
       "1578   igvf_003   igvf_003/nanopore/igvf003_8A_lig-ss_7.fastq.gz   \n",
       "\n",
       "                                md5sum        size       ctime  library_id  \\\n",
       "1587  bb696e63b6a8710ec8435b60ef6c81f8  8886302014  2023-02-07           1   \n",
       "1592  21967ba1a5a3218508961fac482a2952  5866415324  2023-02-07           1   \n",
       "1576  494bd18e8eb9bcef5b79f18f8f21d569  7241241549  2022-12-13           1   \n",
       "1577  af524063452a305fa81585f760ad0cd4  1241893857  2022-12-07           1   \n",
       "1578  0dbdc8a7fe445f8a9e6bbbd7a7d42cfd  7602537389  2022-12-08           1   \n",
       "\n",
       "     sample_id lane_id read_id fragment compression sequencer  \\\n",
       "1587       NaN     NaN      R1      NaN          gz   VH00582   \n",
       "1592       NaN     NaN      R2      NaN          gz   VH00582   \n",
       "1576       NaN     NaN     NaN       11          gz  nanopore   \n",
       "1577       NaN     NaN     NaN        4          gz  nanopore   \n",
       "1578       NaN     NaN     NaN        7          gz  nanopore   \n",
       "\n",
       "                                 sequencer_run flowcell_id lane barcode  \\\n",
       "1587                                         1   AAATMGFHV    1  CAGATC   \n",
       "1592                                         1   AAATMGFHV    1  CAGATC   \n",
       "1576  e0fe4672609d60d67277a22973f5c8534df5fc4c    FAU06484  NaN     NaN   \n",
       "1577  1ab5f6b09adc4af2c1c9a1f93c7a752037dbdfce    FAU11624  NaN     NaN   \n",
       "1578  c3ebdb16d712e04e2ff2f2ea4e7b83be4de41c77    FAU06496  NaN     NaN   \n",
       "\n",
       "     plate_id  \n",
       "1587      003  \n",
       "1592      003  \n",
       "1576      003  \n",
       "1577      003  \n",
       "1578      003  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fastqs = pandas.read_csv(\n",
    "    \"fastq_metadata.tsv\", sep=\"\\t\",\n",
    "    dtype={\n",
    "        \"sample_id\": str,\n",
    "        \"read_id\": str,\n",
    "        \"fragment\": str,\n",
    "        \"sequencer_run\": str,\n",
    "        \"lane\": str,\n",
    "    }).sort_values([\"experiment\", \"library_id\", \"lane_id\", \"read_id\"])\n",
    "fastqs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "parental-violin",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<SplitSeqPlate: IGVF_003>"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "models.SplitSeqPlate.objects.get(name=\"IGVF_003\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "motivated-sister",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CCGTCC'"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "models.Subpool.objects.get(plate__name=\"IGVF_003\", index=10).barcode.get().sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "national-princeton",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<QuerySet [<Subpool: B01_13A>, <Subpool: B01_13B>, <Subpool: B01_13C>, <Subpool: B01_13D>, <Subpool: B01_13E>, <Subpool: B01_13F>, <Subpool: B01_13G>, <Subpool: B01_13H>, <Subpool: B01_13I>, <Subpool: B01_13J>, <Subpool: B01_13K>, <Subpool: B01_13L>, <Subpool: B01_13M>, <Subpool: B01_13N>, <Subpool: B01_13O>, <Subpool: B01_13P>, <Subpool: B01_13Q>, <Subpool: B01_13R>]>"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "models.Subpool.objects.filter(plate__name=\"IGVF_B01\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "republican-strategy",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Barcode mismatch IGVF_003 1 CAGATC nan igvf_003/nanopore/igvf003_8A_lig-ss_11.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC nan igvf_003/nanopore/igvf003_8A_lig-ss_4.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC nan igvf_003/nanopore/igvf003_8A_lig-ss_7.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC nan igvf_003/nanopore/igvf003_8A_lig-ss_6.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC nan igvf_003/nanopore/igvf003_8A_lig-ss_8.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC nan igvf_003/nanopore/igvf003_8A_lig-ss_5.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC nan igvf_003/nanopore/igvf003_8A_lig-ss_1.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC nan igvf_003/nanopore/igvf003_8A_lig-ss_2.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC nan igvf_003/nanopore/igvf003_8A_lig-ss_10.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC nan igvf_003/nanopore/igvf003_8A_lig-ss_9.fastq.gz\n",
      "Barcode mismatch IGVF_003 1 CAGATC nan igvf_003/nanopore/igvf003_8A_lig-ss_3.fastq.gz\n"
     ]
    },
    {
     "ename": "DoesNotExist",
     "evalue": "Subpool matching query does not exist.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mDoesNotExist\u001b[0m                              Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-56-1b004ef68a89>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0mlibrary_id\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlibrary_id\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m     \u001b[0msubpool\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmodels\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSubpool\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobjects\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mplate__name\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mplate_name\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlibrary_id\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m     \u001b[0;32massert\u001b[0m \u001b[0msubpool\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbarcode\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcount\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"{} {}\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msubpool\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlibrary_id\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3/dist-packages/django/db/models/manager.py\u001b[0m in \u001b[0;36mmanager_method\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m     83\u001b[0m         \u001b[0;32mdef\u001b[0m \u001b[0mcreate_method\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     84\u001b[0m             \u001b[0;32mdef\u001b[0m \u001b[0mmanager_method\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 85\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_queryset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     86\u001b[0m             \u001b[0mmanager_method\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     87\u001b[0m             \u001b[0mmanager_method\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__doc__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__doc__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3/dist-packages/django/db/models/query.py\u001b[0m in \u001b[0;36mget\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    433\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mclone\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_result_cache\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    434\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mnum\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 435\u001b[0;31m             raise self.model.DoesNotExist(\n\u001b[0m\u001b[1;32m    436\u001b[0m                 \u001b[0;34m\"%s matching query does not exist.\"\u001b[0m \u001b[0;34m%\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    437\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_meta\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobject_name\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mDoesNotExist\u001b[0m: Subpool matching query does not exist."
     ]
    }
   ],
   "source": [
    "for i, row in fastqs.iterrows():\n",
    "    plate_name = row.experiment.upper()\n",
    "    library_id = int(row.library_id)\n",
    "\n",
    "    subpool = models.Subpool.objects.get(plate__name=plate_name, index=library_id)\n",
    "    assert subpool.barcode.count() == 1, \"{} {}\".format(subpool, library_id)\n",
    "\n",
    "    barcode = subpool.barcode.get()\n",
    "\n",
    "    if barcode.sequence != row.barcode:\n",
    "        print(f\"Barcode mismatch {plate_name} {library_id} {barcode.sequence} {row.barcode} {row.filename}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "gross-cooler",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> /usr/lib/python3/dist-packages/django/db/models/query.py(435)get()\n",
      "    433             return clone._result_cache[0]\n",
      "    434         if not num:\n",
      "--> 435             raise self.model.DoesNotExist(\n",
      "    436                 \"%s matching query does not exist.\" %\n",
      "    437                 self.model._meta.object_name\n",
      "\n",
      "ipdb> up\n",
      "> /usr/lib/python3/dist-packages/django/db/models/manager.py(85)manager_method()\n",
      "     83         def create_method(name, method):\n",
      "     84             def manager_method(self, *args, **kwargs):\n",
      "---> 85                 return getattr(self.get_queryset(), name)(*args, **kwargs)\n",
      "     86             manager_method.__name__ = method.__name__\n",
      "     87             manager_method.__doc__ = method.__doc__\n",
      "\n",
      "ipdb> up\n",
      "> <ipython-input-56-1b004ef68a89>(5)<module>()\n",
      "      3     library_id = int(row.library_id)\n",
      "      4 \n",
      "----> 5     subpool = models.Subpool.objects.get(plate__name=plate_name, index=library_id)\n",
      "      6     assert subpool.barcode.count() == 1, \"{} {}\".format(subpool, library_id)\n",
      "      7 \n",
      "\n",
      "ipdb> p plate_name, library_id\n",
      "('IGVF_012', 1)\n",
      "ipdb> p models.SplitSeqPlate.objects.all()\n",
      "<QuerySet [<SplitSeqPlate: IGVF_003>, <SplitSeqPlate: IGVF_004>, <SplitSeqPlate: IGVF_005>, <SplitSeqPlate: IGVF_006>, <SplitSeqPlate: IGVF_007>, <SplitSeqPlate: IGVF_008>, <SplitSeqPlate: IGVF_B01>, <SplitSeqPlate: IGVF_009>, <SplitSeqPlate: IGVF_010>, <SplitSeqPlate: IGVF_011>]>\n"
     ]
    }
   ],
   "source": [
    "%debug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adjustable-fifty",
   "metadata": {},
   "outputs": [],
   "source": [
    "plate_name, library_id"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "distant-atmosphere",
   "metadata": {},
   "source": [
    "Load information about previously submitted records"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "imported-placement",
   "metadata": {},
   "outputs": [],
   "source": [
    "submitted_sequence_file = {}\n",
    "\n",
    "for accession_prefix in submitted_sheets:\n",
    "    for i, row in submitted_sheets[\"sequence_file\"].iterrows():\n",
    "        if not pandas.isnull(row[\"accession\"]):\n",
    "            if access_prefix == \"igvftst\":\n",
    "                see_also_template = \"https://api.sandbox.igvf.org/sequence-files/{}/\"\n",
    "            elif access_prefix == \"igvf\":\n",
    "                see_also_template = \"https://api.data.igvf.org/sequence-files/{}/\"\n",
    "            \n",
    "            submitted_sequence_file.setdefault(row[\"submitted_file_name\"], []).append({\n",
    "                \"accession_prefix\": accession_prefix,\n",
    "                \"name\": row[\"accession\"],\n",
    "                \"uuid\": row[\"uuid\"],\n",
    "                \"see_also\": see_also_template.format(row[\"accession\"]),\n",
    "            })\n",
    "        \n",
    "print(\"loaded {} submitted sequence_file records\".format(len(submitted_sequence_file)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "numerous-marketplace",
   "metadata": {},
   "outputs": [],
   "source": [
    "submitted_sheets.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "harmful-builder",
   "metadata": {},
   "outputs": [],
   "source": [
    "def built_in_cache(a, b, cache={}):\n",
    "    cache.setdefault(a, b)\n",
    "    print(cache)\n",
    "    \n",
    "built_in_cache('a', 1)\n",
    "built_in_cache('b', 3)\n",
    "built_in_cache('c', 4)\n",
    "built_in_cache('a', 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "second-hopkins",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "supreme-projector",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequencing_experiments = {}\n",
    "sequencing_subpools = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "armed-education",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def get_or_create_sequencing_run(name, run_date, platform, plate, accessions=None):\n",
    "    if name in sequencing_experiments:\n",
    "        return sequencing_experiments[name]\n",
    "    \n",
    "    try:\n",
    "        sequencing_run = models.SequencingRun.objects.get(name=name)\n",
    "    except models.SequencingRun.DoesNotExist:\n",
    "        sequencing_run = models.SequencingRun(\n",
    "            name=name,\n",
    "            run_date=run_date,\n",
    "            platform=platform,\n",
    "            plate=plate,\n",
    "            stranded=models.Stranded.REVERSE,\n",
    "        )\n",
    "        sequencing_run.save()\n",
    "        if accessions is not None:\n",
    "            sequencing_run.accession.set(accessions)\n",
    "            sequencing_run.save\n",
    "        \n",
    "    sequencing_experiments[name] = sequencing_run\n",
    "    return sequencing_run\n",
    "\n",
    "def get_or_create_subpool_in_run(subpool, sequencing_run, raw_reads=None, measurement_set=None):\n",
    "    try:\n",
    "        subpool_in_run = models.SubpoolInRun.objects.get(subpool=subpool, sequencing_run=sequencing_run)\n",
    "    except models.SubpoolInRun.DoesNotExist:\n",
    "        subpool_in_run = models.SubpoolInRun(\n",
    "            subpool=subpool, \n",
    "            sequencing_run=sequencing_run,\n",
    "            raw_reads=None,\n",
    "            status=models.RunStatus.PASS,\n",
    "            measurement_set=measurement_set,\n",
    "        )\n",
    "        subpool_in_run.save()\n",
    "        \n",
    "    sequencing_subpools[name] = subpool_in_run\n",
    "    return subpool_in_run\n",
    "\n",
    "\n",
    "def lane_or_none(value):\n",
    "    if pandas.isnull(value):\n",
    "        return None\n",
    "    elif isinstance(value, float):\n",
    "        return int(value)\n",
    "    elif value.startswith(\"L\"):\n",
    "        return int(value[1:])\n",
    "    else:\n",
    "        return int(value)\n",
    "\n",
    "\n",
    "fastqs = pandas.read_csv(\n",
    "    \"fastq_metadata.tsv\", sep=\"\\t\",\n",
    "    dtype={\n",
    "        \"sample_id\": str,\n",
    "        \"read_id\": str,\n",
    "        \"fragment\": str,\n",
    "        \"sequencer_run\": str,\n",
    "        \"lane\": str,\n",
    "    }).sort_values([\"experiment\", \"library_id\", \"lane_id\", \"read_id\"])\n",
    "fastqs.head()\n",
    "\n",
    "for i, row in fastqs.iterrows():\n",
    "    plate_name = row.experiment.upper()\n",
    "    library_id = int(row.library_id)\n",
    "    fastq_relative = Path(row.filename)\n",
    "    sequencing_run_name = str(Path(fastq_relative.parts[0])/fastq_relative.parts[1])\n",
    "    sequencer = row.get(\"sequencer\")\n",
    "    run_date = row[\"ctime\"]\n",
    "    md5sum = row[\"md5sum\"]\n",
    "\n",
    "    if row[\"read_id\"] in ('I1',):\n",
    "        continue\n",
    "\n",
    "    subpool = models.Subpool.objects.get(plate__name=plate_name, index=library_id)\n",
    "    plate = subpool.plate\n",
    "\n",
    "    platform = {\n",
    "        \"A00850\": models.Platform.objects.get(display_name=\"Novaseq 6000\"),\n",
    "        \"VH00582\": models.Platform.objects.get(display_name=\"Nextseq 2000\"),\n",
    "        \"nanopore\": models.Platform.objects.get(display_name=\"Oxford Nanopore\"),  # TODO: figure out a machine id\n",
    "    }[sequencer]\n",
    "\n",
    "    sequencing_run = get_or_create_sequencing_run(sequencing_run_name, run_date, platform, plate, accessions=None)\n",
    "    subpool_run = get_or_create_subpool_in_run(subpool=subpool, sequencing_run=sequencing_run)\n",
    "    \n",
    "    file_set_name = None\n",
    "    try:\n",
    "        file_set_alias = submitted_sheets[\"sequence_file\"].set_index(\"md5sum\").loc[row.md5sum, \"file_set\"]\n",
    "        file_set_name = file_set_alias.split(\":\")[1]\n",
    "    except KeyError:\n",
    "        \n",
    "        print(f\"Unable to find {row.md5sum}\")\n",
    "\n",
    "    if file_set_name is None:\n",
    "        pass\n",
    "    elif subpool_run.measurement_set is None:\n",
    "        subpool_run.measurement_set = measurement_sets[file_set_name]\n",
    "        subpool_run.save()\n",
    "        print(f\"updating {subpool_run} measurement set to {measurement_sets[file_set_name]}\")\n",
    "    else:\n",
    "        assert subpool_run.measurement_set.name == file_set_name, f\"SubpoolInRun measurement_set mismatch {subpool_run.measurement_set.name} {file_set_name}\"\n",
    "        \n",
    "\n",
    "    accession_info = submitted_sequence_file.get(str(fastq_relative))\n",
    "    subpool_file = models.SubpoolInRunFile(\n",
    "        sequencing_run=sequencing_run,\n",
    "        subpool_run=subpool_run,\n",
    "        md5sum=md5sum,\n",
    "        filename=str(fastq_relative),\n",
    "        flowcell_id=row[\"flowcell_id\"],\n",
    "        lane=lane_or_none(row[\"lane\"]),\n",
    "    )\n",
    "    if not pandas.isnull(row[\"read_id\"]):\n",
    "        subpool_file.read = row[\"read_id\"]\n",
    "    subpool_file.save()\n",
    "\n",
    "    if accession_info is not None:\n",
    "        accession = models.Accession(\n",
    "            namespace=accession_info[\"namespace\"],\n",
    "            name=accession_info[\"name\"],\n",
    "            uuid=accession_info[\"uuid\"],\n",
    "            see_also=accession_info[\"see_also\"],\n",
    "        )\n",
    "        accession.save()                \n",
    "        subpool_file.accession.set([accession])\n",
    "        subpool_file.save()\n",
    "\n",
    "\n",
    "        \n",
    "print(\"total plates\", models.SplitSeqPlate.objects.count())\n",
    "print(\"total sequencing runs\", models.SequencingRun.objects.count())\n",
    "print(\"total subpools\", models.Subpool.objects.count())\n",
    "print(\"total subpool in run\", models.SubpoolInRun.objects.count())\n",
    "print(\"total subpool in run files\", models.SubpoolInRunFile.objects.count())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "revolutionary-hearts",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "preliminary-washington",
   "metadata": {},
   "outputs": [],
   "source": [
    "models.Mouse.objects.count()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
