{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "satisfied-traffic",
   "metadata": {},
   "source": [
    "# Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "nearby-variation",
   "metadata": {},
   "source": [
    "UCI has a complex google doc they've been using to track their samples so far.\n",
    "\n",
    "https://docs.google.com/spreadsheets/d/13M6-Ry6oXgkx94BHZOGioYPI6F_hWDqjGgcaNu2JNYs/edit#gid=1838362486\n",
    "\n",
    "Also there's a metadata spreadsheet for IGVF at\n",
    "https://docs.google.com/spreadsheets/d/1BLMledzmqOqXnJHzpijgw91IOs-9tSlVeZDG_MtXddk/edit#gid=1284120531"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "increased-bottle",
   "metadata": {},
   "outputs": [],
   "source": [
    "import bz2\n",
    "from collections import Counter, namedtuple\n",
    "import datetime\n",
    "import numpy\n",
    "import os\n",
    "import pandas\n",
    "from pathlib import Path\n",
    "import re\n",
    "from subprocess import run, PIPE\n",
    "import sys\n",
    "import zoneinfo\n",
    "\n",
    "os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'mousedemo.settings')\n",
    "os.environ[\"DJANGO_ALLOW_ASYNC_UNSAFE\"] = \"true\"\n",
    "\n",
    "import django\n",
    "from django.contrib.auth import get_user_model\n",
    "from django.db import DEFAULT_DB_ALIAS\n",
    "\n",
    "MOUSEDEMO = str(Path(\"mousedemo\").absolute())\n",
    "if MOUSEDEMO not in sys.path:\n",
    "    sys.path.append(MOUSEDEMO)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "divided-checkout",
   "metadata": {},
   "source": [
    "# Pre-initialization setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "orange-optics",
   "metadata": {},
   "source": [
    "Backup the old ones and make a new database file before we initialize the database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "effective-palace",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a new database\n",
    "result = run([\"python3\", \"manage.py\", \"check\"], capture_output=True, cwd=MOUSEDEMO)\n",
    "\n",
    "if len(result.stderr) > 0:\n",
    "    print(result.stderr.decode(\"utf-8\"))\n",
    "    \n",
    "result.check_returncode()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "documented-cross",
   "metadata": {},
   "outputs": [],
   "source": [
    "initial = Path(\"igvf_mice/migrations/0001_initial.py\")\n",
    "if initial.exists():\n",
    "    initial.unlink()\n",
    "    \n",
    "result = run([\"python3\", \"manage.py\", \"makemigrations\", \"igvf_mice\"], capture_output=True, cwd=MOUSEDEMO)\n",
    "\n",
    "if len(result.stderr) > 0:\n",
    "    print(result.stderr.decode(\"utf-8\"))\n",
    "    \n",
    "result.check_returncode()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "magnetic-soccer",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "renaming mousedemo/db3.sqlite3 to mousedemo/db4.sqlite3\n",
      "renaming mousedemo/db2.sqlite3 to mousedemo/db3.sqlite3\n",
      "renaming mousedemo/db1.sqlite3 to mousedemo/db2.sqlite3\n",
      "renaming mousedemo/db.sqlite3 to mousedemo/db1.sqlite3\n"
     ]
    }
   ],
   "source": [
    "# Backup several old sqlite database files\n",
    "\n",
    "db_name = Path(\"db.sqlite3\")\n",
    "\n",
    "backup_names = []\n",
    "for i in range(0, 4):\n",
    "    backup_names.append(Path(\"db{}.sqlite3\".format(i+1)))\n",
    "\n",
    "if backup_names[-1].exists():\n",
    "    backup_names[-1].unlink()\n",
    "\n",
    "if db_name.exists():    \n",
    "    for i in reversed(range(0, 3)):\n",
    "        if backup_names[i].exists():\n",
    "            print(\"renaming {} to {}\".format(backup_names[i], backup_names[i+1]))\n",
    "            backup_names[i].rename(backup_names[i+1])\n",
    "\n",
    "    db_name.rename(backup_names[0])\n",
    "    print(\"renaming {} to {}\".format(db_name, backup_names[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "organizational-resort",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a new database\n",
    "result = run([\"python3\", \"manage.py\", \"migrate\"], capture_output=True, cwd=MOUSEDEMO)\n",
    "\n",
    "if len(result.stderr) > 0:\n",
    "    print(result.stderr.decode(\"utf-8\"))\n",
    "    \n",
    "result.check_returncode()\n",
    "    \n",
    "assert Path(\"db.sqlite3\").exists()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "blessed-shareware",
   "metadata": {},
   "source": [
    "# Setup users"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "failing-advantage",
   "metadata": {},
   "source": [
    "Now that we have a fresh clean database, lets create user accounts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "aboriginal-reform",
   "metadata": {},
   "outputs": [],
   "source": [
    "django.setup()\n",
    "\n",
    "from mousedemo import settings\n",
    "from igvf_mice import models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "arranged-restaurant",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create accounts.\n",
    "with open(\"pw\", \"rt\") as instream:\n",
    "    pw = instream.read().strip()\n",
    "\n",
    "user_model = get_user_model()\n",
    "user_model._default_manager.db_manager(DEFAULT_DB_ALIAS).create_superuser(\n",
    "    username=\"diane\",\n",
    "    password=pw,   \n",
    ")\n",
    "pw = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "realistic-ocean",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "conservative-recycling",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "canadian-slide",
   "metadata": {},
   "outputs": [],
   "source": [
    "#spreadsheet_name = \"IGVF_Split-seq_20230118.xlsx\"\n",
    "spreadsheet_name = \"https://woldlab.caltech.edu/nextcloud/index.php/s/eEtjBfDqQFnLpSS/download\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "hundred-bookmark",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Planning\n",
      "mice - 8 founders\n",
      "Line information\n",
      "schedule - all\n",
      "Nuclei isolation Schedule with \n",
      "Counting\n",
      "Samples - 8 founders\n",
      "Bridge Samples\n",
      "Samples into experiment\n",
      "Experiment\n",
      "Parvin practice Samples into ex\n",
      "DCC\n",
      "Plate setups\n",
      "PBMCs\n",
      "practice mice\n",
      "IGVF002 Pilot Samples\n",
      "IGVF002 Pulverized Samples\n",
      "IGVF002 Plate Setup\n",
      "IGVF002 Metadata\n",
      "Cerebellum testing\n"
     ]
    }
   ],
   "source": [
    "book = pandas.ExcelFile(spreadsheet_name)\n",
    "for name in book.sheet_names:\n",
    "    print(name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "split-payday",
   "metadata": {},
   "source": [
    "Converters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "occupational-dealing",
   "metadata": {},
   "outputs": [],
   "source": [
    "def truncate(model):\n",
    "    table_name = model._meta.db_table\n",
    "    assert \"\\\\\" not in table_name\n",
    "    with django.db.connection.cursor() as cursor:\n",
    "        # I don't know why the sql params didn't work. this is a sql vulnerability waiting to happen\n",
    "        cursor.execute(\"delete from \\\"{}\\\"\".format(table_name))\n",
    "        cursor.execute(\"DELETE FROM SQLITE_SEQUENCE WHERE name=\\\"{}\\\"\".format(table_name))\n",
    "        cursor.fetchone()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "divine-incident",
   "metadata": {},
   "outputs": [],
   "source": [
    "def int_or_none(x):\n",
    "    if x == \"N/A\":\n",
    "        return None\n",
    "    elif pandas.isnull(x):\n",
    "        return None\n",
    "    else:\n",
    "        return int(x)\n",
    "    \n",
    "def int_or_0(x):\n",
    "    if pandas.isnull(x):\n",
    "        return 0\n",
    "    else:\n",
    "        return int(x)\n",
    "\n",
    "def float_or_none(x):\n",
    "    if pandas.isnull(x):\n",
    "        return None\n",
    "    else:\n",
    "        return float(x)\n",
    "    \n",
    "def str_or_empty(x):\n",
    "    if pandas.isnull(x):\n",
    "        return \"\"\n",
    "    else:\n",
    "        return x\n",
    "\n",
    "def str_or_none(x):\n",
    "    if pandas.isnull(x):\n",
    "        return None\n",
    "    else:\n",
    "        return x\n",
    "    \n",
    "def date_or_none(x):\n",
    "    if pandas.isnull(x):\n",
    "        return None\n",
    "    elif isinstance(x, datetime.datetime):\n",
    "        return x.date()\n",
    "    else:\n",
    "        return x\n",
    "\n",
    "def datetime_or_none(x):\n",
    "    if pandas.isnull(x):\n",
    "        return None\n",
    "    else:\n",
    "        return x\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "noble-personal",
   "metadata": {},
   "source": [
    "# initialize general use tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "psychological-march",
   "metadata": {},
   "outputs": [],
   "source": [
    "source = [\n",
    "    {\"name\": \"The Jackson Laboratory\", \"url\": \"http://www.jax.org/index.html\", \"igvf_id\": \"/sources/jackson-labs\"}\n",
    "    {\"name\": \"Parse Bioscience\", \"url\": \"https://www.parsebiosciences.com/\", \"igvf_id\":},\n",
    "    {\"name\": \"Illumina\", \"url\": \"https://www.illumina.com\", \"igvf_id\": \"/sources/illumina/\"},\n",
    "]\n",
    "\n",
    "if models.Source.objects.count() > 0:\n",
    "    truncate(models.Source)\n",
    "\n",
    "for row in source:\n",
    "    record = models.Source(\n",
    "        name=row[\"name\"],\n",
    "        url=row[\"url\"])\n",
    "    record.save()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "manufactured-account",
   "metadata": {},
   "source": [
    "# Library kit type & version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "innocent-copying",
   "metadata": {},
   "outputs": [],
   "source": [
    "library_construction_kit = [\n",
    "    {\"name\": \"WT Mega\", \"version\": \"v2\", \"source\": \"Parse Bioscience\"},\n",
    "    #{\"name\": \"\", \"version\": \"\", \"source\": , \"Illumina\"},\n",
    "]\n",
    "\n",
    "if models.LibraryConstructionKit.objects.count() > 0:\n",
    "    truncate(models.LibraryConstructionKit)\n",
    "\n",
    "for row in library_construction_kit:\n",
    "    try:\n",
    "        source = models.Source.objects.get(name=row[\"source\"])\n",
    "        \n",
    "        record = models.LibraryConstructionKit(\n",
    "            name=row[\"name\"],\n",
    "            version=row[\"version\"],\n",
    "            source=source,\n",
    "        )\n",
    "        record.save()        \n",
    "    except models.Source.DoesNotExist:\n",
    "        print(f\"Couldn't find {row['source']}\")\n",
    "        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "pressed-payday",
   "metadata": {},
   "source": [
    "# Load LibraryBarcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "convenient-syndication",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<LibraryConstructionKit: WT Mega v2>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "models.LibraryConstructionKit.objects.get(name=\"WT Mega\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "approved-hours",
   "metadata": {},
   "outputs": [],
   "source": [
    "if models.LibraryBarcode.objects.count() > 0:\n",
    "    truncate(models.LibraryBarcode)\n",
    "\n",
    "\n",
    "# Load WT Mega v2 first barcode\n",
    "bc1 = pandas.read_csv(\"bc_data_n192_v4.csv\")\n",
    "\n",
    "kit = models.LibraryConstructionKit.objects.get(name=\"WT Mega\", version=\"v2\")\n",
    "\n",
    "for i, row in bc1.iterrows():\n",
    "    record = models.LibraryBarcode(\n",
    "        kit=kit,\n",
    "        name=row[\"uid\"],\n",
    "        code=row[\"well\"],\n",
    "        sequence=row[\"sequence\"],\n",
    "        barcode_type=row[\"type\"], # TODO what do the codes mean again T,R?\n",
    "    )\n",
    "    record.save()\n",
    "\n",
    "sublibrary=[\n",
    "    (\"1\", \"CAGATC\"),\n",
    "    (\"2\", \"ACTTGA\"),\n",
    "    (\"3\", \"GATCAG\"),\n",
    "    (\"4\", \"TAGCTT\"),\n",
    "    (\"5\", \"ATGTCA\"),\n",
    "    (\"6\", \"CTTGTA\"),\n",
    "    (\"7\", \"AGTCAA\"),\n",
    "    (\"8\", \"AGTTCC\"),\n",
    "    (\"9\", \"GAGTGG\"),\n",
    "    (\"10\", \"CCGTCC\"),\n",
    "    (\"11\", \"GTAGAG\"),\n",
    "    (\"12\", \"GTCCGC\"),\n",
    "    (\"13\", \"GTGAAA\"),\n",
    "    (\"14\", \"GTGGCC\"),\n",
    "    (\"15\", \"GTTTCG\"),\n",
    "    (\"16\", \"CGTACG\"),\n",
    "]\n",
    "sublibrary = pandas.DataFrame(sublibrary, columns=[\"code\",\"sequence\"])\n",
    "\n",
    "for i, row in sublibrary.iterrows():\n",
    "    record = models.LibraryBarcode(\n",
    "        kit=kit,\n",
    "        code=row[\"code\"],\n",
    "        sequence=row[\"sequence\"],\n",
    "    )\n",
    "    record.save()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "hungry-guidance",
   "metadata": {},
   "source": [
    "# Initialize TissueOntology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "dimensional-paintball",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'Hypothalamus/Pituitary': 81,\n",
       "         'Cerebellum': 81,\n",
       "         'Cortex/Hippocampus left': 81,\n",
       "         'Cortex/Hippocampus right': 81,\n",
       "         'Liver': 81,\n",
       "         'Heart': 81,\n",
       "         'Lung': 81,\n",
       "         'Adrenal': 81,\n",
       "         'Kidney': 81,\n",
       "         'Gonads -1 Ovary': 41,\n",
       "         'Gonads -2 Oviduct': 41,\n",
       "         'Perigonadal fat': 81,\n",
       "         'Brown fat': 81,\n",
       "         'Soleus': 81,\n",
       "         'Plantaris': 81,\n",
       "         'Gastrocnemius': 81,\n",
       "         'TA': 81,\n",
       "         'EDL': 81,\n",
       "         'Tail': 81,\n",
       "         'PBMC - WBC': 81,\n",
       "         'Gonads -1 Testis': 40,\n",
       "         'Gonads -2 Epididymis': 40})"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tissue_column = pandas.read_excel(\n",
    "    spreadsheet_name, \n",
    "    \"Samples - 8 founders\",\n",
    "    usecols=[\"Tissue\"])\n",
    "tissue_column.head()\n",
    "\n",
    "tissues = Counter()\n",
    "for i, row in tissue_column.iterrows():\n",
    "    if not pandas.isnull(row[\"Tissue\"]):\n",
    "        tissues[row[\"Tissue\"]] += 1\n",
    "\n",
    "tissues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "green-orange",
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue_dissection_to_ontology_map = {\n",
    "    'Hypothalamus/Pituitary': [(\"UBERON:0001898\",\"hypothalamus\"), (\"UBERON:0000007\",\"pituitary gland\")],\n",
    "    'Cerebellum': [(\"UBERON:0002037\",\"cerebellum\")],\n",
    "    # Hippocampus might be:\n",
    "    #   Hippocampal formation UBERON:0002421 https://www.ebi.ac.uk/ols/ontologies/uberon/terms?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FUBERON_0002421\n",
    "    #   Layer of hippocampus UBERON:0002305 https://www.ebi.ac.uk/ols/ontologies/uberon/terms?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FUBERON_0002305\n",
    "    'Cortex/Hippocampus left': [(\"NTR:0000646\",\"left cerebral cortex\"), (\"NTR:0000750\", \"Hippocampal formation left\")],\n",
    "    'Cortex/Hippocampus right': [(\"NTR:0000647\",\"right cerebral cortex\"), (\"NTR:0000751\", \"Hippocampal formation right\")],\n",
    "    'Liver': [(\"UBERON:0002107\", \"liver\")],\n",
    "    'Heart': [(\"UBERON:0000948\",\"heart\")],\n",
    "    'Lung': [(\"UBERON:0002048\",\"lung\")],\n",
    "    'Adrenal': [(\"UBERON:0002369\",\"adrenal gland\")],\n",
    "    'Kidney': [(\"UBERON:0002113\",\"kidney\")],\n",
    "    'Gonads -1 Ovary': [(\"UBERON:0000992\",\"ovary\")],\n",
    "    'Gonads -2 Oviduct': [(\"UBERON:0000993\",\"oviduct\")],\n",
    "    'Perigonadal fat': [(\"UBERON:0003428\", \"gonadal fat pad\")], #\"is closer\" to dissection\n",
    "    'Brown fat': [(\"UBERON:0001348\",\"brown adipose tissue\")], \n",
    "    'Soleus': [(\"UBERON:0001389\",\"soleus muscle\")],\n",
    "    'Plantaris': [(\"UBERON:0011905\",\"plantaris\")],\n",
    "    'Gastrocnemius': [(\"UBERON:0001388\",\"gastrocnemius\")],\n",
    "    'TA': [(\"UBERON:0001385\",\"tibialis anterior\")],\n",
    "    'EDL': [(\"UBERON:0001386\",\"extensor digitorum longus\")],\n",
    "    'Tail': [(\"UBERON:0002415\",\"tail\")],\n",
    "    'PBMC - WBC': [(\"CL:2000001\", \"peripheral blood mononuclear cell\")],\n",
    "    'Gonads -1 Testis': [(\"UBERON:0000473\",\"testis\")],\n",
    "    'Gonads -2 Epididymis': [(\"UBERON:0001301\",\"epididymis\")],\n",
    "}\n",
    "\n",
    "term_details = pandas.read_csv(\"obo.tsv.bz2\", compression=\"bz2\", sep=\"\\t\", index_col=\"term_id\")\n",
    "\n",
    "if models.OntologyTerm.objects.count() > 0:\n",
    "    truncate(models.OntologyTerm)\n",
    "    \n",
    "for key in tissue_dissection_to_ontology_map:\n",
    "    for term_curie, term_name in tissue_dissection_to_ontology_map[key]:\n",
    "        if term_curie.startswith(\"NTR:\"):\n",
    "            description=None\n",
    "        else:\n",
    "            details = term_details.loc[term_curie]\n",
    "            description = details.description\n",
    "\n",
    "        record = models.OntologyTerm(\n",
    "            curie=term_curie,\n",
    "            name=term_name,\n",
    "            description=description\n",
    "        )\n",
    "        record.save()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "otherwise-runner",
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_urls = {\n",
    "    \"AJ\": \"http://www.informatics.jax.org/inbred_strains/mouse/docs/A.shtml\",\n",
    "    \"B6J\": \"http://www.informatics.jax.org/inbred_strains/mouse/docs/C57BL.shtml\",\n",
    "    \"129S1J\": \"http://www.informatics.jax.org/inbred_strains/mouse/docs/129.shtml\",\n",
    "    \"NODJ\": \"http://www.informatics.jax.org/inbred_strains/mouse/docs/NOD.shtml\",\n",
    "    \"NZOJ\": \"http://www.informatics.jax.org/inbred_strains/mouse/docs/NZO.shtml\",\n",
    "    \"CASTJ\": \"http://www.informatics.jax.org/inbred_strains/mouse/docs/CAST.shtml\",\n",
    "    \"PWKJ\": \"http://www.informatics.jax.org/inbred_strains/mouse/docs/PWK.shtml\",\n",
    "    \"WSBJ\": \"https://www.informatics.jax.org/strain/MGI:2160667\"\n",
    "}\n",
    "\n",
    "strain_background = {\n",
    "    \"AJ\": \"A/J (AJ)\",\n",
    "    \"B6J\": \"C57BL/6J (B6)\",\n",
    "    \"129S1J\": \"129S1/SvImJ (129)\",\n",
    "    \"NODJ\": \"NOD/ShiLtJ (NOD)\",\n",
    "    \"NZOJ\": \"NZO/H1LtJ (NZO)\",\n",
    "    \"CASTJ\": \"CAST/EiJ (CAST)\",\n",
    "    \"PWKJ\": \"PWK/PhJ (PWK)\",\n",
    "    \"WSBJ\": \"WSB/EiJ (WSB)\",\n",
    "    #\"CAST (M. m. castaneus)\",\n",
    "    #\"WSB (M. m. domesticus)\",\n",
    "    #\"PWK (M. m. musculus)\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "committed-dollar",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Designation</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Note</th>\n",
       "      <th>Jax Catalog No</th>\n",
       "      <th>Sample CODE</th>\n",
       "      <th>Strain notes</th>\n",
       "      <th>url</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>A/J</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>000646</td>\n",
       "      <td>AJ</td>\n",
       "      <td>Yellower adrenal gland.</td>\n",
       "      <td>http://www.informatics.jax.org/inbred_strains/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>B</td>\n",
       "      <td>C57BL/6J</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>000664</td>\n",
       "      <td>B6J</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://www.informatics.jax.org/inbred_strains/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>129S1/SvImJ</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>002448</td>\n",
       "      <td>129S1J</td>\n",
       "      <td>Male skin is tougher than female skin. Gallbla...</td>\n",
       "      <td>http://www.informatics.jax.org/inbred_strains/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D</td>\n",
       "      <td>NOD/ShiLtJ</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>001976</td>\n",
       "      <td>NODJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://www.informatics.jax.org/inbred_strains/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>NZO/HlLtJ</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>002105</td>\n",
       "      <td>NZOJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://www.informatics.jax.org/inbred_strains/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>F</td>\n",
       "      <td>CAST/EiJ</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>000928</td>\n",
       "      <td>CASTJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://www.informatics.jax.org/inbred_strains/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>G</td>\n",
       "      <td>PWK/PhJ</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>003715</td>\n",
       "      <td>PWKJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://www.informatics.jax.org/inbred_strains/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>H</td>\n",
       "      <td>WSB/EiJ</td>\n",
       "      <td>CC founder</td>\n",
       "      <td>001145</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.informatics.jax.org/strain/MGI:216...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Designation       Strain        Note Jax Catalog No Sample CODE  \\\n",
       "0           A          A/J  CC founder         000646          AJ   \n",
       "1           B     C57BL/6J  CC founder         000664         B6J   \n",
       "2           C  129S1/SvImJ  CC founder         002448      129S1J   \n",
       "3           D   NOD/ShiLtJ  CC founder         001976        NODJ   \n",
       "4           E    NZO/HlLtJ  CC founder         002105        NZOJ   \n",
       "5           F     CAST/EiJ  CC founder         000928       CASTJ   \n",
       "6           G      PWK/PhJ  CC founder         003715        PWKJ   \n",
       "7           H      WSB/EiJ  CC founder         001145        WSBJ   \n",
       "\n",
       "                                        Strain notes  \\\n",
       "0                            Yellower adrenal gland.   \n",
       "1                                                NaN   \n",
       "2  Male skin is tougher than female skin. Gallbla...   \n",
       "3                                                NaN   \n",
       "4                                                NaN   \n",
       "5                                                NaN   \n",
       "6                                                NaN   \n",
       "7                                                NaN   \n",
       "\n",
       "                                                 url  \n",
       "0  http://www.informatics.jax.org/inbred_strains/...  \n",
       "1  http://www.informatics.jax.org/inbred_strains/...  \n",
       "2  http://www.informatics.jax.org/inbred_strains/...  \n",
       "3  http://www.informatics.jax.org/inbred_strains/...  \n",
       "4  http://www.informatics.jax.org/inbred_strains/...  \n",
       "5  http://www.informatics.jax.org/inbred_strains/...  \n",
       "6  http://www.informatics.jax.org/inbred_strains/...  \n",
       "7  https://www.informatics.jax.org/strain/MGI:216...  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def int_csv_to_hex(x):\n",
    "    r, g, b = x.split(\",\")\n",
    "    \n",
    "    return \"#{:02x}{:02x}{:02x}\".format(int(r), int(g), int(b))\n",
    "\n",
    "def remove_citation(x):\n",
    "    return x.split(\" \")[0]\n",
    "\n",
    "strains = pandas.read_excel(\n",
    "    spreadsheet_name, \n",
    "    sheet_name=\"Line information\", \n",
    "    usecols=[\n",
    "        \"Designation\",\n",
    "        \"Strain\",\n",
    "        \"Note\",\n",
    "        \"Jax Catalog No\",\n",
    "        \"Sample CODE\",\n",
    "        \"Strain notes\",\n",
    "    ],\n",
    "    converters={\n",
    "        \"Strain\": remove_citation,\n",
    "        \"Jax Catalog No\": str,\n",
    "        \"Strain notes\": str_or_empty,\n",
    "    }\n",
    ")\n",
    "\n",
    "# Remove the extra lines with the bridge samples for now. I don't want to track thta level of detail\n",
    "bridge = strains[strains[\"Designation\"] == \"Bridge sample\"].first_valid_index() - 1\n",
    "strains = strains.iloc[0:bridge]\n",
    "\n",
    "# Add in Strain URL\n",
    "strains[\"url\"] = strains[\"Sample CODE\"].apply(lambda x: strain_urls[x])\n",
    "strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "indian-venture",
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_type_lookup = {\n",
    "    \"CC founder\": models.StrainType.CC_FOUNDER,\n",
    "    \"CC Cross\": models.StrainType.CC_CROSS,\n",
    "}\n",
    "\n",
    "if models.MouseStrain.objects.count() > 0:\n",
    "    truncate(models.MouseStrain)\n",
    "\n",
    "for i, row in strains.iterrows():\n",
    "    record = models.MouseStrain(\n",
    "        name=row[\"Strain\"],\n",
    "        code=strain_background[row[\"Sample CODE\"]],\n",
    "        strain_type=strain_type_lookup[row[\"Note\"]],\n",
    "        jax_catalog_number=row[\"Jax Catalog No\"],\n",
    "        notes=row[\"Strain notes\"],\n",
    "        url=row[\"url\"]\n",
    "    )\n",
    "    record.save()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "private-century",
   "metadata": {},
   "source": [
    "This is closest to the Mouse table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "distinct-failure",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def estrus_cycle(x):\n",
    "    if pandas.isnull(x):\n",
    "        return \"NA\"\n",
    "    else:\n",
    "        return x\n",
    "\n",
    "sex_lookup = {\n",
    "    numpy.nan: models.SexEnum.UNKNOWN,\n",
    "    \"Female\": models.SexEnum.FEMALE,\n",
    "    \"Male\": models.SexEnum.MALE,\n",
    "}\n",
    "\n",
    "estrus_stage = {\n",
    "    \"Anestrus\": models.EstrusCycle.ANESTRUS,\n",
    "    \"Anestrus>Proestrus\": models.EstrusCycle.ANESTRUS_PROESTRUS,\n",
    "    \"Proestrus\": models.EstrusCycle.PROESTRUS,\n",
    "    \"Proestrus>Estrus\": models.EstrusCycle.PROESTRUS_ESTRUS,\n",
    "    \"Estrus\": models.EstrusCycle.ESTRUS,\n",
    "    \"Estrus>Metestrus\": models.EstrusCycle.ESTRUS_METESTRUS,\n",
    "    \"Metestrus\": models.EstrusCycle.METESTRUS,\n",
    "    \"Metestrus>Diestrus\": models.EstrusCycle.METESTRUS_DIESTRUS,\n",
    "    \"Diestrus\": models.EstrusCycle.DIESTRUS,\n",
    "    \"Diestrus>Proestrus\": models.EstrusCycle.DIESTRUS_PROESTRUS,\n",
    "}\n",
    "    \n",
    "mice = pandas.read_excel(\n",
    "    spreadsheet_name, \n",
    "    sheet_name=\"mice - 8 founders\", \n",
    "    usecols=range(0,16),\n",
    "    converters={\n",
    "        \"Estres cycle stage\": estrus_cycle,\n",
    "        \"Dissection ID\": int,\n",
    "        \"RMS number\": int_or_none,\n",
    "        \"Housing number\": int_or_none,\n",
    "    }\n",
    ")\n",
    "\n",
    "if models.Mouse.objects.count() > 0:\n",
    "    truncate(models.Mouse)\n",
    "\n",
    "los_angeles_tz = zoneinfo.ZoneInfo(\"America/Los_Angeles\")\n",
    "\n",
    "failed = False\n",
    "for i, row in mice.iterrows():\n",
    "    if not pandas.isnull(row[\"Mouse Name\"]):\n",
    "        name = row[\"Mouse Name\"]\n",
    "        strain_code = row[\"Strain\"]\n",
    "        \n",
    "        if name == \"056_WSBJ_10F->PWKJ_9F\":\n",
    "            name = \"056_PWKJ_9F\"\n",
    "\n",
    "        try:\n",
    "            strain = models.MouseStrain.objects.get(name=strain_code)\n",
    "        except models.MouseStrain.DoesNotExist as e:\n",
    "            print(f\"Unable to find {row['Strain']} for row {i+2}\")\n",
    "            failed = True\n",
    "            continue\n",
    "\n",
    "        if not name[-1] in (\"M\", \"F\"):\n",
    "            raise ValueError(f\"Unrecognized sex field {mouse_name}\")\n",
    "            \n",
    "        name_fields = name.split(\"_\")\n",
    "        \n",
    "        if len(name_fields) != 3:\n",
    "            raise ValueError(f\"Wrong number of attributes in {name} line {i+1}\")\n",
    "            \n",
    "        if not name_fields[1] in strain_background:\n",
    "            raise ValueError(f\"strain background in {name} not recognized line {i+1}\")\n",
    "        \n",
    "        sample_box = \"\" if pandas.isnull(row[\"Sample box\"]) else row[\"Sample box\"]\n",
    "        \n",
    "        record = models.Mouse(\n",
    "            # should i use liz's disection id?\n",
    "            name=name,\n",
    "            strain=strain,\n",
    "            sex=sex_lookup[row[\"Sex\"]],\n",
    "            weight_g=row[\"Weight (g)\"],\n",
    "            date_of_birth=row[\"DOB\"].date(),\n",
    "            operator=row[\"Operator\"],\n",
    "            notes=str_or_empty(row[\"Comments\"]),\n",
    "            sample_box=sample_box,\n",
    "        )\n",
    "        \n",
    "        if not pandas.isnull(row[\"Estres cycle stage\"]):\n",
    "            # Ignore those transitions.\n",
    "            stage = row[\"Estres cycle stage\"]\n",
    "            if stage.endswith(\"?\"):\n",
    "                stage = stage[:-1]\n",
    "            record.estrus_cycle=estrus_stage[stage]\n",
    "            \n",
    "        if not pandas.isnull(row[\"Harvest start time\"]):\n",
    "            record.harvest_start_time=datetime.datetime(\n",
    "                row[\"Harvest Date\"].year,\n",
    "                row[\"Harvest Date\"].month,\n",
    "                row[\"Harvest Date\"].day,\n",
    "                row[\"Harvest start time\"].hour,\n",
    "                row[\"Harvest start time\"].minute,\n",
    "                row[\"Harvest start time\"].second,\n",
    "                tzinfo=los_angeles_tz,\n",
    "            )\n",
    "\n",
    "        if not pandas.isnull(row[\"Harvest finish time\"]):\n",
    "            record.harvest_end_time=datetime.datetime(\n",
    "                row[\"Harvest Date\"].year,\n",
    "                row[\"Harvest Date\"].month,\n",
    "                row[\"Harvest Date\"].day,\n",
    "                row[\"Harvest finish time\"].hour,\n",
    "                row[\"Harvest finish time\"].minute,\n",
    "                row[\"Harvest finish time\"].second,\n",
    "                tzinfo=los_angeles_tz,\n",
    "            )\n",
    "        record.save()\n",
    "        \n",
    "assert not failed, \"Check warning messages\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "hybrid-destruction",
   "metadata": {},
   "source": [
    "# Tissues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "electoral-ability",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>IGVF Sample BOX</th>\n",
       "      <th>Mouse_Tissue ID</th>\n",
       "      <th>Unnamed: 2</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Tube label</th>\n",
       "      <th>tube weight (g)</th>\n",
       "      <th>tube+tissue wight (g)</th>\n",
       "      <th>tissue weight (mg) or blood volume (ul)</th>\n",
       "      <th>cryopreserved blood location</th>\n",
       "      <th>Dissection date</th>\n",
       "      <th>Approx. sac time</th>\n",
       "      <th>Mouse_Tissue ID.1</th>\n",
       "      <th>Dissector</th>\n",
       "      <th>Comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IGVF_Sample_5&amp;6</td>\n",
       "      <td>016_B6J_10F_01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hypothalamus/Pituitary</td>\n",
       "      <td>F</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>B6J</td>\n",
       "      <td>016-01</td>\n",
       "      <td>1.046</td>\n",
       "      <td>1.118</td>\n",
       "      <td>72.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-10-27</td>\n",
       "      <td>09:01:00</td>\n",
       "      <td>016_B6J_10F_01</td>\n",
       "      <td>SK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IGVF_Sample_5&amp;6</td>\n",
       "      <td>016_B6J_10F_02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cerebellum</td>\n",
       "      <td>F</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>B6J</td>\n",
       "      <td>016-02</td>\n",
       "      <td>1.046</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1046.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-10-27</td>\n",
       "      <td>09:01:00</td>\n",
       "      <td>016_B6J_10F_02</td>\n",
       "      <td>SK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IGVF_Sample_5&amp;6</td>\n",
       "      <td>016_B6J_10F_03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cortex/Hippocampus left</td>\n",
       "      <td>F</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>B6J</td>\n",
       "      <td>016-03</td>\n",
       "      <td>1.040</td>\n",
       "      <td>1.181</td>\n",
       "      <td>141.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-10-27</td>\n",
       "      <td>09:01:00</td>\n",
       "      <td>016_B6J_10F_03</td>\n",
       "      <td>SK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IGVF_Sample_5&amp;6</td>\n",
       "      <td>016_B6J_10F_04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cortex/Hippocampus right</td>\n",
       "      <td>F</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>B6J</td>\n",
       "      <td>016-04</td>\n",
       "      <td>1.048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1048.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-10-27</td>\n",
       "      <td>09:01:00</td>\n",
       "      <td>016_B6J_10F_04</td>\n",
       "      <td>SK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IGVF_Sample_5&amp;6</td>\n",
       "      <td>016_B6J_10F_05</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Liver</td>\n",
       "      <td>F</td>\n",
       "      <td>10 weeks</td>\n",
       "      <td>B6J</td>\n",
       "      <td>016-05</td>\n",
       "      <td>1.133</td>\n",
       "      <td>2.099</td>\n",
       "      <td>966.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-10-27</td>\n",
       "      <td>09:01:00</td>\n",
       "      <td>016_B6J_10F_05</td>\n",
       "      <td>GM</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1635</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1636</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1637</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1638</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1639</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1640 rows Ã— 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      IGVF Sample BOX Mouse_Tissue ID Unnamed: 2                    Tissue  \\\n",
       "0     IGVF_Sample_5&6  016_B6J_10F_01        NaN    Hypothalamus/Pituitary   \n",
       "1     IGVF_Sample_5&6  016_B6J_10F_02        NaN                Cerebellum   \n",
       "2     IGVF_Sample_5&6  016_B6J_10F_03        NaN   Cortex/Hippocampus left   \n",
       "3     IGVF_Sample_5&6  016_B6J_10F_04        NaN  Cortex/Hippocampus right   \n",
       "4     IGVF_Sample_5&6  016_B6J_10F_05        NaN                     Liver   \n",
       "...               ...             ...        ...                       ...   \n",
       "1635              NaN             NaN        NaN                       NaN   \n",
       "1636              NaN             NaN        NaN                       NaN   \n",
       "1637              NaN             NaN        NaN                       NaN   \n",
       "1638              NaN             NaN        NaN                       NaN   \n",
       "1639              NaN             NaN        NaN                       NaN   \n",
       "\n",
       "      Sex Timepoint Genotype Tube label  tube weight (g)  \\\n",
       "0       F  10 weeks      B6J     016-01            1.046   \n",
       "1       F  10 weeks      B6J     016-02            1.046   \n",
       "2       F  10 weeks      B6J     016-03            1.040   \n",
       "3       F  10 weeks      B6J     016-04            1.048   \n",
       "4       F  10 weeks      B6J     016-05            1.133   \n",
       "...   ...       ...      ...        ...              ...   \n",
       "1635  NaN       NaN      NaN        NaN              NaN   \n",
       "1636  NaN       NaN      NaN        NaN              NaN   \n",
       "1637  NaN       NaN      NaN        NaN              NaN   \n",
       "1638  NaN       NaN      NaN        NaN              NaN   \n",
       "1639  NaN       NaN      NaN        NaN              NaN   \n",
       "\n",
       "      tube+tissue wight (g)  tissue weight (mg) or blood volume (ul)  \\\n",
       "0                     1.118                                     72.0   \n",
       "1                       NaN                                  -1046.0   \n",
       "2                     1.181                                    141.0   \n",
       "3                       NaN                                  -1048.0   \n",
       "4                     2.099                                    966.0   \n",
       "...                     ...                                      ...   \n",
       "1635                    NaN                                      NaN   \n",
       "1636                    NaN                                      NaN   \n",
       "1637                    NaN                                      NaN   \n",
       "1638                    NaN                                      NaN   \n",
       "1639                    NaN                                      NaN   \n",
       "\n",
       "     cryopreserved blood location Dissection date Approx. sac time  \\\n",
       "0                             NaN      2022-10-27         09:01:00   \n",
       "1                             NaN      2022-10-27         09:01:00   \n",
       "2                             NaN      2022-10-27         09:01:00   \n",
       "3                             NaN      2022-10-27         09:01:00   \n",
       "4                             NaN      2022-10-27         09:01:00   \n",
       "...                           ...             ...              ...   \n",
       "1635                          NaN             NaT              NaN   \n",
       "1636                          NaN             NaT              NaN   \n",
       "1637                          NaN             NaT              NaN   \n",
       "1638                          NaN             NaT              NaN   \n",
       "1639                          NaN             NaT              NaN   \n",
       "\n",
       "     Mouse_Tissue ID.1 Dissector Comment  \n",
       "0       016_B6J_10F_01        SK     NaN  \n",
       "1       016_B6J_10F_02        SK     NaN  \n",
       "2       016_B6J_10F_03        SK     NaN  \n",
       "3       016_B6J_10F_04        SK     NaN  \n",
       "4       016_B6J_10F_05        GM     NaN  \n",
       "...                ...       ...     ...  \n",
       "1635               NaN       NaN     NaN  \n",
       "1636               NaN       NaN     NaN  \n",
       "1637               NaN       NaN     NaN  \n",
       "1638               NaN       NaN     NaN  \n",
       "1639               NaN       NaN     NaN  \n",
       "\n",
       "[1640 rows x 17 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#if models.Tissue.objects.count() > 0:\n",
    "#    truncate(models.Tissue)\n",
    "\n",
    "tissues = pandas.read_excel(\n",
    "    spreadsheet_name, \n",
    "    sheet_name=\"Samples - 8 founders\",\n",
    "    header=0,\n",
    ")\n",
    "tissues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "threatened-procedure",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_timepoint(value):\n",
    "    value, units = value.split(\" \")\n",
    "    value = float(value)\n",
    "    \n",
    "    unit_map = {\n",
    "        \"day\": models.AgeUnitsEnum.DAY,\n",
    "        \"days\": models.AgeUnitsEnum.DAY,\n",
    "        \"week\": models.AgeUnitsEnum.WEEK,\n",
    "        \"weeks\": models.AgeUnitsEnum.WEEK,\n",
    "        \"month\": models.AgeUnitsEnum.MONTH,\n",
    "        \"months\": models.AgeUnitsEnum.MONTH,\n",
    "    }\n",
    "    return (value, unit_map[units])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "relative-corps",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "if models.Tissue.objects.count() > 0:\n",
    "    truncate(models.Tissue)\n",
    "\n",
    "failed = False\n",
    "for i, row in tissues.iterrows():\n",
    "    if not pandas.isnull(row[\"Tissue\"]):\n",
    "        mouse_tissue_id_label = \"Mouse_Tissue ID\"\n",
    "        mouse_tissue_name_fields = row[mouse_tissue_id_label].split(\"_\")\n",
    "        \n",
    "        # some validation on tissue name\n",
    "        if len(mouse_tissue_name_fields) != 4:\n",
    "            raise ValueError(f\"Not enough elements in mouse tissue id {row[mouse_tissue_id_label]}\")\n",
    "            \n",
    "        if not mouse_tissue_name_fields[1] in strain_background:\n",
    "            raise ValueError(f\"tissue strain field in {row[mouse_tissue_id_label]} not recognized\")\n",
    "        \n",
    "        mouse_name = \"_\".join(mouse_tissue_name_fields[0:3])\n",
    "        try:\n",
    "            mouse = models.Mouse.objects.get(name=mouse_name)\n",
    "        except models.Mouse.DoesNotExist:\n",
    "            print(\"row {}, {} was not found\".format(i+2, mouse_name))\n",
    "            failed = True\n",
    "            continue\n",
    "\n",
    "        genotype = strain_background[row[\"Genotype\"]]\n",
    "        \n",
    "        # this is the \"label swap\" on spreadsheet rows 602-605.\n",
    "        if mouse_name == \"092_CASTJ_10F\":\n",
    "            genotype = \"CAST/EiJ (CAST)\"\n",
    "        # this is the other half the swap on spreadsheet rows 1522-1525\n",
    "        elif mouse_name == \"046_NZOJ_10F\":\n",
    "            genotype = 'NZO/H1LtJ (NZO)'\n",
    "\n",
    "        assert mouse.strain.code == genotype, f\"{mouse.mouse_strain.code} != {genotype}\"\n",
    "        tissue_terms = []\n",
    "        for term_curie, term_name in tissue_dissection_to_ontology_map[row[\"Tissue\"]]:\n",
    "            tissue_terms.append(models.OntologyTerm.objects.get(pk=term_curie))\n",
    "        \n",
    "        #age, age_units = parse_timepoint(row[\"Timepoint\"])\n",
    "        \n",
    "        if pandas.isnull(row[\"Approx. sac time\"]):\n",
    "            sac_time = datetime.time(0,0,0)\n",
    "        else:\n",
    "            sac_time = datetime.time(\n",
    "                row[\"Approx. sac time\"].hour,\n",
    "                row[\"Approx. sac time\"].minute,\n",
    "                row[\"Approx. sac time\"].second,    \n",
    "            )\n",
    "        dissection = datetime.datetime(\n",
    "            row[\"Dissection date\"].year,\n",
    "            row[\"Dissection date\"].month,\n",
    "            row[\"Dissection date\"].day,\n",
    "            sac_time.hour,\n",
    "            sac_time.minute,\n",
    "            sac_time.second,\n",
    "            tzinfo=los_angeles_tz,\n",
    "        )\n",
    "        \n",
    "        record = models.Tissue(\n",
    "            mouse=mouse,\n",
    "            name = row[mouse_tissue_id_label],\n",
    "            dissection_time=dissection,\n",
    "            #age=age,\n",
    "            #age_units=age_units,\n",
    "            tube_label=row[\"Tube label\"],\n",
    "            life_stage=models.LifeStageEnum.ADULT,\n",
    "            dissector=row[\"Dissector\"],\n",
    "            dissection_notes=row[\"Comment\"],\n",
    "        )\n",
    "\n",
    "        tube_weight_label = \"tube weight (g)\"\n",
    "        if not pandas.isnull(row[tube_weight_label]):\n",
    "            record.tube_weight_g = float(row[tube_weight_label])\n",
    "\n",
    "        total_weight_label = \"tube+tissue wight (g)\"\n",
    "        if not pandas.isnull(row[total_weight_label]):\n",
    "            record.total_weight_g = float(row[total_weight_label])\n",
    "        \n",
    "        record.save()\n",
    "        record.ontology_term.set(tissue_terms)\n",
    "        record.save()\n",
    "\n",
    "assert not failed, \"Check warning messages.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "several-journey",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AJ': 'A/J (AJ)',\n",
       " 'B6J': 'C57BL/6J (B6)',\n",
       " '129S1J': '129S1/SvImJ (129)',\n",
       " 'NODJ': 'NOD/ShiLtJ (NOD)',\n",
       " 'NZOJ': 'NZO/H1LtJ (NZO)',\n",
       " 'CASTJ': 'CAST/EiJ (CAST)',\n",
       " 'PWKJ': 'PWK/PhJ (PWK)',\n",
       " 'WSBJ': 'WSB/EiJ (WSB)'}"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "strain_background"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "scenic-vermont",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "historic-stand",
   "metadata": {},
   "source": [
    "# Information from Samples into experiment\n",
    "\n",
    "- Total barcoded nuclei (Samples into experiment)\n",
    "\n",
    "Is this FixedTissue, FixedSample, FixedBiosample?\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "central-oriental",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tissue ['046_NZOJ_10F_03'] weight does not match 115.0 150.0. 157\n",
      "Tissue ['046_NZOJ_10F_01'] weight does not match 42.0 76.0. 457\n",
      "Tissue ['046_NZOJ_10F_17'] weight does not match 190.0 116.0. 556\n",
      "Tissue ['016_B6J_10F_10', '016_B6J_10F_11'] weight does not match 33.0 17.0. 600\n",
      "Tissue ['016_B6J_10F_10', '016_B6J_10F_11'] weight does not match 33.0 16.0. 600\n",
      "Tissue ['017_B6J_10M_10', '017_B6J_10M_11'] weight does not match 263.0 180.0. 601\n",
      "Tissue ['017_B6J_10M_10', '017_B6J_10M_11'] weight does not match 263.0 83.0. 601\n",
      "Tissue ['018_B6J_10F_10', '018_B6J_10F_11'] weight does not match 42.0 25.0. 602\n",
      "Tissue ['018_B6J_10F_10', '018_B6J_10F_11'] weight does not match 42.0 17.0. 602\n",
      "Tissue ['019_B6J_10M_10', '019_B6J_10M_11'] weight does not match 280.0 178.0. 603\n",
      "Tissue ['019_B6J_10M_10', '019_B6J_10M_11'] weight does not match 280.0 102.0. 603\n",
      "Tissue ['020_B6J_10F_10', '020_B6J_10F_11'] weight does not match 40.0 20.0. 604\n",
      "Tissue ['020_B6J_10F_10', '020_B6J_10F_11'] weight does not match 40.0 20.0. 604\n",
      "Tissue ['021_B6J_10M_10', '021_B6J_10M_11'] weight does not match 334.0 227.0. 605\n",
      "Tissue ['021_B6J_10M_10', '021_B6J_10M_11'] weight does not match 334.0 107.0. 605\n",
      "Tissue ['024_B6J_10F_10', '024_B6J_10F_11'] weight does not match 30.0 13.0. 606\n",
      "Tissue ['024_B6J_10F_10', '024_B6J_10F_11'] weight does not match 30.0 17.0. 606\n",
      "Tissue ['025_B6J_10M_10', '025_B6J_10M_11'] weight does not match 320.0 211.0. 607\n",
      "Tissue ['025_B6J_10M_10', '025_B6J_10M_11'] weight does not match 320.0 109.0. 607\n",
      "Tissue ['026_AJ_10F_10', '026_AJ_10F_11'] weight does not match 41.0 21.0. 608\n",
      "Tissue ['026_AJ_10F_10', '026_AJ_10F_11'] weight does not match 41.0 20.0. 608\n",
      "Tissue ['029_AJ_10M_10', '029_AJ_10M_11'] weight does not match 231.0 155.0. 609\n",
      "Tissue ['029_AJ_10M_10', '029_AJ_10M_11'] weight does not match 231.0 76.0. 609\n",
      "Tissue ['028_AJ_10F_10', '028_AJ_10F_11'] weight does not match 39.0 21.0. 610\n",
      "Tissue ['028_AJ_10F_10', '028_AJ_10F_11'] weight does not match 39.0 18.0. 610\n",
      "Tissue ['031_AJ_10M_10', '031_AJ_10M_11'] weight does not match 216.0 147.0. 611\n",
      "Tissue ['031_AJ_10M_10', '031_AJ_10M_11'] weight does not match 216.0 69.0. 611\n",
      "Tissue ['030_AJ_10F_10', '030_AJ_10F_11'] weight does not match 35.0 19.0. 612\n",
      "Tissue ['030_AJ_10F_10', '030_AJ_10F_11'] weight does not match 35.0 16.0. 612\n",
      "Tissue ['033_AJ_10M_10', '033_AJ_10M_11'] weight does not match 218.0 150.0. 613\n",
      "Tissue ['033_AJ_10M_10', '033_AJ_10M_11'] weight does not match 218.0 68.0. 613\n",
      "Tissue ['032_AJ_10F_10', '032_AJ_10F_11'] weight does not match 31.0 16.0. 614\n",
      "Tissue ['032_AJ_10F_10', '032_AJ_10F_11'] weight does not match 31.0 15.0. 614\n",
      "Tissue ['035_AJ_10M_10', '035_AJ_10M_11'] weight does not match 264.0 182.0. 615\n",
      "Tissue ['035_AJ_10M_10', '035_AJ_10M_11'] weight does not match 264.0 82.0. 615\n",
      "Tissue ['036_129S1J_10F_10', '036_129S1J_10F_11'] weight does not match 27.0 14.0. 616\n",
      "Tissue ['036_129S1J_10F_10', '036_129S1J_10F_11'] weight does not match 27.0 13.0. 616\n",
      "Tissue ['037_129S1J_10M_10', '037_129S1J_10M_11'] weight does not match 282.0 208.0. 617\n",
      "Tissue ['037_129S1J_10M_10', '037_129S1J_10M_11'] weight does not match 282.0 74.0. 617\n",
      "Tissue ['038_129S1J_10F_10', '038_129S1J_10F_11'] weight does not match 36.0 22.0. 618\n",
      "Tissue ['038_129S1J_10F_10', '038_129S1J_10F_11'] weight does not match 36.0 14.0. 618\n",
      "Tissue ['041_129S1J_10M_10', '041_129S1J_10M_11'] weight does not match 316.0 228.0. 619\n",
      "Tissue ['041_129S1J_10M_10', '041_129S1J_10M_11'] weight does not match 316.0 88.0. 619\n",
      "Tissue ['040_129S1J_10F_10', '040_129S1J_10F_11'] weight does not match 34.0 21.0. 620\n",
      "Tissue ['040_129S1J_10F_10', '040_129S1J_10F_11'] weight does not match 34.0 13.0. 620\n",
      "Tissue ['043_129S1J_10M_10', '043_129S1J_10M_11'] weight does not match 285.0 208.0. 621\n",
      "Tissue ['043_129S1J_10M_10', '043_129S1J_10M_11'] weight does not match 285.0 77.0. 621\n",
      "Tissue ['044_129S1J_10F_10', '044_129S1J_10F_11'] weight does not match 35.0 21.0. 622\n",
      "Tissue ['044_129S1J_10F_10', '044_129S1J_10F_11'] weight does not match 35.0 14.0. 622\n",
      "Tissue ['045_129S1J_10M_10', '045_129S1J_10M_11'] weight does not match 315.0 229.0. 623\n",
      "Tissue ['045_129S1J_10M_10', '045_129S1J_10M_11'] weight does not match 315.0 86.0. 623\n",
      "Tissue ['046_NZOJ_10F_10', '046_NZOJ_10F_11'] weight does not match 44.0 26.0. 624\n",
      "Tissue ['046_NZOJ_10F_10', '046_NZOJ_10F_11'] weight does not match 44.0 18.0. 624\n",
      "Tissue ['047_NZOJ_10M_10', '047_NZOJ_10M_11'] weight does not match 334.0 217.0. 625\n",
      "Tissue ['047_NZOJ_10M_10', '047_NZOJ_10M_11'] weight does not match 334.0 117.0. 625\n",
      "Tissue ['048_NZOJ_10F_10', '048_NZOJ_10F_11'] weight does not match 43.0 26.0. 626\n",
      "Tissue ['048_NZOJ_10F_10', '048_NZOJ_10F_11'] weight does not match 43.0 17.0. 626\n",
      "Tissue ['049_NZOJ_10M_10', '049_NZOJ_10M_11'] weight does not match 365.0 212.0. 627\n",
      "Tissue ['049_NZOJ_10M_10', '049_NZOJ_10M_11'] weight does not match 365.0 153.0. 627\n",
      "Tissue ['050_NZOJ_10F_10', '050_NZOJ_10F_11'] weight does not match 30.0 16.0. 628\n",
      "Tissue ['050_NZOJ_10F_10', '050_NZOJ_10F_11'] weight does not match 30.0 14.0. 628\n",
      "Tissue ['050_NZOJ_10F_10', '050_NZOJ_10F_11'] weight does not match 341.0 16.0. 629\n",
      "Tissue ['050_NZOJ_10F_10', '050_NZOJ_10F_11'] weight does not match 341.0 14.0. 629\n",
      "Tissue ['052_NZOJ_10F_10', '052_NZOJ_10F_11'] weight does not match 40.0 19.0. 630\n",
      "Tissue ['052_NZOJ_10F_10', '052_NZOJ_10F_11'] weight does not match 40.0 21.0. 630\n",
      "Tissue ['053_NZOJ_10M_10', '053_NZOJ_10M_11'] weight does not match 370.0 243.0. 631\n",
      "Tissue ['053_NZOJ_10M_10', '053_NZOJ_10M_11'] weight does not match 370.0 127.0. 631\n",
      "Tissue ['058_WSBJ_10F_10', '058_WSBJ_10F_11'] weight does not match 11.0 8.0. 632\n",
      "Tissue ['058_WSBJ_10F_10', '058_WSBJ_10F_11'] weight does not match 11.0 3.0. 632\n",
      "Tissue ['057_WSBJ_10M_10', '057_WSBJ_10M_11'] weight does not match 168.0 110.0. 633\n",
      "Tissue ['057_WSBJ_10M_10', '057_WSBJ_10M_11'] weight does not match 168.0 58.0. 633\n",
      "Tissue ['060_WSBJ_10F_10', '060_WSBJ_10F_11'] weight does not match 22.0 8.0. 634\n",
      "Tissue ['060_WSBJ_10F_10', '060_WSBJ_10F_11'] weight does not match 22.0 14.0. 634\n",
      "Tissue ['059_WSBJ_10M_10', '059_WSBJ_10M_11'] weight does not match 246.0 172.0. 635\n",
      "Tissue ['059_WSBJ_10M_10', '059_WSBJ_10M_11'] weight does not match 246.0 74.0. 635\n",
      "Tissue ['062_WSBJ_10F_10', '062_WSBJ_10F_11'] weight does not match 19.0 10.0. 636\n",
      "Tissue ['062_WSBJ_10F_10', '062_WSBJ_10F_11'] weight does not match 19.0 9.0. 636\n",
      "Tissue ['061_WSBJ_10M_10', '061_WSBJ_10M_11'] weight does not match 213.0 141.0. 637\n",
      "Tissue ['061_WSBJ_10M_10', '061_WSBJ_10M_11'] weight does not match 213.0 72.0. 637\n",
      "Tissue ['096_WSBJ_10F_10', '096_WSBJ_10F_11'] weight does not match 20.0 14.0. 638\n",
      "Tissue ['096_WSBJ_10F_10', '096_WSBJ_10F_11'] weight does not match 20.0 6.0. 638\n",
      "Tissue ['063_WSBJ_10M_10', '063_WSBJ_10M_11'] weight does not match 170.0 111.0. 639\n",
      "Tissue ['063_WSBJ_10M_10', '063_WSBJ_10M_11'] weight does not match 170.0 59.0. 639\n",
      "Tissue ['066_NODJ_10F_10', '066_NODJ_10F_11'] weight does not match 30.0 19.0. 640\n",
      "Tissue ['066_NODJ_10F_10', '066_NODJ_10F_11'] weight does not match 30.0 11.0. 640\n",
      "Tissue ['067_NODJ_10M_10', '067_NODJ_10M_11'] weight does not match 336.0 231.0. 641\n",
      "Tissue ['067_NODJ_10M_10', '067_NODJ_10M_11'] weight does not match 336.0 105.0. 641\n",
      "Tissue ['068_NODJ_10F_10', '068_NODJ_10F_11'] weight does not match 28.0 16.0. 642\n",
      "Tissue ['068_NODJ_10F_10', '068_NODJ_10F_11'] weight does not match 28.0 12.0. 642\n",
      "Tissue ['069_NODJ_10M_10', '069_NODJ_10M_11'] weight does not match 324.0 219.0. 643\n",
      "Tissue ['069_NODJ_10M_10', '069_NODJ_10M_11'] weight does not match 324.0 105.0. 643\n",
      "Tissue ['070_NODJ_10F_10', '070_NODJ_10F_11'] weight does not match 26.0 17.0. 644\n",
      "Tissue ['070_NODJ_10F_10', '070_NODJ_10F_11'] weight does not match 26.0 9.0. 644\n",
      "Tissue ['071_NODJ_10M_10', '071_NODJ_10M_11'] weight does not match 342.0 221.0. 645\n",
      "Tissue ['071_NODJ_10M_10', '071_NODJ_10M_11'] weight does not match 342.0 121.0. 645\n",
      "Tissue ['072_NODJ_10F_10', '072_NODJ_10F_11'] weight does not match 31.0 18.0. 646\n",
      "Tissue ['072_NODJ_10F_10', '072_NODJ_10F_11'] weight does not match 31.0 13.0. 646\n",
      "Tissue ['073_NODJ_10M_10', '073_NODJ_10M_11'] weight does not match 332.0 211.0. 647\n",
      "Tissue ['073_NODJ_10M_10', '073_NODJ_10M_11'] weight does not match 332.0 121.0. 647\n",
      "Tissue ['076_PWKJ_10F_10', '076_PWKJ_10F_11'] weight does not match 24.0 12.0. 648\n",
      "Tissue ['076_PWKJ_10F_10', '076_PWKJ_10F_11'] weight does not match 24.0 12.0. 648\n",
      "Tissue ['077_PWKJ_10M_10', '077_PWKJ_10M_11'] weight does not match 249.0 175.0. 649\n",
      "Tissue ['077_PWKJ_10M_10', '077_PWKJ_10M_11'] weight does not match 249.0 74.0. 649\n",
      "Tissue ['078_PWKJ_10F_10', '078_PWKJ_10F_11'] weight does not match 28.0 16.0. 650\n",
      "Tissue ['078_PWKJ_10F_10', '078_PWKJ_10F_11'] weight does not match 28.0 12.0. 650\n",
      "Tissue ['079_PWKJ_10M_10', '079_PWKJ_10M_11'] weight does not match 208.0 149.0. 651\n",
      "Tissue ['079_PWKJ_10M_10', '079_PWKJ_10M_11'] weight does not match 208.0 59.0. 651\n",
      "Tissue ['080_PWKJ_10F_10', '080_PWKJ_10F_11'] weight does not match 33.0 13.0. 652\n",
      "Tissue ['080_PWKJ_10F_10', '080_PWKJ_10F_11'] weight does not match 33.0 20.0. 652\n",
      "Tissue ['081_PWKJ_10M_10', '081_PWKJ_10M_11'] weight does not match 242.0 178.0. 653\n",
      "Tissue ['081_PWKJ_10M_10', '081_PWKJ_10M_11'] weight does not match 242.0 64.0. 653\n",
      "Tissue ['084_PWKJ_10F_10', '084_PWKJ_10F_11'] weight does not match 19.0 8.0. 654\n",
      "Tissue ['084_PWKJ_10F_10', '084_PWKJ_10F_11'] weight does not match 19.0 11.0. 654\n",
      "Tissue ['085_PWKJ_10M_10', '085_PWKJ_10M_11'] weight does not match 266.0 200.0. 655\n",
      "Tissue ['085_PWKJ_10M_10', '085_PWKJ_10M_11'] weight does not match 266.0 66.0. 655\n",
      "Tissue ['086_CASTJ_10F_10', '086_CASTJ_10F_11'] weight does not match 28.0 15.0. 656\n",
      "Tissue ['086_CASTJ_10F_10', '086_CASTJ_10F_11'] weight does not match 28.0 13.0. 656\n",
      "Tissue ['087_CASTJ_10M_10', '087_CASTJ_10M_11'] weight does not match 171.0 110.0. 657\n",
      "Tissue ['087_CASTJ_10M_10', '087_CASTJ_10M_11'] weight does not match 171.0 61.0. 657\n",
      "Tissue ['088_CASTJ_10F_10', '088_CASTJ_10F_11'] weight does not match 28.0 13.0. 658\n",
      "Tissue ['088_CASTJ_10F_10', '088_CASTJ_10F_11'] weight does not match 28.0 15.0. 658\n",
      "Tissue ['089_CASTJ_10M_10', '089_CASTJ_10M_11'] weight does not match 166.0 109.0. 659\n",
      "Tissue ['089_CASTJ_10M_10', '089_CASTJ_10M_11'] weight does not match 166.0 57.0. 659\n",
      "Tissue ['090_CASTJ_10F_10', '090_CASTJ_10F_11'] weight does not match 24.0 12.0. 660\n",
      "Tissue ['090_CASTJ_10F_10', '090_CASTJ_10F_11'] weight does not match 24.0 12.0. 660\n",
      "Tissue ['091_CASTJ_10M_10', '091_CASTJ_10M_11'] weight does not match 245.0 144.0. 661\n",
      "Tissue ['091_CASTJ_10M_10', '091_CASTJ_10M_11'] weight does not match 245.0 101.0. 661\n",
      "Tissue ['094_CASTJ_10F_10', '094_CASTJ_10F_11'] weight does not match 23.0 11.0. 662\n",
      "Tissue ['094_CASTJ_10F_10', '094_CASTJ_10F_11'] weight does not match 23.0 12.0. 662\n",
      "Tissue ['093_CASTJ_10M_10', '093_CASTJ_10M_11'] weight does not match 178.0 115.0. 663\n",
      "Tissue ['093_CASTJ_10M_10', '093_CASTJ_10M_11'] weight does not match 178.0 63.0. 663\n"
     ]
    }
   ],
   "source": [
    "samples_into_experiment = pandas.read_excel(\n",
    "    spreadsheet_name, \n",
    "    sheet_name=\"Samples into experiment\",\n",
    "    header=[0,1],\n",
    "    index_col=None,\n",
    ")\n",
    "\n",
    "level0 = []\n",
    "level1 = []\n",
    "for header in samples_into_experiment.columns:\n",
    "    level0.append(\"\" if header[0].startswith(\"Unnamed: \") else header[0])\n",
    "    level1.append(header[1])\n",
    "\n",
    "samples_into_experiment.columns = pandas.MultiIndex.from_arrays([level0, level1], names=[\"phase\", \"name\"])\n",
    "\n",
    "if models.FixedSample.objects.count() > 0:\n",
    "    truncate(models.FixedSample)\n",
    "\n",
    "one_based = 1\n",
    "header_lines = 2\n",
    "failed = False\n",
    "for i, row in samples_into_experiment.iterrows():\n",
    "    line_no = i + one_based + header_lines\n",
    "    box_name = row[(\"\", \"IGVF Fixation BOX\")]\n",
    "    if isinstance(box_name, str):\n",
    "        box_name = box_name.strip()\n",
    "    \n",
    "    if box_name in [\"IGVF_FIX_001\", \"IGVF_FIX_002\"]:\n",
    "        continue\n",
    "\n",
    "    tissue_id = row[(\"\", \"Mouse_Tissue ID\")]\n",
    "    pooled_from = row[(\"after fixation\", \"pooled_from\")]\n",
    "    \n",
    "    if pandas.isnull(tissue_id):\n",
    "        continue\n",
    "    \n",
    "    if (tissue_id.endswith(\"_25\") or tissue_id.endswith(\"_26\")) and pandas.isnull(pooled_from):\n",
    "        print(f\"How are we going to merge these samples? {tissue_id}\")\n",
    "        continue\n",
    "        \n",
    "    if pandas.isnull(pooled_from):\n",
    "        pooled_from = [tissue_id]\n",
    "    else:\n",
    "        pooled_from = pooled_from.split(\",\")\n",
    "    \n",
    "    tissues = []\n",
    "    for pooled_id in pooled_from:\n",
    "        try:\n",
    "            tissues.append(models.Tissue.objects.get(name=pooled_id))\n",
    "        except models.Tissue.DoesNotExist:\n",
    "            print(f\"Tissue {pooled_id} not found in tissue table on {line_no}\")\n",
    "            failed = True\n",
    "\n",
    "    weight = row[(\"\", \"weight (mg)\")]\n",
    "    fixation_date = row[(\"\", \"Fixation date\")]\n",
    "    notes = row[(\"\", \"Notes\")]\n",
    "            \n",
    "    if weight == \"#VALUE!\":\n",
    "        pass\n",
    "    elif weight < 0:\n",
    "        pass\n",
    "    elif pandas.isnull(weight):\n",
    "        pass\n",
    "    else:\n",
    "        for tissue in tissues:\n",
    "            if not pandas.isnull(tissue.weight_mg) and tissue.weight_mg != weight:\n",
    "                print(f\"Tissue {pooled_from} weight does not match {weight} {tissue.weight_mg}. {line_no}\")\n",
    "\n",
    "    sample_id = row[(\"after fixation\", \"Sample ID\")]\n",
    "    if not (pandas.isnull(sample_id) or pandas.isnull(tissue_id)):\n",
    "        if tissue_id != sample_id:\n",
    "            print(f\"If defined, {tissue_id} should equal {sample_id} line {line_no}\")\n",
    "            failed = True\n",
    "\n",
    "    cap_label = row[(\"after fixation\", \"Cap label\")]\n",
    "    if not (pandas.isnull(cap_label) or pandas.isnull(sample_id)):\n",
    "        sample_fields = sample_id.split(\"_\")\n",
    "        predicted_label = \"_\".join([sample_fields[0], sample_fields[-1]])\n",
    "        if predicted_label != cap_label:\n",
    "            print(f\"{cap_label} should equal {predicted_label} line {line_no}\")\n",
    "            failed = True\n",
    "    \n",
    "    well_id = row[(\"Loading\", \"wells in Barcoding plate\")]\n",
    "        \n",
    "    record = models.FixedSample(\n",
    "        name=tissue_id,\n",
    "        tube_label=cap_label,\n",
    "        fixation_name=box_name,\n",
    "        fixation_date=date_or_none(row[(\"\", \"Fixation date\")]),\n",
    "        starting_nuclei=int_or_none(row[(\"before fixation\", \"Total nuclei (x10^6)\")]),\n",
    "        nuclei_into_fixation=int_or_none(row[(\"before fixation\", \"Nuclei into fixation (x10^6)\")]),\n",
    "        fixed_nuclei=int_or_none(row[(\"after fixation\", \"Total fixed nuclei (x10^6)\")]),\n",
    "        aliquots_made=int_or_none(row[(\"after fixation\", \"# aliquots\")]),\n",
    "        aliquot_volume_ul=float_or_none(row[(\"after fixation\", \"uL per aliquot\")]),\n",
    "    )\n",
    "    record.save()\n",
    "    record.tissue.set(tissues)\n",
    "    record.save()\n",
    "\n",
    "assert not failed, \"Check warning messages\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "specific-candle",
   "metadata": {},
   "source": [
    "# Plate layout\n",
    "\n",
    "(the hard thing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "subject-accused",
   "metadata": {},
   "outputs": [],
   "source": [
    "WellContent = namedtuple(\"well_content\", [\"genotype\", \"tissue_id\"])\n",
    "\n",
    "def validate_tissue_ids(contents, expected_genotypes):\n",
    "    for expected, (row_index, row) in zip(expected_genotypes, contents.iterrows()):\n",
    "        for cell in row:\n",
    "            fields = cell.split(\"_\")\n",
    "            assert len(fields) == 4, f\"Field label validation fail {cell}\"\n",
    "            assert fields[1] == expected, f\"{fields[1]} failed to match {expected}\"\n",
    "\n",
    "def find_plate_start(sheet, offsets):\n",
    "    for plate_id_row in sheet[sheet[offsets[\"plate_label\"]].apply(lambda x: not pandas.isnull(x) and x.startswith(\"IGVF_\"))].index:\n",
    "        plate_name = sheet.loc[plate_id_row, offsets[\"plate_label\"]]\n",
    "        for i in range(plate_id_row, plate_id_row + 4):\n",
    "            cell = sheet.loc[i, offsets[\"well_start\"]]\n",
    "            if isinstance(cell, str) and cell.startswith(\"Tissue\"):\n",
    "                yield (plate_name, i)\n",
    "                \n",
    "def get_plate_genotype(sheet, start, offsets):\n",
    "    plate_tissues = sheet.loc[start+2:start+9, offsets[\"tissue_label\"]].tolist()\n",
    "    \n",
    "    return plate_tissues\n",
    "\n",
    "def parse_single_well_block(sheet, plate_start, offsets):\n",
    "    # column ids\n",
    "    simple_plate_column_ids = sheet.loc[plate_start+1, offsets[\"well_range\"]].tolist()\n",
    "\n",
    "    plate_row_labels = sheet.loc[plate_start+2:plate_start+9, offsets[\"well_row_label\"]].tolist()\n",
    "\n",
    "    contents = sheet.loc[plate_start+2:plate_start+9, offsets[\"well_range\"]].copy()\n",
    "    contents.index = plate_row_labels\n",
    "    contents.columns = simple_plate_column_ids\n",
    "\n",
    "    genotypes = get_plate_genotype(sheet, plate_start, offsets)\n",
    "    validate_tissue_ids(contents, genotypes)\n",
    "\n",
    "    well_contents = {}\n",
    "    for genotype, (well_row, row) in zip(genotypes, contents.iterrows()):\n",
    "        for well_column, cell in zip(simple_plate_column_ids, row):\n",
    "            well_contents[(well_row, str(well_column))] = [WellContent(genotype, cell)]\n",
    "    return well_contents\n",
    "\n",
    "def parse_multiplexed_well_block(sheet, start, expected_genotypes, offsets):\n",
    "    well_re = re.compile(\"^[A-H]1?[\\d]$\")\n",
    "    wells = sheet.loc[start+3, offsets[\"well_range\"]].tolist()\n",
    "    for cell in wells:\n",
    "        if pandas.isnull(cell) or well_re.match(cell) is None:\n",
    "            raise ValueError(f\"well value {cell} in row {start+3} does not look correct\")\n",
    "\n",
    "    contents = sheet.loc[start+0:start+1, offsets[\"well_range\"]]\n",
    "    contents.columns = [(x[0], x[1:]) for x in wells]\n",
    "    validate_tissue_ids(contents, expected_genotypes)\n",
    "    contents.index = expected_genotypes\n",
    "\n",
    "    well_contents = {}\n",
    "    for genotype, row in contents.iterrows():\n",
    "        for well, tissue_id in row.iteritems():\n",
    "            well_contents.setdefault(well, []).append(WellContent(genotype, tissue_id))\n",
    "\n",
    "    return well_contents\n",
    "\n",
    "def parse_plate(sheet):\n",
    "    max_rows_from_plate_name_to_well_names = 4\n",
    "\n",
    "    offsets = {\n",
    "        \"plate_label\": 1,\n",
    "        \"tissue_label\": 1,\n",
    "        \"well_row_label\": 2,\n",
    "        \"well_start\": 3,\n",
    "        \"well_range\": slice(3, 10)\n",
    "    }\n",
    "\n",
    "    for plate_name, plate_start in find_plate_start(sheet, offsets):\n",
    "        well_contents = {}\n",
    "        genotypes = get_plate_genotype(sheet, plate_start, offsets)\n",
    "        well_contents.update(parse_single_well_block(sheet, plate_start, offsets))\n",
    "        \n",
    "        complex_blocks = [\n",
    "            (plate_start + 12, slice(0, 2)),\n",
    "            (plate_start + 12 + 5, slice(2, 4)),\n",
    "            (plate_start + 12 + 10, slice(4, 6)),\n",
    "            (plate_start + 12 + 15, slice(6, 8)),\n",
    "        ]\n",
    "        \n",
    "        for complex_start, genotype_slice in complex_blocks:\n",
    "            expected_genotypes = genotypes[genotype_slice]\n",
    "            well_contents.update(parse_multiplexed_well_block(sheet, complex_start, expected_genotypes, offsets))\n",
    "        \n",
    "        yield (plate_name, well_contents)\n",
    "        \n",
    "plate_layout = pandas.read_excel(\n",
    "    spreadsheet_name, \n",
    "    sheet_name=\"Plate setups\",\n",
    "    header=None,\n",
    ")\n",
    "\n",
    "\n",
    "# Validate tissue references:\n",
    "failed = False\n",
    "for plate_name, plate_contents in parse_plate(plate_layout):\n",
    "    for well_id in plate_contents:\n",
    "        well_contents = plate_contents[well_id]\n",
    "        \n",
    "        for well_fraction in well_contents:\n",
    "            try:\n",
    "                biosample = models.Tissue.objects.get(name=well_fraction.tissue_id)\n",
    "            except models.Tissue.DoesNotExist:\n",
    "                print(f\"Unable to find {well_fraction.tissue_id} on plate {plate_name}\")\n",
    "                failed = True\n",
    "                \n",
    "assert not failed, \"Resolve tissues\"\n",
    "\n",
    "# populate database\n",
    "if models.SplitSeqWell.biosample.through.objects.count() > 0:\n",
    "    truncate(models.SplitSeqWell.biosample.through)\n",
    "\n",
    "if models.SplitSeqWell.objects.count() > 0:\n",
    "    truncate(models.SplitSeqWell)\n",
    "    \n",
    "if models.SplitSeqPlate.objects.count() > 0:\n",
    "    truncate(models.SplitSeqPlate)\n",
    "\n",
    "wt_mega_2_kit = models.LibraryConstructionKit.objects.get(name=\"WT Mega\", version=\"v2\")\n",
    "\n",
    "for plate_name, plate_contents in parse_plate(plate_layout):\n",
    "    plate_record = models.SplitSeqPlate(\n",
    "        name=plate_name,\n",
    "        size=models.PlateSizeEnum.size_96,\n",
    "        pool_location=None,\n",
    "        date_performed=None,\n",
    "    )\n",
    "    plate_record.save()\n",
    "    \n",
    "    for well_id in plate_contents:\n",
    "        well_contents = plate_contents[well_id]\n",
    "        \n",
    "        biosamples = []\n",
    "        for well_content in well_contents:\n",
    "            try:\n",
    "                biosamples.append(models.FixedSample.objects.get(name=well_content.tissue_id))\n",
    "            except models.Tissue.DoesNotExist:\n",
    "                print(f\"unable to find tissue {well_content.tissue_id} for {plate_name} {well_id}\")\n",
    "        \n",
    "        barcodes = models.LibraryBarcode.objects.filter(\n",
    "            kit=wt_mega_2_kit,\n",
    "            code=\"{}{}\".format(well_id[0], well_id[1]),\n",
    "        )\n",
    "        \n",
    "        well_record = models.SplitSeqWell(\n",
    "            plate=plate_record,\n",
    "            row=well_id[0],\n",
    "            column=well_id[1],\n",
    "        )\n",
    "        well_record.save()\n",
    "        well_record.biosample.set(biosamples)\n",
    "        well_record.barcode.set(barcodes)\n",
    "        well_record.save()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "endless-compromise",
   "metadata": {},
   "source": [
    "# Sublibraries\n",
    "\n",
    "## Extract meaning from the Experiments tab"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "permanent-italic",
   "metadata": {},
   "source": [
    "- number of nuclei per sublibrary (Experiment) \n",
    "- cDNA amp # PCR cycles (Experiment)\n",
    "- cDNA ng/ul in 25ul (experiment)\n",
    "- total cDNA ng (experiment)\n",
    "- Bioanalyzer cDNA ave bp length (expierment)\n",
    "- Sub library Index PCR # (experiment)\n",
    "- Sublibrary Index (experiment)\n",
    "- Sequence for SampleSheet (experiment)\n",
    "- library ng/ul (experiment)\n",
    "- Bioanalyzer library ave bp length (experiment)\n",
    "- Nextseq run number (experiment)\n",
    "\n",
    "- number of 67k aliquots  (experiment)\n",
    "- number of 8k aliquots (experiment)\n",
    "\n",
    "- QC # raw reads (1 mismatch) (experiment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "coordinated-destruction",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Experiment</th>\n",
       "      <th>mice</th>\n",
       "      <th>Sample Box</th>\n",
       "      <th>Fixation Box</th>\n",
       "      <th>Split-seq prep start date</th>\n",
       "      <th>barcoded [nuclei/ul] cell counter</th>\n",
       "      <th>Volume of barcoded nuclei</th>\n",
       "      <th>Total barcoded nuclei (1st 3 rounds)</th>\n",
       "      <th># of backup 67k aliquots</th>\n",
       "      <th># of backup 8k aliquots</th>\n",
       "      <th>...</th>\n",
       "      <th>Unnamed: 39</th>\n",
       "      <th>Novaseq raw reads</th>\n",
       "      <th>Novaseq L001_1</th>\n",
       "      <th>Novaseq L002_1</th>\n",
       "      <th>Novaseq L003_1</th>\n",
       "      <th>Novaseq L004_1</th>\n",
       "      <th>Novaseq L001_2</th>\n",
       "      <th>Novaseq L002_2</th>\n",
       "      <th>Novaseq L003_2</th>\n",
       "      <th>Novaseq L004_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IGVF_Splitseq_002</td>\n",
       "      <td>B6 (2F+2M)</td>\n",
       "      <td>IGVF_Sample_3</td>\n",
       "      <td>IGVF_FIX_001</td>\n",
       "      <td>2022-08-24 00:00:00</td>\n",
       "      <td>4540</td>\n",
       "      <td>240.0</td>\n",
       "      <td>1089600.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>CAST (2F+2M)</td>\n",
       "      <td>IGVF_Sample_4</td>\n",
       "      <td>IGVF_FIX_002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Fixation V1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1M V1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1000</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1001 rows Ã— 49 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Experiment          mice     Sample Box  Fixation Box  \\\n",
       "0     IGVF_Splitseq_002    B6 (2F+2M)  IGVF_Sample_3  IGVF_FIX_001   \n",
       "1                   NaN  CAST (2F+2M)  IGVF_Sample_4  IGVF_FIX_002   \n",
       "2           Fixation V1           NaN            NaN           NaN   \n",
       "3                 1M V1           NaN            NaN           NaN   \n",
       "4                   NaN           NaN            NaN           NaN   \n",
       "...                 ...           ...            ...           ...   \n",
       "996                 NaN           NaN            NaN           NaN   \n",
       "997                 NaN           NaN            NaN           NaN   \n",
       "998                 NaN           NaN            NaN           NaN   \n",
       "999                 NaN           NaN            NaN           NaN   \n",
       "1000                NaN           NaN            NaN           NaN   \n",
       "\n",
       "     Split-seq prep start date barcoded [nuclei/ul] cell counter  \\\n",
       "0          2022-08-24 00:00:00                              4540   \n",
       "1                          NaN                               NaN   \n",
       "2                          NaN                               NaN   \n",
       "3                          NaN                               NaN   \n",
       "4                          NaN                               NaN   \n",
       "...                        ...                               ...   \n",
       "996                        NaN                               NaN   \n",
       "997                        NaN                               NaN   \n",
       "998                        NaN                               NaN   \n",
       "999                        NaN                               NaN   \n",
       "1000                       NaN                               NaN   \n",
       "\n",
       "      Volume of barcoded nuclei  Total barcoded nuclei (1st 3 rounds)  \\\n",
       "0                         240.0                             1089600.0   \n",
       "1                           NaN                                   NaN   \n",
       "2                           NaN                                   NaN   \n",
       "3                           NaN                                   NaN   \n",
       "4                           NaN                                   NaN   \n",
       "...                         ...                                   ...   \n",
       "996                         NaN                                   NaN   \n",
       "997                         NaN                                   NaN   \n",
       "998                         NaN                                   NaN   \n",
       "999                         NaN                                   NaN   \n",
       "1000                        NaN                                   NaN   \n",
       "\n",
       "      # of backup 67k aliquots  # of backup 8k aliquots  ...  Unnamed: 39  \\\n",
       "0                          3.0                      0.0  ...          NaN   \n",
       "1                          NaN                      NaN  ...          NaN   \n",
       "2                          NaN                      NaN  ...          NaN   \n",
       "3                          NaN                      NaN  ...          NaN   \n",
       "4                          NaN                      NaN  ...          NaN   \n",
       "...                        ...                      ...  ...          ...   \n",
       "996                        NaN                      NaN  ...          NaN   \n",
       "997                        NaN                      NaN  ...          NaN   \n",
       "998                        NaN                      NaN  ...          NaN   \n",
       "999                        NaN                      NaN  ...          NaN   \n",
       "1000                       NaN                      NaN  ...          NaN   \n",
       "\n",
       "      Novaseq raw reads  Novaseq L001_1 Novaseq L002_1 Novaseq L003_1  \\\n",
       "0                   NaN             NaN            NaN            NaN   \n",
       "1                   NaN             NaN            NaN            NaN   \n",
       "2                   NaN             NaN            NaN            NaN   \n",
       "3                   NaN             NaN            NaN            NaN   \n",
       "4                   NaN             NaN            NaN            NaN   \n",
       "...                 ...             ...            ...            ...   \n",
       "996                 NaN             NaN            NaN            NaN   \n",
       "997                 NaN             NaN            NaN            NaN   \n",
       "998                 NaN             NaN            NaN            NaN   \n",
       "999                 NaN             NaN            NaN            NaN   \n",
       "1000                NaN             NaN            NaN            NaN   \n",
       "\n",
       "      Novaseq L004_1  Novaseq L001_2  Novaseq L002_2  Novaseq L003_2  \\\n",
       "0                NaN             NaN             NaN             NaN   \n",
       "1                NaN             NaN             NaN             NaN   \n",
       "2                NaN             NaN             NaN             NaN   \n",
       "3                NaN             NaN             NaN             NaN   \n",
       "4                NaN             NaN             NaN             NaN   \n",
       "...              ...             ...             ...             ...   \n",
       "996              NaN             NaN             NaN             NaN   \n",
       "997              NaN             NaN             NaN             NaN   \n",
       "998              NaN             NaN             NaN             NaN   \n",
       "999              NaN             NaN             NaN             NaN   \n",
       "1000             NaN             NaN             NaN             NaN   \n",
       "\n",
       "     Novaseq L004_2  \n",
       "0               NaN  \n",
       "1               NaN  \n",
       "2               NaN  \n",
       "3               NaN  \n",
       "4               NaN  \n",
       "...             ...  \n",
       "996             NaN  \n",
       "997             NaN  \n",
       "998             NaN  \n",
       "999             NaN  \n",
       "1000            NaN  \n",
       "\n",
       "[1001 rows x 49 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "experiment = pandas.read_excel(\n",
    "    spreadsheet_name, \n",
    "    sheet_name=\"Experiment\",\n",
    "    header=0,\n",
    "    index_col=None,\n",
    "    usecols=range(0, 49)\n",
    ").rename(columns={\n",
    "    \"Unnamed: 40\": \"Novaseq raw reads\",\n",
    "    \"Unnamed: 41\": \"Novaseq L001_1\",\n",
    "    \"Unnamed: 42\": \"Novaseq L002_1\",\n",
    "    \"Unnamed: 43\": \"Novaseq L003_1\",\n",
    "    \"Unnamed: 44\": \"Novaseq L004_1\",\n",
    "    \"Unnamed: 45\": \"Novaseq L001_2\",\n",
    "    \"Unnamed: 46\": \"Novaseq L002_2\",\n",
    "    \"Unnamed: 47\": \"Novaseq L003_2\",\n",
    "    \"Unnamed: 48\": \"Novaseq L004_2\",\n",
    "})\n",
    "\n",
    "experiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "gothic-robert",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('IGVF_Splitseq_002', 0),\n",
       " ('IGVF_Splitseq_003', 17),\n",
       " ('IGVF_Splitseq_004', 35),\n",
       " ('IGVF_Splitseq_005', 53),\n",
       " ('IGVF_Splitseq_006', 71),\n",
       " ('IGVF_Splitseq_007', 90),\n",
       " ('IGVF_Splitseq_008', 108),\n",
       " ('IGVF_Splitseq_009', 127)]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def find_fixation_start(sheet):\n",
    "    for fixation_id_row in sheet[sheet[\"Experiment\"].apply(lambda x: not pandas.isnull(x) and x.startswith(\"IGVF_Splitseq\"))].index:\n",
    "        yield (sheet.loc[fixation_id_row, \"Experiment\"],fixation_id_row)\n",
    "\n",
    "list(find_fixation_start(experiment))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "naval-muslim",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'experiment_name': 'IGVF_Splitseq_003',\n",
       " 'plate_name': 'IGVF_003',\n",
       " 'prep_date': datetime.datetime(2022, 12, 1, 0, 0),\n",
       " 'aliquots_8k': 3,\n",
       " 'aliquots_67k': 22,\n",
       " 'sublibrary': [{'name': '003_8A',\n",
       "   'nuclei': 8000,\n",
       "   'cdna_pcr_rounds': '5 + 10',\n",
       "   'cdna_ng_per_ul_in_25ul': 130.0,\n",
       "   'bioanalyzer_date': Timestamp('2022-12-05 00:00:00'),\n",
       "   'cdna_average_bp_length': 1150,\n",
       "   'index_pcr_no': 12,\n",
       "   'index': '1',\n",
       "   'barcode': 'CAGATC',\n",
       "   'library_ng_per_ul': 11.8,\n",
       "   'library_average_bp_length': 443.0,\n",
       "   'nextseq_run_date': datetime.date(2022, 12, 6),\n",
       "   'nextseq_run_id': 1049,\n",
       "   'nextseq_run_raw_reads': 112237518,\n",
       "   'novaseq_run_date': None,\n",
       "   'novaseq_run_id': None,\n",
       "   'novaseq_run_raw_reads': None},\n",
       "  {'name': '003_67A',\n",
       "   'nuclei': 67000,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul_in_25ul': 239.0,\n",
       "   'bioanalyzer_date': Timestamp('2022-12-05 00:00:00'),\n",
       "   'cdna_average_bp_length': 1149,\n",
       "   'index_pcr_no': 12,\n",
       "   'index': '2',\n",
       "   'barcode': 'ACTTGA',\n",
       "   'library_ng_per_ul': 30.6,\n",
       "   'library_average_bp_length': 413.0,\n",
       "   'nextseq_run_date': datetime.date(2022, 12, 6),\n",
       "   'nextseq_run_id': 1049,\n",
       "   'nextseq_run_raw_reads': 620332161,\n",
       "   'novaseq_run_date': datetime.date(2022, 12, 16),\n",
       "   'novaseq_run_id': '2022-12-16',\n",
       "   'novaseq_run_raw_reads': 1187040511},\n",
       "  {'name': '003_67B',\n",
       "   'nuclei': 67000,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul_in_25ul': 258.0,\n",
       "   'bioanalyzer_date': Timestamp('2022-12-05 00:00:00'),\n",
       "   'cdna_average_bp_length': 1108,\n",
       "   'index_pcr_no': 12,\n",
       "   'index': '3',\n",
       "   'barcode': 'GATCAG',\n",
       "   'library_ng_per_ul': 25.2,\n",
       "   'library_average_bp_length': 430.0,\n",
       "   'nextseq_run_date': datetime.date(2022, 12, 6),\n",
       "   'nextseq_run_id': 1049,\n",
       "   'nextseq_run_raw_reads': 563917519,\n",
       "   'novaseq_run_date': datetime.date(2022, 12, 16),\n",
       "   'novaseq_run_id': '2022-12-16',\n",
       "   'novaseq_run_raw_reads': 1153553060},\n",
       "  {'name': '003_67C',\n",
       "   'nuclei': 67000,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul_in_25ul': 237.0,\n",
       "   'bioanalyzer_date': Timestamp('2022-12-05 00:00:00'),\n",
       "   'cdna_average_bp_length': 1148,\n",
       "   'index_pcr_no': 12,\n",
       "   'index': '4',\n",
       "   'barcode': 'TAGCTT',\n",
       "   'library_ng_per_ul': 20.4,\n",
       "   'library_average_bp_length': 431.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': datetime.date(2022, 12, 16),\n",
       "   'novaseq_run_id': '2022-12-16',\n",
       "   'novaseq_run_raw_reads': 1409040245},\n",
       "  {'name': '003_67D',\n",
       "   'nuclei': 67000,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul_in_25ul': 336.0,\n",
       "   'bioanalyzer_date': Timestamp('2022-12-05 00:00:00'),\n",
       "   'cdna_average_bp_length': 1119,\n",
       "   'index_pcr_no': 12,\n",
       "   'index': '5',\n",
       "   'barcode': 'ATGTCA',\n",
       "   'library_ng_per_ul': 20.2,\n",
       "   'library_average_bp_length': 438.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': datetime.date(2022, 12, 16),\n",
       "   'novaseq_run_id': '2022-12-16',\n",
       "   'novaseq_run_raw_reads': 1344906430},\n",
       "  {'name': '003_67E',\n",
       "   'nuclei': 67000,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul_in_25ul': 284.0,\n",
       "   'bioanalyzer_date': Timestamp('2022-12-05 00:00:00'),\n",
       "   'cdna_average_bp_length': 1140,\n",
       "   'index_pcr_no': 12,\n",
       "   'index': '6',\n",
       "   'barcode': 'CTTGTA',\n",
       "   'library_ng_per_ul': 18.1,\n",
       "   'library_average_bp_length': 431.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': datetime.date(2022, 12, 16),\n",
       "   'novaseq_run_id': '2022-12-16',\n",
       "   'novaseq_run_raw_reads': 1207172770},\n",
       "  {'name': '003_67F',\n",
       "   'nuclei': 67000,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul_in_25ul': 343.0,\n",
       "   'bioanalyzer_date': Timestamp('2022-12-05 00:00:00'),\n",
       "   'cdna_average_bp_length': 1139,\n",
       "   'index_pcr_no': 12,\n",
       "   'index': '7',\n",
       "   'barcode': 'AGTCAA',\n",
       "   'library_ng_per_ul': 22.2,\n",
       "   'library_average_bp_length': 417.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': datetime.date(2022, 12, 16),\n",
       "   'novaseq_run_id': '2022-12-16',\n",
       "   'novaseq_run_raw_reads': 1432433000},\n",
       "  {'name': '003_67G',\n",
       "   'nuclei': 67000,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul_in_25ul': 272.0,\n",
       "   'bioanalyzer_date': Timestamp('2022-12-05 00:00:00'),\n",
       "   'cdna_average_bp_length': 1104,\n",
       "   'index_pcr_no': 12,\n",
       "   'index': '8',\n",
       "   'barcode': 'AGTTCC',\n",
       "   'library_ng_per_ul': 13.8,\n",
       "   'library_average_bp_length': 449.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': datetime.date(2022, 12, 16),\n",
       "   'novaseq_run_id': '2022-12-16',\n",
       "   'novaseq_run_raw_reads': 1291242781},\n",
       "  {'name': '003_67H',\n",
       "   'nuclei': 67000,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul_in_25ul': 304.0,\n",
       "   'bioanalyzer_date': Timestamp('2022-12-05 00:00:00'),\n",
       "   'cdna_average_bp_length': 1128,\n",
       "   'index_pcr_no': 12,\n",
       "   'index': '9',\n",
       "   'barcode': 'GAGTGG',\n",
       "   'library_ng_per_ul': 19.8,\n",
       "   'library_average_bp_length': 421.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': datetime.date(2022, 12, 16),\n",
       "   'novaseq_run_id': '2022-12-16',\n",
       "   'novaseq_run_raw_reads': 1185598644},\n",
       "  {'name': '003_67I',\n",
       "   'nuclei': 67000,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul_in_25ul': 217.0,\n",
       "   'bioanalyzer_date': Timestamp('2022-12-05 00:00:00'),\n",
       "   'cdna_average_bp_length': 1131,\n",
       "   'index_pcr_no': 12,\n",
       "   'index': '10',\n",
       "   'barcode': 'CCGTCC',\n",
       "   'library_ng_per_ul': 14.9,\n",
       "   'library_average_bp_length': 442.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': datetime.date(2022, 12, 16),\n",
       "   'novaseq_run_id': '2022-12-16',\n",
       "   'novaseq_run_raw_reads': 1233335896},\n",
       "  {'name': '003_67J',\n",
       "   'nuclei': 67000,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul_in_25ul': 196.0,\n",
       "   'bioanalyzer_date': Timestamp('2022-12-05 00:00:00'),\n",
       "   'cdna_average_bp_length': 1103,\n",
       "   'index_pcr_no': 12,\n",
       "   'index': '11',\n",
       "   'barcode': 'GTAGAG',\n",
       "   'library_ng_per_ul': 21.2,\n",
       "   'library_average_bp_length': 428.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': datetime.date(2022, 12, 16),\n",
       "   'novaseq_run_id': '2022-12-16',\n",
       "   'novaseq_run_raw_reads': 1335037511},\n",
       "  {'name': '003_67K',\n",
       "   'nuclei': 67000,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul_in_25ul': 249.0,\n",
       "   'bioanalyzer_date': Timestamp('2022-12-05 00:00:00'),\n",
       "   'cdna_average_bp_length': 1188,\n",
       "   'index_pcr_no': 12,\n",
       "   'index': '12',\n",
       "   'barcode': 'GTCCGC',\n",
       "   'library_ng_per_ul': 13.3,\n",
       "   'library_average_bp_length': 463.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': datetime.date(2022, 12, 16),\n",
       "   'novaseq_run_id': '2022-12-16',\n",
       "   'novaseq_run_raw_reads': 1219423479},\n",
       "  {'name': '003_67L',\n",
       "   'nuclei': 67000,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul_in_25ul': 213.0,\n",
       "   'bioanalyzer_date': Timestamp('2022-12-05 00:00:00'),\n",
       "   'cdna_average_bp_length': 1164,\n",
       "   'index_pcr_no': 12,\n",
       "   'index': '13',\n",
       "   'barcode': 'GTGAAA',\n",
       "   'library_ng_per_ul': 14.6,\n",
       "   'library_average_bp_length': 426.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': datetime.date(2022, 12, 16),\n",
       "   'novaseq_run_id': '2022-12-16',\n",
       "   'novaseq_run_raw_reads': 1529723303},\n",
       "  {'name': '003_67M',\n",
       "   'nuclei': 67000,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul_in_25ul': 340.0,\n",
       "   'bioanalyzer_date': Timestamp('2022-12-05 00:00:00'),\n",
       "   'cdna_average_bp_length': 1151,\n",
       "   'index_pcr_no': 12,\n",
       "   'index': '14',\n",
       "   'barcode': 'GTGGCC',\n",
       "   'library_ng_per_ul': 8.86,\n",
       "   'library_average_bp_length': 432.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': datetime.date(2022, 12, 16),\n",
       "   'novaseq_run_id': '2022-12-16',\n",
       "   'novaseq_run_raw_reads': 1307580675},\n",
       "  {'name': '003_67N',\n",
       "   'nuclei': 67000,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul_in_25ul': 317.0,\n",
       "   'bioanalyzer_date': Timestamp('2022-12-05 00:00:00'),\n",
       "   'cdna_average_bp_length': 1119,\n",
       "   'index_pcr_no': 12,\n",
       "   'index': '15',\n",
       "   'barcode': 'GTTTCG',\n",
       "   'library_ng_per_ul': 11.6,\n",
       "   'library_average_bp_length': 424.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': datetime.date(2022, 12, 16),\n",
       "   'novaseq_run_id': '2022-12-16',\n",
       "   'novaseq_run_raw_reads': 1060418846},\n",
       "  {'name': '003_67O',\n",
       "   'nuclei': 67000,\n",
       "   'cdna_pcr_rounds': '5 + 7',\n",
       "   'cdna_ng_per_ul_in_25ul': 255.0,\n",
       "   'bioanalyzer_date': Timestamp('2022-12-05 00:00:00'),\n",
       "   'cdna_average_bp_length': 1157,\n",
       "   'index_pcr_no': 12,\n",
       "   'index': '16',\n",
       "   'barcode': 'CGTACG',\n",
       "   'library_ng_per_ul': 12.8,\n",
       "   'library_average_bp_length': 413.0,\n",
       "   'nextseq_run_date': None,\n",
       "   'nextseq_run_id': None,\n",
       "   'nextseq_run_raw_reads': None,\n",
       "   'novaseq_run_date': datetime.date(2022, 12, 16),\n",
       "   'novaseq_run_id': '2022-12-16',\n",
       "   'novaseq_run_raw_reads': 1302660956}]}"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def parse_fixation(sheet, experiment_name, start):\n",
    "    block = sheet.loc[start: start+15]\n",
    "\n",
    "    experiment_name_fields = experiment_name.split(\"_\")\n",
    "    plate_name = f\"{experiment_name_fields[0]}_{experiment_name_fields[-1]}\"\n",
    "    \n",
    "    experiment = {\n",
    "        \"experiment_name\": experiment_name,\n",
    "        \"plate_name\": plate_name,\n",
    "        \"prep_date\": block.loc[start, \"Split-seq prep start date\"],\n",
    "        \"aliquots_8k\": int_or_0(block.loc[start, \"# of backup 8k aliquots\"]),\n",
    "        \"aliquots_67k\": int_or_0(block.loc[start, \"# of backup 67k aliquots\"]),\n",
    "        \"sublibrary\": [],\n",
    "    }\n",
    "\n",
    "    for i, row in block.iterrows():\n",
    "        if pandas.isnull(row[\"NovaSeq Sequencing date\"]):\n",
    "            novaseq_run_id = None\n",
    "        else:\n",
    "            novaseq_run_id = row[\"NovaSeq Sequencing date\"].isoformat().split(\"T\")[0]\n",
    "\n",
    "        if pandas.isnull(row[\"Novaseq raw reads\"]):\n",
    "            novaseq_raw_reads = None\n",
    "        elif row[\"Novaseq raw reads\"] == \"Total\":\n",
    "            novaseq_raw_reads = None\n",
    "        else:\n",
    "            novaseq_raw_reads = int(row[\"Novaseq raw reads\"])\n",
    "        \n",
    "        sublibrary = {\n",
    "            \"name\": row[\"sublibrary\"],\n",
    "            \"nuclei\": int_or_none(row[\"# of nuclei per sublibrary\"]),\n",
    "            \"cdna_pcr_rounds\": row[\"cDNA amp # PCR cycles\"],\n",
    "            \"cdna_ng_per_ul_in_25ul\": row[\"[cDNA] ng/ul     in 25ul\"],\n",
    "            #total_cdna = cdna_ng_per_ul_in_25ul * 25\n",
    "            \"bioanalyzer_date\": row[\"BA date\"],\n",
    "            \"cdna_average_bp_length\": int_or_none(row[\"Bioanalyzer cDNA ave bp length\"]),\n",
    "            \"index_pcr_no\": int_or_none(row[\"Sub library Index PCR #\"]),\n",
    "            \"index\": str(int_or_none(row[\"Sublibrary Index\"])),\n",
    "            \"barcode\": row[\"Sequence for SampleSheet\"],\n",
    "            \"library_ng_per_ul\": float_or_none(row[\"[library] ng/ul\"]),\n",
    "            \"library_average_bp_length\": float_or_none(row[\"Bioanalyzer library ave bp length\"]),\n",
    "            \"nextseq_run_date\": date_or_none(row[\"QC Sequencing date\"]),\n",
    "            \"nextseq_run_id\": str_or_none(int_or_none(row[\"Nextseq run #\"])),\n",
    "            \"nextseq_run_raw_reads\": int_or_none(row[\"QC # raw reads  (1 mismatch)\"]),\n",
    "            \"novaseq_run_date\": date_or_none(row[\"NovaSeq Sequencing date\"]),\n",
    "            \"novaseq_run_id\": novaseq_run_id,\n",
    "            \"novaseq_run_raw_reads\": novaseq_raw_reads,\n",
    "        }\n",
    "        experiment[\"sublibrary\"].append(sublibrary)\n",
    "        if pandas.isnull(sublibrary[\"nuclei\"]):\n",
    "            continue\n",
    "        elif sublibrary[\"nuclei\"] < 10000:\n",
    "            experiment[\"aliquots_8k\"] += 1\n",
    "        elif sublibrary[\"nuclei\"] > 60000 and sublibrary[\"nuclei\"] < 70000:\n",
    "            experiment[\"aliquots_67k\"] += 1\n",
    "        else:\n",
    "            print(f\"Unusual number of nuclei {sublibrary['name']} {sublibrary['nuclei']}\")\n",
    "    return experiment\n",
    "\n",
    "parse_fixation(experiment, \"IGVF_Splitseq_003\", 17)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "victorian-basics",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unable to find plate IGVF_002\n",
      "Unable to find plate IGVF_007\n",
      "Unable to find plate IGVF_008\n",
      "Unable to find plate IGVF_009\n"
     ]
    }
   ],
   "source": [
    "failed = False\n",
    "for experiment_name, experiment_start in find_fixation_start(experiment):\n",
    "    fixation = parse_fixation(experiment, experiment_name, experiment_start)\n",
    "    \n",
    "    try:\n",
    "        plate = models.SplitSeqPlate.objects.get(name=fixation[\"plate_name\"])\n",
    "    except models.SplitSeqPlate.DoesNotExist:\n",
    "        print(f\"Unable to find plate {fixation['plate_name']}\")\n",
    "        continue\n",
    "        \n",
    "    plate.aliquots_small_made = fixation[\"aliquots_8k\"]\n",
    "    plate.aliquots_large_made = fixation[\"aliquots_67k\"]\n",
    "    plate.save()\n",
    "    \n",
    "    wt_mega_2_kit = models.LibraryConstructionKit.objects.get(name=\"WT Mega\", version=\"v2\")\n",
    "    for sublibrary in fixation[\"sublibrary\"]:\n",
    "        try:\n",
    "            barcodes = models.LibraryBarcode.objects.filter(kit=wt_mega_2_kit, barcode_type=None, code=sublibrary[\"index\"])\n",
    "        except models.LibraryBarcode.DoesNotExist:\n",
    "            print(f\"Unable to find barcode {sublibrary['index']}\")\n",
    "\n",
    "        # validate barcodes\n",
    "        library_barcodes_sequence = {b.sequence for b in barcodes}\n",
    "        expected_sequence = set(sublibrary[\"barcode\"].split(\",\"))\n",
    "        if library_barcodes_sequence != expected_sequence:\n",
    "            print(f\"Database lookup of barcodes {library_barcodes_sequence} doesn't match human entry {expected_sequence}\")\n",
    "            failed = True\n",
    "            continue\n",
    "        \n",
    "        record = models.Sublibrary(\n",
    "            name=sublibrary[\"name\"],\n",
    "            plate=plate,\n",
    "            nuclei=sublibrary[\"nuclei\"],\n",
    "            cdna_pcr_rounds=sublibrary[\"cdna_pcr_rounds\"],\n",
    "            cdna_ng_per_ul_in_25ul=float_or_none(sublibrary[\"cdna_ng_per_ul_in_25ul\"]),\n",
    "            bioanalyzer_date=date_or_none(sublibrary[\"bioanalyzer_date\"]),\n",
    "            cdna_average_bp_length=int_or_none(sublibrary[\"cdna_average_bp_length\"]),\n",
    "            index_pcr_number=int_or_none(sublibrary[\"index_pcr_no\"]),\n",
    "            index=str(int_or_none(sublibrary[\"index\"])),\n",
    "            library_ng_per_ul=float_or_none(sublibrary[\"library_ng_per_ul\"]),\n",
    "            library_average_bp_length=int_or_none(sublibrary[\"library_average_bp_length\"]),\n",
    "        )\n",
    "        record.save()\n",
    "        record.barcode.set(barcodes)\n",
    "        record.save()\n",
    "        \n",
    "        if sublibrary[\"nextseq_run_date\"] is not None:\n",
    "            run = models.SequencingRun(\n",
    "                platform=\"NE\",\n",
    "                name=sublibrary[\"nextseq_run_id\"],\n",
    "                library=record,\n",
    "                run_date=sublibrary[\"nextseq_run_date\"],\n",
    "                raw_reads=sublibrary[\"nextseq_run_raw_reads\"],\n",
    "            )\n",
    "            run.save()\n",
    "            \n",
    "        if sublibrary[\"novaseq_run_date\"] is not None:\n",
    "            run = models.SequencingRun(\n",
    "                platform=\"NO\",\n",
    "                name=sublibrary[\"novaseq_run_id\"],\n",
    "                library=record,\n",
    "                run_date=sublibrary[\"novaseq_run_date\"],\n",
    "                raw_reads=sublibrary[\"novaseq_run_raw_reads\"],\n",
    "            )\n",
    "            run.save()\n",
    "      \n",
    "assert not failed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "marked-probe",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
