{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "floating-raising",
   "metadata": {},
   "source": [
    "# Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "round-television",
   "metadata": {},
   "source": [
    "Using the models we have try to fill in the metadata spreadsheet & submit test records to the IGVF test portal.\n",
    "\n",
    "This try had all the files from each platform type attached to one measurement set. But with Jennifer we decided this is probably a poor idea."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "intended-gender",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "educational-vitamin",
   "metadata": {},
   "outputs": [],
   "source": [
    "import bz2\n",
    "from collections import Counter, namedtuple\n",
    "import datetime\n",
    "import numpy\n",
    "import os\n",
    "import pandas\n",
    "from pathlib import Path\n",
    "import re\n",
    "from subprocess import run, PIPE\n",
    "import sys\n",
    "import zoneinfo\n",
    "\n",
    "os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'mousedemo.settings')\n",
    "os.environ[\"DJANGO_ALLOW_ASYNC_UNSAFE\"] = \"true\"\n",
    "\n",
    "import django\n",
    "from django.contrib.auth import get_user_model\n",
    "from django.db import DEFAULT_DB_ALIAS\n",
    "\n",
    "MOUSEDEMO = str(Path(\"mousedemo\").absolute())\n",
    "if MOUSEDEMO not in sys.path:\n",
    "    sys.path.append(MOUSEDEMO)\n",
    "\n",
    "django.setup()\n",
    "\n",
    "from mousedemo import settings\n",
    "from igvf_mice import models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "signed-percentage",
   "metadata": {},
   "outputs": [],
   "source": [
    "EC = str(Path(\"~/proj/encoded_client\").expanduser())\n",
    "if EC not in sys.path:\n",
    "    sys.path.append(EC)\n",
    "    \n",
    "from encoded_client import encoded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "basic-recipient",
   "metadata": {},
   "outputs": [],
   "source": [
    "server = encoded.ENCODED(\"api.sandbox.igvf.org\")\n",
    "\n",
    "validator = encoded.DCCValidator(server)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "attractive-buffalo",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "representative-observer",
   "metadata": {},
   "outputs": [],
   "source": [
    "award = \"/awards/HG012077/\"\n",
    "labs = [\"/labs/lior-pachter/\", \"/labs/grant-macgregor/\", \"/labs/barbara-wold/\", \"/labs/ali-mortazavi/\"]\n",
    "lab = labs[-1]\n",
    "jax = \"/sources/jackson-labs/\"\n",
    "species = 'Mus musculus'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "psychological-sunday",
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_background = {\n",
    "    \"A/J\": \"A/J (AJ)\",\n",
    "    \"C57BL/6J\": \"C57BL/6J (B6)\",\n",
    "    \"129S1/SvImJ\": \"129S1/SvImJ (129)\",\n",
    "    \"NOD/ShiLtJ\": \"NOD/ShiLtJ (NOD)\",\n",
    "    \"NZO/HlLtJ\": \"NZO/H1LtJ (NZO)\",\n",
    "    \"CAST/EiJ\": \"CAST/EiJ (CAST)\",\n",
    "    \"PWK/PhJ\": \"PWK/PhJ (PWK)\",\n",
    "    \"WSB/EiJ\": \"WSB/EiJ (WSB)\",\n",
    "    #\"CAST (M. m. castaneus)\",\n",
    "    #\"WSB (M. m. domesticus)\",\n",
    "    #\"PWK (M. m. musculus)\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "charming-neighborhood",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "liable-singing",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_run =  \"IGVF_003\"\n",
    "plate_003 = models.SplitSeqPlate.objects.get(name=test_run)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "dimensional-reliance",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<QuerySet [<Sublibrary: 003_8A>]>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plate_003.sublibrary_set.filter(nuclei__lt=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "hidden-adventure",
   "metadata": {},
   "outputs": [],
   "source": [
    "tissues = {}\n",
    "mice = {}\n",
    "for well in plate_003.splitseqwell_set.all():\n",
    "    for biosample in well.biosample.all():\n",
    "        for tissue in biosample.tissue.all():\n",
    "            tissues[tissue.name] = tissue\n",
    "            mice[tissue.mouse.name] = tissue.mouse"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "permanent-expression",
   "metadata": {},
   "source": [
    "# rodent_donor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "welsh-safety",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>uuid</th>\n",
       "      <th>aliases:array</th>\n",
       "      <th>award</th>\n",
       "      <th>lab</th>\n",
       "      <th>taxa</th>\n",
       "      <th>sex</th>\n",
       "      <th>strain</th>\n",
       "      <th>references</th>\n",
       "      <th>url</th>\n",
       "      <th>...</th>\n",
       "      <th>submitter_comment</th>\n",
       "      <th>description</th>\n",
       "      <th>parents</th>\n",
       "      <th>traits</th>\n",
       "      <th>phenotypic_features</th>\n",
       "      <th>external_resources</th>\n",
       "      <th>strain_background</th>\n",
       "      <th>genotype</th>\n",
       "      <th>individual_rodent:boolean</th>\n",
       "      <th>rodent_identifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TSTDO36427294</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:016_B6J_10F</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>female</td>\n",
       "      <td>C57BL/6J</td>\n",
       "      <td>None</td>\n",
       "      <td>http://www.informatics.jax.org/inbred_strains/...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>C57BL/6J (B6)</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>016_B6J_10F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TSTDO96818159</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:017_B6J_10M</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>male</td>\n",
       "      <td>C57BL/6J</td>\n",
       "      <td>None</td>\n",
       "      <td>http://www.informatics.jax.org/inbred_strains/...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>C57BL/6J (B6)</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>017_B6J_10M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:018_B6J_10F</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>female</td>\n",
       "      <td>C57BL/6J</td>\n",
       "      <td>None</td>\n",
       "      <td>http://www.informatics.jax.org/inbred_strains/...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>C57BL/6J (B6)</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>018_B6J_10F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:019_B6J_10M</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>male</td>\n",
       "      <td>C57BL/6J</td>\n",
       "      <td>None</td>\n",
       "      <td>http://www.informatics.jax.org/inbred_strains/...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>C57BL/6J (B6)</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>019_B6J_10M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:020_B6J_10F</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>female</td>\n",
       "      <td>C57BL/6J</td>\n",
       "      <td>None</td>\n",
       "      <td>http://www.informatics.jax.org/inbred_strains/...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>C57BL/6J (B6)</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>020_B6J_10F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:090_CASTJ_10F</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>female</td>\n",
       "      <td>CAST/EiJ</td>\n",
       "      <td>None</td>\n",
       "      <td>http://www.informatics.jax.org/inbred_strains/...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>CAST/EiJ (CAST)</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>090_CASTJ_10F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:091_CASTJ_10M</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>male</td>\n",
       "      <td>CAST/EiJ</td>\n",
       "      <td>None</td>\n",
       "      <td>http://www.informatics.jax.org/inbred_strains/...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>CAST/EiJ (CAST)</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>091_CASTJ_10M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:093_CASTJ_10M</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>male</td>\n",
       "      <td>CAST/EiJ</td>\n",
       "      <td>None</td>\n",
       "      <td>http://www.informatics.jax.org/inbred_strains/...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>CAST/EiJ (CAST)</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>093_CASTJ_10M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:094_CASTJ_10F</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>female</td>\n",
       "      <td>CAST/EiJ</td>\n",
       "      <td>None</td>\n",
       "      <td>http://www.informatics.jax.org/inbred_strains/...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>CAST/EiJ (CAST)</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>094_CASTJ_10F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:096_WSBJ_10F</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>female</td>\n",
       "      <td>WSB/EiJ</td>\n",
       "      <td>None</td>\n",
       "      <td>https://www.informatics.jax.org/strain/MGI:216...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>WSB/EiJ (WSB)</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>096_WSBJ_10F</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession  uuid                aliases:array              award  \\\n",
       "0   TSTDO36427294  None    ali-mortazavi:016_B6J_10F  /awards/HG012077/   \n",
       "1   TSTDO96818159  None    ali-mortazavi:017_B6J_10M  /awards/HG012077/   \n",
       "2    would create  None    ali-mortazavi:018_B6J_10F  /awards/HG012077/   \n",
       "3    would create  None    ali-mortazavi:019_B6J_10M  /awards/HG012077/   \n",
       "4    would create  None    ali-mortazavi:020_B6J_10F  /awards/HG012077/   \n",
       "..            ...   ...                          ...                ...   \n",
       "59   would create  None  ali-mortazavi:090_CASTJ_10F  /awards/HG012077/   \n",
       "60   would create  None  ali-mortazavi:091_CASTJ_10M  /awards/HG012077/   \n",
       "61   would create  None  ali-mortazavi:093_CASTJ_10M  /awards/HG012077/   \n",
       "62   would create  None  ali-mortazavi:094_CASTJ_10F  /awards/HG012077/   \n",
       "63   would create  None   ali-mortazavi:096_WSBJ_10F  /awards/HG012077/   \n",
       "\n",
       "                     lab          taxa     sex    strain references  \\\n",
       "0   /labs/ali-mortazavi/  Mus musculus  female  C57BL/6J       None   \n",
       "1   /labs/ali-mortazavi/  Mus musculus    male  C57BL/6J       None   \n",
       "2   /labs/ali-mortazavi/  Mus musculus  female  C57BL/6J       None   \n",
       "3   /labs/ali-mortazavi/  Mus musculus    male  C57BL/6J       None   \n",
       "4   /labs/ali-mortazavi/  Mus musculus  female  C57BL/6J       None   \n",
       "..                   ...           ...     ...       ...        ...   \n",
       "59  /labs/ali-mortazavi/  Mus musculus  female  CAST/EiJ       None   \n",
       "60  /labs/ali-mortazavi/  Mus musculus    male  CAST/EiJ       None   \n",
       "61  /labs/ali-mortazavi/  Mus musculus    male  CAST/EiJ       None   \n",
       "62  /labs/ali-mortazavi/  Mus musculus  female  CAST/EiJ       None   \n",
       "63  /labs/ali-mortazavi/  Mus musculus  female   WSB/EiJ       None   \n",
       "\n",
       "                                                  url  ... submitter_comment  \\\n",
       "0   http://www.informatics.jax.org/inbred_strains/...  ...              None   \n",
       "1   http://www.informatics.jax.org/inbred_strains/...  ...              None   \n",
       "2   http://www.informatics.jax.org/inbred_strains/...  ...              None   \n",
       "3   http://www.informatics.jax.org/inbred_strains/...  ...              None   \n",
       "4   http://www.informatics.jax.org/inbred_strains/...  ...              None   \n",
       "..                                                ...  ...               ...   \n",
       "59  http://www.informatics.jax.org/inbred_strains/...  ...              None   \n",
       "60  http://www.informatics.jax.org/inbred_strains/...  ...              None   \n",
       "61  http://www.informatics.jax.org/inbred_strains/...  ...              None   \n",
       "62  http://www.informatics.jax.org/inbred_strains/...  ...              None   \n",
       "63  https://www.informatics.jax.org/strain/MGI:216...  ...              None   \n",
       "\n",
       "   description parents traits phenotypic_features external_resources  \\\n",
       "0         None    None   None                None               None   \n",
       "1         None    None   None                None               None   \n",
       "2         None    None   None                None               None   \n",
       "3         None    None   None                None               None   \n",
       "4         None    None   None                None               None   \n",
       "..         ...     ...    ...                 ...                ...   \n",
       "59        None    None   None                None               None   \n",
       "60        None    None   None                None               None   \n",
       "61        None    None   None                None               None   \n",
       "62        None    None   None                None               None   \n",
       "63        None    None   None                None               None   \n",
       "\n",
       "   strain_background genotype individual_rodent:boolean rodent_identifier  \n",
       "0      C57BL/6J (B6)     None                      True       016_B6J_10F  \n",
       "1      C57BL/6J (B6)     None                      True       017_B6J_10M  \n",
       "2      C57BL/6J (B6)     None                      True       018_B6J_10F  \n",
       "3      C57BL/6J (B6)     None                      True       019_B6J_10M  \n",
       "4      C57BL/6J (B6)     None                      True       020_B6J_10F  \n",
       "..               ...      ...                       ...               ...  \n",
       "59   CAST/EiJ (CAST)     None                      True     090_CASTJ_10F  \n",
       "60   CAST/EiJ (CAST)     None                      True     091_CASTJ_10M  \n",
       "61   CAST/EiJ (CAST)     None                      True     093_CASTJ_10M  \n",
       "62   CAST/EiJ (CAST)     None                      True     094_CASTJ_10F  \n",
       "63     WSB/EiJ (WSB)     None                      True      096_WSBJ_10F  \n",
       "\n",
       "[64 rows x 25 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def format_sex(value):\n",
    "    if models.SexEnum.MALE == value:\n",
    "        return \"male\"\n",
    "    elif models.SexEnum.FEMALE == value:\n",
    "        return \"female\"\n",
    "\n",
    "def get_accession_string_or_none(record):\n",
    "    accessions = record.accession.all()\n",
    "    if len(accessions) == 0:\n",
    "        return None\n",
    "    else:\n",
    "        return \",\".join([x.name for x in accessions])\n",
    "    \n",
    "rodent_donor = []\n",
    "for mouse_name in sorted(mice):\n",
    "    mouse = mice[mouse_name]\n",
    "    dcc_row = {\n",
    "        #\"#response\": None,\n",
    "        #\"#response_time\": None,\n",
    "        \"accession\": get_accession_string_or_none(mouse),\n",
    "        \"uuid\": None, \n",
    "        \"aliases:array\": f\"ali-mortazavi:{mouse.name}\",\n",
    "        \"award\": award,\n",
    "        \"lab\": lab,\n",
    "        \"taxa\": species,\n",
    "        \"sex\": format_sex(mouse.sex),\n",
    "        \"strain\": mouse.strain.name,\n",
    "        \"references\": None,\n",
    "        \"url\": mouse.strain.url,\n",
    "        \"source\": jax,\n",
    "        \"lot_id\": None,\n",
    "        \"product_id\": mouse.strain.jax_catalog_number,\n",
    "        \"documents\": None,\n",
    "        \"alternate_accessions\": None,\n",
    "        \"submitter_comment\": None,\n",
    "        \"description\": None,\n",
    "        \"parents\": None,\n",
    "        \"traits\": None,\n",
    "        \"phenotypic_features\": None,\n",
    "        \"external_resources\": None,\n",
    "        \"strain_background\": strain_background[mouse.strain.name],\n",
    "        \"genotype\": None,\n",
    "        \"individual_rodent:boolean\": True,\n",
    "        \"rodent_identifier\": mouse.name,\n",
    "    }\n",
    "    rodent_donor.append(dcc_row)\n",
    "    \n",
    "rodent_donor = pandas.DataFrame(rodent_donor)\n",
    "\n",
    "dry_run = True\n",
    "created = server.post_sheet(\"rodent_donor\", rodent_donor, dry_run=dry_run, verbose=True, validator=validator)\n",
    "if len(created) > 0 and not dry_run:\n",
    "    rodent_donor.to_excel(\"rodent_donor.xlsx\")\n",
    "rodent_donor"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "modern-diversity",
   "metadata": {},
   "source": [
    "# tissue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "approved-longer",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>uuid</th>\n",
       "      <th>aliases:array</th>\n",
       "      <th>award</th>\n",
       "      <th>lab</th>\n",
       "      <th>source</th>\n",
       "      <th>donors:array</th>\n",
       "      <th>taxa</th>\n",
       "      <th>biosample_term</th>\n",
       "      <th>term_names:skip</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TSTSM69861046</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:016_B6J_10F_03_NTR_0000646</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:016_B6J_10F</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>/sample-terms/NTR_0000646/</td>\n",
       "      <td>left cerebral cortex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TSTSM75831182</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:016_B6J_10F_03_NTR_0000750</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:016_B6J_10F</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>/sample-terms/NTR_0000750/</td>\n",
       "      <td>Hippocampal formation left</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TSTSM69542835</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:017_B6J_10M_03_NTR_0000646</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:017_B6J_10M</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>/sample-terms/NTR_0000646/</td>\n",
       "      <td>left cerebral cortex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TSTSM28050193</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:017_B6J_10M_03_NTR_0000750</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:017_B6J_10M</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>/sample-terms/NTR_0000750/</td>\n",
       "      <td>Hippocampal formation left</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:018_B6J_10F_03_NTR_0000646</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:018_B6J_10F</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>/sample-terms/NTR_0000646/</td>\n",
       "      <td>left cerebral cortex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:051_NZOJ_10M_06_UBERON_0000948</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:051_NZOJ_10M</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>/sample-terms/UBERON_0000948/</td>\n",
       "      <td>heart</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:096_WSBJ_10F_06_UBERON_0000948</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:096_WSBJ_10F</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>/sample-terms/UBERON_0000948/</td>\n",
       "      <td>heart</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:052_NZOJ_10F_06_UBERON_0000948</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:052_NZOJ_10F</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>/sample-terms/UBERON_0000948/</td>\n",
       "      <td>heart</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:063_WSBJ_10M_06_UBERON_0000948</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:063_WSBJ_10M</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>/sample-terms/UBERON_0000948/</td>\n",
       "      <td>heart</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:053_NZOJ_10M_06_UBERON_0000948</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:053_NZOJ_10M</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>/sample-terms/UBERON_0000948/</td>\n",
       "      <td>heart</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>192 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         accession  uuid                                 aliases:array  \\\n",
       "0    TSTSM69861046  None      ali-mortazavi:016_B6J_10F_03_NTR_0000646   \n",
       "1    TSTSM75831182  None      ali-mortazavi:016_B6J_10F_03_NTR_0000750   \n",
       "2    TSTSM69542835  None      ali-mortazavi:017_B6J_10M_03_NTR_0000646   \n",
       "3    TSTSM28050193  None      ali-mortazavi:017_B6J_10M_03_NTR_0000750   \n",
       "4     would create  None      ali-mortazavi:018_B6J_10F_03_NTR_0000646   \n",
       "..             ...   ...                                           ...   \n",
       "187   would create  None  ali-mortazavi:051_NZOJ_10M_06_UBERON_0000948   \n",
       "188   would create  None  ali-mortazavi:096_WSBJ_10F_06_UBERON_0000948   \n",
       "189   would create  None  ali-mortazavi:052_NZOJ_10F_06_UBERON_0000948   \n",
       "190   would create  None  ali-mortazavi:063_WSBJ_10M_06_UBERON_0000948   \n",
       "191   would create  None  ali-mortazavi:053_NZOJ_10M_06_UBERON_0000948   \n",
       "\n",
       "                 award                   lab                  source  \\\n",
       "0    /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "1    /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "2    /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "3    /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "4    /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "..                 ...                   ...                     ...   \n",
       "187  /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "188  /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "189  /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "190  /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "191  /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "\n",
       "                   donors:array          taxa                 biosample_term  \\\n",
       "0     ali-mortazavi:016_B6J_10F  Mus musculus     /sample-terms/NTR_0000646/   \n",
       "1     ali-mortazavi:016_B6J_10F  Mus musculus     /sample-terms/NTR_0000750/   \n",
       "2     ali-mortazavi:017_B6J_10M  Mus musculus     /sample-terms/NTR_0000646/   \n",
       "3     ali-mortazavi:017_B6J_10M  Mus musculus     /sample-terms/NTR_0000750/   \n",
       "4     ali-mortazavi:018_B6J_10F  Mus musculus     /sample-terms/NTR_0000646/   \n",
       "..                          ...           ...                            ...   \n",
       "187  ali-mortazavi:051_NZOJ_10M  Mus musculus  /sample-terms/UBERON_0000948/   \n",
       "188  ali-mortazavi:096_WSBJ_10F  Mus musculus  /sample-terms/UBERON_0000948/   \n",
       "189  ali-mortazavi:052_NZOJ_10F  Mus musculus  /sample-terms/UBERON_0000948/   \n",
       "190  ali-mortazavi:063_WSBJ_10M  Mus musculus  /sample-terms/UBERON_0000948/   \n",
       "191  ali-mortazavi:053_NZOJ_10M  Mus musculus  /sample-terms/UBERON_0000948/   \n",
       "\n",
       "                term_names:skip  \n",
       "0          left cerebral cortex  \n",
       "1    Hippocampal formation left  \n",
       "2          left cerebral cortex  \n",
       "3    Hippocampal formation left  \n",
       "4          left cerebral cortex  \n",
       "..                          ...  \n",
       "187                       heart  \n",
       "188                       heart  \n",
       "189                       heart  \n",
       "190                       heart  \n",
       "191                       heart  \n",
       "\n",
       "[192 rows x 10 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_accession_list_or_none(record, length):\n",
    "    accessions = record.accession.all()\n",
    "    if len(accessions) == 0:\n",
    "        return [None] * length\n",
    "    elif len(accessions) == length:\n",
    "        return [x.name for x in accessions]\n",
    "    else:\n",
    "        raise ValueError(\"Unexpected number of accessions {} {}\".format(accessions, length))\n",
    "\n",
    "tissue_sheet = []\n",
    "for tissue_name in tissues:\n",
    "    tissue = tissues[tissue_name]\n",
    "    donor_alias = f\"ali-mortazavi:{tissue.mouse.name}\"\n",
    "    ontology_terms = tissue.ontology_term.all()\n",
    "    accessions = get_accession_list_or_none(tissue, len(ontology_terms))\n",
    "    for term, accession_id in zip(ontology_terms, accessions):\n",
    "        curie = term.curie.replace(\":\", \"_\")\n",
    "        alias = f\"ali-mortazavi:{tissue.name}_{curie}\"\n",
    "        sample_term = \"/sample-terms/{}/\".format(curie)\n",
    "        if numpy.all(pandas.isnull(accession_id)):\n",
    "            accession_id = None\n",
    "        dcc_row = {\n",
    "            \"accession\": accession_id,\n",
    "            \"uuid\": None,\n",
    "            \"aliases:array\": alias,\n",
    "            \"award\": award,\n",
    "            \"lab\": lab,\n",
    "            \"source\": jax,\n",
    "            \"donors:array\": donor_alias,\n",
    "            \"taxa\": species,\n",
    "            \"biosample_term\": sample_term,\n",
    "            \"term_names:skip\": term.name,\n",
    "        }\n",
    "        tissue_sheet.append(dcc_row)\n",
    "    \n",
    "tissue_sheet = pandas.DataFrame(tissue_sheet)\n",
    "\n",
    "dry_run=True\n",
    "created = server.post_sheet(\"tissue\", tissue_sheet, dry_run=dry_run, verbose=True, validator=validator)\n",
    "if len(created) > 0 and not dry_run:\n",
    "    tissue_sheet.to_excel(\"tissue_sheet.xlsx\")\n",
    "tissue_sheet"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "received-conversion",
   "metadata": {},
   "source": [
    "# measurement_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "incorporated-shakespeare",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nanopore igvf_003/nanopore IGVF_003\n",
      "nextseq igvf_003/nextseq IGVF_003\n",
      "novaseq igvf_003/nova1 IGVF_003\n",
      "novaseq igvf_003/nova2 IGVF_003\n",
      "nextseq igvf_004/nextseq IGVF_004\n",
      "novaseq igvf_004/nova1 IGVF_004\n",
      "novaseq igvf_004/nova2 IGVF_004\n",
      "nextseq igvf_005/nextseq IGVF_005\n",
      "novaseq igvf_005/nova1 IGVF_005\n",
      "novaseq igvf_005/nova2 IGVF_005\n",
      "nextseq igvf_006/nextseq IGVF_006\n"
     ]
    }
   ],
   "source": [
    "for run in models.SequencingRun.objects.all():\n",
    "    print(run.platform.name, run.name, run.plate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "placed-swift",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ali-mortazavi:igvf_003_nextseq': {<SplitSeqPlate: IGVF_003>}, 'ali-mortazavi:igvf_003_novaseq': {<SplitSeqPlate: IGVF_003>}, 'ali-mortazavi:igvf_004_nextseq': {<SplitSeqPlate: IGVF_004>}, 'ali-mortazavi:igvf_004_novaseq': {<SplitSeqPlate: IGVF_004>}, 'ali-mortazavi:igvf_005_nextseq': {<SplitSeqPlate: IGVF_005>}, 'ali-mortazavi:igvf_005_novaseq': {<SplitSeqPlate: IGVF_005>}, 'ali-mortazavi:igvf_006_nextseq': {<SplitSeqPlate: IGVF_006>}}\n"
     ]
    }
   ],
   "source": [
    "\n",
    "measurement_plates = {}\n",
    "for run in models.SequencingRun.objects.all():\n",
    "    platform = run.platform.name\n",
    "    if platform != \"nanopore\":\n",
    "        name = \"{}_{}\".format(run.plate.name.lower(), run.platform.name)\n",
    "        alias = f\"ali-mortazavi:{name}\"\n",
    "        measurement_plates.setdefault(alias, set()).add(run.plate)\n",
    "    \n",
    "    \n",
    "for alias in measurement_plates:\n",
    "    assert len(measurement_plates[alias]) == 1\n",
    "\n",
    "print(measurement_plates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "polished-handle",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_samples_donors(plate):\n",
    "    samples = []\n",
    "    donors = {}\n",
    "    for well in plate.splitseqwell_set.all():\n",
    "        for biosample in well.biosample.all():\n",
    "            for tissue in biosample.tissue.all():\n",
    "                mouse_alias = f\"ali-mortazavi:{tissue.mouse.name}\"\n",
    "                # I need an ordered set of donors\n",
    "                donors[mouse_alias] = None\n",
    "                for term in tissue.ontology_term.all():\n",
    "                    curie = term.curie.replace(\":\", \"_\")\n",
    "                    tissue_alias = f\"ali-mortazavi:{tissue.name}_{curie}\"\n",
    "                    samples.append(tissue_alias)\n",
    "\n",
    "    return {\n",
    "        \"samples\": samples,\n",
    "        \"donor\": list(donors.keys())\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "seeing-reform",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>uuid</th>\n",
       "      <th>aliases:array</th>\n",
       "      <th>award</th>\n",
       "      <th>lab</th>\n",
       "      <th>assay_term</th>\n",
       "      <th>documents</th>\n",
       "      <th>alternate_accessions</th>\n",
       "      <th>submitter_comment</th>\n",
       "      <th>description</th>\n",
       "      <th>samples:array</th>\n",
       "      <th>protocol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:igvf_003_nextseq</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/assay-terms/OBI_0003109/</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:016_B6J_10F_03_NTR_0000646,ali-m...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:igvf_003_novaseq</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/assay-terms/OBI_0003109/</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:016_B6J_10F_03_NTR_0000646,ali-m...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession  uuid                   aliases:array              award  \\\n",
       "0  would create  None  ali-mortazavi:igvf_003_nextseq  /awards/HG012077/   \n",
       "1  would create  None  ali-mortazavi:igvf_003_novaseq  /awards/HG012077/   \n",
       "\n",
       "                    lab                 assay_term documents  \\\n",
       "0  /labs/ali-mortazavi/  /assay-terms/OBI_0003109/      None   \n",
       "1  /labs/ali-mortazavi/  /assay-terms/OBI_0003109/      None   \n",
       "\n",
       "  alternate_accessions submitter_comment description  \\\n",
       "0                 None              None        None   \n",
       "1                 None              None        None   \n",
       "\n",
       "                                       samples:array protocol  \n",
       "0  ali-mortazavi:016_B6J_10F_03_NTR_0000646,ali-m...     None  \n",
       "1  ali-mortazavi:016_B6J_10F_03_NTR_0000646,ali-m...     None  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "measurements = []\n",
    "\n",
    "for alias in measurement_plates:\n",
    "    plate = next(iter(measurement_plates[alias]))\n",
    "    if plate.name == test_run:\n",
    "        plate_details = get_samples_donors(plate)\n",
    "        \n",
    "        dcc_row = {\n",
    "            \"accession\": None,\n",
    "            \"uuid\": None,\n",
    "            \"aliases:array\": alias,\n",
    "            \"award\": award,\n",
    "            \"lab\": lab,\n",
    "            \"assay_term\": \"/assay-terms/OBI_0003109/\", # single-nucleus RNA sequencing assay\n",
    "            \"documents\": None,\n",
    "            \"alternate_accessions\": None,\n",
    "            \"submitter_comment\": None,\n",
    "            \"description\": None,\n",
    "            \"samples:array\": \",\".join(plate_details[\"samples\"]),\n",
    "            #\"samples_len:skip\": len(plate_details[\"samples\"]),\n",
    "            #\"donors:array\": \",\".join(plate_details[\"donor\"]),\n",
    "            #\"donors_len:skip\": len(plate_details[\"donor\"]),\n",
    "            \"protocol\": None,\n",
    "        }\n",
    "        measurements.append(dcc_row)\n",
    "        \n",
    "measurements = pandas.DataFrame(measurements)\n",
    "dry_run=True\n",
    "created = server.post_sheet(\"measurement_set\", measurements, dry_run=dry_run, verbose=True, validator=validator)\n",
    "if len(created) > 0 and not dry_run:\n",
    "    measurements.to_excel(\"measurement_set.xlsx\", index=False)\n",
    "measurements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "determined-webmaster",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "exciting-creativity",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "median-niagara",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "differential-savage",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "daily-expansion",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cordless-warrant",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "respective-accommodation",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "closing-cause",
   "metadata": {},
   "source": [
    "# Link files to runs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "center-imperial",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['igvf_004', 'igvf_005', 'igvf_003'])\n"
     ]
    }
   ],
   "source": [
    "known_fastqs = {}\n",
    "with open(\"current_fastq_runs.txt\", \"rt\") as instream:\n",
    "    for line in instream:\n",
    "        filename = Path(line.rstrip())\n",
    "        known_fastqs.setdefault(filename.parts[0], {}).setdefault(filename.parts[1], []).append(filename)\n",
    "\n",
    "print (known_fastqs.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "controlling-figure",
   "metadata": {},
   "outputs": [],
   "source": [
    "known_md5s = {}\n",
    "\n",
    "for md5sum_name in [\"igvf_003.md5sums\"]:\n",
    "    with open(md5sum_name, \"rt\") as instream:\n",
    "        run_name = md5sum_name.split(\".\")[0]\n",
    "        for line in instream:\n",
    "            records = line.split()\n",
    "            md5sum = records[0]\n",
    "            filename = Path(records[1])\n",
    "            known_md5s.setdefault(run_name, {}).setdefault(filename.parts[0], {})[filename.name] = md5sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "solid-coordinate",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_groups = {\n",
    "    \"ali-mortazavi:igvf_003_nextseq\": [\"nextseq\"],\n",
    "    \"ali-mortazavi:igvf_003_novaseq\": [\"nova1\", \"nova2\"],\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "dominant-person",
   "metadata": {},
   "outputs": [],
   "source": [
    "nextseq_re = re.compile(r\"(?P<sublibrary_id>[\\d]+_[\\d]+[A-Z])_(?P<read>R[12])\\.fastq\\.gz\")\n",
    "#print(nextseq_re.match(\"003_67A_R1.fastq.gz\").groups())\n",
    "nova_re = re.compile(r\"(?P<sublibrary_id>Sublibrary_[\\d]+)_S[\\d+]+_(?P<lane>L[\\d]+)_(?P<read>[IR][123])_(?P<fragment>[\\d]+)\\.fastq\\.gz\")\n",
    "#print(nova_re.match(\"Sublibrary_11_S10_L004_R1_001.fastq.gz\").groups())\n",
    "file_patterns = {\n",
    "    \"nextseq\": nextseq_re,\n",
    "    \"nova1\": nova_re,\n",
    "    \"nova2\": nova_re\n",
    "}\n",
    "\n",
    "path_to_platform = {\n",
    "    \"nextseq\": \"nextseq\",\n",
    "    \"nova1\": \"novaseq\",\n",
    "    \"nova2\": \"novaseq\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "impressed-bicycle",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "246\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>uuid</th>\n",
       "      <th>award</th>\n",
       "      <th>lab</th>\n",
       "      <th>md5sum</th>\n",
       "      <th>file_format</th>\n",
       "      <th>file_set</th>\n",
       "      <th>content_type</th>\n",
       "      <th>sequencing_run:integer</th>\n",
       "      <th>submitted_file_name</th>\n",
       "      <th>illumina_read_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>bb696e63b6a8710ec8435b60ef6c81f8</td>\n",
       "      <td>fastq</td>\n",
       "      <td>ali-mortazavi:igvf_003_nextseq</td>\n",
       "      <td>reads</td>\n",
       "      <td>1</td>\n",
       "      <td>igvf_003/nextseq/003_8A_R1.fastq.gz</td>\n",
       "      <td>R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>21967ba1a5a3218508961fac482a2952</td>\n",
       "      <td>fastq</td>\n",
       "      <td>ali-mortazavi:igvf_003_nextseq</td>\n",
       "      <td>reads</td>\n",
       "      <td>1</td>\n",
       "      <td>igvf_003/nextseq/003_8A_R2.fastq.gz</td>\n",
       "      <td>R2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>35350d53cf858267d294bce69e475c83</td>\n",
       "      <td>fastq</td>\n",
       "      <td>ali-mortazavi:igvf_003_nextseq</td>\n",
       "      <td>reads</td>\n",
       "      <td>2</td>\n",
       "      <td>igvf_003/nextseq/003_67A_R1.fastq.gz</td>\n",
       "      <td>R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>00c68ff0f9e0217d422c57e8948d4bb4</td>\n",
       "      <td>fastq</td>\n",
       "      <td>ali-mortazavi:igvf_003_nextseq</td>\n",
       "      <td>reads</td>\n",
       "      <td>2</td>\n",
       "      <td>igvf_003/nextseq/003_67A_R2.fastq.gz</td>\n",
       "      <td>R2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>79db765143bb23e14c613c9e1339e9f7</td>\n",
       "      <td>fastq</td>\n",
       "      <td>ali-mortazavi:igvf_003_nextseq</td>\n",
       "      <td>reads</td>\n",
       "      <td>3</td>\n",
       "      <td>igvf_003/nextseq/003_67B_R1.fastq.gz</td>\n",
       "      <td>R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>7358c7000d036cff3372417ee3166da7</td>\n",
       "      <td>fastq</td>\n",
       "      <td>ali-mortazavi:igvf_003_novaseq</td>\n",
       "      <td>reads</td>\n",
       "      <td>118</td>\n",
       "      <td>igvf_003/nova2/Sublibrary_16_S15_L002_R2_001.f...</td>\n",
       "      <td>R2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>892394a7d256d6945147353bf2a2fc35</td>\n",
       "      <td>fastq</td>\n",
       "      <td>ali-mortazavi:igvf_003_novaseq</td>\n",
       "      <td>reads</td>\n",
       "      <td>119</td>\n",
       "      <td>igvf_003/nova2/Sublibrary_16_S15_L003_R1_001.f...</td>\n",
       "      <td>R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>cb32b3cbec11a08530765b9cf95c16ac</td>\n",
       "      <td>fastq</td>\n",
       "      <td>ali-mortazavi:igvf_003_novaseq</td>\n",
       "      <td>reads</td>\n",
       "      <td>119</td>\n",
       "      <td>igvf_003/nova2/Sublibrary_16_S15_L003_R2_001.f...</td>\n",
       "      <td>R2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>76f59bde3742e3e426f9424e1836b594</td>\n",
       "      <td>fastq</td>\n",
       "      <td>ali-mortazavi:igvf_003_novaseq</td>\n",
       "      <td>reads</td>\n",
       "      <td>120</td>\n",
       "      <td>igvf_003/nova2/Sublibrary_16_S15_L004_R1_001.f...</td>\n",
       "      <td>R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>15fbfcfd8fd035fd7e0d8e2326ddbf53</td>\n",
       "      <td>fastq</td>\n",
       "      <td>ali-mortazavi:igvf_003_novaseq</td>\n",
       "      <td>reads</td>\n",
       "      <td>120</td>\n",
       "      <td>igvf_003/nova2/Sublibrary_16_S15_L004_R2_001.f...</td>\n",
       "      <td>R2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>246 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession  uuid              award                   lab  \\\n",
       "0    would create  None  /awards/HG012077/  /labs/ali-mortazavi/   \n",
       "1    would create  None  /awards/HG012077/  /labs/ali-mortazavi/   \n",
       "2    would create  None  /awards/HG012077/  /labs/ali-mortazavi/   \n",
       "3    would create  None  /awards/HG012077/  /labs/ali-mortazavi/   \n",
       "4    would create  None  /awards/HG012077/  /labs/ali-mortazavi/   \n",
       "..            ...   ...                ...                   ...   \n",
       "241  would create  None  /awards/HG012077/  /labs/ali-mortazavi/   \n",
       "242  would create  None  /awards/HG012077/  /labs/ali-mortazavi/   \n",
       "243  would create  None  /awards/HG012077/  /labs/ali-mortazavi/   \n",
       "244  would create  None  /awards/HG012077/  /labs/ali-mortazavi/   \n",
       "245  would create  None  /awards/HG012077/  /labs/ali-mortazavi/   \n",
       "\n",
       "                               md5sum file_format  \\\n",
       "0    bb696e63b6a8710ec8435b60ef6c81f8       fastq   \n",
       "1    21967ba1a5a3218508961fac482a2952       fastq   \n",
       "2    35350d53cf858267d294bce69e475c83       fastq   \n",
       "3    00c68ff0f9e0217d422c57e8948d4bb4       fastq   \n",
       "4    79db765143bb23e14c613c9e1339e9f7       fastq   \n",
       "..                                ...         ...   \n",
       "241  7358c7000d036cff3372417ee3166da7       fastq   \n",
       "242  892394a7d256d6945147353bf2a2fc35       fastq   \n",
       "243  cb32b3cbec11a08530765b9cf95c16ac       fastq   \n",
       "244  76f59bde3742e3e426f9424e1836b594       fastq   \n",
       "245  15fbfcfd8fd035fd7e0d8e2326ddbf53       fastq   \n",
       "\n",
       "                           file_set content_type  sequencing_run:integer  \\\n",
       "0    ali-mortazavi:igvf_003_nextseq        reads                       1   \n",
       "1    ali-mortazavi:igvf_003_nextseq        reads                       1   \n",
       "2    ali-mortazavi:igvf_003_nextseq        reads                       2   \n",
       "3    ali-mortazavi:igvf_003_nextseq        reads                       2   \n",
       "4    ali-mortazavi:igvf_003_nextseq        reads                       3   \n",
       "..                              ...          ...                     ...   \n",
       "241  ali-mortazavi:igvf_003_novaseq        reads                     118   \n",
       "242  ali-mortazavi:igvf_003_novaseq        reads                     119   \n",
       "243  ali-mortazavi:igvf_003_novaseq        reads                     119   \n",
       "244  ali-mortazavi:igvf_003_novaseq        reads                     120   \n",
       "245  ali-mortazavi:igvf_003_novaseq        reads                     120   \n",
       "\n",
       "                                   submitted_file_name illumina_read_type  \n",
       "0                  igvf_003/nextseq/003_8A_R1.fastq.gz                 R1  \n",
       "1                  igvf_003/nextseq/003_8A_R2.fastq.gz                 R2  \n",
       "2                 igvf_003/nextseq/003_67A_R1.fastq.gz                 R1  \n",
       "3                 igvf_003/nextseq/003_67A_R2.fastq.gz                 R2  \n",
       "4                 igvf_003/nextseq/003_67B_R1.fastq.gz                 R1  \n",
       "..                                                 ...                ...  \n",
       "241  igvf_003/nova2/Sublibrary_16_S15_L002_R2_001.f...                 R2  \n",
       "242  igvf_003/nova2/Sublibrary_16_S15_L003_R1_001.f...                 R1  \n",
       "243  igvf_003/nova2/Sublibrary_16_S15_L003_R2_001.f...                 R2  \n",
       "244  igvf_003/nova2/Sublibrary_16_S15_L004_R1_001.f...                 R1  \n",
       "245  igvf_003/nova2/Sublibrary_16_S15_L004_R2_001.f...                 R2  \n",
       "\n",
       "[246 rows x 11 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def run_key(x, pattern):\n",
    "    if isinstance(x, Path):\n",
    "        x = x.name\n",
    "\n",
    "    match = pattern.match(x)\n",
    "    assert match is not None\n",
    "    groups = match.groupdict()\n",
    "    if groups[\"sublibrary_id\"].startswith(\"Sublibrary\"):\n",
    "        sublibrary_id = int(groups[\"sublibrary_id\"].split(\"_\")[1])\n",
    "    else:\n",
    "        run, library = groups[\"sublibrary_id\"].split(\"_\")\n",
    "        head, tail = library[:-1], library[-1:]\n",
    "        sublibrary_id = (int(run), int(head), tail)\n",
    "        \n",
    "    if \"lane\" in groups:\n",
    "        lane_id = int(groups[\"lane\"][1:])\n",
    "    else:\n",
    "        lane_id = None\n",
    "        \n",
    "    read = groups[\"read\"]\n",
    "\n",
    "    return(sublibrary_id, lane_id, read)\n",
    "        \n",
    "sequencing_data = []\n",
    "plate_name = \"igvf_003\"\n",
    "run_counts = {\n",
    "    \"nextseq\": 1,\n",
    "    \"novaseq\": 1,\n",
    "}\n",
    "run_breaks = {\n",
    "    \"nextseq\": None,\n",
    "    \"novaseq\": None,\n",
    "}\n",
    "for alias in run_groups:\n",
    "    for run_name in run_groups[alias]:\n",
    "        pattern = file_patterns[run_name]\n",
    "        \n",
    "        for filename in sorted(known_fastqs[plate_name][run_name], key=lambda x: run_key(x, pattern)):\n",
    "            sublibrary_id, lane_id, read = run_key(filename, pattern)\n",
    "            \n",
    "            if read == \"I1\":\n",
    "                continue\n",
    "            \n",
    "            # Group by sublibrary_id & lane_id\n",
    "            if run_breaks[path_to_platform[run_name]] is None:\n",
    "                run_breaks[path_to_platform[run_name]] = (sublibrary_id, lane_id)\n",
    "            elif run_breaks[path_to_platform[run_name]] != (sublibrary_id, lane_id):\n",
    "                run_counts[path_to_platform[run_name]] += 1\n",
    "                run_breaks[path_to_platform[run_name]] = (sublibrary_id, lane_id)\n",
    "\n",
    "            dcc_row = {\n",
    "                \"accession\": None,\n",
    "                \"uuid\": None,\n",
    "                #\"aliases\": None,\n",
    "                \"award\": award,\n",
    "                \"lab\": lab,\n",
    "                \"md5sum\": known_md5s[plate_name][run_name][filename.name],\n",
    "                \"file_format\": \"fastq\",\n",
    "                \"file_set\": \"ali-mortazavi:{}_{}\".format(plate_name, path_to_platform[run_name]),\n",
    "                \"content_type\": \"reads\",\n",
    "                \"sequencing_run:integer\": run_counts[path_to_platform[run_name]],\n",
    "                #\"documents:array\": None,\n",
    "                #\"submitter_comment\": None,\n",
    "                #\"description\": None,\n",
    "                #\"dbxrefs\": None,\n",
    "                #\"derived_from\": None,\n",
    "                #\"file_format_specifications\": None,\n",
    "                \"submitted_file_name\": filename,\n",
    "                \"illumina_read_type\": read,\n",
    "            }\n",
    "            sequencing_data.append(dcc_row)\n",
    "            \n",
    "sequencing_data = pandas.DataFrame(sequencing_data)\n",
    "sequencing_data.to_excel(\"sequence_data.xlsx\", index=False)\n",
    "created = server.post_sheet(\"sequence_data\", sequencing_data, dry_run=True, verbose=True, validator=validator)\n",
    "if len(created) > 0:\n",
    "    print((len(created)))\n",
    "    \n",
    "sequencing_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "vertical-gibson",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "collaborative-gateway",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
