{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "floating-raising",
   "metadata": {},
   "source": [
    "# Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "round-television",
   "metadata": {},
   "source": [
    "Using the models we have try to fill in the metadata spreadsheet & submit test records to the IGVF test portal.\n",
    "\n",
    "This try had all the files from each platform type attached to one measurement set. But with Jennifer we decided this is probably a poor idea."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "intended-gender",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "educational-vitamin",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "module 'django.contrib.admin' has no attribute 'display'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-fe3bbdcf016a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     24\u001b[0m     \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mMOUSEDEMO\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     25\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 26\u001b[0;31m \u001b[0mdjango\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msetup\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     27\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     28\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mmousedemo\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0msettings\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3/dist-packages/django/__init__.py\u001b[0m in \u001b[0;36msetup\u001b[0;34m(set_prefix)\u001b[0m\n\u001b[1;32m     22\u001b[0m             \u001b[0;34m'/'\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0msettings\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mFORCE_SCRIPT_NAME\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0msettings\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mFORCE_SCRIPT_NAME\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     23\u001b[0m         )\n\u001b[0;32m---> 24\u001b[0;31m     \u001b[0mapps\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpopulate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msettings\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mINSTALLED_APPS\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/usr/lib/python3/dist-packages/django/apps/registry.py\u001b[0m in \u001b[0;36mpopulate\u001b[0;34m(self, installed_apps)\u001b[0m\n\u001b[1;32m    112\u001b[0m             \u001b[0;31m# Phase 2: import models modules.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    113\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mapp_config\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapp_configs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 114\u001b[0;31m                 \u001b[0mapp_config\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mimport_models\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    115\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    116\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclear_cache\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3/dist-packages/django/apps/config.py\u001b[0m in \u001b[0;36mimport_models\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    209\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mmodule_has_submodule\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmodule\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mMODELS_MODULE_NAME\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    210\u001b[0m             \u001b[0mmodels_module_name\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'%s.%s'\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mMODELS_MODULE_NAME\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 211\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmodels_module\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mimport_module\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodels_module_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    212\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    213\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mready\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.9/importlib/__init__.py\u001b[0m in \u001b[0;36mimport_module\u001b[0;34m(name, package)\u001b[0m\n\u001b[1;32m    125\u001b[0m                 \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    126\u001b[0m             \u001b[0mlevel\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 127\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0m_bootstrap\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_gcd_import\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mlevel\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpackage\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlevel\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    128\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    129\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.9/importlib/_bootstrap.py\u001b[0m in \u001b[0;36m_gcd_import\u001b[0;34m(name, package, level)\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.9/importlib/_bootstrap.py\u001b[0m in \u001b[0;36m_find_and_load\u001b[0;34m(name, import_)\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.9/importlib/_bootstrap.py\u001b[0m in \u001b[0;36m_find_and_load_unlocked\u001b[0;34m(name, import_)\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.9/importlib/_bootstrap.py\u001b[0m in \u001b[0;36m_load_unlocked\u001b[0;34m(spec)\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.9/importlib/_bootstrap_external.py\u001b[0m in \u001b[0;36mexec_module\u001b[0;34m(self, module)\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.9/importlib/_bootstrap.py\u001b[0m in \u001b[0;36m_call_with_frames_removed\u001b[0;34m(f, *args, **kwds)\u001b[0m\n",
      "\u001b[0;32m~/proj/igvf_mice/igvf_mice/models.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 12\u001b[0;31m \u001b[0;32mclass\u001b[0m \u001b[0mAccessionNamespace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodels\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mModel\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     13\u001b[0m     \u001b[0;34m\"\"\"Describe a potential source of accession ids.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/proj/igvf_mice/igvf_mice/models.py\u001b[0m in \u001b[0;36mAccessionNamespace\u001b[0;34m()\u001b[0m\n\u001b[1;32m     34\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\" \"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"_\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     35\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 36\u001b[0;31m     \u001b[0;34m@\u001b[0m\u001b[0madmin\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdisplay\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     37\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mlink\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     38\u001b[0m         return format_html(\n",
      "\u001b[0;31mAttributeError\u001b[0m: module 'django.contrib.admin' has no attribute 'display'"
     ]
    }
   ],
   "source": [
    "import bz2\n",
    "from collections import Counter, namedtuple\n",
    "import datetime\n",
    "import numpy\n",
    "import os\n",
    "import pandas\n",
    "from pathlib import Path\n",
    "import re\n",
    "from subprocess import run, PIPE\n",
    "import sys\n",
    "import time\n",
    "from tqdm import tqdm\n",
    "import zoneinfo\n",
    "\n",
    "os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'mousedemo.settings')\n",
    "os.environ[\"DJANGO_ALLOW_ASYNC_UNSAFE\"] = \"true\"\n",
    "\n",
    "import django\n",
    "from django.contrib.auth import get_user_model\n",
    "from django.db import DEFAULT_DB_ALIAS\n",
    "\n",
    "MOUSEDEMO = str(Path(\"mousedemo\").absolute())\n",
    "if MOUSEDEMO not in sys.path:\n",
    "    sys.path.append(MOUSEDEMO)\n",
    "\n",
    "django.setup()\n",
    "\n",
    "from mousedemo import settings\n",
    "from igvf_mice import models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "signed-percentage",
   "metadata": {},
   "outputs": [],
   "source": [
    "EC = str(Path(\"~/proj/encoded_client\").expanduser())\n",
    "if EC not in sys.path:\n",
    "    sys.path.append(EC)\n",
    "    \n",
    "from encoded_client import encoded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "basic-recipient",
   "metadata": {},
   "outputs": [],
   "source": [
    "server = encoded.ENCODED(\"api.sandbox.igvf.org\")\n",
    "\n",
    "validator = encoded.DCCValidator(server)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "attractive-buffalo",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "representative-observer",
   "metadata": {},
   "outputs": [],
   "source": [
    "award = \"/awards/HG012077/\"\n",
    "labs = [\"/labs/lior-pachter/\", \"/labs/grant-macgregor/\", \"/labs/barbara-wold/\", \"/labs/ali-mortazavi/\"]\n",
    "lab = labs[-1]\n",
    "jax = \"/sources/jackson-labs/\"\n",
    "species = 'Mus musculus'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "psychological-sunday",
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_background = {\n",
    "    \"A/J\": \"A/J (AJ)\",\n",
    "    \"C57BL/6J\": \"C57BL/6J (B6)\",\n",
    "    \"129S1/SvImJ\": \"129S1/SvImJ (129)\",\n",
    "    \"NOD/ShiLtJ\": \"NOD/ShiLtJ (NOD)\",\n",
    "    \"NZO/HlLtJ\": \"NZO/H1LtJ (NZO)\",\n",
    "    \"CAST/EiJ\": \"CAST/EiJ (CAST)\",\n",
    "    \"PWK/PhJ\": \"PWK/PhJ (PWK)\",\n",
    "    \"WSB/EiJ\": \"WSB/EiJ (WSB)\",\n",
    "    #\"CAST (M. m. castaneus)\",\n",
    "    #\"WSB (M. m. domesticus)\",\n",
    "    #\"PWK (M. m. musculus)\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "distributed-webcam",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "involved-desperate",
   "metadata": {},
   "outputs": [],
   "source": [
    "#plate_name =  \"IGVF_B01\"\n",
    "plate_name = \"IGVF_B01\"\n",
    "plate = models.SplitSeqPlate.objects.get(name=plate_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "hidden-adventure",
   "metadata": {},
   "outputs": [],
   "source": [
    "tissues = {}\n",
    "mice = {}\n",
    "for well in plate.splitseqwell_set.all():\n",
    "    for biosample in well.biosample.all():\n",
    "        for tissue in biosample.tissue.all():\n",
    "            tissues[tissue.name] = tissue\n",
    "            mice[tissue.mouse.name] = tissue.mouse\n",
    "            \n",
    "print(len(tissues))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "permanent-expression",
   "metadata": {},
   "source": [
    "# rodent_donor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "welsh-safety",
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_sex(value):\n",
    "    if models.SexEnum.MALE == value:\n",
    "        return \"male\"\n",
    "    elif models.SexEnum.FEMALE == value:\n",
    "        return \"female\"\n",
    "\n",
    "def get_accession_string_or_none(record):\n",
    "    accessions = record.accession.all()\n",
    "    if len(accessions) == 0:\n",
    "        return None\n",
    "    else:\n",
    "        return \",\".join([x.name for x in accessions])\n",
    "    \n",
    "rodent_donor = []\n",
    "for mouse_name in sorted(mice):\n",
    "    mouse = mice[mouse_name]\n",
    "    dcc_row = {\n",
    "        #\"#response\": None,\n",
    "        #\"#response_time\": None,\n",
    "        \"accession\": get_accession_string_or_none(mouse),\n",
    "        \"uuid\": None, \n",
    "        \"aliases:array\": f\"ali-mortazavi:{mouse.name}\",\n",
    "        \"award\": award,\n",
    "        \"lab\": lab,\n",
    "        \"taxa\": species,\n",
    "        \"sex\": format_sex(mouse.sex),\n",
    "        \"strain\": mouse.strain.name,\n",
    "        \"references\": None,\n",
    "        \"url\": mouse.strain.url,\n",
    "        \"source\": jax,\n",
    "        \"lot_id\": None,\n",
    "        \"product_id\": mouse.strain.jax_catalog_number,\n",
    "        \"documents\": None,\n",
    "        \"alternate_accessions\": None,\n",
    "        \"submitter_comment\": None,\n",
    "        \"description\": None,\n",
    "        \"parents\": None,\n",
    "        \"traits\": None,\n",
    "        \"phenotypic_features\": None,\n",
    "        \"external_resources\": None,\n",
    "        \"strain_background\": strain_background[mouse.strain.name],\n",
    "        \"genotype\": None,\n",
    "        \"individual_rodent:boolean\": True,\n",
    "        \"rodent_identifier\": mouse.name,\n",
    "    }\n",
    "    rodent_donor.append(dcc_row)\n",
    "    \n",
    "rodent_donor = pandas.DataFrame(rodent_donor)\n",
    "\n",
    "dry_run = True\n",
    "created = server.post_sheet(\"rodent_donor\", rodent_donor, dry_run=dry_run, verbose=True, validator=validator)\n",
    "if len(created) > 0 and not dry_run:\n",
    "    rodent_donor.to_excel(\"{}_rodent_donor.xlsx\".format(plate_name))\n",
    "rodent_donor"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "modern-diversity",
   "metadata": {},
   "source": [
    "# tissue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "approved-longer",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def get_accession_list_or_none(record, length):\n",
    "    accessions = record.accession.all()\n",
    "    if len(accessions) == 0:\n",
    "        return [None] * length\n",
    "    elif len(accessions) == length:\n",
    "        return [x.name for x in accessions]\n",
    "    else:\n",
    "        raise ValueError(\"Unexpected number of accessions {} {}\".format(accessions, length))\n",
    "\n",
    "tissue_sheet = []\n",
    "for tissue_name in tissues:\n",
    "    tissue = tissues[tissue_name]\n",
    "    donor_alias = f\"ali-mortazavi:{tissue.mouse.name}\"\n",
    "    ontology_terms = tissue.ontology_term.all()\n",
    "    accessions = get_accession_list_or_none(tissue, len(ontology_terms))\n",
    "    for term, accession_id in zip(ontology_terms, accessions):\n",
    "        curie = term.curie.replace(\":\", \"_\")\n",
    "        alias = f\"ali-mortazavi:{tissue.name}_{curie}\"\n",
    "        sample_term = \"/sample-terms/{}/\".format(curie)\n",
    "        if numpy.all(pandas.isnull(accession_id)):\n",
    "            accession_id = None\n",
    "        dcc_row = {\n",
    "            \"accession\": accession_id,\n",
    "            \"uuid\": None,\n",
    "            \"aliases:array\": alias,\n",
    "            \"award\": award,\n",
    "            \"lab\": lab,\n",
    "            \"source\": jax,\n",
    "            \"donors:array\": donor_alias,\n",
    "            #\"taxa\": species,\n",
    "            \"biosample_term\": sample_term,\n",
    "            \"term_names:skip\": term.name,\n",
    "        }\n",
    "        tissue_sheet.append(dcc_row)\n",
    "    \n",
    "tissue_sheet = pandas.DataFrame(tissue_sheet)\n",
    "\n",
    "dry_run = True\n",
    "created = server.post_sheet(\"tissue\", tissue_sheet, dry_run=dry_run, verbose=True, validator=validator)\n",
    "if len(created) > 0 and not dry_run:\n",
    "    tissue_sheet.to_excel(\"{}_tissue_sheet.xlsx\".format(plate_name))\n",
    "tissue_sheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bored-dealing",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "infrared-oasis",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "received-conversion",
   "metadata": {},
   "source": [
    "# measurement_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "polished-handle",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_samples_donors(plate):\n",
    "    samples = set()\n",
    "    donors = {}\n",
    "    for well in plate.splitseqwell_set.all():\n",
    "        for biosample in well.biosample.all():\n",
    "            for tissue in biosample.tissue.all():\n",
    "                mouse_alias = f\"ali-mortazavi:{tissue.mouse.name}\"\n",
    "                # I need an ordered set of donors\n",
    "                donors[mouse_alias] = None\n",
    "                for term in tissue.ontology_term.all():\n",
    "                    curie = term.curie.replace(\":\", \"_\")\n",
    "                    tissue_alias = f\"ali-mortazavi:{tissue.name}_{curie}\"\n",
    "                    samples.add(tissue_alias)\n",
    "\n",
    "    return {\n",
    "        \"samples\": samples,\n",
    "        \"donor\": list(donors.keys())\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "incorrect-norway",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "knowing-terminal",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "measurement_set = []\n",
    "\n",
    "plate_details = get_samples_donors(models.SplitSeqPlate.objects.get(name=plate_name))\n",
    "families = models.Platform.objects.values(\"family\").distinct()\n",
    "\n",
    "for subpool in models.Subpool.objects.filter(plate__name=plate_name).order_by(\"name\"):\n",
    "    for platform in families:\n",
    "        subpoolinrun_set = subpool.subpoolinrun_set.filter(sequencing_run__platform__family=platform[\"family\"])\n",
    "        if subpoolinrun_set.exists():\n",
    "            subpoolinrun = subpoolinrun_set.first()\n",
    "            if subpoolinrun.measurement_set is not None:\n",
    "                accession = \",\".join([x.name for x in subpoolinrun.measurement_set.accession.all()])\n",
    "            else:\n",
    "                accesssion = None\n",
    "                \n",
    "            alias = \"ali-mortazavi:{}_{}\".format(subpool.name, platform[\"family\"])\n",
    "            dcc_row = {\n",
    "                \"accession\": accession,\n",
    "                \"uuid\": None,\n",
    "                \"aliases:array\": alias,\n",
    "                \"award\": award,\n",
    "                \"lab\": lab,\n",
    "                \"assay_term\": \"/assay-terms/OBI_0003109/\", # single-nucleus RNA sequencing assay\n",
    "                \"documents\": None,\n",
    "                \"alternate_accessions\": None,\n",
    "                \"submitter_comment\": None,\n",
    "                \"description\": None,\n",
    "                \"samples:array\": \",\".join(plate_details[\"samples\"]),\n",
    "                #\"samples_len:skip\": len(plate_details[\"samples\"]),\n",
    "                #\"donors:array\": \",\".join(plate_details[\"donor\"]),\n",
    "                #\"donors_len:skip\": len(plate_details[\"donor\"]),\n",
    "                \"protocol\": None,\n",
    "            }\n",
    "            measurement_set.append(dcc_row)\n",
    "\n",
    "measurement_set = pandas.DataFrame(measurement_set)\n",
    "dry_run = True\n",
    "created = server.post_sheet(\"measurement_set\", measurement_set, dry_run=dry_run, verbose=True, validator=validator)\n",
    "if len(created) > 0 and not dry_run:\n",
    "    measurement_set.to_excel(\"{}_measurement_set.xlsx\".format(plate_name))\n",
    "\n",
    "measurement_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "executed-agent",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, row in measurement_set.iterrows():\n",
    "    measurement = server.get_json(\"/measurement_set/{}/\".format(row.accession))\n",
    "    samples = row[\"samples:array\"].split(\",\")\n",
    "    print(row.accession, len(measurement[\"samples\"]), len(samples))\n",
    "    if len(samples) > len(measurement[\"samples\"]):\n",
    "        print(server.patch_json(measurement[\"@id\"], {\"samples\": samples}))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "compatible-rubber",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "families = models.Platform.objects.values(\"family\").distinct()\n",
    "\n",
    "t0 = time.monotonic()\n",
    "sequence_file = []\n",
    "for subpool in models.Subpool.objects.filter(plate__name=plate_name).order_by(\"name\"):\n",
    "    #alias = \"ali-mortazavi:{}_{}\".format(subpool.name, platform[\"family\"])\n",
    "    sequencing_run_break = None\n",
    "    sequencing_run_count = 0\n",
    "    last_family = None\n",
    "    for run in subpool.subpoolinrun_set.all().order_by(\"id\"):\n",
    "        platform_term_id = run.sequencing_run.platform.igvf_id\n",
    "        family = run.sequencing_run.platform.family\n",
    "        if family == \"nanopore\":\n",
    "            continue\n",
    "\n",
    "        if family != last_family:\n",
    "            sequencing_run_count = 0\n",
    "            last_family = family\n",
    "        measurement_set_alias = \"ali-mortazavi:{}_{}\".format(subpool.name, family)\n",
    "        for subpool_in_run_file in run.subpoolinrunfile_set.all().order_by(\"subpool_run\", \"lane\", \"fragment\", \"read\"):\n",
    "            \n",
    "            #, subpool_in_run_file.lane, subpool_in_run_file.fragment\n",
    "            current_sequencing_run = (subpool_in_run_file.subpool_run)\n",
    "            if sequencing_run_break != current_sequencing_run:\n",
    "                sequencing_run_break = current_sequencing_run\n",
    "                sequencing_run_count += 1\n",
    "                \n",
    "            #print(subpool_in_run_file.filename, current_sequencing_run, sequencing_run_count)\n",
    "            \n",
    "            illumina_read = subpool_in_run_file.read if family == \"illumina\" else None\n",
    "            \n",
    "            if subpool_in_run_file.accession.exists():\n",
    "                accession = \",\".join([accession.name for accession in subpool_in_run_file.accession.all()])\n",
    "            else:\n",
    "                accession = None\n",
    "\n",
    "            dcc_row = {\n",
    "                \"accession\": accession,\n",
    "                \"uuid\": None,\n",
    "                #\"aliases\": None,\n",
    "                \"award\": award,\n",
    "                \"lab\": lab,\n",
    "                \"md5sum\": subpool_in_run_file.md5sum,\n",
    "                \"file_format\": \"fastq\",\n",
    "                \"file_set\": measurement_set_alias,\n",
    "                \"content_type\": \"reads\",\n",
    "                \"flowcell_id\": subpool_in_run_file.flowcell_id,\n",
    "                \"lane:integer\": subpool_in_run_file.lane,\n",
    "                \"sequencing_run:integer\": sequencing_run_count,\n",
    "                #\"documents:array\": None,\n",
    "                #\"submitter_comment\": None,\n",
    "                #\"description\": None,\n",
    "                #\"dbxrefs\": None,\n",
    "                #\"derived_from\": None,\n",
    "                #\"file_format_specifications\": None,\n",
    "                \"submitted_file_name\": subpool_in_run_file.filename,\n",
    "                \"illumina_read_type\": illumina_read,\n",
    "                \"sequencing_platform\": platform_term_id,\n",
    "            }\n",
    "            sequence_file.append(dcc_row)\n",
    "\n",
    "print(time.monotonic() - t0)\n",
    "sequence_file = pandas.DataFrame(sequence_file)\n",
    "\n",
    "dry_run = True\n",
    "created = server.post_sheet(\"sequence_file\", sequence_file, dry_run=dry_run, verbose=True, validator=validator)\n",
    "if len(created) > 0 and not dry_run:\n",
    "    sequence_file.to_excel(\"{}_sequence_file.xlsx\".format(plate_name))\n",
    "\n",
    "sequence_file\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ahead-federation",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "collaborative-gateway",
   "metadata": {},
   "outputs": [],
   "source": [
    "submission_book_name = \"{}_submission.xlsx\".format(plate_name)\n",
    "print(submission_book_name)\n",
    "with pandas.ExcelWriter(submission_book_name) as book:\n",
    "    rodent_donor.to_excel(book, sheet_name=\"rodent_donor\", index=False)\n",
    "    tissue_sheet.to_excel(book, sheet_name=\"tissue\", index=False)\n",
    "    measurement_set.to_excel(book, sheet_name=\"measurement_set\", index=False)\n",
    "    sequence_file.to_excel(book, sheet_name=\"sequence_file\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "coordinate-enemy",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "conventional-mechanics",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
