{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "visible-leadership",
   "metadata": {},
   "source": [
    "# Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "starting-dress",
   "metadata": {},
   "source": [
    "Using the models we have try to fill in the metadata spreadsheet & submit test records to the IGVF test portal.\n",
    "\n",
    "This try had all the files from each platform type attached to one measurement set. But with Jennifer we decided this is probably a poor idea."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "suffering-component",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "educational-vitamin",
   "metadata": {},
   "outputs": [],
   "source": [
    "import bz2\n",
    "from collections import Counter, namedtuple\n",
    "import datetime\n",
    "import json\n",
    "import numpy\n",
    "import os\n",
    "import pandas\n",
    "from pathlib import Path\n",
    "import re\n",
    "from subprocess import run, PIPE\n",
    "from tqdm import tqdm\n",
    "import sys\n",
    "import zoneinfo\n",
    "\n",
    "os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'mousedemo.settings')\n",
    "os.environ[\"DJANGO_ALLOW_ASYNC_UNSAFE\"] = \"true\"\n",
    "\n",
    "import django\n",
    "from django.contrib.auth import get_user_model\n",
    "from django.db import DEFAULT_DB_ALIAS\n",
    "\n",
    "MOUSEDEMO = str(Path(\"mousedemo\").absolute())\n",
    "if MOUSEDEMO not in sys.path:\n",
    "    sys.path.append(MOUSEDEMO)\n",
    "\n",
    "django.setup()\n",
    "\n",
    "from mousedemo import settings\n",
    "from igvf_mice import models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "signed-percentage",
   "metadata": {},
   "outputs": [],
   "source": [
    "EC = str(Path(\"~/proj/encoded_client\").expanduser())\n",
    "if EC not in sys.path:\n",
    "    sys.path.append(EC)\n",
    "    \n",
    "from encoded_client import encoded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "basic-recipient",
   "metadata": {},
   "outputs": [],
   "source": [
    "#server_name = \"api.sandbox.igvf.org\"\n",
    "# accession_prefix = \"igvftst\"\n",
    "# upload_target = \"sandbox\"\n",
    "server_name = \"api.data.igvf.org\"\n",
    "accession_prefix = \"igvf\"\n",
    "upload_target = \"production\"\n",
    "\n",
    "#server_name = \"igvfd-test-igvf-532-sandbox-db.demo.igvf.org\"\n",
    "assert \"/\" not in server_name\n",
    "assert \"..\" not in server_name\n",
    "\n",
    "server = encoded.ENCODED(server_name)\n",
    "\n",
    "validator = encoded.DCCValidator(server)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "representative-observer",
   "metadata": {},
   "outputs": [],
   "source": [
    "award = \"/awards/HG012077/\"\n",
    "labs = [\"/labs/lior-pachter/\", \"/labs/grant-macgregor/\", \"/labs/barbara-wold/\", \"/labs/ali-mortazavi/\"]\n",
    "lab = labs[-1]\n",
    "jax = \"/sources/jackson-labs/\"\n",
    "species = 'Mus musculus'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "charming-neighborhood",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "liable-singing",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_run_name =  \"IGVF_B01\"\n",
    "plate = models.SplitSeqPlate.objects.get(name=test_run_name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "hidden-adventure",
   "metadata": {},
   "outputs": [],
   "source": [
    "tissues = {}\n",
    "mice = {}\n",
    "for well in plate.splitseqwell_set.all():\n",
    "    for biosample in well.biosample.all():\n",
    "        for tissue in biosample.tissue.all():\n",
    "            tissues[tissue.name] = tissue\n",
    "            mice[tissue.mouse.name] = tissue.mouse"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "permanent-expression",
   "metadata": {},
   "source": [
    "# rodent_donor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "welsh-safety",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>uuid</th>\n",
       "      <th>aliases:array</th>\n",
       "      <th>award</th>\n",
       "      <th>lab</th>\n",
       "      <th>taxa</th>\n",
       "      <th>sex</th>\n",
       "      <th>strain</th>\n",
       "      <th>references</th>\n",
       "      <th>url</th>\n",
       "      <th>...</th>\n",
       "      <th>submitter_comment</th>\n",
       "      <th>description</th>\n",
       "      <th>parents</th>\n",
       "      <th>traits</th>\n",
       "      <th>phenotypic_features</th>\n",
       "      <th>external_resources</th>\n",
       "      <th>strain_background</th>\n",
       "      <th>genotype</th>\n",
       "      <th>individual_rodent:boolean</th>\n",
       "      <th>rodent_identifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:105_CASTJ_10M</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>male</td>\n",
       "      <td>CASTJ</td>\n",
       "      <td>None</td>\n",
       "      <td>https://www.jax.org/strain/000928</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>CAST/EiJ (CAST)</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>105_CASTJ_10M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:106_CASTJ_10F</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>female</td>\n",
       "      <td>CASTJ</td>\n",
       "      <td>None</td>\n",
       "      <td>https://www.jax.org/strain/000928</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>CAST/EiJ (CAST)</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>106_CASTJ_10F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:107_CASTJ_10M</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>male</td>\n",
       "      <td>CASTJ</td>\n",
       "      <td>None</td>\n",
       "      <td>https://www.jax.org/strain/000928</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>CAST/EiJ (CAST)</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>107_CASTJ_10M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:108_CASTJ_10F</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>female</td>\n",
       "      <td>CASTJ</td>\n",
       "      <td>None</td>\n",
       "      <td>https://www.jax.org/strain/000928</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>CAST/EiJ (CAST)</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>108_CASTJ_10F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:117_B6J_10M</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>male</td>\n",
       "      <td>B6J</td>\n",
       "      <td>None</td>\n",
       "      <td>https://www.jax.org/strain/000664</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>C57BL/6J (B6)</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>117_B6J_10M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:118_B6J_10F</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>female</td>\n",
       "      <td>B6J</td>\n",
       "      <td>None</td>\n",
       "      <td>https://www.jax.org/strain/000664</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>C57BL/6J (B6)</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>118_B6J_10F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:119_B6J_10M</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>male</td>\n",
       "      <td>B6J</td>\n",
       "      <td>None</td>\n",
       "      <td>https://www.jax.org/strain/000664</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>C57BL/6J (B6)</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>119_B6J_10M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:120_B6J_10F</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>female</td>\n",
       "      <td>B6J</td>\n",
       "      <td>None</td>\n",
       "      <td>https://www.jax.org/strain/000664</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>C57BL/6J (B6)</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>120_B6J_10F</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession  uuid                aliases:array              award  \\\n",
       "0  would create  None  ali-mortazavi:105_CASTJ_10M  /awards/HG012077/   \n",
       "1  would create  None  ali-mortazavi:106_CASTJ_10F  /awards/HG012077/   \n",
       "2  would create  None  ali-mortazavi:107_CASTJ_10M  /awards/HG012077/   \n",
       "3  would create  None  ali-mortazavi:108_CASTJ_10F  /awards/HG012077/   \n",
       "4  would create  None    ali-mortazavi:117_B6J_10M  /awards/HG012077/   \n",
       "5  would create  None    ali-mortazavi:118_B6J_10F  /awards/HG012077/   \n",
       "6  would create  None    ali-mortazavi:119_B6J_10M  /awards/HG012077/   \n",
       "7  would create  None    ali-mortazavi:120_B6J_10F  /awards/HG012077/   \n",
       "\n",
       "                    lab          taxa     sex strain references  \\\n",
       "0  /labs/ali-mortazavi/  Mus musculus    male  CASTJ       None   \n",
       "1  /labs/ali-mortazavi/  Mus musculus  female  CASTJ       None   \n",
       "2  /labs/ali-mortazavi/  Mus musculus    male  CASTJ       None   \n",
       "3  /labs/ali-mortazavi/  Mus musculus  female  CASTJ       None   \n",
       "4  /labs/ali-mortazavi/  Mus musculus    male    B6J       None   \n",
       "5  /labs/ali-mortazavi/  Mus musculus  female    B6J       None   \n",
       "6  /labs/ali-mortazavi/  Mus musculus    male    B6J       None   \n",
       "7  /labs/ali-mortazavi/  Mus musculus  female    B6J       None   \n",
       "\n",
       "                                 url  ... submitter_comment description  \\\n",
       "0  https://www.jax.org/strain/000928  ...              None        None   \n",
       "1  https://www.jax.org/strain/000928  ...              None        None   \n",
       "2  https://www.jax.org/strain/000928  ...              None        None   \n",
       "3  https://www.jax.org/strain/000928  ...              None        None   \n",
       "4  https://www.jax.org/strain/000664  ...              None        None   \n",
       "5  https://www.jax.org/strain/000664  ...              None        None   \n",
       "6  https://www.jax.org/strain/000664  ...              None        None   \n",
       "7  https://www.jax.org/strain/000664  ...              None        None   \n",
       "\n",
       "  parents traits phenotypic_features external_resources strain_background  \\\n",
       "0    None   None                None               None   CAST/EiJ (CAST)   \n",
       "1    None   None                None               None   CAST/EiJ (CAST)   \n",
       "2    None   None                None               None   CAST/EiJ (CAST)   \n",
       "3    None   None                None               None   CAST/EiJ (CAST)   \n",
       "4    None   None                None               None     C57BL/6J (B6)   \n",
       "5    None   None                None               None     C57BL/6J (B6)   \n",
       "6    None   None                None               None     C57BL/6J (B6)   \n",
       "7    None   None                None               None     C57BL/6J (B6)   \n",
       "\n",
       "  genotype individual_rodent:boolean rodent_identifier  \n",
       "0     None                      True     105_CASTJ_10M  \n",
       "1     None                      True     106_CASTJ_10F  \n",
       "2     None                      True     107_CASTJ_10M  \n",
       "3     None                      True     108_CASTJ_10F  \n",
       "4     None                      True       117_B6J_10M  \n",
       "5     None                      True       118_B6J_10F  \n",
       "6     None                      True       119_B6J_10M  \n",
       "7     None                      True       120_B6J_10F  \n",
       "\n",
       "[8 rows x 25 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def format_sex(value):\n",
    "    if models.SexEnum.MALE == value:\n",
    "        return \"male\"\n",
    "    elif models.SexEnum.FEMALE == value:\n",
    "        return \"female\"\n",
    "\n",
    "def get_accession_string_or_none(record, accession_prefix):\n",
    "    accessions = record.accession.filter(namespace__accession_prefix=accession_prefix)\n",
    "    if len(accessions) == 0:\n",
    "        return None\n",
    "    elif len(accessions) == 1:\n",
    "        return accessions[0]\n",
    "    else:\n",
    "        raise ValueError(\"There should only be one accession per namespace {}\".format(accessions))\n",
    "    \n",
    "rodent_donor = []\n",
    "for mouse_name in sorted(mice):\n",
    "    mouse = mice[mouse_name]\n",
    "    dcc_row = {\n",
    "        #\"#response\": None,\n",
    "        #\"#response_time\": None,\n",
    "        \"accession\": get_accession_string_or_none(mouse, accession_prefix),\n",
    "        \"uuid\": None, \n",
    "        \"aliases:array\": f\"ali-mortazavi:{mouse.name}\",\n",
    "        \"award\": award,\n",
    "        \"lab\": lab,\n",
    "        \"taxa\": species,\n",
    "        \"sex\": format_sex(mouse.sex),\n",
    "        \"strain\": mouse.strain.name,\n",
    "        \"references\": None,\n",
    "        \"url\": mouse.strain.see_also,\n",
    "        \"sources:array\": jax,\n",
    "        \"lot_id\": None,\n",
    "        \"product_id\": mouse.strain.jax_catalog_number,\n",
    "        \"documents\": None,\n",
    "        \"alternate_accessions\": None,\n",
    "        \"submitter_comment\": None,\n",
    "        \"description\": None,\n",
    "        \"parents\": None,\n",
    "        \"traits\": None,\n",
    "        \"phenotypic_features\": None,\n",
    "        \"external_resources\": None,\n",
    "        \"strain_background\": mouse.strain.igvf_id,\n",
    "        \"genotype\": None,\n",
    "        \"individual_rodent:boolean\": True,\n",
    "        \"rodent_identifier\": mouse.name,\n",
    "    }\n",
    "    rodent_donor.append(dcc_row)\n",
    "    \n",
    "rodent_donor = pandas.DataFrame(rodent_donor)\n",
    "\n",
    "dry_run = True\n",
    "created = server.post_sheet(\"rodent_donor\", rodent_donor, dry_run=dry_run, verbose=True, validator=validator)\n",
    "if len(created) > 0 and not dry_run:\n",
    "    rodent_donor.to_excel(\"rodent_donor.{}.xlsx\".format(server_name))\n",
    "rodent_donor"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "modern-diversity",
   "metadata": {},
   "source": [
    "# tissue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "approved-longer",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>uuid</th>\n",
       "      <th>aliases:array</th>\n",
       "      <th>award</th>\n",
       "      <th>lab</th>\n",
       "      <th>sources:array</th>\n",
       "      <th>donors:array</th>\n",
       "      <th>taxa</th>\n",
       "      <th>sample_terms:array</th>\n",
       "      <th>term_names:skip</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:118_B6J_10F_21_NTR_0000646</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:118_B6J_10F</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>/sample-terms/NTR_0000646/</td>\n",
       "      <td>left cerebral cortex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:117_B6J_10M_21_NTR_0000646</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:117_B6J_10M</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>/sample-terms/NTR_0000646/</td>\n",
       "      <td>left cerebral cortex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:106_CASTJ_10F_21_NTR_0000646</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:106_CASTJ_10F</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>/sample-terms/NTR_0000646/</td>\n",
       "      <td>left cerebral cortex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:105_CASTJ_10M_21_NTR_0000646</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:105_CASTJ_10M</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>/sample-terms/NTR_0000646/</td>\n",
       "      <td>left cerebral cortex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:120_B6J_10F_21_NTR_0000646</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:120_B6J_10F</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>/sample-terms/NTR_0000646/</td>\n",
       "      <td>left cerebral cortex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:119_B6J_10M_21_NTR_0000646</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:119_B6J_10M</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>/sample-terms/NTR_0000646/</td>\n",
       "      <td>left cerebral cortex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:108_CASTJ_10F_21_NTR_0000646</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:108_CASTJ_10F</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>/sample-terms/NTR_0000646/</td>\n",
       "      <td>left cerebral cortex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:107_CASTJ_10M_21_NTR_0000646</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:107_CASTJ_10M</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>/sample-terms/NTR_0000646/</td>\n",
       "      <td>left cerebral cortex</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession  uuid                               aliases:array  \\\n",
       "0  would create  None    ali-mortazavi:118_B6J_10F_21_NTR_0000646   \n",
       "1  would create  None    ali-mortazavi:117_B6J_10M_21_NTR_0000646   \n",
       "2  would create  None  ali-mortazavi:106_CASTJ_10F_21_NTR_0000646   \n",
       "3  would create  None  ali-mortazavi:105_CASTJ_10M_21_NTR_0000646   \n",
       "4  would create  None    ali-mortazavi:120_B6J_10F_21_NTR_0000646   \n",
       "5  would create  None    ali-mortazavi:119_B6J_10M_21_NTR_0000646   \n",
       "6  would create  None  ali-mortazavi:108_CASTJ_10F_21_NTR_0000646   \n",
       "7  would create  None  ali-mortazavi:107_CASTJ_10M_21_NTR_0000646   \n",
       "\n",
       "               award                   lab           sources:array  \\\n",
       "0  /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "1  /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "2  /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "3  /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "4  /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "5  /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "6  /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "7  /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "\n",
       "                  donors:array          taxa          sample_terms:array  \\\n",
       "0    ali-mortazavi:118_B6J_10F  Mus musculus  /sample-terms/NTR_0000646/   \n",
       "1    ali-mortazavi:117_B6J_10M  Mus musculus  /sample-terms/NTR_0000646/   \n",
       "2  ali-mortazavi:106_CASTJ_10F  Mus musculus  /sample-terms/NTR_0000646/   \n",
       "3  ali-mortazavi:105_CASTJ_10M  Mus musculus  /sample-terms/NTR_0000646/   \n",
       "4    ali-mortazavi:120_B6J_10F  Mus musculus  /sample-terms/NTR_0000646/   \n",
       "5    ali-mortazavi:119_B6J_10M  Mus musculus  /sample-terms/NTR_0000646/   \n",
       "6  ali-mortazavi:108_CASTJ_10F  Mus musculus  /sample-terms/NTR_0000646/   \n",
       "7  ali-mortazavi:107_CASTJ_10M  Mus musculus  /sample-terms/NTR_0000646/   \n",
       "\n",
       "        term_names:skip  \n",
       "0  left cerebral cortex  \n",
       "1  left cerebral cortex  \n",
       "2  left cerebral cortex  \n",
       "3  left cerebral cortex  \n",
       "4  left cerebral cortex  \n",
       "5  left cerebral cortex  \n",
       "6  left cerebral cortex  \n",
       "7  left cerebral cortex  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_accession_list_or_none(record, accession_prefix, length):\n",
    "    accessions = record.accession.filter(namespace__accession_prefix=accession_prefix)\n",
    "    if len(accessions) == 0:\n",
    "        return [None] * length\n",
    "    elif len(accessions) == length:\n",
    "        return [x.name for x in accessions]\n",
    "    else:\n",
    "        raise ValueError(\"Unexpected number of accessions {} {}\".format(accessions, length))\n",
    "\n",
    "tissue_sheet = []\n",
    "for tissue_name in tissues:\n",
    "    tissue = tissues[tissue_name]\n",
    "    donor_alias = f\"ali-mortazavi:{tissue.mouse.name}\"\n",
    "    ontology_terms = tissue.ontology_term.all()\n",
    "    accessions = get_accession_list_or_none(tissue, accession_prefix, len(ontology_terms))\n",
    "    for term, accession_id in zip(ontology_terms, accessions):\n",
    "        curie = term.curie.replace(\":\", \"_\")\n",
    "        alias = f\"ali-mortazavi:{tissue.name}_{curie}\"\n",
    "        sample_term = \"/sample-terms/{}/\".format(curie)\n",
    "        if numpy.all(pandas.isnull(accession_id)):\n",
    "            accession_id = None\n",
    "        dcc_row = {\n",
    "            \"accession\": accession_id,\n",
    "            \"uuid\": None,\n",
    "            \"aliases:array\": alias,\n",
    "            \"award\": award,\n",
    "            \"lab\": lab,\n",
    "            \"sources:array\": jax,\n",
    "            \"donors:array\": donor_alias,\n",
    "            \"taxa\": species,\n",
    "            \"sample_terms:array\": sample_term,\n",
    "            \"term_names:skip\": term.name,\n",
    "        }\n",
    "        tissue_sheet.append(dcc_row)\n",
    "    \n",
    "tissue_sheet = pandas.DataFrame(tissue_sheet)\n",
    "\n",
    "dry_run=True\n",
    "created = server.post_sheet(\"tissue\", tissue_sheet, dry_run=dry_run, verbose=True, validator=validator)\n",
    "if len(created) > 0 and not dry_run:\n",
    "    tissue_sheet.to_excel(\"tissue_sheet.{}.xlsx\".format(server_name))\n",
    "tissue_sheet"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "finite-hypothetical",
   "metadata": {},
   "source": [
    "# multiplexed_tissue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "painful-rouge",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "polished-handle",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_samples_donors(plate):\n",
    "    samples = {}\n",
    "    donors = {}\n",
    "\n",
    "    for well in plate.splitseqwell_set.all():\n",
    "        for biosample in well.biosample.all():\n",
    "            for tissue in biosample.tissue.all():\n",
    "                mouse_alias = f\"ali-mortazavi:{tissue.mouse.name}\"\n",
    "                # I need an ordered set of donors\n",
    "                donors[mouse_alias] = tissue.mouse\n",
    "                for term in tissue.ontology_term.all():\n",
    "                    curie = term.curie.replace(\":\", \"_\")\n",
    "                    tissue_alias = f\"ali-mortazavi:{tissue.name}_{curie}\"\n",
    "                    samples[tissue_alias] = None\n",
    "\n",
    "    return {\n",
    "        \"samples\": list(samples.keys()),\n",
    "        \"donor\": list(donors.keys()),\n",
    "        \"source\": set([donors[donor_id].source.igvf_id for donor_id in donors],)\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "emerging-sensitivity",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "B01_13A 2\n",
      "B01_13B 2\n",
      "B01_13C 2\n",
      "B01_13D 2\n",
      "B01_13E 2\n",
      "B01_13F 2\n",
      "B01_13G 2\n",
      "B01_13H 2\n",
      "B01_13I 0\n",
      "B01_13J 0\n",
      "B01_13K 0\n",
      "B01_13L 0\n",
      "B01_13M 0\n",
      "B01_13N 0\n",
      "B01_13O 0\n",
      "B01_13P 0\n",
      "B01_13Q 0\n",
      "B01_13R 0\n"
     ]
    }
   ],
   "source": [
    "for subpool in plate.subpool_set.all():\n",
    "    print(subpool.name, subpool.subpoolinrun_set.count())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "contained-clark",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>uuid</th>\n",
       "      <th>aliases:array</th>\n",
       "      <th>cellular_sub_pool</th>\n",
       "      <th>award</th>\n",
       "      <th>lab</th>\n",
       "      <th>sources:array</th>\n",
       "      <th>multiplexed_samples:array</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:subpool_b01_13a</td>\n",
       "      <td>B01_13A</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:118_B6J_10F_21_NTR_0000646,ali-m...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:subpool_b01_13b</td>\n",
       "      <td>B01_13B</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:118_B6J_10F_21_NTR_0000646,ali-m...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:subpool_b01_13c</td>\n",
       "      <td>B01_13C</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:118_B6J_10F_21_NTR_0000646,ali-m...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:subpool_b01_13d</td>\n",
       "      <td>B01_13D</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:118_B6J_10F_21_NTR_0000646,ali-m...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:subpool_b01_13e</td>\n",
       "      <td>B01_13E</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:118_B6J_10F_21_NTR_0000646,ali-m...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:subpool_b01_13f</td>\n",
       "      <td>B01_13F</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:118_B6J_10F_21_NTR_0000646,ali-m...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:subpool_b01_13g</td>\n",
       "      <td>B01_13G</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:118_B6J_10F_21_NTR_0000646,ali-m...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:subpool_b01_13h</td>\n",
       "      <td>B01_13H</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/sources/jackson-labs/</td>\n",
       "      <td>ali-mortazavi:118_B6J_10F_21_NTR_0000646,ali-m...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession  uuid                  aliases:array cellular_sub_pool  \\\n",
       "0  would create  None  ali-mortazavi:subpool_b01_13a           B01_13A   \n",
       "1  would create  None  ali-mortazavi:subpool_b01_13b           B01_13B   \n",
       "2  would create  None  ali-mortazavi:subpool_b01_13c           B01_13C   \n",
       "3  would create  None  ali-mortazavi:subpool_b01_13d           B01_13D   \n",
       "4  would create  None  ali-mortazavi:subpool_b01_13e           B01_13E   \n",
       "5  would create  None  ali-mortazavi:subpool_b01_13f           B01_13F   \n",
       "6  would create  None  ali-mortazavi:subpool_b01_13g           B01_13G   \n",
       "7  would create  None  ali-mortazavi:subpool_b01_13h           B01_13H   \n",
       "\n",
       "               award                   lab           sources:array  \\\n",
       "0  /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "1  /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "2  /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "3  /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "4  /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "5  /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "6  /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "7  /awards/HG012077/  /labs/ali-mortazavi/  /sources/jackson-labs/   \n",
       "\n",
       "                           multiplexed_samples:array  \n",
       "0  ali-mortazavi:118_B6J_10F_21_NTR_0000646,ali-m...  \n",
       "1  ali-mortazavi:118_B6J_10F_21_NTR_0000646,ali-m...  \n",
       "2  ali-mortazavi:118_B6J_10F_21_NTR_0000646,ali-m...  \n",
       "3  ali-mortazavi:118_B6J_10F_21_NTR_0000646,ali-m...  \n",
       "4  ali-mortazavi:118_B6J_10F_21_NTR_0000646,ali-m...  \n",
       "5  ali-mortazavi:118_B6J_10F_21_NTR_0000646,ali-m...  \n",
       "6  ali-mortazavi:118_B6J_10F_21_NTR_0000646,ali-m...  \n",
       "7  ali-mortazavi:118_B6J_10F_21_NTR_0000646,ali-m...  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_subpool_accession(alias, subpool):\n",
    "    if subpool is None:\n",
    "        try:\n",
    "            posted = server.get_json(alias)\n",
    "            return posted[\"accession\"]\n",
    "        except encoded.HTTPError as e:\n",
    "            if e.response.status_code == 404:\n",
    "                pass\n",
    "            else:\n",
    "                print(\"HTTPError\", e)\n",
    "        return None\n",
    "\n",
    "    accessions = subpool.accession.filter(namespace__accession_prefix=accession_prefix)\n",
    "    if len(accessions) > 1:\n",
    "        raise RuntimeError(\"Too many accessions\")\n",
    "    elif len(accessions) == 1:\n",
    "        return accessions[0].name\n",
    "\n",
    "    return None\n",
    "\n",
    "\n",
    "multiplexed_tissue = []\n",
    "\n",
    "plate_details = get_samples_donors(plate)\n",
    "assert len(plate_details[\"source\"]) == 1, \"Not sure what do with mixed sources yet.\"\n",
    "source = next(iter(plate_details[\"source\"]))\n",
    "multiplexed_sample = []\n",
    "for subpool in plate.subpool_set.all():\n",
    "    if subpool.subpoolinrun_set.count() > 0:\n",
    "        alias = \"ali-mortazavi:subpool_{}\".format(subpool.name.lower())\n",
    "        multiplexed_sample.append({\n",
    "            \"accession\": get_subpool_accession(alias, subpool),\n",
    "            \"uuid\": None,\n",
    "            \"aliases:array\": alias,\n",
    "            \"cellular_sub_pool\": subpool.name,\n",
    "            \"award\": award,\n",
    "            \"lab\": lab,\n",
    "            \"sources:array\": source,\n",
    "            \"multiplexed_samples:array\": \",\".join(plate_details[\"samples\"]),\n",
    "        })\n",
    "#print(json.dumps(multiplexed_sample, indent=2))\n",
    "multiplexed_sample = pandas.DataFrame(multiplexed_sample)\n",
    "\n",
    "\n",
    "dry_run=True\n",
    "created = server.post_sheet(\"multiplexed_sample\", multiplexed_sample, dry_run=dry_run, verbose=True, validator=validator)\n",
    "if len(created) > 0 and not dry_run:\n",
    "    multiplexed_sample.to_excel(\"multiplexed_sample.{}.xlsx\".format(server_name), index=False)\n",
    "multiplexed_sample\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "received-conversion",
   "metadata": {},
   "source": [
    "# measurement_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "incorporated-shakespeare",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nextseq2000 igvf_003/nextseq IGVF_003 3\n",
      "nanopore igvf_003/nanopore IGVF_003 1\n",
      "novaseq6000 igvf_003/nova1 IGVF_003 15\n",
      "novaseq6000 igvf_003/nova2 IGVF_003 15\n",
      "nextseq2000 igvf_004/nextseq IGVF_004 3\n",
      "novaseq6000 igvf_004/nova1 IGVF_004 15\n",
      "novaseq6000 igvf_004/nova2 IGVF_004 15\n",
      "nextseq2000 igvf_005/nextseq IGVF_005 3\n",
      "novaseq6000 igvf_005/nova1 IGVF_005 15\n",
      "novaseq6000 igvf_005/nova2 IGVF_005 15\n",
      "novaseq6000 igvf_007/nova1 IGVF_007 15\n",
      "novaseq6000 igvf_007/nova2 IGVF_007 15\n",
      "nextseq2000 igvf_008/nextseq IGVF_008 3\n",
      "novaseq6000 igvf_008/nova1 IGVF_008 9\n",
      "nextseq2000 igvf_b01/next1 IGVF_B01 8\n",
      "nextseq2000 igvf_b01/next2 IGVF_B01 8\n"
     ]
    }
   ],
   "source": [
    "for run in models.SequencingRun.objects.all():\n",
    "    print(run.platform.name, run.name, run.plate, run.subpoolinrun_set.count())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "lyric-child",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "run.subpoolinrun_set.count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "placed-swift",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "measurement_sets = {}\n",
    "for run in models.SequencingRun.objects.all():\n",
    "    platform = run.platform.family\n",
    "    for subpoolinrun in run.subpoolinrun_set.all():\n",
    "        #print(run.name, run.plate.name, run.platform.name, subpoolrun.id, subpoolrun.subpool.name, subpoolrun.subpoolrunfile_set.count())\n",
    "        name = \"{}_{}\".format(subpoolinrun.subpool.name, platform)\n",
    "        alias = f\"ali-mortazavi:{name}\"\n",
    "        measurement_sets.setdefault(alias, set()).add(subpoolinrun)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "tight-wichita",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>uuid</th>\n",
       "      <th>aliases:array</th>\n",
       "      <th>award</th>\n",
       "      <th>lab</th>\n",
       "      <th>assay_term</th>\n",
       "      <th>sequencing_library_type:array</th>\n",
       "      <th>documents</th>\n",
       "      <th>alternate_accessions</th>\n",
       "      <th>description</th>\n",
       "      <th>samples:array</th>\n",
       "      <th>protocol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:B01_13A_illumina</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/assay-terms/OBI_0003109/</td>\n",
       "      <td>mRNA enriched</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:subpool_b01_13a</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:B01_13B_illumina</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/assay-terms/OBI_0003109/</td>\n",
       "      <td>mRNA enriched</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:subpool_b01_13b</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:B01_13C_illumina</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/assay-terms/OBI_0003109/</td>\n",
       "      <td>mRNA enriched</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:subpool_b01_13c</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:B01_13D_illumina</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/assay-terms/OBI_0003109/</td>\n",
       "      <td>mRNA enriched</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:subpool_b01_13d</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:B01_13E_illumina</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/assay-terms/OBI_0003109/</td>\n",
       "      <td>mRNA enriched</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:subpool_b01_13e</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:B01_13F_illumina</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/assay-terms/OBI_0003109/</td>\n",
       "      <td>mRNA enriched</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:subpool_b01_13f</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:B01_13G_illumina</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/assay-terms/OBI_0003109/</td>\n",
       "      <td>exome capture</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:subpool_b01_13g</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>would create</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:B01_13H_illumina</td>\n",
       "      <td>/awards/HG012077/</td>\n",
       "      <td>/labs/ali-mortazavi/</td>\n",
       "      <td>/assay-terms/OBI_0003109/</td>\n",
       "      <td>mRNA enriched</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ali-mortazavi:subpool_b01_13h</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession  uuid                   aliases:array              award  \\\n",
       "0  would create  None  ali-mortazavi:B01_13A_illumina  /awards/HG012077/   \n",
       "1  would create  None  ali-mortazavi:B01_13B_illumina  /awards/HG012077/   \n",
       "2  would create  None  ali-mortazavi:B01_13C_illumina  /awards/HG012077/   \n",
       "3  would create  None  ali-mortazavi:B01_13D_illumina  /awards/HG012077/   \n",
       "4  would create  None  ali-mortazavi:B01_13E_illumina  /awards/HG012077/   \n",
       "5  would create  None  ali-mortazavi:B01_13F_illumina  /awards/HG012077/   \n",
       "6  would create  None  ali-mortazavi:B01_13G_illumina  /awards/HG012077/   \n",
       "7  would create  None  ali-mortazavi:B01_13H_illumina  /awards/HG012077/   \n",
       "\n",
       "                    lab                 assay_term  \\\n",
       "0  /labs/ali-mortazavi/  /assay-terms/OBI_0003109/   \n",
       "1  /labs/ali-mortazavi/  /assay-terms/OBI_0003109/   \n",
       "2  /labs/ali-mortazavi/  /assay-terms/OBI_0003109/   \n",
       "3  /labs/ali-mortazavi/  /assay-terms/OBI_0003109/   \n",
       "4  /labs/ali-mortazavi/  /assay-terms/OBI_0003109/   \n",
       "5  /labs/ali-mortazavi/  /assay-terms/OBI_0003109/   \n",
       "6  /labs/ali-mortazavi/  /assay-terms/OBI_0003109/   \n",
       "7  /labs/ali-mortazavi/  /assay-terms/OBI_0003109/   \n",
       "\n",
       "  sequencing_library_type:array documents alternate_accessions description  \\\n",
       "0                 mRNA enriched      None                 None        None   \n",
       "1                 mRNA enriched      None                 None        None   \n",
       "2                 mRNA enriched      None                 None        None   \n",
       "3                 mRNA enriched      None                 None        None   \n",
       "4                 mRNA enriched      None                 None        None   \n",
       "5                 mRNA enriched      None                 None        None   \n",
       "6                 exome capture      None                 None        None   \n",
       "7                 mRNA enriched      None                 None        None   \n",
       "\n",
       "                   samples:array protocol  \n",
       "0  ali-mortazavi:subpool_b01_13a     None  \n",
       "1  ali-mortazavi:subpool_b01_13b     None  \n",
       "2  ali-mortazavi:subpool_b01_13c     None  \n",
       "3  ali-mortazavi:subpool_b01_13d     None  \n",
       "4  ali-mortazavi:subpool_b01_13e     None  \n",
       "5  ali-mortazavi:subpool_b01_13f     None  \n",
       "6  ali-mortazavi:subpool_b01_13g     None  \n",
       "7  ali-mortazavi:subpool_b01_13h     None  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_measurement_set_accession(alias, measurement_set):\n",
    "    if measurement_set is None:\n",
    "        try:\n",
    "            posted = server.get_json(alias)\n",
    "            return posted[\"accession\"]\n",
    "        except encoded.HTTPError as e:\n",
    "            if e.response.status_code == 404:\n",
    "                pass\n",
    "            else:\n",
    "                print(\"HTTPError\", e)\n",
    "        return None\n",
    "\n",
    "    accessions = measurement_set.accession.filter(namespace__accession_prefix=accession_prefix)\n",
    "    if len(accessions) > 1:\n",
    "        raise RuntimeError(\"Too many accessions\")\n",
    "    elif len(accessions) == 1:\n",
    "        return accessions[0].name\n",
    "\n",
    "    return None\n",
    "\n",
    "igvf_test_ns = models.AccessionNamespace.objects.get(name=\"IGVF test\")\n",
    "measurements = []\n",
    "sequence_data = []\n",
    "#plate_cache = {}\n",
    "\n",
    "sequencing_library_type_map = {\n",
    "    \"NO\": 'mRNA enriched',\n",
    "    \"EX\": \"exome capture\",\n",
    "}\n",
    "\n",
    "for alias in measurement_sets:\n",
    "    run = next(iter(measurement_sets[alias]))\n",
    "    subpool = run.subpool\n",
    "    plate = run.sequencing_run.plate\n",
    "    if plate.name == test_run_name:   \n",
    "        #plate_details = plate_cache.setdefault(plate, get_samples_donors(plate))        \n",
    "        measurement_row = {\n",
    "            \"accession\": get_measurement_set_accession(alias, run.measurement_set),\n",
    "            \"uuid\": None,\n",
    "            \"aliases:array\": alias,\n",
    "            \"award\": award,\n",
    "            \"lab\": lab,\n",
    "            \"assay_term\": \"/assay-terms/OBI_0003109/\", # single-nucleus RNA sequencing assay\n",
    "            \"sequencing_library_type:array\": \",\".join([sequencing_library_type_map[subpool.selection_type]]),\n",
    "            \"documents\": None,\n",
    "            \"alternate_accessions\": None,\n",
    "            #\"submitter_comment\": None,\n",
    "            \"description\": None,\n",
    "            # old attach all the samples to the measurement_set\n",
    "            #\"samples:array\": \",\".join(plate_details[\"samples\"]),\n",
    "            # New attach the multiplexed_sample to the measurement_set\n",
    "            \"samples:array\": \"ali-mortazavi:subpool_{}\".format(subpool.name.lower()),\n",
    "            \"protocol\": None,\n",
    "        }\n",
    "        measurements.append(measurement_row)\n",
    "            \n",
    "measurements = pandas.DataFrame(measurements)\n",
    "if 1:\n",
    "    measurements.to_excel(\"measurement_set.{}.xlsx\".format(server_name), index=False)\n",
    "dry_run=True\n",
    "created = server.post_sheet(\"measurement_set\", measurements, dry_run=dry_run, verbose=True, validator=validator)\n",
    "if len(created) > 0 and not dry_run:\n",
    "    measurements.to_excel(\"measurement_set.{}.xlsx\".format(server_name), index=False)\n",
    "measurements\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "foreign-trash",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sir = next(iter(measurement_sets[alias]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "chronic-closer",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sir.subpool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "specialized-arbitration",
   "metadata": {},
   "outputs": [],
   "source": [
    "def resolve_aliases(entities):\n",
    "    resolved = []\n",
    "    for element in entities:\n",
    "        if \":\" in element:\n",
    "            obj = server.get_json(element)\n",
    "            resolved.append(obj[\"@id\"])\n",
    "        else:\n",
    "            resolved.append(element)\n",
    "    return resolved\n",
    "\n",
    "#resolve_aliases([\"ali-mortazavi:118_B6J_10F_21_NTR_0000646\",\"ali-mortazavi:117_B6J_10M_21_NTR_0000646\",\"ali-mortazavi:106_CASTJ_10F_21_NTR_0000646\"])\n",
    "                 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "eleven-annotation",
   "metadata": {},
   "outputs": [],
   "source": [
    "# if need to fix sample list in measurement set.\n",
    "if 0:\n",
    "    for i, row in measurements.iterrows():\n",
    "        measurement_set = server.get_json(row[\"accession\"])\n",
    "        desired_biosample = resolve_aliases(row[\"samples:array\"].split(\",\"))\n",
    "\n",
    "        payload = {}\n",
    "        different_biosample = set(desired_biosample).symmetric_difference(measurement_set[\"samples\"])\n",
    "        if len(different_biosample) > 0:\n",
    "            payload.update({\"samples\": desired_biosample})\n",
    "           \n",
    "\n",
    "        sequencing_library_type = row[\"sequencing_library_type:array\"].split(\",\")\n",
    "        if measurement_set.get(\"sequencing_library_type\") != sequencing_library_type:\n",
    "            payload.update({\"sequencing_library_type\": sequencing_library_type})\n",
    "            \n",
    "        print(measurement_set[\"@id\"], payload)\n",
    "        #if True and len(payload) > 0:\n",
    "        #    print(server.patch_json(measurement_set[\"@id\"], payload))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "dramatic-worry",
   "metadata": {},
   "outputs": [],
   "source": [
    "measurements.to_excel(\"measurement_set.{}.xlsx\".format(server_name), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "finite-patient",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, row in measurements.iterrows():\n",
    "    if not (pandas.isnull(row[\"accession\"]) or row[\"accession\"] == \"would create\"):\n",
    "        dcc = server.get_json(\"/measurement_set/{}/\".format(row[\"accession\"]))\n",
    "        print(row[\"accession\"], len(dcc[\"samples\"]), len(row[\"samples:array\"].split(\",\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "seeing-reform",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "sequence_data = []\n",
    "plate_cache = {}\n",
    "\n",
    "for alias in measurement_sets:\n",
    "    run_break = None\n",
    "    run_id = 1    \n",
    "    for run in measurement_sets[alias]:\n",
    "        #print(type(run), run.id, run.subpool.name, run.sequencing_run.name, run.measurement_set)\n",
    "        plate = run.sequencing_run.plate\n",
    "        if plate.name == test_run_name:\n",
    "            plate_details = plate_cache.setdefault(plate, get_samples_donors(plate))\n",
    "            \n",
    "            for run_file in run.subpoolinrunfile_set.order_by(\"subpool_run\", \"lane\", \"read\"):\n",
    "                # Need to lookup accession\n",
    "                \n",
    "                current_break = (run_file.subpool_run.sequencing_run.id, run_file.subpool_run.id, run_file.lane)\n",
    "                if run_break is None:\n",
    "                    run_break = current_break\n",
    "                elif run_break != current_break:\n",
    "                    run_id += 1\n",
    "                    run_break = current_break\n",
    "                \n",
    "                accession = run_file.accession.filter(namespace__accession_prefix=accession_prefix)\n",
    "                if len(accession) > 1:\n",
    "                    print(\"Several accessions? {} {}\".format(run_file.filename, \",\".join(accessions)))\n",
    "                    accession = accession[0].name\n",
    "                elif len(accession) == 1:\n",
    "                    accession = accession[0].name\n",
    "                else:\n",
    "                    accession = None\n",
    "                \n",
    "                file_row = {\n",
    "                    \"accession\": accession,\n",
    "                    \"uuid\": None,\n",
    "                    #\"aliases\": None,\n",
    "                    \"award\": award,\n",
    "                    \"lab\": lab,\n",
    "                    \"md5sum\": run_file.md5sum,\n",
    "                    \"file_format\": \"fastq\",\n",
    "                    \"file_set\": alias,\n",
    "                    \"content_type\": \"reads\",\n",
    "                    \"sequencing_run:integer\": run_id,\n",
    "                    \"documents:array\": None,\n",
    "                    \"submitter_comment\": None,\n",
    "                    \"description\": None,\n",
    "                    \"dbxrefs\": None,\n",
    "                    \"derived_from\": None,\n",
    "                    \"seqspec\": None,\n",
    "                    \"submitted_file_name\": run_file.filename,\n",
    "                    \"illumina_read_type\": run_file.read,\n",
    "                }\n",
    "                sequence_data.append(file_row)                \n",
    "        \n",
    "\n",
    "sequence_data = pandas.DataFrame(sequence_data)\n",
    "sequence_data.to_excel(\"sequence_data.{}.xlsx\".format(server_name), index=False)\n",
    "sequence_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "respected-findings",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "saving-romance",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "rising-nepal",
   "metadata": {},
   "outputs": [],
   "source": [
    "with pandas.ExcelWriter(\"{}_{}.xlsx\".format(test_run_name, upload_target)) as book:\n",
    "    rodent_donor.to_excel(book, sheet_name=\"rodent_donor\", index=None)\n",
    "    tissue_sheet.to_excel(book, sheet_name=\"tissue\", index=None)\n",
    "    multiplexed_sample.to_excel(book, sheet_name=\"multiplexed_sample\", index=None)\n",
    "    measurements.to_excel(book, sheet_name=\"measurement_set\", index=None)\n",
    "    sequence_data.to_excel(book, sheet_name=\"sequence_data\", index=None)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "continued-connecticut",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
